// h1z1-server/out/servers/ZoneServer2016/classes/collectingentity.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.CollectingEntity = void 0;
const enums_1 = require("../models/enums");
const loadouts_1 = require("../data/loadouts");
function getSubEntityData(entity, child) {
    switch (entity.itemDefinitionId) {
        case enums_1.Items.BEE_BOX:
            entity.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.bee_box;
            child.workingEffect = enums_1.Effects.EFX_Bees_BeeHiveBox;
            break;
        case enums_1.Items.DEW_COLLECTOR:
            entity.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.dew_collector;
            child.workingEffect = 0;
            break;
        case enums_1.Items.ANIMAL_TRAP:
            entity.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.animal_trap;
            child.workingEffect = 0;
            break;
    }
}
class CollectingEntity {
    /** Id of the parent that the CollectingEntity sits on - used for deck foundations and ground tampers */
    parentObjectCharacterId;
    workingEffect = 0; // bee box
    dictionary;
    requiredTicks = 4; // 20 min to fill
    currentTicks = 0;
    requiredHoneycombTicks = 72; // 6 hours to collect honeycomb
    currentHoneycombTicks = 0;
    wasUsed = false;
    isWorking = true; // some values below arent used, keeping them to avoid compile errors related to 2 different subclasses
    isSmelting = true;
    allowedFuel = [];
    filterId = 0;
    subType = "CollectingEntity"; // for saving identification
    constructor(parentObject, server) {
        this.parentObjectCharacterId = parentObject.characterId;
        if (!parentObject.getParent(server)) {
            this.dictionary = server._worldLootableConstruction;
        }
        else
            this.dictionary = server._lootableConstruction;
    }
    OnInteractionString(server, client) {
        server.sendData(client, "Command.InteractionString", {
            guid: this.parentObjectCharacterId,
            stringId: enums_1.StringIds.OPEN
        });
    }
    OnFullCharacterDataRequest(server, client) {
        const parentObject = server._lootableConstruction[this.parentObjectCharacterId];
        getSubEntityData(parentObject, this);
        if (parentObject.itemDefinitionId != enums_1.Items.BEE_BOX)
            return;
        const container = parentObject.getContainer();
        if (!container)
            return;
        for (const a in container.items) {
            const item = container.items[a];
            if (item.itemDefinitionId == enums_1.Items.HONEY ||
                item.itemDefinitionId == enums_1.Items.WAX) {
                server.sendData(client, "Command.PlayDialogEffect", {
                    characterId: parentObject.characterId,
                    effectId: this.workingEffect
                });
                return;
            }
        }
    }
}
exports.CollectingEntity = CollectingEntity;
