// h1z1-server/out/servers/ZoneServer2016/classes/loadoutItem.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadoutItem = void 0;
const enums_1 = require("../models/enums");
const baseItem_1 = require("./baseItem");
class LoadoutItem extends baseItem_1.BaseItem {
    loadoutItemOwnerGuid;
    constructor(item, loadoutSlotId, loadoutItemOwnerGuid) {
        super(item.itemDefinitionId, item.itemGuid, item.currentDurability, item.stackCount);
        this.slotId = loadoutSlotId;
        this.containerGuid = "0xFFFFFFFFFFFFFFFF";
        this.stackCount = 1;
        this.weapon = item.weapon;
        this.loadoutItemOwnerGuid = loadoutItemOwnerGuid;
    }
    transferLoadoutItem(server, targetCharacterId, newSlotId) {
        const client = server.getClientByCharId(targetCharacterId);
        if (!client)
            return;
        const oldLoadoutItem = client.character._loadout[newSlotId];
        if (!server.validateLoadoutSlot(this.itemDefinitionId, newSlotId, client.character.loadoutId)) {
            server.containerError(client, enums_1.ContainerErrors.INVALID_LOADOUT_SLOT);
            return;
        }
        if (oldLoadoutItem) {
            if (!server.removeLoadoutItem(client.character, oldLoadoutItem.slotId)) {
                server.containerError(client, enums_1.ContainerErrors.NO_ITEM_IN_SLOT);
                return;
            }
        }
        if (!server.removeLoadoutItem(client.character, this.slotId)) {
            server.containerError(client, enums_1.ContainerErrors.NO_ITEM_IN_SLOT);
            return;
        }
        if (oldLoadoutItem) {
            client.character.equipItem(server, oldLoadoutItem, true, this.slotId);
        }
        client.character.equipItem(server, this, true, newSlotId);
    }
}
exports.LoadoutItem = LoadoutItem;
