// h1z1-server/out/servers/ZoneServer2016/classes/loadoutcontainer.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoadoutContainer = void 0;
const basefullcharacter_1 = require("../entities/basefullcharacter");
const enums_1 = require("../models/enums");
const loadoutItem_1 = require("./loadoutItem");
// helper functions
function combineItemStack(server, client, oldStackCount, targetContainer, item, count, targetCharacter) {
    if (oldStackCount == count) {
        // if full stack is moved
        server.addContainerItem(targetCharacter, item, targetContainer, false);
        return;
    }
    // if only partial stack is moved
    server.addContainerItem(targetCharacter, server.generateItem(item.itemDefinitionId, count), targetContainer, false);
}
class LoadoutContainer extends loadoutItem_1.LoadoutItem {
    /** Id of the container attached to the item - See ContainerDefinitions.json for more information */
    containerDefinitionId;
    /** HashMap of items in the container - uses itemGuid (string) for indexing */
    items = {};
    /** Determines if the player can transfer items inside of the container - used for props mainly */
    canAcceptItems = true;
    /** Array of item id's that can bypass a container not allowing items */
    acceptedItems = [];
    /** Returns false if the items state is static */
    isMutable = true;
    constructor(item, containerDefinitionId) {
        super(item, item.slotId, item.loadoutItemOwnerGuid);
        this.containerDefinitionId = containerDefinitionId;
    }
    /**
     * Gets the used bulk of this container.
     * @param server The ZoneServer instance.
     * @returns Returns the amount of bulk used.
     */
    getUsedBulk(server) {
        let bulk = 0;
        for (const item of Object.values(this.items)) {
            const itemDefinition = server.getItemDefinition(item.itemDefinitionId);
            if (!itemDefinition)
                continue;
            bulk += itemDefinition.BULK * item.stackCount;
        }
        return bulk;
    }
    /**
     * Gets the maximum bulk that this container can hold.
     * @param server The ZoneServer instance.
     */
    getMaxBulk(server) {
        return server.getContainerDefinition(this.containerDefinitionId).MAX_BULK;
    }
    /**
     * Gets the available bulk for this container.
     * @param server The ZoneServer instance.
     * @returns Returns the amount of bulk available.
     */
    getAvailableBulk(server) {
        const availableBulk = this.getMaxBulk(server) - this.getUsedBulk(server);
        // prevents returning a negative available bulk for containers with 0 max bulk
        return availableBulk <= 0 ? 0 : availableBulk;
    }
    /**
     * Returns a boolean if this container has enough space for a given amount of a certain item.
     * @param server The ZoneServer instance.
     * @param itemDefinitionId The definiton id of the item to check.
     * @param count The amount of the item to check.
     */
    getHasSpace(server, itemDefinitionId, count) {
        if (this.itemDefinitionId == enums_1.Items.CONTAINER_REPAIR_BOX &&
            ![
                enums_1.Items.WOOD_LOG,
                enums_1.Items.WOOD_PLANK,
                enums_1.Items.NAIL,
                enums_1.Items.METAL_BRACKET,
                enums_1.Items.METAL_SHEET,
                enums_1.Items.SHARD_METAL,
                enums_1.Items.WOOD_STICK
            ].includes(itemDefinitionId))
            return false;
        if (this.getMaxBulk(server) == 0)
            return true; // for external containers
        const itemDefinition = server.getItemDefinition(itemDefinitionId);
        if (!itemDefinition)
            return false;
        return !!(this.getMaxBulk(server) -
            (this.getUsedBulk(server) + itemDefinition.BULK * count) >=
            0);
    }
    /**
     * Gets an item stack in a container that has space for a specified item.
     * @param server The ZoneServer instance.
     * @param itemDefId The item definition ID of the item stack to check.
     * @param count The amount of items to fit into the stack.
     * @param slotId Optional: The slotId of a specific item stack to check.
     * @returns Returns the itemGuid of the item stack.
     */
    getAvailableItemStack(server, itemDefId, count) {
        //
        // if slotId is defined, then only an item with the same slotId will be returned
        if (server.getItemDefinition(itemDefId)?.MAX_STACK_SIZE == 1)
            return "";
        for (const item of Object.values(this.items)) {
            if (item.itemDefinitionId == itemDefId &&
                (server.getItemDefinition(item.itemDefinitionId)?.MAX_STACK_SIZE ??
                    1) >=
                    item.stackCount + count) {
                return item.itemGuid;
            }
        }
        return "";
    }
    /**
     * Gets the maximum slots that this container can hold.
     * @param server The ZoneServer instance.
     */
    getMaxSlots(server) {
        return server.getContainerDefinition(this.containerDefinitionId)
            .MAXIMUM_SLOTS;
    }
    // transfers an item from this container to another
    transferItem(server, targetContainer, item, newSlotId, count) {
        if (!count)
            count = item.stackCount;
        const oldStackCount = item.stackCount, // saves stack count before it gets altered
        sourceCharacter = server.getEntity(this.loadoutItemOwnerGuid), targetCharacter = server.getEntity(targetContainer.loadoutItemOwnerGuid);
        if (!(sourceCharacter instanceof basefullcharacter_1.BaseFullCharacter) ||
            !(targetCharacter instanceof basefullcharacter_1.BaseFullCharacter)) {
            return;
        }
        const client = server.getClientByCharId(sourceCharacter.characterId) ||
            server.getClientByCharId(targetCharacter.characterId);
        if (!client)
            return;
        if (!targetContainer.canAcceptItems) {
            server.containerError(client, enums_1.ContainerErrors.DOES_NOT_ACCEPT_ITEMS);
            return;
        }
        if (targetContainer.acceptedItems.length &&
            !targetContainer.acceptedItems.includes(item.itemDefinitionId)) {
            server.containerError(client, enums_1.ContainerErrors.UNACCEPTED_ITEM);
            return;
        }
        if (!this.isMutable || !targetContainer.isMutable) {
            server.containerError(client, enums_1.ContainerErrors.NOT_MUTABLE);
            return;
        }
        if (this.containerGuid != targetContainer.containerGuid &&
            !targetContainer.getHasSpace(server, item.itemDefinitionId, count)) {
            // allows items in the same container but different stacks to be stacked
            return;
        }
        if (targetContainer.getMaxBulk(server) > 0) {
            const availableSpace = targetContainer.getAvailableBulk(server), itemBulk = server.getItemDefinition(item.itemDefinitionId)?.BULK ?? 1;
            let lootCount = Math.floor(availableSpace / itemBulk);
            if (lootCount) {
                if (lootCount > item.stackCount) {
                    lootCount = item.stackCount;
                }
            }
            else
                return;
            if (count > lootCount) {
                count = lootCount;
            }
        }
        if (!server.removeContainerItem(sourceCharacter, item, this, count)) {
            server.containerError(client, enums_1.ContainerErrors.NO_ITEM_IN_SLOT);
            return;
        }
        if (item.weapon) {
            clearTimeout(item.weapon.reloadTimer);
            delete item.weapon.reloadTimer;
        }
        const itemStack = targetContainer.getAvailableItemStack(server, item.itemDefinitionId, count);
        if (itemStack) {
            // add to existing item stack
            const item = targetContainer.items[itemStack];
            item.stackCount += count;
            server.updateContainerItem(targetCharacter, item, targetContainer);
        }
        else {
            // add item to end
            combineItemStack(server, client, oldStackCount, targetContainer, item, count, targetCharacter);
        }
    }
}
exports.LoadoutContainer = LoadoutContainer;
