// h1z1-server/out/servers/ZoneServer2016/classes/smeltingentity.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmeltingEntity = void 0;
const enums_1 = require("../models/enums");
const loadouts_1 = require("../data/loadouts");
function getAllowedFuel(itemDefinitionId) {
    switch (itemDefinitionId) {
        case enums_1.Items.FURNACE:
            return [
                enums_1.Items.WOOD_LOG,
                enums_1.Items.WEAPON_BRANCH,
                enums_1.Items.WOOD_PLANK,
                enums_1.Items.WOOD_STICK,
                enums_1.Items.CHARCOAL
            ];
        case enums_1.Items.BARBEQUE:
            return [
                enums_1.Items.WEAPON_BRANCH,
                enums_1.Items.WOOD_STICK,
                enums_1.Items.WOOD_PLANK,
                enums_1.Items.CHARCOAL
            ];
        case enums_1.Items.CAMPFIRE:
            return [
                enums_1.Items.WOOD_LOG,
                enums_1.Items.WEAPON_BRANCH,
                enums_1.Items.WOOD_PLANK,
                enums_1.Items.WOOD_STICK,
                enums_1.Items.CHARCOAL
            ];
        default:
            return [
                enums_1.Items.WOOD_LOG,
                enums_1.Items.WEAPON_BRANCH,
                enums_1.Items.WOOD_PLANK,
                enums_1.Items.CHARCOAL
            ];
    }
}
function getSmeltingEntityData(entity, child) {
    switch (entity.itemDefinitionId) {
        case enums_1.Items.FURNACE:
            child.filterId = enums_1.FilterIds.FURNACE;
            entity.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.furnace;
            child.workingEffect = enums_1.Effects.EFX_Fire_Furnace;
            break;
        case enums_1.Items.CAMPFIRE:
            child.filterId = enums_1.FilterIds.COOKING;
            entity.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.campfire;
            child.workingEffect = enums_1.Effects.EFX_Fire_Campfire;
            break;
        case enums_1.Items.BARBEQUE:
            child.filterId = enums_1.FilterIds.COOKING;
            entity.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.barbeque;
            child.workingEffect = enums_1.Effects.EFX_Fire_Barbecue;
            break;
        default:
            child.filterId = enums_1.FilterIds.FURNACE;
            entity.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.furnace;
            child.workingEffect = enums_1.Effects.EFX_Fire_Furnace;
            break;
    }
}
class SmeltingEntity {
    /** The CharacterId of the parent that the entity sits on */
    parentObjectCharacterId;
    /** Array of item id's that can be used as fuel */
    allowedFuel;
    /** Determines if a smelting entity is for cooking or smelting - Default: smelting */
    filterId = enums_1.FilterIds.FURNACE;
    /** Id of the effect the smelting entity will have, Default: Furance EFX
     * - See Effects enum for more information */
    workingEffect = enums_1.Effects.EFX_Fire_Furnace;
    /** Returns true when the smelting entity has been ignited */
    isWorking = false;
    /** Returns true when the item is smelted */
    isSmelting = false;
    /** Global HashMap of SmeltingEntities - uses CharacterId (string) for indexing */
    dictionary;
    /** For identification upon saving */
    subType = "SmeltingEntity";
    constructor(parentObject, server) {
        this.parentObjectCharacterId = parentObject.characterId;
        this.allowedFuel = getAllowedFuel(parentObject.itemDefinitionId);
        getSmeltingEntityData(parentObject, this);
        if (!parentObject.getParent(server)) {
            this.dictionary = server._worldLootableConstruction;
        }
        else
            this.dictionary = server._lootableConstruction;
    }
    OnInteractionString(server, client) {
        server.sendData(client, "Command.InteractionString", {
            guid: this.parentObjectCharacterId,
            stringId: enums_1.StringIds.USE_IGNITABLE
        });
    }
    OnFullCharacterDataRequest(server, client) {
        if (!this.isWorking)
            return;
        server.sendData(client, "Command.PlayDialogEffect", {
            characterId: this.parentObjectCharacterId,
            effectId: this.workingEffect
        });
    }
}
exports.SmeltingEntity = SmeltingEntity;
