// h1z1-server/out/servers/ZoneServer2016/classes/zoneclient.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoneClient2016 = void 0;
const utils_1 = require("../../../utils/utils");
const character_1 = require("../entities/character");
//import { h1z1PacketsType2016 } from "../../../types/packets";
//import { zone2016packets } from "../../../types/zone2016packets";
class ZoneClient2016 {
    guid;
    character;
    currentPOI;
    firstLoading = true;
    isLoading = true;
    characterReleased = false;
    firstCharacterReleased = true;
    isSynced = false;
    isInteracting = false;
    isAdmin = false;
    isDebugMode = false;
    isDecoy = false;
    banType = "";
    HWID = "";
    posAtLastRoutine = new Float32Array();
    posAtTimerStart = new Float32Array();
    oldPos = {
        position: new Float32Array(),
        time: 0
    };
    enableChecks = true;
    speedWarnsNumber = 0;
    pvpStats = { shotsFired: 0, shotsHit: 0, head: 0, spine: 0, legs: 0, hands: 0 };
    clientLogs = [];
    reports = 0;
    lastDeathReport;
    hudTimer = null;
    spawnedDTOs = [];
    spawnedEntities = new Set();
    sentInteractionCounter = 1;
    searchedProps = [];
    managedObjects = [];
    vehicle = {};
    radio = false;
    npcsToSpawnTimer;
    loginSessionId;
    clearHudTimer;
    clearTimers;
    sessionId;
    soeClientId;
    lastKeepAliveTime = 0;
    pings = [];
    avgPing = 0;
    avgPingLen = 4;
    pingWarnings = 0;
    isWeaponLock = false;
    lastMovementImpared = 0;
    avgPingReady = false;
    chunkRenderDistance = 400;
    routineCounter = 0;
    zonePings = [];
    properlyLogout = false;
    permissionLevel = 0;
    fireHints = {};
    isInAir = false;
    startLoc = 0;
    startingPos;
    firstReleased = true;
    /*(lightWeightNpcQueue: {
      packetName: h1z1PacketsType2016;
      data: zone2016packets;
    }[] = [];*/
    isMuted = false;
    blockedPositionUpdates = 0;
    flaggedShots = 0;
    isFairPlayFlagged = false;
    assetIntegrityKickTimer;
    isInVoiceChat = false;
    voiceChatTimer;
    heartBeatTimer;
    afkTimer;
    movementSet = new Set();
    static minMovementForAfk = 20;
    static afkTime = 10 * 60_000;
    gotAfkWarning = false;
    constructor(sessionId, soeClientId, loginSessionId, characterId, transientId, server) {
        this.sessionId = sessionId;
        this.soeClientId = soeClientId;
        this.loginSessionId = loginSessionId;
        this.spawnedEntities = new Set();
        this.managedObjects = [];
        this.clearTimers = () => {
            clearTimeout(this.npcsToSpawnTimer);
        };
        this.clearHudTimer = () => {
            if (this.hudTimer) {
                clearTimeout(this.hudTimer);
            }
            this.hudTimer = null;
            this.isInteracting = false;
        };
        this.character = new character_1.Character2016(characterId, transientId, server);
    }
    afk(server) {
        if (this.movementSet.size < ZoneClient2016.minMovementForAfk) {
            if (this.gotAfkWarning) {
                server.kickPlayer(this);
                return;
            }
            server.sendAlert(this, "You will be kicked soon for AFK");
            this.gotAfkWarning = true;
            setTimeout(() => {
                this.afk(server);
            }, 20_000);
        }
        else {
            this.movementSet.clear();
            if (this.gotAfkWarning) {
                this.gotAfkWarning = false;
                server.sendAlert(this, "You're no longer detected as AFK");
            }
        }
    }
    addPing(ping) {
        if (ping > 0) {
            this.pings.push(ping);
        }
        if (this.pings.length > this.avgPingLen) {
            this.pings.shift();
        }
        if (this.pings.length === this.avgPingLen) {
            this.updateAvgPing();
        }
        else {
            this.avgPingReady = false;
        }
    }
    updateAvgPing() {
        this.avgPing = (0, utils_1.toInt)(utils_1._.sum(this.pings) / this.pings.length);
        this.avgPingReady = true;
    }
}
exports.ZoneClient2016 = ZoneClient2016;
