// h1z1-server/out/servers/ZoneServer2016/managers/abilitiesmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbilitiesManager = void 0;
const enums_1 = require("../models/enums");
const baselightweightcharacter_1 = require("../entities/baselightweightcharacter");
const npc_1 = require("../entities/npc");
const vehicleAbilities = require("../../../../data/2016/dataSources/VehicleAbilities.json");
class AbilitiesManager {
    sendVehicleAbilities(server, client, vehicle) {
        let abilityId;
        switch (vehicle.vehicleId) {
            case enums_1.VehicleIds.OFFROADER:
                abilityId = enums_1.AbilityIds.TURBO_OFFROADER;
                break;
            case enums_1.VehicleIds.PICKUP:
                abilityId = enums_1.AbilityIds.TURBO_PICKP_TRUCK;
                break;
            case enums_1.VehicleIds.POLICECAR:
                abilityId = enums_1.AbilityIds.TURBO_POLICE_CAR;
                break;
            case enums_1.VehicleIds.ATV:
                abilityId = enums_1.AbilityIds.TURBO_ATV;
                break;
        }
        if (abilityId) {
            vehicleAbilities.abilities[2].unknownArray1 = [
                {
                    unknownDword1: abilityId,
                    unknownDword2: abilityId,
                    unknownDword3: 0
                }
            ];
        }
        server.sendData(client, "Abilities.SetVehicleActivatableAbilityManager", vehicleAbilities);
    }
    processAbilityInit(server, client, packetData) {
        const hitLocation = packetData.abilityData?.hitLocation;
        client.character.checkCurrentInteractionGuid();
        const characterId = packetData.abilityData?.hitLocation ??
            client.character.currentInteractionGuid;
        if (hitLocation) {
            client.character.abilityInitTime = Date.now();
            client.character.meleeHit = {
                abilityHitLocation: hitLocation,
                characterId: characterId
            };
            return;
        }
        switch (packetData.abilityId) {
            case enums_1.AbilityIds.NV_GOGGLES:
                const index = client.character.screenEffects.indexOf("NIGHTVISION");
                if (index <= -1) {
                    if (client.character._loadout[29] &&
                        client.character._loadout[29].itemDefinitionId == enums_1.Items.NV_GOGGLES) {
                        client.character.screenEffects.push("NIGHTVISION");
                        server.addScreenEffect(client, server._screenEffects["NIGHTVISION"]);
                    }
                }
                else {
                    client.character.screenEffects.splice(index, 1);
                    server.removeScreenEffect(client, server._screenEffects["NIGHTVISION"]);
                }
                this.deactivateAbility(server, client, enums_1.AbilityIds.NV_GOGGLES, 3, 3);
                return;
        }
        const vehicle = server._vehicles[client.vehicle.mountedVehicle ?? ""], isDriver = vehicle?.getDriver(server) == client.character;
        if (!vehicle || !isDriver) {
            client.character.abilityInitTime = Date.now();
            return;
        }
        server.abilitiesManager.processVehicleAbilityInit(server, client, vehicle, packetData);
    }
    processVehicleAbilityInit(server, client, vehicle, packetData) {
        switch (packetData.abilityId) {
            case enums_1.Abilities.VEHICLE_HEADLIGHTS:
                vehicle.setHeadlightState(server, true, client);
                break;
            case enums_1.Abilities.VEHICLE_SIREN:
                vehicle.setSirenState(server, true, client);
                break;
            case enums_1.Abilities.VEHICLE_TURBO:
                break;
            case enums_1.Abilities.VEHICLE_HORN:
                vehicle.setHornState(server, true, client);
                break;
        }
    }
    processAbilityUninit(server, client, vehicle, packetData) {
        switch (packetData.abilityId) {
            case enums_1.Abilities.VEHICLE_HEADLIGHTS:
                vehicle.setHeadlightState(server, false, client);
                break;
            case enums_1.Abilities.VEHICLE_SIREN:
                vehicle.setSirenState(server, false, client);
                break;
            case enums_1.Abilities.VEHICLE_TURBO:
                break;
            case enums_1.Abilities.VEHICLE_HORN:
                vehicle.setHornState(server, false, client);
                break;
        }
        server.sendData(client, "Abilities.VehicleDeactivateAbility", {
            abilityId: packetData.abilityId,
            unknownDword1: 12
        });
    }
    processAbilityUpdate(server, client, packetData, entity) {
        // todo: validate time between init and update
        if (!client.character.abilityInitTime) {
            // This happen a lot in production but i'm unable to reproduce it
            // Maybe sometimes the hitLocation field in abilityInit is empty
            // console.log(
            //   `${client.character.characterId} tried to update a non-initialized ability!`
            // );
            return;
        }
        client.character.checkCurrentInteractionGuid();
        const weaponItem = client.character.getEquippedWeapon();
        if (!weaponItem)
            return;
        this.handleMeleeHit(server, client, entity, weaponItem);
        client.character.meleeHit = {
            abilityHitLocation: "",
            characterId: ""
        };
    }
    handleMeleeHit(server, client, entity, weaponItem) {
        client.character.abilityInitTime = 0;
        if (!weaponItem.weapon)
            return;
        if (entity instanceof npc_1.Npc &&
            entity.isAlive &&
            server.isHeadshotOnly &&
            client.character.meleeHit.abilityHitLocation != "HEAD")
            return;
        // Zombies should be able to get hit anywhere when they're dead even if the gamerule is headshot only.
        // TODO: CHECK MELEE BLOCK TIME FOR EACH WEAPON
        // CHECK MELEE RANGE ALSO
        //if (client.character.meleeBlocked()) return true;
        client.character.lastMeleeHitTime = Date.now();
        // TODO: calculate this based on melee weapondefinition
        const baseDamage = 1000;
        if (entity instanceof baselightweightcharacter_1.BaseLightweightCharacter && entity.flags.knockedOut)
            return;
        let effectString = "MAT_";
        let meleeType;
        let damage = baseDamage;
        switch (weaponItem.itemDefinitionId) {
            case enums_1.Items.WEAPON_MACHETE01:
            case enums_1.Items.WEAPON_AXE_FIRE:
            case enums_1.Items.WEAPON_AXE_WOOD:
            case enums_1.Items.WEAPON_HATCHET:
            case enums_1.Items.WEAPON_HATCHET_MAKESHIFT:
            case enums_1.Items.WEAPON_KATANA:
            case enums_1.Items.WEAPON_HISREGARD_MACHETE:
                effectString += "Blade_ForehandSlash";
                meleeType = enums_1.MeleeTypes.BLADE;
                damage *= 3;
                break;
            case enums_1.Items.WEAPON_BAT_ALUM:
            case enums_1.Items.WEAPON_BAT_WOOD:
            case enums_1.Items.WEAPON_MOURNING_WOOD:
            case enums_1.Items.WEAPON_BRANCH:
            case enums_1.Items.WEAPON_CROWBAR:
            case enums_1.Items.WEAPON_HAMMER:
            case enums_1.Items.WEAPON_HAMMER_DEMOLITION:
            case enums_1.Items.WEAPON_PIPE:
            case enums_1.Items.WEAPON_WRENCH:
                effectString += "Blunt_ForehandChop";
                meleeType = enums_1.MeleeTypes.BLUNT;
                damage *= 2;
                break;
            case enums_1.Items.WEAPON_COMBATKNIFE:
            case enums_1.Items.SKINNING_KNIFE:
            case enums_1.Items.WEAPON_TOXIC_COMBATKNIFE:
                effectString += "Knife_ForehandSlash";
                meleeType = enums_1.MeleeTypes.KNIFE;
                damage *= 3;
                break;
            case enums_1.Items.WEAPON_GUITAR:
                effectString += "Guitar_ForehandSlash";
                meleeType = enums_1.MeleeTypes.GUITAR;
                damage *= 2;
                break;
            default:
                effectString += "Fists_RightHook";
                meleeType = enums_1.MeleeTypes.FISTS;
                break;
        }
        switch (client.character.meleeHit.abilityHitLocation) {
            case "HEAD":
            case "GLASSES":
            case "NECK":
                damage *= 2;
                break;
        }
        const damageInfo = {
            entity: client.character.characterId,
            weapon: weaponItem.itemDefinitionId,
            damage: damage, // need to figure out a good number for this
            causeBleed: false, // another method for melees to apply bleeding
            meleeType: meleeType,
            hitReport: {
                sessionProjectileCount: 0,
                characterId: client.character.characterId,
                position: client.character.state.position,
                unknownFlag1: 0,
                unknownByte2: 0,
                totalShotCount: 0,
                hitLocation: client.character.meleeHit.abilityHitLocation
            }
        };
        entity.OnMeleeHit(server, damageInfo);
        if (entity.materialType != enums_1.MaterialTypes.ZOMBIE &&
            entity.materialType != enums_1.MaterialTypes.FLESH)
            return;
        const effectId = enums_1.Effects[effectString];
        const dictionary = server.getEntityDictionary(entity.characterId);
        if (!dictionary)
            return;
        if (entity.effectTags.includes(effectId)) {
            server.sendDataToAllWithSpawnedEntity(dictionary, entity.characterId, "Character.RemoveEffectTagCompositeEffect", {
                characterId: entity.characterId,
                newEffectId: 0,
                effectId: effectId
            });
        }
        else {
            entity.effectTags.push(effectId);
        }
        server.sendDataToAllWithSpawnedEntity(dictionary, entity.characterId, "Character.AddEffectTagCompositeEffect", {
            characterId: entity.characterId,
            unknownDword1: effectId,
            effectId: effectId,
            unknownGuid: client.character.characterId,
            unknownDword2: 3
        });
    }
    processAddEffectPacket(server, client, packetData) {
        const abilityEffectId = packetData.effectData.abilityEffectId2 ?? 0, clientEffect = server._clientEffectsData[abilityEffectId];
        if (clientEffect.typeName == "RequestAnimation") {
            const animationName = clientEffect.animationName;
            server.sendDataToAllOthersWithSpawnedEntity(server._characters, client, client.character.characterId, "Character.PlayAnimation", {
                characterId: client.character.characterId,
                animationName: animationName,
                animationType: ""
            });
            return;
        }
        const vehicleAbilityEffectId = packetData.effectData.abilityEffectId1, vehicle = server._vehicles[client.vehicle.mountedVehicle ?? ""];
        if (!vehicle)
            return;
        switch (vehicleAbilityEffectId) {
            case enums_1.VehicleEffects.MOTOR_RUN_OFFROADER_1:
                vehicle.checkEngineRequirements(server);
                break;
            case enums_1.VehicleEffects.TURBO_OFFROADER:
            case enums_1.VehicleEffects.TURBO_PICKUP_TRUCK:
            case enums_1.VehicleEffects.TURBO_POLICE_CAR:
            case enums_1.VehicleEffects.TURBO_ATV:
                vehicle.setTurboState(server, client, true);
                break;
        }
    }
    processRemoveEffectPacket(server, client, packetData) {
        const vehicleAbilityEffectId = packetData.abilityEffectData.abilityEffectId1, vehicle = server._vehicles[packetData.targetCharacterId ?? ""];
        if (!vehicle)
            return;
        switch (vehicleAbilityEffectId) {
            case enums_1.VehicleEffects.MOTOR_RUN_OFFROADER_1:
                vehicle.stopEngine(server);
                break;
            case enums_1.VehicleEffects.TURBO_OFFROADER:
            case enums_1.VehicleEffects.TURBO_PICKUP_TRUCK:
            case enums_1.VehicleEffects.TURBO_POLICE_CAR:
            case enums_1.VehicleEffects.TURBO_ATV:
                this.sendRemoveEffectPacket(server, packetData, server._vehicles);
                vehicle.setTurboState(server, client, false);
                break;
        }
    }
    addEffectTag(server, client, entity, effectId, dictionary) {
        const index = entity.effectTags.indexOf(effectId);
        if (index <= -1) {
            server.sendDataToAllOthersWithSpawnedEntity(dictionary, client, entity.characterId, "Character.AddEffectTagCompositeEffect", {
                characterId: entity.characterId,
                effectId: effectId,
                unknownDword1: effectId,
                unknownDword2: effectId
            });
            entity.effectTags.push(effectId);
        }
    }
    removeEffectTag(server, client, entity, effectId, dictionary) {
        const index = entity.effectTags.indexOf(effectId);
        if (index > -1)
            server.sendDataToAllOthersWithSpawnedEntity(dictionary, client, entity.characterId, "Character.RemoveEffectTagCompositeEffect", {
                characterId: entity.characterId,
                effectId: effectId,
                newEffectId: 0
            });
        entity.effectTags.splice(index, 1);
    }
    sendRemoveEffectPacket(server, packetData, dictionary) {
        server.sendDataToAllWithSpawnedEntity(dictionary, packetData.targetCharacterData.characterId, "Effect.RemoveEffect", {
            abilityEffectData: {
                unknownDword1: 4,
                abilityEffectId1: packetData.abilityEffectData.abilityEffectId1,
                abilityEffectId2: packetData.abilityEffectData.abilityEffectId2
            },
            targetCharacterData: {
                characterId: packetData.targetCharacterId
            },
            guid2: "0x0",
            targetCharacterId: packetData.targetCharacterData.characterId
            //unknownVector1: [0, 0, 0, 0]
        });
    }
    deactivateAbility(server, client, abilityId, unk1, unk2) {
        server.sendData(client, "Abilities.UninitAbility", {
            unknownDword1: unk1 ?? 4,
            abilityId: abilityId,
            unknownDword2: unk2 ?? 3
        });
    }
}
exports.AbilitiesManager = AbilitiesManager;
