// h1z1-server/out/servers/ZoneServer2016/managers/accountinventorymanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountInventoryManager = void 0;
const utils_1 = require("../../../utils/utils");
const node_fs_1 = __importDefault(require("node:fs"));
const accountCrates_1 = require("../data/accountCrates");
class AccountInventoryManager {
    isInSoloMode;
    // need to be initialized before using the manager in !solomode
    mongodbCollection;
    _soloDataPath = `${(0, utils_1.getAppDataFolderPath)()}/single_player_accountitems.json`;
    constructor(zoneServer) {
        this.isInSoloMode = zoneServer._soloMode;
    }
    init(collection) {
        this.mongodbCollection = collection;
    }
    _getSoloAccountItems(loginSessionId) {
        return JSON.parse(node_fs_1.default.readFileSync(this._soloDataPath).toString()).filter((e) => e.loginSessionId === loginSessionId);
    }
    _saveSoloAccountItems(items) {
        node_fs_1.default.writeFileSync(this._soloDataPath, JSON.stringify(items));
    }
    async getAccountItems(loginSessionId) {
        let accountItems, soloUpdate = false;
        if (this.isInSoloMode) {
            accountItems = this._getSoloAccountItems(loginSessionId);
        }
        else {
            accountItems = await this.mongodbCollection
                .find({ loginSessionId })
                .toArray();
        }
        for (const item of accountItems) {
            const newId = accountCrates_1.LEGACY_CRATE_CONVERSION[item.itemDefinitionId];
            if (newId && newId != item.itemDefinitionId) {
                if (!this.isInSoloMode) {
                    await this.mongodbCollection.updateOne({
                        loginSessionId,
                        itemDefinitionId: item.itemDefinitionId
                    }, {
                        $set: {
                            itemDefinitionId: newId
                        }
                    });
                }
                else {
                    soloUpdate = true;
                }
                item.itemDefinitionId = newId;
            }
        }
        if (this.isInSoloMode && soloUpdate) {
            this._saveSoloAccountItems(accountItems);
        }
        return accountItems;
    }
    async getAccountItem(loginSessionId, itemDefinitionId) {
        if (this.isInSoloMode) {
            const accountItems = this._getSoloAccountItems(loginSessionId);
            return (accountItems.find((v) => {
                return v.itemDefinitionId === itemDefinitionId;
            }) ?? null);
        }
        else {
            return await this.mongodbCollection.findOne({
                loginSessionId,
                itemDefinitionId
            });
        }
    }
    async addAccountItem(loginSessionId, item) {
        let accountItems;
        if (this.isInSoloMode) {
            accountItems = this._getSoloAccountItems(loginSessionId);
        }
        else {
            accountItems = await this.getAccountItems(loginSessionId);
        }
        const newCrateItem = accountCrates_1.LEGACY_CRATE_CONVERSION[item.itemDefinitionId];
        if (newCrateItem && newCrateItem != item.itemDefinitionId) {
            item.itemDefinitionId = newCrateItem;
        }
        const index = accountItems.findIndex((v) => {
            return v.itemDefinitionId === item.itemDefinitionId;
        });
        const storedItem = accountItems[index];
        if (this.isInSoloMode) {
            if (storedItem) {
                storedItem.stackCount++;
            }
            else {
                accountItems.push({ loginSessionId, ...item });
            }
            this._saveSoloAccountItems(accountItems);
        }
        else {
            if (storedItem) {
                storedItem.stackCount++;
                return await this.mongodbCollection.updateOne({
                    loginSessionId,
                    itemDefinitionId: item.itemDefinitionId
                }, {
                    $set: {
                        stackCount: storedItem.stackCount
                    }
                });
            }
            else {
                return await this.mongodbCollection.insertOne({
                    loginSessionId,
                    ...item
                });
            }
        }
    }
    async updateAccountItem(loginSessionId, item) {
        if (this.isInSoloMode) {
            const accountItems = this._getSoloAccountItems(loginSessionId);
            for (let i = 0; i < accountItems.length; i++) {
                const originalItem = accountItems[i];
                if (originalItem.itemDefinitionId === item.itemDefinitionId) {
                    accountItems[i] = { loginSessionId, ...item };
                    break;
                }
            }
            this._saveSoloAccountItems(accountItems);
        }
        else {
            return await this.mongodbCollection.updateOne({
                loginSessionId,
                itemDefinitionId: item.itemDefinitionId,
                itemGuid: item.itemGuid
            }, {
                $set: {
                    ...item
                }
            });
        }
    }
    async removeAccountItem(loginSessionId, item) {
        if (this.isInSoloMode) {
            const accountItems = this._getSoloAccountItems(loginSessionId);
            for (let i = 0; i < accountItems.length; i++) {
                const originalItem = accountItems[i];
                if (originalItem.itemDefinitionId === item.itemDefinitionId) {
                    accountItems.splice(i, 1);
                    break;
                }
            }
            this._saveSoloAccountItems(accountItems);
        }
        else {
            return await this.mongodbCollection.deleteOne({
                loginSessionId,
                itemDefinitionId: item.itemDefinitionId
            });
        }
    }
    async rollbackItems(loginSessionId, removedItems) {
        for (const { inventoryItem, count } of removedItems) {
            inventoryItem.stackCount += count;
            await this.updateAccountItem(loginSessionId, inventoryItem);
        }
    }
}
exports.AccountInventoryManager = AccountInventoryManager;
