// h1z1-server/out/servers/ZoneServer2016/managers/aimanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.AiManager = void 0;
const utils_1 = require("../../../utils/utils");
const character_1 = require("../entities/character");
const explosiveentity_1 = require("../entities/explosiveentity");
const trapentity_1 = require("../entities/trapentity");
const degradeTrapsCallTime = 1300_000;
const ttlExplosives = 3600_000 * 3;
class AiManager {
    server;
    trapEntities = new Set();
    playerEntities = new Set();
    explosiveEntities = new Set();
    systemsCallsTime = new Map();
    now = 0;
    constructor(server) {
        this.server = server;
    }
    addEntity(entity) {
        switch (true) {
            case entity instanceof trapentity_1.TrapEntity: {
                this.trapEntities.add(entity);
                break;
            }
            case entity instanceof character_1.Character2016: {
                this.playerEntities.add(entity);
                break;
            }
            case entity instanceof explosiveentity_1.ExplosiveEntity: {
                this.explosiveEntities.add(entity);
                break;
            }
        }
    }
    removeEntity(entity) {
        switch (true) {
            case entity instanceof trapentity_1.TrapEntity: {
                this.trapEntities.delete(entity);
                break;
            }
            case entity instanceof character_1.Character2016: {
                this.playerEntities.delete(entity);
                break;
            }
            case entity instanceof explosiveentity_1.ExplosiveEntity: {
                this.explosiveEntities.delete(entity);
                break;
            }
        }
    }
    getEntitiesTotalNumber() {
        return (this.trapEntities.size +
            this.playerEntities.size +
            this.explosiveEntities.size);
    }
    getEntitiesStats() {
        return `Players: ${this.playerEntities.size}\nTraps: ${this.trapEntities.size}\nExplosive: ${this.explosiveEntities.size}`;
    }
    checkTraps() {
        this.playerEntities.forEach((player) => {
            this.trapEntities.forEach((trap) => {
                if (trap.lastTrigger + trap.cooldown > this.now) {
                    return;
                }
                const inRadius = (0, utils_1.isPosInRadiusWithY)(trap.triggerRadiusX, player.state.position, trap.state.position, trap.triggerRadiusY);
                if (player.isAlive && inRadius) {
                    trap.detonate(player.characterId);
                }
            });
        });
    }
    checkExplosive() {
        this.playerEntities.forEach((player) => {
            this.explosiveEntities.forEach((explosive) => {
                const inRadius = (0, utils_1.isPosInRadiusWithY)(0.6, player.state.position, explosive.state.position, 0.5);
                if (player.isAlive && inRadius) {
                    explosive.detonate(player.characterId);
                }
            });
        });
    }
    degradeTraps() {
        this.trapEntities.forEach((trap) => {
            trap.damage(this.server, {
                damage: (trap.maxHealth * degradeTrapsCallTime) / trap.degradationTime,
                entity: "Server.degradeTraps"
            });
        });
    }
    triggerOldExplosives() {
        this.explosiveEntities.forEach((explosive) => {
            if (explosive.creationTime + ttlExplosives < this.now) {
                explosive.detonate();
            }
        });
    }
    executeScheduled(fn) {
        this.systemsCallsTime.set(fn.name, this.now);
        fn.bind(this)();
    }
    scheduleExecute(fn, time) {
        if (this.systemsCallsTime.has(fn.name)) {
            const lastCall = this.systemsCallsTime.get(fn.name);
            if (lastCall + time < this.now) {
                this.executeScheduled(fn);
            }
        }
        else {
            this.executeScheduled(fn);
        }
    }
    run() {
        this.now = Date.now();
        this.checkTraps();
        this.checkExplosive();
        this.scheduleExecute(this.degradeTraps, degradeTrapsCallTime);
        this.scheduleExecute(this.triggerOldExplosives, 60_000);
    }
}
exports.AiManager = AiManager;
