// h1z1-server/out/servers/ZoneServer2016/managers/challengemanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChallengeManager = exports.ChallengeDifficulty = exports.ChallengeStatus = exports.ChallengeType = void 0;
const node_crypto_1 = require("node:crypto");
var ChallengeType;
(function (ChallengeType) {
    ChallengeType[ChallengeType["NONE"] = 0] = "NONE";
    ChallengeType[ChallengeType["TREE_HATER"] = 1] = "TREE_HATER";
    ChallengeType[ChallengeType["NO_WASTE"] = 2] = "NO_WASTE";
    ChallengeType[ChallengeType["DAWN_ITS_TASTY"] = 3] = "DAWN_ITS_TASTY";
    ChallengeType[ChallengeType["CARDIO_ISSUES"] = 4] = "CARDIO_ISSUES";
    ChallengeType[ChallengeType["BRAIN_DEAD"] = 5] = "BRAIN_DEAD";
    ChallengeType[ChallengeType["RECYCLING"] = 6] = "RECYCLING";
    ChallengeType[ChallengeType["MY_HOME"] = 7] = "MY_HOME";
    ChallengeType[ChallengeType["TIRED_BUDDY"] = 8] = "TIRED_BUDDY";
    ChallengeType[ChallengeType["LIGHTER"] = 9] = "LIGHTER";
    ChallengeType[ChallengeType["MY_LAND"] = 10] = "MY_LAND";
    ChallengeType[ChallengeType["PV_PD_SURVIVAL"] = 11] = "PV_PD_SURVIVAL";
    ChallengeType[ChallengeType["GLOBAL_DISARMAMENT"] = 12] = "GLOBAL_DISARMAMENT";
    ChallengeType[ChallengeType["ROCKY"] = 13] = "ROCKY";
    ChallengeType[ChallengeType["ROCKSTAR"] = 14] = "ROCKSTAR";
    ChallengeType[ChallengeType["IED"] = 15] = "IED";
    ChallengeType[ChallengeType["RANCHITO"] = 16] = "RANCHITO";
    ChallengeType[ChallengeType["SWIZZLE"] = 17] = "SWIZZLE";
})(ChallengeType || (exports.ChallengeType = ChallengeType = {}));
var ChallengeStatus;
(function (ChallengeStatus) {
    ChallengeStatus[ChallengeStatus["CURRENT"] = 1] = "CURRENT";
    ChallengeStatus[ChallengeStatus["DONE"] = 2] = "DONE";
    ChallengeStatus[ChallengeStatus["TIMED_OUT"] = 3] = "TIMED_OUT";
})(ChallengeStatus || (exports.ChallengeStatus = ChallengeStatus = {}));
var ChallengeDifficulty;
(function (ChallengeDifficulty) {
    ChallengeDifficulty[ChallengeDifficulty["EASY"] = 1] = "EASY";
    ChallengeDifficulty[ChallengeDifficulty["MEDIUM"] = 2] = "MEDIUM";
    ChallengeDifficulty[ChallengeDifficulty["HARD"] = 3] = "HARD";
})(ChallengeDifficulty || (exports.ChallengeDifficulty = ChallengeDifficulty = {}));
class ChallengeManager {
    server;
    challenges;
    challengesCollection;
    // managed by config
    challengesPerDay = 3;
    enabled = true;
    constructor(server) {
        this.server = server;
        this.challenges = [
            {
                type: ChallengeType.TREE_HATER,
                difficulty: ChallengeDifficulty.EASY,
                name: "Tree hater",
                description: "Cut 10 trees",
                neededPoints: 10,
                pvpOnly: false
            },
            {
                type: ChallengeType.NO_WASTE,
                difficulty: ChallengeDifficulty.EASY,
                name: "No waste",
                description: "Repair a gun",
                neededPoints: 1,
                pvpOnly: false
            },
            {
                type: ChallengeType.DAWN_ITS_TASTY,
                difficulty: ChallengeDifficulty.EASY,
                name: "Dawn it's tasty",
                description: "Harvest 40 blackberries",
                neededPoints: 40,
                pvpOnly: false
            },
            {
                type: ChallengeType.CARDIO_ISSUES,
                difficulty: ChallengeDifficulty.EASY,
                name: "Cardio deficiency detected",
                description: "Run out of stamina",
                neededPoints: 1,
                pvpOnly: false
            },
            {
                type: ChallengeType.RANCHITO,
                difficulty: ChallengeDifficulty.EASY,
                name: "Wait... Why am i here again?",
                description: "Visit Ranchito",
                neededPoints: 1,
                pvpOnly: false
            },
            {
                type: ChallengeType.BRAIN_DEAD,
                difficulty: ChallengeDifficulty.MEDIUM,
                name: "They're brain-dead anyway",
                description: "Kill 50 zombies",
                neededPoints: 50,
                pvpOnly: false
            },
            {
                type: ChallengeType.RECYCLING,
                difficulty: ChallengeDifficulty.MEDIUM,
                name: "RECYCLING",
                description: "Get 40 scraps from vehicles",
                neededPoints: 40,
                pvpOnly: false
            },
            {
                type: ChallengeType.MY_HOME,
                difficulty: ChallengeDifficulty.MEDIUM,
                name: "This is my home",
                description: "Craft a shack",
                neededPoints: 1,
                pvpOnly: false
            },
            {
                type: ChallengeType.TIRED_BUDDY,
                difficulty: ChallengeDifficulty.MEDIUM,
                name: "Tired buddy",
                description: "Survive until being tired",
                neededPoints: 1,
                pvpOnly: false
            },
            // Too easy to abuse rn
            // {
            //   type: ChallengeType.LIGHTER,
            //   difficulty: ChallengeDifficulty.MEDIUM,
            //   name: "You light up my life",
            //   description: "Find 10 lighters",
            //   neededPoints: 10,
            //   pvpOnly: false
            // },
            {
                type: ChallengeType.MY_LAND,
                difficulty: ChallengeDifficulty.MEDIUM,
                name: "My land!",
                description: "Craft a deck foundation",
                neededPoints: 1,
                pvpOnly: false
            },
            {
                type: ChallengeType.SWIZZLE,
                difficulty: ChallengeDifficulty.MEDIUM,
                name: "Shady buisness",
                description: "Consume some swizzle",
                neededPoints: 1,
                pvpOnly: false
            },
            {
                type: ChallengeType.PV_PD_SURVIVAL,
                difficulty: ChallengeDifficulty.HARD,
                name: "Proving a point",
                description: "Survive 30 minutes in pv pd",
                neededPoints: 30,
                pvpOnly: true
            },
            {
                type: ChallengeType.GLOBAL_DISARMAMENT,
                difficulty: ChallengeDifficulty.HARD,
                name: "GLOBAL DISARMAMENT",
                description: "Fire 60 bullets of any gun",
                neededPoints: 60,
                pvpOnly: false
            },
            {
                type: ChallengeType.ROCKY,
                difficulty: ChallengeDifficulty.HARD,
                name: "Rocky",
                description: "Kill 2 players with your fists",
                neededPoints: 2,
                pvpOnly: true
            },
            {
                type: ChallengeType.ROCKSTAR,
                difficulty: ChallengeDifficulty.HARD,
                name: "Rockstar!!!",
                description: "Kill someone with a guitar",
                neededPoints: 1,
                pvpOnly: true
            },
            {
                type: ChallengeType.IED,
                difficulty: ChallengeDifficulty.HARD,
                name: "Let's blow this joint",
                description: "Craft 15 IEDS",
                neededPoints: 15,
                pvpOnly: false
            }
        ];
    }
    scheduleExpires(time) {
        if (!time) {
            const now = new Date();
            const midnight = new Date(now);
            midnight.setHours(23, 59, 59, 0);
            time = midnight.getTime() - now.getTime();
        }
        console.log(`Challenges expires in ${(time / 60_000).toFixed(0)} minutes`);
        setTimeout(() => {
            this.expireChallenges();
            this.scheduleExpires(24 * 3600 * 1000);
        }, time);
    }
    async expireChallenges() {
        await this.challengesCollection.updateMany({ date: { $lt: new Date() } }, { $set: { status: ChallengeStatus.TIMED_OUT } });
        for (const c in this.server._clients) {
            const client = this.server._clients[c];
            this.affectChallenge(client);
        }
    }
    init(collection) {
        if (this.enabled) {
            this.challengesCollection = collection;
            this.scheduleExpires();
        }
    }
    getChallengeInfo(type) {
        return this.challenges.find((e) => e.type === type);
    }
    async loadChallenges(client) {
        if (this.server._soloMode || !this.enabled) {
            return;
        }
        const currentChallenge = await this.getCurrentChallengeData(client);
        if (!currentChallenge) {
            this.affectChallenge(client);
        }
        else {
            client.character.currentChallenge = currentChallenge.type;
            this.displayChallengeInfos(client);
        }
    }
    async getCurrentChallengeData(client) {
        if (this.server._soloMode) {
            this.server.sendAlert(client, "Challenges aren't available for solomode yet.");
        }
        else {
            const query = {
                status: ChallengeStatus.CURRENT,
                serverId: this.server._worldId,
                playerGuid: client.loginSessionId
            };
            const challengeData = await this.challengesCollection.findOne(query);
            return challengeData ?? undefined;
        }
    }
    async displayChallengeInfos(client) {
        const cInfos = this.getChallengeInfo(client.character.currentChallenge);
        let message;
        if (cInfos) {
            const challengeData = await this.getCurrentChallengeData(client);
            message = `Challenge "${cInfos.name}": ${cInfos.description}`;
            if (cInfos.neededPoints > 1 &&
                cInfos.neededPoints !== challengeData?.points) {
                message += `\n Progression: ${challengeData?.points}/${cInfos.neededPoints}`;
            }
            else if (cInfos.neededPoints === challengeData?.points) {
                message += ` Accomplished!`;
            }
        }
        else {
            message = `No more challenges for today. (${this.challengesPerDay})`;
        }
        this.server.sendAlert(client, message);
    }
    async registerChallengeProgression(client, challengeType, pointsToAdd) {
        if (challengeType !== client.character.currentChallenge) {
            return;
        }
        const currentChallenge = await this.getCurrentChallengeData(client);
        if (!currentChallenge) {
            this.server.sendAlert(client, "no current challenge");
            return;
        }
        const points = currentChallenge.points + pointsToAdd;
        const challengeInfo = this.getChallengeInfo(currentChallenge.type);
        if (!challengeInfo) {
            return;
        }
        if (this.server._soloMode) {
            // nothing
        }
        else {
            await this.challengesCollection.updateOne({ _id: currentChallenge._id }, { $set: { points } });
            await this.displayChallengeInfos(client);
            if (points >= challengeInfo.neededPoints) {
                this.finishChallenge(client);
            }
        }
    }
    async finishChallenge(client) {
        if (client.character.currentChallenge === ChallengeType.NONE) {
            return;
        }
        client.character.currentChallenge = ChallengeType.NONE;
        const query = {
            status: ChallengeStatus.CURRENT,
            serverId: this.server._worldId,
            playerGuid: client.loginSessionId
        };
        await this.challengesCollection.updateOne(query, {
            $set: { status: ChallengeStatus.DONE }
        });
        this.server.rewardManager.dropReward(client);
        setTimeout(() => {
            this.affectChallenge(client);
        }, 5000);
    }
    async affectChallenge(client) {
        const now = Date.now();
        const startOfDay = new Date(now);
        startOfDay.setHours(0, 0, 0, 0);
        const endOfDay = new Date(now);
        endOfDay.setHours(23, 59, 59, 999);
        const challengesToday = await this.challengesCollection
            .find({
            date: { $gte: startOfDay, $lt: endOfDay },
            serverId: this.server._worldId,
            status: { $ne: ChallengeStatus.TIMED_OUT },
            playerGuid: client.loginSessionId
        })
            .toArray();
        if (challengesToday.length >= this.challengesPerDay) {
            client.character.currentChallenge = ChallengeType.NONE;
            this.displayChallengeInfos(client);
            return;
        }
        const challengesTypesDoneToday = challengesToday.map((e) => {
            return e.type;
        });
        const currentProgression = challengesToday.length / this.challengesPerDay;
        let nextDifficultyChallenge;
        if (currentProgression < 0.3) {
            nextDifficultyChallenge = ChallengeDifficulty.EASY;
        }
        else if (currentProgression < 0.6) {
            nextDifficultyChallenge = ChallengeDifficulty.MEDIUM;
        }
        else {
            nextDifficultyChallenge = ChallengeDifficulty.HARD;
        }
        const challengesAvailable = this.challenges.filter((v) => {
            return (!challengesTypesDoneToday.includes(v.type) &&
                (!v.pvpOnly || !this.server.isPvE) &&
                v.difficulty === nextDifficultyChallenge);
        });
        if (challengesAvailable.length) {
            const rnd_index = (0, node_crypto_1.randomInt)(challengesAvailable.length);
            const challenge = challengesAvailable[rnd_index];
            if (challenge) {
                const challengeData = {
                    serverId: this.server._worldId,
                    type: challenge.type,
                    date: new Date(),
                    status: ChallengeStatus.CURRENT,
                    playerGuid: client.loginSessionId,
                    points: 0
                };
                await this.challengesCollection.insertOne(challengeData);
                client.character.currentChallenge = challenge.type;
            }
            else {
                client.character.currentChallenge = ChallengeType.NONE;
            }
        }
        else {
            client.character.currentChallenge = ChallengeType.NONE;
        }
        this.displayChallengeInfos(client);
    }
}
exports.ChallengeManager = ChallengeManager;
