// h1z1-server/out/servers/ZoneServer2016/managers/chatmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatManager = void 0;
const enums_1 = require("../../../utils/enums");
const utils_1 = require("../../../utils/utils");
const blacklist = require("../../../..//data/2016/sampleData/blacklisted_words.json");
class ChatManager {
    sendChatText(server, client, message, clearChat = false) {
        if (clearChat) {
            server.sendData(client, "Chat.ChatText", {
                message: `\n\n\n\n\n\n`,
                unknownDword1: 0,
                color: [255, 255, 255, 0],
                unknownDword2: 13951728,
                unknownBoolean1: false,
                unknownBoolean2: true
            });
        }
        server.sendData(client, "Chat.ChatText", {
            message: message,
            unknownDword1: 0,
            color: [255, 255, 255, 0],
            unknownDword2: 13951728,
            unknownBoolean1: false,
            unknownBoolean2: true
        });
    }
    sendChatTextToAllOthers(server, client, message, clearChat = false) {
        for (const a in server._clients) {
            if (client != server._clients[a]) {
                this.sendChatText(server, server._clients[a], message, clearChat);
            }
        }
    }
    sendChatTextToAdmins(server, message, clearChat = false) {
        for (const a in server._clients) {
            if (server._clients[a].isAdmin) {
                this.sendChatText(server, server._clients[a], message, clearChat);
            }
        }
    }
    sendGlobalChatText(server, message, clearChat = false) {
        for (const a in server._clients) {
            this.sendChatText(server, server._clients[a], message, clearChat);
        }
    }
    sendChatToAllInRange(server, client, message, range) {
        const substitutions = {
            "@": "a",
            "4": "a",
            "3": "e",
            "1": "i",
            "!": "i",
            "0": "o",
            $: "s"
        };
        const sanitizedMessage = message
            .toLowerCase()
            .replace(/[@431!0$]/g, (match) => substitutions[match] || match);
        const detectedWords = blacklist.filter((word) => {
            const regex = new RegExp(`\\b${word.replace(/[@431!0$]/g, (match) => substitutions[match] || match)}\\b`, "i");
            return regex.test(sanitizedMessage);
        });
        if (detectedWords.length > 0) {
            message = "I love you";
        }
        server.sendDataToAllInRange(range, client.character.state.position, "Chat.ChatText", {
            message: `${client.character.name}: ${message}`,
            unknownDword1: 0,
            color: [255, 255, 255, 0],
            unknownDword2: 13951728,
            unknownBoolean1: false,
            unknownBoolean2: true
        });
    }
    muteClient(server, client, reason, adminName, timestamp) {
        const object = {
            name: client.character.name.toLowerCase() || "",
            muteReason: reason ? reason : "no reason",
            loginSessionId: client.loginSessionId,
            adminName: adminName ? adminName : "",
            expirationDate: 0,
            active: true,
            unmuteAdminName: ""
        };
        if (timestamp) {
            object.expirationDate = timestamp;
        }
        server._db?.collection(enums_1.DB_COLLECTIONS.MUTED).insertOne(object);
        if (timestamp) {
            server.sendChatText(client, reason
                ? `You have been muted until: ${(0, utils_1.getDateString)(timestamp)}. Reason: ${reason}`
                : `You have been muted until: ${(0, utils_1.getDateString)(timestamp)}`);
            server.sendChatTextToAllOthers(client, reason
                ? `${client.character.name} has been muted until: ${(0, utils_1.getDateString)(timestamp)}. Reason: ${reason}`
                : `${client.character.name} has been muted until: ${(0, utils_1.getDateString)(timestamp)}`);
        }
        else {
            server.sendChatText(client, reason
                ? `You have been permanently muted. Reason: ${reason}`
                : "You have been permanently muted.");
            server.sendChatTextToAllOthers(client, reason
                ? `${client.character.name} has been muted! Reason: ${reason}`
                : `${client.character.name} has been muted!`);
        }
    }
    async checkMute(server, client) {
        const mutedClient = (await server._db
            ?.collection(enums_1.DB_COLLECTIONS.MUTED)
            .findOne({
            name: client.character.name.toLowerCase(),
            active: true
        }));
        if (mutedClient) {
            if (mutedClient.expirationDate &&
                mutedClient.expirationDate <= Date.now()) {
                await server._db
                    ?.collection(enums_1.DB_COLLECTIONS.MUTED)
                    .findOneAndUpdate({ name: client.character.name.toLowerCase(), active: true }, { $set: { active: false, unmuteAdminName: "SERVER" } });
                server.sendChatText(client, "You have been unmuted!");
                return false;
            }
            return true;
        }
        return false;
    }
}
exports.ChatManager = ChatManager;
