// h1z1-server/out/servers/ZoneServer2016/managers/constructionmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructionManager = void 0;
const Z1_vehicles = require("../../../../data/2016/zoneData/Z1_vehicleLocations.json"), spawnLocations2 = require("../../../../data/2016/zoneData/Z1_gridSpawns.json");
const utils_1 = require("../../../utils/utils");
const baseentity_1 = require("../entities/baseentity");
const constructionchildentity_1 = require("../entities/constructionchildentity");
const constructiondoor_1 = require("../entities/constructiondoor");
const constructionparententity_1 = require("../entities/constructionparententity");
const explosiveentity_1 = require("../entities/explosiveentity");
const lootableconstructionentity_1 = require("../entities/lootableconstructionentity");
const plant_1 = require("../entities/plant");
const plantingdiameter_1 = require("../entities/plantingdiameter");
const temporaryentity_1 = require("../entities/temporaryentity");
const trapentity_1 = require("../entities/trapentity");
const enums_1 = require("../models/enums");
class ConstructionManager {
    overridePlacementItems = [
        enums_1.Items.IED,
        enums_1.Items.LANDMINE,
        enums_1.Items.SEED_CORN,
        enums_1.Items.SEED_WHEAT
    ];
    shackItems = [
        enums_1.Items.SHACK,
        enums_1.Items.SHACK_SMALL,
        enums_1.Items.SHACK_BASIC
    ];
    /** MANAGED BY CONFIGMANAGER - See defaultConfig.yaml for more information */
    allowPOIPlacement;
    allowStackedPlacement;
    allowOutOfBoundsPlacement;
    placementRange;
    spawnPointBlockedPlacementRange;
    vehicleSpawnPointBlockedPlacementRange;
    playerFoundationBlockedPlacementRange;
    playerShackBlockedPlacementRange;
    lowerStrongholdDefenses;
    sendConstructionData(server, client) {
        const unknownArray1 = [46, 45, 47, 48, 49, 50, 12, 7, 15], unknownArray2 = [...unknownArray1, 5, 10, 44, 57, 27, 2, 55, 56];
        server.sendData(client, "Construction.Unknown", {
            unknownArray1: unknownArray1.map((value) => {
                return { unknownDword1: value };
            }),
            /* this array affects certain items placed on direct
            ground ex. punji sticks, furnace, flare, etc
            */
            unknownArray2: unknownArray2.map((value) => {
                return { unknownDword1: value };
            })
        });
    }
    sendPlacementFinalize(server, client, status) {
        server.sendData(client, "Construction.PlacementFinalizeResponse", {
            status: status,
            unknownString1: ""
        });
    }
    placementError(server, client, error) {
        server.sendAlert(client, `Construction Error: ${error}`);
    }
    detectStackedPlacement(server, parentObjectCharacterId, position, itemDefinitionId) {
        // disallow construction stacking
        // world constructions may not be placed within 1 radius, this problem wont affect stuff inside any foundation
        if (this.allowStackedPlacement)
            return false;
        if (!Number(parentObjectCharacterId) &&
            !this.overridePlacementItems.includes(itemDefinitionId)) {
            for (const a in server._worldSimpleConstruction) {
                const c = server._worldSimpleConstruction[a];
                const diff = Math.abs(c.state.position[1] - position[1]);
                if ((0, utils_1.isPosInRadiusWithY)(1, c.state.position, position, 1.5) &&
                    diff > 0.3) {
                    return true;
                }
            }
            for (const a in server._constructionSimple) {
                const c = server._constructionSimple[a];
                const diff = Math.abs(c.state.position[1] - position[1]);
                if ((0, utils_1.isPosInRadiusWithY)(1, c.state.position, position, 1.5) &&
                    diff > 0.3) {
                    return true;
                }
            }
            for (const a in server._lootableConstruction) {
                const c = server._lootableConstruction[a];
                const diff = Math.abs(c.state.position[1] - position[1]);
                if ((0, utils_1.isPosInRadiusWithY)(1, c.state.position, position, 1.5) &&
                    diff > 0.3) {
                    return true;
                }
                if (c.itemDefinitionId == enums_1.Items.STORAGE_BOX &&
                    (itemDefinitionId == enums_1.Items.BEE_BOX ||
                        itemDefinitionId == enums_1.Items.FURNACE)) {
                    if ((0, utils_1.isPosInRadiusWithY)(0.5, c.state.position, position, 1.5) &&
                        diff < 0.5) {
                        return true;
                    }
                }
            }
            for (const a in server._worldLootableConstruction) {
                const c = server._worldLootableConstruction[a];
                const diff = Math.abs(c.state.position[1] - position[1]);
                if ((0, utils_1.isPosInRadiusWithY)(1, c.state.position, position, 1.5) &&
                    diff > 0.3) {
                    return true;
                }
                if (c.itemDefinitionId == enums_1.Items.STORAGE_BOX &&
                    (itemDefinitionId == enums_1.Items.BEE_BOX ||
                        itemDefinitionId == enums_1.Items.FURNACE)) {
                    if ((0, utils_1.isPosInRadiusWithY)(0.5, c.state.position, position, 1.5) &&
                        diff < 0.5) {
                        return true;
                    }
                }
            }
        }
        return false;
    }
    detectStackedTamperPlacement(server, item, position) {
        if ([
            enums_1.Items.GROUND_TAMPER,
            enums_1.Items.FOUNDATION,
            enums_1.Items.HAND_SHOVEL,
            enums_1.Items.FOUNDATION_EXPANSION
        ].includes(item.itemDefinitionId)) {
            // fix for tamper stacking
            let tampersInRadius = 0;
            for (const a in server._constructionFoundations) {
                const foundation = server._constructionFoundations[a];
                // Prevent stacking / hiding hidden stashes under tampers / foundations
                if (item.itemDefinitionId == enums_1.Items.HAND_SHOVEL &&
                    [
                        enums_1.Items.GROUND_TAMPER,
                        enums_1.Items.FOUNDATION,
                        enums_1.Items.FOUNDATION_EXPANSION
                    ].includes(foundation.itemDefinitionId) &&
                    (0, utils_1.getDistance)(foundation.state.position, position) <= 22)
                    return true;
                if ((0, utils_1.isPosInRadius)(22, foundation.state.position, position) &&
                    foundation.itemDefinitionId == enums_1.Items.GROUND_TAMPER)
                    tampersInRadius++;
            }
            if (tampersInRadius >= 3) {
                return true;
            }
            // Prevent stacking / hiding hidden stashes under tampers / foundations
            return (Object.values(server._worldLootableConstruction).filter((lc) => lc.actorModelId == enums_1.ModelIds.HAND_SHOVEL &&
                (0, utils_1.getDistance)(lc.state.position, position) <= 15).length > 0);
        }
        return false;
    }
    detectStackedTreePlacement(server, item, position) {
        const disallowedItems = [enums_1.Items.STORAGE_BOX, enums_1.Items.BEE_BOX, enums_1.Items.FURNACE];
        if (disallowedItems.includes(item.itemDefinitionId)) {
            server.speedtreeManager._speedTreesList.forEach((v) => {
                if ((0, utils_1.isPosInRadius)(1, position, v.position)) {
                    return true;
                }
            });
        }
        return false;
    }
    detectOutOfRange(client, item, position) {
        if (!this.placementRange)
            return false;
        if (item.itemDefinitionId != enums_1.Items.GROUND_TAMPER &&
            item.itemDefinitionId != enums_1.Items.FOUNDATION &&
            item.itemDefinitionId != enums_1.Items.FOUNDATION_EXPANSION &&
            !(0, utils_1.isPosInRadius)(this.placementRange, client.character.state.position, position)) {
            return true;
        }
        return false;
    }
    detectSpawnPointPlacement(itemDefinitionId, client, position, isInsidePermissionedFoundation) {
        if (client.isDebugMode)
            return false;
        if (!this.spawnPointBlockedPlacementRange)
            return false;
        let isInSpawnPoint = false;
        spawnLocations2.forEach((point) => {
            if ((0, utils_1.isPosInRadius)(this.spawnPointBlockedPlacementRange, position, point))
                isInSpawnPoint = true;
        });
        if (isInSpawnPoint &&
            !isInsidePermissionedFoundation &&
            !this.overridePlacementItems.includes(itemDefinitionId)) {
            return true;
        }
        return false;
    }
    detectVehicleSpawnPointPlacement(itemDefinitionId, position, client, isInsidePermissionedFoundation) {
        if (client.isDebugMode)
            return false;
        if (!this.vehicleSpawnPointBlockedPlacementRange)
            return false;
        let isInVehicleSpawnPoint = false;
        Z1_vehicles.forEach((vehicleSpawn) => {
            if ((0, utils_1.isPosInRadius)(this.vehicleSpawnPointBlockedPlacementRange, position, vehicleSpawn.position))
                isInVehicleSpawnPoint = true;
        });
        if (isInVehicleSpawnPoint &&
            !isInsidePermissionedFoundation &&
            !this.overridePlacementItems.includes(itemDefinitionId)) {
            return true;
        }
        return false;
    }
    detectOutOfBoundsPlacement(server, position) {
        if (this.allowOutOfBoundsPlacement)
            return false;
        let inMapBounds = false;
        server._spawnGrid.forEach((cell) => {
            if (position[0] >= cell.position[0] - cell.width / 2 &&
                position[0] <= cell.position[0] + cell.width / 2 &&
                position[2] >= cell.position[2] - cell.height / 2 &&
                position[2] <= cell.position[2] + cell.height / 2) {
                inMapBounds = true;
            }
        });
        if (!inMapBounds) {
            return true;
        }
        return false;
    }
    detectPOIPlacement(itemDefinitionId, position, client, isInsidePermissionedFoundation) {
        if (client.isDebugMode)
            return false;
        if (this.overridePlacementItems.includes(itemDefinitionId))
            return false;
        const isInPoi = (0, utils_1.isPosInPoi)(position);
        // allow placement in poi if object is parented to a foundation
        if (isInPoi && !isInsidePermissionedFoundation) {
            return true;
        }
        return false;
    }
    handleInvalidPlacement(server, client, itemDefinitionId, position, isInsidePermissionedFoundation) {
        if (this.detectSpawnPointPlacement(itemDefinitionId, client, position, isInsidePermissionedFoundation)) {
            this.sendPlacementFinalize(server, client, false);
            server.sendAlert(client, "You may not place this object this close to a spawn point");
            return true;
        }
        if (this.detectVehicleSpawnPointPlacement(itemDefinitionId, position, client, isInsidePermissionedFoundation)) {
            this.sendPlacementFinalize(server, client, false);
            server.sendAlert(client, "You may not place this object this close to a vehicle spawn point");
            return true;
        }
        if (this.detectOutOfBoundsPlacement(server, position)) {
            this.sendPlacementFinalize(server, client, false);
            server.sendAlert(client, "You may not place this object this close to edge of the map");
            return true;
        }
        if (server.isNoBuildInPois &&
            this.detectPOIPlacement(itemDefinitionId, position, client, isInsidePermissionedFoundation)) {
            this.sendPlacementFinalize(server, client, false);
            server.sendAlert(client, "You may not place this object this close to a town or point of interest.");
            return true;
        }
        return false;
    }
    getFreeplaceParentCharacterId(server, position) {
        // for construction entities that don't have a parentObjectCharacterId from the client
        let freeplaceParentCharacterId = "";
        // TODO: SEARCH FOUNDATIONS IN GRID RANGE INSTEAD OF ALL OF THEM
        // TODO: CHECK DECKS BEFORE TAMPERS SO OBJECTS PLACED ON A DECK DON'T GET INCORRECTLY
        // PARENTED TO THE TAMPER A DECK IS ON
        for (const a in server._constructionFoundations) {
            const foundation = server._constructionFoundations[a];
            // check if inside a shelter even if not inside foundation (large shelters can extend it)
            Object.values(foundation.occupiedShelterSlots).forEach((shelter) => {
                // check uppers first so entity is not incorrectly parented to top of a lower (isOn)
                Object.values(shelter.occupiedShelterSlots).forEach((upperShelter) => {
                    if (upperShelter.isInside(position) || upperShelter.isOn(position)) {
                        freeplaceParentCharacterId = upperShelter.characterId;
                    }
                });
                if (!freeplaceParentCharacterId &&
                    (shelter.isInside(position) || shelter.isOn(position))) {
                    freeplaceParentCharacterId = shelter.characterId;
                }
            });
            // for disconnected upper shelters
            if (!Number(freeplaceParentCharacterId)) {
                Object.values(foundation.freeplaceEntities).forEach((freeplace) => {
                    if (freeplace instanceof constructionchildentity_1.ConstructionChildEntity &&
                        (freeplace.isInside(position) || freeplace.isOn(position))) {
                        freeplaceParentCharacterId = freeplace.characterId;
                    }
                });
            }
            // check deck last in case it's parented to a shelter or upper first
            if (!Number(freeplaceParentCharacterId) &&
                foundation.isInside(position)) {
                return foundation.characterId;
            }
            if (Number(freeplaceParentCharacterId))
                return freeplaceParentCharacterId;
        }
        return freeplaceParentCharacterId;
    }
    handleClosePlacement(server, client, foundation, position, itemDefinitionId, isInsidePermissionedFoundation) {
        let allowBuild = false;
        const hasBuildPermission = foundation.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.BUILD);
        if (client.isDebugMode || hasBuildPermission)
            allowBuild = true;
        if ((0, utils_1.isPosInRadius)(foundation.itemDefinitionId === enums_1.Items.FOUNDATION ||
            foundation.itemDefinitionId === enums_1.Items.GROUND_TAMPER
            ? this.playerFoundationBlockedPlacementRange
            : this.playerShackBlockedPlacementRange, position, foundation.state.position) &&
            //!isInFoundation &&
            !allowBuild &&
            !this.overridePlacementItems.includes(itemDefinitionId) &&
            !isInsidePermissionedFoundation) {
            server.sendAlert(client, "You may not place this object this close to another players foundation");
            this.sendPlacementFinalize(server, client, false);
            return true;
        }
        return false;
    }
    getIsOnPermissionedFoundation(server, client, parentObjectCharacterId, freeplaceParentCharacterId) {
        const characterId = !!Number(parentObjectCharacterId)
            ? parentObjectCharacterId
            : freeplaceParentCharacterId, parent = server._constructionFoundations[characterId] ||
            server._constructionSimple[characterId];
        if (!parent)
            return false;
        return parent.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.BUILD);
    }
    placement(server, client, itemDefinitionId, modelId, position, rotation, scale, parentObjectCharacterId, BuildingSlot) {
        const item = client.character.getItemById(itemDefinitionId);
        if (!item) {
            this.sendPlacementFinalize(server, client, false);
            return;
        }
        // invalid placement checks that don't require a parentCharacterId
        if (this.detectStackedPlacement(server, parentObjectCharacterId, position, itemDefinitionId)) {
            this.sendPlacementFinalize(server, client, false);
            this.placementError(server, client, enums_1.ConstructionErrors.STACKED);
            return;
        }
        if (this.detectStackedTamperPlacement(server, item, position)) {
            this.sendPlacementFinalize(server, client, false);
            this.placementError(server, client, item.itemDefinitionId == enums_1.Items.HAND_SHOVEL
                ? enums_1.ConstructionErrors.OVERLAP
                : enums_1.ConstructionErrors.STACKED);
            return;
        }
        // takes too long to process for now
        /*
        if (this.detectStackedTreePlacement(server, item, position)) {
          this.sendPlacementFinalize(server, client, 0);
          this.placementError(server, client, ConstructionErrors.TREESTACKED);
          return;
        }
        */
        if (this.detectOutOfRange(client, item, position)) {
            this.sendPlacementFinalize(server, client, false);
            this.placementError(server, client, enums_1.ConstructionErrors.OUT_OF_RANGE);
            return;
        }
        // for construction entities that don't have a parentObjectCharacterId from the client
        const freeplaceParentCharacterId = this.getFreeplaceParentCharacterId(server, position), isOnPermissionedFoundation = this.getIsOnPermissionedFoundation(server, client, parentObjectCharacterId, freeplaceParentCharacterId);
        if ((!!Number(parentObjectCharacterId) ||
            !!Number(freeplaceParentCharacterId)) &&
            !isOnPermissionedFoundation &&
            !client.isDebugMode &&
            !this.overridePlacementItems.includes(itemDefinitionId)) {
            this.placementError(server, client, enums_1.ConstructionErrors.BUILD_PERMISSION);
            this.sendPlacementFinalize(server, client, false);
            return;
        }
        for (const a in server._constructionFoundations) {
            const foundation = server._constructionFoundations[a];
            if (this.handleClosePlacement(server, client, foundation, position, itemDefinitionId, isOnPermissionedFoundation)) {
                return;
            }
        }
        if (![enums_1.Items.TRAP_FIRE, enums_1.Items.TRAP_FLASH, enums_1.Items.WOODEN_BARRICADE].includes(itemDefinitionId) &&
            this.handleInvalidPlacement(server, client, itemDefinitionId, position, isOnPermissionedFoundation)) {
            return;
        }
        if (!this.handleConstructionPlacement(server, client, itemDefinitionId, modelId, position, rotation, scale, parentObjectCharacterId, BuildingSlot, freeplaceParentCharacterId)) {
            this.sendPlacementFinalize(server, client, false);
            return;
        }
        server.removeInventoryItem(client.character, item);
        this.sendPlacementFinalize(server, client, true);
        this.constructionPermissionsManager(server, client);
    }
    handleConstructionPlacement(server, client, itemDefinitionId, modelId, position, rotation, scale, parentObjectCharacterId, BuildingSlot, freeplaceParentCharacterId) {
        switch (itemDefinitionId) {
            case enums_1.Items.BARBED_WIRE:
            case enums_1.Items.SNARE:
            case enums_1.Items.PUNJI_STICKS:
            case enums_1.Items.PUNJI_STICK_ROW:
            case enums_1.Items.TRAP_FIRE:
            case enums_1.Items.TRAP_FLASH:
                return this.placeTrap(server, itemDefinitionId, modelId, position, (0, utils_1.fixEulerOrder)(rotation), false, client.character.characterId);
            case enums_1.Items.RIGGED_LIGHT:
                return this.placeTemporaryEntity(server, modelId, position, (0, utils_1.fixEulerOrder)(rotation), 86400000);
            case enums_1.Items.FLARE:
                return this.placeTemporaryEntity(server, modelId, position, (0, utils_1.eul2quat)(rotation), 120000);
            case enums_1.Items.CANDLE:
                return this.placeTemporaryEntity(server, modelId, position, (0, utils_1.eul2quat)(rotation), 3600000);
            case enums_1.Items.IED:
            case enums_1.Items.LANDMINE:
                return this.placeExplosiveEntity(server, itemDefinitionId, modelId, position, (0, utils_1.eul2quat)(rotation), client.character.characterId);
            case enums_1.Items.METAL_GATE:
            case enums_1.Items.DOOR_BASIC:
            case enums_1.Items.DOOR_WOOD:
            case enums_1.Items.DOOR_METAL:
                return this.placeConstructionDoor(server, client, itemDefinitionId, modelId, parentObjectCharacterId, BuildingSlot);
            case enums_1.Items.SHACK_SMALL:
                return false;
            case enums_1.Items.GROUND_TAMPER:
            case enums_1.Items.SHACK_BASIC:
            case enums_1.Items.SHACK:
            case enums_1.Items.FOUNDATION:
            case enums_1.Items.FOUNDATION_EXPANSION:
                return this.placeConstructionFoundation(server, client, itemDefinitionId, modelId, position, (0, utils_1.fixEulerOrder)(rotation), parentObjectCharacterId, BuildingSlot);
            case enums_1.Items.FOUNDATION_RAMP:
                return this.placeConstructionRamp(server, client, itemDefinitionId, modelId, parentObjectCharacterId, BuildingSlot);
            case enums_1.Items.FOUNDATION_STAIRS:
                return this.placeConstructionStairs(server, client, itemDefinitionId, modelId, (0, utils_1.fixEulerOrder)(rotation), parentObjectCharacterId, BuildingSlot);
            case enums_1.Items.STORAGE_BOX:
            case enums_1.Items.REPAIR_BOX:
                return this.placeLootableConstruction(server, itemDefinitionId, modelId, position, (0, utils_1.fixEulerOrder)(rotation), freeplaceParentCharacterId);
            case enums_1.Items.FURNACE:
            case enums_1.Items.BARBEQUE:
            case enums_1.Items.CAMPFIRE:
                return this.placeSmeltingEntity(server, itemDefinitionId, modelId, position, (0, utils_1.fixEulerOrder)(rotation), new Float32Array([1, 1, 1, 1]), freeplaceParentCharacterId);
            case enums_1.Items.BEE_BOX:
            case enums_1.Items.DEW_COLLECTOR:
            case enums_1.Items.ANIMAL_TRAP:
                return this.placeCollectingEntity(server, itemDefinitionId, modelId, position, (0, utils_1.fixEulerOrder)(rotation), freeplaceParentCharacterId);
            case enums_1.Items.METAL_WALL:
            case enums_1.Items.METAL_WALL_UPPER:
            case enums_1.Items.METAL_DOORWAY:
                return this.placeConstructionWall(server, client, itemDefinitionId, modelId, parentObjectCharacterId, BuildingSlot);
            case enums_1.Items.SHELTER:
            case enums_1.Items.SHELTER_LARGE:
            case enums_1.Items.SHELTER_UPPER:
            case enums_1.Items.SHELTER_UPPER_LARGE:
            case enums_1.Items.STRUCTURE_STAIRS:
            case enums_1.Items.STRUCTURE_STAIRS_UPPER:
            case enums_1.Items.LOOKOUT_TOWER:
                return this.placeConstructionShelter(server, client, itemDefinitionId, modelId, (0, utils_1.fixEulerOrder)(rotation), parentObjectCharacterId, BuildingSlot);
            case enums_1.Items.GROUND_TILLER:
                return this.placePlantingDiameter(server, modelId, position, (0, utils_1.fixEulerOrder)(rotation), client.character.characterId);
            case enums_1.Items.SEED_WHEAT:
            case enums_1.Items.SEED_CORN:
                return this.placePlantOnDiameter(server, modelId, position, rotation, BuildingSlot, parentObjectCharacterId, itemDefinitionId);
            case enums_1.Items.HAND_SHOVEL:
                return Boolean(this.placeStashEntity(server, itemDefinitionId, modelId, position, (0, utils_1.fixEulerOrder)(rotation), new Float32Array([1, 1, 1, 1]), freeplaceParentCharacterId));
            default:
                //this.placementError(client, ConstructionErrors.UNKNOWN_CONSTRUCTION);
                return this.placeSimpleConstruction(server, modelId, position, rotation, scale, freeplaceParentCharacterId ?? "", itemDefinitionId);
        }
    }
    placeConstructionShelter(server, client, itemDefinitionId, modelId, rotation, parentObjectCharacterId, BuildingSlot) {
        const parent = server._constructionFoundations[parentObjectCharacterId] ||
            server._constructionSimple[parentObjectCharacterId];
        if (!Number(parentObjectCharacterId) || !parent) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_PARENT);
            return false;
        }
        if (parent instanceof constructionparententity_1.ConstructionParentEntity) {
            BuildingSlot = parent.getAdjustedShelterSlotId(BuildingSlot);
        }
        if (parent &&
            parent.isSlotOccupied(parent.occupiedShelterSlots, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
            this.placementError(server, client, enums_1.ConstructionErrors.OVERLAP);
            return false;
        }
        if (parent &&
            this.isSlotOnPlacementCooldown(parent.shelterSlotsPlacementTimer, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
            server.sendAlert(client, `You cant place in this slot for the next ${((parent.shelterSlotsPlacementTimer[(0, utils_1.getConstructionSlotId)(BuildingSlot)] - Date.now()) / 1000).toFixed(2)} seconds`);
            return false;
        }
        if (parent &&
            parent.isSlotOccupied(parent.occupiedShelterSlots, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
            this.placementError(server, client, enums_1.ConstructionErrors.OVERLAP);
            return false;
        }
        if (!parent.isShelterSlotValid(BuildingSlot, itemDefinitionId)) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        const position = parent.getSlotPosition(BuildingSlot, parent.shelterSlots);
        if (!position) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        const characterId = server.generateGuid(), transientId = 1, // dont think its needed
        shelter = new constructionchildentity_1.ConstructionChildEntity(characterId, transientId, modelId, position, rotation, server, itemDefinitionId, parentObjectCharacterId, BuildingSlot);
        const result = parent.setShelterSlot(server, shelter);
        if (!result) {
            this.placementError(server, client, enums_1.ConstructionErrors.WALL_SLOT_FAILED);
            return false;
        }
        server._constructionSimple[characterId] = shelter;
        server.executeFuncForAllReadyClientsInRange((client) => {
            this.spawnSimpleConstruction(server, client, shelter);
        }, shelter);
        return true;
    }
    placeConstructionWall(server, client, itemDefinitionId, modelId, parentObjectCharacterId, BuildingSlot) {
        const parent = server._constructionFoundations[parentObjectCharacterId] ||
            server._constructionSimple[parentObjectCharacterId];
        if (!Number(parentObjectCharacterId) || !parent) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_PARENT);
            return false;
        }
        if (!parent.isWallSlotValid(BuildingSlot, itemDefinitionId)) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        let position, rotation;
        if (itemDefinitionId == enums_1.Items.METAL_WALL_UPPER) {
            if (parent &&
                parent.isSlotOccupied(parent.occupiedUpperWallSlots, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
                this.placementError(server, client, enums_1.ConstructionErrors.OVERLAP);
                return false;
            }
            if (parent &&
                this.isSlotOnPlacementCooldown(parent.upperWallSlotsPlacementTimer, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
                server.sendAlert(client, `You cant place in this slot for the next ${((parent.upperWallSlotsPlacementTimer[(0, utils_1.getConstructionSlotId)(BuildingSlot)] - Date.now()) / 1000).toFixed(2)} seconds`);
                return false;
            }
            position = parent.getSlotPosition(BuildingSlot, parent.upperWallSlots);
            rotation = parent.getSlotRotation(BuildingSlot, parent.upperWallSlots);
        }
        else {
            if (parent &&
                parent.isSlotOccupied(parent.occupiedWallSlots, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
                this.placementError(server, client, enums_1.ConstructionErrors.OVERLAP);
                return false;
            }
            if (parent &&
                this.isSlotOnPlacementCooldown(parent.wallSlotsPlacementTimer, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
                server.sendAlert(client, `You cant place in this slot for the next ${((parent.wallSlotsPlacementTimer[(0, utils_1.getConstructionSlotId)(BuildingSlot)] - Date.now()) / 1000).toFixed(2)} seconds`);
                return false;
            }
            position = parent.getSlotPosition(BuildingSlot, parent.wallSlots);
            rotation = parent.getSlotRotation(BuildingSlot, parent.wallSlots);
        }
        if (!position || !rotation) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        const characterId = server.generateGuid(), transientId = 1, // dont think its needed
        wall = new constructionchildentity_1.ConstructionChildEntity(characterId, transientId, modelId, position, rotation, server, itemDefinitionId, parentObjectCharacterId, BuildingSlot);
        parent.setWallSlot(server, wall);
        server._constructionSimple[characterId] = wall;
        server.executeFuncForAllReadyClientsInRange((client) => {
            this.spawnSimpleConstruction(server, client, wall);
        }, wall);
        return true;
    }
    placeConstructionRamp(server, client, itemDefinitionId, modelId, parentObjectCharacterId, BuildingSlot) {
        const parentFoundation = server._constructionFoundations[parentObjectCharacterId];
        if (!Number(parentObjectCharacterId) || !parentFoundation) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_PARENT);
            return false;
        }
        if (parentFoundation &&
            parentFoundation.isSlotOccupied(parentFoundation.occupiedRampSlots, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
            this.placementError(server, client, enums_1.ConstructionErrors.OVERLAP);
            return false;
        }
        if (!parentFoundation.isRampSlotValid(BuildingSlot, itemDefinitionId)) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        const position = parentFoundation.getSlotPosition(BuildingSlot, parentFoundation.rampSlots), rotation = parentFoundation.getSlotRotation(BuildingSlot, parentFoundation.rampSlots);
        if (!position || !rotation) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        const characterId = server.generateGuid(), transientId = 1, // dont think its needed
        ramp = new constructionchildentity_1.ConstructionChildEntity(characterId, transientId, modelId, position, rotation, server, itemDefinitionId, parentObjectCharacterId, BuildingSlot);
        parentFoundation.setRampSlot(ramp);
        server._constructionSimple[characterId] = ramp;
        server.executeFuncForAllReadyClientsInRange((client) => {
            this.spawnSimpleConstruction(server, client, ramp);
        }, ramp);
        return true;
    }
    placeConstructionStairs(server, client, itemDefinitionId, modelId, rotation, parentObjectCharacterId, BuildingSlot) {
        const parentFoundation = server._constructionFoundations[parentObjectCharacterId];
        if (!Number(parentObjectCharacterId) || !parentFoundation) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_PARENT);
            return false;
        }
        if (parentFoundation &&
            parentFoundation.isSlotOccupied(parentFoundation.occupiedRampSlots, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
            this.placementError(server, client, enums_1.ConstructionErrors.OVERLAP);
            return false;
        }
        if (!parentFoundation.isRampSlotValid(BuildingSlot, itemDefinitionId)) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        const position = parentFoundation.getSlotPosition(BuildingSlot, parentFoundation.rampSlots);
        if (!position) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        // rotation is not slot-locked yet
        const characterId = server.generateGuid(), transientId = 1, // dont think its needed
        stairs = new constructionchildentity_1.ConstructionChildEntity(characterId, transientId, modelId, position, rotation, server, itemDefinitionId, parentObjectCharacterId, BuildingSlot);
        parentFoundation.setRampSlot(stairs);
        server._constructionSimple[characterId] = stairs;
        server.executeFuncForAllReadyClientsInRange((client) => {
            this.spawnSimpleConstruction(server, client, stairs);
        }, stairs);
        return true;
    }
    placeConstructionDoor(server, client, itemDefinitionId, modelId, parentObjectCharacterId, BuildingSlot) {
        const parent = server._constructionFoundations[parentObjectCharacterId] ||
            server._constructionSimple[parentObjectCharacterId];
        if (!Number(parentObjectCharacterId) || !parent) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_PARENT);
            return false;
        }
        if (parent &&
            parent.isSlotOccupied(parent.occupiedWallSlots, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
            this.placementError(server, client, enums_1.ConstructionErrors.OVERLAP);
            return false;
        }
        if (parent &&
            this.isSlotOnPlacementCooldown(parent.wallSlotsPlacementTimer, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
            server.sendAlert(client, `You cant place in this slot for the next ${((parent.wallSlotsPlacementTimer[(0, utils_1.getConstructionSlotId)(BuildingSlot)] - Date.now()) / 1000).toFixed(2)} seconds`);
            return false;
        }
        if (!parent.isWallSlotValid(BuildingSlot, itemDefinitionId)) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        const position = parent.getSlotPosition(BuildingSlot, parent.wallSlots), rotation = parent.getSlotRotation(BuildingSlot, parent.wallSlots);
        if (server.constructionManager.lowerStrongholdDefenses &&
            position &&
            (itemDefinitionId == enums_1.Items.METAL_GATE ||
                itemDefinitionId == enums_1.Items.METAL_WALL ||
                itemDefinitionId == enums_1.Items.METAL_WALL_UPPER))
            position[1] -= 0.1;
        if (!position || !rotation) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
            return false;
        }
        const characterId = server.generateGuid(), transientId = server.getTransientId(characterId), door = new constructiondoor_1.ConstructionDoor(characterId, transientId, modelId, position, new Float32Array([rotation[1], 0, 0, 0]), server, itemDefinitionId, client.character.characterId, parentObjectCharacterId, BuildingSlot);
        const result = parent.setWallSlot(server, door);
        if (!result) {
            this.placementError(server, client, enums_1.ConstructionErrors.WALL_SLOT_FAILED);
            return false;
        }
        server._constructionDoors[characterId] = door;
        server.executeFuncForAllReadyClientsInRange((client) => {
            this.spawnConstructionDoor(server, client, door);
        }, door);
        return true;
    }
    placeConstructionFoundation(server, client, itemDefinitionId, modelId, position, rotation, parentObjectCharacterId, BuildingSlot) {
        if (itemDefinitionId == enums_1.Items.FOUNDATION_EXPANSION &&
            (!parentObjectCharacterId || !BuildingSlot)) {
            // prevent expansions from being placed without a deck if client check is bypassed
            return false;
        }
        if (BuildingSlot &&
            server._constructionFoundations[parentObjectCharacterId]
                ?.occupiedExpansionSlots[(0, utils_1.getConstructionSlotId)(BuildingSlot)]) {
            this.placementError(server, client, enums_1.ConstructionErrors.OVERLAP);
            return false;
        }
        const parentFoundation = server._constructionFoundations[parentObjectCharacterId];
        if (Number(parentObjectCharacterId) && !parentFoundation) {
            this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_PARENT);
            return false;
        }
        if (parentFoundation && BuildingSlot) {
            if (parentFoundation &&
                parentFoundation.isSlotOccupied(parentFoundation.occupiedExpansionSlots, (0, utils_1.getConstructionSlotId)(BuildingSlot))) {
                this.placementError(server, client, enums_1.ConstructionErrors.OVERLAP);
                return false;
            }
            if (!parentFoundation.isExpansionSlotValid(BuildingSlot, itemDefinitionId)) {
                this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
                return false;
            }
            const pos = parentFoundation.getSlotPosition(BuildingSlot || "", parentFoundation.expansionSlots), rot = parentFoundation.getSlotRotation(BuildingSlot || "", parentFoundation.expansionSlots);
            if (!pos || !rot) {
                this.placementError(server, client, enums_1.ConstructionErrors.UNKNOWN_SLOT);
                return false;
            }
            position = pos;
            rotation = rot;
        }
        let ownerCharacterId = client.character.characterId, ownerName = client.character.name;
        if (itemDefinitionId == enums_1.Items.FOUNDATION_EXPANSION) {
            ownerCharacterId = parentFoundation.ownerCharacterId;
            ownerName = "";
        }
        const characterId = server.generateGuid(), transientId = server.getTransientId(characterId), npc = new constructionparententity_1.ConstructionParentEntity(characterId, transientId, modelId, position, rotation, server, itemDefinitionId, ownerCharacterId, ownerName, parentObjectCharacterId, BuildingSlot);
        if (parentFoundation && BuildingSlot) {
            parentFoundation.setExpansionSlot(npc);
            npc.permissions = parentFoundation.permissions;
        }
        server._constructionFoundations[characterId] = npc;
        server.executeFuncForAllReadyClientsInRange((client) => {
            this.spawnConstructionParent(server, client, npc);
        }, npc);
        return true;
    }
    placeTemporaryEntity(server, modelId, position, rotation, time) {
        const characterId = server.generateGuid(), transientId = 1, // dont think its needed
        npc = new temporaryentity_1.TemporaryEntity(characterId, transientId, modelId, position, rotation, server);
        npc.setDespawnTimer(server, time);
        server._temporaryObjects[characterId] = npc;
        server.spawnSimpleNpcForAllInRange(npc);
        return true;
    }
    placeTrap(server, itemDefinitionId, modelId, position, rotation, worldOwned = false, owner = "") {
        const characterId = server.generateGuid(), transientId = 1, // dont think its needed here
        npc = new trapentity_1.TrapEntity(characterId, transientId, modelId, position, rotation, server, itemDefinitionId, worldOwned, owner);
        npc.arm();
        //temporarily disabled
        server._traps[characterId] = npc;
        server.spawnSimpleNpcForAllInRange(npc);
        return true;
    }
    placeExplosiveEntity(server, itemDefinitionId, modelId, position, rotation, ownerCharacterId) {
        const characterId = server.generateGuid(), transientId = 1, // dont think its needed
        npc = new explosiveentity_1.ExplosiveEntity(characterId, transientId, modelId, position, rotation, server, itemDefinitionId, ownerCharacterId);
        if (npc.isLandmine()) {
            npc.arm(server);
        }
        server._explosives[characterId] = npc;
        server.spawnSimpleNpcForAllInRange(npc);
        return true;
    }
    // used by multiple construction classes that don't extend each other
    undoPlacementInteractionString(server, entity, client) {
        server.sendData(client, "Command.InteractionString", {
            guid: entity.characterId,
            stringId: enums_1.StringIds.UNDO_PLACEMENT
        });
    }
    placeLootableConstruction(server, itemDefinitionId, modelId, position, rotation, parentObjectCharacterId) {
        const characterId = server.generateGuid(), transientId = server.getTransientId(characterId);
        const obj = new lootableconstructionentity_1.LootableConstructionEntity(characterId, transientId, modelId, position, rotation, server, new Float32Array([1, 1, 1, 1]), itemDefinitionId, parentObjectCharacterId || "", "");
        const parent = obj.getParent(server);
        if (parent) {
            server._lootableConstruction[characterId] = obj;
            parent.addFreeplaceConstruction(obj);
        }
        else {
            server._worldLootableConstruction[characterId] = obj;
        }
        obj.equipLoadout(server);
        if (itemDefinitionId == enums_1.Items.REPAIR_BOX) {
            const container = obj.getContainer();
            if (container) {
                container.acceptedItems = [];
                server.decayManager.dailyRepairMaterials.forEach((material) => {
                    container.acceptedItems.push(material.itemDefinitionId);
                });
            }
        }
        server.executeFuncForAllReadyClientsInRange((client) => {
            if (this.shouldHideEntity(server, client, obj)) {
                return;
            }
            this.spawnLootableConstruction(server, client, obj);
        }, obj);
        return true;
    }
    placeSmeltingEntity(server, itemDefinitionId, modelId, position, rotation, scale, parentObjectCharacterId, isProp = false) {
        const characterId = server.generateGuid(), transientId = server.getTransientId(characterId);
        const obj = new lootableconstructionentity_1.LootableConstructionEntity(characterId, transientId, modelId, position, rotation, server, scale, itemDefinitionId, parentObjectCharacterId || "", "SmeltingEntity", isProp);
        const parent = obj.getParent(server);
        if (parent) {
            server._lootableConstruction[characterId] = obj;
            parent.addFreeplaceConstruction(obj);
        }
        else {
            server._worldLootableConstruction[characterId] = obj;
        }
        obj.equipLoadout(server);
        server.executeFuncForAllReadyClientsInRange((client) => {
            if (this.shouldHideEntity(server, client, obj)) {
                return;
            }
            this.spawnLootableConstruction(server, client, obj);
        }, obj);
        return true;
    }
    placeCollectingEntity(server, itemDefinitionId, modelId, position, rotation, parentObjectCharacterId) {
        const characterId = server.generateGuid(), transientId = server.getTransientId(characterId);
        const obj = new lootableconstructionentity_1.LootableConstructionEntity(characterId, transientId, modelId, position, rotation, server, new Float32Array([1, 1, 1, 1]), itemDefinitionId, parentObjectCharacterId || "", "CollectingEntity");
        const parent = obj.getParent(server);
        if (parent) {
            server._lootableConstruction[characterId] = obj;
            parent.addFreeplaceConstruction(obj);
        }
        else {
            server._worldLootableConstruction[characterId] = obj;
        }
        obj.equipLoadout(server);
        const container = obj.getContainer();
        if (container) {
            switch (obj.itemDefinitionId) {
                case enums_1.Items.ANIMAL_TRAP:
                    container.canAcceptItems = false;
                    break;
                case enums_1.Items.DEW_COLLECTOR:
                    container.acceptedItems = [enums_1.Items.WATER_EMPTY];
                case enums_1.Items.BEE_BOX:
                    container.acceptedItems = [enums_1.Items.WATER_EMPTY, enums_1.Items.HONEYCOMB];
            }
        }
        server.executeFuncForAllReadyClientsInRange((client) => {
            if (this.shouldHideEntity(server, client, obj)) {
                return;
            }
            this.spawnLootableConstruction(server, client, obj);
        }, obj);
        server.smeltingManager._collectingEntities[characterId] = characterId;
        return true;
    }
    placePlantingDiameter(server, modelId, position, rotation, ownerCharacterId) {
        const characterId = server.generateGuid(), transientId = 1;
        const obj = new plantingdiameter_1.PlantingDiameter(characterId, transientId, modelId, position, rotation, server, ownerCharacterId);
        server._temporaryObjects[characterId] = obj;
        server.spawnSimpleNpcForAllInRange(obj);
        return true;
    }
    placePlantOnDiameter(server, modelId, position, rotation, slot, parentObjectCharacterId, itemDefinitionId) {
        const item = server.generateItem(itemDefinitionId);
        if (!item)
            return false;
        const characterId = server.generateGuid(), transientId = server.getTransientId(characterId);
        if (!server._temporaryObjects[parentObjectCharacterId])
            return false;
        const parent = server._temporaryObjects[parentObjectCharacterId];
        if (parent.seedSlots[slot]) {
            return false;
        }
        const obj = new plant_1.Plant(characterId, transientId, modelId, position, (0, utils_1.eul2quat)(rotation), server, 0, item, parentObjectCharacterId, slot);
        parent.seedSlots[slot] = obj;
        server._plants[characterId] = obj;
        server.spawnSimpleNpcForAllInRange(obj);
        return true;
    }
    placeSimpleConstruction(server, modelId, position, rotation, scale, parentObjectCharacterId, itemDefinitionId) {
        const characterId = server.generateGuid(), transientId = 1, // not needed hopefully probably maybe it'll break one day we'll see
        construction = new constructionchildentity_1.ConstructionChildEntity(characterId, transientId, modelId, position, (0, utils_1.fixEulerOrder)(rotation), server, itemDefinitionId, parentObjectCharacterId, "");
        construction.scale = scale;
        const parent = construction.getParent(server);
        if (parent) {
            server._constructionSimple[characterId] = construction;
            parent.addFreeplaceConstruction(construction);
        }
        else {
            server._worldSimpleConstruction[characterId] = construction;
        }
        server.executeFuncForAllReadyClientsInRange((client) => {
            if (this.shouldHideEntity(server, client, construction)) {
                return;
            }
            this.spawnSimpleConstruction(server, client, construction);
        }, construction);
        return true;
    }
    placeStashEntity(server, itemDefinitionId, modelId, position, rotation, scale, parentObjectCharacterId) {
        const characterId = server.generateGuid(), transientId = server.getTransientId(characterId);
        const obj = new lootableconstructionentity_1.LootableConstructionEntity(characterId, transientId, modelId, position, rotation, server, scale, itemDefinitionId, parentObjectCharacterId || "", "");
        server._worldLootableConstruction[characterId] = obj;
        obj.equipLoadout(server);
        server.executeFuncForAllReadyClientsInRange((client) => {
            if (this.shouldHideEntity(server, client, obj)) {
                return;
            }
            this.spawnLootableConstruction(server, client, obj);
        }, obj);
        return obj;
    }
    checkFoundationPermission(server, client, foundation) {
        // under foundation check
        const hasVisitPermission = foundation.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.VISIT);
        if (foundation.itemDefinitionId == enums_1.Items.FOUNDATION ||
            foundation.itemDefinitionId == enums_1.Items.FOUNDATION_EXPANSION) {
            if (foundation.isUnder(client.character.state.position))
                if (hasVisitPermission) {
                    this.tpPlayerOutsideFoundation(server, client, foundation, true);
                }
                else {
                    this.tpPlayerOutsideFoundation(server, client, foundation, false);
                }
        }
        if (!foundation.isSecured)
            return false;
        let allowed = false;
        if (hasVisitPermission)
            allowed = true;
        if (foundation.itemDefinitionId == enums_1.Items.SHACK ||
            foundation.itemDefinitionId == enums_1.Items.SHACK_SMALL ||
            foundation.itemDefinitionId == enums_1.Items.SHACK_BASIC) {
            if (foundation.isInside(client.character.state.position)) {
                if (allowed) {
                    this.constructionHidePlayer(server, client, foundation.characterId, true);
                    return true;
                }
                else if (!client.isAdmin || !client.isDebugMode) {
                    this.tpPlayerOutsideFoundation(server, client, foundation);
                }
            }
        }
        if (allowed)
            return false;
        const bufferZone = 0.15;
        const position = client.character.state.position;
        const positions = [];
        // Create the Float32Arrays
        positions.push(new Float32Array([position[0] + bufferZone, position[1], position[2], 1])); // Move X by +5
        positions.push(new Float32Array([position[0] - bufferZone, position[1], position[2], 1])); // Move X by -5
        positions.push(new Float32Array([position[0], position[1], position[2] + bufferZone, 1])); // Move Y by +5
        positions.push(new Float32Array([position[0], position[1], position[2] - bufferZone, 1])); // Move Y by -5
        for (let x = 0; x < positions.length; x++) {
            if (foundation.isInside(positions[x]) &&
                (!client.isAdmin || !client.isDebugMode)) {
                this.tpPlayerOutsideFoundation(server, client, foundation);
                return false;
            }
        }
        return false;
    }
    checkConstructionChildEntityPermission(server, client, construction) {
        const allowedIds = [
            enums_1.Items.SHELTER,
            enums_1.Items.SHELTER_LARGE,
            enums_1.Items.SHELTER_UPPER,
            enums_1.Items.SHELTER_UPPER_LARGE
        ];
        if (!allowedIds.includes(construction.itemDefinitionId))
            return false;
        let allowed = false;
        if (!construction.isSecured)
            return false;
        let foundation;
        if (server._constructionFoundations[construction.parentObjectCharacterId]) {
            foundation =
                server._constructionFoundations[construction.parentObjectCharacterId];
        }
        else if (server._constructionSimple[construction.parentObjectCharacterId] &&
            server._constructionFoundations[server._constructionSimple[construction.parentObjectCharacterId]
                .parentObjectCharacterId]) {
            foundation =
                server._constructionFoundations[server._constructionSimple[construction.parentObjectCharacterId]
                    .parentObjectCharacterId];
        }
        else {
            for (const a in server._constructionFoundations) {
                const b = server._constructionFoundations[a];
                if (!b.isInside(construction.state.position))
                    continue;
                foundation = b;
            }
        }
        if (!foundation)
            return false;
        const hasVisitPermission = foundation.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.VISIT);
        if (hasVisitPermission)
            allowed = true;
        for (const f in construction.freeplaceEntities) {
            const freePlacedEntity = construction.freeplaceEntities[f];
            if (construction.isInside(freePlacedEntity.state.position)) {
                if (freePlacedEntity instanceof baseentity_1.BaseEntity) {
                    if (allowed) {
                        this.constructionHideEntities(server, client, construction.characterId, freePlacedEntity, true);
                    }
                }
            }
        }
        if (construction.isInside(client.character.state.position)) {
            if (allowed) {
                this.constructionHidePlayer(server, client, construction.characterId, true);
                return true;
            }
            else if (!client.isAdmin || !client.isDebugMode) {
                const damageInfo = {
                    entity: "Server.Permissions",
                    damage: 99999
                };
                server.killCharacter(client, damageInfo);
                return false;
            }
        }
        return false;
    }
    constructionHidePlayer(server, client, constructionGuid, state) {
        if (state) {
            //TODO: Leaving a group will not remove the player because the client already has the isHidden flag.
            //To reproduce: Create a group and give the other player visitor perm, before disbanding the group remove other players' perms.
            if (!client.character.isHidden) {
                client.character.isHidden = constructionGuid;
                for (const a in server._clients) {
                    const iteratedClient = server._clients[a];
                    const constructionEntity = server.getConstructionEntity(constructionGuid);
                    let hasVisitPermission = false;
                    if (constructionEntity) {
                        hasVisitPermission = constructionEntity.getHasPermission(server, iteratedClient.character.characterId, enums_1.ConstructionPermissionIds.VISIT);
                    }
                    const isSameGroup = client.character.groupId != 0 &&
                        iteratedClient.character.groupId != 0 &&
                        client.character.groupId === iteratedClient.character.groupId;
                    const hasPermission = isSameGroup && hasVisitPermission;
                    if (iteratedClient.spawnedEntities.has(client.character) &&
                        iteratedClient.character.isHidden != client.character.isHidden &&
                        !hasPermission) {
                        server.sendData(iteratedClient, "Character.RemovePlayer", {
                            characterId: client.character.characterId
                        });
                        iteratedClient.spawnedEntities.delete(client.character);
                    }
                }
            }
            else
                return;
        }
        else if (client.character.isHidden)
            client.character.isHidden = "";
    }
    constructionHideEntities(server, client, constructionGuid, freePlacedEntity, state) {
        if (state) {
            if (!freePlacedEntity.isHidden) {
                freePlacedEntity.isHidden = constructionGuid;
                for (const a in server._clients) {
                    const iteratedClient = server._clients[a];
                    const constructionEntity = server.getConstructionEntity(constructionGuid);
                    let hasVisitPermission = false;
                    if (constructionEntity) {
                        hasVisitPermission = constructionEntity.getHasPermission(server, iteratedClient.character.characterId, enums_1.ConstructionPermissionIds.VISIT);
                    }
                    const isSameGroup = client.character.groupId != 0 &&
                        iteratedClient.character.groupId != 0 &&
                        client.character.groupId === iteratedClient.character.groupId;
                    const hasPermission = isSameGroup && hasVisitPermission;
                    if (iteratedClient.spawnedEntities.has(client.character) &&
                        iteratedClient.character.isHidden != freePlacedEntity.isHidden &&
                        !hasPermission) {
                        server.sendData(iteratedClient, "Character.RemovePlayer", {
                            characterId: client.character.characterId
                        });
                        iteratedClient.spawnedEntities.delete(client.character);
                    }
                }
            }
            else
                return;
        }
        else if (freePlacedEntity.isHidden)
            freePlacedEntity.isHidden = "";
    }
    tpPlayerOutsideFoundation(server, client, foundation, tpUp = false) {
        const currentAngle = Math.atan2(client.character.state.position[2] - foundation.state.position[2], client.character.state.position[0] - foundation.state.position[0]);
        if (tpUp) {
            server.sendChatText(client, "Construction: Stuck under foundation");
            const foundationY = foundation.state.position[1], yOffset = foundation.itemDefinitionId == enums_1.Items.FOUNDATION ? 2.2 : 0.1;
            client.startLoc = foundationY + yOffset;
            server.sendData(client, "ClientUpdate.UpdateLocation", {
                position: new Float32Array([
                    client.character.state.position[0],
                    foundationY + yOffset,
                    client.character.state.position[2],
                    1
                ]),
                triggerLoadingScreen: false
            });
            client.enableChecks = false;
            client.isInAir = false;
            setTimeout(() => {
                client.enableChecks = true;
            }, 500);
            return;
        }
        const newPos = (0, utils_1.movePoint)(client.character.state.position, currentAngle, 2.5);
        server.sendChatText(client, "Construction: no visitor permission");
        if (client.vehicle.mountedVehicle) {
            server.dismountVehicle(client);
        }
        client.character.state.position = new Float32Array([
            newPos[0],
            client.character.state.position[1],
            newPos[2],
            1
        ]);
        server.sendData(client, "ClientUpdate.UpdateLocation", {
            position: client.character.state.position,
            triggerLoadingScreen: false
        });
        client.enableChecks = false;
        setTimeout(() => {
            client.enableChecks = true;
        }, 500);
        setTimeout(() => {
            if (foundation.isSecured &&
                foundation.isInside(client.character.state.position)) {
                const damageInfo = {
                    entity: "Server.Permissions",
                    damage: 99999
                };
                client.character.damage(server, damageInfo);
            }
        }, 2000);
        this.recheckClientInsideShelter(client, server, currentAngle);
        this.checkFoundationPermission(server, client, foundation);
    }
    recheckClientInsideShelter(client, server, tpDirection) {
        for (const a in server._constructionSimple) {
            const simple = server._constructionSimple[a];
            const shelters = [
                enums_1.Items.SHELTER,
                enums_1.Items.SHELTER_LARGE,
                enums_1.Items.SHELTER_UPPER,
                enums_1.Items.SHELTER_UPPER_LARGE
            ];
            if (!shelters.includes(simple.itemDefinitionId))
                continue;
            if (simple.isInside(client.character.state.position)) {
                const newPos = (0, utils_1.movePoint)(client.character.state.position, tpDirection, 2.5);
                client.character.state.position = new Float32Array([
                    newPos[0],
                    client.character.state.position[1],
                    newPos[2],
                    1
                ]);
                server.sendData(client, "ClientUpdate.UpdateLocation", {
                    position: client.character.state.position,
                    triggerLoadingScreen: false
                });
                this.recheckClientInsideShelter(client, server, tpDirection);
                return;
            }
        }
    }
    plantManager(server) {
        const date = new Date().getTime();
        for (const characterId in server._temporaryObjects) {
            const object = server._temporaryObjects[characterId];
            if (object instanceof plantingdiameter_1.PlantingDiameter) {
                if (object.disappearTimestamp < date &&
                    Object.values(object.seedSlots).length === 0) {
                    server.deleteEntity(object.characterId, server._temporaryObjects);
                }
                else if (object.disappearTimestamp < date)
                    object.disappearTimestamp = date + 86400000;
                if (object.fertilizedTimestamp < date)
                    object.isFertilized = false;
                Object.values(object.seedSlots).forEach((plant) => {
                    if (plant.nextStateTime < date)
                        plant.grow(server);
                });
            }
        }
    }
    shouldHideEntity(server, client, entity) {
        if (!(entity instanceof lootableconstructionentity_1.LootableConstructionEntity ||
            entity instanceof constructionchildentity_1.ConstructionChildEntity)) {
            return false;
        }
        // these entities can be handled as freeplace but should never be hidden
        switch (entity.itemDefinitionId) {
            case enums_1.Items.SHELTER_UPPER:
            case enums_1.Items.SHELTER_UPPER_LARGE:
            case enums_1.Items.STRUCTURE_STAIRS_UPPER:
            case enums_1.Items.METAL_WALL_UPPER:
                return false;
        }
        const parent = entity.getParent(server);
        if (!parent)
            return false;
        // foundations, expansions, and tampers should never hide entities
        switch (parent.itemDefinitionId) {
            case enums_1.Items.FOUNDATION:
            case enums_1.Items.FOUNDATION_EXPANSION:
            case enums_1.Items.GROUND_TAMPER:
                return false;
        }
        const parentSecured = parent.isSecured, hasVisitPermission = parent.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.VISIT), isInside = parent.isInside(entity.state.position);
        return (!client.isDebugMode && parentSecured && isInside && !hasVisitPermission);
        // TODO: check if character is in secured shelter / shack
    }
    spawnConstructionFreeplace(server, client, parentEntity) {
        for (const entity of Object.values(parentEntity.freeplaceEntities)) {
            if (!(0, utils_1.isPosInRadius)(entity.npcRenderDistance || server.charactersRenderDistance, entity.state.position, client.character.state.position) ||
                this.shouldHideEntity(server, client, entity))
                continue;
            if (entity instanceof constructionchildentity_1.ConstructionChildEntity) {
                this.spawnSimpleConstruction(server, client, entity);
            }
            else if (entity instanceof constructiondoor_1.ConstructionDoor) {
                this.spawnConstructionDoor(server, client, entity);
            }
            else if (entity instanceof lootableconstructionentity_1.LootableConstructionEntity) {
                this.spawnLootableConstruction(server, client, entity);
            }
        }
    }
    spawnConstructionParent(server, client, entity) {
        if (!client.spawnedEntities.has(entity)) {
            server.addSimpleNpc(client, entity);
            client.spawnedEntities.add(entity);
        }
        // slotted construction spawning
        this.spawnConstructionTree(server, client, entity);
        // freeplace construction spawning
        this.spawnConstructionFreeplace(server, client, entity);
    }
    spawnConstructionDoor(server, client, entity) {
        if (client.spawnedEntities.has(entity) || !client.isSynced)
            return;
        server.addLightweightNpc(client, entity, server.getItemDefinition(entity.itemDefinitionId)?.NAME_ID);
        client.spawnedEntities.add(entity);
        server.updateResource(client, entity.characterId, (entity.health / entity.maxHealth) * 1000000, enums_1.ResourceIds.CONSTRUCTION_CONDITION, enums_1.ResourceTypes.CONDITION);
        if (entity.isOpen) {
            server.sendData(client, "PlayerUpdatePosition", {
                transientId: entity.transientId,
                positionUpdate: {
                    sequenceTime: 0,
                    unknown3_int8: 0,
                    position: entity.state.position,
                    orientation: entity.openAngle
                }
            });
        }
    }
    spawnSimpleConstruction(server, client, entity, spawnTree = true) {
        if (!client.spawnedEntities.has(entity)) {
            server.addSimpleNpc(client, entity);
            client.spawnedEntities.add(entity);
        }
        if (!spawnTree)
            return;
        // slotted construction spawning
        this.spawnConstructionTree(server, client, entity);
        // freeplace construction spawning
        this.spawnConstructionFreeplace(server, client, entity);
    }
    spawnLootableConstruction(server, client, entity) {
        if (client.spawnedEntities.has(entity))
            return;
        server.addSimpleNpc(client, entity);
        client.spawnedEntities.add(entity);
    }
    spawnConstructionTree(server, client, parentEntity) {
        for (const slotMap of parentEntity.getOccupiedSlotMaps()) {
            for (const entity of Object.values(slotMap)) {
                if ((0, utils_1.isPosInRadius)(entity.npcRenderDistance
                    ? entity.npcRenderDistance
                    : server.charactersRenderDistance, client.character.state.position, entity.state.position)) {
                    if (entity instanceof constructionchildentity_1.ConstructionChildEntity) {
                        this.spawnSimpleConstruction(server, client, entity);
                    }
                    else if (entity instanceof constructiondoor_1.ConstructionDoor) {
                        this.spawnConstructionDoor(server, client, entity);
                    }
                }
            }
        }
    }
    constructionPermissionsManager(server, client) {
        let hide = false;
        client.character.insideBuilding = "";
        for (const object of client.spawnedEntities) {
            if (object instanceof constructionparententity_1.ConstructionParentEntity) {
                if (object.isInside(client.character.state.position))
                    client.character.insideBuilding = object.characterId;
                if (this.checkFoundationPermission(server, client, object)) {
                    hide = true;
                    continue;
                }
            }
            if (object instanceof constructionchildentity_1.ConstructionChildEntity) {
                if (object.isInside(client.character.state.position))
                    client.character.insideBuilding = object.characterId;
                if (this.checkConstructionChildEntityPermission(server, client, object)) {
                    hide = true;
                    continue;
                }
            }
        }
        if (!hide && client.character.isHidden) {
            client.character.isHidden = "";
            server.spawnCharacterToOtherClients(client.character);
        }
    }
    repairChildEntity(server, entity) {
        let accumulatedItemDamage = 0;
        if (entity instanceof constructionchildentity_1.ConstructionChildEntity) {
            Object.values(entity.occupiedShelterSlots).forEach((slot) => {
                accumulatedItemDamage += this.repairChildEntity(server, slot);
            });
            Object.values(entity.occupiedWallSlots).forEach((wall) => {
                accumulatedItemDamage += this.repairChildEntity(server, wall);
            });
            Object.values(entity.occupiedUpperWallSlots).forEach((slot) => {
                accumulatedItemDamage += this.repairChildEntity(server, slot);
            });
            accumulatedItemDamage += this.repairFreeplaceEntities(server, entity);
        }
        if (entity.health >= entity.maxHealth)
            return accumulatedItemDamage;
        this.repairConstruction(server, entity, entity.maxHealth / 10);
        accumulatedItemDamage += 15;
        return accumulatedItemDamage;
    }
    fullyRepairShelterSlots(server, entity) {
        Object.values(entity.occupiedShelterSlots).forEach((child) => {
            this.fullyRepairChildEntity(server, child);
        });
    }
    fullyRepairWallSlots(server, occupiedWallSlotMap) {
        Object.values(occupiedWallSlotMap).forEach((child) => {
            this.fullyRepairChildEntity(server, child);
        });
    }
    fullyRepairChildEntity(server, entity) {
        if (entity instanceof constructionchildentity_1.ConstructionChildEntity) {
            this.fullyRepairShelterSlots(server, entity);
            this.fullyRepairWallSlots(server, entity.occupiedWallSlots);
            this.fullyRepairWallSlots(server, entity.occupiedUpperWallSlots);
            this.fullyRepairFreeplaceEntities(server, entity);
        }
        this.fullyRepairConstruction(server, entity);
    }
    repairConstruction(server, entity, amount) {
        const damage = entity.health + amount > entity.maxHealth
            ? (entity.maxHealth - entity.health) * -1
            : (amount *= -1), damageInfo = {
            entity: "Server.RepairConstruction",
            damage
        };
        entity.damage(server, damageInfo);
    }
    fullyRepairConstruction(server, entity) {
        entity.health = entity.maxHealth;
        server.sendDataToAllWithSpawnedEntity(server.getConstructionDictionary(entity.characterId), entity.characterId, "Character.UpdateSimpleProxyHealth", entity.pGetSimpleProxyHealth());
        entity.isDecayProtected = true;
    }
    repairFreeplaceEntities(server, entity) {
        let accumulatedItemDamage = 0;
        Object.values(entity.freeplaceEntities).forEach((child) => {
            if (child.health < child.maxHealth) {
                this.repairConstruction(server, child, child.maxHealth / 10);
                accumulatedItemDamage += 5;
            }
        });
        return accumulatedItemDamage;
    }
    repairShelterSlots(server, entity) {
        let accumulatedItemDamage = 0;
        Object.values(entity.occupiedShelterSlots).forEach((child) => {
            accumulatedItemDamage += this.repairChildEntity(server, child);
        });
        return accumulatedItemDamage;
    }
    repairWallSlots(server, entity) {
        let accumulatedItemDamage = 0;
        Object.values(entity.occupiedWallSlots).forEach((child) => {
            accumulatedItemDamage += this.repairChildEntity(server, child);
        });
        return accumulatedItemDamage;
    }
    demolishConstructionEntity(server, client, entity, weaponItem) {
        switch (entity.itemDefinitionId) {
            case enums_1.Items.FOUNDATION:
            case enums_1.Items.FOUNDATION_EXPANSION:
            case enums_1.Items.GROUND_TAMPER:
            case enums_1.Items.FOUNDATION_RAMP:
            case enums_1.Items.FOUNDATION_STAIRS:
                return;
        }
        const permission = entity.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.DEMOLISH);
        if (!permission) {
            this.placementError(server, client, enums_1.ConstructionErrors.DEMOLISH_PERMISSION);
            return;
        }
        if (entity.canUndoPlacement(server, client)) {
            // give back item only if can undo
            client.character.lootItem(server, server.generateItem(entity.itemDefinitionId));
            entity.destroy(server);
        }
        entity.damage(server, {
            entity: client.character.characterId,
            weapon: weaponItem.itemDefinitionId,
            damage: entity.maxHealth / 3 + 10
        });
        server.damageItem(client.character, weaponItem, 50);
    }
    hammerConstructionEntity(server, client, entity, weaponItem) {
        const foundation = entity.getParentFoundation(server);
        let timeDif;
        if (foundation)
            timeDif = Date.now() - foundation.lastDamagedTimestamp;
        if (entity instanceof constructionparententity_1.ConstructionParentEntity)
            timeDif = Date.now() - entity.lastDamagedTimestamp;
        const cooldownTime = 30000;
        if (timeDif && timeDif < cooldownTime) {
            server.sendAlert(client, `You cant repair this base for the next ${(cooldownTime / 1000 - Number(timeDif / 1000)).toFixed(2)} seconds`);
            return;
        }
        let accumulatedItemDamage = 0;
        server.sendCompositeEffectToAllInRange(15, client.character.characterId, entity.state.position, enums_1.Effects.PFX_Impact_Blade_Wood);
        if (entity instanceof constructionparententity_1.ConstructionParentEntity) {
            Object.values(entity.occupiedExpansionSlots).forEach((expansion) => {
                const timeDiffExpansion = Date.now() - expansion.lastDamagedTimestamp;
                if (timeDiffExpansion < 30000) {
                    server.sendAlert(client, `You cant repair this base for the next ${(cooldownTime / 1000 - Number(timeDiffExpansion / 1000)).toFixed(2)} seconds`);
                    return;
                }
                // repair every object on each expansion
                accumulatedItemDamage += this.repairShelterSlots(server, expansion);
                accumulatedItemDamage += this.repairWallSlots(server, expansion);
                accumulatedItemDamage += this.repairFreeplaceEntities(server, expansion);
            });
        }
        if (entity instanceof constructionchildentity_1.ConstructionChildEntity) {
            // repair every object on main foundation
            accumulatedItemDamage += this.repairShelterSlots(server, entity);
            accumulatedItemDamage += this.repairWallSlots(server, entity);
            accumulatedItemDamage += this.repairFreeplaceEntities(server, entity);
        }
        if (entity.health < entity.maxHealth) {
            this.repairConstruction(server, entity, entity.maxHealth / 10);
            accumulatedItemDamage += 15;
        }
        server.damageItem(client.character, weaponItem, Math.ceil(accumulatedItemDamage / 4));
        client.character.lastMeleeHitTime = Date.now();
    }
    fullyRepairFreeplaceEntities(server, entity) {
        Object.values(entity.freeplaceEntities).forEach((child) => {
            if (child.health < child.maxHealth) {
                this.fullyRepairConstruction(server, child);
            }
        });
    }
    fullyRepairFoundation(server, entity) {
        Object.values(entity.occupiedExpansionSlots).forEach((expansion) => {
            // repair every object on each expansion
            this.fullyRepairShelterSlots(server, expansion);
            this.fullyRepairWallSlots(server, expansion.occupiedWallSlots);
            this.fullyRepairFreeplaceEntities(server, expansion);
        });
        // repair every object on main foundation
        this.fullyRepairShelterSlots(server, entity);
        this.fullyRepairWallSlots(server, entity.occupiedWallSlots);
        this.fullyRepairFreeplaceEntities(server, entity);
        if (entity.health >= entity.maxHealth)
            return;
        this.fullyRepairConstruction(server, entity);
    }
    isConstructionInSecuredArea(server, construction) {
        const gates = [
            enums_1.Items.METAL_GATE,
            enums_1.Items.METAL_WALL,
            enums_1.Items.METAL_WALL_UPPER,
            enums_1.Items.METAL_DOORWAY
        ];
        const doors = [
            enums_1.Items.DOOR_BASIC,
            enums_1.Items.DOOR_METAL,
            enums_1.Items.DOOR_WOOD
        ];
        const parentFoundation = construction.getParentFoundation(server);
        if (!parentFoundation || !parentFoundation.isSecured)
            return false;
        if (!gates.includes(construction.itemDefinitionId) &&
            !doors.includes(construction.itemDefinitionId)) {
            return true;
        }
        if (((parentFoundation.itemDefinitionId == enums_1.Items.FOUNDATION_EXPANSION ||
            parentFoundation.itemDefinitionId == enums_1.Items.GROUND_TAMPER) &&
            !doors.includes(construction.itemDefinitionId)) ||
            parentFoundation.itemDefinitionId == enums_1.Items.SHACK ||
            parentFoundation.itemDefinitionId == enums_1.Items.SHACK_BASIC ||
            parentFoundation.itemDefinitionId == enums_1.Items.SHACK_SMALL)
            return false;
        if (!(construction instanceof lootableconstructionentity_1.LootableConstructionEntity) &&
            (parentFoundation.itemDefinitionId == enums_1.Items.FOUNDATION ||
                parentFoundation.itemDefinitionId == enums_1.Items.GROUND_TAMPER)) {
            switch (construction.getSlotNumber()) {
                case 4:
                case 5:
                case 6:
                    if (parentFoundation.occupiedExpansionSlots["1"]?.isSecured) {
                        return true;
                    }
                    break;
                case 1:
                case 2:
                case 3:
                    if (parentFoundation.occupiedExpansionSlots["2"]?.isSecured) {
                        return true;
                    }
                    break;
                case 10:
                case 11:
                case 12:
                    if (parentFoundation.occupiedExpansionSlots["3"]?.isSecured) {
                        return true;
                    }
                    break;
                case 7:
                case 8:
                case 9:
                    if (parentFoundation.occupiedExpansionSlots["4"]?.isSecured) {
                        return true;
                    }
                    break;
            }
            return false;
        }
        else if (parentFoundation.itemDefinitionId == enums_1.Items.FOUNDATION_EXPANSION) {
            if (doors.includes(construction.itemDefinitionId)) {
                if (parentFoundation.isSecured) {
                    return true;
                }
                else {
                    return false;
                }
            }
        }
        return false;
    }
    OnMeleeHit(server, damageInfo, construction) {
        const client = server.getClientByCharId(damageInfo.entity), weapon = client?.character.getEquippedWeapon();
        if (!client || !weapon)
            return;
        switch (weapon.itemDefinitionId) {
            case enums_1.Items.WEAPON_HAMMER_DEMOLITION:
                this.demolishConstructionEntity(server, client, construction, weapon);
                return;
            case enums_1.Items.WEAPON_HAMMER:
                this.hammerConstructionEntity(server, client, construction, weapon);
                return;
        }
        if (this.isConstructionInSecuredArea(server, construction)) {
            if (client) {
                server.constructionManager.sendBaseSecuredMessage(server, client);
            }
        }
    }
    sendBaseSecuredMessage(server, client, rayCastProtectionType = 0) {
        switch (rayCastProtectionType) {
            case 1:
                server.sendAlert(client, "Base protection unbroken, damage reduced");
                return;
            case 2:
                server.sendAlert(client, "You must destroy the base's gate before placing explosives inside.");
                return;
            default:
                server.sendAlert(client, "You must destroy the base's gate before affecting interior structures.");
                return;
        }
    }
    checkConstructionDamage(server, constructionObject, damage, position, entityPosition, itemDefinitionId) {
        switch (itemDefinitionId) {
            case enums_1.Items.IED:
            case enums_1.Items.LANDMINE:
                break; // use base damage
            case enums_1.Items.FUEL_ETHANOL:
                damage /= 2.7;
                break;
            case enums_1.Items.FUEL_BIOFUEL:
                damage /= 3;
                break;
            default: // vehicles
                damage /= 12;
                break;
        }
        const distance = (0, utils_1.getDistance)(entityPosition, position);
        constructionObject.damage(server, {
            entity: "",
            damage: distance < constructionObject.damageRange
                ? damage
                : damage / Math.sqrt(distance)
        });
    }
    isSlotOnPlacementCooldown(slots, slot) {
        return slots[slot] > Date.now();
    }
    hasOwnedBases(server, characterId) {
        for (const a in server._constructionFoundations) {
            const foundation = server._constructionFoundations[a];
            if (foundation.ownerCharacterId === characterId) {
                return true;
            }
        }
        return false;
    }
    removeOwnerFromBases(server, characterId) {
        for (const a in server._constructionFoundations) {
            const foundation = server._constructionFoundations[a];
            if (foundation.ownerCharacterId === characterId) {
                foundation.ownerCharacterId = "";
            }
        }
    }
}
exports.ConstructionManager = ConstructionManager;
