// h1z1-server/out/servers/ZoneServer2016/managers/craftmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.CraftManager = void 0;
const enums_1 = require("../models/enums");
const utils_1 = require("../../../utils/utils");
const challengemanager_1 = require("./challengemanager");
const itemobject_1 = require("../entities/itemobject");
const debug = require("debug")("ZoneServer");
/**
 * Retrieves the craft components data source from the client's inventory and mounted container.
 * @param client The client to get the craft components data source for.
 * @param server ZoneServer pointer.
 * @returns The craft components data source object.
 */
function getCraftComponentsDataSource(client, server) {
    // ignoring proximity container items for now
    // todo: include other datasources when they are available ex. proximity items, accessed container
    const inventory = {};
    Object.keys(client.character._containers).forEach((loadoutSlotId) => {
        const container = client.character._containers[Number(loadoutSlotId)];
        Object.keys(container.items).forEach((itemGuid) => {
            const item = container.items[itemGuid];
            if (inventory[item.itemDefinitionId]) {
                inventory[item.itemDefinitionId].stackCount += item.stackCount;
            }
            else {
                inventory[item.itemDefinitionId] = {
                    ...item
                }; // push new itemstack
            }
        });
    });
    const proximityItems = server.getCraftingProximityItems(client)?.items;
    if (proximityItems) {
        proximityItems.forEach((item) => {
            if (inventory[item.itemDefinitionId]) {
                inventory[item.itemDefinitionId].stackCount += item.itemData.count;
            }
            else {
                inventory[item.itemDefinitionId] = {
                    ...item.itemData,
                    stackCount: item.itemData.count
                }; // push new itemstack
            }
        });
    }
    if (!client.character.mountedContainer)
        return inventory;
    const container = client.character.mountedContainer.getContainer();
    if (!container)
        return inventory; // should never trigger
    Object.keys(container.items).forEach((itemGuid) => {
        const item = container.items[itemGuid];
        if (inventory[item.itemDefinitionId]) {
            inventory[item.itemDefinitionId].stackCount += item.stackCount;
        }
        else {
            inventory[item.itemDefinitionId] = {
                ...item
            }; // push new itemstack
        }
    });
    return inventory;
}
/**
 * CraftManager handles the crafting of a recipe by a client.
 */
class CraftManager {
    /** Iterated count of craft actions in a loop */
    craftLoopCount = 0;
    maxCraftLoopCount = 500;
    /** HashMap of item components that make-up the parent item,
     * uses itemDefinitionId (number) for indexing
     */
    componentsDataSource = {};
    /**
     * Constructs a new instance of the CraftManager class.
     * @param client - The client object.
     * @param server - The server object.
     * @param recipeId - The ID of the recipe to craft.
     * @param count - The number of times to craft the recipe.
     */
    constructor(client, server, recipeId = 0, count = 0) {
        this.componentsDataSource = getCraftComponentsDataSource(client, server);
        this.start(client, server, recipeId, count);
    }
    /**
     * Removes simulated craft components from the internal recipe components data source.
     * @param itemDefinitionId The item definition ID of the craft component to remove.
     * @param count The number of craft components to remove.
     * @returns A boolean indicating if the removal was successful.
     */
    removeSimulatedCraftComponent(itemDefinitionId, count) {
        const removeItem = this.componentsDataSource[itemDefinitionId];
        if (!removeItem)
            return false;
        if (removeItem.stackCount == count) {
            delete this.componentsDataSource[itemDefinitionId];
        }
        else if (removeItem.stackCount > count) {
            this.componentsDataSource[itemDefinitionId].stackCount -= count;
        }
        else {
            // if removeItem.stackCount < count
            return false;
        }
        return true;
    }
    /**
     * Removes a craft component from the character and updates the remaining items.
     * @param server The ZoneServer2016 instance.
     * @param itemDS The item data source containing the craft component.
     * @param count The number of craft components to remove.
     * @returns A boolean indicating if the removal was successful.
     */
    async removeCraftComponent(server, itemDS, count) {
        if (await server.removeInventoryItem(itemDS.character, itemDS.item, count))
            return true;
        const item = itemDS.item;
        const remainder = item?.stackCount - count;
        if (remainder <= 0) {
            return server.deleteEntity(item.ownerCharacterId, server._spawnedItems);
        }
        const entity = server.getEntity(item.ownerCharacterId);
        if (entity instanceof itemobject_1.ItemObject) {
            entity.item.stackCount -= count;
            const client = server.getClientByCharId(itemDS.character.characterId);
            if (!client)
                return true;
            server.sendData(client, "ClientUpdate.ProximateItems", server.getProximityItems(client));
            return true;
        }
        return false;
    }
    /**
     * Restores removed items to the character's inventory.
     * @param server The ZoneServer2016 instance.
     * @param client The client whose inventory needs restoration.
     * @param removedItems Array of removed items to restore.
     */
    async restoreRemovedItems(server, client, removedItems) {
        for (const removed of removedItems) {
            await client.character.lootItem(server, server.generateItem(removed.itemDS.item.itemDefinitionId, removed.count));
        }
    }
    /**
     * Generates the craft queue based on the recipe and its components.
     * @param server The ZoneServer2016 instance.
     * @param client The client performing the craft.
     * @param recipe The recipe object.
     * @param recipeCount The number of times to repeat the recipe.
     * @param recipeId The ID of the recipe being crafted.
     * @param craftCount The total number of items to craft.
     * @returns A promise resolving to a boolean indicating if the craft queue generation was successful.
     */
    async generateCraftQueue(server, client, recipe, recipeCount, recipeId, craftCount) {
        for (const component of recipe.components) {
            const remainingItems = component.requiredAmount * recipeCount;
            // if component isn't found at all
            if (!this.componentsDataSource[component.itemDefinitionId]) {
                const componentRecipe = client.character.recipes[component.itemDefinitionId], componentBundleCount = componentRecipe?.bundleCount || 1;
                if (!componentRecipe) {
                    debug(`[CraftManager] ${client.character.name} tried to craft an invalid recipe ${component.itemDefinitionId}!`);
                    return false; // no valid recipe to craft component
                }
                if (component.requiredAmount / componentBundleCount < 1) {
                    if (!(await this.craftItem(server, client, component.itemDefinitionId, Math.ceil((component.requiredAmount / componentBundleCount) * recipeCount)))) {
                        return false; // craftItem returned some error
                    }
                }
                else {
                    for (let i = 0; i < recipeCount; i++) {
                        if (!(await this.craftItem(server, client, component.itemDefinitionId, Math.ceil(component.requiredAmount / componentBundleCount)))) {
                            return false; // craftItem returned some error
                        }
                    }
                }
                // if there is only some of the component
            }
            else if (this.componentsDataSource[component.itemDefinitionId].stackCount <
                remainingItems) {
                const componentRecipe = client.character.recipes[component.itemDefinitionId], componentBundleCount = componentRecipe?.bundleCount || 1;
                if (!componentRecipe) {
                    debug(`[CraftManager] ${client.character.name} tried to craft an invalid recipe ${component.itemDefinitionId}!`);
                    return false; // no valid recipe to craft component
                }
                let stackCount = this.componentsDataSource[component.itemDefinitionId].stackCount;
                if (component.requiredAmount / componentBundleCount < 1) {
                    let craft = 0;
                    for (let i = 0; i < recipeCount; i++) {
                        let craftAmount = 0; // amount that needs crafted per recipe iteration
                        if (stackCount >= component.requiredAmount) {
                            // use some of stack and don't craft any component through one recipe iteration
                            stackCount -= Math.ceil(component.requiredAmount / componentBundleCount);
                        }
                        else {
                            // use all of stack if required component amount is greater than stackCount
                            craftAmount =
                                Math.ceil(component.requiredAmount / componentBundleCount) -
                                    stackCount;
                            stackCount = 0;
                        }
                        craft += Math.ceil(craftAmount / componentBundleCount);
                    }
                    if (!(await this.craftItem(server, client, component.itemDefinitionId, craft))) {
                        return false; // craftItem returned some error
                    }
                }
                else {
                    for (let i = 0; i < recipeCount; i++) {
                        let craftAmount = 0; // amount that needs crafted per recipe iteration
                        if (stackCount >= component.requiredAmount) {
                            // use some of stack and don't craft any component through one recipe iteration
                            stackCount -= component.requiredAmount;
                        }
                        else {
                            // use all of stack if required component amount is greater than stackCount
                            craftAmount = component.requiredAmount - stackCount;
                            stackCount = 0;
                        }
                        if (!(await this.craftItem(server, client, component.itemDefinitionId, Math.ceil(craftAmount / componentBundleCount)))) {
                            return false; // craftItem returned some error
                        }
                    }
                }
            }
            if (!this.removeSimulatedCraftComponent(component.itemDefinitionId, remainingItems)) {
                server.containerError(client, enums_1.ContainerErrors.NO_ITEM_IN_SLOT);
                return false;
            }
        }
        // push dummy item
        if (this.componentsDataSource[recipeId]) {
            this.componentsDataSource[recipeId].stackCount += craftCount;
        }
        else {
            this.componentsDataSource[recipeId] = {
                itemDefinitionId: recipeId,
                stackCount: craftCount
            };
        }
        return true;
    }
    /**
     * Retrieves the inventory data source for a character.
     * @param character The Character2016 instance.
     * @returns The inventory data source object.
     */
    getInventoryDataSource(character) {
        const inv = {};
        Object.values(character.getInventoryAsContainer()).forEach((items) => {
            inv[items[0].itemDefinitionId] = items.map((item) => {
                return {
                    item: item,
                    character: character
                };
            });
        });
        const mountedContainer = character.mountedContainer;
        if (!mountedContainer)
            return inv;
        const container = mountedContainer.getContainer();
        if (!container)
            return inv; // should never trigger
        for (const item of Object.values(container.items)) {
            const itemDS = {
                item,
                character: mountedContainer
            };
            if (inv[item.itemDefinitionId]) {
                inv[item.itemDefinitionId].push(itemDS);
                continue;
            }
            inv[item.itemDefinitionId] = [itemDS];
        }
        return inv;
    }
    /**
     * Crafts an item using the given recipe and adds it to the client's inventory.
     * @param server The ZoneServer2016 instance.
     * @param client The client performing the craft.
     * @param recipeId The ID of the recipe being crafted.
     * @param recipeCount The number of times to repeat the recipe.
     * @returns A promise resolving to a boolean indicating if the crafting process was successful.
     */
    async craftItem(server, client, recipeId, recipeCount) {
        if (!recipeCount)
            return false;
        this.craftLoopCount++;
        if (this.craftLoopCount > this.maxCraftLoopCount) {
            return false;
        }
        const recipe = client.character.recipes[recipeId], bundleCount = recipe?.bundleCount || 1, // the amount of an item crafted from 1 recipe (ex. crafting 1 stick recipe gives you 2)
        craftCount = recipeCount * bundleCount; // the actual amount of items to craft
        if (!recipe)
            return false;
        switch (recipe.filterId) {
            case enums_1.FilterIds.COOKING:
            case enums_1.FilterIds.FURNACE:
                server.sendAlert(client, "This recipe requires a furnace, barbeque, or campfire to craft");
                return false;
        }
        if (recipe.requireWorkbench) {
            if (!(0, utils_1.checkConstructionInRange)(server._constructionSimple, client.character.state.position, 3, enums_1.Items.WORKBENCH) &&
                !(0, utils_1.checkConstructionInRange)(server._worldSimpleConstruction, client.character.state.position, 3, enums_1.Items.WORKBENCH)) {
                server.sendAlert(client, "You must be near a workbench to complete this recipe");
                return false;
            }
        }
        if (recipe.requireWeaponWorkbench) {
            if (!(0, utils_1.checkConstructionInRange)(server._constructionSimple, client.character.state.position, 3, enums_1.Items.WORKBENCH_WEAPON) &&
                !(0, utils_1.checkConstructionInRange)(server._worldSimpleConstruction, client.character.state.position, 3, enums_1.Items.WORKBENCH_WEAPON)) {
                server.sendAlert(client, "You must be near a weapon workbench to complete this recipe");
                return false;
            }
        }
        if (!(await this.generateCraftQueue(server, client, recipe, recipeCount, recipeId, craftCount))) {
            return false;
        }
        //#region CRAFTING
        await server.pUtilizeHudTimer(client, server.getItemDefinition(recipeId)?.NAME_ID ?? 0, 1000 * recipeCount, 0);
        const r = client.character.recipes[recipeId];
        const removedItems = [];
        let craftSuccess = true;
        for (const component of r.components) {
            const inventory = this.getInventoryDataSource(client.character);
            const proximityItems = server.getCraftingProximityItems(client);
            if (proximityItems?.items) {
                const character = client.character;
                proximityItems.items.forEach((item) => {
                    if (!inventory[item.itemDefinitionId]) {
                        inventory[item.itemDefinitionId] = [
                            {
                                item: {
                                    ...item.itemData,
                                    stackCount: item?.itemData?.count ?? 0
                                },
                                character
                            }
                        ];
                    }
                    else if (inventory[item.itemDefinitionId]) {
                        inventory[item.itemDefinitionId].push({
                            item: {
                                ...item.itemData,
                                stackCount: item?.itemData?.count ?? 0
                            },
                            character
                        });
                    }
                });
            }
            let remainingItems = component.requiredAmount * recipeCount, stackCount = 0;
            if (!inventory[component.itemDefinitionId]) {
                server.containerError(client, enums_1.ContainerErrors.NO_ITEM_IN_SLOT);
                craftSuccess = false;
                break;
            }
            for (const itemDS of inventory[component.itemDefinitionId]) {
                stackCount += itemDS.item.stackCount;
            }
            if (remainingItems > stackCount) {
                server.containerError(client, enums_1.ContainerErrors.NO_ITEM_IN_SLOT);
                craftSuccess = false;
                break;
            }
            for (const itemDS of inventory[component.itemDefinitionId]) {
                if (itemDS.item.stackCount >= remainingItems) {
                    if (!(await this.removeCraftComponent(server, itemDS, remainingItems))) {
                        server.containerError(client, enums_1.ContainerErrors.NO_ITEM_IN_SLOT);
                        craftSuccess = false; // return if not enough items
                        break;
                    }
                    removedItems.push({
                        itemDS,
                        count: Math.min(itemDS.item.stackCount, remainingItems)
                    });
                    remainingItems = 0;
                }
                else {
                    if (await this.removeCraftComponent(server, itemDS, itemDS.item.stackCount)) {
                        removedItems.push({
                            itemDS,
                            count: Math.min(itemDS.item.stackCount, remainingItems)
                        });
                        remainingItems -= itemDS.item.stackCount;
                    }
                    else {
                        server.containerError(client, enums_1.ContainerErrors.NO_ITEM_IN_SLOT);
                        craftSuccess = false; // return if not enough items
                        break;
                    }
                }
                if (!remainingItems)
                    break;
            }
        }
        if (!craftSuccess) {
            await this.restoreRemovedItems(server, client, removedItems);
            return false;
        }
        client.character.lootItem(server, server.generateItem(recipeId, craftCount, true));
        if (recipe.leftOverItems) {
            recipe.leftOverItems.forEach((id) => {
                client.character.lootItem(server, server.generateItem(id, 1));
            });
        }
        switch (recipeId) {
            case enums_1.Items.IED:
                server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.IED, 1);
                break;
            case enums_1.Items.FOUNDATION:
                server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.MY_LAND, 1);
                break;
            case enums_1.Items.SHACK:
                server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.MY_HOME, 1);
                break;
        }
        return true;
        //#endregion
    }
    async start(client, server, recipeId, count) {
        for (let i = 0; i < count; i++) {
            if (!(await this.craftItem(server, client, recipeId, 1)))
                return;
        }
    }
}
exports.CraftManager = CraftManager;
