// h1z1-server/out/servers/ZoneServer2016/managers/decaymanager.js
"use strict";
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecayManager = void 0;
const enums_1 = require("../models/enums");
const lootableconstructionentity_1 = require("../entities/lootableconstructionentity");
const utils_1 = require("../../../utils/utils");
class DecayManager {
    /** Used for tracking the tick amount needed before decay damage occurs on the construction */
    constructionDamageTickCount = 0;
    /** Used for tracking the tick amount needed before decay damage occurs on the vehicle */
    vehicleDamageTickCount = 0; // used to run vehicle damaging once every x loops
    /** Timer used for determining the interval for decay ticks */
    runTimer;
    /** MANAGED BY CONFIGMANAGER - See defaultConfig.yaml for more information */
    decayTickInterval;
    constructionDamageTicks;
    ticksToFullDecay;
    worldFreeplaceDecayMultiplier;
    vehicleDamageTicks;
    vacantFoundationTicks;
    griefFoundationTimer;
    griefCheckSlotAmount;
    baseVehicleDamage;
    maxVehiclesPerArea;
    vehicleDamageRange;
    dailyRepairMaterials;
    clearTimers() {
        if (this.runTimer)
            clearTimeout(this.runTimer);
    }
    run(server) {
        this.contructionExpirationCheck(server);
        if (this.constructionDamageTickCount >= this.constructionDamageTicks) {
            this.contructionDecayDamage(server);
            this.constructionDamageTickCount = -1;
        }
        this.constructionDamageTickCount++;
        if (this.vehicleDamageTickCount >= this.vehicleDamageTicks) {
            this.vehicleDecayDamage(server);
            this.vehicleDamageTickCount = -1;
        }
        this.vehicleDamageTickCount++;
        this.runTimer = setTimeout(() => {
            this.run(server);
        }, this.decayTickInterval);
    }
    contructionExpirationCheck(server) {
        let destroyedGriefFoundations = 0;
        for (const a in server._constructionFoundations) {
            const foundation = server._constructionFoundations[a];
            if (foundation.itemDefinitionId == enums_1.Items.FOUNDATION ||
                foundation.itemDefinitionId == enums_1.Items.GROUND_TAMPER) {
                if (Date.now() - foundation.placementTime >=
                    this.griefFoundationTimer * 3600000) {
                    if (Object.keys(foundation.occupiedWallSlots).length <
                        this.griefCheckSlotAmount &&
                        Object.keys(foundation.occupiedShelterSlots).length == 0 &&
                        Object.keys(foundation.occupiedExpansionSlots).length == 0) {
                        for (const a in foundation.occupiedWallSlots) {
                            foundation.occupiedWallSlots[a].destroy(server);
                        }
                        // clear floating entities
                        for (const a in foundation.freeplaceEntities) {
                            foundation.freeplaceEntities[a].destroy(server);
                        }
                        foundation.destroy(server);
                        destroyedGriefFoundations++;
                    }
                }
            }
            if (foundation.itemDefinitionId != enums_1.Items.FOUNDATION &&
                foundation.itemDefinitionId != enums_1.Items.GROUND_TAMPER) {
                continue;
            }
            let expansionsEmpty = true;
            Object.values(foundation.occupiedExpansionSlots).forEach((exp) => {
                if (Object.keys(exp.occupiedWallSlots).length != 0 ||
                    Object.keys(exp.occupiedShelterSlots).length != 0 ||
                    Object.keys(exp.occupiedUpperWallSlots).length != 0) {
                    expansionsEmpty = false;
                }
            });
            if (!expansionsEmpty)
                continue;
            if (Object.keys(foundation.occupiedWallSlots).length == 0 &&
                Object.keys(foundation.occupiedShelterSlots).length == 0 &&
                Object.keys(foundation.occupiedUpperWallSlots).length == 0) {
                if (foundation.ticksWithoutObjects >= this.vacantFoundationTicks) {
                    for (const a in foundation.occupiedExpansionSlots) {
                        const expansion = foundation.occupiedExpansionSlots[a];
                        for (const a in expansion.occupiedRampSlots) {
                            expansion.occupiedRampSlots[a].destroy(server);
                        }
                        // clear floating shelters / other entities on expansion
                        for (const a in expansion.freeplaceEntities) {
                            expansion.freeplaceEntities[a].destroy(server);
                        }
                        expansion.destroy(server);
                    }
                    for (const a in foundation.occupiedRampSlots) {
                        foundation.occupiedRampSlots[a].destroy(server);
                    }
                    // clear floating shelters / other entities
                    for (const a in foundation.freeplaceEntities) {
                        foundation.freeplaceEntities[a].destroy(server);
                    }
                    Object.values(foundation.freeplaceEntities).forEach((entity) => {
                        entity.destroy(server);
                    });
                    foundation.destroy(server);
                }
                foundation.ticksWithoutObjects++;
            }
            else {
                foundation.ticksWithoutObjects = 0;
            }
        }
        if (destroyedGriefFoundations > 0) {
            console.log(`Destroyed ${destroyedGriefFoundations} grief foundations`);
        }
    }
    decayDamage(server, entity, freeplaceDecayMultiplier = 1) {
        if (entity.isDecayProtected) {
            entity.isDecayProtected = false;
            return;
        }
        entity.damage(server, {
            entity: "Server.DecayManager",
            damage: entity.maxHealth / (this.ticksToFullDecay / freeplaceDecayMultiplier)
        });
    }
    // uses repair box if one is detected on the base and it has the required materials
    useRepairBox(server, foundation) {
        for (const b in foundation.freeplaceEntities) {
            const freePlace = foundation.freeplaceEntities[b];
            if (freePlace.itemDefinitionId != enums_1.Items.REPAIR_BOX ||
                !(freePlace instanceof lootableconstructionentity_1.LootableConstructionEntity)) {
                continue;
            }
            const container = freePlace.getContainer();
            if (!container)
                continue;
            let hasMaterials = true;
            const itemsToRemove = [];
            this.dailyRepairMaterials.forEach((material) => {
                let materialPresent = false;
                for (const c in container.items) {
                    const item = container.items[c];
                    if (item.itemDefinitionId == material.itemDefinitionId &&
                        item.stackCount >= material.requiredCount) {
                        materialPresent = true;
                        itemsToRemove.push({
                            item: item,
                            count: material.requiredCount
                        });
                    }
                }
                if (!materialPresent)
                    hasMaterials = false;
            });
            if (!hasMaterials)
                continue;
            itemsToRemove.forEach((itemToRemove) => {
                server.removeContainerItem(freePlace, itemToRemove.item, container, itemToRemove.count);
            });
            server.constructionManager.fullyRepairFoundation(server, foundation);
            return;
        }
    }
    contructionDecayDamage(server) {
        for (const a in server._constructionFoundations) {
            const foundation = server._constructionFoundations[a];
            this.useRepairBox(server, foundation);
            if (foundation.itemDefinitionId != enums_1.Items.FOUNDATION &&
                foundation.itemDefinitionId != enums_1.Items.GROUND_TAMPER &&
                foundation.itemDefinitionId != enums_1.Items.FOUNDATION_EXPANSION) {
                this.decayDamage(server, foundation);
            }
        }
        for (const a in server._worldLootableConstruction) {
            this.decayDamage(server, server._worldLootableConstruction[a], this.worldFreeplaceDecayMultiplier);
        }
        for (const a in server._worldSimpleConstruction) {
            this.decayDamage(server, server._worldSimpleConstruction[a], this.worldFreeplaceDecayMultiplier);
        }
        for (const a in server._constructionSimple) {
            const simple = server._constructionSimple[a];
            if (simple.itemDefinitionId == enums_1.Items.FOUNDATION_RAMP ||
                simple.itemDefinitionId == enums_1.Items.FOUNDATION_STAIRS) {
                continue;
            }
            this.decayDamage(server, server._constructionSimple[a]);
        }
        for (const a in server._lootableConstruction) {
            this.decayDamage(server, server._lootableConstruction[a]);
        }
        for (const a in server._constructionDoors) {
            const door = server._constructionDoors[a];
            this.decayDamage(server, door);
        }
    }
    getCloseVehicles(server, vehicle) {
        const vehicles = [];
        for (const characterId in server._vehicles) {
            const v = server._vehicles[characterId];
            if (!vehicle)
                continue;
            if ((0, utils_1.getDistance)(vehicle.state.position, v.state.position) <=
                this.vehicleDamageRange) {
                vehicles.push(v.characterId);
            }
        }
        return vehicles;
    }
    vehicleDecayDamage(server) {
        for (const characterId in server._vehicles) {
            const vehicle = server._vehicles[characterId];
            if (!vehicle)
                continue;
            const closeVehicles = this.getCloseVehicles(server, vehicle);
            let damage = this.baseVehicleDamage;
            if (closeVehicles.length > this.maxVehiclesPerArea) {
                damage *= closeVehicles.length - this.maxVehiclesPerArea + 1;
            }
            vehicle.damage(server, {
                entity: "Server.DecayManager",
                damage: damage
            });
            server.updateResourceToAllWithSpawnedEntity(vehicle.characterId, vehicle._resources[enums_1.ResourceIds.CONDITION], enums_1.ResourceIds.CONDITION, enums_1.ResourceTypes.CONDITION, server._vehicles);
            if (vehicle.getHealth() > 0)
                continue;
            vehicle.destroy(server);
        }
    }
}
exports.DecayManager = DecayManager;
