// h1z1-server/out/servers/ZoneServer2016/managers/playtimemanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayTimeManager = void 0;
class PlayTimeManager {
    interval;
    server;
    updatePlaytime() {
        const chars = Object.values(this.server._characters);
        for (const char of chars) {
            char.playTime += 1;
        }
    }
    init(server) {
        this.server = server;
        this.start();
    }
    start() {
        // if somehow it was already running
        if (this.interval) {
            this.stop();
        }
        this.interval = setInterval(this.updatePlaytime.bind(this), 60_000);
    }
    stop() {
        clearInterval(this.interval);
    }
}
exports.PlayTimeManager = PlayTimeManager;
