// h1z1-server/out/servers/ZoneServer2016/managers/pluginmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginManager = exports.BasePlugin = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const child_process_1 = require("child_process");
const utils_1 = require("../../../utils/utils");
/**
 * Abstract class representing a base plugin.
 */
class BasePlugin {
    dir;
}
exports.BasePlugin = BasePlugin;
/**
 * Checks if a folder exists at the specified path.
 * @param path - The path of the folder to check.
 * @returns A boolean indicating whether the folder exists.
 */
function folderExists(path) {
    // shoutout chatGPT
    try {
        const stats = fs.statSync(path);
        return stats.isDirectory();
    }
    catch (err) {
        if (err.code === "ENOENT") {
            // Folder doesn't exist
            return false;
        }
        else {
            // Other error occurred
            throw err;
        }
    }
}
/**
 * Searches for a folder within a directory.
 * @param directory - The directory to search in.
 * @param folderName - The name of the folder to search for.
 * @returns The path of the found folder, or null if not found.
 */
function searchFolder(directory, folderName) {
    // shoutout chatGPT
    const files = fs.readdirSync(directory);
    for (const file of files) {
        // this only puts a single backslash for some reason, so manually replace all with a double
        const filePath = path.join(directory, file).replaceAll("\\", "\\\\");
        const stats = fs.statSync(filePath);
        if (stats.isDirectory()) {
            if (file === folderName) {
                return filePath; // Found the folder!
            }
            else {
                const subFolder = searchFolder(filePath, folderName);
                if (subFolder) {
                    return subFolder; // Found the folder in a sub-folder!
                }
            }
        }
    }
    return null; // Folder not found
}
/**
 * Recursively traverses a directory and replaces a search string with a replace string in all files.
 * @param directory - The directory to traverse.
 * @param searchString - The string to search for.
 * @param replaceString - The string to replace with.
 */
function traverseAndReplace(directory, searchString, replaceString) {
    // shoutout chatGPT
    if (process.platform === "win32") {
        // https://stackoverflow.com/a/70560464
        replaceString = path.resolve(replaceString).split(path.sep).join("/");
    }
    const files = fs.readdirSync(directory);
    files.forEach((file) => {
        const filePath = path.join(directory, file);
        const stats = fs.statSync(filePath);
        if (stats.isDirectory()) {
            traverseAndReplace(filePath, searchString, replaceString);
        }
        else if (stats.isFile() && file.endsWith(".js")) {
            replaceInFile(filePath, searchString, replaceString);
        }
    });
}
/**
 * Replaces a search string with a replace string in a specific file.
 * @param filePath - The path of the file.
 * @param searchString - The string to search for.
 * @param replaceString - The string to replace with.
 */
function replaceInFile(filePath, searchString, replaceString) {
    // shoutout chatGPT
    const content = fs.readFileSync(filePath, "utf8"), replacedContent = content.replace(new RegExp(searchString, "g"), replaceString);
    fs.writeFileSync(filePath, replacedContent, "utf8");
}
class PluginManager {
    plugins = [];
    get pluginCount() {
        return this.plugins.length;
    }
    pluginsDir = process.env.PLUGINS_DIR || path.join(process.cwd(), "plugins");
    moduleDir = process.env.PLUGIN_MODULES_DIR ||
        searchFolder(process.cwd(), "h1z1-server") ||
        process.cwd();
    /**
     * Checks if the plugins directory exists and creates it if it doesn't.
     */
    checkPluginsFolder() {
        if (!fs.existsSync(this.pluginsDir)) {
            fs.mkdirSync(this.pluginsDir);
            console.log(`[PluginManager] Created plugins folder: ${this.pluginsDir}`);
        }
    }
    /**
     * Retrieves the dependencies for a plugin.
     * @param projectPath - The path of the plugin.
     * @returns An array of dependencies.
     */
    getDependencies(projectPath) {
        // shoutout chatGPT
        const packageJsonPath = path.join(this.pluginsDir, projectPath, "package.json"), packageJsonContent = fs.readFileSync(packageJsonPath, "utf-8"), packageJson = JSON.parse(packageJsonContent), dependencies = Object.keys(packageJson.dependencies || {});
        return dependencies;
    }
    /**
     * Installs the dependencies for a plugin.
     * @param pluginPath - The path of the plugin.
     * @param dependencies - An array of dependencies to install.
     */
    async installDependencies(pluginPath) {
        return new Promise((resolve, reject) => {
            // shoutout chatGPT
            console.log("[PluginManager] Installing dependencies...");
            const installCommand = `npm install`;
            try {
                (0, child_process_1.execSync)(installCommand, {
                    cwd: path.join(this.pluginsDir, pluginPath),
                    stdio: "inherit"
                });
                console.log("[PluginManager] Dependencies installed successfully.");
                resolve();
            }
            catch (error) {
                console.error(error);
                console.error("[PluginManager] Failed to install dependencies.");
                reject();
                process.exit(1);
            }
        });
    }
    /**
     * Loads a plugin from the specified path.
     * @param pluginPath - The path of the plugin.
     * @returns A promise that resolves when the plugin is loaded.
     */
    async loadPlugin(pluginPath) {
        const runPath = path.join(this.pluginsDir, pluginPath, "plugin.js");
        if (!folderExists(path.join(this.pluginsDir, pluginPath, "node_modules")) ||
            process.env.PLUGIN_FORCE_BUILD) {
            // Install dependencies into the node_modules directory
            await this.installDependencies(pluginPath);
        }
        else {
            console.log(`[PluginManager] Dependencies loaded for ${pluginPath}`);
        }
        traverseAndReplace(path.join(this.pluginsDir, pluginPath), "C:/H1Z1/H1EmuServerFiles/h1z1-server-QuickStart-master/node_modules/h1z1-server", this.moduleDir);
        delete require.cache[require.resolve(runPath)];
        const module = await Promise.resolve(`${runPath}`).then(s => __importStar(require(s)));
        const plugin = new module.default();
        plugin.dir = path.join(this.pluginsDir, pluginPath);
        this.plugins.push(plugin);
        console.log(`[PluginManager] Loaded: ${plugin.name}`);
    }
    /**
     * Loads all plugins from the plugins directory.
     * @returns A promise that resolves when all plugins are loaded.
     */
    async loadPlugins() {
        this.checkPluginsFolder();
        const pluginFolders = fs.readdirSync(this.pluginsDir);
        for (const folder of pluginFolders) {
            if (!folderExists(path.join(this.pluginsDir, folder)))
                continue; // if not folder
            try {
                await this.loadPlugin(folder);
            }
            catch (e) {
                console.error(e);
                console.log(`[PluginManager] Plugin "${folder}" not loaded, make sure you've compiled the plugin before attempting to load it.`);
            }
        }
    }
    /**
     * Loads the configuration for a plugin.
     * @param server - The ZoneServer2016 instance.
     * @param plugin - The plugin instance.
     */
    async loadPluginConfig(server, plugin) {
        const defaultConfigPath = path.join(plugin.dir, "data", "defaultconfig.yaml"), defaultConfig = server.configManager.loadYaml(defaultConfigPath, false), fileName = `${plugin.name
            .toLowerCase()
            .replaceAll(" ", "-")}-config.yaml`, configPath = path.join(this.pluginsDir, fileName);
        if (!(0, utils_1.fileExists)(path.join(this.pluginsDir, fileName))) {
            console.log(`[PluginManager] Generated config file for ${plugin.name} at base plugins directory: ${fileName}`);
            await (0, utils_1.copyFile)(defaultConfigPath, configPath);
        }
        const configFile = server.configManager.loadYaml(configPath, false);
        const config = {
            // in case a config file is outdated, load missing values using default
            ...defaultConfig,
            ...configFile
        };
        plugin.loadConfig(config);
    }
    /**
     * Registers all commands defined by the plugin.
     * @param server - The ZoneServer2016 instance.
     * @param plugin - The plugin instance.
     */
    registerPluginCommands(server, plugin) {
        if (!plugin.commands)
            return;
        plugin.commands.forEach((command) => {
            this.registerCommand(plugin, server, command);
        });
    }
    /**
     * Initializes the plugins and loads their configurations.
     * @param server - The ZoneServer2016 instance.
     */
    async initializePlugins(server) {
        // Used in tests
        if (process.env.DISABLE_PLUGINS) {
            return;
        }
        if (!this.moduleDir) {
            console.error("[PluginManager] moduleDir is undefined!");
            console.log(`[PluginManager] No plugins loaded.`);
            return;
        }
        this.plugins = [];
        await this.loadPlugins();
        for (const plugin of this.plugins) {
            try {
                await this.loadPluginConfig(server, plugin);
                console.time(`Loading ${plugin.name} plugin`);
                await plugin.init(server);
                console.timeEnd(`Loading ${plugin.name} plugin`);
                this.registerPluginCommands(server, plugin);
                console.log(`[PluginManager] ${plugin.name} initialized!`);
            }
            catch (e) {
                console.error(e);
            }
        }
        if (this.plugins.length == 0) {
            console.log(`[PluginManager] No plugins loaded.`);
        }
    }
    //#region PLUGIN HELPERS
    /**
     * Hooks a method by overriding its default behavior.
     * @param plugin - The plugin instance ("this")
     * @param thisArg - The object on which the method is defined.
     * @param methodName - The name of the method to be hooked.
     * @param hook - A function that will be called during the hooking process.
     * @param options - Options for controlling the hooking behavior. Specifies whether to call the original method before/after the hook.
     * @returns
     */
    hookMethod(plugin, thisArg, methodName, hook, options) {
        const originalFunction = thisArg[methodName];
        if (!originalFunction) {
            console.log(`\n\n\n[PluginManager] Plugin ${plugin.name} tried to hook an invalid method!\n\n\n`);
            return;
        }
        thisArg[methodName] = function (...args) {
            if (options.callBefore)
                originalFunction.call(thisArg, ...args);
            const ret = hook(...args);
            if (ret !== false && options.callAfter)
                originalFunction.call(thisArg, ...args);
        };
    }
    /**
     * Registers a custom command to be used in-game.
     * @param plugin - The plugin instance ("this").
     * @param server - The ZoneServer2016 instance.
     * @param command - The command to register.
     * @deprecated Add your command to the commands array instead. Refer to documentation for more info.
     */
    // to be made private, keeping public for a bit in case plugins use legacy command adding
    registerCommand(plugin, server, command) {
        server.commandHandler.commands[(0, utils_1.flhash)(command.name.toUpperCase())] =
            command;
        console.log(`[PluginManager] Plugin ${plugin.name} registered a command: /${command.name}`);
    }
    /**
     * Lists and sends commands from a plugin to a player's console.
     *
     * This function iterates through the list of commands provided by a plugin
     * and sends each command to a client's console in a game server.
     *
     * @param server - The ZoneServer2016 instance.
     * @param client - The client requesting the command list.
     * @param plugin - The plugin instance ("this").
     */
    listCommands(server, client, plugin) {
        if (!plugin.commands)
            return;
        const commands = plugin.commands.map((command) => {
            return `/${command.name}: ${command.description}`;
        });
        // workaround for possible h1z1 console text limit?
        commands.forEach((command) => {
            server.sendConsoleText(client, `${command}`, true);
        });
    }
}
exports.PluginManager = PluginManager;
