// h1z1-server/out/servers/ZoneServer2016/managers/randomeventsmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.RandomEventsManager = void 0;
const node_crypto_1 = require("node:crypto");
const utils_1 = require("../../../utils/utils");
class RandomEventsManager {
    server;
    interval;
    // managed by config
    enabled = true;
    constructor(server) {
        this.server = server;
    }
    start() {
        if (this.enabled) {
            this.interval = setInterval(() => this.run(), 600_000);
        }
    }
    stop() {
        clearInterval(this.interval);
    }
    weightedRandom(items) {
        let random = Math.random() * items.reduce((sum, item) => sum + item.weight, 0);
        return items.find((item) => (random -= item.weight) < 0).value;
    }
    spawnRandomAirdrop() {
        const cellIndex = (0, node_crypto_1.randomInt)(100);
        const spg = this.server._spawnGrid[cellIndex];
        const rnd_index = (0, node_crypto_1.randomInt)(spg.spawnPoints.length);
        const pos = spg.spawnPoints[rnd_index];
        const airdropTypes = [
            { value: "Farmer", weight: 15 },
            { value: "Demolitioner", weight: 5 },
            { value: "Medic", weight: 25 },
            { value: "Builder", weight: 25 },
            { value: "Fighter", weight: 10 },
            { value: "Supplier", weight: 20 }
        ];
        const airdropType = this.weightedRandom(airdropTypes);
        this.server.airdropManager.spawnAirdrop(pos, airdropType);
        const cellName = (0, utils_1.getCellName)(cellIndex, 10);
        this.server.sendAlertToAll(`Random airdrop on ${cellName}`);
        if (!this.server._soloMode) {
            this.server._db.collection("random_aidrops_logs").insertOne({
                position: pos,
                serverId: this.server._worldId
            });
        }
    }
    run() {
        // Determine player count multiplier
        const playerCount = Object.keys(this.server._clients).length;
        let multiplier = 1;
        if (playerCount >= 100) {
            multiplier = 2.5;
        }
        else if (playerCount >= 60) {
            multiplier = 2;
        }
        else if (playerCount >= 30) {
            multiplier = 1.5;
        }
        // Use a larger range for more granularity
        const threshold = 10 * multiplier;
        if (this.server.airdropManager.allowedAirdropSpawn() &&
            (0, node_crypto_1.randomInt)(100) < threshold) {
            this.spawnRandomAirdrop();
        }
    }
}
exports.RandomEventsManager = RandomEventsManager;
