// h1z1-server/out/servers/ZoneServer2016/managers/rconmanager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RConManager = exports.RconMessageType = void 0;
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
const ws_1 = require("ws");
const node_events_1 = require("node:events");
const node_http_1 = require("node:http");
var RconMessageType;
(function (RconMessageType) {
    RconMessageType[RconMessageType["ExecCommand"] = 1] = "ExecCommand";
    RconMessageType[RconMessageType["DisplayUserMessage"] = 2] = "DisplayUserMessage";
    RconMessageType[RconMessageType["BroadcastLog"] = 3] = "BroadcastLog";
})(RconMessageType || (exports.RconMessageType = RconMessageType = {}));
class RConManager extends node_events_1.EventEmitter {
    server;
    wss;
    // Managed by config
    wssPort = 0;
    password = "";
    protocolVersion = 1;
    constructor() {
        super();
        this.server = (0, node_http_1.createServer)();
    }
    start() {
        if (process.env.FORCE_DISABLE_WS || !this.wssPort || !this.password) {
            console.log("RConManager disabled");
            return;
        }
        console.log("RConManager start");
        this.wss = new ws_1.WebSocketServer({ noServer: true });
        this.wss.on("connection", (ws) => {
            ws.on("message", (message) => {
                try {
                    const rconMessage = JSON.parse(message.toString());
                    if (this.protocolVersion === rconMessage.protocolVersion) {
                        this.emit("message", ws, rconMessage);
                    }
                    else {
                        console.error("Protocol missmatch");
                        ws.send("Protocol missmatch");
                    }
                }
                catch (e) {
                    console.error(e);
                }
            });
        });
        this.server.on("upgrade", (req, socket, head) => {
            socket.on("error", console.error);
            if (req.headers.authorization?.replace("Bearer ", "") !== this.password) {
                socket.write("HTTP/1.1 401 Unauthorized\r\n\r\n");
                socket.destroy();
                return;
            }
            this.wss.handleUpgrade(req, socket, head, (ws) => {
                this.wss.emit("connection", ws);
            });
        });
        this.server.listen(this.wssPort);
    }
    createMessage(type, payload) {
        const rconMessage = {
            type,
            payload,
            protocolVersion: this.protocolVersion
        };
        return JSON.stringify(rconMessage);
    }
    broadcastLog(message) {
        const clients = this.wss.clients;
        const rconMessage = this.createMessage(RconMessageType.BroadcastLog, message);
        clients.forEach((ws) => {
            ws.send(rconMessage);
        });
    }
    async stop() {
        // const close = promisify(this.wss.close);
        // await close();
    }
}
exports.RConManager = RConManager;
