// h1z1-server/out/servers/ZoneServer2016/managers/rewardmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.RewardManager = void 0;
const timers_1 = require("timers");
const enums_1 = require("../models/enums");
const utils_1 = require("../../../utils/utils");
class RewardManager {
    server;
    rewards;
    playTimerewards;
    timer;
    constructor(server) {
        this.server = server;
        if (server?.isBattleRoyale()) {
            this.rewards = [];
            this.playTimerewards = [];
            return;
        }
        this.rewards = [
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_MARAUDER,
                dropChances: 5
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_SHOWDOWN,
                dropChances: 5
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_INVITATIONAL,
                dropChances: 5
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_INFERNAL,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_ALPHA_LAUNCH,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_PREDATOR,
                dropChances: 5
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_EZW,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_RENEGADE,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_WASTELAND,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_RONIN,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_MERCENARY,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_WEARABLES,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_INVITATIONAL_2016,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_VIGILANTE,
                dropChances: 6
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_H1EMU,
                dropChances: 5
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_SWIRL,
                dropChances: 5
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_BREAKOUT,
                dropChances: 5
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_VICTORY,
                dropChances: 5
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_H1EMUEXCLUSIVE,
                dropChances: 0
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_INFERNAL_2025,
                dropChances: 0
            },
            {
                itemId: enums_1.AccountItems.REWARD_CRATE_FROST,
                dropChances: 0
            }
        ];
        this.playTimerewards = [
            {
                itemId: enums_1.AccountItems.MYSTERY_BAG_1,
                dropChances: 40
            },
            {
                itemId: enums_1.AccountItems.ELITE_BAG_HARDCORE,
                dropChances: 20
            },
            {
                itemId: enums_1.AccountItems.MYSTERY_BAG_V2,
                dropChances: 40
            }
        ];
    }
    startInterval() {
        this.timer = (0, timers_1.setInterval)(this.update.bind(this), 1_000);
    }
    start() {
        this.startInterval();
    }
    stop() {
        clearInterval(this.timer);
    }
    addRewardToPlayer(client, rewardId) {
        const item = this.server.generateAccountItem(rewardId);
        if (item) {
            this.server.lootAccountItem(this.server, client, item);
        }
        else {
            console.log("Server failed to generate reward account item");
        }
    }
    dropReward(client) {
        if (this.server?.isBattleRoyale())
            return;
        let random = Math.random() * 100;
        for (const reward of this.rewards) {
            if (reward.dropChances === 0) {
                continue;
            }
            random -= reward.dropChances;
            if (random <= 0) {
                const rewardId = reward.itemId;
                this.addRewardToPlayer(client, rewardId);
                break;
            }
        }
    }
    dropPlayTimeReward(client) {
        client.character.lastDropPlaytime = client.character.playTime;
        if ((0, utils_1.isHalloween)()) {
            const rewardId = enums_1.AccountItems.REWARD_CRATE_INFERNAL;
            this.addRewardToPlayer(client, rewardId);
        }
        else {
            let rewardId = enums_1.AccountItems.MYSTERY_BAG_1;
            let random = Math.random() * 100;
            for (const reward of this.playTimerewards) {
                if (reward.dropChances === 0) {
                    continue;
                }
                random -= reward.dropChances;
                if (random <= 0) {
                    rewardId = reward.itemId;
                    break;
                }
            }
            this.addRewardToPlayer(client, rewardId);
        }
    }
    update() {
        for (const clientKey in this.server._clients) {
            const client = this.server._clients[clientKey];
            if (client.character.playTime - client.character.lastDropPlaytime >
                ((0, utils_1.isHalloween)() ? 60 : 120) &&
                (0, utils_1.isPosInPoi)(client.character.state.position)) {
                this.dropPlayTimeReward(client);
            }
        }
    }
}
exports.RewardManager = RewardManager;
