// h1z1-server/out/servers/ZoneServer2016/managers/smeltingmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmeltingManager = void 0;
function getRewardId(itemDefinitionId) {
    switch (itemDefinitionId) {
        case enums_1.Items.DEW_COLLECTOR:
            return enums_1.Items.WATER_STAGNANT;
        case enums_1.Items.BEE_BOX:
            return enums_1.Items.HONEY;
    }
    return;
}
const enums_1 = require("../models/enums");
const collectingentity_1 = require("../classes/collectingentity");
const smeltingentity_1 = require("../classes/smeltingentity");
const Recipes_1 = require("../data/Recipes");
const promises_1 = require("timers/promises");
class SmeltingManager {
    /** HashMap of all SmeltingEntities,
     * uses CharacterId (string) for indexing
     */
    _smeltingEntities = {};
    /** HashMap of all CollectingEntities,
     * uses CharacterId (string) for indexing
     */
    _collectingEntities = {};
    /** The time (milliseconds) it takes for a CollectingEntity to fill water/honey - 5 min x 4 ticks = 20 mins */
    collectingTickTime = 300000;
    /** The time (milliseconds) at which the most recent qualified item was "burned" */
    lastBurnTime = 0;
    /** The timer for checking all smeltable entities  */
    checkSmeltablesTimer;
    /** The timer to check for honeycomb inside of all collectable entities - 5 min x 72 ticks = 6 hours for honeycomb */
    checkCollectorsTimer;
    /** MANAGED BY CONFIGMANAGER - See defaultConfig.yaml for more information */
    /** The time (milliseconds) it takes for a fuel entity to burn - 2 minutes seconds by default */
    burnTime;
    /** The time (milliseconds) it takes for a non-fuel entity to smelt - 7 seconds by default */
    smeltTime;
    clearTimers() {
        if (this.checkSmeltablesTimer) {
            clearTimeout(this.checkSmeltablesTimer);
        }
        if (this.checkCollectorsTimer) {
            clearTimeout(this.checkCollectorsTimer);
        }
    }
    checkSmeltables(server) {
        this.lastBurnTime = Date.now();
        for (const a in this._smeltingEntities) {
            const entity = this.getTrueEntity(server, this._smeltingEntities[a]);
            if (!entity) {
                delete this._smeltingEntities[a];
                continue;
            }
            const subEntity = entity.subEntity;
            if (!(subEntity instanceof smeltingentity_1.SmeltingEntity))
                continue;
            const container = entity.getContainer();
            if (!container) {
                delete this._smeltingEntities[a];
                continue;
            }
            const items = Object.values(container.items);
            //const dictionary = subEntity?.dictionary;
            if (items.length <= 0) {
                subEntity.isWorking = false;
                this.extinguish(server, entity /*dictionary*/);
                continue;
            }
            else {
                if (!this.checkFuel(server, entity)) {
                    this.extinguish(server, entity /*dictionary*/);
                    continue;
                }
                else if (!subEntity.isSmelting) {
                    subEntity.isSmelting = true;
                    this.smelt(server, entity);
                }
            }
            server.sendDataToAllWithSpawnedEntity(subEntity.dictionary, entity.characterId, "Character.PlayWorldCompositeEffect", {
                characterId: entity.characterId,
                effectId: entity.subEntity.workingEffect,
                position: entity.state.position,
                effectTime: Math.ceil(this.burnTime / 1000)
            });
        }
        this.checkSmeltablesTimer = setTimeout(() => {
            this.checkSmeltables(server);
        }, this.burnTime);
    }
    fillDewCollectors(server) {
        for (const a in this._collectingEntities) {
            const entity = this.getTrueEntity(server, this._collectingEntities[a]);
            if (!entity) {
                delete this._smeltingEntities[a];
                continue;
            }
            const subEntity = entity.subEntity;
            if (!(subEntity instanceof collectingentity_1.CollectingEntity)) {
                delete this._smeltingEntities[a];
                continue;
            }
            const container = entity.getContainer();
            if (!container) {
                delete this._smeltingEntities[a];
                continue;
            }
            if (entity.itemDefinitionId === enums_1.Items.DEW_COLLECTOR) {
                for (const a in container.items) {
                    const item = container.items[a];
                    if (item.itemDefinitionId != enums_1.Items.WATER_EMPTY)
                        continue;
                    if (!server.removeContainerItem(entity, item, container, 1)) {
                        return;
                    }
                    const reward = getRewardId(entity.itemDefinitionId);
                    if (reward) {
                        entity.lootContainerItem(server, server.generateItem(reward), 1, false);
                    }
                }
            }
        }
    }
    checkCollectors(server) {
        for (const a in this._collectingEntities) {
            const entity = this.getTrueEntity(server, this._collectingEntities[a]);
            if (!entity) {
                delete this._smeltingEntities[a];
                continue;
            }
            const subEntity = entity.subEntity;
            if (!(subEntity instanceof collectingentity_1.CollectingEntity)) {
                delete this._smeltingEntities[a];
                continue;
            }
            const container = entity.getContainer();
            if (!container) {
                delete this._smeltingEntities[a];
                continue;
            }
            switch (entity.itemDefinitionId) {
                case enums_1.Items.BEE_BOX:
                    this.checkCollector(server, entity, subEntity, container);
                    this.checkHoneycomb(server, entity, subEntity, container);
                    break;
                case enums_1.Items.DEW_COLLECTOR:
                    this.checkCollector(server, entity, subEntity, container);
                    break;
                case enums_1.Items.ANIMAL_TRAP:
                    this.checkAnimalTrap(server, entity, subEntity, container);
                    break;
            }
        }
        this.checkCollectorsTimer = setTimeout(() => {
            this.checkCollectors(server);
        }, this.collectingTickTime);
    }
    getTrueEntity(server, characterId) {
        return (server._lootableConstruction[characterId] ||
            server._worldLootableConstruction[characterId] ||
            undefined);
    }
    extinguish(server, entity
    //dictionary: any
    ) {
        delete this._smeltingEntities[entity.characterId];
        //this.sendWorkingEffect(server, 0, entity.characterId, dictionary);
        entity.subEntity.isWorking = false;
    }
    /*private sendWorkingEffect(
      server: ZoneServer2016,
      effect: number,
      characterId: string,
      dictionary: any
    ) {
      server.sendDataToAllWithSpawnedEntity(
        dictionary,
        characterId,
        "Command.PlayDialogEffect",
        {
          characterId: characterId,
          effectId: effect,
        }
      );
    }*/
    checkFuel(server, entity) {
        const container = entity.getContainer();
        for (const a in container.items) {
            const item = container.items[a];
            if (entity.subEntity.allowedFuel.includes(item.itemDefinitionId)) {
                server.removeContainerItem(entity, item, entity.getContainer(), 1);
                if (item.itemDefinitionId == enums_1.Items.WOOD_LOG) {
                    // give charcoal if wood log was burned
                    entity.lootContainerItem(server, server.generateItem(enums_1.Items.CHARCOAL), 1, false);
                }
                return true;
            }
        }
        return false;
    }
    async smelt(server, entity) {
        await promises_1.scheduler.wait(this.smeltTime, {});
        if (!entity.subEntity?.isWorking) {
            entity.subEntity.isSmelting = false;
            return;
        }
        const container = entity.getContainer();
        if (!container)
            return;
        if (!entity.subEntity)
            return;
        for (const a in Recipes_1.smeltingData) {
            const recipe = Recipes_1.smeltingData[a];
            if (recipe.filterId == entity.subEntity.filterId) {
                const fulfilledComponents = [];
                const itemsToRemove = [];
                for (const a in recipe.components) {
                    const component = recipe.components[a];
                    let requiredAmount = component.requiredAmount;
                    for (const a in container.items) {
                        const item = container.items[a];
                        if (!fulfilledComponents.includes(component)) {
                            if (component.itemDefinitionId == item.itemDefinitionId) {
                                if (requiredAmount > item.stackCount) {
                                    requiredAmount -= item.stackCount;
                                    itemsToRemove.push({ item: item, count: item.stackCount });
                                }
                                else {
                                    fulfilledComponents.push(component);
                                    itemsToRemove.push({ item: item, count: requiredAmount });
                                }
                                if (fulfilledComponents.length == recipe.components.length) {
                                    itemsToRemove.forEach((item) => {
                                        server.removeContainerItem(entity, item.item, entity.getContainer(), item.count);
                                    });
                                    entity.lootContainerItem(server, server.generateItem(recipe.rewardId), 1, false);
                                    this.smelt(server, entity);
                                    return;
                                }
                            }
                        }
                    }
                }
            }
        }
        this.smelt(server, entity);
        return;
    }
    checkCollector(server, entity, subEntity, container) {
        for (const a in container.items) {
            const item = container.items[a];
            // check if the current item is an empty water bottle and that the container is either empty
            // or currently not containing honeycomb, in that case then exit the scope of the for loop
            if (entity.itemDefinitionId == enums_1.Items.BEE_BOX) {
                if (item.itemDefinitionId == enums_1.Items.WATER_EMPTY &&
                    (Object.keys(container.items).length == 1 ||
                        !Object.values(container.items).some((item) => item.itemDefinitionId == enums_1.Items.HONEYCOMB)))
                    continue;
            }
            if (item.itemDefinitionId != enums_1.Items.WATER_EMPTY)
                continue;
            if (subEntity.currentTicks >= subEntity.requiredTicks) {
                subEntity.currentTicks = 0;
                if (!server.removeContainerItem(entity, item, container, 1)) {
                    return;
                }
                const reward = getRewardId(entity.itemDefinitionId);
                if (reward) {
                    if (reward == enums_1.Items.HONEY) {
                        if (item.stackCount >= 10)
                            return;
                        const honeycombItem = entity.getItemById(enums_1.Items.HONEYCOMB);
                        if (!honeycombItem ||
                            !server.removeContainerItem(entity, honeycombItem, container, 1)) {
                            return;
                        }
                        entity.lootContainerItem(server, server.generateItem(enums_1.Items.WAX), 1, false);
                        server.sendDataToAllWithSpawnedEntity(subEntity.dictionary, entity.characterId, "Command.PlayDialogEffect", {
                            characterId: entity.characterId,
                            effectId: subEntity.workingEffect
                        });
                    }
                    entity.lootContainerItem(server, server.generateItem(reward), 1, false);
                }
                return;
            }
            subEntity.currentTicks++;
        }
    }
    checkHoneycomb(server, entity, subEntity, container) {
        for (const a in container.items) {
            const item = container.items[a];
            if (item.itemDefinitionId == enums_1.Items.HONEYCOMB) {
                if (item.stackCount >= 5)
                    return;
                server.sendDataToAllWithSpawnedEntity(subEntity.dictionary, entity.characterId, "Character.PlayWorldCompositeEffect", {
                    characterId: entity.characterId,
                    effectId: entity.subEntity.workingEffect,
                    position: entity.state.position,
                    effectTime: Math.ceil(this.collectingTickTime / 1000)
                });
            }
        }
        if (subEntity.currentHoneycombTicks >= subEntity.requiredHoneycombTicks) {
            subEntity.currentHoneycombTicks = 0;
            server.sendDataToAllWithSpawnedEntity(subEntity.dictionary, entity.characterId, "Character.PlayWorldCompositeEffect", {
                characterId: entity.characterId,
                effectId: entity.subEntity.workingEffect,
                position: entity.state.position,
                effectTime: Math.ceil(this.collectingTickTime / 1000)
            });
            entity.lootContainerItem(server, server.generateItem(enums_1.Items.HONEYCOMB), 1, false);
            return;
        }
        subEntity.currentHoneycombTicks++;
    }
    checkAnimalTrap(server, entity, subEntity, container) {
        const checkEmpty = Object.keys(container.items).length === 0;
        if (checkEmpty) {
            if (subEntity.wasUsed) {
                entity.destroy(server);
                return;
            }
            if (subEntity.currentTicks >= subEntity.requiredTicks) {
                subEntity.currentTicks = 0;
                entity.lootContainerItem(server, server.generateItem(enums_1.Items.MEAT_RABBIT), 1, false);
                entity.actorModelId = 35;
                server.sendDataToAllWithSpawnedEntity(subEntity.dictionary, entity.characterId, "Character.ReplaceBaseModel", {
                    characterId: entity.characterId,
                    modelId: entity.actorModelId
                });
                subEntity.wasUsed = true;
                return;
            }
            subEntity.currentTicks++;
        }
    }
}
exports.SmeltingManager = SmeltingManager;
