// h1z1-server/out/servers/ZoneServer2016/managers/speedtreemanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpeedTreeManager = void 0;
const utils_1 = require("../../../utils/utils");
const enums_1 = require("../models/enums");
const challengemanager_1 = require("./challengemanager");
class SpeedTreeManager {
    /** HashMap of destroyed trees,
     * uses ObjectId (number) for indexing
     */
    _speedTreesDestroyed = {};
    /** The amount of spawned trees in the world */
    _speedTreesCounter = {};
    /** HashMap of all spawned trees in the world,
     * uses ObjectId (number) for indexing
     */
    _speedTreesList = new Map();
    /** MANAGED BY CONFIGMANAGER - See defaultConfig.yaml for more information */
    minBlackberryHarvest;
    maxBlackberryHarvest;
    branchHarvestChance;
    minStickHarvest;
    maxStickHarvest;
    treeRespawnTimeMS;
    minWoodLogHarvest;
    maxWoodLogHarvest;
    minTreeHits;
    maxTreeHits;
    initiateList() {
        const Z1_speedTrees = (0, utils_1.loadJson)(__dirname + "/../../../../data/2016/zoneData/Z1_speedTrees.json");
        Z1_speedTrees.forEach((tree) => {
            this._speedTreesList.set(tree.uniqueId, {
                objectId: tree.uniqueId,
                position: tree.position
            });
        });
    }
    customize(DTOArray) {
        for (const object in this._speedTreesDestroyed) {
            const DTO = this._speedTreesDestroyed[object];
            const DTOinstance = {
                objectId: DTO.objectId,
                replacementModel: DTO.modelName.concat(".Stump")
            };
            DTOArray.push(DTOinstance);
        }
    }
    use(server, client, objectId, treeId, name) {
        const speedtreeDestroyed = this._speedTreesDestroyed[objectId];
        let destroy = false;
        let count = 1;
        if (speedtreeDestroyed)
            return;
        let itemDefId = 0;
        switch (treeId) {
            case enums_1.TreeIds.BLACKBERRY:
                server.startInteractionTimer(client, 0, 0, 9);
                itemDefId = enums_1.Items.BLACKBERRY;
                if (Math.random() <= this.branchHarvestChance) {
                    client.character.lootItem(server, server.generateItem(enums_1.Items.WEAPON_BRANCH));
                }
                server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.DAWN_ITS_TASTY, 1);
                destroy = true;
                count = (0, utils_1.randomIntFromInterval)(this.minBlackberryHarvest, this.maxBlackberryHarvest);
                break;
            case enums_1.TreeIds.DEVILCLUB:
            case enums_1.TreeIds.VINEMAPLE:
                server.startInteractionTimer(client, 0, 0, 9);
                itemDefId = enums_1.Items.WOOD_STICK;
                destroy = true;
                count = (0, utils_1.randomIntFromInterval)(this.minStickHarvest, this.maxStickHarvest);
                break;
            case enums_1.TreeIds.REDMAPLE:
            case enums_1.TreeIds.WESTERNCEDAR:
            case enums_1.TreeIds.GREENMAPLE:
            case enums_1.TreeIds.GREENMAPLEDEAD:
            case enums_1.TreeIds.WESTERNCEDARSAPLING:
            case enums_1.TreeIds.SAPLINGMAPLE:
            case enums_1.TreeIds.WHITEBIRCH:
            case enums_1.TreeIds.REDCEDAR:
            case enums_1.TreeIds.PAPERBIRCH:
            case enums_1.TreeIds.OREGONOAK:
                const wep = client.character.getEquippedWeapon();
                if (!wep)
                    return;
                const durabilityDamage = server.getDurabilityDamage(wep.itemDefinitionId);
                switch (wep.itemDefinitionId) {
                    case enums_1.Items.WEAPON_HATCHET:
                    case enums_1.Items.WEAPON_HATCHET_MAKESHIFT:
                    case enums_1.Items.WEAPON_AXE_FIRE:
                    case enums_1.Items.WEAPON_AXE_WOOD:
                        break;
                    case enums_1.Items.WEAPON_FISTS:
                    case enums_1.Items.WEAPON_FLASHLIGHT:
                    case enums_1.Items.WEAPON_BINOCULARS:
                        return;
                    default:
                        server.sendAlert(client, "This tool is not sharp enough for this!");
                        return;
                }
                server.damageItem(client.character, wep, durabilityDamage);
                if (!this._speedTreesCounter[objectId]) {
                    this._speedTreesCounter[objectId] = {
                        hitPoints: (0, utils_1.randomIntFromInterval)(this.minTreeHits - 1, this.maxTreeHits - 1)
                    }; // add a new tree key with random level of hitpoints
                }
                if (this._speedTreesCounter[objectId].hitPoints-- == 0) {
                    server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.TREE_HATER, 1);
                    destroy = true;
                    delete this._speedTreesCounter[objectId]; // If out of health destroy tree and delete its key
                    itemDefId = enums_1.Items.WOOD_LOG;
                    count = (0, utils_1.randomIntFromInterval)(this.minWoodLogHarvest, this.maxWoodLogHarvest);
                    server.lootCrateWithChance(client, 2);
                }
                break;
            default: // boulders (do nothing);
                return;
        }
        if (itemDefId) {
            client.character.lootContainerItem(server, server.generateItem(itemDefId, count));
        }
        if (destroy) {
            this.destroy(server, objectId, name);
        }
    }
    destroy(server, objectId, name) {
        server.sendDataToAll("DtoStateChange", {
            objectId: objectId,
            modelName: name.concat(".Stump"),
            effectId: 0,
            unk3: 0,
            unk4: true
        });
        this._speedTreesDestroyed[objectId] = {
            objectId: objectId,
            modelName: name
        };
        setTimeout(() => {
            server.sendDataToAll("DtoStateChange", {
                objectId: objectId,
                modelName: this._speedTreesDestroyed[objectId].modelName,
                effectId: 0,
                unk3: 0,
                unk4: true
            });
            delete this._speedTreesDestroyed[objectId];
        }, this.treeRespawnTimeMS);
    }
}
exports.SpeedTreeManager = SpeedTreeManager;
