// h1z1-server/out/servers/ZoneServer2016/managers/tasksmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskManager = void 0;
class ScheduledTask {
    time;
    cb;
    lastRun = -1;
    constructor(time, cb) {
        this.time = time;
        this.cb = cb;
    }
}
class TaskManager {
    queue = [];
    priorityQueue = [];
    scheduleQueue = [];
    get taskCount() {
        return this.queue.length + this.priorityQueue.length;
    }
    register(cb, prioritize = false) {
        if (prioritize) {
            this.priorityQueue.push(cb);
        }
        else {
            this.queue.push(cb);
        }
    }
    checkElapsed(startTime, allowedTime) {
        return process.hrtime.bigint() - startTime < allowedTime;
    }
    process(time_ms) {
        const startTime = process.hrtime.bigint();
        const allowedTime = BigInt(time_ms) * 1000000n;
        this.process_schedule();
        if (!this.checkElapsed(startTime, allowedTime)) {
            console.warn("[TaskManager] didn't had time to process tasks after schedule");
            return;
        }
        for (let index = 0; index < this.priorityQueue.length; index++) {
            if (!this.checkElapsed(startTime, allowedTime)) {
                console.warn("[TaskManager] didn't had time to process all tasks, stopped in priority queue");
                break;
            }
            const element = this.priorityQueue.shift();
            element();
        }
        for (let index = 0; index < this.queue.length; index++) {
            if (!this.checkElapsed(startTime, allowedTime)) {
                console.warn("[TaskManager] didn't had time to process all tasks");
                break;
            }
            const element = this.queue.shift();
            element();
        }
    }
    register_schedule(cb, time) {
        this.scheduleQueue.push(new ScheduledTask(time, cb));
    }
    process_schedule() {
        const now = Date.now();
        for (let index = 0; index < this.scheduleQueue.length; index++) {
            const schedule = this.scheduleQueue[index];
            if (now - schedule.lastRun > schedule.time) {
                schedule.cb();
                schedule.lastRun = now;
            }
        }
    }
}
exports.TaskManager = TaskManager;
