// h1z1-server/out/servers/ZoneServer2016/managers/voicechatmanager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VoiceChatManager = void 0;
class VoiceChatManager {
    /* MANAGED BY CONFIGMANAGER */
    useVoiceChatV2;
    joinVoiceChatOnConnect;
    serverAccessToken;
    serverAddress;
    sendVoiceChatError(server, client, error) {
        server.sendChatText(client, `[Voicechat Error] ${error}`);
    }
    sendVoiceChatState(server, client, instance = 1) {
        server.sendData(client, "H1emu.VoiceState", {
            message: JSON.stringify({
                name: client.character.name,
                world: server._worldId,
                instance: instance // somebody else can implement radio logic using this. It's implemented serverside already
                // +10 to make it a radio channel, so radio chan 1 is instance: 11, chan 2 instance 12 etc etc
            })
        });
        server.sendData(client, "UpdateWeatherData", server.weatherManager.weather);
    }
    handleVoiceChatInit(server, client) {
        if (!this.useVoiceChatV2) {
            server.sendChatText(client, `voicechat is disabled in the configuration`);
        }
        server.sendChatText(client, `connecting to voice chat`);
        if (!client.isInVoiceChat) {
            server.sendData(client, "H1emu.VoiceInit", {});
            setTimeout(() => {
                this.sendVoiceChatState(server, client);
                client.voiceChatTimer = setInterval(() => {
                    this.sendVoiceChatState(server, client);
                }, 20000);
                client.isInVoiceChat = true;
            }, 4000);
        }
        else {
            server.sendChatText(client, `already initialized, please run /vc leave before retrying`);
        }
    }
    handleVoiceChatDisconnect(server, client) {
        server.sendChatText(client, `disconnecting from voicechat`);
        client.isInVoiceChat = false;
        //client.voiceChatTimer = null;
        clearInterval(client.voiceChatTimer);
    }
    handleVoiceChatCommand(server, client, args) {
        if (!args[0]) {
            server.sendChatText(client, "Missing command, valid commands are: status, join, leave");
            return;
        }
        if (args[0] == "leave" && !client.isInVoiceChat) {
            server.sendChatText(client, "You have already left.");
            return;
        }
        switch (args[0]) {
            case "status":
                server.sendChatText(client, `isInVoiceChat: ${client.isInVoiceChat ? "allegedly" : "no"}`);
                break;
            case "join":
                this.handleVoiceChatInit(server, client);
                break;
            case "leave":
                this.handleVoiceChatDisconnect(server, client);
                break;
            default:
                server.sendChatText(client, "Unknown command, valid commands are: status, join, leave");
                break;
        }
    }
    handleVoiceChatAdminCommand(server, client, args) {
        if (!args[0]) {
            server.sendChatText(client, "Missing command, valid commands are: status, list, mute, unmute");
            return;
        }
        if (args[0] == "leave" && !client.isInVoiceChat) {
            server.sendChatText(client, "You have already left.");
            return;
        }
        switch (args[0]) {
            case "status":
                server.sendChatText(client, `not implemented yet`);
                break;
            case "list":
                server.sendChatText(client, `not implemented yet`);
                break;
            case "mute":
                server.sendChatText(client, `not implemented yet`);
                break;
            case "unmute":
                server.sendChatText(client, `not implemented yet`);
                break;
            default:
                server.sendChatText(client, "Unknown command, valid commands are: status, list, mute, unmute");
                break;
        }
    }
}
exports.VoiceChatManager = VoiceChatManager;
