// h1z1-server/out/servers/ZoneServer2016/managers/worldobjectmanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorldObjectManager = void 0;
exports.getRandomSkin = getRandomSkin;
exports.getRandomItem = getRandomItem;
const Z1_doors = require("../../../../data/2016/zoneData/Z1_doors.json");
const Z1_items = require("../../../../data/2016/zoneData/Z1_items.json");
const Z1_vehicles = require("../../../../data/2016/zoneData/Z1_vehicleLocations.json");
const Z1_npcs = require("../../../../data/2016/zoneData/Z1_npcs.json");
const Z1_lootableProps = require("../../../../data/2016/zoneData/Z1_lootableProps.json");
const Z1_taskProps = require("../../../../data/2016/zoneData/Z1_taskProps.json");
const Z1_crates = require("../../../../data/2016/zoneData/Z1_crates.json");
const Z1_destroyables = require("../../../../data/2016/zoneData/Z1_destroyables.json");
const models = require("../../../../data/2016/dataSources/Models.json");
// const bannedZombieModels = require("../../../../data/2016/sampleData/bannedZombiesModels.json");
const utils_1 = require("../../../utils/utils");
const enums_1 = require("../models/enums");
const vehicle_1 = require("../entities/vehicle");
const itemobject_1 = require("../entities/itemobject");
const doorentity_1 = require("../entities/doorentity");
const explosiveentity_1 = require("../entities/explosiveentity");
const lootspawns_1 = require("../data/lootspawns");
const lootbag_1 = require("../entities/lootbag");
const lootableprop_1 = require("../entities/lootableprop");
const taskprop_1 = require("../entities/taskprop");
const crate_1 = require("../entities/crate");
const destroyable_1 = require("../entities/destroyable");
const watersource_1 = require("../entities/watersource");
const treasurechest_1 = require("../entities/treasurechest");
const npc_1 = require("../entities/npc");
//import { EntityType } from "h1emu-ai";
const promises_1 = require("node:timers/promises");
const debug = require("debug")("ZoneServer");
function getRandomSkin(itemDefinitionId) {
    let itemDefId = 0;
    let arr = [];
    switch (itemDefinitionId) {
        case enums_1.Items.CONVEYS_BLUE:
            arr = Object.keys(enums_1.DefaultSkinsConveys);
            break;
        case enums_1.Items.BACKPACK_BLUE_ORANGE:
            arr = Object.keys(enums_1.DefaultSkinsBackpack);
            break;
        case enums_1.Items.HELMET_MOTORCYCLE:
            arr = Object.keys(enums_1.DefaultSkinsMotorHelmet);
            break;
        case enums_1.Items.ZEDS_WHITE:
            arr = Object.keys(enums_1.DefaultSkinsZeds);
            break;
        case enums_1.Items.GATORS_RED:
            arr = Object.keys(enums_1.DefaultSkinsGators);
            break;
        case enums_1.Items.BOOTS_GRAY_BLUE:
            arr = Object.keys(enums_1.DefaultSkinsBoots);
            break;
        default:
            return itemDefinitionId;
    }
    itemDefId = Number(arr[Math.floor((Math.random() * arr.length) / 2)]);
    return itemDefId;
}
function getRandomItem(items) {
    const totalWeight = items.reduce((total, item) => total + item.weight, 0), randomWeight = Math.random() * totalWeight;
    let currentWeight = 0;
    for (let i = 0; i < items.length; i++) {
        currentWeight += items[i].weight;
        if (currentWeight > randomWeight) {
            return items[i];
        }
    }
}
class WorldObjectManager {
    /** HashMap of all spawned NPCs in the world - uses spawnerId (number) for indexing */
    spawnedNpcs = {};
    /** HashMap of all spawned objects in the world - uses spawnerId (number) for indexing */
    spawnedLootObjects = {};
    /** Global respawn timers */
    _lastLootRespawnTime = 0;
    _lastVehicleRespawnTime = 0;
    _lastNpcRespawnTime = 0;
    _lastWaterSourceReplenishTime = 0;
    /** MANAGED BY CONFIGMANAGER - See defaultConfig.yaml for more information */
    vehicleSpawnCap;
    lootRespawnTimer;
    vehicleRespawnTimer;
    npcRespawnTimer;
    hasCustomLootRespawnTime;
    itemDespawnTimer;
    lootDespawnTimer;
    deadNpcDespawnTimer;
    lootbagDespawnTimer;
    vehicleSpawnRadius;
    npcSpawnRadius;
    chanceNpc;
    chanceScreamer;
    chanceWornLetter;
    waterSourceReplenishTimer;
    waterSourceRefillAmount;
    zombieSlots = [
        enums_1.EquipSlots.HEAD,
        enums_1.EquipSlots.CHEST,
        enums_1.EquipSlots.LEGS,
        enums_1.EquipSlots.HANDS,
        enums_1.EquipSlots.FEET,
        enums_1.EquipSlots.HAIR
    ];
    static itemSpawnersChances = {};
    getItemRespawnTimer(server) {
        if (this.hasCustomLootRespawnTime)
            return;
        const playerCount = utils_1._.size(server._characters);
        if (playerCount >= 75) {
            this.lootRespawnTimer = 600_000; // 10 min
        }
        else if (playerCount >= 50) {
            this.lootRespawnTimer = 900_000; // 15 min
        }
        else if (playerCount >= 25) {
            this.lootRespawnTimer = 1_200_000; // 20 min
        }
        else if (playerCount >= 1) {
            this.lootRespawnTimer = 1_500_000; // 25 min
        }
        else {
            this.lootRespawnTimer = 1_500_000; // 25 min
        }
    }
    async run(server) {
        debug("WOM::Run");
        if (server.isSurvival()) {
            this.getItemRespawnTimer(server);
            if (this._lastLootRespawnTime + this.lootRespawnTimer <= Date.now()) {
                this.refillScrapInChunks(server);
                this.createLoot(server);
                this.createContainerLoot(server);
                this._lastLootRespawnTime = Date.now();
                server.divideLargeCells(700);
            }
            if (this._lastNpcRespawnTime + this.npcRespawnTimer <= Date.now()) {
                this.createNpcs(server);
                this._lastNpcRespawnTime = Date.now();
            }
            if (this._lastVehicleRespawnTime + this.vehicleRespawnTimer <=
                Date.now()) {
                this.createVehicles(server);
                this._lastVehicleRespawnTime = Date.now();
            }
            if (this._lastWaterSourceReplenishTime + this.waterSourceReplenishTimer <=
                Date.now()) {
                this.replenishWaterSources(server);
                this._lastWaterSourceReplenishTime = Date.now();
            }
            this.updateQuestContainers(server);
        }
        if (server.isSurvival()) {
            this.despawnEntities(server);
        }
    }
    async npcDespawner(server) {
        let counter = 0;
        for (const characterId in server._npcs) {
            if (counter > 30) {
                counter = 0;
                await promises_1.scheduler.wait(30);
            }
            counter++;
            const npc = server._npcs[characterId];
            // dead npc despawner
            if (npc &&
                npc.flags.knockedOut &&
                Date.now() - npc.deathTime >= this.deadNpcDespawnTimer) {
                server.deleteEntity(npc.characterId, server._npcs);
            }
        }
    }
    lootbagDespawner(server) {
        for (const characterId in server._lootbags) {
            // lootbag despawner
            const lootbag = server._lootbags[characterId];
            if (Date.now() - lootbag.creationTime >= this.lootbagDespawnTimer) {
                server.deleteEntity(lootbag.characterId, server._lootbags);
            }
        }
    }
    async itemDespawner(server) {
        let counter = 0;
        for (const characterId in server._spawnedItems) {
            if (counter > 100) {
                counter = 0;
                await promises_1.scheduler.wait(30);
            }
            counter++;
            const itemObject = server._spawnedItems[characterId];
            if (!itemObject)
                return;
            // dropped item despawner
            const despawnTime = itemObject.spawnerId == -1
                ? this.itemDespawnTimer
                : this.lootDespawnTimer;
            if (Date.now() - itemObject.creationTime >= despawnTime) {
                server.deleteEntity(itemObject.characterId, server._spawnedItems);
                switch (itemObject.item.itemDefinitionId) {
                    case enums_1.Items.FUEL_ETHANOL:
                    case enums_1.Items.FUEL_BIOFUEL:
                        server.deleteEntity(itemObject.characterId, server._explosives);
                        break;
                }
                if (itemObject.spawnerId != -1)
                    delete this.spawnedLootObjects[itemObject.spawnerId];
                server.sendCompositeEffectToAllWithSpawnedEntity(server._spawnedItems, itemObject, server.getItemDefinition(itemObject.item.itemDefinitionId)
                    ?.PICKUP_EFFECT ?? 5151);
            }
        }
    }
    despawnEntities(server) {
        this.npcDespawner(server);
        this.lootbagDespawner(server);
        this.itemDespawner(server);
    }
    createNpc(server, modelId, position, rotation, spawnerId = 0) {
        const characterId = (0, utils_1.generateRandomGuid)();
        const npc = new npc_1.Npc(characterId, server.getTransientId(characterId), modelId, position, rotation, server, spawnerId);
        // doesn't work anymore
        // this.equipRandomSkins(server, npc, this.zombieSlots, bannedZombieModels);
        server._npcs[characterId] = npc;
        if (spawnerId)
            this.spawnedNpcs[spawnerId] = characterId;
        return npc;
    }
    createLootEntity(server, item, position, rotation, itemSpawnerId = -1) {
        if (!item) {
            debug(`[ERROR] Tried to createLootEntity with invalid item object`);
            return;
        }
        const itemDef = server.getItemDefinition(item.itemDefinitionId);
        if (!itemDef) {
            debug(`[ERROR] Tried to createLootEntity for invalid itemDefId: ${item.itemDefinitionId}`);
            return;
        }
        const characterId = (0, utils_1.generateRandomGuid)(), modelId = itemDef.WORLD_MODEL_ID || 9, lootObj = new itemobject_1.ItemObject(characterId, server.getTransientId(characterId), modelId, position, rotation, server, itemSpawnerId || 0, item);
        lootObj.isWorldItem = true;
        server._spawnedItems[characterId] = lootObj;
        server._spawnedItems[characterId].nameId = itemDef.NAME_ID;
        switch (item.itemDefinitionId) {
            case enums_1.Items.FUEL_ETHANOL:
            case enums_1.Items.FUEL_BIOFUEL:
                lootObj.flags.projectileCollision = 1;
                server._explosives[characterId] = new explosiveentity_1.ExplosiveEntity(characterId, server.getTransientId(characterId), modelId, position, rotation, server, item.itemDefinitionId);
                break;
        }
        if (itemSpawnerId)
            this.spawnedLootObjects[itemSpawnerId] = characterId;
        lootObj.creationTime = Date.now();
        return lootObj;
    }
    createLootbag(server, entity) {
        const characterId = (0, utils_1.generateRandomGuid)(), isCharacter = !!server._characters[entity.characterId], items = entity.getDeathItems(server);
        if (!utils_1._.size(items))
            return; // don't spawn lootbag if inventory is empty
        const pos = entity.state.position, lootbag = new lootbag_1.Lootbag(characterId, server.getTransientId(characterId), isCharacter ? enums_1.ModelIds.LOOT_BAG_BLOODY : enums_1.ModelIds.LOOT_BAG_CLEAN, new Float32Array([pos[0], pos[1] + 0.1, pos[2]]), new Float32Array([0, 0, 0, 0]), server);
        const container = lootbag.getContainer();
        if (container) {
            container.items = items;
        }
        server._lootbags[characterId] = lootbag;
        server.executeFuncForAllReadyClientsInRange((client) => {
            server.addLightweightNpc(client, lootbag);
            client.spawnedEntities.add(lootbag);
        }, lootbag);
    }
    createAirdropContainer(server, pos, forceAirdrop = "") {
        const airdropTypes = [
            "Farmer",
            "Demolitioner",
            "Medic",
            "Builder",
            "Fighter",
            "Supplier"
        ];
        const experimentalWeapons = [
            { weapon: enums_1.Items.WEAPON_REAPER, ammo: enums_1.Items.AMMO_308 },
            { weapon: enums_1.Items.WEAPON_BLAZE, ammo: enums_1.Items.AMMO_223 },
            { weapon: enums_1.Items.WEAPON_FROSTBITE, ammo: enums_1.Items.AMMO_762 },
            { weapon: enums_1.Items.WEAPON_NAGAFENS_RAGE, ammo: enums_1.Items.AMMO_12GA }
        ];
        const experimentalSrubs = [
            enums_1.Items.QUEST_HAPPY_SKULL_SCRUBS_CAP,
            enums_1.Items.QUEST_HAPPY_SKULL_SCRUBS_SHIRT,
            enums_1.Items.QUEST_HAPPY_SKULL_SCRUBS_PANTS,
            enums_1.Items.QUEST_HUGZ_NEEDED_SCRUBS_CAP,
            enums_1.Items.QUEST_HUGZ_NEEDED_SCRUBS_SHIRT,
            enums_1.Items.QUEST_HUGZ_NEEDED_SCRUBS_PANTS,
            enums_1.Items.QUEST_KURAMA_MEDICAL_SCRUBS_CAP,
            enums_1.Items.QUEST_KURAMA_MEDICAL_SCRUBS_SHIRT,
            enums_1.Items.QUEST_KURAMA_MEDICAL_SCRUBS_PANTS
        ];
        const index = Math.floor(Math.random() * airdropTypes.length);
        let airdropType = airdropTypes[index];
        let lootSpawner = lootspawns_1.containerLootSpawners[airdropType];
        if (forceAirdrop.length > 0) {
            airdropType = forceAirdrop;
            lootSpawner = lootspawns_1.containerLootSpawners[forceAirdrop];
        }
        const characterId = (0, utils_1.generateRandomGuid)();
        const lootbag = new lootbag_1.Lootbag(characterId, server.getTransientId(characterId), enums_1.ModelIds.MILITARY_CRATE, new Float32Array([pos[0], pos[1] + 0.1, pos[2]]), new Float32Array([0, 0, 0, 0]), server);
        const container = lootbag.getContainer();
        if (container) {
            lootSpawner.items.forEach((item) => {
                server.addContainerItem(lootbag, server.generateItem(item.item, item.spawnCount.max), container);
            });
        }
        let effectId = enums_1.Effects.Smoke_Green; // default
        switch (airdropType) {
            case "Farmer":
                effectId = enums_1.Effects.Smoke_Green;
                break;
            case "Demolitioner":
                effectId = enums_1.Effects.Smoke_Orange;
                break;
            case "Medic":
                effectId = enums_1.Effects.Smoke_Blue;
                break;
            case "Builder":
                effectId = enums_1.Effects.Smoke_Purple;
                break;
            case "Fighter":
                effectId = enums_1.Effects.Smoke_Red;
                if (container) {
                    const experimental = experimentalWeapons[Math.floor(Math.random() * experimentalWeapons.length)];
                    server.addContainerItem(lootbag, server.generateItem(experimental.weapon, 1), container);
                    server.addContainerItem(lootbag, server.generateItem(experimental.ammo, 30), container);
                }
                break;
            case "Supplier":
                effectId = enums_1.Effects.Smoke_Yellow;
                if (container) {
                    const experimental = experimentalWeapons[Math.floor(Math.random() * experimentalWeapons.length)];
                    server.addContainerItem(lootbag, server.generateItem(experimental.weapon, 1), container);
                    server.addContainerItem(lootbag, server.generateItem(experimental.ammo, 30), container);
                }
            case "Hospital":
                effectId = enums_1.Effects.Smoke_Orange;
                if (container) {
                    const randomIndex = Math.floor(Math.random() * experimentalSrubs.length);
                    server.addContainerItem(lootbag, server.generateItem(experimentalSrubs[randomIndex], 1), container);
                }
                break;
        }
        const smokePos = new Float32Array([pos[0], pos[1] + 0.3, pos[2], 1]);
        for (const a in server._clients) {
            const c = server._clients[a];
            server.sendData(c, "Character.PlayWorldCompositeEffect", {
                characterId: c.character.characterId,
                effectId: effectId,
                position: smokePos,
                effectTime: 60
            });
            if ((0, utils_1.isPosInRadius)(3, c.character.state.position, pos)) {
                server.killCharacter(c, { damage: 99999, entity: "aidrop" });
            }
        }
        server._lootbags[characterId] = lootbag;
    }
    createProps(server) {
        if (server.isSurvival()) {
            Z1_lootableProps.forEach((propType) => {
                propType.instances.forEach((propInstance) => {
                    const itemMap = {
                        36: enums_1.Items.FURNACE,
                        9205: enums_1.Items.BARBEQUE,
                        9041: enums_1.Items.CAMPFIRE
                    };
                    if (Object.keys(itemMap).includes(propInstance.modelId.toString())) {
                        server.constructionManager.placeSmeltingEntity(server, itemMap[propInstance.modelId], propInstance.modelId, new Float32Array(propInstance.position), new Float32Array((0, utils_1.fixEulerOrder)(propInstance.rotation)), new Float32Array(propInstance.scale), server._serverGuid, true);
                        return;
                    }
                    const characterId = (0, utils_1.generateRandomGuid)();
                    const obj = new (propInstance.modelId == 9347 ? treasurechest_1.TreasureChest : lootableprop_1.LootableProp)(characterId, server.getTransientId(characterId), // need transient generated for Interaction Replication
                    propInstance.modelId, new Float32Array(propInstance.position), new Float32Array([
                        propInstance.rotation[1],
                        propInstance.rotation[0],
                        propInstance.rotation[2],
                        0
                    ]), server, new Float32Array(propInstance.scale), propInstance.id, Number(propType.renderDistance));
                    server._lootableProps[characterId] = obj;
                    obj.equipItem(server, server.generateItem(obj.containerId), false);
                    if (![
                        enums_1.ModelIds.HOSPITAL_LAB_WORKBENCH,
                        enums_1.ModelIds.TREASURE_CHEST,
                        enums_1.ModelIds.CAMPFIRE,
                        enums_1.ModelIds.FURNACE
                    ].includes(propInstance.modelId)) {
                        const container = obj.getContainer();
                        if (container) {
                            container.canAcceptItems = false;
                        }
                        obj.nameId =
                            server.getItemDefinition(obj.containerId)?.NAME_ID ?? 0;
                    }
                });
            });
            Z1_taskProps.forEach((propType) => {
                propType.instances.forEach((propInstance) => {
                    const characterId = (0, utils_1.generateRandomGuid)();
                    let obj;
                    switch (propType.actorDefinition) {
                        case "Common_Props_SpikeTrap.adr":
                            server.constructionManager.placeTrap(server, enums_1.Items.PUNJI_STICKS, propType.modelId, new Float32Array(propInstance.position), (0, utils_1.fixEulerOrder)(propInstance.rotation), true);
                            break;
                        case "Common_Props_BarbedWire.adr":
                        case "Common_Props_BarbedWire_Posts.adr":
                            server.constructionManager.placeTrap(server, enums_1.Items.BARBED_WIRE, propType.modelId, new Float32Array(propInstance.position), (0, utils_1.fixEulerOrder)(propInstance.rotation), true);
                            break;
                        case "Common_Props_Cabinets_BathroomSink.adr":
                        case "Common_Props_Bathroom_Toilet01.adr":
                        case "Common_Props_Dam_WaterValve01.adr":
                        case "Common_Props_Well.adr":
                        case "Common_Props_FireHydrant.adr":
                            obj = new watersource_1.WaterSource(characterId, server.getTransientId(characterId), // need transient generated for Interaction Replication
                            propType.modelId, new Float32Array(propInstance.position), new Float32Array((0, utils_1.fixEulerOrder)(propInstance.rotation)), server, new Float32Array(propInstance.scale), propInstance.id, propType.renderDistance, propType.actorDefinition, this.waterSourceRefillAmount);
                            break;
                        case "Common_Props_WorkBench01.adr":
                            server.constructionManager.placeSimpleConstruction(server, propType.modelId, new Float32Array(propInstance.position), new Float32Array((0, utils_1.fixEulerOrder)(propInstance.rotation)), new Float32Array([1, 1, 1, 1]), server._serverGuid, enums_1.Items.WORKBENCH);
                            break;
                        case "Common_Props_Gravestone01.adr":
                            obj = new taskprop_1.TaskProp(characterId, server.getTransientId(characterId), // need transient generated for Interaction Replication
                            propType.modelId, new Float32Array(propInstance.position), new Float32Array((0, utils_1.fixEulerOrder)(propInstance.rotation)), server, new Float32Array(propInstance.scale), propInstance.id, propType.renderDistance, propType.actorDefinition);
                            if (propType.tribute) {
                                const thisObj = obj;
                                obj.OnInteractionString = (server, client) => {
                                    server.sendData(client, "Command.InteractionString", {
                                        guid: thisObj.characterId,
                                        stringId: 0
                                    });
                                };
                                obj.getTaskPropData = () => {
                                    thisObj.nameId = 66;
                                    thisObj.rewardItems = [];
                                };
                                obj.OnPlayerSelect = (server, client) => {
                                    server.utilizeHudTimer(client, 66, 60000, // Minute of silence
                                    0, () => {
                                        server.sendChatText(client, "In loving memory of our dear friend, you will be deeply missed.");
                                    });
                                };
                                // punish shooting at the grave
                                obj.OnProjectileHit = (server, damageInfo) => {
                                    const assholeId = damageInfo.entity;
                                    const asshole = server._characters[assholeId];
                                    damageInfo.damage = damageInfo.damage * 2;
                                    asshole.damage(server, damageInfo);
                                };
                                obj.OnMeleeHit = obj.OnProjectileHit;
                            }
                            break;
                        default:
                            obj = new taskprop_1.TaskProp(characterId, server.getTransientId(characterId), // need transient generated for Interaction Replication
                            propType.modelId, new Float32Array(propInstance.position), new Float32Array((0, utils_1.fixEulerOrder)(propInstance.rotation)), server, new Float32Array(propInstance.scale), propInstance.id, propType.renderDistance, propType.actorDefinition);
                    }
                    if (obj)
                        server._taskProps[characterId] = obj;
                });
            });
        }
        Z1_crates.forEach((propType) => {
            propType.instances.forEach((propInstance) => {
                const characterId = (0, utils_1.generateRandomGuid)();
                const obj = new crate_1.Crate(characterId, server.getTransientId(characterId), // need transient generated for Interaction Replication
                (0, crate_1.getActorModelId)(propType.actorDefinition), new Float32Array(propInstance.position), new Float32Array([
                    propInstance.rotation[1],
                    propInstance.rotation[0],
                    propInstance.rotation[2],
                    0
                ]), server, new Float32Array(propInstance.scale), propInstance.zoneId, Number(propType.renderDistance));
                server._crates[characterId] = obj;
            });
        });
        Z1_destroyables.forEach((propType) => {
            propType.instances.forEach((propInstance) => {
                const characterId = (0, utils_1.generateRandomGuid)();
                const obj = new destroyable_1.Destroyable(characterId, server.getTransientId(characterId), // need transient generated for Interaction Replication
                propInstance.modelId, new Float32Array(propInstance.position), new Float32Array([
                    propInstance.rotation[1],
                    propInstance.rotation[0],
                    propInstance.rotation[2],
                    0
                ]), server, new Float32Array(propInstance.scale), propInstance.id, Number(propType.renderDistance));
                server._destroyables[characterId] = obj;
                server._destroyableDTOlist.push(propInstance.id);
            });
        });
        debug("All props created");
    }
    async replenishWaterSources(server) {
        let counter = 0;
        for (const a in server._taskProps) {
            if (counter > 9) {
                counter = 0;
                await promises_1.scheduler.wait(60);
            }
            counter++;
            const propInstance = server._taskProps[a];
            if (propInstance instanceof watersource_1.WaterSource)
                propInstance.replenish();
        }
    }
    createDoor(server, modelID, position, rotation, scale, spawnerId) {
        const characterId = (0, utils_1.generateRandomGuid)();
        server._doors[characterId] = new doorentity_1.DoorEntity(characterId, server.getTransientId(characterId), modelID, position, rotation, server, scale, spawnerId);
    }
    createDoors(server) {
        Z1_doors.forEach((doorType) => {
            const modelId = utils_1._.find(models, (model) => {
                return (model.MODEL_FILE_NAME ===
                    doorType.actorDefinition.replace("_Placer", ""));
            })?.ID;
            doorType.instances.forEach((doorInstance) => {
                this.createDoor(server, modelId ? modelId : 9183, new Float32Array(doorInstance.position), new Float32Array(doorInstance.rotation), new Float32Array(doorInstance.scale), 
                // doorInstance.id doesn't exist
                0);
            });
        });
        debug("All doors objects created");
    }
    setSpawnchance(server, entity, percentage, item) {
        if (percentage <= 0)
            return false;
        const randomNumber = Math.random() * 100;
        if (randomNumber <= percentage || percentage >= 100) {
            entity.lootItem(server, server.generateItem(item));
        }
    }
    createVehicleAtPos(server, position, rotation = new Float32Array([0, 0, 0, 1]), vehicleId = 1, maxSpawnChance = false) {
        const characterId = server.generateGuid(), vehicleData = new vehicle_1.Vehicle2016(characterId, server.getTransientId(characterId), 0, new Float32Array(position), new Float32Array(rotation), server, (0, utils_1.getCurrentServerTimeWrapper)().getTruncatedU32(), vehicleId);
        this.createVehicle(server, vehicleData, maxSpawnChance);
    }
    createVehicle(server, vehicle, maxSpawnChance = false) {
        if (!(vehicle instanceof vehicle_1.Vehicle2016)) {
            const characterId = server.generateGuid();
            const vehicleData = new vehicle_1.Vehicle2016(characterId, server.getTransientId(characterId), 0, new Float32Array(vehicle.position), new Float32Array(vehicle.rotation), server, (0, utils_1.getCurrentServerTimeWrapper)().getTruncatedU32(), vehicle.vehicleId);
            this.createVehicle(server, vehicleData, maxSpawnChance);
            return;
        }
        vehicle.equipLoadout(server);
        this.setSpawnchance(server, vehicle, maxSpawnChance ? 100 : 50, enums_1.Items.BATTERY);
        this.setSpawnchance(server, vehicle, maxSpawnChance ? 100 : 50, enums_1.Items.SPARKPLUGS);
        if (vehicle.vehicleId != enums_1.VehicleIds.PARACHUTE) {
            this.setSpawnchance(server, vehicle, maxSpawnChance ? 100 : 30, enums_1.Items.VEHICLE_KEY);
            this.setSpawnchance(server, vehicle, maxSpawnChance ? 100 : 20, enums_1.Items.FUEL_BIOFUEL);
        }
        this.setSpawnchance(server, vehicle, maxSpawnChance ? 100 : 30, vehicle.getHeadlightsItemId());
        this.setSpawnchance(server, vehicle, maxSpawnChance ? 100 : 30, vehicle.getTurboItemId());
        server._vehicles[vehicle.characterId] = vehicle;
    }
    createVehicles(server, maxSpawnChance = false) {
        if (utils_1._.size(server._vehicles) >= this.vehicleSpawnCap)
            return;
        const respawnAmount = Math.ceil((this.vehicleSpawnCap - utils_1._.size(server._vehicles)) / 8);
        for (let x = 0; x < respawnAmount; x++) {
            const dataVehicle = Z1_vehicles[(0, utils_1.randomIntFromInterval)(0, Z1_vehicles.length - 1)];
            let spawn = true;
            Object.values(server._vehicles).forEach((spawnedVehicle) => {
                if (!spawn)
                    return;
                if ((0, utils_1.isPosInRadius)(this.vehicleSpawnRadius, dataVehicle.position, spawnedVehicle.state.position)) {
                    spawn = false;
                }
            });
            if (!spawn) {
                continue;
            }
            const characterId = (0, utils_1.generateRandomGuid)(), vehicleData = new vehicle_1.Vehicle2016(characterId, server.getTransientId(characterId), 0, new Float32Array(dataVehicle.position), new Float32Array(dataVehicle.rotation), server, (0, utils_1.getCurrentServerTimeWrapper)().getTruncatedU32(), dataVehicle.vehicleId);
            vehicleData.positionUpdate.orientation = dataVehicle.orientation;
            this.createVehicle(server, vehicleData, maxSpawnChance); // save vehicle
        }
        debug("All vehicles created");
    }
    async createNpcs(server) {
        // This is only for giving the world some life
        for (const spawnerType of Z1_npcs) {
            const authorizedModelId = [];
            switch (spawnerType.actorDefinition) {
                case "NPCSpawner_ZombieLazy.adr":
                    authorizedModelId.push(9510);
                    authorizedModelId.push(9634);
                    break;
                case "NPCSpawner_ZombieWalker.adr":
                    authorizedModelId.push(9510);
                    authorizedModelId.push(9634);
                    break;
                case "NPCSpawner_Deer001.adr":
                    authorizedModelId.push(9002);
                    authorizedModelId.push(9253);
                    break;
                case "NPCSpawner_Wolf001.adr":
                    authorizedModelId.push(9003);
                    break;
                case "Bear_Brown.adr":
                    authorizedModelId.push(9187);
                    break;
                default:
                    break;
            }
            if (!authorizedModelId.length)
                continue;
            for (const npcInstance of spawnerType.instances) {
                let spawn = true;
                let counter = 0;
                for (const a in server._npcs) {
                    if (counter > 150) {
                        counter = 0;
                        await promises_1.scheduler.wait(30);
                    }
                    counter++;
                    if (!server._npcs[a])
                        continue;
                    if ((0, utils_1.isPosInRadius)(this.npcSpawnRadius, npcInstance.position, server._npcs[a].state.position)) {
                        spawn = false;
                        break;
                    }
                }
                if (!spawn)
                    continue;
                const spawnchance = Math.floor(Math.random() * 100) + 1; // temporary spawnchance
                if (spawnchance <= this.chanceNpc) {
                    const screamerChance = Math.floor(Math.random() * 1000) + 1; // temporary spawnchance
                    if (screamerChance <= this.chanceScreamer) {
                        authorizedModelId.push(9667);
                    }
                    this.createNpc(server, authorizedModelId[Math.floor(Math.random() * authorizedModelId.length)], new Float32Array(npcInstance.position), new Float32Array((0, utils_1.eul2quat)(npcInstance.rotation)), npcInstance.id);
                }
            }
        }
        debug("All npcs objects created");
    }
    refillScrapInChunks(server) {
        for (let x = 0; x < server._grid.length; x++) {
            const chunk = server._grid[x];
            chunk.availableScrap += 20;
            if (chunk.availableScrap > 50)
                chunk.availableScrap = 50;
        }
    }
    async createLoot(server, lTables = lootspawns_1.lootTables) {
        let counter = 0;
        for (const spawnerType of Z1_items) {
            const lootTable = lTables[spawnerType.actorDefinition];
            if (lootTable) {
                for (const itemInstance of spawnerType.instances) {
                    if (counter > 9) {
                        counter = 0;
                        await promises_1.scheduler.wait(60);
                    }
                    counter++;
                    if (this.spawnedLootObjects[itemInstance.id])
                        continue;
                    const chance = Math.floor(Math.random() * 100) + 1;
                    if (chance <=
                        lootTable.spawnChance *
                            (1 - (0, utils_1.isLootNerfedLoc)(itemInstance.position) / 100)) {
                        if (!WorldObjectManager.itemSpawnersChances[itemInstance.id]) {
                            const realSpawnChance = ((lootTable.spawnChance / lootTable.items.length) *
                                spawnerType.instances.length) /
                                100;
                            WorldObjectManager.itemSpawnersChances[spawnerType.actorDefinition] = realSpawnChance;
                        }
                        const item = getRandomItem(lootTable.items);
                        if (item) {
                            this.createLootEntity(server, server.generateItem(getRandomSkin(item.item), (0, utils_1.randomIntFromInterval)(item.spawnCount.min, item.spawnCount.max)), new Float32Array(itemInstance.position), new Float32Array(itemInstance.rotation), itemInstance.id);
                        }
                    }
                }
            }
        }
    }
    async updateQuestContainers(server) {
        let counter = 0;
        for (const a in server._lootableProps) {
            if (counter > 100) {
                counter = 0;
                await promises_1.scheduler.wait(30); // Await the wait function to pause
            }
            counter++;
            const prop = server._lootableProps[a];
            switch (prop.actorModelId) {
                case enums_1.ModelIds.HOSPITAL_LAB_WORKBENCH:
                    if (prop.hasItem(enums_1.Items.SYRINGE_INFECTED_BLOOD) &&
                        prop.hasItem(enums_1.Items.EMPTY_SPECIMEN_BAG) &&
                        prop.hasItem(enums_1.Items.BRAIN_INFECTED) &&
                        prop.hasItem(enums_1.Items.VIAL_H1Z1_REDUCER)) {
                        const req1 = prop.getItemById(enums_1.Items.SYRINGE_INFECTED_BLOOD), req2 = prop.getItemById(enums_1.Items.EMPTY_SPECIMEN_BAG), req3 = prop.getItemById(enums_1.Items.BRAIN_INFECTED), req4 = prop.getItemById(enums_1.Items.VIAL_H1Z1_REDUCER);
                        if (!req1 || !req2 || !req3 || !req4)
                            continue;
                        if (!server.removeInventoryItem(prop, req1) ||
                            !server.removeInventoryItem(prop, req2) ||
                            !server.removeInventoryItem(prop, req3) ||
                            !server.removeInventoryItem(prop, req4)) {
                            continue;
                        }
                        const obj = server.generateItem(enums_1.Items.BRAIN_TREATED, 1);
                        prop.lootItem(server, obj, 1, false);
                    }
                    if (prop.hasItem(enums_1.Items.QUEST_HAPPY_SKULL_SCRUBS_SHIRT) &&
                        prop.hasItem(enums_1.Items.QUEST_HAPPY_SKULL_SCRUBS_PANTS) &&
                        prop.hasItem(enums_1.Items.QUEST_HAPPY_SKULL_SCRUBS_CAP)) {
                        const req1 = prop.getItemById(enums_1.Items.QUEST_HAPPY_SKULL_SCRUBS_SHIRT), req2 = prop.getItemById(enums_1.Items.QUEST_HAPPY_SKULL_SCRUBS_PANTS), req3 = prop.getItemById(enums_1.Items.QUEST_HAPPY_SKULL_SCRUBS_CAP);
                        if (!req1 || !req2 || !req3)
                            continue;
                        if (!server.removeInventoryItem(prop, req1) ||
                            !server.removeInventoryItem(prop, req2) ||
                            !server.removeInventoryItem(prop, req3)) {
                            continue;
                        }
                        const obj1 = server.generateItem(enums_1.Items.REWARD_HAPPY_SKULL_SCRUBS_SHIRT, 1);
                        prop.lootItem(server, obj1, 1, false);
                        const obj2 = server.generateItem(enums_1.Items.REWARD_HAPPY_SKULL_SCRUBS_PANTS, 1);
                        prop.lootItem(server, obj2, 1, false);
                        const obj3 = server.generateItem(enums_1.Items.REWARD_HAPPY_SKULL_SCRUBS_CAP, 1);
                        prop.lootItem(server, obj3, 1, false);
                    }
                    if (prop.hasItem(enums_1.Items.QUEST_HUGZ_NEEDED_SCRUBS_SHIRT) &&
                        prop.hasItem(enums_1.Items.QUEST_HUGZ_NEEDED_SCRUBS_PANTS) &&
                        prop.hasItem(enums_1.Items.QUEST_HUGZ_NEEDED_SCRUBS_CAP)) {
                        const req1 = prop.getItemById(enums_1.Items.QUEST_HUGZ_NEEDED_SCRUBS_SHIRT), req2 = prop.getItemById(enums_1.Items.QUEST_HUGZ_NEEDED_SCRUBS_PANTS), req3 = prop.getItemById(enums_1.Items.QUEST_HUGZ_NEEDED_SCRUBS_CAP);
                        if (!req1 || !req2 || !req3)
                            continue;
                        if (!server.removeInventoryItem(prop, req1) ||
                            !server.removeInventoryItem(prop, req2) ||
                            !server.removeInventoryItem(prop, req3)) {
                            continue;
                        }
                        const obj4 = server.generateItem(enums_1.Items.REWARD_HUGZ_NEEDED_SCRUBS_SHIRT, 1);
                        prop.lootItem(server, obj4, 1, false);
                        const obj5 = server.generateItem(enums_1.Items.REWARD_HUGZ_NEEDED_SCRUBS_PANTS, 1);
                        prop.lootItem(server, obj5, 1, false);
                        const obj6 = server.generateItem(enums_1.Items.REWARD_HUGZ_NEEDED_SCRUBS_CAP, 1);
                        prop.lootItem(server, obj6, 1, false);
                    }
                    if (prop.hasItem(enums_1.Items.QUEST_KURAMA_MEDICAL_SCRUBS_SHIRT) &&
                        prop.hasItem(enums_1.Items.QUEST_KURAMA_MEDICAL_SCRUBS_PANTS) &&
                        prop.hasItem(enums_1.Items.QUEST_KURAMA_MEDICAL_SCRUBS_CAP)) {
                        const req1 = prop.getItemById(enums_1.Items.QUEST_KURAMA_MEDICAL_SCRUBS_SHIRT), req2 = prop.getItemById(enums_1.Items.QUEST_KURAMA_MEDICAL_SCRUBS_PANTS), req3 = prop.getItemById(enums_1.Items.QUEST_KURAMA_MEDICAL_SCRUBS_CAP);
                        if (!req1 || !req2 || !req3)
                            continue;
                        if (!server.removeInventoryItem(prop, req1) ||
                            !server.removeInventoryItem(prop, req2) ||
                            !server.removeInventoryItem(prop, req3)) {
                            continue;
                        }
                        const obj10 = server.generateItem(enums_1.Items.REWARD_KURAMA_MEDICAL_SCRUBS_SHIRT, 1);
                        prop.lootItem(server, obj10, 1, false);
                        const obj11 = server.generateItem(enums_1.Items.REWARD_KURAMA_MEDICAL_SCRUBS_PANTS, 1);
                        prop.lootItem(server, obj11, 1, false);
                        const obj12 = server.generateItem(enums_1.Items.REWARD_KURAMA_MEDICAL_SCRUBS_CAP, 1);
                        prop.lootItem(server, obj12, 1, false);
                    }
                    if (prop.hasItem(enums_1.Items.QUEST_MILITARY_SCRUBS_SHIRT) &&
                        prop.hasItem(enums_1.Items.QUEST_MILITARY_SCRUBS_PANTS) &&
                        prop.hasItem(enums_1.Items.QUEST_MILITARY_SCRUBS_CAP)) {
                        const req1 = prop.getItemById(enums_1.Items.QUEST_MILITARY_SCRUBS_SHIRT), req2 = prop.getItemById(enums_1.Items.QUEST_MILITARY_SCRUBS_PANTS), req3 = prop.getItemById(enums_1.Items.QUEST_MILITARY_SCRUBS_CAP);
                        if (!req1 || !req2 || !req3)
                            continue;
                        if (!server.removeInventoryItem(prop, req1) ||
                            !server.removeInventoryItem(prop, req2) ||
                            !server.removeInventoryItem(prop, req3)) {
                            continue;
                        }
                        const obj7 = server.generateItem(enums_1.Items.REWARD_MILITARY_SCRUBS_SHIRT, 1);
                        prop.lootItem(server, obj7, 1, false);
                        const obj8 = server.generateItem(enums_1.Items.REWARD_MILITARY_SCRUBS_PANTS, 1);
                        prop.lootItem(server, obj8, 1, false);
                        const obj9 = server.generateItem(enums_1.Items.REWARD_MILITARY_SCRUBS_CAP, 1);
                        prop.lootItem(server, obj9, 1, false);
                    }
                    break;
                case enums_1.ModelIds.TREASURE_CHEST:
                    const rewardChest = server._lootableProps[a];
                    if (rewardChest)
                        rewardChest.triggerRewards(server);
                    if (rewardChest.clearChestTimer) {
                        clearTimeout(rewardChest.clearChestTimer);
                    }
                    rewardChest.clearChestTimer = setTimeout(() => {
                        // give the player 5 minutes to loot before clearing out the treasure chest. also check
                        // if no players are currently accessing the chest
                        if (!rewardChest.mountedCharacter) {
                            const container = rewardChest.getContainer();
                            for (const a in container.items) {
                                const item = container.items[a];
                                if (item.itemDefinitionId === rewardChest.requiredItemId)
                                    continue; // skip worn letters
                                server.removeContainerItem(rewardChest, item, container, item.stackCount);
                            }
                        }
                    }, 300_000);
                    break;
            }
        }
    }
    async createContainerLoot(server) {
        let counter = 0;
        for (const a in server._lootableProps) {
            if (counter > 9) {
                counter = 0;
                await promises_1.scheduler.wait(60); // Await the wait function to pause
            }
            counter++;
            const prop = server._lootableProps[a];
            const container = prop.getContainer();
            if (!container)
                continue;
            if (!!Object.keys(container.items).length)
                continue; // skip if container is not empty
            if (!prop.shouldSpawnLoot)
                continue; // skip medical stations and treasure chests
            const lootTable = lootspawns_1.containerLootSpawners[prop.lootSpawner];
            if (lootTable) {
                for (let x = 0; x < lootTable.maxItems; x++) {
                    const item = getRandomItem(lootTable.items);
                    if (!item)
                        continue;
                    const chance = Math.floor(Math.random() * 100) + 1; // temporary spawnchance
                    let allow = true;
                    Object.values(container.items).forEach((spawnedItem) => {
                        if (item.item == spawnedItem.itemDefinitionId)
                            allow = false; // dont allow the same item to be added twice
                    });
                    if (allow) {
                        if (chance <=
                            item.weight * (1 - (0, utils_1.isLootNerfedLoc)(prop.state.position) / 100)) {
                            const count = Math.floor(Math.random() *
                                (item.spawnCount.max - item.spawnCount.min + 1) +
                                item.spawnCount.min);
                            // temporary spawnchance
                            server.addContainerItem(prop, server.generateItem(getRandomSkin(item.item), count), container);
                        }
                    }
                    else {
                        x--;
                    }
                }
            }
            if (Object.keys(container.items).length != 0) {
                // mark prop as unsearched for clients
                Object.values(server._clients).forEach((client) => {
                    const index = client.searchedProps.indexOf(prop);
                    if (index > -1) {
                        client.searchedProps.splice(index, 1);
                    }
                });
            }
        }
    }
}
exports.WorldObjectManager = WorldObjectManager;
