// h1z1-server/out/servers/ZoneServer2016/managers/worlddatamanager.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = __importStar(require("node:test"));
const worlddatamanager_1 = require("./worlddatamanager");
const assert_1 = __importDefault(require("assert"));
const zoneserver_1 = require("../zoneserver");
const constructionparententity_1 = require("../entities/constructionparententity");
const h1emu_core_1 = require("h1emu-core");
const constructionchildentity_1 = require("../entities/constructionchildentity");
const constructiondoor_1 = require("../entities/constructiondoor");
const enums_1 = require("../models/enums");
const isMongoTests = process.env.MONGO_TESTS === "true";
process.env.FORCE_DISABLE_WS = "true";
function removeUndefinedKeys(obj) {
    for (const key in obj) {
        if (obj[key] === undefined) {
            delete obj[key];
        }
        else if (typeof obj[key] === "object") {
            removeUndefinedKeys(obj[key]);
        }
    }
    return obj;
}
function removeMongoDbId(obj) {
    for (const key in obj) {
        if (key === "_id") {
            delete obj[key];
        }
        else if (typeof obj[key] === "object") {
            removeMongoDbId(obj[key]);
        }
    }
    return obj;
}
async function worldSaveUnitTests(t, mongoAddress) {
    const worldId = 81;
    const zone = new zoneserver_1.ZoneServer2016(0, Buffer.from("fake"), mongoAddress, worldId);
    await zone.start();
    const wdmanager = new worlddatamanager_1.WorldDataManager();
    await t.test("check world save version", async () => {
        assert_1.default.equal(wdmanager.worldSaveVersion, 2);
    });
    await t.test("init world", async () => {
        await wdmanager.initialize(worldId, mongoAddress);
    });
    const world = {
        lastGuidItem: 1000n,
        vehicles: [],
        characters: [],
        constructions: [],
        worldConstructions: [],
        crops: [],
        traps: []
    };
    await t.test("convert vehicles", async () => {
        world.vehicles = worlddatamanager_1.WorldDataManager.convertVehiclesToSaveData(Object.values(zone._vehicles), worldId);
    });
    const constructionNb = 1;
    await t.test("convert constructions", async () => {
        // clear the construction list
        zone._constructionFoundations = {};
        assert_1.default.deepStrictEqual(Object.keys(zone._constructionFoundations).length, 0);
        for (let i = 0; i < constructionNb; i++) {
            let characterId = (0, h1emu_core_1.generate_random_guid)();
            let transientId = zone.getTransientId(characterId);
            const pos = new Float32Array([0, 0, 0, 0]);
            const foundation = new constructionparententity_1.ConstructionParentEntity(characterId, transientId, 1, pos, pos, zone, enums_1.Items.FOUNDATION, "1", "name", "", "");
            for (let j = 0; j < 4; j++) {
                characterId = (0, h1emu_core_1.generate_random_guid)();
                transientId = zone.getTransientId(characterId);
                const wall = new constructionchildentity_1.ConstructionChildEntity(characterId, transientId, 1, pos, pos, zone, enums_1.Items.METAL_WALL, foundation.characterId, "");
                foundation.setWallSlot(zone, wall);
            }
            characterId = (0, h1emu_core_1.generate_random_guid)();
            transientId = zone.getTransientId(characterId);
            const door = new constructiondoor_1.ConstructionDoor(characterId, transientId, 1, pos, pos, zone, enums_1.Items.DOOR_METAL, foundation.characterId, "", "");
            foundation.setWallSlot(zone, door);
            zone._constructionFoundations[characterId] = foundation;
        }
        Object.values(zone._constructionFoundations).forEach((entity) => {
            const construction = worlddatamanager_1.WorldDataManager.getConstructionParentSaveData(entity, zone._worldId);
            world.constructions.push(construction);
        });
        assert_1.default.deepStrictEqual(world.constructions.length, constructionNb);
    });
    await t.test("save world", async () => {
        await wdmanager.saveWorld({ ...world });
    });
    await t.test("sanitize world object", async () => {
        world.vehicles.forEach((vehicle) => {
            // some keys are undefined, so we need to remove them
            // there like undefined but defined at the same time weird to explain
            // the value is undefined but the key is defined
            removeUndefinedKeys(vehicle);
        });
    });
    await t.test("load world data", async () => {
        const loadedWorldData = await wdmanager.getServerData(worldId);
        assert_1.default.deepEqual(loadedWorldData?.serverId, worldId);
        assert_1.default.deepEqual(BigInt(loadedWorldData?.lastItemGuid), world.lastGuidItem);
    });
    await t.test("load vehicles", async () => {
        const loadedVehicles = await wdmanager.loadVehiclesData();
        removeMongoDbId(loadedVehicles);
        assert_1.default.deepStrictEqual(loadedVehicles, world.vehicles);
    });
    await t.test("load constructions", async () => {
        const loadedConstructions = await wdmanager.loadConstructionData();
        assert_1.default.deepStrictEqual(loadedConstructions?.length, world.constructions.length);
        assert_1.default.deepStrictEqual(loadedConstructions?.length, constructionNb);
        removeMongoDbId(loadedConstructions);
        removeMongoDbId(world.constructions);
        // sort per characterId since mongo mangle the order
        loadedConstructions.sort((a, b) => {
            return a.characterId.localeCompare(b.characterId);
        });
        world.constructions.sort((a, b) => {
            return a.characterId.localeCompare(b.characterId);
        });
        assert_1.default.deepStrictEqual(loadedConstructions, world.constructions);
    });
    await t.test("load world constructions", async () => {
        const loadedWorldConstructions = await wdmanager.loadWorldFreeplaceConstruction();
        assert_1.default.deepEqual(loadedWorldConstructions, world.worldConstructions);
    });
    await t.test("load crops", async () => {
        const loadedCrops = await wdmanager.loadCropData();
        assert_1.default.deepEqual(loadedCrops, world.crops);
    });
}
(0, node_test_1.default)("WorldDataManager", { timeout: 10000 }, async (t) => {
    await worldSaveUnitTests(t, "");
});
(0, node_test_1.default)("WorldDataManager-mongo", { timeout: 10000, skip: !isMongoTests }, async (t) => {
    await worldSaveUnitTests(t, "mongodb://localhost:27017");
});
(0, node_test_1.after)(() => {
    setImmediate(() => {
        process.exit(0);
    });
});
