// h1z1-server/out/servers/ZoneServer2016/managers/worlddatamanagerthread.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
const worker_1 = require("threads/worker");
const worlddatamanager_1 = require("./worlddatamanager");
const worldDataManager = new worlddatamanager_1.WorldDataManager();
(0, worker_1.expose)({
    initialize(worldId, mongoAddress) {
        return worldDataManager.initialize(worldId, mongoAddress);
    },
    getServerData(serverId) {
        return worldDataManager.getServerData(serverId);
    },
    fetchWorldData() {
        return worldDataManager.fetchWorldData();
    },
    fetchCharacterData(characterId) {
        return worldDataManager.fetchCharacterData(characterId);
    },
    saveWorld(world) {
        return worldDataManager.saveWorld(world);
    },
    insertWorld(lastItemGuid) {
        return worldDataManager.insertWorld(lastItemGuid);
    },
    deleteWorld() {
        return worldDataManager.deleteWorld();
    },
    saveCharacters(characters) {
        return worldDataManager.saveCharacters(characters);
    },
    saveCharacterData(character, lastItemGuid) {
        return worldDataManager.saveCharacterData(character, lastItemGuid);
    },
    kill() {
        process.exit(0);
    }
});
