// h1z1-server/out/servers/ZoneServer2016/managers/worlddatamanager.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorldDataManager = void 0;
exports.constructLoadout = constructLoadout;
exports.constructContainer = constructContainer;
exports.constructContainers = constructContainers;
const mongodb_1 = require("mongodb");
const utils_1 = require("../../../utils/utils");
const loadoutItem_1 = require("../classes/loadoutItem");
const loadoutcontainer_1 = require("../classes/loadoutcontainer");
const baseItem_1 = require("../classes/baseItem");
const weapon_1 = require("../classes/weapon");
const constructionchildentity_1 = require("../entities/constructionchildentity");
const constructionparententity_1 = require("../entities/constructionparententity");
const lootableconstructionentity_1 = require("../entities/lootableconstructionentity");
const constructiondoor_1 = require("../entities/constructiondoor");
const plantingdiameter_1 = require("../entities/plantingdiameter");
const plant_1 = require("../entities/plant");
const enums_1 = require("../../../utils/enums");
const constants_1 = require("../../../utils/constants");
const enums_2 = require("../models/enums");
const vehicle_1 = require("../entities/vehicle");
const trapentity_1 = require("../entities/trapentity");
const explosiveentity_1 = require("../entities/explosiveentity");
const fs = require("node:fs");
const debug = require("debug")("ZoneServer");
function constructLoadout(savedLoadout, entityLoadout) {
    Object.values(savedLoadout).forEach((item) => {
        const loadoutItem = new loadoutItem_1.LoadoutItem(new baseItem_1.BaseItem(item.itemDefinitionId, item.itemGuid, item.currentDurability, item.stackCount), item.slotId, item.loadoutItemOwnerGuid);
        loadoutItem.weapon = item.weapon
            ? new weapon_1.Weapon(loadoutItem, item.weapon.ammoCount)
            : undefined;
        entityLoadout[item.slotId] = loadoutItem;
    });
}
function constructContainer(container) {
    const loadoutContainer = new loadoutcontainer_1.LoadoutContainer(new loadoutItem_1.LoadoutItem(new baseItem_1.BaseItem(container.itemDefinitionId, container.itemGuid, container.currentDurability, container.stackCount), container.slotId, container.loadoutItemOwnerGuid), container.containerDefinitionId);
    Object.values(container.items).forEach((item) => {
        const i = new baseItem_1.BaseItem(item.itemDefinitionId, item.itemGuid, item.currentDurability, item.stackCount);
        i.slotId = item.slotId;
        i.containerGuid = item.containerGuid;
        i.weapon = item.weapon ? new weapon_1.Weapon(i, item.weapon.ammoCount) : undefined;
        loadoutContainer.items[item.itemGuid] = i;
    });
    return loadoutContainer;
}
function constructContainers(savedContainers, entityContainers) {
    Object.values(savedContainers).forEach((container) => {
        entityContainers[container.slotId] = constructContainer(container);
    });
}
class WorldDataManager {
    _db;
    _appDataFolder = (0, utils_1.getAppDataFolderPath)();
    _worldId = 0;
    _soloMode = false;
    worldSaveVersion = 2;
    /** MANAGED BY CONFIGMANAGER - See defaultConfig.yaml for more information */
    /*saveTimeInterval: number = 600000;
  
    nextSaveTime: number = Date.now() + this.saveTimeInterval;*/
    static async getDatabase(mongoAddress) {
        const mongoClient = new mongodb_1.MongoClient(mongoAddress, {
            maxPoolSize: 200
        });
        try {
            await mongoClient.connect();
        }
        catch (e) {
            console.error(e);
            throw debug("[ERROR]Unable to connect to mongo server " + mongoAddress);
        }
        debug("connected to mongo !");
        // if no collections exist on h1server database , fill it with samples
        if (!(await mongoClient.db(constants_1.DB_NAME).collections()).length) {
            await (0, utils_1.initMongo)(mongoClient, "ZoneServer");
        }
        return [mongoClient.db(constants_1.DB_NAME), mongoClient];
    }
    async initialize(worldId, mongoAddress) {
        this._soloMode = !mongoAddress;
        this._worldId = worldId;
        if (!this._soloMode) {
            [this._db] = await WorldDataManager.getDatabase(mongoAddress);
        }
    }
    async insertWorld(lastItemGuid) {
        if (this._soloMode)
            return;
        if (!this._worldId) {
            const worldCount = (await this._db?.collection(enums_1.DB_COLLECTIONS.WORLDS).countDocuments()) ||
                0;
            this._worldId = worldCount + 1;
            await this._db?.collection(enums_1.DB_COLLECTIONS.WORLDS).insertOne({
                worldId: this._worldId,
                lastItemGuid: (0, utils_1.toBigHex)(lastItemGuid),
                worldSaveVersion: this.worldSaveVersion
            });
            debug("Existing world was not found, created one.");
        }
        else {
            await this._db?.collection(enums_1.DB_COLLECTIONS.WORLDS).insertOne({
                worldId: this._worldId,
                lastItemGuid: (0, utils_1.toBigHex)(lastItemGuid),
                worldSaveVersion: this.worldSaveVersion
            });
        }
    }
    async fetchWorldData() {
        const vehicles = (await this.loadVehiclesData());
        const constructionParents = (await this.loadConstructionData());
        const freeplace = (await this.loadWorldFreeplaceConstruction());
        const crops = (await this.loadCropData());
        const traps = (await this.loadTrapData());
        debug("World fetched!");
        return {
            constructionParents,
            freeplace,
            crops,
            traps,
            lastTransientId: 0,
            vehicles
        };
    }
    async deleteServerData() {
        if (this._soloMode) {
            fs.writeFileSync(`${this._appDataFolder}/worlddata/world.json`, JSON.stringify({}, null, 2));
        }
        else {
            await this._db?.collection(enums_1.DB_COLLECTIONS.WORLDS).deleteOne({
                worldId: this._worldId
            });
        }
    }
    async deleteCharacters() {
        if (this._soloMode) {
            fs.writeFileSync(`${this._appDataFolder}/single_player_characters2016.json`, JSON.stringify([], null, 2));
        }
        else {
            await this._db?.collection(enums_1.DB_COLLECTIONS.CHARACTERS).updateMany({
                serverId: this._worldId
            }, { $set: { status: 0 } });
        }
    }
    async deleteWorld() {
        await this.deleteServerData();
        await this.deleteCharacters();
        debug("World deleted!");
    }
    async saveWorld(world) {
        console.time("WDM: saveWorld");
        await this.saveVehicles(world.vehicles.filter((vehicle) => ![enums_2.VehicleIds.SPECTATE, enums_2.VehicleIds.PARACHUTE].includes(vehicle.vehicleId)));
        await this.saveServerData(world.lastGuidItem);
        await this.saveCharacters(world.characters);
        await this.saveConstructionData(world.constructions);
        await this.saveWorldFreeplaceConstruction(world.worldConstructions);
        await this.saveCropData(world.crops);
        await this.saveTrapData(world.traps);
        console.timeEnd("WDM: saveWorld");
    }
    //#region DATA GETTER HELPER FUNCTIONS
    static getBaseSaveData(serverId) {
        return {
            serverId: serverId
        };
    }
    static getBaseEntityUpdateSaveData(entity) {
        return {
            position: Array.from(entity.state.position),
            rotation: Array.from(entity.state.rotation)
        };
    }
    static getPositionUpdateSaveData(entity) {
        return {
            orientation: entity.positionUpdate.orientation,
            frontTilt: entity.positionUpdate.frontTilt,
            sideTilt: entity.positionUpdate.sideTilt
        };
    }
    static getBaseFullEntitySaveData(entity, serverId) {
        return {
            ...this.getBaseEntityUpdateSaveData(entity),
            ...this.getBaseSaveData(serverId),
            characterId: entity.characterId,
            actorModelId: entity.actorModelId
        };
    }
    static getWeaponSaveData(weapon) {
        return {
            ammoCount: weapon.ammoCount
        };
    }
    static getItemSaveData(item) {
        return {
            itemDefinitionId: item.itemDefinitionId,
            slotId: item.slotId,
            itemGuid: item.itemGuid,
            containerGuid: item.containerGuid,
            currentDurability: item.currentDurability,
            stackCount: item.stackCount,
            weapon: item.weapon ? this.getWeaponSaveData(item.weapon) : undefined
        };
    }
    static getLoadoutItemSaveData(item) {
        return {
            ...this.getItemSaveData(item),
            loadoutItemOwnerGuid: item.loadoutItemOwnerGuid
        };
    }
    static getLoadoutContainerSaveData(container) {
        const items = {};
        Object.values(container.items).forEach((item) => {
            items[item.itemGuid] = {
                ...this.getItemSaveData(item)
            };
        });
        return {
            ...this.getLoadoutItemSaveData(container),
            containerDefinitionId: container.containerDefinitionId,
            items: items
        };
    }
    static getBaseFullCharacterUpdateSaveData(entity, worldSaveVersion) {
        const loadout = {}, containers = {};
        Object.values(entity._loadout).forEach((item) => {
            loadout[item.slotId] = {
                ...this.getLoadoutItemSaveData(item)
            };
        });
        Object.values(entity._containers).forEach((container) => {
            containers[container.slotId] = {
                ...this.getLoadoutContainerSaveData(container)
            };
        });
        return {
            ...this.getBaseEntityUpdateSaveData(entity),
            _loadout: loadout,
            _containers: containers,
            _resources: entity._resources,
            worldSaveVersion: worldSaveVersion
        };
    }
    //#endregion
    //#region SERVER DATA
    async getServerData(serverId) {
        let serverData;
        if (this._soloMode) {
            serverData = require(`${this._appDataFolder}/worlddata/world.json`);
            if (!serverData?.serverId) {
                debug("World data not found in file, aborting.");
                return null;
            }
        }
        else {
            serverData = await this._db
                ?.collection(enums_1.DB_COLLECTIONS.WORLDS)
                .findOne({ worldId: serverId });
            if (!serverData || !serverData.serverId) {
                debug("World data not found in mongo, aborting.");
                return null;
            }
        }
        return serverData ?? null;
    }
    async saveServerData(lastItemGuid) {
        const saveData = {
            serverId: this._worldId,
            lastItemGuid: (0, utils_1.toBigHex)(lastItemGuid),
            worldSaveVersion: this.worldSaveVersion
        };
        if (this._soloMode) {
            fs.writeFileSync(`${this._appDataFolder}/worlddata/world.json`, JSON.stringify(saveData, null, 2));
        }
        else {
            await this._db?.collection(enums_1.DB_COLLECTIONS.WORLDS).updateOne({ worldId: this._worldId }, {
                $set: {
                    ...saveData
                }
            });
        }
    }
    //#endregion
    //#region CHARACTER DATA
    async fetchCharacterData(characterId) {
        let savedCharacter;
        if (this._soloMode) {
            delete require.cache[require.resolve(`${this._appDataFolder}/single_player_characters2016.json`)];
            const SinglePlayerCharacters = require(`${this._appDataFolder}/single_player_characters2016.json`);
            savedCharacter = SinglePlayerCharacters.find((character) => character.characterId === characterId);
            if (!savedCharacter) {
                throw `[ERROR] Single player character not found! characterId: ${characterId}`;
            }
            savedCharacter.spawnGridData =
                savedCharacter.spawnGridData || Array(100).fill(0);
        }
        else {
            const loadedCharacter = await this._db
                ?.collection("characters")
                .findOne({ characterId: characterId });
            if (!loadedCharacter) {
                throw `[ERROR] Mongo character not found! characterId: ${characterId}`;
            }
            savedCharacter = {
                serverId: loadedCharacter.serverId,
                creationDate: loadedCharacter.creationDate,
                lastLoginDate: loadedCharacter.lastLoginDate,
                characterId: loadedCharacter.characterId,
                ownerId: loadedCharacter.ownerId,
                characterName: loadedCharacter.characterName,
                actorModelId: loadedCharacter.actorModelId,
                headActor: loadedCharacter.headActor,
                hairModel: loadedCharacter.hairModel,
                gender: loadedCharacter.gender,
                isRespawning: loadedCharacter.isRespawning || false,
                spawnGridData: loadedCharacter.spawnGridData || Array(100).fill(0),
                position: loadedCharacter.position,
                rotation: loadedCharacter.rotation,
                _loadout: loadedCharacter._loadout || {},
                _containers: loadedCharacter._containers || {},
                _resources: loadedCharacter._resources || {},
                mutedCharacters: loadedCharacter.mutedCharacters || [],
                metrics: loadedCharacter.metrics || {},
                playTime: loadedCharacter.playTime ?? 0,
                lastDropPlayTime: loadedCharacter.lastDropPlayTime ?? 0,
                status: 1,
                worldSaveVersion: this.worldSaveVersion
            };
        }
        return savedCharacter;
    }
    static convertCharactersToSaveData(characters, worldId) {
        const charactersSaveData = [];
        for (let i = 0; i < characters.length; i++) {
            const character = characters[i];
            const characterSave = this.convertToCharacterSaveData(character, worldId);
            // TODO: this is a temp solution, a deepclone slow down the save process :(
            (0, utils_1.removeUntransferableFields)(characterSave);
            charactersSaveData.push(characterSave);
        }
        return charactersSaveData;
    }
    static convertVehiclesToSaveData(characters, worldId) {
        const vehiclesSaveData = [];
        for (let i = 0; i < characters.length; i++) {
            const vehicle = characters[i];
            const vehicleSave = this.convertToVehicleSaveData(vehicle, worldId);
            // TODO: this is a temp solution, a deepclone slow down the save process :(
            (0, utils_1.removeUntransferableFields)(vehicleSave);
            vehiclesSaveData.push(vehicleSave);
        }
        return vehiclesSaveData;
    }
    static convertToVehicleSaveData(vehicle, worldId) {
        const saveData = {
            ...WorldDataManager.getBaseFullCharacterUpdateSaveData(vehicle, worldId),
            vehicleId: vehicle.vehicleId,
            actorModelId: vehicle.actorModelId,
            shaderGroupId: vehicle.shaderGroupId,
            characterId: vehicle.characterId,
            serverId: worldId,
            rotation: Array.from(vehicle.state.lookAt),
            positionUpdate: WorldDataManager.getPositionUpdateSaveData(vehicle)
        };
        return saveData;
    }
    static convertToCharacterSaveData(character, worldId) {
        const saveData = {
            ...WorldDataManager.getBaseFullCharacterUpdateSaveData(character, worldId),
            characterId: character.characterId,
            rotation: Array.from(character.state.lookAt),
            isRespawning: character.isRespawning,
            playTime: character.playTime,
            lastDropPlayTime: character.lastDropPlaytime,
            spawnGridData: character.spawnGridData,
            mutedCharacters: character.mutedCharacters,
            metrics: character.metrics
        };
        return saveData;
    }
    async saveCharacterData(characterSaveData, lastItemGuid) {
        /*
          lastItemGuid MUST be saved whenever a character is saved (a character is saved on logout)
          in case of a crash so that a player can't end up with an item in their inventory with an
          itemGuid lower than the saved lastItemGuid
        */
        if (lastItemGuid)
            this.saveServerData(lastItemGuid);
        if (this._soloMode) {
            const singlePlayerCharacters = require(`${this._appDataFolder}/single_player_characters2016.json`);
            let singlePlayerCharacter = singlePlayerCharacters.find((character) => character.characterId === character.characterId);
            if (!singlePlayerCharacter) {
                console.log("[ERROR] Single player character savedata not found!");
                return;
            }
            singlePlayerCharacter = {
                ...singlePlayerCharacter,
                ...characterSaveData
            };
            fs.writeFileSync(`${this._appDataFolder}/single_player_characters2016.json`, JSON.stringify([singlePlayerCharacter], null, 2));
        }
        else {
            await this._db?.collection(enums_1.DB_COLLECTIONS.CHARACTERS).updateOne({
                serverId: this._worldId,
                characterId: characterSaveData.characterId
            }, {
                $set: {
                    ...characterSaveData
                }
            }, { upsert: process.env.MONGO_TESTS ? true : false });
        }
    }
    async saveCharacters(characters) {
        const promises = [];
        for (let i = 0; i < characters.length; i++) {
            const character = characters[i];
            promises.push(this.saveCharacterData(character));
        }
        await Promise.all(promises);
    }
    //#endregion
    //#region VEHICLE DATA
    // private async loadVehicleData() {
    //   let vehicles: Array<FullVehicleSaveData>;
    //   if (this._soloMode) {
    //     vehicles = require(`${this._appDataFolder}/worlddata/vehicles.json`);
    //     if (!vehicles) {
    //       debug("Vehicle data not found in file, aborting.");
    //       return;
    //     }
    //   } else {
    //     vehicles = <any>(
    //       await this._db
    //         ?.collection("vehicles")
    //         .find({ worldId: this._worldId })
    //         .toArray()
    //     );
    //   }
    //   vehicles.forEach((vehicle) => {
    //     const transientId = server.getTransientId(vehicle.characterId);
    //     const vehicleData = new Vehicle2016(
    //       vehicle.characterId,
    //       transientId,
    //       0,
    //       new Float32Array(vehicle.position),
    //       new Float32Array(vehicle.rotation),
    //       server,
    //       server._gameTime,
    //       vehicle.vehicleId
    //     );
    //     constructLoadout(vehicle._loadout, vehicleData._loadout);
    //     constructContainers(vehicle._containers, vehicleData._containers);
    //     vehicleData._resources = vehicle._resources;
    //     server.worldObjectManager.createVehicle(server, vehicleData);
    //   });
    // }
    // async saveVehicles(server: ZoneServer2016) {
    //   if (!server.enableWorldSaves) return;
    //   const vehicles: Array<FullVehicleSaveData> = Object.values(
    //     server._vehicles
    //   ).map((vehicle) => {
    //     return {
    //       ...WorldDataManager.getBaseFullCharacterUpdateSaveData(server, vehicle),
    //       ...WorldDataManager.getBaseFullEntitySaveData(server, vehicle),
    //       characterId: vehicle.characterId,
    //       actorModelId: vehicle.actorModelId,
    //       vehicleId: vehicle.vehicleId,
    //       worldSaveVersion: server.worldSaveVersion,
    //     };
    //   });
    //   if (server._soloMode) {
    //     fs.writeFileSync(
    //       `${this._appDataFolder}/worlddata/vehicles.json`,
    //       JSON.stringify(vehicles, null, 2)
    //     );
    //   } else {
    //     const collection = server._db?.collection(DB_COLLECTIONS.VEHICLES);
    //     collection?.deleteMany({ serverId: server._worldId }); // clear vehicles
    //     if (vehicles.length) collection?.insertMany(vehicles);
    //   }
    // }
    //#endregion
    //#region CONSTRUCTION DATA
    static loadConstructionDoorEntity(server, entityData) {
        const transientId = server.getTransientId(entityData.characterId), entity = new constructiondoor_1.ConstructionDoor(entityData.characterId, transientId, entityData.actorModelId, new Float32Array(entityData.position), new Float32Array(entityData.rotation), server, entityData.itemDefinitionId, entityData.ownerCharacterId, entityData.parentObjectCharacterId, entityData.slot);
        entity.health = entityData.health;
        entity.passwordHash = entityData.passwordHash;
        entity.grantedAccess = entityData.grantedAccess;
        entity.placementTime = entityData.placementTime;
        server._constructionDoors[entity.characterId] = entity;
        return entity;
    }
    static loadLootableConstructionEntity(server, entityData, isWorldConstruction = false) {
        const transientId = server.getTransientId(entityData.characterId), entity = new lootableconstructionentity_1.LootableConstructionEntity(entityData.characterId, transientId, entityData.actorModelId, new Float32Array(entityData.position), new Float32Array(entityData.rotation), server, new Float32Array([1, 1, 1, 1]), entityData.itemDefinitionId, entityData.parentObjectCharacterId, entityData.subEntityType);
        entity.health = entityData.health;
        entity.placementTime = entityData.placementTime;
        if (entityData.container) {
            const container = constructContainer(entityData.container);
            entity._loadout["31"] = container;
            entity._containers["31"] = container;
        }
        if (entityData.subEntityType === "CollectingEntity") {
            server.smeltingManager._collectingEntities[entityData.characterId] =
                entityData.characterId;
            const container = entity.getContainer();
            switch (entityData.itemDefinitionId) {
                case enums_2.Items.ANIMAL_TRAP:
                    if (container)
                        container.canAcceptItems = false;
                    break;
                case enums_2.Items.BEE_BOX:
                    if (container)
                        container.acceptedItems = [enums_2.Items.HONEYCOMB, enums_2.Items.WATER_EMPTY];
                    break;
                case enums_2.Items.DEW_COLLECTOR:
                    if (container)
                        container.acceptedItems = [enums_2.Items.WATER_EMPTY];
                    break;
            }
        }
        if (isWorldConstruction) {
            server._worldLootableConstruction[entity.characterId] = entity;
            return entity;
        }
        server._lootableConstruction[entity.characterId] = entity;
        return entity;
    }
    static loadConstructionChildSlots(server, parent, entityData) {
        Object.values(entityData.occupiedWallSlots).forEach((wallData) => {
            let wall;
            if ("occupiedWallSlots" in wallData) {
                wall = this.loadConstructionChildEntity(server, wallData);
            }
            else {
                wall = this.loadConstructionDoorEntity(server, wallData);
            }
            parent.setWallSlot(server, wall);
        });
        Object.values(entityData.occupiedUpperWallSlots).forEach((wallData) => {
            const wall = this.loadConstructionChildEntity(server, wallData);
            parent.setWallSlot(server, wall);
        });
        Object.values(entityData.occupiedShelterSlots).forEach((shelterData) => {
            const shelter = this.loadConstructionChildEntity(server, shelterData);
            parent.setShelterSlot(server, shelter);
        });
        Object.values(entityData.freeplaceEntities).forEach((freeplaceData) => {
            let freeplace;
            if ("occupiedWallSlots" in freeplaceData) {
                freeplace = this.loadConstructionChildEntity(server, freeplaceData);
            }
            else if ("passwordHash" in freeplaceData) {
                freeplace = this.loadConstructionDoorEntity(server, freeplaceData);
            }
            else {
                freeplace = this.loadLootableConstructionEntity(server, freeplaceData);
            }
            parent.addFreeplaceConstruction(freeplace);
        });
    }
    static loadConstructionChildEntity(server, entityData) {
        const transientId = server.getTransientId(entityData.characterId), entity = new constructionchildentity_1.ConstructionChildEntity(entityData.characterId, transientId, entityData.actorModelId, new Float32Array(entityData.position), new Float32Array([0, entityData.eulerAngle, 0, 0]), server, entityData.itemDefinitionId, entityData.parentObjectCharacterId, entityData.slot, entityData.eulerAngle);
        entity.health = entityData.health;
        entity.placementTime = entityData.placementTime;
        server._constructionSimple[entity.characterId] = entity;
        this.loadConstructionChildSlots(server, entity, entityData);
        return entity;
    }
    static loadConstructionParentEntity(server, entityData) {
        const transientId = server.getTransientId(entityData.characterId), foundation = new constructionparententity_1.ConstructionParentEntity(entityData.characterId, transientId, entityData.actorModelId, new Float32Array(entityData.position), new Float32Array([0, entityData.eulerAngle, 0, 0]), server, entityData.itemDefinitionId, entityData.ownerCharacterId, "", entityData.parentObjectCharacterId, entityData.slot, entityData.eulerAngle);
        foundation.health = entityData.health;
        foundation.placementTime = entityData.placementTime;
        foundation.permissions = entityData.permissions;
        server._constructionFoundations[foundation.characterId] = foundation;
        this.loadConstructionChildSlots(server, foundation, entityData);
        Object.values(entityData.occupiedExpansionSlots).forEach((expansionData) => {
            const expansion = WorldDataManager.loadConstructionParentEntity(server, expansionData);
            foundation.setExpansionSlot(expansion);
        });
        Object.values(entityData.occupiedRampSlots).forEach((rampData) => {
            const ramp = this.loadConstructionChildEntity(server, rampData);
            foundation.setRampSlot(ramp);
        });
        foundation.updateSecuredState(server);
        return foundation;
    }
    async loadVehiclesData() {
        let vehicles = [];
        if (this._soloMode) {
            vehicles = require(`${this._appDataFolder}/worlddata/vehicles.json`);
            if (!vehicles) {
                debug("vehicles data not found in file, aborting.");
                return;
            }
        }
        else {
            vehicles = (await this._db
                ?.collection(enums_1.DB_COLLECTIONS.VEHICLES)
                .find({ serverId: this._worldId })
                .toArray());
        }
        return vehicles;
    }
    async loadConstructionData() {
        let constructionParents = [];
        if (this._soloMode) {
            constructionParents = require(`${this._appDataFolder}/worlddata/construction.json`);
            if (!constructionParents) {
                debug("Construction data not found in file, aborting.");
                return;
            }
        }
        else {
            constructionParents = (await this._db
                ?.collection(enums_1.DB_COLLECTIONS.CONSTRUCTION)
                .find({ serverId: this._worldId })
                .toArray());
        }
        return constructionParents;
    }
    static loadConstructionParentEntities(constructionParents, server) {
        constructionParents.forEach((parent) => {
            WorldDataManager.loadConstructionParentEntity(server, parent);
        });
    }
    static getBaseConstructionSaveData(entity, serverId) {
        return {
            ...this.getBaseFullEntitySaveData(entity, serverId),
            health: entity.health,
            placementTime: entity.placementTime,
            parentObjectCharacterId: entity.parentObjectCharacterId,
            itemDefinitionId: entity.itemDefinitionId,
            slot: entity instanceof lootableconstructionentity_1.LootableConstructionEntity ? "" : entity.slot
        };
    }
    static getConstructionDoorSaveData(entity, serverId) {
        return {
            ...this.getBaseConstructionSaveData(entity, serverId),
            ownerCharacterId: entity.ownerCharacterId,
            passwordHash: entity.passwordHash,
            grantedAccess: entity.grantedAccess,
            rotation: Array.from(entity.startRot) // override quaternion rotation
        };
    }
    static getLootableConstructionSaveData(entity, serverId) {
        return {
            ...this.getBaseConstructionSaveData(entity, serverId),
            container: entity.getContainer(),
            subEntityType: entity.subEntity?.subType || ""
        };
    }
    static getConstructionChildSaveData(entity, serverId) {
        const wallSlots = {}, upperWallSlots = {}, shelterSlots = {}, freePlaceEntities = {};
        Object.values(entity.occupiedWallSlots).forEach((wall) => {
            if (wall instanceof constructiondoor_1.ConstructionDoor) {
                wallSlots[wall.getSlotNumber()] = this.getConstructionDoorSaveData(wall, serverId);
            }
            else {
                wallSlots[wall.getSlotNumber()] = this.getConstructionChildSaveData(wall, serverId);
            }
        });
        Object.values(entity.occupiedUpperWallSlots).forEach((wall) => {
            upperWallSlots[wall.getSlotNumber()] = this.getConstructionChildSaveData(wall, serverId);
        });
        Object.values(entity.occupiedShelterSlots).forEach((shelter) => {
            shelterSlots[shelter.getSlotNumber()] = this.getConstructionChildSaveData(shelter, serverId);
        });
        Object.values(entity.freeplaceEntities).forEach((entity) => {
            if (entity instanceof constructiondoor_1.ConstructionDoor) {
                freePlaceEntities[entity.characterId] =
                    this.getConstructionDoorSaveData(entity, serverId);
            }
            else if (entity instanceof lootableconstructionentity_1.LootableConstructionEntity) {
                freePlaceEntities[entity.characterId] =
                    this.getLootableConstructionSaveData(entity, serverId);
            }
            else {
                freePlaceEntities[entity.characterId] =
                    this.getConstructionChildSaveData(entity, serverId);
            }
        });
        return {
            ...this.getBaseConstructionSaveData(entity, serverId),
            eulerAngle: entity.eulerAngle,
            occupiedWallSlots: wallSlots,
            occupiedUpperWallSlots: upperWallSlots,
            occupiedShelterSlots: shelterSlots,
            freeplaceEntities: freePlaceEntities
        };
    }
    static getConstructionParentSaveData(entity, serverId) {
        const expansionSlots = {}, rampSlots = {};
        Object.values(entity.occupiedExpansionSlots).forEach((expansion) => {
            expansionSlots[expansion.getSlotNumber()] =
                this.getConstructionParentSaveData(expansion, serverId);
        });
        Object.values(entity.occupiedRampSlots).forEach((ramp) => {
            rampSlots[ramp.getSlotNumber()] = this.getConstructionChildSaveData(ramp, serverId);
        });
        return {
            ...this.getConstructionChildSaveData(entity, serverId),
            permissions: entity.permissions,
            ownerCharacterId: entity.ownerCharacterId,
            occupiedExpansionSlots: expansionSlots,
            occupiedRampSlots: rampSlots
        };
    }
    async saveConstructionData(constructions) {
        if (!constructions.length)
            return;
        if (this._soloMode) {
            fs.writeFileSync(`${this._appDataFolder}/worlddata/construction.json`, JSON.stringify(constructions, null, 2));
            return;
        }
        const collection = this._db?.collection(enums_1.DB_COLLECTIONS.CONSTRUCTION);
        // 1. Build ordered bulkWrite operations
        const ops = constructions.map((c) => ({
            updateOne: {
                filter: { characterId: c.characterId, serverId: this._worldId },
                update: { $set: c },
                upsert: true
            }
        }));
        // 2. Execute bulkWrite in order for consistency
        await collection.bulkWrite(ops, { ordered: true });
        // 3. Remove old records not present in the current constructions
        const allCharacterIds = constructions.map((c) => c.characterId);
        await collection.deleteMany({
            serverId: this._worldId,
            characterId: { $nin: allCharacterIds }
        });
    }
    static getPlantSaveData(entity, serverId) {
        return {
            ...this.getBaseFullEntitySaveData(entity, serverId),
            growState: entity.growState,
            nextStateTime: entity.nextStateTime,
            parentObjectCharacterId: entity.parentObjectCharacterId,
            slot: entity.slot,
            item: this.getItemSaveData(entity.item)
        };
    }
    static getPlantingDiameterSaveData(entity, serverId) {
        const slots = {};
        Object.values(entity.seedSlots).forEach((plant) => {
            slots[plant.slot] = this.getPlantSaveData(plant, serverId);
        });
        return {
            ...this.getBaseFullEntitySaveData(entity, serverId),
            seedSlots: slots,
            placementTime: entity.placementTime,
            fertilizedTimestamp: entity.fertilizedTimestamp,
            isFertilized: entity.isFertilized,
            ownerCharacterId: entity.ownerCharacterId
        };
    }
    async saveCropData(crops) {
        if (this._soloMode) {
            fs.writeFileSync(`${this._appDataFolder}/worlddata/crops.json`, JSON.stringify(crops, null, 2));
        }
        else {
            const collection = this._db?.collection(enums_1.DB_COLLECTIONS.CROPS);
            const updatePromises = [];
            for (let i = 0; i < crops.length; i++) {
                const construction = crops[i];
                updatePromises.push(collection.updateOne({ characterId: construction.characterId, serverId: this._worldId }, { $set: construction }, { upsert: true }));
            }
            await Promise.all(updatePromises);
            const allCharactersIds = crops.map((crop) => {
                return crop.characterId;
            });
            await collection.deleteMany({
                serverId: this._worldId,
                characterId: { $nin: allCharactersIds }
            });
        }
    }
    static loadPlant(server, parent, entityData) {
        const item = new baseItem_1.BaseItem(entityData.item.itemDefinitionId, entityData.item.itemGuid, entityData.item.currentDurability, entityData.item.stackCount), transientId = server.getTransientId(entityData.characterId), plant = new plant_1.Plant(entityData.characterId, transientId, entityData.actorModelId, new Float32Array(entityData.position), new Float32Array(entityData.rotation), server, 0, item, parent.characterId, entityData.slot);
        plant.growState = entityData.growState;
        plant.nextStateTime = entityData.nextStateTime;
        server._plants[plant.characterId] = plant;
        return plant;
    }
    static loadVehicles(server, entityData) {
        const transientId = server.getTransientId(entityData.characterId), vehicle = new vehicle_1.Vehicle2016(entityData.characterId, transientId, entityData.actorModelId, new Float32Array(entityData.position), new Float32Array(entityData.rotation), server, (0, utils_1.getCurrentServerTimeWrapper)().getTruncatedU32(), entityData.vehicleId, entityData?.shaderGroupId ?? 0);
        vehicle._resources = entityData._resources;
        Object.assign(vehicle.positionUpdate, entityData.positionUpdate);
        constructLoadout(entityData._loadout, vehicle._loadout);
        constructContainers(entityData._containers, vehicle._containers);
        server._vehicles[vehicle.characterId] = vehicle;
    }
    static loadPlantingDiameter(server, entityData) {
        const transientId = server.getTransientId(entityData.characterId), plantingDiameter = new plantingdiameter_1.PlantingDiameter(entityData.characterId, transientId, entityData.actorModelId, new Float32Array(entityData.position), new Float32Array(entityData.rotation), server, entityData?.ownerCharacterId ?? "");
        plantingDiameter.fertilizedTimestamp = entityData.fertilizedTimestamp;
        plantingDiameter.isFertilized = entityData.isFertilized;
        server._temporaryObjects[plantingDiameter.characterId] = plantingDiameter;
        Object.values(entityData.seedSlots).forEach((plant) => {
            plantingDiameter.seedSlots[plant.slot] = this.loadPlant(server, plantingDiameter, plant);
        });
    }
    async loadCropData() {
        let crops = [];
        if (this._soloMode) {
            crops = require(`${this._appDataFolder}/worlddata/crops.json`);
            if (!crops) {
                debug("Crop data not found in file, aborting.");
                return;
            }
        }
        else {
            crops = (await this._db
                ?.collection(enums_1.DB_COLLECTIONS.CROPS)
                .find({ serverId: this._worldId })
                .toArray());
        }
        return crops;
    }
    async saveVehicles(vehicles) {
        if (this._soloMode) {
            fs.writeFileSync(`${this._appDataFolder}/worlddata/vehicles.json`, JSON.stringify(vehicles, null, 2));
        }
        else {
            const collection = this._db?.collection(enums_1.DB_COLLECTIONS.VEHICLES);
            const updatePromises = [];
            for (let i = 0; i < vehicles.length; i++) {
                const vehicle = vehicles[i];
                updatePromises.push(collection.updateOne({ characterId: vehicle.characterId, serverId: this._worldId }, { $set: vehicle }, { upsert: true }));
            }
            await Promise.all(updatePromises);
            const allCharactersIds = vehicles.map((vehicle) => {
                return vehicle.characterId;
            });
            await collection.deleteMany({
                serverId: this._worldId,
                characterId: { $nin: allCharactersIds }
            });
        }
    }
    async saveWorldFreeplaceConstruction(freeplaces) {
        if (this._soloMode) {
            fs.writeFileSync(`${this._appDataFolder}/worlddata/worldconstruction.json`, JSON.stringify(freeplaces, null, 2));
        }
        else {
            const collection = this._db?.collection(enums_1.DB_COLLECTIONS.WORLD_CONSTRUCTIONS);
            const updatePromises = [];
            for (let i = 0; i < freeplaces.length; i++) {
                const construction = freeplaces[i];
                updatePromises.push(collection.updateOne({ characterId: construction.characterId, serverId: this._worldId }, { $set: construction }, { upsert: true }));
            }
            await Promise.all(updatePromises);
            const allCharactersIds = freeplaces.map((freeplace) => {
                return freeplace.characterId;
            });
            await collection.deleteMany({
                serverId: this._worldId,
                characterId: { $nin: allCharactersIds }
            });
        }
    }
    async loadWorldFreeplaceConstruction() {
        //worldconstruction
        let freeplace = [];
        if (this._soloMode) {
            freeplace = require(`${this._appDataFolder}/worlddata/worldconstruction.json`);
            if (!freeplace) {
                debug("World construction data not found in file, aborting.");
                return;
            }
        }
        else {
            freeplace = (await this._db
                ?.collection(enums_1.DB_COLLECTIONS.WORLD_CONSTRUCTIONS)
                .find({ serverId: this._worldId })
                .toArray());
        }
        return freeplace;
    }
    static getTrapSaveData(entity, serverId) {
        return {
            ...this.getBaseFullEntitySaveData(entity, serverId),
            ownerCharacterId: entity.ownerCharacterId,
            itemDefinitionId: entity.itemDefinitionId,
            health: entity.health
        };
    }
    async saveTrapData(traps) {
        if (this._soloMode) {
            fs.writeFileSync(`${this._appDataFolder}/worlddata/traps.json`, JSON.stringify(traps, null, 2));
        }
        else {
            const collection = this._db?.collection(enums_1.DB_COLLECTIONS.TRAPS);
            const updatePromises = [];
            for (let i = 0; i < traps.length; i++) {
                const construction = traps[i];
                updatePromises.push(collection.updateOne({ characterId: construction.characterId, serverId: this._worldId }, { $set: construction }, { upsert: true }));
            }
            await Promise.all(updatePromises);
            const allCharactersIds = traps.map((trap) => {
                return trap.characterId;
            });
            await collection.deleteMany({
                serverId: this._worldId,
                characterId: { $nin: allCharactersIds }
            });
        }
    }
    async loadTrapData() {
        let traps = [];
        if (this._soloMode) {
            traps = require(`${this._appDataFolder}/worlddata/traps.json`);
            if (!traps) {
                debug("trap data not found in file, aborting.");
                return;
            }
        }
        else {
            traps = (await this._db
                ?.collection(enums_1.DB_COLLECTIONS.TRAPS)
                .find({ serverId: this._worldId })
                .toArray());
        }
        return traps;
    }
    static loadTraps(server, entityData) {
        const transientId = server.getTransientId(entityData.characterId);
        switch (entityData.itemDefinitionId) {
            case enums_2.Items.LANDMINE:
                const explosive = new explosiveentity_1.ExplosiveEntity(entityData.characterId, transientId, entityData.actorModelId, new Float32Array(entityData.position), new Float32Array(entityData.rotation), server, entityData.itemDefinitionId, entityData.ownerCharacterId);
                server._explosives[entityData.characterId] = explosive;
                explosive.arm(server);
                //temporarily Disabled
                break;
            default:
                const trap = new trapentity_1.TrapEntity(entityData.characterId, transientId, entityData.actorModelId, new Float32Array(entityData.position), new Float32Array(entityData.rotation), server, entityData.itemDefinitionId, false, entityData.ownerCharacterId);
                trap.health = entityData.health;
                server._traps[trap.characterId] = trap;
                trap.arm();
        }
    }
}
exports.WorldDataManager = WorldDataManager;
