// h1z1-server/out/servers/ZoneServer2016/handlers/commands/internalcommands.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.internalCommands = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const enums_1 = require("../../models/enums");
const vehicle_1 = require("../../entities/vehicle");
const types_1 = require("./types");
const utils_1 = require("../../../../utils/utils");
const constants_1 = require("../../../../utils/constants");
exports.internalCommands = [
    //#region DEFAULT PERMISSIONS
    {
        name: "respawn",
        permissionLevel: types_1.PermissionLevels.DEFAULT,
        execute: async (server, client, packetData) => {
            await server.hookManager.checkAsyncHook("OnPlayerRespawn", client);
            if (server.isBattleRoyale()) {
                const pos = client.character.awaitingTeleportLocation ??
                    new Float32Array([3824.41, 148.55, -3999.0, 1]);
                server.respawnPlayer(client, pos);
                return;
            }
            const { gridPosition } = packetData;
            if (!gridPosition) {
                return;
            }
            const matchingCell = server._spawnGrid.find((cell) => (0, utils_1.isPosInRadius)(50, cell.position, gridPosition));
            if (matchingCell) {
                const pos = server.calculatePosFromSpawnCell(matchingCell);
                server.respawnPlayer(client, pos);
            }
        }
    },
    {
        name: "spectate",
        permissionLevel: types_1.PermissionLevels.MODERATOR,
        execute: (server, client, packetData) => {
            client.character.isSpectator = !client.character.isSpectator;
            server.hookManager.checkHook("OnPlayerSpectate", client);
            if (client.character.isSpectator) {
                const vehicle = new vehicle_1.Vehicle2016(constants_1.OBSERVER_GUID, server.getTransientId(constants_1.OBSERVER_GUID), 9371, client.character.state.position, client.character.state.lookAt, server, (0, utils_1.getCurrentServerTimeWrapper)().getTruncatedU32(), enums_1.VehicleIds.SPECTATE);
                for (const a in server._clients) {
                    const iteratedClient = server._clients[a];
                    if (iteratedClient.spawnedEntities.has(client.character)) {
                        server.sendData(iteratedClient, "Character.RemovePlayer", {
                            characterId: client.character.characterId
                        });
                        iteratedClient.spawnedEntities.delete(client.character);
                    }
                }
                server.sendData(client, "Spectator.Enable", {});
                server.sendData(client, "AddLightweightVehicle", {
                    ...vehicle,
                    npcData: {
                        ...vehicle,
                        ...vehicle.state,
                        actorModelId: vehicle.actorModelId
                    }
                });
                server.sendData(client, "Mount.MountResponse", {
                    characterId: client.character.characterId,
                    vehicleGuid: vehicle.characterId,
                    seatId: 0,
                    isDriver: 1,
                    identity: {}
                });
                server.sendData(client, "Character.ManagedObject", {
                    objectCharacterId: vehicle.characterId,
                    characterId: client.character.characterId
                });
                server.sendData(client, "ClientUpdate.ManagedObjectResponseControl", {
                    control: true,
                    objectCharacterId: vehicle.characterId
                });
            }
            else {
                server.sendData(client, "Mount.DismountResponse", {
                    characterId: client.character.characterId
                });
                server.sendData(client, "Character.RemovePlayer", {
                    characterId: constants_1.OBSERVER_GUID
                });
                for (const a in server._decoys) {
                    const decoy = server._decoys[a];
                    if (decoy.transientId == client.character.transientId) {
                        server.sendDataToAll("Character.RemovePlayer", {
                            characterId: decoy.characterId
                        });
                        client.isDecoy = false;
                        server.sendChatText(client, `Decoy removed`, false);
                    }
                }
                return;
            }
            server.sendAlert(client, `Set spectate state to ${client.character.isSpectator}`);
        }
    },
    {
        name: "run",
        permissionLevel: types_1.PermissionLevels.MODERATOR,
        execute: (server, client, packetData) => {
            if (packetData.runSpeed &&
                packetData.runSpeed > 10 &&
                !client.character.isGodMode()) {
                server.setGodMode(client, true);
            }
            server.sendData(client, "Command.RunSpeed", {
                runSpeed: packetData.runSpeed
            });
        }
    },
    {
        name: "vehicle",
        permissionLevel: types_1.PermissionLevels.ADMIN,
        execute: (server, client, packetData) => {
            const allowedIds = [
                enums_1.VehicleIds.POLICECAR,
                enums_1.VehicleIds.PICKUP,
                enums_1.VehicleIds.ATV,
                enums_1.VehicleIds.OFFROADER
            ];
            const vehicleId = packetData.vehicleId;
            if (!vehicleId || !allowedIds.includes(vehicleId)) {
                server.sendChatText(client, "[ERROR] Invalid vehicleId, please choose one of listed below:");
                server.sendChatText(client, `OFFROADER: ${enums_1.VehicleIds.OFFROADER}, PICKUP: ${enums_1.VehicleIds.PICKUP}, POLICECAR: ${enums_1.VehicleIds.POLICECAR}, ATV: ${enums_1.VehicleIds.ATV}`);
                return;
            }
            const characterId = server.generateGuid(), position = packetData.position;
            if (!position)
                return;
            const vehicle = new vehicle_1.Vehicle2016(characterId, server.getTransientId(characterId), 0, position, client.character.state.lookAt, server, (0, utils_1.getCurrentServerTimeWrapper)().getTruncatedU32(), vehicleId, 0);
            server.worldObjectManager.createVehicle(server, vehicle, true);
            client.character.ownedVehicle = vehicle.characterId;
        }
    }
];
