// h1z1-server/out/packets/ClientProtocol/ClientProtocol_860/ragdoll.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ragdollPackets = void 0;
const shared_1 = require("./shared");
exports.ragdollPackets = [
    [
        "Ragdoll.Start",
        0xd00100,
        {
            fields: [
                {
                    name: "characterId",
                    type: "uint64string",
                    defaultValue: "0x0000000000000000"
                }
            ]
        }
    ],
    [
        "Ragdoll.UpdatePose",
        0xd00101,
        {
            fields: [
                {
                    name: "characterId",
                    type: "uint64string",
                    defaultValue: "0x0000000000000000"
                },
                {
                    name: "positionUpdate",
                    type: "custom",
                    parser: shared_1.readPositionUpdateData,
                    packer: shared_1.packPositionUpdateData
                }
            ]
        }
    ],
    [
        "Ragdoll.Unk",
        0xd00118,
        {
            fields: [
                {
                    name: "characterId",
                    type: "uint64string",
                    defaultValue: "0x0000000000000000"
                },
                { name: "unk1", type: "int32", defaultValue: 1 },
                {
                    name: "unkArray1",
                    type: "array",
                    defaultValue: [],
                    fields: [
                        { name: "unknownByte1", type: "uint8", defaultValue: 0 },
                        {
                            name: "position",
                            type: "floatvector3",
                            defaultValue: [0, 50, 0]
                        },
                        {
                            name: "rotation",
                            type: "floatvector4",
                            defaultValue: [0, 0, 0, 0]
                        }
                    ]
                }
            ]
        }
    ],
    [
        "Ragdoll.Unk2",
        0xd0010b,
        {
            fields: [
                {
                    name: "characterId",
                    type: "uint64string",
                    defaultValue: "0x0000000000000000"
                },
                { name: "unk1", type: "int32", defaultValue: 1 },
                {
                    name: "unkArray1",
                    type: "array",
                    defaultValue: [],
                    fields: [
                        { name: "unknownByte1", type: "uint8", defaultValue: 1 },
                        {
                            name: "position",
                            type: "floatvector3",
                            defaultValue: [0, 50, 0]
                        },
                        {
                            name: "rotation",
                            type: "floatvector4",
                            defaultValue: [0, 0, 0, 0]
                        }
                    ]
                },
                {
                    name: "positionUpdate",
                    type: "custom",
                    parser: shared_1.readPositionUpdateData,
                    packer: shared_1.packPositionUpdateData
                }
            ]
        }
    ],
    [
        "Ragdoll.Stop",
        0xd002,
        {
            fields: [
                { name: "unknown3", type: "uint8", defaultValue: 0 },
                {
                    name: "unknown4",
                    type: "uint64string",
                    defaultValue: "0x0000000000000000"
                },
                {
                    name: "array1",
                    type: "array",
                    fields: [
                        { name: "unknown5", type: "uint8", defaultValue: 0 },
                        { name: "unknown6", type: "uint32", defaultValue: 0 }
                    ]
                },
                {
                    name: "array2",
                    type: "array",
                    fields: [{ name: "unknown7", type: "uint32", defaultValue: 0 }]
                }
            ]
        }
    ]
];
