// h1z1-server/out/servers/LoginZoneConnection/shared/baselzconnection.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseLZConnection = void 0;
const node_events_1 = require("node:events");
const lzconnectionprotocol_1 = require("../../../protocols/lzconnectionprotocol");
const lzconnectionclient_1 = require("./lzconnectionclient");
const node_dgram_1 = __importDefault(require("node:dgram"));
const debug = require("debug")("LZConnection");
class BaseLZConnection extends node_events_1.EventEmitter {
    _serverAddress;
    _serverPort;
    _protocol;
    _udpLength = 512;
    _clients = {};
    _connection;
    _pingTime = 5000; // ms
    _pingTimeout = 12000;
    _pingTimer;
    constructor(serverPort) {
        super();
        this._serverAddress = process.env.SERVER_BIND_ADDRESS || "0.0.0.0";
        this._serverPort = serverPort;
        this._protocol = new lzconnectionprotocol_1.LZConnectionProtocol();
        const isWindows = process.platform === "win32";
        this._connection = node_dgram_1.default.createSocket({
            type: "udp4",
            reuseAddr: true,
            reusePort: !isWindows
        });
    }
    clientHandler(remote, opcode) {
        let client;
        const clientId = `${remote.address}:${remote.port}`;
        if (!this._clients[clientId]) {
            // if client doesn't exist yet, only accept sessionrequest or sessionreply
            if (opcode !== 0x01 && opcode !== 0x02)
                return;
            client = this._clients[clientId] = new lzconnectionclient_1.LZConnectionClient(remote);
            this.updateClientLastPing(clientId);
        }
        else {
            client = this._clients[clientId];
        }
        return client;
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    messageHandler(data, client) {
        throw new Error("You need to implement messageHandler !");
    }
    /* eslint-enable @typescript-eslint/no-unused-vars */
    connectionHandler(data, remote) {
        const client = this.clientHandler(remote, data[0]);
        if (client) {
            this.messageHandler(data, client);
        }
        else {
            debug(`Connection rejected from remote ${remote.address}:${remote.port}`);
        }
    }
    async start() {
        this._connection.on("message", (message, remoteInfo) => this.connectionHandler(message, remoteInfo));
        return await new Promise((resolve) => {
            this._connection.bind(this._serverPort, this._serverAddress, () => {
                resolve();
            });
        });
    }
    async stop() {
        await new Promise((resolve) => {
            this._connection.close(() => {
                resolve(true);
            });
        });
    }
    sendData(client, packetName, obj) {
        // blocks zone from sending packet without open session
        if (!client || (!client.serverId && packetName !== "SessionRequest"))
            return;
        const data = this._protocol.pack(packetName, obj);
        if (data) {
            this._connection.send(data, client.port, client.address);
        }
    }
    ping(client) {
        this.sendData(client, "Ping", {});
    }
    updateClientLastPing(clientId) {
        this._clients[clientId].lastPing = Date.now();
    }
}
exports.BaseLZConnection = BaseLZConnection;
exports.BaseLZConnection = BaseLZConnection;
