// h1z1-server/out/servers/GatewayServer/gatewayserver.threaded.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatewayServerThreaded = exports.GatewayServerThreadedInternalEvents = void 0;
const node_events_1 = require("node:events");
const node_worker_threads_1 = require("node:worker_threads");
const promises_1 = require("node:timers/promises");
const node_path_1 = __importDefault(require("node:path"));
var GatewayServerThreadedInternalEvents;
(function (GatewayServerThreadedInternalEvents) {
    GatewayServerThreadedInternalEvents[GatewayServerThreadedInternalEvents["START"] = 0] = "START";
    GatewayServerThreadedInternalEvents[GatewayServerThreadedInternalEvents["STOP"] = 1] = "STOP";
})(GatewayServerThreadedInternalEvents || (exports.GatewayServerThreadedInternalEvents = GatewayServerThreadedInternalEvents = {}));
class GatewayServerThreaded extends node_events_1.EventEmitter {
    worker;
    appDataChannel = new node_worker_threads_1.MessageChannel();
    loginChannel = new node_worker_threads_1.MessageChannel();
    disconnectChannel = new node_worker_threads_1.MessageChannel();
    clientInfoChannel = new node_worker_threads_1.MessageChannel();
    internalChannel = new node_worker_threads_1.MessageChannel();
    reqCount = 0;
    requestQueue = new Map();
    constructor(serverPort, gatewayKey) {
        super();
        const workerPath = node_path_1.default.join(__dirname, "gatewayserver.worker.js");
        this.worker = new node_worker_threads_1.Worker(workerPath, {
            workerData: {
                serverPort,
                gatewayKey,
                appDataChannel: this.appDataChannel.port1,
                disconnectChannel: this.disconnectChannel.port1,
                loginChannel: this.loginChannel.port1,
                internalChannel: this.internalChannel.port1,
                clientInfoChannel: this.clientInfoChannel.port1
            },
            transferList: [
                this.appDataChannel.port1,
                this.disconnectChannel.port1,
                this.loginChannel.port1,
                this.internalChannel.port1,
                this.clientInfoChannel.port1
            ]
        });
        this.clientInfoChannel.port2.on("message", (msg) => {
            // Resolve the promise with the result
            this.requestQueue.get(msg.requestId)(msg.result);
            this.requestQueue.delete(msg.requestId);
        });
        this.appDataChannel.port2.on("message", (msg) => {
            this.emit("tunneldata", msg.sessionId, Buffer.from(msg.data), msg.channel);
        });
        this.disconnectChannel.port2.on("message", (sessionId) => {
            this.emit("disconnect", sessionId);
        });
        this.loginChannel.port2.on("message", (msg) => {
            this.emit("login", msg.soeClientId, msg.characterId, msg.ticket, msg.client_protocol);
        });
        this.internalChannel.port2.on("message", (msg) => {
            console.log(msg);
        });
    }
    start() {
        this.internalChannel.port2.postMessage(GatewayServerThreadedInternalEvents.START);
    }
    sendTunnelData(soeClientId, data, channel) {
        this.appDataChannel.port2.postMessage({
            soeClientId,
            data,
            channel
        }
        // FIXME:  transfering the buffer create weird behavior need to investigate
        // [data.buffer]
        );
    }
    async stop() {
        this.internalChannel.port2.postMessage(GatewayServerThreadedInternalEvents.STOP);
        await promises_1.scheduler.wait(1000);
        this.worker.terminate();
    }
    async getSoeClientAvgPing(soeClientId) {
        const fnName = this.getSoeClientAvgPing.name;
        return this.askGatewayThread(fnName, soeClientId);
    }
    async getSoeClientNetworkStats(soeClientId) {
        const fnName = this.getSoeClientNetworkStats.name;
        return this.askGatewayThread(fnName, soeClientId);
    }
    askGatewayThread(fnName, soeClientId) {
        const reqId = this.reqCount++;
        this.clientInfoChannel.port2.postMessage({
            fnName,
            soeClientId,
            requestId: reqId
        });
        return new Promise((resolve) => {
            this.requestQueue.set(reqId, resolve);
        });
    }
    async getSoeClientSessionId(soeClientId) {
        const fnName = this.getSoeClientSessionId.name;
        return this.askGatewayThread(fnName, soeClientId);
    }
    async getSoeClientNetworkInfos(soeClientId) {
        const fnName = this.getSoeClientNetworkInfos.name;
        return this.askGatewayThread(fnName, soeClientId);
    }
    async deleteSoeClient(soeClientId) {
        const fnName = this.deleteSoeClient.name;
        return this.askGatewayThread(fnName, soeClientId);
    }
}
exports.GatewayServerThreaded = GatewayServerThreaded;
