// h1z1-server/out/servers/ZoneServer2016/entities/basesimplenpc.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseSimpleNpc = void 0;
const baseentity_1 = require("./baseentity");
class BaseSimpleNpc extends baseentity_1.BaseEntity {
    /** Maximum health alloted BaseSimpleNPC */
    maxHealth = 100000;
    /** Health initialized by maxHealth to determine the health percentage in pGetSimpleProxyHealth */
    health = this.maxHealth;
    constructor(characterId, transientId, actorModelId, position, rotation, server) {
        super(characterId, transientId, actorModelId, position, rotation, server);
    }
    pGetSimpleNpc() {
        return {
            characterId: this.characterId,
            transientId: this.transientId,
            position: this.state.position,
            rotation: this.state.rotation,
            modelId: this.actorModelId,
            scale: this.scale,
            health: this.pGetSimpleProxyHealth().healthPercentage
        };
    }
    pGetSimpleProxyHealth() {
        return {
            characterId: this.characterId,
            healthPercentage: (this.health / this.maxHealth) * 100
        };
    }
}
exports.BaseSimpleNpc = BaseSimpleNpc;
