// h1z1-server/out/servers/ZoneServer2016/entities/baseentity.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseEntity = void 0;
const enums_1 = require("../models/enums");
function getRenderDistance(actorModelId) {
    let range = 0;
    switch (actorModelId) {
        case enums_1.ModelIds.GROUND_TAMPER:
            range = 1000;
            break;
        case enums_1.ModelIds.DECK_EXPANSION:
        case enums_1.ModelIds.METAL_SHACK_DOOR:
        case enums_1.ModelIds.METAL_SHACK:
        case enums_1.ModelIds.SMALL_SHACK:
        case enums_1.ModelIds.DEW_COLLECTOR:
        case enums_1.ModelIds.WOOD_SHACK:
        case enums_1.ModelIds.WOOD_SHACK_DOOR:
            range = 1000;
            break;
        case enums_1.ModelIds.RAMP:
            range = 450;
            break;
        case enums_1.ModelIds.FOUNDATION_STAIRS:
        case enums_1.ModelIds.METAL_GATE:
        case enums_1.ModelIds.METAL_WALL:
        case enums_1.ModelIds.UPPER_METAL_WALL:
        case enums_1.ModelIds.SHELTER:
        case enums_1.ModelIds.LARGE_SHELTER:
        case enums_1.ModelIds.UPPER_LEVEL_SHELTER:
        case enums_1.ModelIds.UPPER_LEVEL_LARGE_SHELTER:
        case enums_1.ModelIds.STRUCTURE_STAIRS:
        case enums_1.ModelIds.TOWER:
        case enums_1.ModelIds.DECK_FOUNDATION:
            range = 1000;
            break;
    }
    return range ? range : 1000;
}
class BaseEntity {
    /** Universal Identifier, used for locating the home of an entity */
    characterId;
    /** Required for interactable entities, npc and interaction components packet */
    transientId;
    /** Id of the model that corresponds to the entity */
    actorModelId;
    temporaryActorModelId;
    /** State of the BaseLightweightCharacter, includes: state (Float32Array),
     * rotation(Float32Array), lookAt(Float32Array), and yaw (number) */
    state;
    /** Physical size of the entity based on the entity model */
    scale = new Float32Array([1, 1, 1, 1]);
    /** Distance (H1Z1 meters) where the entity will render,
     * when undefined, uses the zoneserver._charactersRenderDistance value instead
     */
    npcRenderDistance;
    /** Distance (H1Z1 meters) at which the player can interact with the entity */
    interactionDistance;
    /** Used for vehicle abilities, array of the corresponding vehicles effects */
    effectTags = [];
    /** The physical material the entity is made of - See enums.ts/MaterialTypes for more information */
    materialType;
    /** The guid of the secured shelter the entity is inside */
    isHidden = "";
    server;
    constructor(characterId, transientId, actorModelId, position, rotation, server) {
        this.characterId = characterId;
        this.transientId = transientId;
        this.actorModelId = actorModelId;
        this.state = {
            position: position,
            rotation: rotation
        };
        this.npcRenderDistance =
            getRenderDistance(actorModelId) || server.charactersRenderDistance;
        this.interactionDistance = server.interactionDistance;
        this.materialType = this.getMaterialType(server, this.actorModelId);
        server.pushToGridCell(this);
        this.server = server;
    }
    getMaterialType(server, actorModelId) {
        const modelData = server._modelsData[actorModelId];
        if (!modelData) {
            return 0;
        }
        return modelData.materialType;
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    damage(server, damageInfo) {
        // default: do nothing
    }
    OnPlayerSelect(server, client, isInstant) {
        // default: do nothing
    }
    OnInteractionString(server, client) {
        // default: do nothing
    }
    OnProjectileHit(server, damageInfo) {
        // default: do nothing
    }
    OnMeleeHit(server, damageInfo) {
        // default: do nothing
    }
    destroy(server) {
        console.log(`Attempted to call destroy() on an entity an undefined destroy method! actorModelId ${this.actorModelId}`);
        return false;
    }
    OnExplosiveHit(server, sourceEntity, client) {
        // default: do nothing
    }
}
exports.BaseEntity = BaseEntity;
