// h1z1-server/out/servers/ZoneServer2016/entities/baselootableentity.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseLootableEntity = void 0;
const enums_1 = require("../models/enums");
const basefullcharacter_1 = require("./basefullcharacter");
class BaseLootableEntity extends basefullcharacter_1.BaseFullCharacter {
    /** The current character interacting with the BaseLootableEntity */
    mountedCharacter;
    /** Distance (H1Z1 meters) at which the player can interact with the BaseLootableEntity */
    interactionDistance = 4;
    /** Determines if the entity BaseLootableEntity should be a lootbag,
     * Returns true if the modelId is included in the array below
     */
    isLootbag;
    /** Used to determine whether the modelId should spawn loot when the global respawns occur,
     * Returns false if the modelId is not included in the array below
     */
    shouldSpawnLoot;
    constructor(characterId, transientId, actorModelId, position, rotation, server) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.isLootbag = [
            enums_1.ModelIds.LOOT_BAG_CLEAN,
            enums_1.ModelIds.LOOT_BAG_BLOODY
        ].includes(actorModelId);
        this.useSimpleStruct = true;
        this.shouldSpawnLoot = ![
            enums_1.ModelIds.HOSPITAL_LAB_WORKBENCH,
            enums_1.ModelIds.TREASURE_CHEST,
            enums_1.ModelIds.CAMPFIRE,
            enums_1.ModelIds.FURNACE,
            enums_1.ModelIds.HAND_SHOVEL
        ].includes(actorModelId);
    }
    getContainer() {
        return Object.values(this._containers)[0];
    }
    OnInteractionString(server, client) {
        server.sendData(client, "Command.InteractionString", {
            guid: this.characterId,
            stringId: enums_1.StringIds.OPEN
        });
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        if (client.character.characterId == this.mountedCharacter) {
            client.character.dismountContainer(server);
            delete this.mountedCharacter; // the check below wont fix container if characterId is the same
            return; // as the one that opened it, so just to be sure we delete it here
        }
        if (this.mountedCharacter) {
            if (!server.getClientByCharId(this.mountedCharacter)) {
                delete this.mountedCharacter;
            }
            else {
                server.containerError(client, enums_1.ContainerErrors.IN_USE);
                return;
            }
        }
        client.character.mountContainer(server, this);
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    OnFullCharacterDataRequest(server, client) {
        // do nothing for now
    }
}
exports.BaseLootableEntity = BaseLootableEntity;
