// h1z1-server/out/servers/ZoneServer2016/entities/character.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.Character2016 = void 0;
const enums_1 = require("../models/enums");
const basefullcharacter_1 = require("./basefullcharacter");
const utils_1 = require("../../../utils/utils");
const loadouts_1 = require("../data/loadouts");
const vehicle_1 = require("../entities/vehicle");
const constants_1 = require("../../../utils/constants");
const Recipes_1 = require("../data/Recipes");
const constructionchildentity_1 = require("./constructionchildentity");
const constructionparententity_1 = require("./constructionparententity");
const explosiveentity_1 = require("./explosiveentity");
const projectileentity_1 = require("./projectileentity");
const challengemanager_1 = require("../managers/challengemanager");
const lootableconstructionentity_1 = require("./lootableconstructionentity");
const stats = require("../../../../data/2016/sampleData/stats.json");
class Character2016 extends basefullcharacter_1.BaseFullCharacter {
    /** The players in-game name */
    name;
    /** The location the player spawned at */
    spawnLocation;
    /** Used to update the status of the players resources */
    resourcesUpdater;
    factionId = 2;
    playTime = 0;
    lastDropPlaytime = 0;
    set godMode(state) {
        this.characterStates.invincibility = state;
    }
    get godMode() {
        return this.characterStates.invincibility;
    }
    /** Determines several states of the player such as being in water or knocked out */
    characterStates;
    awaitingTeleportLocation;
    /** States set by StanceFlags */
    isRunning = false;
    isSitting = false;
    /** Used for resources */
    isBleeding = false;
    isBandaged = false;
    isExhausted = false;
    isPoisoned = false;
    isCoffeeSugared = false;
    isDeerScented = false;
    /** Last time (milliseconds) the player melee'd */
    lastMeleeHitTime = 0;
    /** Used to determine if a player has aimlock */
    aimVectorWarns = 0;
    /** Set by isAlive(state), determines whether the player is alive */
    static isAlive = true;
    /** Sets the knocked out state of the character to determine aliveness */
    set isAlive(state) {
        this.characterStates.knockedOut = !state;
    }
    /** Checks whether the player is alive */
    get isAlive() {
        return !this.characterStates.knockedOut;
    }
    /** Determines if a player is currently moving */
    isMoving = false;
    /** Values used upon character creation */
    hairModel;
    isRespawning = false;
    isReady = false;
    creationDate;
    /** Last login date for the player */
    lastLoginDate;
    /** Last player that the user has whispered to */
    lastWhisperedPlayer;
    /** Returns true if a player has recently melee'd a bee box */
    hasAlertedBees = false;
    /** Time (milliseconds) at which the player last exited a vehicle */
    vehicleExitDate = new Date().getTime();
    /** Current selected slot inside the player's hotbar */
    currentLoadoutSlot = enums_1.LoadoutSlots.FISTS;
    /** Current player/vehicle loadout */
    loadoutId = enums_1.LoadoutIds.CHARACTER;
    /** Used for applying the interval of healing to be applied for any heal type */
    healingIntervals = {
        1: null,
        2: null,
        3: null
    };
    /** Used for applying multiple healing types at once */
    healType = {
        1: {
            healingTicks: 0,
            healingMaxTicks: 0
        },
        2: {
            healingTicks: 0,
            healingMaxTicks: 0
        },
        3: {
            healingTicks: 0,
            healingMaxTicks: 0
        }
    };
    /** Used for applying the interval of healing to be applied for any heal type */
    starthealingInterval;
    timeouts;
    /** Handles the current position of the player */
    positionUpdate;
    /** Admin tools */
    tempGodMode = false;
    isVanished = false;
    isSpectator = false;
    /** Called once SendSelfToClient has been sent */
    initialized = false;
    /** Data for the spawn grid map, keeps track of which grids to block out */
    spawnGridData = [];
    /** Values used by Fairplay */
    lastJumpTime = 0;
    lastSitTime = 0;
    sitCount = 0;
    /** Values used for detecting Enas movement (spamming crouch for an advantage) */
    crouchCount = 0;
    lastCrouchTime = 0;
    isCrouching = false;
    /** Current stance of the player while holding a weapon */
    weaponStance = 1;
    /** Current stance of the player: jumping, running etc. (See getStanceFlags for all possible stances) */
    stance;
    /** Tracks if the player is currently playing an emote */
    currentEmote = 0; // effectId of the current emote, 0 = no emote
    lastEmoteTime = 0; // timestamp of when the emote was started
    /** Metrics of miscellaneous attributes */
    metrics = {
        recipesDiscovered: 0,
        zombiesKilled: 0,
        wildlifeKilled: 0,
        startedSurvivingTP: Date.now(),
        vehiclesDestroyed: 0,
        playersKilled: 0
    };
    /** Tracks combat with other players/entities */
    combatlog = [];
    /** CharacterId of the vehicle spawned by /hax drive or spawnVehicle */
    ownedVehicle;
    /** Values determined by what entity the player is looking at and in range of */
    currentInteractionGuid = "";
    lastInteractionRequestGuid;
    lastInteractionStringTime = 0;
    lastInteractionTime = 0;
    /** The current container attached in the top left of the player's inventory  */
    mountedContainer;
    /** Default loadout the player will spawn with */
    defaultLoadout = loadouts_1.characterDefaultLoadout;
    /** List of ignored players's characterIds */
    mutedCharacters = [];
    /** Id of the player's group */
    groupId = 0;
    /** Used for applied effects to a player - burning, movement speed etc. */
    _characterEffects = {};
    /** The time at which the most recent unlock attempt, failed */
    lastLockFailure = 0;
    /** Array of the player's applied HUD indicators */
    resourceHudIndicators = [];
    /** Indicator's for the bottom right HUD: bleeding, bandage, BEES! etc. */
    hudIndicators = {};
    /** Screen effects applied to a player: bleeding, night vision, etc. (see ScreenEffects.json) */
    screenEffects = [];
    /** The time (milliseconds) at which a user has sent a second melee hit */
    abilityInitTime = 0;
    /** Used for keeping track of the location and characterId of the first registered melee hit */
    meleeHit = {
        abilityHitLocation: "",
        characterId: ""
    };
    lastRepairTime = 0;
    /** HashMap of all recipes on a server
     * uses recipeId (number) for indexing
     */
    recipes = Recipes_1.recipes;
    currentChallenge = challengemanager_1.ChallengeType.NONE;
    hasAirdropClearance = false;
    constructor(characterId, transientId, server) {
        super(characterId, transientId, 0, new Float32Array([0, 0, 0, 1]), new Float32Array([0, 0, 0, 1]), server);
        /** The distance at which the character will render, exceeding the renderDistance and the object renders away */
        this.npcRenderDistance = 310;
        /** Default resource amounts applied after respawning */
        this._resources = {
            [enums_1.ResourceIds.HEALTH]: 10000,
            [enums_1.ResourceIds.STAMINA]: 600,
            [enums_1.ResourceIds.BLEEDING]: 0
        };
        if (server.isSurvival()) {
            this._resources[enums_1.ResourceIds.HUNGER] = 10000;
            this._resources[enums_1.ResourceIds.HYDRATION] = 10000;
            this._resources[enums_1.ResourceIds.VIRUS] = 0;
            this._resources[enums_1.ResourceIds.COMFORT] = 5000;
            this._resources[enums_1.ResourceIds.ENDURANCE] = 8000;
        }
        if (server.isBattleRoyale()) {
            this._resources[enums_1.ResourceIds.TOXICITY] = 0;
        }
        this.characterStates = {
            knockedOut: false,
            inWater: false,
            invincibility: false
        };
        this.timeouts = {};
        this.starthealingInterval = (client, server, healType) => {
            client.character.healingIntervals[healType] = setTimeout(() => {
                if (!server._clients[client.sessionId]) {
                    return;
                }
                let typeName = "";
                switch (healType) {
                    case 1:
                        typeName = "HEALING_BANDAGE_DRESSED";
                        break;
                    case 2:
                        typeName = "HEALING_BANDAGE";
                        break;
                    case 3:
                        typeName = "HEALING_FIRST_AID";
                        break;
                }
                if (!typeName)
                    return;
                const index = this.resourceHudIndicators.indexOf(typeName);
                if (index <= -1) {
                    this.resourceHudIndicators.push(typeName);
                    server.sendHudIndicators(client);
                }
                client.character._resources[enums_1.ResourceIds.HEALTH] += 100;
                if (client.character._resources[enums_1.ResourceIds.HEALTH] > 10000) {
                    client.character._resources[enums_1.ResourceIds.HEALTH] = 10000;
                }
                server.updateResource(client, client.character.characterId, client.character._resources[enums_1.ResourceIds.HEALTH], enums_1.ResourceIds.HEALTH);
                if (client.character.healType[healType].healingTicks++ <
                    client.character.healType[healType].healingMaxTicks) {
                    client.character.healingIntervals[healType]?.refresh();
                }
                else {
                    client.character.healType[healType].healingMaxTicks = 0;
                    client.character.healType[healType].healingTicks = 0;
                    clearTimeout(client.character.healingIntervals[healType]);
                    this.resourceHudIndicators.splice(index, 1);
                    server.sendHudIndicators(client);
                    client.character.healingIntervals[healType] = null;
                }
            }, 1000);
        };
        this.materialType = enums_1.MaterialTypes.FLESH;
    }
    getShaderGroup() {
        switch (this.headActor) {
            case "SurvivorMale_Head_02.adr":
            case "SurvivorFemale_Head_02.adr":
                return 125;
            case "SurvivorMale_Head_03.adr":
            case "SurvivorFemale_Head_03.adr":
                return 129;
            default:
                return 122;
        }
    }
    pGetRecipes(server) {
        const recipeKeys = Object.keys(this.recipes);
        const recipes = [];
        let i = 0;
        for (const recipe of Object.values(this.recipes)) {
            const recipeDef = server.getItemDefinition(Number(recipeKeys[i]));
            if (!recipeDef)
                continue;
            if (server.isBattleRoyale() &&
                ![enums_1.Items.BANDAGE, enums_1.Items.BACKPACK_SATCHEL, enums_1.Items.ARMOR_PLATED].includes(recipeDef.ID))
                continue;
            i++;
            recipes.push({
                recipeId: recipeDef.ID,
                itemDefinitionId: recipeDef.ID,
                nameId: recipeDef.NAME_ID,
                iconId: recipeDef.IMAGE_SET_ID,
                unknownDword1: 0, // idk
                descriptionId: recipeDef.DESCRIPTION_ID,
                unknownDword2: 1, // idk
                bundleCount: recipe.bundleCount || 1,
                membersOnly: false, // could be used for admin-only recipes?
                filterId: recipe.filterId,
                components: recipe.components
                    .map((component) => {
                    const componentDef = server.getItemDefinition(component.itemDefinitionId);
                    if (!componentDef)
                        return true;
                    return {
                        unknownDword1: 0, // idk
                        nameId: componentDef.NAME_ID,
                        iconId: componentDef.IMAGE_SET_ID,
                        unknownDword2: 0, // idk
                        requiredAmount: component.requiredAmount,
                        unknownQword1: "0x0", // idk
                        unknownDword3: 0, // idk
                        itemDefinitionId: componentDef.ID
                    };
                })
                    .filter((component) => component !== true)
            });
        }
        return recipes;
    }
    startResourceUpdater(client, server) {
        client.character.resourcesUpdater = setTimeout(() => this.updateResources(client, server), 3000);
    }
    updateResources(client, server) {
        let effectId;
        for (const a in this.hudIndicators) {
            const indicator = this.hudIndicators[a];
            if (Date.now() > indicator.expirationTime) {
                delete this.hudIndicators[a];
                server.sendHudIndicators(client);
            }
        }
        for (const a in this._characterEffects) {
            const characterEffect = this._characterEffects[a];
            if (characterEffect.duration < Date.now()) {
                if (characterEffect.endCallback)
                    characterEffect.endCallback(server, this);
                effectId = 0;
                delete this._characterEffects[a];
                continue;
            }
            if (characterEffect.callback)
                characterEffect.callback(server, this);
            effectId = characterEffect.id;
        }
        if (effectId == 0 && effectId != undefined) {
            server.sendDataToAllWithSpawnedEntity(server._characters, this.characterId, "Command.PlayDialogEffect", {
                characterId: this.characterId,
                effectId: effectId
            });
        }
        if (this.isGodMode()) {
            client.character.resourcesUpdater.refresh();
            return;
        }
        if (!server._clients[client.sessionId]) {
            return;
        }
        const hunger = this._resources[enums_1.ResourceIds.HUNGER], hydration = this._resources[enums_1.ResourceIds.HYDRATION], health = this._resources[enums_1.ResourceIds.HEALTH], virus = this._resources[enums_1.ResourceIds.VIRUS], stamina = this._resources[enums_1.ResourceIds.STAMINA], bleeding = this._resources[enums_1.ResourceIds.BLEEDING], energy = this._resources[enums_1.ResourceIds.ENDURANCE], comfort = this._resources[enums_1.ResourceIds.COMFORT], toxicity = this._resources[enums_1.ResourceIds.TOXICITY];
        if (client.character.isRunning &&
            (client.vehicle.mountedVehicle == "" || !client.vehicle.mountedVehicle)) {
            if (server.isSurvival()) {
                client.character._resources[enums_1.ResourceIds.STAMINA] -= 4;
                client.character.isExhausted =
                    client.character._resources[enums_1.ResourceIds.STAMINA] < 120;
            }
        }
        else if (!client.character.isBleeding || !client.character.isMoving) {
            client.character._resources[enums_1.ResourceIds.STAMINA] += 12;
        }
        if (client.character.isSitting &&
            server.isSurvival() &&
            ((0, utils_1.checkConstructionInRange)(server._lootableConstruction, client.character.state.position, 4, enums_1.Items.CAMPFIRE) ||
                (0, utils_1.checkConstructionInRange)(server._worldLootableConstruction, client.character.state.position, 4, enums_1.Items.CAMPFIRE))) {
            client.character._resources[enums_1.ResourceIds.COMFORT] += 30;
        }
        if (server.isSurvival()) {
            client.character._resources[enums_1.ResourceIds.HUNGER] -= 2;
            client.character._resources[enums_1.ResourceIds.ENDURANCE] -= 2;
            client.character._resources[enums_1.ResourceIds.COMFORT] -= 3;
            client.character._resources[enums_1.ResourceIds.HYDRATION] -= 4;
            let desiredEnergyIndicator = "";
            const energyIndicators = [
                enums_1.ResourceIndicators.EXHAUSTED,
                enums_1.ResourceIndicators.VERY_TIRED,
                enums_1.ResourceIndicators.TIRED
            ];
            switch (true) {
                case energy <= 801:
                    server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.TIRED_BUDDY, 1);
                    desiredEnergyIndicator = enums_1.ResourceIndicators.EXHAUSTED;
                    client.character._resources[enums_1.ResourceIds.STAMINA] -= 20;
                    break;
                case energy <= 2601 && energy > 801:
                    server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.TIRED_BUDDY, 1);
                    desiredEnergyIndicator = enums_1.ResourceIndicators.VERY_TIRED;
                    client.character._resources[enums_1.ResourceIds.STAMINA] -= 14;
                    break;
                case energy <= 3501 && energy > 2601:
                    server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.TIRED_BUDDY, 1);
                    desiredEnergyIndicator = enums_1.ResourceIndicators.TIRED;
                    break;
                case energy > 3501:
                    desiredEnergyIndicator = "";
                    break;
                default:
                    desiredEnergyIndicator = "";
                    break;
            }
            let desiredComfortIndicator = "";
            const comfortIndicators = [
                enums_1.ResourceIndicators.COMFORT_PLUS,
                enums_1.ResourceIndicators.COMFORT_PLUSPLUS
            ];
            switch (true) {
                case comfort > 2001:
                    desiredComfortIndicator = enums_1.ResourceIndicators.COMFORT_PLUSPLUS;
                    client.character._resources[enums_1.ResourceIds.HEALTH] += 10;
                    client.character._resources[enums_1.ResourceIds.STAMINA] += 2;
                    break;
                case comfort >= 751 && comfort <= 2001:
                    desiredComfortIndicator = enums_1.ResourceIndicators.COMFORT_PLUS;
                    client.character._resources[enums_1.ResourceIds.HEALTH] += 5;
                    client.character._resources[enums_1.ResourceIds.STAMINA] += 1;
                    break;
                case comfort < 751:
                    desiredComfortIndicator = "";
                    break;
                default:
                    desiredComfortIndicator = "";
                    break;
            }
            this.checkResource(server, enums_1.ResourceIds.ENDURANCE);
            this.checkResource(server, enums_1.ResourceIds.STAMINA);
            this.checkResource(server, enums_1.ResourceIds.COMFORT);
            [...energyIndicators, ...comfortIndicators].forEach((indicator) => {
                const index = this.resourceHudIndicators.indexOf(indicator);
                const desiredIndicator = indicator === desiredEnergyIndicator
                    ? desiredEnergyIndicator
                    : indicator === desiredComfortIndicator
                        ? desiredComfortIndicator
                        : null;
                if (index > -1 && indicator != desiredIndicator) {
                    this.resourceHudIndicators.splice(index, 1);
                    server.sendHudIndicators(client);
                }
                else if (indicator == desiredIndicator && index <= -1) {
                    this.resourceHudIndicators.push(desiredIndicator);
                    server.sendHudIndicators(client);
                }
            });
        }
        let desiredBleedingIndicator = "";
        const bleedingIndicators = [
            enums_1.ResourceIndicators.BLEEDING_LIGHT,
            enums_1.ResourceIndicators.BLEEDING_MODERATE,
            enums_1.ResourceIndicators.BLEEDING_SEVERE
        ];
        switch (true) {
            case bleeding > 0 && bleeding < 30:
                desiredBleedingIndicator = enums_1.ResourceIndicators.BLEEDING_LIGHT;
                break;
            case bleeding >= 30 && bleeding < 60:
                desiredBleedingIndicator = enums_1.ResourceIndicators.BLEEDING_MODERATE;
                break;
            case bleeding >= 60:
                desiredBleedingIndicator = enums_1.ResourceIndicators.BLEEDING_SEVERE;
                break;
            default:
                desiredBleedingIndicator = "";
                break;
        }
        bleedingIndicators.forEach((indicator) => {
            const index = this.resourceHudIndicators.indexOf(indicator);
            if (index > -1 && indicator != desiredBleedingIndicator) {
                this.resourceHudIndicators.splice(index, 1);
                server.sendHudIndicators(client);
            }
            else if (indicator == desiredBleedingIndicator && index <= -1) {
                this.resourceHudIndicators.push(desiredBleedingIndicator);
                server.sendHudIndicators(client);
            }
            const index2 = this.screenEffects.indexOf(indicator);
            if (index2 > -1 && indicator != desiredBleedingIndicator) {
                this.screenEffects.splice(index2, 1);
                server.removeScreenEffect(client, server._screenEffects[indicator]);
            }
            else if (indicator == desiredBleedingIndicator && index2 <= -1) {
                this.screenEffects.push(desiredBleedingIndicator);
                server.addScreenEffect(client, server._screenEffects[desiredBleedingIndicator]);
            }
        });
        if (client.character._resources[enums_1.ResourceIds.BLEEDING] > 0) {
            this.damage(server, {
                entity: "Server.Character.Bleeding",
                damage: Math.ceil(client.character._resources[enums_1.ResourceIds.BLEEDING] / 40) *
                    100
            });
        }
        this.checkResource(server, enums_1.ResourceIds.BLEEDING);
        if (server.isSurvival()) {
            this.checkResource(server, enums_1.ResourceIds.HUNGER, () => {
                this.damage(server, { entity: "Server.Character.Hunger", damage: 100 });
            });
            const indexHunger = this.resourceHudIndicators.indexOf(enums_1.ResourceIndicators.STARVING);
            if (hunger == 0) {
                if (indexHunger <= -1) {
                    this.resourceHudIndicators.push(enums_1.ResourceIndicators.STARVING);
                    server.sendHudIndicators(client);
                }
            }
            else {
                if (indexHunger > -1) {
                    this.resourceHudIndicators.splice(indexHunger, 1);
                    server.sendHudIndicators(client);
                }
            }
            const indexFoodPoison = this.resourceHudIndicators.indexOf(enums_1.ResourceIndicators.FOOD_POISONING);
            if (this.isPoisoned) {
                if (indexFoodPoison <= -1) {
                    this.resourceHudIndicators.push(enums_1.ResourceIndicators.FOOD_POISONING);
                    server.sendHudIndicators(client);
                }
            }
            else {
                if (indexFoodPoison > -1) {
                    this.resourceHudIndicators.splice(indexFoodPoison);
                    server.sendHudIndicators(client);
                }
            }
            const indexCoffeeSugared = this.resourceHudIndicators.indexOf(enums_1.ResourceIndicators.COFFEE_SUGAR);
            if (this.isCoffeeSugared) {
                if (indexCoffeeSugared <= -1) {
                    this.resourceHudIndicators.push(enums_1.ResourceIndicators.COFFEE_SUGAR);
                    server.sendHudIndicators(client);
                }
            }
            else {
                if (indexCoffeeSugared > -1) {
                    this.resourceHudIndicators.splice(indexFoodPoison);
                    server.sendHudIndicators(client);
                }
            }
            this.checkResource(server, enums_1.ResourceIds.HUNGER, () => {
                this.damage(server, { entity: "Server.Character.Hunger", damage: 100 });
            });
            this.checkResource(server, enums_1.ResourceIds.HYDRATION, () => {
                this.damage(server, {
                    entity: "Server.Character.Hydration",
                    damage: 100
                });
            });
            const indexDehydrated = this.resourceHudIndicators.indexOf(enums_1.ResourceIndicators.DEHYDRATED);
            if (hydration == 0) {
                if (indexDehydrated <= -1) {
                    this.resourceHudIndicators.push(enums_1.ResourceIndicators.DEHYDRATED);
                    server.sendHudIndicators(client);
                }
            }
            else {
                if (indexDehydrated > -1) {
                    this.resourceHudIndicators.splice(indexDehydrated, 1);
                    server.sendHudIndicators(client);
                }
            }
        }
        this.checkResource(server, enums_1.ResourceIds.HEALTH);
        this.updateResource(server, client, enums_1.ResourceIds.HEALTH, enums_1.ResourceTypes.HEALTH, health);
        this.updateResource(server, client, enums_1.ResourceIds.BLEEDING, enums_1.ResourceTypes.BLEEDING, bleeding);
        if (server.isSurvival()) {
            this.updateResource(server, client, enums_1.ResourceIds.STAMINA, enums_1.ResourceTypes.STAMINA, stamina);
            this.updateResource(server, client, enums_1.ResourceIds.HUNGER, enums_1.ResourceTypes.HUNGER, hunger);
            this.updateResource(server, client, enums_1.ResourceIds.HYDRATION, enums_1.ResourceTypes.HYDRATION, hydration);
            this.updateResource(server, client, enums_1.ResourceIds.VIRUS, enums_1.ResourceTypes.VIRUS, virus);
            if (stamina <= 2) {
                server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.CARDIO_ISSUES, 1);
            }
            this.updateResource(server, client, enums_1.ResourceIds.ENDURANCE, enums_1.ResourceTypes.ENDURANCE, energy);
            this.updateResource(server, client, enums_1.ResourceIds.COMFORT, enums_1.ResourceTypes.COMFORT, comfort);
        }
        if (server.isBattleRoyale()) {
            this.updateResource(server, client, enums_1.ResourceIds.TOXICITY, enums_1.ResourceTypes.TOXICITY, toxicity);
        }
        client.character.resourcesUpdater.refresh();
    }
    checkResource(server, resourceId, damageCallback) {
        const minValue = resourceId == enums_1.ResourceIds.BLEEDING ? -40 : 0, maxValue = server.getResourceMaxValue(resourceId);
        if (this._resources[resourceId] > maxValue) {
            this._resources[resourceId] = maxValue;
        }
        else if (this._resources[resourceId] < minValue) {
            this._resources[resourceId] =
                minValue + resourceId == enums_1.ResourceIds.ENDURANCE ? 1 : 0;
            if (damageCallback) {
                damageCallback();
            }
        }
    }
    updateResource(server, client, resourceId, resouceType, oldValue) {
        const resource = client.character._resources[resourceId];
        if (resource == oldValue)
            return;
        // only network stamina to other clients
        if (resourceId == enums_1.ResourceIds.STAMINA) {
            server.updateResourceToAllWithSpawnedEntity(client.character.characterId, resource > 0 ? resource : 0, resourceId, resouceType, server._characters);
            return;
        }
        server.updateResource(client, this.characterId, resource > 0 ? resource : 0, resourceId, resouceType);
    }
    isGodMode() {
        return this.godMode || this.tempGodMode;
    }
    clearReloadTimeout() {
        const weaponItem = this.getEquippedWeapon();
        if (!weaponItem || !weaponItem.weapon || !weaponItem.weapon.reloadTimer)
            return;
        clearTimeout(weaponItem.weapon.reloadTimer);
        weaponItem.weapon.reloadTimer = undefined;
    }
    addCombatlogEntry(entry) {
        this.combatlog.push(entry);
        if (this.combatlog.length > 10) {
            this.combatlog.shift();
        }
    }
    getCombatLog() {
        return this.combatlog;
    }
    updateLoadout(server, sendPacketToLocalClient = true) {
        const client = server.getClientByContainerAccessor(this);
        if (!client || !client.character.initialized)
            return;
        if (sendPacketToLocalClient) {
            server.sendData(client, "Loadout.SetLoadoutSlots", this.pGetLoadoutSlots());
        }
        server.sendDataToAllOthersWithSpawnedEntity(server._characters, client, this.characterId, "Loadout.SetLoadoutSlots", this.pGetLoadoutSlots());
        server.sendData(client, "Abilities.SetActivatableAbilityManager", {
            abilities: this.pGetActivatableAbilities(server)
        });
    }
    /**
     * Gets the LightweightPC packet fields for use in SelfSendToClient and AddLightweightPC
     */
    pGetLightweight() {
        return {
            ...super.pGetLightweight(),
            rotation: this.state.lookAt,
            identity: {
                characterName: this.name
            },
            shaderGroupId: this.getShaderGroup()
        };
    }
    pGetLightweightPC(server, client) {
        const vehicleId = client.vehicle.mountedVehicle, vehicle = server._vehicles[vehicleId ?? ""], mountSeatId = vehicle?.getCharacterSeat(this.characterId);
        return {
            ...this.pGetLightweight(),
            mountGuid: vehicleId || "",
            mountSeatId: mountSeatId == -1 ? 0 : mountSeatId,
            mountRelatedDword1: vehicle ? 1 : 0,
            flags1: {
                isAdmin: client.isAdmin ? 1 : 0
            }
        };
    }
    pGetSendSelf(server, client, accountItems) {
        return {
            data: {
                ...this.pGetLightweight(),
                guid: client.loginSessionId,
                hairModel: this.hairModel,
                isRespawning: this.isRespawning,
                gender: this.gender,
                creationDate: this.creationDate,
                lastLoginDate: this.lastLoginDate,
                identity: {
                    characterName: this.name
                },
                inventory: {
                    items: this.pGetInventoryItems(server)
                    //unknownDword1: 2355
                },
                recipes: this.pGetRecipes(server),
                stats: this.getStats(),
                loadoutSlots: this.pGetLoadoutSlots(),
                equipmentSlots: this.pGetEquipment(),
                characterResources: this.pGetResources(),
                containers: this.pGetContainers(server),
                //unknownQword1: this.characterId,
                //unknownDword38: 1,
                //vehicleLoadoutRelatedQword: this.characterId,
                //unknownQword3: this.characterId,
                //vehicleLoadoutRelatedDword: 1,
                //unknownDword40: 1
                isAdmin: client.isAdmin,
                firstPersonOnly: server.isFirstPersonOnly,
                shaderGroupId: this.getShaderGroup(),
                accountItems: {
                    items: accountItems.map((item) => {
                        return {
                            itemId: item.itemGuid,
                            itemData: {
                                itemId: item.itemGuid,
                                itemDefinitionId: item.itemDefinitionId,
                                itemCount: item.stackCount
                            }
                        };
                    })
                }
                //profileId: 270,
                //unknownDword15: 165449,
                //unknownDword16: 260,
                //unknownDword18: 1,
                //unknownDword19: 218,
                //unknownDword26: 2,
                //unknownDword23: 91,
                //unknownDword24: 1,
                //unknownBoolean5: true,
                //dailyRibbonCount: 300,
                //profiles: [
                //  {
                //    profileId: 2,
                //    nameId: 12287,
                //    descriptionId: 12271,
                //    type: 3,
                //    iconId: 171,
                //    unknownByte1: 8,
                //    unknownByte2: 128,
                //    unknownArray1: [],
                //    unknownByte3: 2,
                //    unknownDword7: 1.7,
                //    unknownDword8: 0.95,
                //    unknownDword10: 1,
                //    unknownDword11: 15
                //  }
                //]
            }
        };
    }
    pGetRemoteWeaponData(server, item) {
        const itemDefinition = server.getItemDefinition(item.itemDefinitionId), weaponDefinition = server.getWeaponDefinition(itemDefinition?.PARAM1 ?? 0), firegroups = weaponDefinition.FIRE_GROUPS || [];
        return {
            weaponDefinitionId: weaponDefinition.ID,
            equipmentSlotId: this.getActiveEquipmentSlot(item),
            firegroups: firegroups.map((firegroup) => {
                const firegroupDef = server.getFiregroupDefinition(firegroup.FIRE_GROUP_ID), firemodes = firegroupDef?.FIRE_MODES || [];
                if (!firemodes) {
                    console.error(`firegroupDef missing for`);
                    console.log(firegroup);
                }
                return {
                    firegroupId: firegroup.FIRE_GROUP_ID,
                    unknownArray1: firegroup
                        ? firemodes.map((firemode, j) => {
                            return {
                                unknownDword1: j,
                                unknownDword2: firemode.FIRE_MODE_ID
                            };
                        })
                        : [] // probably firemodes
                };
            })
        };
    }
    pGetRemoteWeaponExtraData(server, item) {
        const itemDefinition = server.getItemDefinition(item.itemDefinitionId), weaponDefinition = server.getWeaponDefinition(itemDefinition?.PARAM1 ?? 0), firegroups = weaponDefinition.FIRE_GROUPS;
        return {
            guid: item.itemGuid,
            unknownByte1: 0, // firegroupIndex (default 0)?
            unknownByte2: 0, // MOST LIKELY firemodeIndex?
            unknownByte3: -1,
            unknownByte4: -1,
            unknownByte5: 1,
            unknownDword1: 0,
            unknownByte6: 0,
            unknownDword2: 0,
            unknownArray1: firegroups.map(() => {
                // same len as firegroups in remoteweapons
                return {
                    // setting unknownDword1 makes the 308 sound when fullpc packet it sent
                    unknownDword1: 0, //firegroup.FIRE_GROUP_ID,
                    unknownBoolean1: false,
                    unknownBoolean2: false
                };
            })
        };
    }
    pGetRemoteWeaponsData(server) {
        const remoteWeapons = [];
        Object.values(this._loadout).forEach((item) => {
            if (server.isWeapon(item.itemDefinitionId)) {
                remoteWeapons.push({
                    guid: item.itemGuid,
                    ...this.pGetRemoteWeaponData(server, item)
                });
            }
        });
        return remoteWeapons;
    }
    pGetRemoteWeaponsExtraData(server) {
        const remoteWeaponsExtra = [];
        Object.values(this._loadout).forEach((item) => {
            if (server.isWeapon(item.itemDefinitionId)) {
                remoteWeaponsExtra.push(this.pGetRemoteWeaponExtraData(server, item));
            }
        });
        return remoteWeaponsExtra;
    }
    pGetContainers(server) {
        if (!this.mountedContainer)
            return super.pGetContainers(server);
        // to avoid a mounted container being dismounted if container list is updated while mounted
        const containers = super.pGetContainers(server), mountedContainer = this.mountedContainer.getContainer();
        if (!mountedContainer)
            return containers;
        /*containers.push({
          loadoutSlotId: mountedContainer.slotId,
          containerData: super.pGetContainerData(server, mountedContainer),
        });*/
        return containers;
    }
    pGetLoadoutSlots() {
        //if (!this.mountedContainer) return super.pGetLoadoutSlots();
        // to avoid a mounted container being dismounted if loadout is updated while mounted
        const loadoutSlots = Object.values(this.getLoadoutSlots()).map((slotId) => {
            return this.pGetLoadoutSlot(slotId);
        });
        //const mountedContainer = this.mountedContainer.getContainer();
        //if (mountedContainer) {}
        /*loadoutSlots.push(
            this.mountedContainer.pGetLoadoutSlot(mountedContainer.slotId)
          );*/
        return {
            characterId: this.characterId,
            loadoutId: this.loadoutId, // needs to be 3
            loadoutData: {
                loadoutSlots: loadoutSlots
            },
            currentSlotId: this.currentLoadoutSlot
        };
    }
    pGetActivatableAbility(slotId, itemDefinition, abilityLineId) {
        const WEAPON_FISTS = enums_1.Items.WEAPON_FISTS, itemDefinitionId = itemDefinition.ID, abilityId = itemDefinition.ACTIVATABLE_ABILITY_ID, abilityEntry = {
            unknownDword1: abilityId,
            unknownDword2: abilityId,
            unknownDword3: 0
        };
        return {
            loadoutSlotId: slotId,
            abilityLineId: abilityLineId,
            unknownArray1: itemDefinitionId == WEAPON_FISTS
                ? [
                    {
                        unknownDword1: 1111278,
                        unknownDword2: 1111278,
                        unknownDword3: 0
                    },
                    abilityEntry
                ]
                : [abilityEntry],
            unknownDword3: 2,
            itemDefinitionId: itemDefinitionId,
            unknownByte: 64
        };
    }
    pGetActivatableAbilities(server) {
        const abilities = [
            {
                loadoutSlotId: 1,
                abilityLineId: 1,
                unknownArray1: [
                    {
                        unknownDword1: 1111164,
                        unknownDword2: 1111164,
                        unknownDword3: 0
                    }
                ],
                unknownDword3: 2,
                itemDefinitionId: 83,
                unknownByte: 64
            }
            // hardcoded one weapon ability to fix fists after respawning
        ];
        let abilityLineId = 1;
        Object.values(this._loadout).forEach((slot) => {
            const itemDefinition = server.getItemDefinition(slot.itemDefinitionId);
            if (!itemDefinition)
                return;
            const { slotId } = slot;
            abilities.push(this.pGetActivatableAbility(slotId, itemDefinition, abilityLineId));
        });
        return abilities;
    }
    resetMetrics() {
        this.metrics.zombiesKilled = 0;
        this.metrics.wildlifeKilled = 0;
        this.metrics.recipesDiscovered = 0;
        this.metrics.startedSurvivingTP = Date.now();
        this.metrics.vehiclesDestroyed = 0;
        this.metrics.playersKilled = 0;
    }
    resetResources(server) {
        this._resources[enums_1.ResourceIds.HEALTH] = 10000;
        this._resources[enums_1.ResourceIds.STAMINA] = 600;
        this._resources[enums_1.ResourceIds.BLEEDING] = 0;
        if (server.isSurvival()) {
            this._resources[enums_1.ResourceIds.HUNGER] = 10000;
            this._resources[enums_1.ResourceIds.HYDRATION] = 10000;
            this._resources[enums_1.ResourceIds.ENDURANCE] = 8000;
            this._resources[enums_1.ResourceIds.VIRUS] = 0;
            this._resources[enums_1.ResourceIds.COMFORT] = 5000;
        }
        if (server.isBattleRoyale()) {
            this._resources[enums_1.ResourceIds.TOXICITY] = 0;
        }
        for (const a in this.healType) {
            const healType = this.healType[a];
            healType.healingTicks = 0;
            healType.healingMaxTicks = 0;
        }
        this.hudIndicators = {};
        this.resourcesUpdater?.refresh();
        const client = server.getClientByCharId(this.characterId);
        if (!client)
            return;
        server.sendHudIndicators(client);
        server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.HEALTH], enums_1.ResourceIds.HEALTH);
        server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.STAMINA], enums_1.ResourceIds.STAMINA);
        server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.BLEEDING], enums_1.ResourceIds.BLEEDING);
        if (server.isSurvival()) {
            server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.HUNGER], enums_1.ResourceIds.HUNGER);
            server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.HYDRATION], enums_1.ResourceIds.HYDRATION);
            server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.ENDURANCE], enums_1.ResourceIds.ENDURANCE);
            server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.VIRUS], enums_1.ResourceIds.VIRUS);
            server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.COMFORT], enums_1.ResourceIds.COMFORT);
        }
        if (server.isBattleRoyale()) {
            server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.TOXICITY], enums_1.ResourceIds.TOXICITY);
        }
    }
    getHealth() {
        return this._resources[enums_1.ResourceIds.HEALTH];
    }
    async damage(server, damageInfo) {
        if (server.isPvE &&
            damageInfo.hitReport?.characterId &&
            server._characters[damageInfo.hitReport?.characterId])
            return;
        const client = server.getClientByCharId(this.characterId);
        if (!client)
            return;
        server.hookManager.checkHook("OnPlayerDamage", client, damageInfo);
        const damage = damageInfo.damage, oldHealth = this._resources[enums_1.ResourceIds.HEALTH];
        if (this.isGodMode() || !this.isAlive || this.isRespawning || damage < 1)
            return;
        // Don't damage players inside shelters
        if (this.isHidden) {
            const entity = server.getConstructionEntity(this.isHidden);
            if (entity instanceof constructionchildentity_1.ConstructionChildEntity ||
                entity instanceof constructionparententity_1.ConstructionParentEntity) {
                if (damage > 0 &&
                    !damageInfo.entity.includes("Server.") &&
                    entity.isInside(this.state.position)) {
                    return;
                }
            }
        }
        if (damageInfo.causeBleed) {
            if ((0, utils_1.randomIntFromInterval)(0, 100) < damage / 100 && damage > 500) {
                this._resources[enums_1.ResourceIds.BLEEDING] += 41;
                if (damage > 4000) {
                    this._resources[enums_1.ResourceIds.BLEEDING] += 41;
                }
                server.updateResourceToAllWithSpawnedEntity(this.characterId, this._resources[enums_1.ResourceIds.BLEEDING], enums_1.ResourceIds.BLEEDING, enums_1.ResourceTypes.BLEEDING, server._characters);
            }
        }
        const sourceEntity = server.getEntity(damageInfo.entity);
        this._resources[enums_1.ResourceIds.HEALTH] -= damage;
        if (this._resources[enums_1.ResourceIds.HEALTH] <= 0) {
            this._resources[enums_1.ResourceIds.HEALTH] = 0;
            server.killCharacter(client, damageInfo);
            if (sourceEntity instanceof Character2016 &&
                sourceEntity.characterId != client.character.characterId) {
                sourceEntity.metrics.playersKilled++;
            }
            else if (sourceEntity instanceof projectileentity_1.ProjectileEntity) {
                const sourceCharacter = server.getEntity(sourceEntity.managerCharacterId);
                if (sourceCharacter instanceof Character2016 &&
                    sourceCharacter.characterId != client.character.characterId) {
                    sourceCharacter.metrics.playersKilled++;
                }
            }
        }
        server.updateResource(client, this.characterId, this._resources[enums_1.ResourceIds.HEALTH], enums_1.ResourceIds.HEALTH);
        const orientation = (0, utils_1.calculateOrientation)(this.state.position, sourceEntity?.state.position || this.state.position // send damaged screen effect during falling/hunger etc
        );
        server.sendData(client, "ClientUpdate.DamageInfo", {
            transientId: 0,
            orientationToSource: orientation,
            unknownDword2: 100
        });
        server.sendChatText(client, `Received ${damage} damage`);
        const damageRecord = await server.generateDamageRecord(this.characterId, damageInfo, oldHealth);
        this.addCombatlogEntry(damageRecord);
        //server.combatLog(client);
        const sourceClient = server.getClientByCharId(damageInfo.entity);
        if (!sourceClient?.character)
            return;
        sourceClient.character.addCombatlogEntry(damageRecord);
        //server.combatLog(sourceClient);
    }
    /**
     *
     * @param {ZoneServer2016} server - The current server
     * @param {BaseLootableEntity} lootableEntity - The lootable entity the player will mount to
     * @returns
     */
    mountContainer(server, lootableEntity) {
        const client = server.getClientByCharId(this.characterId);
        if (!client)
            return;
        const container = lootableEntity.getContainer();
        if (!container) {
            server.containerError(client, enums_1.ContainerErrors.NOT_CONSTRUCTED);
            return;
        }
        if (!(lootableEntity instanceof vehicle_1.Vehicle2016) &&
            !(0, utils_1.isPosInRadius)(lootableEntity.interactionDistance, this.state.position, lootableEntity.state.position)) {
            server.containerError(client, enums_1.ContainerErrors.INTERACTION_VALIDATION);
            return;
        }
        // construction container permissions
        const lootableConstruction = server._lootableConstruction[lootableEntity.characterId];
        if (lootableConstruction && lootableConstruction.parentObjectCharacterId) {
            const parent = lootableConstruction.getParent(server);
            const c = server.getClientByCharId(this.characterId);
            if (parent &&
                parent.isSecured &&
                !c?.isDebugMode &&
                !parent.getHasPermission(server, this.characterId, enums_1.ConstructionPermissionIds.CONTAINERS)) {
                server.containerError(client, enums_1.ContainerErrors.NO_PERMISSION);
                return;
            }
        }
        const oldMount = this.mountedContainer?.characterId;
        // Only mount container if none is mounted
        if (!lootableEntity.mountedCharacter) {
            lootableEntity.mountedCharacter = this.characterId;
        }
        this.mountedContainer = lootableEntity;
        // Change accessor so others in the vehicle will also see the loot but can't interact with
        let accessor = client.character.characterId;
        if (lootableEntity instanceof vehicle_1.Vehicle2016 &&
            server._vehicles[lootableEntity.characterId]) {
            const vehicle = server._vehicles[lootableEntity.characterId], driver = vehicle.getDriver(server);
            if (driver)
                accessor = driver.characterId;
            else
                accessor = "0x0";
        }
        server.sendData(client, "AccessedCharacter.BeginCharacterAccess", {
            objectCharacterId: lootableEntity instanceof vehicle_1.Vehicle2016
                ? lootableEntity.characterId
                : constants_1.EXTERNAL_CONTAINER_GUID,
            mutatorCharacterId: accessor,
            dontOpenInventory: lootableEntity instanceof vehicle_1.Vehicle2016 ? true : !!oldMount,
            itemsData: {
                items: Object.values(container.items).map((item) => {
                    return lootableEntity.pGetItemData(server, item, container.containerDefinitionId);
                }),
                unknownDword1: 92 // idk
            }
        });
        server.initializeContainerList(client, lootableEntity);
        Object.values(lootableEntity._loadout).forEach((item) => {
            server.addItem(client, item, constants_1.LOADOUT_CONTAINER_ID, lootableEntity);
        });
        Object.values(container.items).forEach((item) => {
            server.addItem(client, item, container.containerDefinitionId, lootableEntity);
        });
        server.sendData(client, "Loadout.SetLoadoutSlots", {
            characterId: lootableEntity instanceof vehicle_1.Vehicle2016
                ? lootableEntity.characterId
                : constants_1.EXTERNAL_CONTAINER_GUID,
            loadoutId: lootableEntity instanceof vehicle_1.Vehicle2016 ? lootableEntity.loadoutId : 5,
            loadoutData: {
                loadoutSlots: Object.values(lootableEntity.getLoadoutSlots()).map((slotId) => {
                    return lootableEntity.pGetLoadoutSlot(slotId);
                })
            },
            currentSlotId: lootableEntity.currentLoadoutSlot
        });
        // Just to update available crafting items.
        if (lootableEntity instanceof lootableconstructionentity_1.LootableConstructionEntity) {
            server.sendData(client, "ClientUpdate.ProximateItems", server.getProximityItems(client));
        }
    }
    dismountContainer(server) {
        const client = server.getClientByCharId(this.characterId);
        if (!client || !this.mountedContainer)
            return;
        const container = this.mountedContainer.getContainer();
        if (!container) {
            server.containerError(client, enums_1.ContainerErrors.NOT_CONSTRUCTED);
            return;
        }
        server.deleteItem(this, container.itemGuid);
        Object.values(container.items).forEach((item) => {
            if (!this.mountedContainer)
                return;
            server.deleteItem(this, item.itemGuid);
        });
        if (this.mountedContainer.isLootbag && !utils_1._.size(container.items)) {
            server.deleteEntity(this.mountedContainer.characterId, server._lootbags);
        }
        server.sendData(client, "AccessedCharacter.EndCharacterAccess", {
            characterId: this.mountedContainer.characterId || ""
        });
        delete this.mountedContainer.mountedCharacter;
        delete this.mountedContainer;
        this.updateLoadout(server);
        server.initializeContainerList(client);
    }
    getStats() {
        return stats.map((stat) => {
            return {
                statId: stat.statData.statId,
                statData: {
                    statId: stat.statData.statId,
                    statValue: {
                        type: (0, utils_1.isFloat)(stat.statData.statValue.value.base) ||
                            (0, utils_1.isFloat)(stat.statData.statValue.value.modifier)
                            ? 1
                            : 0,
                        value: {
                            base: stat.statData.statValue.value.base,
                            modifier: stat.statData.statValue.value.modifier
                        }
                    }
                }
            };
        });
    }
    updateEquipmentSlot(server, slotId) {
        if (!server.getClientByCharId(this.characterId)?.character.initialized)
            return;
        server.sendDataToAllWithSpawnedEntity(server._characters, this.characterId, "Equipment.SetCharacterEquipmentSlot", this.pGetEquipmentSlotFull(slotId));
    }
    meleeBlocked(delay = 1000) {
        return this.lastMeleeHitTime + delay >= Date.now();
    }
    checkCurrentInteractionGuid() {
        // mainly for melee workaround (3s timeout)
        if (this.currentInteractionGuid &&
            this.lastInteractionStringTime + 1000 <= Date.now()) {
            this.currentInteractionGuid = "";
        }
    }
    pGetEquipmentSlotFull(slotId) {
        const slot = this._equipment[slotId];
        if (!slot)
            return;
        return {
            characterData: {
                characterId: this.characterId
            },
            equipmentSlot: this.pGetEquipmentSlot(slotId),
            attachmentData: this.pGetAttachmentSlot(slotId)
        };
    }
    updateEquipment(server) {
        if (!server.getClientByCharId(this.characterId)?.character.initialized)
            return;
        server.sendDataToAllWithSpawnedEntity(server._characters, this.characterId, "Equipment.SetCharacterEquipment", this.pGetEquipment());
    }
    pGetEquipment() {
        return {
            characterData: {
                profileId: 5,
                characterId: this.characterId
            },
            unknownDword1: 0,
            tintAlias: "Default",
            decalAlias: "#",
            equipmentSlots: this.pGetEquipmentSlots(),
            attachmentData: this.pGetAttachmentSlots(),
            unknownBoolean1: true
        };
    }
    pGetAttachmentSlots() {
        return Object.keys(this._equipment).map((slotId) => {
            return this.pGetAttachmentSlot(Number(slotId));
        });
    }
    pGetAttachmentSlot(slotId) {
        const slot = this._equipment[slotId];
        let shaderParams = slot?.SHADER_PARAMETER_GROUP ?? [];
        if (slotId == 3 && slot.SHADER_PARAMETER_GROUP?.length == 4) {
            shaderParams =
                this.hoodState == "Down"
                    ? [slot.SHADER_PARAMETER_GROUP[0], slot.SHADER_PARAMETER_GROUP[1]]
                    : [slot.SHADER_PARAMETER_GROUP[2], slot.SHADER_PARAMETER_GROUP[3]];
        }
        return slot
            ? {
                modelName: slot.modelName.replace(/Up|Down/g, this.hoodState == "Down" ? "Up" : "Down"),
                textureAlias: slot.textureAlias || "",
                effectId: slot.effectId || 0,
                tintAlias: slot.tintAlias || "Default",
                decalAlias: slot.decalAlias || "#",
                slotId: slot.slotId,
                SHADER_PARAMETER_GROUP: shaderParams
            }
            : undefined;
    }
    pGetItemData(server, item, containerDefId) {
        return {
            itemDefinitionId: item.itemDefinitionId,
            tintId: 0,
            guid: item.itemGuid,
            count: item.stackCount,
            itemSubData: {
                hasSubData: false
            },
            containerGuid: item.containerGuid,
            containerDefinitionId: containerDefId,
            containerSlotId: item.slotId,
            baseDurability: server.getItemBaseDurability(item.itemDefinitionId),
            currentDurability: item.currentDurability,
            maxDurabilityFromDefinition: server.getItemBaseDurability(item.itemDefinitionId),
            unknownBoolean1: true,
            ownerCharacterId: this.characterId,
            unknownDword9: 1,
            weaponData: this.pGetItemWeaponData(server, item)
        };
    }
    OnFullCharacterDataRequest(server, client) {
        server.sendData(client, "LightweightToFullPc", {
            useCompression: false,
            fullPcData: {
                transientId: this.transientId,
                attachmentData: this.pGetAttachmentSlots(),
                headActor: this.headActor,
                hairModel: this.hairModel,
                resources: { data: this.pGetResources() },
                remoteWeapons: { data: this.pGetRemoteWeaponsData(server) }
            },
            positionUpdate: {
                ...this.positionUpdate,
                sequenceTime: (0, utils_1.getCurrentServerTimeWrapper)().getTruncatedU32(),
                position: this.state.position // trying to fix invisible characters/vehicles until they move
            },
            stats: this.getStats().map((stat) => {
                return stat.statData;
            }),
            remoteWeaponsExtra: this.pGetRemoteWeaponsExtraData(server)
        });
        // needed so all weapons replicate reload and projectile impact
        Object.values(this._loadout).forEach((item) => {
            if (!server.isWeapon(item.itemDefinitionId))
                return;
            server.sendRemoteWeaponUpdateData(client, this.transientId, item.itemGuid, "Update.SwitchFireMode", {
                firegroupIndex: 0,
                firemodeIndex: 0
            });
        });
        server.sendData(client, "Character.WeaponStance", {
            characterId: this.characterId,
            stance: this.weaponStance
        });
        const c = server.getClientByCharId(this.characterId);
        if (c && !c.firstLoading) {
            server.updateCharacterState(client, this.characterId, this.characterStates, false);
            //TODO: This is necessary to fix the audio on other clients. Have to figure out a better way of doing so but currently it works.
            setTimeout(() => {
                c.character.updateFootwearAudio(server);
            }, 7500);
        }
        if (this.onReadyCallback) {
            this.onReadyCallback(client);
            delete this.onReadyCallback;
        }
    }
    applySpecialWeaponEffect(server, client, weaponDefinitionId) {
        switch (weaponDefinitionId) {
            case enums_1.WeaponDefinitionIds.WEAPON_BLAZE:
                server.applyCharacterEffect(this, enums_1.Effects.PFX_Fire_Person_loop, 400, 10000);
                break;
            case enums_1.WeaponDefinitionIds.WEAPON_FROSTBITE:
                const effectTime = 5000;
                if (!this._characterEffects[enums_1.Effects.PFX_Seasonal_Holiday_Snow_skel]) {
                    server.sendData(client, "ClientUpdate.ModifyMovementSpeed", {
                        speed: 0.5
                    });
                    setTimeout(() => {
                        server.sendData(client, "ClientUpdate.ModifyMovementSpeed", {
                            speed: 2
                        });
                    }, effectTime);
                }
                server.applyCharacterEffect(this, enums_1.Effects.PFX_Seasonal_Holiday_Snow_skel, 0, effectTime);
                break;
        }
    }
    OnProjectileHit(server, damageInfo) {
        if (!this.isAlive || server.isPvE)
            return;
        if (server.isHeadshotOnly) {
            switch (damageInfo.hitReport?.hitLocation) {
                case "HEAD":
                case "GLASSES":
                case "NECK":
                    break;
                default:
                    return;
            }
        }
        const itemDefinition = server.getItemDefinition(damageInfo.weapon);
        if (!itemDefinition)
            return;
        const weaponDefinitionId = itemDefinition.PARAM1;
        const sourceClient = server.getClientByCharId(damageInfo.entity), // source
        targetClient = server.getClientByCharId(this.characterId); // target
        if (!sourceClient || !targetClient || !damageInfo.hitReport) {
            return;
        }
        server.fairPlayManager.hitMissFairPlayCheck(server, sourceClient, true, damageInfo.hitReport?.hitLocation || "");
        const hasHelmetBefore = this.hasHelmet(server);
        const hasArmorBefore = this.hasArmor(server);
        let damage = damageInfo.damage, canStopBleed, weaponDmgModifier, headshotDmgMultiplier;
        // these should be configurable
        const weaponDmgModifierDefault = 4, weaponDmgModifierShotgun = 10, weaponDmgModifierSniper = 1, headshotDmgMultiplierDefault = 4, headshotDmgMultiplierShotgun = 2, headshotDmgMultiplierSniper = 6;
        switch (weaponDefinitionId) {
            case enums_1.WeaponDefinitionIds.WEAPON_SHOTGUN:
                weaponDmgModifier = weaponDmgModifierShotgun;
                headshotDmgMultiplier = headshotDmgMultiplierShotgun;
                break;
            case enums_1.WeaponDefinitionIds.WEAPON_308:
                weaponDmgModifier = weaponDmgModifierSniper;
                headshotDmgMultiplier = headshotDmgMultiplierSniper;
                break;
            default:
                weaponDmgModifier = weaponDmgModifierDefault;
                headshotDmgMultiplier = headshotDmgMultiplierDefault;
                break;
        }
        switch (damageInfo.hitReport?.hitLocation) {
            case "HEAD":
            case "GLASSES":
            case "NECK":
                damage = damage *= headshotDmgMultiplier;
                damage = server.applyHelmetDamageReduction(this, damage, 1);
                break;
            default:
                damage = server.applyArmorDamageReduction(this, damage, weaponDmgModifier);
                canStopBleed = true;
                break;
        }
        if (this.isAlive) {
            server.sendHitmarker(sourceClient, damageInfo.hitReport?.hitLocation, this.hasHelmet(server), this.hasArmor(server), hasHelmetBefore, hasArmorBefore);
        }
        this.applySpecialWeaponEffect(server, targetClient, weaponDefinitionId);
        this.damage(server, {
            ...damageInfo,
            damage: damage,
            causeBleed: !(canStopBleed && this.hasArmor(server))
        });
    }
    OnMeleeHit(server, damageInfo) {
        if ((server.isPvE && !server._npcs[damageInfo.entity]) || this.isRespawning)
            return;
        let damage = damageInfo.damage / 2;
        let bleedingChance = 5;
        switch (damageInfo.meleeType) {
            case enums_1.MeleeTypes.BLADE:
                bleedingChance = 35;
                break;
            case enums_1.MeleeTypes.BLUNT:
                bleedingChance = 15;
                break;
            case enums_1.MeleeTypes.FISTS:
                bleedingChance = 5;
                break;
            case enums_1.MeleeTypes.GUITAR:
                bleedingChance = 15;
                break;
            case enums_1.MeleeTypes.KNIFE:
                bleedingChance = 35;
                break;
        }
        switch (damageInfo.hitReport?.hitLocation) {
            case "HEAD":
            case "GLASSES":
            case "NECK":
                damage = server.applyHelmetDamageReduction(this, damage, 1);
                break;
            default:
                damage = server.applyArmorDamageReduction(this, damage, 4);
                break;
        }
        if ((0, utils_1.randomIntFromInterval)(0, 100) <= bleedingChance) {
            this._resources[enums_1.ResourceIds.BLEEDING] += 20;
            server.updateResourceToAllWithSpawnedEntity(this.characterId, this._resources[enums_1.ResourceIds.BLEEDING], enums_1.ResourceIds.BLEEDING, enums_1.ResourceIds.BLEEDING, server._characters);
        }
        this.damage(server, { ...damageInfo, damage });
    }
    OnExplosiveHit(server, sourceEntity) {
        let damage = 10000, weapon, entity;
        switch (true) {
            case sourceEntity instanceof explosiveentity_1.ExplosiveEntity:
                damage = 50000;
                weapon = sourceEntity.itemDefinitionId;
                entity = sourceEntity.characterId;
                break;
            case sourceEntity instanceof projectileentity_1.ProjectileEntity:
                damage = sourceEntity.actorModelId == 0 ? 8000 : 10000;
                weapon = sourceEntity.itemDefinitionId;
                entity = sourceEntity.managerCharacterId;
                break;
            default:
                damage = 10000;
                entity = sourceEntity.characterId;
                break;
        }
        const distance = (0, utils_1.getDistance)(sourceEntity.state.position, this.state.position);
        if (distance > 1)
            damage /= distance;
        this.damage(server, {
            entity: entity,
            weapon: weapon,
            damage: damage
        });
    }
}
exports.Character2016 = Character2016;
