// h1z1-server/out/servers/ZoneServer2016/entities/character.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_test_1 = __importStar(require("node:test"));
const zoneserver_1 = require("../zoneserver");
const node_assert_1 = __importDefault(require("node:assert"));
const test_utils_1 = require("../../../utils/test.utils");
const npc_1 = require("./npc");
process.env.FORCE_DISABLE_WS = "true";
(0, node_test_1.default)("Damage-pve", { timeout: 10000 }, async (t) => {
    const zone = new zoneserver_1.ZoneServer2016(0);
    zone.isPvE = true;
    await zone.start();
    node_assert_1.default.equal(zone.isPvE, true);
    const character = (0, test_utils_1.createFakeCharacter)(zone);
    (0, test_utils_1.createFakeZoneClient)(zone, character);
    const damager = (0, test_utils_1.createFakeCharacter)(zone);
    (0, test_utils_1.createFakeZoneClient)(zone, damager);
    await t.test("Damage from entity", async () => {
        const oldHealth = character.getHealth();
        const dmg = 26;
        const damageInfo = {
            entity: damager.characterId,
            damage: dmg,
            hitReport: {
                characterId: damager.characterId,
                sessionProjectileCount: 0,
                totalShotCount: 0,
                position: new Float32Array([0, 0, 0]),
                unknownByte2: 0,
                unknownFlag1: 0
            }
        };
        character.damage(zone, damageInfo);
        node_assert_1.default.equal(character.getHealth(), oldHealth);
    });
    await zone.stop();
});
(0, node_test_1.default)("Damage-npc-pvp", { timeout: 10000 }, async (t) => {
    const zone = new zoneserver_1.ZoneServer2016(0);
    await zone.start();
    const character = (0, test_utils_1.createFakeCharacter)(zone);
    (0, test_utils_1.createFakeZoneClient)(zone, character);
    const zguid = zone.generateGuid();
    const damager = new npc_1.Npc(zguid, zone.getTransientId(zguid), 9510, new Float32Array([0, 0, 0, 1]), new Float32Array([0, 0, 0, 1]), zone);
    zone._npcs[zguid] = damager;
    await t.test("Damage from entity", async () => {
        const oldHealth = character.getHealth();
        const dmg = 26;
        const damageInfo = {
            entity: damager.characterId,
            damage: dmg,
            hitReport: {
                characterId: damager.characterId,
                sessionProjectileCount: 0,
                totalShotCount: 0,
                position: new Float32Array([0, 0, 0]),
                unknownByte2: 0,
                unknownFlag1: 0
            }
        };
        character.damage(zone, damageInfo);
        node_assert_1.default.equal(character.getHealth(), oldHealth - dmg);
    });
    await zone.stop();
});
(0, node_test_1.default)("Damage-npc-pve", { timeout: 10000 }, async (t) => {
    const zone = new zoneserver_1.ZoneServer2016(0);
    zone.isPvE = true;
    await zone.start();
    node_assert_1.default.equal(zone.isPvE, true);
    const character = (0, test_utils_1.createFakeCharacter)(zone);
    (0, test_utils_1.createFakeZoneClient)(zone, character);
    const zguid = zone.generateGuid();
    const damager = new npc_1.Npc(zguid, zone.getTransientId(zguid), 9510, new Float32Array([0, 0, 0, 1]), new Float32Array([0, 0, 0, 1]), zone);
    zone._npcs[zguid] = damager;
    await t.test("Damage from entity", async () => {
        const oldHealth = character.getHealth();
        const dmg = 26;
        const damageInfo = {
            entity: damager.characterId,
            damage: dmg,
            hitReport: {
                characterId: damager.characterId,
                sessionProjectileCount: 0,
                totalShotCount: 0,
                position: new Float32Array([0, 0, 0]),
                unknownByte2: 0,
                unknownFlag1: 0
            }
        };
        character.damage(zone, damageInfo);
        node_assert_1.default.equal(character.getHealth(), oldHealth - dmg);
    });
    await zone.stop();
});
(0, node_test_1.default)("Damage-pvp", { timeout: 10000 }, async (t) => {
    const zone = new zoneserver_1.ZoneServer2016(0);
    await zone.start();
    node_assert_1.default.equal(zone.isPvE, false);
    const character = (0, test_utils_1.createFakeCharacter)(zone);
    (0, test_utils_1.createFakeZoneClient)(zone, character);
    const damager = (0, test_utils_1.createFakeCharacter)(zone);
    (0, test_utils_1.createFakeZoneClient)(zone, damager);
    await t.test("Damage from entity", async () => {
        const oldHealth = character.getHealth();
        const dmg = 26;
        const damageInfo = {
            entity: damager.characterId,
            damage: dmg,
            hitReport: {
                characterId: damager.characterId,
                sessionProjectileCount: 0,
                totalShotCount: 0,
                position: new Float32Array([0, 0, 0]),
                unknownByte2: 0,
                unknownFlag1: 0
            }
        };
        character.damage(zone, damageInfo);
        node_assert_1.default.equal(character.getHealth(), oldHealth - dmg);
    });
    await zone.stop();
});
(0, node_test_1.after)(() => {
    setImmediate(() => {
        process.exit(0);
    });
});
