// h1z1-server/out/servers/ZoneServer2016/entities/constructiondoor.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructionDoor = void 0;
const doorentity_1 = require("./doorentity");
const enums_1 = require("../models/enums");
const utils_1 = require("../../../utils/utils");
const enums_2 = require("../../../utils/enums");
const explosiveentity_1 = require("./explosiveentity");
function getDamageRange(definitionId) {
    switch (definitionId) {
        case enums_1.Items.METAL_GATE:
            return 4.3;
        case enums_1.Items.DOOR_WOOD:
        case enums_1.Items.DOOR_METAL:
        case enums_1.Items.DOOR_BASIC:
            return 2.5;
        default:
            return 2;
    }
}
function getMaxHealth(itemDefinitionId) {
    switch (itemDefinitionId) {
        case enums_1.Items.METAL_GATE:
            return 1000000;
        case enums_1.Items.DOOR_METAL:
            return 500000;
        case enums_1.Items.DOOR_WOOD:
            return 250000;
        case enums_1.Items.DOOR_BASIC:
            return 150000;
        default:
            return 1000000;
    }
}
class ConstructionDoor extends doorentity_1.DoorEntity {
    /** CharacterId of the player that owns the door entity */
    ownerCharacterId;
    /** Hashed password for the door */
    passwordHash = 0;
    /** List of all accessed players on a door entity: demolish permission also grants access */
    grantedAccess = [];
    /** The parent object the door entity is attached to */
    parentObjectCharacterId;
    /** ItemDefinitionId of the door */
    itemDefinitionId;
    /** Current slot from the parent object the door is attached to */
    slot;
    /** Range at which the door entity will receive damage from explosions */
    damageRange;
    /** Fixed position the door remains at while it's closed */
    fixedPosition;
    /** Time (milliseconds) the door was placed at */
    placementTime = Date.now();
    /** Used to determine whether the parent object is secured - having the door open makes the parent object unsecured */
    isSecured = true;
    /** Used by DecayManager, determines if the entity will be damaged the next decay tick */
    isDecayProtected = false;
    constructor(characterId, transientId, actorModelId, position, rotation, server, itemDefinitionId, ownerCharacterId, parentObjectCharacterId, slot) {
        super(characterId, transientId, actorModelId, position, rotation, server, new Float32Array([1, 1, 1, 1]), 0);
        this.ownerCharacterId = ownerCharacterId;
        this.itemDefinitionId = itemDefinitionId;
        this.parentObjectCharacterId = parentObjectCharacterId;
        this.slot = slot;
        this.profileId = enums_2.CUSTOM_PROFILES_IDS.CONSTRUCTION; /// mark as construction
        this.maxHealth = getMaxHealth(itemDefinitionId);
        this.health = this.maxHealth;
        this.damageRange = getDamageRange(this.itemDefinitionId);
        this.fixedPosition = (0, utils_1.movePoint)(this.state.position, -this.openAngle, this.itemDefinitionId == enums_1.Items.DOOR_METAL ||
            this.itemDefinitionId == enums_1.Items.DOOR_WOOD
            ? 0.625
            : 2.5);
        const itemDefinition = server.getItemDefinition(this.itemDefinitionId);
        if (itemDefinition)
            this.nameId = itemDefinition.NAME_ID;
        this.grantedAccess.push(ownerCharacterId);
        this.npcRenderDistance = 350;
        switch (this.itemDefinitionId) {
            case enums_1.Items.METAL_GATE:
                this.interactionDistance = 5;
        }
    }
    damage(server, damageInfo) {
        this.health -= damageInfo.damage;
        server.updateResourceToAllWithSpawnedEntity(this.characterId, (this.health / this.maxHealth) * 1000000, enums_1.ResourceIds.CONSTRUCTION_CONDITION, enums_1.ResourceTypes.CONDITION, server._constructionDoors);
        if (damageInfo.damage > 0 &&
            !this.getHasPermission(server, damageInfo.entity, enums_1.ConstructionPermissionIds.DEMOLISH)) {
            const timestamp = Date.now();
            const parentFoundation = this.getParentFoundation(server);
            if (parentFoundation)
                parentFoundation.lastDamagedTimestamp = timestamp;
        }
        if (this.health > 0)
            return;
        this.destroy(server, 3000);
    }
    destroy(server, destructTime = 0, slotCooldown = 30000) {
        const deleted = server.deleteEntity(this.characterId, server._constructionDoors, 242, destructTime);
        const parent = this.getParent(server);
        if (!parent)
            return deleted;
        if (parent.freeplaceEntities[this.characterId]) {
            delete parent.freeplaceEntities[this.characterId];
        }
        let slotMap, updateSecured = false;
        switch (this.itemDefinitionId) {
            case enums_1.Items.METAL_GATE:
            case enums_1.Items.DOOR_BASIC:
            case enums_1.Items.DOOR_WOOD:
            case enums_1.Items.DOOR_METAL:
                slotMap = parent.occupiedWallSlots;
                updateSecured = true;
                parent.wallSlotsPlacementTimer[this.getSlotNumber()] = slotCooldown;
                break;
        }
        if (slotMap)
            parent.clearSlot(this.getSlotNumber(), slotMap);
        if (updateSecured)
            parent.updateSecuredState(server);
        return deleted;
    }
    canUndoPlacement(server, client) {
        const weapon = client.character.getEquippedWeapon();
        if (!weapon)
            return false;
        return (this.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.BUILD) &&
            Date.now() < this.placementTime + 120000 &&
            weapon.itemDefinitionId == enums_1.Items.WEAPON_HAMMER_DEMOLITION);
    }
    getParent(server) {
        return (server._constructionSimple[this.parentObjectCharacterId] ||
            server._constructionFoundations[this.parentObjectCharacterId]);
    }
    getParentFoundation(server) {
        const parent = this.getParent(server);
        if (!parent)
            return;
        if (server._constructionSimple[parent.characterId]) {
            return server._constructionSimple[parent.characterId].getParentFoundation(server);
        }
        return server._constructionFoundations[parent.characterId];
    }
    getHasPermission(server, characterId, permission) {
        return (this.getParentFoundation(server)?.getHasPermission(server, characterId, permission) || false);
    }
    getSlotNumber() {
        if (!this.slot)
            return 0;
        return (0, utils_1.getConstructionSlotId)(this.slot);
    }
    OnPlayerSelect(server, client, isInstant) {
        if (this.canUndoPlacement(server, client)) {
            this.destroy(server);
            client.character.lootItem(server, server.generateItem(this.itemDefinitionId));
            return;
        }
        if (isInstant) {
            if (this.passwordHash == 0 ||
                this.grantedAccess.includes(client.character.characterId) ||
                client.character.characterId === this.ownerCharacterId ||
                this.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.CONTAINERS // Container permissions player is able to open constructiondoors
                ) ||
                (client.isAdmin && client.isDebugMode) // debug mode open all doors/gates
            ) {
                if (this.moving) {
                    return;
                }
                this.moving = true;
                // eslint-disable-next-line
                const door = this; // for setTimeout callback
                setTimeout(function () {
                    door.moving = false;
                }, 1000);
                server.sendDataToAllWithSpawnedEntity(server._constructionDoors, this.characterId, "PlayerUpdatePosition", {
                    transientId: this.transientId,
                    positionUpdate: {
                        sequenceTime: 0,
                        unknown3_int8: 0,
                        position: this.state.position,
                        orientation: this.isOpen ? this.closedAngle : this.openAngle
                    }
                });
                server.sendDataToAllWithSpawnedEntity(server._constructionDoors, this.characterId, "Command.PlayDialogEffect", {
                    characterId: this.characterId,
                    effectId: this.isOpen ? this.closeSound : this.openSound
                });
                this.isOpen = !this.isOpen;
                this.isSecured = !this.isOpen;
                const parent = this.getParent(server);
                if (parent) {
                    parent.updateSecuredState(server);
                    // spawn hidden characters immediately after door opens
                    const allowedConstruction = [
                        enums_1.Items.SHELTER,
                        enums_1.Items.SHELTER_LARGE,
                        enums_1.Items.SHELTER_UPPER,
                        enums_1.Items.SHELTER_UPPER_LARGE,
                        enums_1.Items.SHACK,
                        enums_1.Items.SHACK_BASIC,
                        enums_1.Items.SHACK_SMALL
                    ];
                    if (this.isOpen &&
                        allowedConstruction.includes(parent.itemDefinitionId)) {
                        for (const a in server._clients) {
                            const client = server._clients[a];
                            if (client.character.isHidden == parent.characterId)
                                server.constructionManager.constructionPermissionsManager(server, client);
                        }
                    }
                }
                return;
            }
            else {
                server.sendData(client, "Locks.ShowMenu", {
                    characterId: client.character.characterId,
                    unknownDword1: 2,
                    lockType: 2,
                    objectCharacterId: this.characterId
                });
                return;
            }
        }
        else if (!isInstant) {
            if (client.character.characterId === this.ownerCharacterId ||
                this.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.DEMOLISH)) {
                server.sendData(client, "Locks.ShowMenu", {
                    characterId: client.character.characterId,
                    unknownDword1: 2,
                    lockType: 1,
                    objectCharacterId: this.characterId
                });
                return;
            }
            else if (!this.grantedAccess.includes(client.character.characterId)) {
                server.sendData(client, "Locks.ShowMenu", {
                    characterId: client.character.characterId,
                    unknownDword1: 2,
                    lockType: 2,
                    objectCharacterId: this.characterId
                });
            }
        }
    }
    OnInteractionString(server, client) {
        if (this.canUndoPlacement(server, client)) {
            server.constructionManager.undoPlacementInteractionString(server, this, client);
            return;
        }
        if (client.character.characterId === this.ownerCharacterId ||
            this.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.DEMOLISH)) {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.OPEN_AND_LOCK
            });
        }
        else if (!this.grantedAccess.includes(client.character.characterId)) {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.ENTER_ACCESS_CODE
            });
        }
        else {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.OPEN
            });
        }
    }
    OnProjectileHit() {
        // do nothing for now
    }
    OnMeleeHit(server, damageInfo) {
        server.constructionManager.OnMeleeHit(server, damageInfo, this);
    }
    OnExplosiveHit(server, sourceEntity, client, useRaycast) {
        if (server.isPvE) {
            return;
        }
        const itemDefinitionId = sourceEntity instanceof explosiveentity_1.ExplosiveEntity
            ? sourceEntity.itemDefinitionId
            : 0;
        if (!(0, utils_1.isPosInRadius)(this.damageRange, this.fixedPosition ? this.fixedPosition : this.state.position, sourceEntity.state.position)) {
            return;
        }
        if (server.constructionManager.isConstructionInSecuredArea(server, this)) {
            if (useRaycast) {
                if (!client)
                    return;
                server.constructionManager.sendBaseSecuredMessage(server, client);
                return;
            }
            else {
                if (!client)
                    return;
                server.constructionManager.sendBaseSecuredMessage(server, client);
                return;
            }
        }
        server.constructionManager.checkConstructionDamage(server, this, server.baseConstructionDamage, sourceEntity.state.position, this.fixedPosition ? this.fixedPosition : this.state.position, itemDefinitionId);
    }
}
exports.ConstructionDoor = ConstructionDoor;
