// h1z1-server/out/servers/ZoneServer2016/entities/constructionchildentity.js
"use strict";
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructionChildEntity = void 0;
function getRenderDistance(itemDefinitionId) {
    let range = 0;
    switch (itemDefinitionId) {
        case enums_1.Items.SHACK:
        case enums_1.Items.SHACK_SMALL:
        case enums_1.Items.SHACK_BASIC:
        case enums_1.Items.FOUNDATION:
        case enums_1.Items.FOUNDATION_EXPANSION:
        case enums_1.Items.GROUND_TAMPER:
        case enums_1.Items.METAL_DOORWAY:
        case enums_1.Items.METAL_WALL:
        case enums_1.Items.METAL_WALL_UPPER:
        case enums_1.Items.DOOR_BASIC:
        case enums_1.Items.DOOR_METAL:
        case enums_1.Items.DOOR_WOOD:
        case enums_1.Items.SHELTER:
        case enums_1.Items.SHELTER_LARGE:
        case enums_1.Items.SHELTER_UPPER:
        case enums_1.Items.SHELTER_UPPER_LARGE:
        case enums_1.Items.METAL_GATE:
        case enums_1.Items.STRUCTURE_STAIRS:
        case enums_1.Items.STRUCTURE_STAIRS_UPPER:
            range = 420;
            break;
        case enums_1.Items.FURNACE:
        case enums_1.Items.WORKBENCH:
        case enums_1.Items.WORKBENCH_WEAPON:
        case enums_1.Items.BEE_BOX:
        case enums_1.Items.DEW_COLLECTOR:
            range = 420;
            break;
        case enums_1.Items.STORAGE_BOX:
        case enums_1.Items.ANIMAL_TRAP:
            range = 20;
            break;
        default:
            range = 420;
            break;
    }
    return range ? range : 1000;
}
const baselightweightcharacter_1 = require("./baselightweightcharacter");
const enums_1 = require("../models/enums");
const utils_1 = require("../../../utils/utils");
const constructionslots_1 = require("../data/constructionslots");
const constructiondoor_1 = require("./constructiondoor");
const explosiveentity_1 = require("./explosiveentity");
const enums_2 = require("../../../utils/enums");
function getDamageRange(definitionId) {
    switch (definitionId) {
        case enums_1.Items.METAL_WALL:
        case enums_1.Items.METAL_WALL_UPPER:
        case enums_1.Items.SHELTER:
        case enums_1.Items.SHELTER_UPPER:
            return 4.3;
        case enums_1.Items.SHELTER_LARGE:
        case enums_1.Items.SHELTER_UPPER_LARGE:
            return 6.5;
        default:
            return 2;
    }
}
function getMaxHealth(itemDefinitionId) {
    switch (itemDefinitionId) {
        case enums_1.Items.SHELTER:
        case enums_1.Items.SHELTER_LARGE:
        case enums_1.Items.SHELTER_UPPER:
        case enums_1.Items.SHELTER_UPPER_LARGE:
        case enums_1.Items.STRUCTURE_STAIRS:
        case enums_1.Items.STRUCTURE_STAIRS_UPPER:
        case enums_1.Items.LOOKOUT_TOWER:
        case enums_1.Items.METAL_WALL:
        case enums_1.Items.METAL_WALL_UPPER:
        case enums_1.Items.METAL_DOORWAY:
        case enums_1.Items.FOUNDATION_RAMP:
        case enums_1.Items.FOUNDATION_STAIRS:
            return 1000000;
        case enums_1.Items.WORKBENCH:
        case enums_1.Items.WORKBENCH_WEAPON:
            return 500000;
        default:
            return 10000;
    }
}
function getInteractionDistance(itemDefinitionId) {
    switch (itemDefinitionId) {
        case enums_1.Items.SHELTER_LARGE:
        case enums_1.Items.SHELTER_UPPER_LARGE:
        case enums_1.Items.LOOKOUT_TOWER:
        case enums_1.Items.FOUNDATION_RAMP:
        case enums_1.Items.FOUNDATION_STAIRS:
            return 6;
        default:
            return 4;
    }
}
class ConstructionChildEntity extends baselightweightcharacter_1.BaseLightweightCharacter {
    /** Id of the ConstructionChildEntity - See ServerItemDefinitions.json for more information */
    itemDefinitionId;
    /** The parent object the ConstructionChildEntity is attached to */
    parentObjectCharacterId;
    /** Used for manipulating the X, Y, and Z axes for the ConstructionChildEntity */
    eulerAngle;
    /** The sockets the ConstructionChildEntity is occupying */
    slot;
    /** Returns true if the ConstructionChildEntity is secured by a door */
    isSecured;
    /** Range that the ConstructionChildEntity will take damage from explosives */
    damageRange;
    /** Static position of the ConstructionChildEntity */
    fixedPosition;
    /** Time (milliseconds) the ConstructionChildEntity was placed */
    placementTime = Date.now();
    /** 3d boundaries of the space the ConstructionParentEntity occupies (8 vertice points) */
    cubebounds;
    boundsOn;
    /** Time (milliseconds) the player has to undo placement on the ConstructionChildEntity */
    undoPlacementTime = 600000;
    destroyedEffect = enums_1.Effects.PFX_Death_Barricade01;
    /** Used by DecayManager, determines if the entity will be damaged the next decay tick */
    isDecayProtected = false;
    /** FOR DOORS ON SHELTERS / DOORWAYS / LOOKOUT */
    wallSlots = {};
    occupiedWallSlots = {};
    shelterSlotsPlacementTimer = {};
    wallSlotsPlacementTimer = {};
    upperWallSlotsPlacementTimer = {};
    /** FOR UPPER WALL ON WALLS / DOORWAYS */
    upperWallSlots = {};
    occupiedUpperWallSlots = {};
    shelterSlots = {};
    occupiedShelterSlots = {};
    /** Objects that don't occupy any sockets inside of the ConstructionChildEntity,
     * uses CharacterId (string) for indexing */
    freeplaceEntities = {};
    constructor(characterId, transientId, actorModelId, position, rotation, server, itemDefinitionId, parentObjectCharacterId, slot, overrideEulerAngle) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        if (overrideEulerAngle) {
            this.state.rotation = rotation;
            this.eulerAngle = overrideEulerAngle;
        }
        else {
            this.state.rotation = rotation;
            this.eulerAngle = rotation[1];
        }
        this.itemDefinitionId = itemDefinitionId;
        this.parentObjectCharacterId = parentObjectCharacterId;
        this.slot = slot;
        this.profileId = 999; /// mark as construction
        this.damageRange = getDamageRange(this.itemDefinitionId);
        this.isSecured = this.itemDefinitionId == enums_1.Items.METAL_WALL ? true : false;
        this.npcRenderDistance = getRenderDistance(this.itemDefinitionId);
        this.interactionDistance = getInteractionDistance(this.itemDefinitionId);
        this.useSimpleStruct = true;
        this.maxHealth = getMaxHealth(this.itemDefinitionId);
        this.health = this.maxHealth;
        (0, utils_1.registerConstructionSlots)(this, this.wallSlots, constructionslots_1.wallSlotDefinitions);
        Object.seal(this.wallSlots);
        (0, utils_1.registerConstructionSlots)(this, this.upperWallSlots, constructionslots_1.upperWallSlotDefinitions);
        Object.seal(this.upperWallSlots);
        (0, utils_1.registerConstructionSlots)(this, this.shelterSlots, constructionslots_1.shelterSlotDefinitions);
        Object.seal(this.shelterSlots);
        const angle = -this.state.rotation[1];
        switch (itemDefinitionId) {
            case enums_1.Items.SHELTER_LARGE:
            case enums_1.Items.SHELTER_UPPER_LARGE:
                const centerPoint = (0, utils_1.movePoint)(position, angle + (90 * Math.PI) / 180, 2.5);
                this.fixedPosition = centerPoint;
                this.cubebounds = (0, utils_1.getCubeBounds)(centerPoint, 10, 5, angle, position[1], position[1] + 1.8);
                const pos = position[1] + 2.4;
                this.boundsOn = (0, utils_1.getCubeBounds)(centerPoint, 10, 5, angle, pos, pos + 1.8);
                break;
            case enums_1.Items.SHELTER:
            case enums_1.Items.SHELTER_UPPER:
                this.cubebounds = (0, utils_1.getCubeBounds)(position, 5, 5, angle, position[1], position[1] + 1.8);
                const p = position[1] + 2.4;
                this.boundsOn = (0, utils_1.getCubeBounds)(position, 5, 5, angle, p, p + 1.8);
                break;
            case enums_1.Items.METAL_DOORWAY:
            case enums_1.Items.METAL_WALL:
                this.fixedPosition = (0, utils_1.movePoint)(position, -(rotation[1] + 1.575), 2.5);
                break;
        }
        const itemDefinition = server.getItemDefinition(this.itemDefinitionId);
        if (itemDefinition)
            this.nameId = itemDefinition.NAME_ID;
    }
    getOccupiedSlotMaps() {
        return [
            this.occupiedWallSlots,
            this.occupiedUpperWallSlots,
            this.occupiedShelterSlots
        ];
    }
    getSlotPosition(slot, slots) {
        if (typeof slot == "string") {
            slot = (0, utils_1.getConstructionSlotId)(slot);
        }
        if (slot == 101)
            slot = 1; // upper wall slot
        return slots[slot]?.position || undefined;
    }
    getSlotRotation(slot, slots) {
        if (typeof slot == "string") {
            slot = (0, utils_1.getConstructionSlotId)(slot);
        }
        if (slot == 101)
            slot = 1; // upper wall slot
        return slots[slot]?.rotation || undefined;
    }
    updateSecuredState(server) {
        switch (this.itemDefinitionId) {
            case enums_1.Items.METAL_DOORWAY: // for parent foundation
                const door = this.occupiedWallSlots[1];
                if (!door || !(door instanceof constructiondoor_1.ConstructionDoor) || door.isOpen) {
                    this.isSecured = false;
                }
                else {
                    this.isSecured = true;
                }
                const parent = this.getParentFoundation(server);
                if (!parent)
                    return;
                parent.updateSecuredState(server);
                break;
            case enums_1.Items.SHELTER_LARGE:
            case enums_1.Items.SHELTER_UPPER_LARGE:
            case enums_1.Items.SHELTER:
            case enums_1.Items.SHELTER_UPPER:
                const doorslot = this.occupiedWallSlots[1];
                if (!doorslot ||
                    !(doorslot instanceof constructiondoor_1.ConstructionDoor) ||
                    doorslot.isOpen) {
                    this.isSecured = false;
                    return;
                }
                this.isSecured = true;
                break;
            case enums_1.Items.METAL_WALL:
                this.isSecured = true;
                break;
        }
    }
    isSlotOccupied(slotMap, slot) {
        return !!slotMap[slot];
    }
    isSlotsEmpty() {
        return (Object.values(this.occupiedShelterSlots).length +
            Object.values(this.occupiedUpperWallSlots).length +
            Object.values(this.occupiedWallSlots).length ==
            0);
    }
    isSlotValid(slot, definitions, slotMap, itemDefinitionId) {
        if (itemDefinitionId == enums_1.Items.METAL_WALL_UPPER) {
            slot = 1;
        }
        const slots = definitions[this.itemDefinitionId];
        if (!slots) {
            return false;
        }
        if (!slots.authorizedItems.includes(itemDefinitionId)) {
            return false;
        }
        if (!slotMap[slot]) {
            return false;
        }
        return true;
    }
    setSlot(entity, definitions, slotMap, occupiedSlots) {
        const slot = entity.getSlotNumber();
        if (!this.isSlotValid(slot, definitions, slotMap, entity.itemDefinitionId)) {
            return false;
        }
        occupiedSlots[slot] = entity;
        return true;
    }
    clearSlot(slot, occupiedSlots) {
        delete occupiedSlots[slot];
    }
    isWallSlotValid(buildingSlot, itemDefinitionId) {
        let slot = 0;
        if (typeof buildingSlot == "string") {
            slot = (0, utils_1.getConstructionSlotId)(buildingSlot);
        }
        if (slot == 101) {
            // UPPER WALLS
            return this.isSlotValid(1, constructionslots_1.upperWallSlotDefinitions, this.upperWallSlots, itemDefinitionId);
        }
        return this.isSlotValid(slot, constructionslots_1.wallSlotDefinitions, this.wallSlots, itemDefinitionId);
    }
    logSlotInvalid(server, wall, type) {
        if (!server._soloMode) {
            server._db.collection(enums_2.DB_COLLECTIONS.CONSTRUCTION_LOGS).insertOne({
                serverId: server._worldId,
                slotNumber: wall.getSlotNumber(),
                constructionCharId: this.characterId,
                parentCharId: this.parentObjectCharacterId,
                type
            });
        }
    }
    setWallSlot(server, wall) {
        if (wall.itemDefinitionId == enums_1.Items.METAL_WALL_UPPER) {
            return this.setSlot(wall, constructionslots_1.upperWallSlotDefinitions, this.upperWallSlots, this.occupiedUpperWallSlots);
        }
        const set = this.setSlot(wall, constructionslots_1.wallSlotDefinitions, this.wallSlots, this.occupiedWallSlots);
        if (set)
            this.updateSecuredState(server);
        if (!set) {
            this.logSlotInvalid(server, wall, "wall");
        }
        return set;
    }
    isShelterSlotValid(buildingSlot, itemDefinitionId) {
        let slot = 0;
        if (typeof buildingSlot == "string") {
            slot = (0, utils_1.getConstructionSlotId)(buildingSlot);
        }
        return this.isSlotValid(slot, constructionslots_1.shelterSlotDefinitions, this.shelterSlots, itemDefinitionId);
    }
    setShelterSlot(server, shelter) {
        const set = this.setSlot(shelter, constructionslots_1.shelterSlotDefinitions, this.shelterSlots, this.occupiedShelterSlots);
        if (set)
            this.updateSecuredState(server);
        if (!set) {
            this.logSlotInvalid(server, shelter, "shelter");
        }
        return set;
    }
    addFreeplaceConstruction(entity) {
        this.freeplaceEntities[entity.characterId] = entity;
    }
    damage(server, damageInfo) {
        switch (this.itemDefinitionId) {
            case enums_1.Items.FOUNDATION_RAMP:
            case enums_1.Items.FOUNDATION_STAIRS:
                return;
        }
        const dictionary = server.getEntityDictionary(this.characterId);
        if (!dictionary) {
            return;
        }
        this.health -= damageInfo.damage;
        server.sendDataToAllWithSpawnedEntity(dictionary, this.characterId, "Character.UpdateSimpleProxyHealth", this.pGetSimpleProxyHealth());
        const hasPerms = this.getHasPermission(server, damageInfo.entity, enums_1.ConstructionPermissionIds.DEMOLISH);
        if (damageInfo.damage > 0 && !hasPerms) {
            const timestamp = Date.now();
            const parent = this.getParent(server);
            if (parent)
                parent.lastDamagedTimestamp = timestamp;
            const parentFoundation = this.getParentFoundation(server);
            if (parentFoundation)
                parentFoundation.lastDamagedTimestamp = timestamp;
        }
        if (this.health > 0)
            return;
        this.destroy(server, 3000, hasPerms ? 30000 : 0);
    }
    isInside(position) {
        switch (this.itemDefinitionId) {
            case enums_1.Items.SHELTER:
            case enums_1.Items.SHELTER_LARGE:
            case enums_1.Items.SHELTER_UPPER:
            case enums_1.Items.SHELTER_UPPER_LARGE:
                if (!this.cubebounds) {
                    console.error(`ERROR: CONSTRUCTION CUBEBOUNDS IS NOT DEFINED FOR ${this.itemDefinitionId} ${this.characterId}`);
                    return false; // this should never occur
                }
                return (0, utils_1.isInsideCube)(Array.from(position), this.cubebounds);
            default:
                return false;
        }
    }
    isOn(position) {
        switch (this.itemDefinitionId) {
            case enums_1.Items.SHELTER:
            case enums_1.Items.SHELTER_LARGE:
            case enums_1.Items.SHELTER_UPPER:
            case enums_1.Items.SHELTER_UPPER_LARGE:
                if (!this.boundsOn) {
                    console.error(`ERROR: CONSTRUCTION boundsOn IS NOT DEFINED FOR ${this.itemDefinitionId} ${this.characterId}`);
                    return false; // this should never occur
                }
                return (0, utils_1.isInsideCube)(Array.from(position), this.boundsOn);
            default:
                return false;
        }
    }
    destroy(server, destructTime = 0, slotCooldown = 30000) {
        const deleted = server.deleteEntity(this.characterId, server._constructionSimple[this.characterId]
            ? server._constructionSimple
            : server._worldSimpleConstruction, this.destroyedEffect, destructTime);
        const parent = this.getParent(server);
        if (!parent)
            return deleted;
        if (parent.freeplaceEntities[this.characterId]) {
            delete parent.freeplaceEntities[this.characterId];
        }
        let slotMap, updateSecured = false;
        switch (this.itemDefinitionId) {
            case enums_1.Items.METAL_GATE:
            case enums_1.Items.DOOR_BASIC:
            case enums_1.Items.DOOR_WOOD:
            case enums_1.Items.DOOR_METAL:
            case enums_1.Items.METAL_WALL:
            case enums_1.Items.METAL_DOORWAY:
                slotMap = parent.occupiedWallSlots;
                updateSecured = true;
                parent.wallSlotsPlacementTimer[this.getSlotNumber()] = slotCooldown;
                break;
            case enums_1.Items.METAL_WALL_UPPER:
                slotMap = parent.occupiedUpperWallSlots;
                parent.upperWallSlotsPlacementTimer[this.getSlotNumber()] =
                    slotCooldown;
                break;
            case enums_1.Items.SHELTER:
            case enums_1.Items.SHELTER_LARGE:
            case enums_1.Items.SHELTER_UPPER:
            case enums_1.Items.SHELTER_UPPER_LARGE:
            case enums_1.Items.STRUCTURE_STAIRS:
            case enums_1.Items.STRUCTURE_STAIRS_UPPER:
            case enums_1.Items.LOOKOUT_TOWER:
                slotMap = parent.occupiedShelterSlots;
                parent.shelterSlotsPlacementTimer[this.getSlotNumber()] = slotCooldown;
                break;
            case enums_1.Items.FOUNDATION_RAMP:
            case enums_1.Items.FOUNDATION_STAIRS:
                slotMap = parent.occupiedRampSlots;
                break;
        }
        let freeplace = [];
        this.getOccupiedSlotMaps().forEach((slotMap) => {
            freeplace = [...freeplace, ...Object.values(slotMap)];
        });
        if (slotMap)
            parent.clearSlot(this.getSlotNumber(), slotMap);
        if (updateSecured)
            parent.updateSecuredState(server);
        // re-register now disconnected slotted entities as freeplace entities
        const parentFoundation = this.getParentFoundation(server);
        if (parentFoundation) {
            freeplace.forEach((entity) => {
                entity.parentObjectCharacterId = parentFoundation.characterId;
                parentFoundation.freeplaceEntities[entity.characterId] = entity;
            });
            // move free placed entities to parent foundation
            for (const a in this.freeplaceEntities) {
                const freePlacedEntity = this.freeplaceEntities[a];
                freePlacedEntity.parentObjectCharacterId = parentFoundation.characterId;
                parentFoundation.freeplaceEntities[freePlacedEntity.characterId] =
                    freePlacedEntity;
            }
        }
        return deleted;
    }
    getParent(server) {
        return (server._constructionFoundations[this.parentObjectCharacterId] ||
            server._constructionSimple[this.parentObjectCharacterId]);
    }
    canUndoPlacement(server, client) {
        const weapon = client.character.getEquippedWeapon();
        if (!weapon)
            return false;
        return (this.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.BUILD) &&
            Date.now() < this.placementTime + this.undoPlacementTime &&
            weapon.itemDefinitionId == enums_1.Items.WEAPON_HAMMER_DEMOLITION);
    }
    getHasPermission(server, characterId, permission) {
        return (this.getParentFoundation(server)?.getHasPermission(server, characterId, permission) || false);
    }
    getParentFoundation(server) {
        const parent = this.getParent(server);
        if (!parent)
            return;
        if (server._constructionSimple[parent.characterId]) {
            return server._constructionSimple[parent.characterId].getParentFoundation(server);
        }
        return server._constructionFoundations[parent.characterId];
    }
    getSlotNumber() {
        if (!this.slot)
            return 0;
        return (0, utils_1.getConstructionSlotId)(this.slot);
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        if (this.canUndoPlacement(server, client)) {
            this.destroy(server);
            client.character.lootItem(server, server.generateItem(this.itemDefinitionId));
            return;
        }
        if (client.character._resources[enums_1.ResourceIds.ENDURANCE] <= 3501 &&
            this.itemDefinitionId == enums_1.Items.SLEEPING_MAT) {
            server.utilizeHudTimer(client, enums_1.StringIds.RESTING, 20000, 0, () => {
                server.sleep(client);
            });
        }
    }
    OnInteractionString(server, client) {
        if (this.canUndoPlacement(server, client)) {
            server.constructionManager.undoPlacementInteractionString(server, this, client);
            return;
        }
        if (client.character._resources[enums_1.ResourceIds.ENDURANCE] <= 3501 &&
            this.itemDefinitionId == enums_1.Items.SLEEPING_MAT) {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.REST
            });
        }
    }
    OnProjectileHit() {
        // do nothing for now
    }
    OnMeleeHit(server, damageInfo) {
        server.constructionManager.OnMeleeHit(server, damageInfo, this);
    }
    OnExplosiveHit(server, sourceEntity, client, useRaycast) {
        if (server.isPvE) {
            return;
        }
        if (this.itemDefinitionId == enums_1.Items.FOUNDATION_RAMP ||
            this.itemDefinitionId == enums_1.Items.FOUNDATION_STAIRS)
            return;
        const itemDefinitionId = sourceEntity instanceof explosiveentity_1.ExplosiveEntity
            ? sourceEntity.itemDefinitionId
            : 0;
        if (server._worldSimpleConstruction[this.characterId] &&
            (0, utils_1.isPosInRadius)(4, this.state.position, sourceEntity.state.position)) {
            server.constructionManager.checkConstructionDamage(server, this, server.baseConstructionDamage, sourceEntity.state.position, this.state.position, itemDefinitionId);
            return;
        }
        if (!(0, utils_1.isPosInRadius)(this.damageRange * 1.5, this.fixedPosition ? this.fixedPosition : this.state.position, sourceEntity.state.position)) {
            return;
        }
        if (server.constructionManager.isConstructionInSecuredArea(server, this)) {
            if (useRaycast) {
                let damage = server.baseConstructionDamage;
                switch (this.itemDefinitionId) {
                    case enums_1.Items.SHELTER:
                    case enums_1.Items.SHELTER_LARGE:
                    case enums_1.Items.SHELTER_UPPER:
                    case enums_1.Items.SHELTER_UPPER_LARGE:
                        damage *= 30 / 100;
                        break;
                    default:
                        damage = 0;
                        break;
                }
                server.constructionManager.checkConstructionDamage(server, this, damage, sourceEntity.state.position, this.fixedPosition ? this.fixedPosition : this.state.position, itemDefinitionId);
                if (!client)
                    return;
                server.constructionManager.sendBaseSecuredMessage(server, client, 1);
                return;
            }
            else {
                if (!client)
                    return;
                server.constructionManager.sendBaseSecuredMessage(server, client);
                return;
            }
        }
        server.constructionManager.checkConstructionDamage(server, this, server.baseConstructionDamage, sourceEntity.state.position, this.fixedPosition ? this.fixedPosition : this.state.position, itemDefinitionId);
    }
}
exports.ConstructionChildEntity = ConstructionChildEntity;
