// h1z1-server/out/servers/ZoneServer2016/entities/constructionparententity.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructionParentEntity = void 0;
const constructionchildentity_1 = require("./constructionchildentity");
const lootableconstructionentity_1 = require("./lootableconstructionentity");
const enums_1 = require("../models/enums");
const utils_1 = require("../../../utils/utils");
const constructionslots_1 = require("../data/constructionslots");
const explosiveentity_1 = require("./explosiveentity");
function getDamageRange(definitionId) {
    switch (definitionId) {
        case enums_1.Items.SHACK:
            return 4.5;
        case enums_1.Items.SHACK_SMALL:
            return 4;
        case enums_1.Items.SHACK_BASIC:
            return 3;
        default:
            return 4.5;
    }
}
function getMaxHealth(itemDefinitionId) {
    switch (itemDefinitionId) {
        case enums_1.Items.SHACK:
        case enums_1.Items.SHACK_SMALL:
            return 1000000;
        case enums_1.Items.SHACK_BASIC:
            return 250000;
        default:
            return 1000000;
    }
}
class ConstructionParentEntity extends constructionchildentity_1.ConstructionChildEntity {
    /** Allowed permissions for players on the ConstructionParentEntity,
     * determines if a player can visit, use containers, build or demolish*/
    permissions = {};
    /** CharacterId of the player who placed the ConstructionParentEntity */
    ownerCharacterId;
    /** Used by decay manager to determine the amount of ticks a ConstructionParentEntity has spent without any occupied slots */
    ticksWithoutObjects = 0;
    /** Data on deck expansions - includes [slot: number] of position (Float32Array) and rotation (Float32Array)  */
    expansionSlots = {};
    /** HashMap of occupied expansion slots for a deck foundation (1 per side - 4 total),
     * uses slot (number) for indexing
     */
    occupiedExpansionSlots = {};
    /** Data on a ramp - includes: [slot: number] of position (Float32Array) and rotation (Float32Array) */
    rampSlots = {};
    /** HashMap of occupied ramp slots for a deck foundation (3 per side - 12 total),
     * uses slot (number) for indexing
     */
    occupiedRampSlots = {};
    /** Last time the ConstructionParentEntity was damaged */
    lastDamagedTimestamp = 0;
    /** Id of the ConstructionParentEntity - See ServerItemDefinitions.json for more information */
    itemDefinitionId;
    /** Index of the parent slot, used by ConstructionChildEntity - also determines CubeBounds  */
    slot;
    /** Range that the ConstructionParentEntity will take damage from explosives */
    damageRange;
    /** For detecting players / objects underneath a foundation */
    boundsUnder;
    constructor(characterId, transientId, actorModelId, position, rotation, server, itemDefinitionId, ownerCharacterId, ownerName, parentObjectCharacterId, BuildingSlot, overrideEulerAngle) {
        super(characterId, transientId, actorModelId, position, rotation, server, itemDefinitionId, parentObjectCharacterId, "", overrideEulerAngle);
        this.maxHealth = getMaxHealth(itemDefinitionId);
        this.health = this.maxHealth;
        this.ownerCharacterId = ownerCharacterId;
        this.useSimpleStruct = true;
        if (itemDefinitionId != enums_1.Items.FOUNDATION_EXPANSION) {
            const ownerPermission = {
                characterId: ownerCharacterId,
                characterName: ownerName,
                useContainers: true,
                build: true,
                demolish: true,
                visit: true
            };
            this.permissions[ownerPermission.characterId] = ownerPermission;
        }
        this.itemDefinitionId = itemDefinitionId;
        this.parentObjectCharacterId = parentObjectCharacterId;
        this.slot = BuildingSlot || "";
        this.damageRange = getDamageRange(this.itemDefinitionId);
        switch (this.itemDefinitionId) {
            case enums_1.Items.GROUND_TAMPER:
                this.cubebounds = this.getCubeBoundsFromWallSlots(server, [1, 5, 9, 13], 0, 10);
                this.interactionDistance = 16; // fix tamper interact distance if deck is placed on it
                break;
            case enums_1.Items.FOUNDATION:
                this.cubebounds = this.getCubeBoundsFromWallSlots(server, [1, 4, 7, 10], -4, 11);
                this.boundsUnder = this.getCubeBoundsFromWallSlots(server, [1, 4, 7, 10], -4, 2.1);
                this.interactionDistance = 10;
                break;
            case enums_1.Items.FOUNDATION_EXPANSION:
                this.cubebounds = this.getCubeBoundsFromWallSlots(server, [1, 2, 5, 0], -6, 9);
                this.boundsUnder = this.getCubeBoundsFromWallSlots(server, [1, 2, 5, 0], -6, -0.1);
                this.interactionDistance = 8;
                break;
            case enums_1.Items.SHACK:
                this.cubebounds = (0, utils_1.getCubeBounds)(position, 4.7, 5, -this.eulerAngle, position[1] + 0.7, position[1] + 2.8);
                this.interactionDistance = 4;
                break;
            case enums_1.Items.SHACK_SMALL:
                this.cubebounds = (0, utils_1.getCubeBounds)(position, 3.5, 2.5, -this.eulerAngle, position[1] + 0.7, position[1] + 2.8);
                this.interactionDistance = 4;
                break;
            case enums_1.Items.SHACK_BASIC:
                this.cubebounds = (0, utils_1.getCubeBounds)(position, 1.6, 1.6, -this.eulerAngle, position[1], position[1] + 1.7);
                this.interactionDistance = 4;
                break;
        }
        (0, utils_1.registerConstructionSlots)(this, this.wallSlots, constructionslots_1.wallSlotDefinitions);
        Object.seal(this.wallSlots);
        (0, utils_1.registerConstructionSlots)(this, this.expansionSlots, constructionslots_1.foundationExpansionSlotDefinitions);
        Object.seal(this.expansionSlots);
        (0, utils_1.registerConstructionSlots)(this, this.rampSlots, constructionslots_1.foundationRampSlotDefinitions);
        Object.seal(this.rampSlots);
        (0, utils_1.registerConstructionSlots)(this, this.shelterSlots, constructionslots_1.shelterSlotDefinitions);
        Object.seal(this.shelterSlots);
        const itemDefinition = server.getItemDefinition(this.itemDefinitionId);
        if (itemDefinition)
            this.nameId = itemDefinition.NAME_ID;
    }
    getOccupiedSlotMaps() {
        return [
            ...super.getOccupiedSlotMaps(),
            this.occupiedExpansionSlots,
            this.occupiedRampSlots
        ];
    }
    getSquareBounds(slots) {
        const bounds = [
            [0, 0],
            [0, 0],
            [0, 0],
            [0, 0]
        ];
        slots.forEach((slot, idx) => {
            const pos = this.getSlotPosition(slot, this.wallSlots);
            if (pos)
                bounds[idx] = [pos[0], pos[2]];
        });
        return bounds;
    }
    getExpansionDependentWallPos(server) {
        const parent = this.getParentFoundation(server);
        if (!parent)
            return;
        return parent.getSlotPosition(this.getExpansion4thBoundWall(), parent.wallSlots);
    }
    getExpansionSquareBounds(server) {
        // 1, 2, 5, 3RD dependent foundation wall
        const bounds = this.getSquareBounds([1, 2, 5, 0]), pos = this.getExpansionDependentWallPos(server);
        if (!pos)
            return;
        bounds[3] = [pos[0], pos[2]];
        return bounds;
    }
    getCubeBoundsFromWallSlots(server, slots, lowerYOffset, upperYOffset) {
        const bounds = [
            [0, 0, 0],
            [0, 0, 0],
            [0, 0, 0],
            [0, 0, 0],
            [0, 0, 0],
            [0, 0, 0],
            [0, 0, 0],
            [0, 0, 0]
        ], lowerY = this.state.position[1] + lowerYOffset, upperY = this.state.position[1] + upperYOffset;
        for (let i = 0; i < 4; i++) {
            const adjustedIdx = i >= 4 ? i - 4 : i;
            let pos = this.getSlotPosition(slots[adjustedIdx], this.wallSlots);
            if (slots[adjustedIdx] == 0) {
                // for expansion slot #4
                pos = this.getExpansionDependentWallPos(server);
            }
            if (!pos)
                return;
            bounds[i] = [pos[0], lowerY, pos[2]];
            bounds[i + 4] = [pos[0], upperY, pos[2]];
        }
        return bounds;
    }
    getAdjustedShelterSlotId(buildingSlot) {
        let slot = (0, utils_1.getConstructionSlotId)(buildingSlot);
        if (this.itemDefinitionId == enums_1.Items.GROUND_TAMPER) {
            switch (true) {
                case slot >= 11 && slot <= 14:
                    slot = slot - 6;
                    break;
                case slot >= 21 && slot <= 24:
                    slot = slot - 12;
                    break;
                case slot >= 31 && slot <= 34:
                    slot = slot - 18;
                    break;
            }
        }
        else if (this.itemDefinitionId == enums_1.Items.FOUNDATION) {
            switch (true) {
                case slot >= 11 && slot <= 13:
                    slot = slot - 7;
                    break;
                case slot >= 21 && slot <= 23:
                    slot = slot - 14;
                    break;
            }
        }
        return `Structure${slot > 9 ? slot.toString() : `0${slot.toString()}`}`;
    }
    /**
     * [Expansions only] Returns the wall slot on the deck that is required to complete the expansion's square bounds
     */
    getExpansion4thBoundWall() {
        switch (this.getSlotNumber()) {
            case 1:
                return 7;
            case 2:
                return 4;
            case 3:
                return 1;
            case 4:
                return 10;
        }
        return 1;
    }
    /**
     * [Deck expansions only] Returns an array containing the parent foundation walls that a given expansion depends on to be secured.
     * @param expansion The expansion to check.
     */
    getDependentWalls() {
        switch (this.getSlotNumber()) {
            case 1:
                return [4, 5, 6];
            case 2:
                return [1, 2, 3];
            case 3:
                return [10, 11, 12];
            case 4:
                return [7, 8, 9];
        }
        return [];
    }
    getDependentWallsExternal(entity) {
        switch (entity.getSlotNumber()) {
            case 1:
                return [4, 5, 6];
            case 2:
                return [1, 2, 3];
            case 3:
                return [10, 11, 12];
            case 4:
                return [7, 8, 9];
        }
        return [];
    }
    /**
     * [Deck foundations only] Returns the slotId of an expansion on the same side as a given wall.
     * @param expansion The expansion to check.
     */
    getDependentExpansion(slotId) {
        switch (slotId) {
            case 4:
            case 5:
            case 6:
                return 1;
            case 1:
            case 2:
            case 3:
                return 2;
            case 10:
            case 11:
            case 12:
                return 3;
            case 7:
            case 8:
            case 9:
                return 4;
        }
        return 0; // should never get here
    }
    isSideSecure(side) {
        if (this.itemDefinitionId != enums_1.Items.FOUNDATION)
            return false;
        let secure = true;
        Object.keys(this.wallSlots).forEach((slot) => {
            const wall = this.occupiedWallSlots[Number(slot) + 1];
            if (side == this.getDependentExpansion(Number(slot) + 1) &&
                (!wall || !wall.isSecured)) {
                secure = false;
                return;
            }
        });
        return secure;
    }
    /**
     * Tests if all walls slots of this foundation are occupied and secured.
     * @returns boolean
     */
    getWallsSecured() {
        const wallSlots = Object.values(this.occupiedWallSlots);
        // check if all wall slots are occupied
        if (wallSlots.length != Object.values(this.wallSlots).length) {
            return false;
        }
        // check if any walls are gates / if they're open
        for (const wall of wallSlots) {
            if (!wall.isSecured) {
                return false;
            }
        }
        return true;
    }
    updateSecuredState(server) {
        // doesn't work correctly yet -Meme
        /*
        switch(this.itemDefinitionId) {
          case Items.FOUNDATION:
            for (const expansion of Object.values(this.occupiedExpansionSlots)) {
              expansion.updateSecuredState(server);
            }
            
            for(let i = 1; i < 5; i++) {
              const expansion = this.occupiedExpansionSlots[i];
              if(!this.isSideSecure(i) && (!expansion || !expansion.getWallsSecured())) {
                if(expansion) expansion.isSecured = false;
                this.isSecured = false;
                server.sendAlertToAll("NOT SECURE");
                return;
              }
            }
            server.sendAlertToAll("SECURE");
            this.isSecured = true;
            for (const expansion of Object.values(this.occupiedExpansionSlots)) {
              expansion.updateSecuredState(server);
            }
            return;
          case Items.FOUNDATION_EXPANSION:
            const parent =
              server._constructionFoundations[this.parentObjectCharacterId];
            if (parent) {
              if(parent.isSecured && this.getWallsSecured()) {
                this.isSecured = true;
                return;
              }
    
              for (const slot of this.getDependentWalls()) {
                const wall = parent.occupiedWallSlots[slot];
                if (!wall || !wall.isSecured) {
                  this.isSecured = false;
                  return;
                }
              }
            }
            else {
              this.isSecured = false;
              return;
            }
            break;
          default:
            if(!this.getWallsSecured()) {
              this.isSecured = false;
              return;
            }
            break;
        }
        this.isSecured = true;
        */
        // update secured state for all attached expansions
        if (this.itemDefinitionId == enums_1.Items.FOUNDATION_EXPANSION) {
            const parent = server._constructionFoundations[this.parentObjectCharacterId];
            if (parent) {
                parent.updateSecuredState(server);
            }
            return;
        }
        if (this.itemDefinitionId == enums_1.Items.FOUNDATION) {
            for (const a in this.occupiedExpansionSlots) {
                let allowSecured = true;
                const wallSlots = Object.values(this.occupiedExpansionSlots[a].occupiedWallSlots);
                // check if all wall slots are occupied
                if (wallSlots.length !=
                    Object.values(this.occupiedExpansionSlots[a].wallSlots).length) {
                    allowSecured = false;
                }
                // check if any walls are gates / if they're open
                for (const wall of wallSlots) {
                    if (!wall.isSecured) {
                        allowSecured = false;
                    }
                }
                this.occupiedExpansionSlots[a].isSecured = allowSecured;
            }
            // if it ain't broke, don't fix it.
            let side01 = false;
            let side02 = false;
            let side03 = false;
            let side04 = false;
            if (this.occupiedWallSlots["4"] &&
                this.occupiedWallSlots["5"] &&
                this.occupiedWallSlots["6"]) {
                if (this.occupiedWallSlots["4"].isSecured &&
                    this.occupiedWallSlots["5"].isSecured &&
                    this.occupiedWallSlots["6"].isSecured) {
                    side01 = true;
                }
                else if (this.occupiedExpansionSlots["1"] &&
                    this.occupiedExpansionSlots["1"].isSecured)
                    side01 = true;
            }
            else if (this.occupiedExpansionSlots["1"] &&
                this.occupiedExpansionSlots["1"].isSecured)
                side01 = true;
            if (this.occupiedWallSlots["1"] &&
                this.occupiedWallSlots["2"] &&
                this.occupiedWallSlots["3"]) {
                if (this.occupiedWallSlots["1"].isSecured &&
                    this.occupiedWallSlots["2"].isSecured &&
                    this.occupiedWallSlots["3"].isSecured) {
                    side02 = true;
                }
                else if (this.occupiedExpansionSlots["2"] &&
                    this.occupiedExpansionSlots["2"].isSecured)
                    side02 = true;
            }
            else if (this.occupiedExpansionSlots["2"] &&
                this.occupiedExpansionSlots["2"].isSecured)
                side02 = true;
            if (this.occupiedWallSlots["10"] &&
                this.occupiedWallSlots["11"] &&
                this.occupiedWallSlots["12"]) {
                if (this.occupiedWallSlots["10"].isSecured &&
                    this.occupiedWallSlots["11"].isSecured &&
                    this.occupiedWallSlots["12"].isSecured) {
                    side03 = true;
                }
                else if (this.occupiedExpansionSlots["3"] &&
                    this.occupiedExpansionSlots["3"].isSecured)
                    side03 = true;
            }
            else if (this.occupiedExpansionSlots["3"] &&
                this.occupiedExpansionSlots["3"].isSecured)
                side03 = true;
            if (this.occupiedWallSlots["7"] &&
                this.occupiedWallSlots["8"] &&
                this.occupiedWallSlots["9"]) {
                if (this.occupiedWallSlots["7"].isSecured &&
                    this.occupiedWallSlots["8"].isSecured &&
                    this.occupiedWallSlots["9"].isSecured) {
                    side04 = true;
                }
                else if (this.occupiedExpansionSlots["4"] &&
                    this.occupiedExpansionSlots["4"].isSecured)
                    side04 = true;
            }
            else if (this.occupiedExpansionSlots["4"] &&
                this.occupiedExpansionSlots["4"].isSecured)
                side04 = true;
            if (side01 && side02 && side03 && side04) {
                this.isSecured = true;
            }
            else
                this.isSecured = false;
            if (!this.isSecured) {
                if (this.occupiedExpansionSlots["1"]) {
                    for (const slot of this.getDependentWallsExternal(this.occupiedExpansionSlots["1"])) {
                        if (!side02 || !side03 || !side04) {
                            const wall = this.occupiedWallSlots[slot];
                            if (!wall || !wall.isSecured) {
                                this.occupiedExpansionSlots["1"].isSecured = false;
                                break;
                            }
                        }
                    }
                }
                if (this.occupiedExpansionSlots["2"]) {
                    for (const slot of this.getDependentWallsExternal(this.occupiedExpansionSlots["2"])) {
                        if (!side01 || !side03 || !side04) {
                            const wall = this.occupiedWallSlots[slot];
                            if (!wall || !wall.isSecured) {
                                this.occupiedExpansionSlots["2"].isSecured = false;
                                break;
                            }
                        }
                    }
                }
                if (this.occupiedExpansionSlots["3"]) {
                    for (const slot of this.getDependentWallsExternal(this.occupiedExpansionSlots["3"])) {
                        if (!side02 || !side01 || !side04) {
                            const wall = this.occupiedWallSlots[slot];
                            if (!wall || !wall.isSecured) {
                                this.occupiedExpansionSlots["3"].isSecured = false;
                                break;
                            }
                        }
                    }
                }
                if (this.occupiedExpansionSlots["4"]) {
                    for (const slot of this.getDependentWallsExternal(this.occupiedExpansionSlots["4"])) {
                        if (!side02 || !side03 || !side01) {
                            const wall = this.occupiedWallSlots[slot];
                            if (!wall || !wall.isSecured) {
                                this.occupiedExpansionSlots["4"].isSecured = false;
                                break;
                            }
                        }
                    }
                }
            }
            return;
        }
        const wallSlots = Object.values(this.occupiedWallSlots);
        // check if all wall slots are occupied
        if (wallSlots.length != Object.values(this.wallSlots).length) {
            this.isSecured = false;
            return;
        }
        // check if any walls are gates / if they're open
        for (const wall of wallSlots) {
            if (!wall.isSecured) {
                this.isSecured = false;
                return;
            }
        }
        /* TODO: for deck foundations ONLY, need to check each side to see if it's secure,
        and if not, check if the expansion is secure (without checking for dependent deck walls)
        */
        // if this is an expansion, check dependent parent foundation walls
        const parent = server._constructionFoundations[this.parentObjectCharacterId];
        if (parent) {
            for (const slot of this.getDependentWalls()) {
                const wall = parent.occupiedWallSlots[slot];
                if (!wall || !wall.isSecured) {
                    this.isSecured = false;
                    return;
                }
            }
        }
        this.isSecured = true;
    }
    isExpansionSlotValid(buildingSlot, itemDefinitionId) {
        let slot = 0;
        if (typeof buildingSlot == "string") {
            slot = (0, utils_1.getConstructionSlotId)(buildingSlot);
        }
        return this.isSlotValid(slot, constructionslots_1.foundationExpansionSlotDefinitions, this.expansionSlots, itemDefinitionId);
    }
    setExpansionSlot(expansion) {
        return this.setSlot(expansion, constructionslots_1.foundationExpansionSlotDefinitions, this.expansionSlots, this.occupiedExpansionSlots);
    }
    isRampSlotValid(buildingSlot, itemDefinitionId) {
        let slot = 0;
        if (typeof buildingSlot == "string") {
            slot = (0, utils_1.getConstructionSlotId)(buildingSlot);
        }
        return this.isSlotValid(slot, constructionslots_1.foundationRampSlotDefinitions, this.rampSlots, itemDefinitionId);
    }
    setRampSlot(ramp) {
        return this.setSlot(ramp, constructionslots_1.foundationRampSlotDefinitions, this.rampSlots, this.occupiedRampSlots);
    }
    isInside(position) {
        switch (this.itemDefinitionId) {
            case enums_1.Items.FOUNDATION:
            case enums_1.Items.FOUNDATION_EXPANSION:
            case enums_1.Items.GROUND_TAMPER:
            case enums_1.Items.SHACK:
            case enums_1.Items.SHACK_BASIC:
            case enums_1.Items.SHACK_SMALL:
                if (!this.cubebounds) {
                    console.error(`ERROR: CONSTRUCTION CUBE BOUNDS IS NOT DEFINED FOR ${this.itemDefinitionId} ${this.characterId}`);
                    return false; // this should never occur
                }
                return (0, utils_1.isInsideCube)(Array.from(position), this.cubebounds);
            default:
                return false;
        }
    }
    isUnder(position) {
        switch (this.itemDefinitionId) {
            case enums_1.Items.FOUNDATION:
            case enums_1.Items.FOUNDATION_EXPANSION:
                if (!this.boundsUnder) {
                    console.error(`ERROR: CONSTRUCTION BOUNDS UNDER IS NOT DEFINED FOR ${this.itemDefinitionId} ${this.characterId}`);
                    return false; // this should never occur
                }
                return (0, utils_1.isInsideCube)(Array.from(position), this.boundsUnder);
            default:
                return false;
        }
    }
    isOn() {
        // prevents isOn not defined messages if deck is checked by mistake
        // only used for ConstructionChildEntity
        return false;
    }
    destroy(server, destructTime = 0, slotCooldown = 30000) {
        const deleted = server.deleteEntity(this.characterId, server._constructionFoundations, 242, destructTime);
        if (!deleted)
            return false;
        if (this.itemDefinitionId == enums_1.Items.SHACK ||
            this.itemDefinitionId == enums_1.Items.SHACK_SMALL ||
            this.itemDefinitionId == enums_1.Items.SHACK_BASIC) {
            for (const entity of Object.values(this.freeplaceEntities)) {
                if (entity instanceof constructionchildentity_1.ConstructionChildEntity) {
                    server._worldSimpleConstruction[entity.characterId] = entity;
                    delete server._constructionSimple[entity.characterId];
                }
                else if (entity instanceof lootableconstructionentity_1.LootableConstructionEntity) {
                    server._worldLootableConstruction[entity.characterId] = entity;
                    delete server._lootableConstruction[entity.characterId];
                }
            }
        }
        const parent = server._constructionFoundations[this.parentObjectCharacterId];
        if (!parent)
            return deleted;
        if (!this.slot || !this.parentObjectCharacterId)
            return deleted;
        switch (this.itemDefinitionId) {
            case enums_1.Items.METAL_GATE:
            case enums_1.Items.DOOR_BASIC:
            case enums_1.Items.DOOR_WOOD:
            case enums_1.Items.DOOR_METAL:
            case enums_1.Items.METAL_WALL:
            case enums_1.Items.METAL_DOORWAY:
                parent.wallSlotsPlacementTimer[this.getSlotNumber()] = slotCooldown;
                break;
            case enums_1.Items.METAL_WALL_UPPER:
                parent.upperWallSlotsPlacementTimer[this.getSlotNumber()] =
                    slotCooldown;
                break;
            case enums_1.Items.SHELTER:
            case enums_1.Items.SHELTER_LARGE:
            case enums_1.Items.SHELTER_UPPER:
            case enums_1.Items.SHELTER_UPPER_LARGE:
            case enums_1.Items.STRUCTURE_STAIRS:
            case enums_1.Items.STRUCTURE_STAIRS_UPPER:
            case enums_1.Items.LOOKOUT_TOWER:
                parent.shelterSlotsPlacementTimer[this.getSlotNumber()] = slotCooldown;
                break;
            case enums_1.Items.FOUNDATION_RAMP:
            case enums_1.Items.FOUNDATION_STAIRS:
                break;
        }
        parent.clearSlot(this.getSlotNumber(), parent.occupiedExpansionSlots);
        return deleted;
    }
    isExpansionSlotsEmpty() {
        return Object.values(this.occupiedExpansionSlots).length == 0;
    }
    isSlotsEmpty() {
        return (super.isSlotsEmpty() &&
            this.isExpansionSlotsEmpty() &&
            Object.values(this.occupiedRampSlots).length == 0);
    }
    canUndoPlacement(server, client) {
        const weapon = client.character.getEquippedWeapon();
        if (!weapon)
            return false;
        return (this.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.BUILD) &&
            Date.now() < this.placementTime + this.undoPlacementTime &&
            weapon.itemDefinitionId == enums_1.Items.WEAPON_HAMMER_DEMOLITION &&
            this.isSlotsEmpty());
    }
    getHasPermission(server, characterId, permission) {
        if (characterId == this.ownerCharacterId)
            return true;
        if (!this.permissions[characterId])
            return false;
        switch (permission) {
            case enums_1.ConstructionPermissionIds.BUILD:
                return this.permissions[characterId]?.build;
            case enums_1.ConstructionPermissionIds.DEMOLISH:
                return this.permissions[characterId]?.demolish;
            case enums_1.ConstructionPermissionIds.CONTAINERS:
                return this.permissions[characterId]?.useContainers;
            case enums_1.ConstructionPermissionIds.VISIT:
                return this.permissions[characterId]?.visit;
        }
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        if (!this.ownerCharacterId) {
            this.ownerCharacterId = client.character.characterId;
            return;
        }
        if (this.canUndoPlacement(server, client)) {
            this.destroy(server);
            client.character.lootItem(server, server.generateItem(this.itemDefinitionId));
            return;
        }
        if (!this.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.DEMOLISH) && // if the player has permission to build, show the permissions menu
            this.ownerCharacterId != client.character.characterId &&
            (!client.isAdmin || !client.isDebugMode) // allows debug mode
        )
            return;
        server.sendData(client, "NpcFoundationPermissionsManagerBase.ShowPermissions", {
            characterId: this.characterId,
            characterId2: this.characterId,
            permissions: Object.values(this.permissions).filter((perm) => perm.characterId != this.ownerCharacterId)
        });
    }
    OnInteractionString(server, client) {
        if (!this.ownerCharacterId) {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.CLAIM_TARGET
            });
            return;
        }
        if (this.canUndoPlacement(server, client)) {
            server.constructionManager.undoPlacementInteractionString(server, this, client);
            return;
        }
        if (!this.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.DEMOLISH) && // if the player has permission to build, show the permissions menu interaction string
            this.ownerCharacterId != client.character.characterId &&
            (!client.isAdmin || !client.isDebugMode))
            return; // debug mode give permission interact string
        server.sendData(client, "Command.InteractionString", {
            guid: this.characterId,
            stringId: enums_1.StringIds.PERMISSIONS_TARGET
        });
    }
    damage(server, damageInfo) {
        switch (this.itemDefinitionId) {
            case enums_1.Items.FOUNDATION:
            case enums_1.Items.FOUNDATION_EXPANSION:
            case enums_1.Items.GROUND_TAMPER:
                return;
        }
        this.health -= damageInfo.damage;
        server.sendDataToAllWithSpawnedEntity(server._constructionFoundations, this.characterId, "Character.UpdateSimpleProxyHealth", this.pGetSimpleProxyHealth());
        if (damageInfo.damage > 0 &&
            !this.getHasPermission(server, damageInfo.entity, enums_1.ConstructionPermissionIds.DEMOLISH)) {
            const timestamp = Date.now();
            const parent = this.getParent(server);
            if (parent)
                parent.lastDamagedTimestamp = timestamp;
            const parentFoundation = this.getParentFoundation(server);
            if (parentFoundation)
                parentFoundation.lastDamagedTimestamp = timestamp;
        }
        if (this.health > 0)
            return;
        this.destroy(server, 3000);
    }
    OnExplosiveHit(server, sourceEntity) {
        if (server.isPvE) {
            return;
        }
        if (!(0, utils_1.isPosInRadius)(this.damageRange * 1.5, this.state.position, sourceEntity.state.position))
            return;
        const itemDefinitionId = sourceEntity instanceof explosiveentity_1.ExplosiveEntity
            ? sourceEntity.itemDefinitionId
            : 0;
        switch (this.itemDefinitionId) {
            case enums_1.Items.SHACK:
            case enums_1.Items.SHACK_SMALL:
            case enums_1.Items.SHACK_BASIC:
                server.constructionManager.checkConstructionDamage(server, this, server.baseConstructionDamage, sourceEntity.state.position, this.state.position, itemDefinitionId);
        }
    }
}
exports.ConstructionParentEntity = ConstructionParentEntity;
