// h1z1-server/out/servers/ZoneServer2016/entities/destroyable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Destroyable = void 0;
const utils_1 = require("../../../utils/utils");
const enums_1 = require("../models/enums");
const basesimplenpc_1 = require("./basesimplenpc");
function getDestroyedModels(actorModelId) {
    switch (actorModelId) {
        case enums_1.ModelIds.GLASS_WINDOW_01:
            return [
                enums_1.ModelIds.GLASS_WINDOW_DESTROYED_1,
                enums_1.ModelIds.GLASS_WINDOW_DESTROYED_2,
                enums_1.ModelIds.GLASS_WINDOW_DESTROYED_3
            ];
        case enums_1.ModelIds.TINTED_WINDOW_01:
            return [
                enums_1.ModelIds.TINTED_WINDOW_DESTROYED_1,
                enums_1.ModelIds.TINTED_WINDOW_DESTROYED_2,
                enums_1.ModelIds.TINTED_WINDOW_DESTROYED_3
            ];
        default:
            return [];
    }
}
function getMaxHealth(actorModelId) {
    switch (actorModelId) {
        case enums_1.ModelIds.GLASS_WINDOW_01:
        case enums_1.ModelIds.TINTED_WINDOW_01:
            return 2000;
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_PLANK:
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_POSTS_1X1:
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_1X1:
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_POSTS_1X2:
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_GAP_1X1:
        // temp hack so it can't be destroyed ( or you really wanted to )
        // return MAX_UINT32;
        default:
            return 5000;
    }
}
function getDestroyedEffectId(actorModelId) {
    switch (actorModelId) {
        case enums_1.ModelIds.GLASS_WINDOW_01:
        case enums_1.ModelIds.GLASS_WINDOW_DESTROYED_1:
        case enums_1.ModelIds.GLASS_WINDOW_DESTROYED_2:
        case enums_1.ModelIds.GLASS_WINDOW_DESTROYED_3:
        case enums_1.ModelIds.TINTED_WINDOW_01:
        case enums_1.ModelIds.TINTED_WINDOW_DESTROYED_1:
        case enums_1.ModelIds.TINTED_WINDOW_DESTROYED_2:
        case enums_1.ModelIds.TINTED_WINDOW_DESTROYED_3:
            return enums_1.Effects.PFX_Damage_GlassWindow_House;
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_PLANK:
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_POSTS_1X1:
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_1X1:
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_POSTS_1X2:
        case enums_1.ModelIds.FENCES_WOOD_PLANKS_GREY_GAP_1X1:
            return enums_1.Effects.PFX_Damage_Fence_ResidTall01;
        case enums_1.ModelIds.CHAIN_LINK_FENCE_1X1:
        case enums_1.ModelIds.CHAIN_LINK_FENCE_1X2:
            return enums_1.Effects.PFX_Damage_Fence_ChainLink_01;
        case enums_1.ModelIds.BARBED_WIRE_FENCE_1X1:
            return enums_1.Effects.PFX_Damage_Z1_BarbedWireFence1x1;
        case enums_1.ModelIds.BARBED_WIRE_FENCE_1X2:
            return enums_1.Effects.PFX_Damage_Z1_BarbedWireFence1x2;
        case enums_1.ModelIds.FENCE_RANCH_1X1:
        case enums_1.ModelIds.FENCE_RANCH_1X2:
            return enums_1.Effects.PFX_Damage_Fence_Farm01_Post;
        case enums_1.ModelIds.TRAFFIC_SIGNS_STOP_SIGN:
            return enums_1.Effects.PFX_Damage_TrafficSigns_StopSign;
        case enums_1.ModelIds.SIGNS_STOP_SIGN_01:
        case enums_1.ModelIds.SIGNS_STOP_SIGN_01_DESTROYED:
            return enums_1.Effects.PFX_Damage_Sign_Stop;
        case enums_1.ModelIds.TRAFFIC_SIGN_POLE:
        case enums_1.ModelIds.TRAFFIC_SIGN_POLE_DOUBLE:
            return enums_1.Effects.PFX_Damage_TrafficSigns_StreetSign;
        case enums_1.ModelIds.TRAFFIC_SIGNS_TOWN:
            return enums_1.Effects.PFX_Damage_TrafficSigns_Town;
        case enums_1.ModelIds.TRAFFIC_SIGNS_HIGHWAY:
            return enums_1.Effects.PFX_Damage_TrafficSigns_Highway;
        case enums_1.ModelIds.TRAFFIC_SIGNS_HIGHWAY_LARGE_SIGNS_ROAD_TWO_POST_01:
            return enums_1.Effects.PFX_Damage_TrafficSigns_Highway_TwoPosts;
        case enums_1.ModelIds.TRAFFIC_SIGNS_HIGHWAY_LARGE_SIGNS_WOOD_TWO_POST_01:
            return enums_1.Effects.PFX_Damage_TrafficSigns_Highway_Wood_TwoPosts;
        case enums_1.ModelIds.TRAFFIC_SIGN_HIGHWAY_WOOD_BASE:
            return enums_1.Effects.PFX_Damage_TrafficSigns_Highway_Wood_TwoPosts;
        case enums_1.ModelIds.STREET_SIGNS_STREET_NAMES_SIGN:
            return enums_1.Effects.PFX_Damage_TrafficSigns_StreetSign;
        case enums_1.ModelIds.STREET_SIGN_POLE:
            return enums_1.Effects.PFX_Damage_TrafficSigns_StreetSign;
        case enums_1.ModelIds.STREET_SIGNS_STREET_NAMES_SIGN_POST:
            return enums_1.Effects.PFX_Damage_TrafficSigns_StreetSign_StreetName;
        default:
            console.log(`[WARN] No destroyed effect for modelId: ${actorModelId}`);
            return enums_1.Effects.TEST_Null;
    }
}
class Destroyable extends basesimplenpc_1.BaseSimpleNpc {
    spawnerId;
    maxHealth;
    health;
    destroyedModel;
    destroyedModels = [];
    destroyed = false;
    destroyedEffectId;
    constructor(characterId, transientId, actorModelId, position, rotation, server, scale, zoneId, renderDistance) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.spawnerId = zoneId;
        this.scale = scale;
        this.npcRenderDistance = renderDistance;
        this.destroyedEffectId = getDestroyedEffectId(this.actorModelId);
        this.destroyedModels = getDestroyedModels(this.actorModelId);
        this.destroyedModel =
            this.destroyedModels[(this.destroyedModels.length * Math.random()) | 0];
        this.maxHealth = getMaxHealth(actorModelId);
        this.health = this.maxHealth;
    }
    damage(server, damageInfo) {
        this.health -= damageInfo.damage;
        server.sendDataToAllWithSpawnedEntity(server._destroyables, this.characterId, "Character.UpdateSimpleProxyHealth", this.pGetSimpleProxyHealth());
        if (this.health > 0)
            return;
        this.destroy(server, true);
    }
    destroy(server, useDestroyedModel = false) {
        if (!this.destroyed) {
            this.destroyed = true;
            server.sendDataToAllWithSpawnedEntity(server._destroyables, this.characterId, "Character.RemovePlayer", {
                characterId: this.characterId,
                unknownWord1: 1,
                effectId: this.destroyedEffectId
            });
            if (this.destroyedModel && useDestroyedModel) {
                server.sendDataToAllWithSpawnedEntity(server._destroyables, this.characterId, "AddLightweightNpc", this.pGetLightweight());
            }
            return true;
        }
        server.deleteEntity(this.characterId, server._destroyables, enums_1.Effects.PFX_Damage_GlassWindow_House);
        return true;
    }
    OnProjectileHit(server, damageInfo) {
        if (this.destroyed)
            return;
        this.damage(server, damageInfo);
    }
    OnMeleeHit(server, damageInfo) {
        if (this.destroyed)
            return;
        this.damage(server, damageInfo);
        const client = server.getClientByCharId(damageInfo.entity), weapon = client?.character.getEquippedWeapon();
        if (!client || !weapon)
            return;
        const durabilityDamage = server.getDurabilityDamage(weapon.itemDefinitionId);
        server.damageItem(client.character, weapon, durabilityDamage);
    }
    pGetSimpleNpc() {
        return {
            characterId: this.characterId,
            transientId: this.transientId,
            position: this.state.position,
            rotation: this.state.rotation,
            modelId: this.destroyed ? this.destroyedModel : this.actorModelId,
            scale: this.scale,
            health: (this.health / this.maxHealth) * 100
        };
    }
    pGetLightweight() {
        return {
            characterId: this.characterId,
            transientId: this.transientId,
            actorModelId: this.destroyed ? this.destroyedModel : this.actorModelId,
            position: new Float32Array(Array.from(this.state.position).map((pos, idx) => {
                return idx == 1 ? pos++ : pos;
            })),
            rotation: (0, utils_1.eul2quat)(new Float32Array([this.state.rotation[1], 0, 0])),
            scale: this.scale,
            isLightweight: true,
            flags: {
                flags1: {},
                flags2: {},
                flags3: {}
            },
            attachedObject: {}
        };
    }
    OnExplosiveHit(server, sourceEntity) {
        const distance = (0, utils_1.getDistance)(sourceEntity.state.position, this.state.position);
        if (distance > 5)
            return;
        this.damage(server, {
            entity: sourceEntity.characterId,
            damage: this.health
        });
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) { }
}
exports.Destroyable = Destroyable;
