// h1z1-server/out/servers/ZoneServer2016/entities/doorentity.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.DoorEntity = void 0;
const utils_1 = require("../../../utils/utils");
const baselightweightcharacter_1 = require("./baselightweightcharacter");
const enums_1 = require("../models/enums");
function getDestroyedModels(actorModel) {
    switch (actorModel) {
        case enums_1.ModelIds.RESTAURANT_DOOR_1:
            return [
                enums_1.ModelIds.RESTAURANT_DOOR_GLASS_BROKEN_1,
                enums_1.ModelIds.RESTAURANT_DOOR_GLASS_BROKEN_2,
                enums_1.ModelIds.RESTAURANT_DOOR_GLASS_BROKEN_3
            ];
        case enums_1.ModelIds.BUISNESS_DOORS_GLASS:
            return [
                enums_1.ModelIds.BUISNESS_DOORS_GLASS_BROKEN_1,
                enums_1.ModelIds.BUISNESS_DOORS_GLASS_BROKEN_2,
                enums_1.ModelIds.BUISNESS_DOORS_GLASS_BROKEN_3
            ];
        case enums_1.ModelIds.CABIN_DOOR:
            return [enums_1.ModelIds.CABIN_DOOR_BROKEN];
        case enums_1.ModelIds.BUISNESS_DOOR_GLASS:
            return [
                enums_1.ModelIds.BUISNESS_DOOR_GLASS_BROKEN_1,
                enums_1.ModelIds.BUISNESS_DOOR_GLASS_BROKEN_2,
                enums_1.ModelIds.BUISNESS_DOOR_GLASS_BROKEN_3
            ];
        case enums_1.ModelIds.CAMPER_DOOR_1:
            return [
                enums_1.ModelIds.CAMPER_DOOR_GLASS_BROKEN_1,
                enums_1.ModelIds.CAMPER_DOOR_GLASS_BROKEN_2
            ];
        default:
            return [];
    }
}
function getDoorSound(actorModelId) {
    let openSound = enums_1.Effects.SFX_Door_Wood_Open;
    let closeSound = enums_1.Effects.SFX_Door_Wood_Close;
    switch (actorModelId) {
        case enums_1.ModelIds.PLYWOOD_METAL_GATE:
            openSound = enums_1.Effects.SFX_Gate_Placeable_Metal_Open;
            closeSound = enums_1.Effects.SFX_Gate_Placeable_Metal_Close;
            break;
        case enums_1.ModelIds.RESIDENTIAL_DOOR_1:
        case enums_1.ModelIds.RESIDENTIAL_DOOR_2:
        case enums_1.ModelIds.RESIDENTIAL_DOOR_3:
        case enums_1.ModelIds.RESIDENTIAL_DOOR_4:
        case enums_1.ModelIds.RESIDENTIAL_DOOR_5:
        case enums_1.ModelIds.RESIDENTIAL_FRONT_DOOR_1:
        case enums_1.ModelIds.RESIDENTIAL_FRONT_DOOR_2:
        case enums_1.ModelIds.RESIDENTIAL_INTERIOR_DOOR:
        case enums_1.ModelIds.CAMPER_DOOR_1:
        case enums_1.ModelIds.CAMPER_DOOR_GLASS_BROKEN_1:
        case enums_1.ModelIds.CAMPER_DOOR_GLASS_BROKEN_2:
            openSound = enums_1.Effects.SFX_Door_Wood_Open;
            closeSound = enums_1.Effects.SFX_Door_Wood_Close;
            break;
        case enums_1.ModelIds.CHURCH_DOOR:
            openSound = enums_1.Effects.SFX_Door_Wood_Church_Open;
            closeSound = enums_1.Effects.SFX_Door_Wood_Church_Close;
            break;
        case enums_1.ModelIds.OFFICE_DOOR:
        case enums_1.ModelIds.MENS_LEFT_DOOR:
        case enums_1.ModelIds.WOMENS_LEFT_DOOR:
            openSound = enums_1.Effects.SFX_Door_Office_Open;
            closeSound = enums_1.Effects.SFX_Door_Office_Close;
            break;
        case enums_1.ModelIds.FREEZER_DOOR:
            openSound = enums_1.Effects.SFX_Door_Metal_Freezer_Open;
            closeSound = enums_1.Effects.SFX_Door_Metal_Freezer_Close;
            break;
        case enums_1.ModelIds.INDUSTRIAL_DOOR_1:
        case enums_1.ModelIds.INDUSTRIAL_DOOR_2:
        case enums_1.ModelIds.PUSH_DOOR:
            openSound = enums_1.Effects.SFX_Door_Metal_Industrial_Open;
            closeSound = enums_1.Effects.SFX_Door_Metal_Industrial_Close;
            break;
        case enums_1.ModelIds.RESTAURANT_DOOR_GLASS_BROKEN_1:
        case enums_1.ModelIds.RESTAURANT_DOOR_GLASS_BROKEN_2:
        case enums_1.ModelIds.RESTAURANT_DOOR_GLASS_BROKEN_3:
        case enums_1.ModelIds.RESTAURANT_DOOR_1:
            openSound = enums_1.Effects.SFX_Door_Wood_Restaurant_Open;
            closeSound = enums_1.Effects.SFX_Door_Wood_Restaurant_Close;
            break;
        case enums_1.ModelIds.BUISNESS_DOOR_GLASS:
        case enums_1.ModelIds.BUISNESS_DOOR_GLASS_BROKEN_1:
        case enums_1.ModelIds.BUISNESS_DOOR_GLASS_BROKEN_2:
        case enums_1.ModelIds.BUISNESS_DOOR_GLASS_BROKEN_3:
            openSound = enums_1.Effects.SFX_Door_Glass_Business_Open;
            closeSound = enums_1.Effects.SFX_Door_Glass_Business_Close;
            break;
        default:
            break;
    }
    return { openSound, closeSound };
}
class DoorEntity extends baselightweightcharacter_1.BaseLightweightCharacter {
    spawnerId;
    /** Distance (H1Z1 meters) where the DoorEntity will spawn for the player */
    npcRenderDistance = 150;
    openAngle;
    closedAngle;
    /** Default rotation for the  */
    startRot;
    /** Returns true while the door is the process of closing/opening */
    moving = false;
    /** Returns true if the player opens the door */
    isOpen = false;
    openSound;
    closeSound;
    /** Returns true when a window is broken  */
    destroyed = false;
    destroyedModel;
    destroyedModels;
    constructor(characterId, transientId, actorModelId, position, rotation, server, scale, spawnerId) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.positionUpdateType = enums_1.PositionUpdateType.MOVABLE;
        this.flags.projectileCollision = 1;
        this.scale = new Float32Array(scale);
        this.spawnerId = spawnerId;
        this.startRot = rotation;
        this.state.rotation = new Float32Array((0, utils_1.eul2quat)(new Float32Array([rotation[0], rotation[1], rotation[2], rotation[3]])));
        this.openAngle = this.startRot[0] + 1.575;
        this.closedAngle = this.startRot[0];
        const { openSound, closeSound } = getDoorSound(this.actorModelId);
        this.openSound = openSound;
        this.closeSound = closeSound;
        this.destroyedModels = getDestroyedModels(this.actorModelId);
        this.destroyedModel =
            this.destroyedModels[(this.destroyedModels.length * Math.random()) | 0];
        this.health = 2000;
    }
    pGetLightweight() {
        return {
            characterId: this.characterId,
            transientId: this.transientId,
            actorModelId: this.destroyed ? this.destroyedModel : this.actorModelId,
            position: new Float32Array(Array.from(this.state.position).map((pos, idx) => {
                return idx == 1 ? pos++ : pos;
            })),
            rotation: this.state.rotation,
            scale: this.scale,
            positionUpdateType: this.positionUpdateType,
            profileId: this.profileId,
            isLightweight: this.isLightweight,
            flags: {
                flags1: this.flags,
                flags2: this.flags,
                flags3: this.flags
            },
            headActor: this.headActor,
            attachedObject: {}
        };
    }
    OnProjectileHit(server, damageInfo) {
        if (!this.destroyedModel || this.destroyed)
            return;
        this.health -= damageInfo.damage;
        if (this.health > 0)
            return;
        this.destroyed = true;
        server.sendDataToAllWithSpawnedEntity(server._doors, this.characterId, "Character.Destroyed", {
            characterId: this.characterId,
            destroyedModel: this.destroyedModel,
            disableWeirdPhysic: true,
            destroyedEffect2: enums_1.Effects.PFX_Damage_GlassWindow_House
        });
    }
    OnMeleeHit(server, damageInfo) {
        this.OnProjectileHit(server, damageInfo);
    }
    OnPlayerSelect(server, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    client, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    isInstant) {
        if (this.moving) {
            return;
        }
        this.moving = true;
        // eslint-disable-next-line
        const door = this; // for setTimeout callback
        setTimeout(function () {
            door.moving = false;
        }, 1000);
        server.sendDataToAllWithSpawnedEntity(server._doors, this.characterId, "PlayerUpdatePosition", {
            transientId: this.transientId,
            positionUpdate: {
                sequenceTime: 0,
                unknown3_int8: 0,
                position: this.state.position,
                orientation: this.isOpen ? this.closedAngle : this.openAngle
            }
        });
        server.sendDataToAllWithSpawnedEntity(server._doors, this.characterId, "Command.PlayDialogEffect", {
            characterId: this.characterId,
            effectId: this.isOpen ? this.closeSound : this.openSound
        });
        this.isOpen = !this.isOpen;
    }
    OnInteractionString(server, client) {
        server.sendData(client, "Command.InteractionString", {
            guid: this.characterId,
            stringId: enums_1.StringIds.USE_DOOR
        });
    }
    destroy(server) {
        return server.deleteEntity(this.characterId, server._doors);
    }
}
exports.DoorEntity = DoorEntity;
