// h1z1-server/out/servers/ZoneServer2016/entities/explosiveentity.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExplosiveEntity = void 0;
const utils_1 = require("../../../utils/utils");
const enums_1 = require("../models/enums");
const baselightweightcharacter_1 = require("./baselightweightcharacter");
const promises_1 = require("node:timers/promises");
class ExplosiveEntity extends baselightweightcharacter_1.BaseLightweightCharacter {
    /** Id of the item - See ServerItemDefinitions.json for more information */
    itemDefinitionId;
    /** The distance where the explosive will render for the player */
    npcRenderDistance = 300;
    /** Returns true upon explosion of the explosive */
    detonated = false;
    /** Used for shooting explosion with projectiles, 1 for IEDS and a coinflip between 1 and 2 for biofuel/ethanol */
    triggerExplosionShots = this.isLandmine() || this.isIED() ? 1 : Math.floor(Math.random() * 2) + 1;
    /** the characterId from who place this to keep track */
    ownerCharacterId;
    creationTime = Date.now();
    isAwaitingExplosion = false;
    constructor(characterId, transientId, actorModelId, position, rotation, server, itemDefinitionId, ownerCharacterId = "") {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.itemDefinitionId = itemDefinitionId;
        this.ownerCharacterId = ownerCharacterId;
    }
    isIED() {
        return this.itemDefinitionId == enums_1.Items.IED;
    }
    isLandmine() {
        return this.itemDefinitionId == enums_1.Items.LANDMINE;
    }
    ignite(server, client) {
        if (!this.isIED()) {
            return;
        }
        const pos = this.state.position;
        server.sendDataToAllWithSpawnedEntity(server._explosives, this.characterId, "Character.PlayWorldCompositeEffect", {
            characterId: this.characterId,
            effectId: enums_1.Effects.PFX_Fire_Lighter,
            position: new Float32Array([pos[0], pos[1] + 0.1, pos[2], 1]),
            effectTime: 8
        });
        server.sendDataToAllWithSpawnedEntity(server._explosives, this.characterId, "Character.PlayWorldCompositeEffect", {
            characterId: this.characterId,
            effectId: enums_1.Effects.EFX_Candle_Flame_01,
            position: new Float32Array([pos[0], pos[1] + 0.1, pos[2], 1]),
            effectTime: 8
        });
        setTimeout(() => {
            this.detonate(client.character.characterId);
        }, 10000);
    }
    detonate(characterId = "") {
        let client = this.server.getClientByCharId(characterId);
        if (!this.server._explosives[this.characterId] || this.detonated)
            return;
        this.detonated = true;
        this.server.sendCompositeEffectToAllInRange(600, "", this.state.position, enums_1.Effects.PFX_Impact_Explosion_Landmine_Dirt_10m);
        if ((0, utils_1.isChristmasSeason)()) {
            this.server.sendCompositeEffectToAllInRange(600, "", this.state.position, enums_1.Effects.PFX_Seasonal_Holiday_Snow_skel);
        }
        queueMicrotask(() => {
            this.server.deleteEntity(this.characterId, this.server._explosives);
        });
        this.server.explosionDamage(this, client);
    }
    /** Used by landmines to arm their explosivenss */
    async arm(server) {
        // Wait 10 seconds before activating the trap
        await promises_1.scheduler.wait(10_000);
        server.aiManager.addEntity(this);
    }
    /** Called when the explosive gets hit by a projectile (bullet, arrow, etc.) */
    OnProjectileHit(server, damageInfo) {
        this.triggerExplosionShots -= 1;
        if (damageInfo.weapon == enums_1.Items.WEAPON_SHOTGUN ||
            damageInfo.weapon == enums_1.Items.WEAPON_NAGAFENS_RAGE) {
            // prevent shotguns one shotting gas cans
            const randomInt = (0, utils_1.randomIntFromInterval)(0, 100);
            if (randomInt < 90)
                this.triggerExplosionShots += 1;
        }
        if (this.triggerExplosionShots > 0)
            return;
        this.detonate(damageInfo.entity);
    }
    async OnExplosiveHit(server, sourceEntity, client, waitTime = 0, useRaycast = false) {
        if (useRaycast && this.isAwaitingExplosion)
            return;
        this.isAwaitingExplosion = true;
        if (this.characterId == sourceEntity.characterId)
            return;
        if (!useRaycast) {
            if ((0, utils_1.getDistance)(sourceEntity.state.position, this.state.position) >= 2)
                return;
        }
        await promises_1.scheduler.wait(waitTime);
        if (server._spawnedItems[this.characterId]) {
            const itemObject = server._spawnedItems[this.characterId];
            server.deleteEntity(this.characterId, server._spawnedItems);
            delete server.worldObjectManager.spawnedLootObjects[itemObject.spawnerId];
        }
        if (this.detonated)
            return;
        this.detonate(client?.character.characterId);
    }
    destroy() {
        this.detonate();
        return true;
    }
}
exports.ExplosiveEntity = ExplosiveEntity;
