// h1z1-server/out/servers/ZoneServer2016/entities/lootableconstructionentity.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.LootableConstructionEntity = void 0;
const enums_1 = require("../models/enums");
const baselootableentity_1 = require("./baselootableentity");
const smeltingentity_1 = require("../classes/smeltingentity");
const loadouts_1 = require("../data/loadouts");
const collectingentity_1 = require("../classes/collectingentity");
const constants_1 = require("../../../utils/constants");
const promises_1 = require("timers/promises");
const utils_1 = require("../../../utils/utils");
const explosiveentity_1 = require("./explosiveentity");
function getMaxHealth(itemDefinitionId) {
    switch (itemDefinitionId) {
        case enums_1.Items.CAMPFIRE:
        case enums_1.Items.DEW_COLLECTOR:
            return 100000;
        default:
            return 250000;
    }
}
class LootableConstructionEntity extends baselootableentity_1.BaseLootableEntity {
    /** Time (milliseconds) when the LootableConstructionEntity was placed */
    placementTime = Date.now();
    /** Parent foundation that the LootableConstructionEntity is on top of */
    parentObjectCharacterId;
    loadoutId = 5;
    /** Id of the LootableConstructionId - See ServerItemDefinitions.json for more information */
    itemDefinitionId;
    /** Range at which the LootableConstructionEntity will receive damage from explosions */
    damageRange = 1.5;
    /** Distance (H1Z1 meters) at which the player can interact with the LootableConstructionEntity */
    interactionDistance = 3;
    /** Determines if the LootableConstructionEntity is a SmeltingEntity or CollectingEntity */
    subEntity;
    /** Used by DecayManager, determines if the entity will be damaged the next decay tick */
    isDecayProtected = false;
    isProp = false;
    constructor(characterId, transientId, actorModelId, position, rotation, server, scale, itemDefinitionId, parentObjectCharacterId, subEntityType, isProp = false) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.parentObjectCharacterId = parentObjectCharacterId || "";
        this.itemDefinitionId = itemDefinitionId;
        const itemDefinition = server.getItemDefinition(itemDefinitionId);
        if (itemDefinition)
            this.nameId = itemDefinition.NAME_ID;
        this.profileId = 999; /// mark as construction
        this.isProp = isProp;
        this.maxHealth = getMaxHealth(this.itemDefinitionId);
        this.health = this.maxHealth;
        switch (this.itemDefinitionId) {
            case enums_1.Items.REPAIR_BOX:
                this.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.repair_box;
                break;
            case enums_1.Items.HAND_SHOVEL:
                this.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.stash;
                break;
            default:
                this.defaultLoadout = loadouts_1.lootableContainerDefaultLoadouts.storage;
                break;
        }
        if (subEntityType === "SmeltingEntity") {
            this.subEntity = new smeltingentity_1.SmeltingEntity(this, server);
            this.npcRenderDistance = 250;
        }
        else if (subEntityType === "CollectingEntity") {
            this.subEntity = new collectingentity_1.CollectingEntity(this, server);
            this.interactionDistance = 5;
            this.npcRenderDistance = 200;
        }
        else {
            this.npcRenderDistance = 20;
        }
        this.scale = scale;
    }
    damage(server, damageInfo) {
        const dictionary = server.getEntityDictionary(this.characterId);
        if (!dictionary) {
            return;
        }
        this.health -= damageInfo.damage;
        server.sendDataToAllWithSpawnedEntity(dictionary, this.characterId, "Character.UpdateSimpleProxyHealth", this.pGetSimpleProxyHealth());
        if (this.health > 0)
            return;
        this.destroy(server, 3000);
    }
    getParent(server) {
        return (server._constructionFoundations[this.parentObjectCharacterId] ||
            server._constructionSimple[this.parentObjectCharacterId] ||
            undefined);
    }
    getParentFoundation(server) {
        const parent = this.getParent(server);
        if (!parent)
            return;
        if (server._constructionSimple[parent.characterId]) {
            return server._constructionSimple[parent.characterId].getParentFoundation(server);
        }
        return server._constructionFoundations[parent.characterId];
    }
    canUndoPlacement(server, client) {
        const weapon = client.character.getEquippedWeapon();
        if (!weapon)
            return false;
        return (this.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.BUILD) &&
            Date.now() < this.placementTime + 120000 &&
            weapon.itemDefinitionId == enums_1.Items.WEAPON_HAMMER_DEMOLITION);
    }
    destroy(server, destructTime = 0) {
        const deleted = server.deleteEntity(this.characterId, server._lootableConstruction[this.characterId]
            ? server._lootableConstruction
            : server._worldLootableConstruction, 242, destructTime);
        const parent = this.getParent(server);
        if (parent && parent.freeplaceEntities[this.characterId]) {
            delete parent.freeplaceEntities[this.characterId];
        }
        server.worldObjectManager.createLootbag(server, this);
        const container = this.getContainer();
        if (container) {
            container.items = {};
            for (const a in server._characters) {
                const character = server._characters[a];
                if (character.mountedContainer == this) {
                    character.dismountContainer(server);
                }
            }
        }
        return deleted;
    }
    getHasPermission(server, characterId, permission) {
        return (this.getParentFoundation(server)?.getHasPermission(server, characterId, permission) || false);
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        if (this.canUndoPlacement(server, client)) {
            this.destroy(server);
            client.character.lootItem(server, server.generateItem(this.itemDefinitionId));
            return;
        }
        if (server.fairPlayManager.useFairPlay) {
            for (const a in server._constructionFoundations) {
                const foundation = server._constructionFoundations[a];
                if (foundation.itemDefinitionId != enums_1.Items.FOUNDATION &&
                    foundation.itemDefinitionId != enums_1.Items.FOUNDATION_EXPANSION)
                    continue;
                if (foundation.isInside(this.state.position)) {
                    let pos = foundation.state.position[1];
                    if (foundation.parentObjectCharacterId &&
                        server._constructionFoundations[foundation.parentObjectCharacterId]) {
                        pos =
                            server._constructionFoundations[foundation.parentObjectCharacterId].state.position[1];
                    }
                    if (this.state.position[1] - (pos + 2) < 0) {
                        server.sendChatText(client, `FairPlay: Glitched storage detected`);
                        return;
                    }
                }
            }
        }
        super.OnPlayerSelect(server, client);
        if (this.itemDefinitionId == enums_1.Items.REPAIR_BOX) {
            server.sendData(client, "Character.DailyRepairMaterials", {
                characterId: this.characterId,
                containerId: constants_1.EXTERNAL_CONTAINER_GUID,
                materials: server.decayManager.dailyRepairMaterials
            });
        }
    }
    OnInteractionString(server, client) {
        if (this.canUndoPlacement(server, client)) {
            server.constructionManager.undoPlacementInteractionString(server, this, client);
            return;
        }
        if (this.subEntity) {
            this.subEntity.OnInteractionString(server, client);
            return;
        }
        server.sendData(client, "Command.InteractionString", {
            guid: this.characterId,
            stringId: enums_1.StringIds.OPEN
        });
    }
    OnFullCharacterDataRequest(server, client) {
        if (this.subEntity) {
            this.subEntity.OnFullCharacterDataRequest(server, client);
        }
    }
    async handleBeeboxSwarm(server, damageInfo) {
        const dictionary = server.getEntityDictionary(this.characterId);
        if (!dictionary)
            return;
        // Find all clients within the radius
        const clientsInRange = Object.values(server._clients).filter((c) => {
            if (!c.character || !c.character.isAlive)
                return false;
            return (0, utils_1.isPosInRadius)(5, this.state.position, c.character.state.position);
        });
        //  Bee FX for all affected clients
        for (const c of clientsInRange) {
            server.sendDataToAllWithSpawnedEntity(dictionary, this.characterId, "Character.PlayWorldCompositeEffect", {
                characterId: this.characterId,
                effectId: enums_1.Effects.PFX_Bee_Swarm_Attack,
                position: c.character.state.position,
                effectTime: 5
            });
        }
        // Only show HUD and deal damage if bees are angry
        if (damageInfo && damageInfo.damage > 0) {
            let hudIndicator = server._hudIndicators[enums_1.ResourceIndicators.BEES];
            if (hudIndicator) {
                for (const c of clientsInRange) {
                    if (c.character.hudIndicators[hudIndicator.typeName]) {
                        c.character.hudIndicators[hudIndicator.typeName].expirationTime +=
                            5000;
                    }
                    else {
                        c.character.hudIndicators[hudIndicator.typeName] = {
                            typeName: hudIndicator.typeName,
                            expirationTime: Date.now() + 5000
                        };
                        server.sendHudIndicators(c);
                    }
                }
            }
            // Damage all affected clients over time
            for (let i = 0; i < 12; i++) {
                for (const c of clientsInRange) {
                    const dmgInfo = {
                        entity: c.character.characterId,
                        damage: 20
                    };
                    c.character.damage(server, dmgInfo);
                }
                await promises_1.scheduler.wait(500);
            }
        }
    }
    OnMeleeHit(server, damageInfo) {
        if (this.isProp)
            return;
        if (this.itemDefinitionId == enums_1.Items.BEE_BOX &&
            damageInfo.weapon != enums_1.Items.WEAPON_HAMMER_DEMOLITION) {
            this.handleBeeboxSwarm(server, damageInfo);
        }
        server.constructionManager.OnMeleeHit(server, damageInfo, this);
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    OnProjectileHit(server, damageInfo) {
        if (this.isProp)
            return;
        if (this.itemDefinitionId == enums_1.Items.BEE_BOX) {
            this.handleBeeboxSwarm(server, damageInfo);
        }
        // ...existing or future projectile damage logic...
    }
    OnExplosiveHit(server, sourceEntity, client, useRaycast) {
        if (server.isPvE) {
            return;
        }
        if (!(0, utils_1.isPosInRadius)(2, this.state.position, sourceEntity.state.position))
            return;
        const itemDefinitionId = sourceEntity instanceof explosiveentity_1.ExplosiveEntity
            ? sourceEntity.itemDefinitionId
            : 0;
        if (server._worldLootableConstruction[this.characterId]) {
            server.constructionManager.checkConstructionDamage(server, this, server.baseConstructionDamage, sourceEntity.state.position, this.state.position, itemDefinitionId);
            return;
        }
        if (!useRaycast) {
            const parent = this.getParent(server);
            if (parent && parent.isSecured) {
                if (!client)
                    return;
                server.constructionManager.sendBaseSecuredMessage(server, client);
                return;
            }
        }
        server.constructionManager.checkConstructionDamage(server, this, server.baseConstructionDamage, sourceEntity.state.position, this.state.position, itemDefinitionId);
    }
}
exports.LootableConstructionEntity = LootableConstructionEntity;
