// h1z1-server/out/servers/ZoneServer2016/entities/lootableprop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LootableProp = void 0;
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2022 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
const baselootableentity_1 = require("./baselootableentity");
const enums_1 = require("../models/enums");
const utils_1 = require("../../../utils/utils");
const challengemanager_1 = require("../managers/challengemanager");
function getContainerAndTime(entity) {
    switch (entity.actorModelId) {
        case enums_1.ModelIds.WRECKED_VAN:
            entity.containerId = enums_1.Items.CONTAINER_WRECKED_VAN;
            entity.searchTime = 1500;
            entity.lootSpawner = "Wrecked Van";
            break;
        case enums_1.ModelIds.WRECKED_CAR:
        case enums_1.ModelIds.WRECKED_CAR_OD_01:
            entity.containerId = enums_1.Items.CONTAINER_WRECKED_CAR;
            entity.searchTime = 1500;
            entity.lootSpawner = "Wrecked Car";
            break;
        case enums_1.ModelIds.WRECKED_TRUCK_01:
        case enums_1.ModelIds.WRECKED_TRUCK_OD_01:
            entity.containerId = enums_1.Items.CONTAINER_WRECKED_TRUCK;
            entity.searchTime = 1500;
            entity.lootSpawner = "Wrecked Truck";
            break;
        case enums_1.ModelIds.LOCKERS_LOCKER_WEAPONS:
            entity.containerId = enums_1.Items.CONTAINER_WEAPONS_LOCKER;
            entity.searchTime = 500;
            entity.lootSpawner = "Weapons Locker";
            break;
        case enums_1.ModelIds.LOCKERS_LOCKER_POLICE:
            entity.containerId = enums_1.Items.CONTAINER_LOCKER;
            entity.searchTime = 500;
            entity.lootSpawner = "Locker";
            break;
        case enums_1.ModelIds.OFFICE_CABINETS_02:
        case enums_1.ModelIds.DESK:
            entity.containerId = enums_1.Items.CONTAINER_DESK;
            entity.searchTime = 500;
            entity.lootSpawner = "Desk";
            break;
        case enums_1.ModelIds.LAB_GLASS_CABINET:
            entity.containerId = enums_1.Items.CONTAINER_CABINETS;
            entity.searchTime = 700;
            entity.lootSpawner = "Cabinets";
            break;
        case enums_1.ModelIds.OFFICE_CUBE_CABINET:
            entity.containerId = enums_1.Items.CONTAINER_CABINETS_CUBE;
            entity.searchTime = 500;
            entity.lootSpawner = "Cabinets Cube";
            break;
        case enums_1.ModelIds.CABINET_SET_03:
        case enums_1.ModelIds.CABINET_SET_05:
        case enums_1.ModelIds.CABINET_SET_06:
        case enums_1.ModelIds.CABINET_SET_07:
        case enums_1.ModelIds.CABINET_SET_08:
        case enums_1.ModelIds.CABINET_SET_09:
            entity.containerId = enums_1.Items.CONTAINER_CABINETS_KITCHEN;
            entity.searchTime = 5000;
            entity.lootSpawner = "Cabinets Kitchen";
            break;
        case enums_1.ModelIds.CABINETS_KITCHEN_02:
        case enums_1.ModelIds.CABINETS_KITCHEN_03:
        case enums_1.ModelIds.CABINETS_KITCHEN_05:
        case enums_1.ModelIds.CABINETS_KITCHEN_04:
        case enums_1.ModelIds.CABINETS_KITCHEN_01:
        case enums_1.ModelIds.CABINETS_KITCHEN_06:
        case enums_1.ModelIds.CABINET_SET_01:
        case enums_1.ModelIds.CABINET_SET_02:
        case enums_1.ModelIds.CABINET_SET_04:
        case enums_1.ModelIds.CABINET_SET_10:
        case enums_1.ModelIds.CABINET_SET_11:
            entity.containerId = enums_1.Items.CONTAINER_CABINETS_KITCHEN;
            entity.searchTime = 1000;
            entity.lootSpawner = "Cabinets Kitchen";
            break;
        case enums_1.ModelIds.CABINETS_BATHROOM_01:
        case enums_1.ModelIds.CABINETS_BATHROOM_SINK:
        case enums_1.ModelIds.CABINETS_BATHROOM_02:
            entity.containerId = enums_1.Items.CONTAINER_CABINETS_BATHROOM;
            entity.searchTime = 500;
            entity.lootSpawner = "Cabinets Bathroom";
            break;
        case enums_1.ModelIds.BLUE_TOOL_CABINET:
        case enums_1.ModelIds.RED_SILVER_TOOL_CABINET:
        case enums_1.ModelIds.TOOL_CABINET_01:
        case enums_1.ModelIds.TOOL_CABINET_02:
            entity.containerId = enums_1.Items.CONTAINER_TOOL_CABINETS;
            entity.searchTime = 500;
            entity.lootSpawner = "Tool Cabinet";
            break;
        case enums_1.ModelIds.DUMPSTER:
            entity.containerId = enums_1.Items.CONTAINER_DUMPSTER;
            entity.searchTime = 500;
            entity.lootSpawner = "Dumpster";
            break;
        case enums_1.ModelIds.OFFICE_CABINETS_01:
        case enums_1.ModelIds.OFFICE_CABINETS_03:
        case enums_1.ModelIds.OFFICE_FILING_CABINET:
        case enums_1.ModelIds.FILE_CABINETS:
            entity.containerId = enums_1.Items.CONTAINER_FILE_CABINET;
            entity.searchTime = 500;
            entity.lootSpawner = "File Cabinet";
            break;
        case enums_1.ModelIds.KITCHEN_FRIDGE:
            entity.containerId = enums_1.Items.CONTAINER_FRIDGE;
            entity.searchTime = 500;
            entity.lootSpawner = "Fridge";
            break;
        case enums_1.ModelIds.WASHER:
            entity.containerId = enums_1.Items.CONTAINER_WASHING_MACHINE;
            entity.searchTime = 500;
            entity.lootSpawner = "Washer";
            break;
        case enums_1.ModelIds.DRYER:
            entity.containerId = enums_1.Items.CONTAINER_DRYER;
            entity.searchTime = 500;
            entity.lootSpawner = "Dryer";
            break;
        case enums_1.ModelIds.OTTOMAN_01:
        case enums_1.ModelIds.OTTOMAN_02:
            entity.containerId = enums_1.Items.CONTAINER_OTTOMAN;
            entity.searchTime = 500;
            entity.lootSpawner = "Ottoman";
            break;
        case enums_1.ModelIds.DRESSER_01:
        case enums_1.ModelIds.DRESSER_02:
            entity.containerId = enums_1.Items.CONTAINER_DRESSER;
            entity.searchTime = 500;
            entity.lootSpawner = "Dresser";
            break;
        case enums_1.ModelIds.ARMOIRE_02:
        case enums_1.ModelIds.ARMOIRE_01:
            entity.containerId = enums_1.Items.CONTAINER_ARMOIRE;
            entity.searchTime = 500;
            entity.lootSpawner = "Armoire";
            break;
        case enums_1.ModelIds.GARBAGE_CAN_01:
            entity.containerId = enums_1.Items.CONTAINER_GARBAGE_CAN;
            entity.searchTime = 500;
            entity.lootSpawner = "Garbage Can";
            break;
        case enums_1.ModelIds.HOSPITAL_DRUG_CABNINET:
            entity.containerId = enums_1.Items.CONTAINER_DRUG_CABINET;
            entity.searchTime = 1000;
            entity.lootSpawner = "Drug Cabinets";
            break;
        case enums_1.ModelIds.HOSPITAL_LAB_WORKBENCH:
            entity.containerId = enums_1.Items.CONTAINER_MEDICAL_STATION;
            entity.searchTime = 0;
            entity.lootSpawner = "Medical Station";
            break;
        case enums_1.ModelIds.HOSPITAL_DESK:
        case enums_1.ModelIds.HOSPITAL_LAB_COUNTERTOP:
            entity.containerId = enums_1.Items.CONTAINER_HOSPITAL_DESK;
            entity.searchTime = 1000;
            entity.lootSpawner = "Hospital Desk";
            break;
        case enums_1.ModelIds.HOSPITAL_GROSSING_STATION:
            entity.containerId = enums_1.Items.CONTAINER_GROSSING_STATION;
            entity.searchTime = 1000;
            entity.lootSpawner = "Grossing Station";
            break;
        case enums_1.ModelIds.HOSPITAL_REFRIGERATOR:
            entity.containerId = enums_1.Items.CONTAINER_HOSPITAL_REFRIGERATOR;
            entity.searchTime = 1000;
            entity.lootSpawner = "Hospital Refrigerator";
            break;
        case enums_1.ModelIds.HOSPITAL_SINK_COUNTERTOP:
            entity.containerId = enums_1.Items.CONTAINER_HOSPITAL_CABINET;
            entity.searchTime = 1000;
            entity.lootSpawner = "Hospital Cabinets";
            break;
        case enums_1.ModelIds.TREASURE_CHEST:
            entity.containerId = enums_1.Items.CONTAINER_LOOT_CACHE;
            entity.nameId = enums_1.Items.CONTAINER_LOOT_CACHE;
            entity.searchTime = 1000;
        case enums_1.ModelIds.METAL_STORAGE_CHEST:
            entity.containerId = enums_1.Items.CONTAINER_STORAGE;
            entity.searchTime = 500;
            entity.lootSpawner = "Dresser";
            break;
        case enums_1.ModelIds.FURNACE:
            entity.containerId = enums_1.Items.CONTAINER_FURNACE;
            entity.searchTime = 0;
            break;
        case enums_1.ModelIds.CAMPFIRE:
            entity.containerId = enums_1.Items.CONTAINER_CAMPFIRE;
            entity.searchTime = 0;
            break;
        case enums_1.ModelIds.BARBEQUE:
            entity.containerId = enums_1.Items.BARBEQUE;
            entity.searchTime = 0;
            break;
        default:
            entity.containerId = enums_1.Items.CONTAINER_STORAGE;
            entity.searchTime = 500;
            return;
    }
}
class LootableProp extends baselootableentity_1.BaseLootableEntity {
    spawnerId;
    npcRenderDistance = 150;
    containerId = enums_1.Items.CONTAINER_STORAGE;
    /** Determines the loot table to distribute to the LootableProp */
    lootSpawner = "Wrecked Car";
    /** Time (milliseconds) it takes before the container loads for the player */
    searchTime = 1000;
    constructor(characterId, transientId, actorModelId, position, rotation, server, scale, spawnerId, renderDistance) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.scale = new Float32Array(scale);
        this.spawnerId = spawnerId;
        this.npcRenderDistance = renderDistance;
        this.loadoutId = 5;
        getContainerAndTime(this);
        /*switch (this.lootSpawner) {
          case "Wrecked Van":
          case "Wrecked Car":
          case "Wrecked Truck":
            this.useSimpleStruct = false;
            this.state.rotation = eul2quat(
              new Float32Array([
                this.state.rotation[1],
                this.state.rotation[0],
                this.state.rotation[2],
                0
              ])
            );
            break;
        }*/
    }
    pGetSimpleNpc() {
        return {
            characterId: this.characterId,
            transientId: this.transientId,
            position: this.state.position,
            rotation: this.state.rotation,
            modelId: this.actorModelId,
            scale: this.scale,
            health: (this.health / this.maxHealth) * 100
            //terrainObjectId: this.spawnerId
        };
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        if (!client.searchedProps.includes(this)) {
            server.utilizeHudTimer(client, server.getItemDefinition(this.getContainer()?.itemDefinitionId)
                ?.NAME_ID ?? 0, this.searchTime, 0, () => {
                super.OnPlayerSelect(server, client);
                client.searchedProps.push(this);
            });
        }
        else {
            super.OnPlayerSelect(server, client);
        }
    }
    OnInteractionString(server, client) {
        if (this.actorModelId == enums_1.ModelIds.HOSPITAL_LAB_WORKBENCH) {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.USE_TARGET
            });
            return;
        }
        if (client.searchedProps.includes(this)) {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.OPEN
            });
        }
        else {
            switch (this.actorModelId) {
                case enums_1.ModelIds.CABINET_SET_03:
                case enums_1.ModelIds.CABINET_SET_05:
                case enums_1.ModelIds.CABINET_SET_06:
                case enums_1.ModelIds.CABINET_SET_07:
                case enums_1.ModelIds.CABINET_SET_08:
                case enums_1.ModelIds.CABINET_SET_09:
                    server.sendData(client, "Command.InteractionString", {
                        guid: this.characterId,
                        stringId: enums_1.StringIds.SEARCH_ALL_CABINETS
                    });
                    break;
                default:
                    server.sendData(client, "Command.InteractionString", {
                        guid: this.characterId,
                        stringId: enums_1.StringIds.SEARCH
                    });
                    return;
            }
        }
    }
    destroy(server) {
        return server.deleteEntity(this.characterId, server._lootableProps);
    }
    OnMeleeHit(server, damageInfo) {
        switch (this.lootSpawner) {
            case "Wrecked Van":
            case "Wrecked Car":
            case "Wrecked Truck":
                break;
            default:
                return;
        }
        const client = server.getClientByCharId(damageInfo.entity);
        const weapon = client?.character.getEquippedWeapon();
        if (!client || !weapon || weapon.itemDefinitionId != enums_1.Items.WEAPON_CROWBAR) {
            return;
        }
        for (let x = 0; x < server._grid.length; x++) {
            const grid = server._grid[x];
            const index = grid.objects.indexOf(this);
            if (index > -1) {
                if (grid.availableScrap) {
                    if ((0, utils_1.randomIntFromInterval)(0, 100) <= server.crowbarHitRewardChance) {
                        grid.availableScrap--;
                        client.character.lootItem(server, server.generateItem(enums_1.Items.METAL_SCRAP));
                        server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.RECYCLING, 1);
                        server.lootCrateWithChance(client, 2);
                    }
                }
                else {
                    server.sendChatText(client, `There is no metal scrap left in this area`);
                }
            }
        }
        server.damageItem(client.character, weapon, server.crowbarHitDamage);
    }
}
exports.LootableProp = LootableProp;
