// h1z1-server/out/servers/ZoneServer2016/entities/lootbag.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lootbag = void 0;
const baselootableentity_1 = require("./baselootableentity");
const loadouts_1 = require("../data/loadouts");
const enums_1 = require("../models/enums");
class Lootbag extends baselootableentity_1.BaseLootableEntity {
    /** Time (milliseconds) when the lootbag is created */
    creationTime = Date.now();
    /** Returns true if the player can transfer items to the lootbag */
    canAcceptItems = false;
    loadoutId = 5;
    constructor(characterId, transientId, actorModelId, position, rotation, server) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        const container = this.getContainer();
        if (container)
            container.canAcceptItems = false;
        if (actorModelId != enums_1.ModelIds.MILITARY_CRATE)
            this.flags.noCollide = 1;
        this.npcRenderDistance =
            this.actorModelId != enums_1.ModelIds.MILITARY_CRATE ? 50 : 200;
        this.defaultLoadout =
            this.actorModelId != enums_1.ModelIds.MILITARY_CRATE
                ? loadouts_1.lootableContainerDefaultLoadouts.lootbag
                : loadouts_1.lootableContainerDefaultLoadouts.military_crate;
        this.equipLoadout(server);
        this.useSimpleStruct = false;
    }
    OnInteractionString(server, client) {
        if (client.searchedProps.includes(this)) {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.OPEN
            });
        }
        else {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.SEARCH
            });
        }
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        if (!client.searchedProps.includes(this)) {
            server.utilizeHudTimer(client, server.getItemDefinition(this._containers["31"].itemDefinitionId)
                ?.NAME_ID ?? 0, this.actorModelId != enums_1.ModelIds.MILITARY_CRATE ? 0 : 10000, 0, () => {
                super.OnPlayerSelect(server, client);
                client.searchedProps.push(this);
            });
        }
        else {
            super.OnPlayerSelect(server, client);
        }
    }
    destroy(server) {
        return server.deleteEntity(this.characterId, server._lootbags);
    }
}
exports.Lootbag = Lootbag;
