// h1z1-server/out/servers/ZoneServer2016/entities/npc.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.Npc = void 0;
const basefullcharacter_1 = require("./basefullcharacter");
const utils_1 = require("../../../utils/utils");
const enums_1 = require("../../../utils/enums");
const enums_2 = require("../models/enums");
const challengemanager_1 = require("../managers/challengemanager");
const projectileentity_1 = require("./projectileentity");
const lootbag_1 = require("../entities/lootbag");
class Npc extends basefullcharacter_1.BaseFullCharacter {
    health;
    npcRenderDistance = 100;
    spawnerId;
    deathTime = 0;
    npcId = 0;
    rewardItems = [];
    flags = {
        bit0: 0,
        bit1: 0,
        bit2: 0,
        bit3: 0,
        bit4: 0,
        bit5: 0,
        bit6: 0,
        bit7: 0,
        nonAttackable: 0, // disables melee flinch
        bit9: 0,
        bit10: 0,
        bit11: 0,
        projectileCollision: 1,
        bit13: 0, // causes a crash if 1 with noCollide 1
        bit14: 0,
        bit15: 0,
        bit16: 0,
        bit17: 0,
        bit18: 0,
        bit19: 0,
        noCollide: 0, // determines if NpcCollision packet gets sent on player collide
        knockedOut: 0, // knockedOut = 1 will not show the entity if the value is sent immediatly at 1
        bit22: 0,
        bit23: 0
    };
    get isAlive() {
        return this.deathTime == 0;
    }
    server;
    npcMeleeDamage;
    isSelected = false;
    constructor(characterId, transientId, actorModelId, position, rotation, server, spawnerId = 0) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.positionUpdateType = enums_2.PositionUpdateType.MOVABLE;
        this.spawnerId = spawnerId;
        this.health = 10000;
        this.initNpcData();
        this.server = server;
        switch (actorModelId) {
            case enums_2.ModelIds.ZOMBIE_FEMALE_WALKER:
            case enums_2.ModelIds.ZOMBIE_MALE_WALKER:
                this.materialType = enums_2.MaterialTypes.ZOMBIE;
                this.npcMeleeDamage = 2000;
                break;
            case enums_2.ModelIds.ZOMBIE_SCREAMER:
                this.materialType = enums_2.MaterialTypes.ZOMBIE;
                this.npcMeleeDamage = 3000;
                break;
            case enums_2.ModelIds.DEER:
            case enums_2.ModelIds.DEER_BUCK:
                this.materialType = enums_2.MaterialTypes.FLESH;
                this.npcMeleeDamage = 0;
                break;
            case enums_2.ModelIds.WOLF:
                this.materialType = enums_2.MaterialTypes.FLESH;
                this.npcMeleeDamage = 2000;
                break;
            case enums_2.ModelIds.BEAR:
                this.materialType = enums_2.MaterialTypes.FLESH;
                this.npcMeleeDamage = 4000;
                break;
            default:
                this.materialType = enums_2.MaterialTypes.FLESH;
                this.npcMeleeDamage = 0;
                break;
        }
        server.aiManager.addEntity(this);
    }
    playAnimation(animationName) {
        this.server.sendDataToAllWithSpawnedEntity(this.server._npcs, this.characterId, "Character.PlayAnimation", {
            characterId: this.characterId,
            animationName: animationName
        });
    }
    applyDamage(characterId) {
        const client = this.server.getClientByCharId(characterId);
        if (client) {
            const damageInfo = {
                entity: this.characterId,
                weapon: enums_2.Items.WEAPON_MACHETE01,
                damage: this.npcMeleeDamage,
                causeBleed: false, // another method for melees to apply bleeding
                meleeType: enums_2.MeleeTypes.BLADE,
                hitReport: {
                    sessionProjectileCount: 0,
                    characterId: client.character.characterId,
                    position: client.character.state.position,
                    unknownFlag1: 0,
                    unknownByte2: 0,
                    totalShotCount: 0,
                    hitLocation: client.character.meleeHit.abilityHitLocation
                }
            };
            client.character.OnMeleeHit(this.server, damageInfo);
        }
        else {
            console.log(`CharacterId ${characterId} not found when applying damage from npc`);
        }
    }
    async damage(server, damageInfo) {
        let client = server.getClientByCharId(damageInfo.entity);
        if (!client) {
            const sourceEntity = server.getEntity(damageInfo.entity);
            if (sourceEntity instanceof projectileentity_1.ProjectileEntity) {
                client = server.getClientByCharId(sourceEntity.managerCharacterId);
            }
            else {
                return;
            }
        }
        const oldHealth = this.health;
        if ((this.health -= damageInfo.damage) <= 0 && this.isAlive) {
            this.deathTime = Date.now();
            this.flags.knockedOut = 1;
            // Custom lootbag for zombies
            switch (this.actorModelId) {
                case enums_2.ModelIds.ZOMBIE_FEMALE_WALKER:
                case enums_2.ModelIds.ZOMBIE_MALE_WALKER:
                case enums_2.ModelIds.ZOMBIE_SCREAMER: {
                    this.addZombieLoot(server);
                    break;
                }
                default:
                    server.worldObjectManager.createLootbag(server, this);
                    break;
            }
            if (client) {
                if (this.npcId === enums_2.NpcIds.ZOMBIE) {
                    server.challengeManager.registerChallengeProgression(client, challengemanager_1.ChallengeType.BRAIN_DEAD, 1);
                }
                if (!server._soloMode) {
                    (0, utils_1.logClientActionToMongo)(server._db.collection(enums_1.DB_COLLECTIONS.KILLS), client, server._worldId, {
                        type: this.npcId == enums_2.NpcIds.ZOMBIE
                            ? enums_1.KILL_TYPE.ZOMBIE
                            : enums_1.KILL_TYPE.WILDLIFE
                    });
                }
                if (this.npcId == enums_2.NpcIds.ZOMBIE)
                    client.character.metrics.zombiesKilled++;
                else
                    client.character.metrics.wildlifeKilled++;
            }
            for (const a in server._clients) {
                const c = server._clients[a];
                if (c.spawnedEntities.has(this)) {
                    if (!c.isLoading) {
                        server.sendData(c, "Character.StartMultiStateDeath", {
                            data: {
                                characterId: this.characterId,
                                flag: 128,
                                managerCharacterId: c.character.characterId
                            }
                        });
                        server.sendData(c, "Character.ManagedObject", {
                            objectCharacterId: this.characterId,
                            characterId: c.character.characterId
                        });
                    }
                    else {
                        server.sendData(c, "Character.StartMultiStateDeath", {
                            data: {
                                characterId: this.characterId,
                                flag: 0
                            }
                        });
                    }
                }
            }
        }
        if (client) {
            const damageRecord = await server.generateDamageRecord(this.characterId, damageInfo, oldHealth);
            client.character.addCombatlogEntry(damageRecord);
        }
    }
    addZombieLoot(server) {
        const lootItems = [];
        const wornLetters = [
            enums_2.Items.WORN_LETTER_CHURCH_PV,
            enums_2.Items.WORN_LETTER_LJ_PV,
            enums_2.Items.WORN_LETTER_MISTY_DAM,
            enums_2.Items.WORN_LETTER_RADIO,
            enums_2.Items.WORN_LETTER_RUBY_LAKE,
            enums_2.Items.WORN_LETTER_TOXIC_LAKE,
            enums_2.Items.WORN_LETTER_VILLAS,
            enums_2.Items.WORN_LETTER_WATER_TOWER
        ];
        // Worn letter (4% chance, up to 2)
        if ((0, utils_1.chance)(50)) {
            const randomWornLetter = wornLetters[(0, utils_1.randomIntFromInterval)(0, wornLetters.length - 1)];
            const wornLetterItem = server.generateItem(randomWornLetter, 1);
            if (wornLetterItem) {
                lootItems.push(wornLetterItem);
            }
        }
        const GoodammoTypes = [
            enums_2.Items.AMMO_12GA,
            enums_2.Items.AMMO_223,
            enums_2.Items.AMMO_308,
            enums_2.Items.AMMO_762
        ];
        // Ammo (5% chance)
        if ((0, utils_1.chance)(50)) {
            for (let i = 0; i < 2; i++) {
                const randomAmmo = GoodammoTypes[(0, utils_1.randomIntFromInterval)(0, GoodammoTypes.length - 1)];
                const ammoCount = (0, utils_1.randomIntFromInterval)(1, 3);
                const ammoItem = server.generateItem(randomAmmo, ammoCount);
                if (ammoItem) {
                    lootItems.push(ammoItem);
                }
            }
        }
        const ammoTypes = [enums_2.Items.AMMO_380, enums_2.Items.AMMO_9MM, enums_2.Items.AMMO_45];
        // Ammo (10% chance)
        if ((0, utils_1.chance)(100)) {
            for (let i = 0; i < ammoTypes.length - 1; i++) {
                const randomAmmo = ammoTypes[(0, utils_1.randomIntFromInterval)(0, ammoTypes.length - 1)];
                const ammoCount = (0, utils_1.randomIntFromInterval)(1, 5);
                const ammoItem = server.generateItem(randomAmmo, ammoCount);
                if (ammoItem) {
                    lootItems.push(ammoItem);
                }
            }
        }
        const specialItems = [
            enums_2.Items.WEAPON_BOW_MAKESHIFT,
            enums_2.Items.BACKPACK_BLUE_ORANGE,
            enums_2.Items.CRUMPLED_NOTE,
            enums_2.Items.REFRIGERATOR_NOTE
        ];
        // Special item (15% chance)
        if ((0, utils_1.chance)(150)) {
            for (let i = 0; i < specialItems.length - 1; i++) {
                const randomSpecial = specialItems[(0, utils_1.randomIntFromInterval)(0, specialItems.length - 1)];
                const specialItem = server.generateItem(randomSpecial, 1);
                if (specialItem) {
                    lootItems.push(specialItem);
                }
            }
        }
        // Prototype item (1% chance)
        const PrototypeItems = [
            enums_2.Items.PROTOTYPE_MECHANISM,
            enums_2.Items.PROTOTYPE_RECEIVER,
            enums_2.Items.PROTOTYPE_TRIGGER_ASSEMBLY
        ];
        if ((0, utils_1.chance)(10)) {
            const randomSpecial = PrototypeItems[(0, utils_1.randomIntFromInterval)(0, PrototypeItems.length - 1)];
            const specialItem = server.generateItem(randomSpecial, 1);
            if (specialItem) {
                lootItems.push(specialItem);
            }
        }
        // Cloth (80% chance)
        if ((0, utils_1.chance)(800)) {
            const clothCount = (0, utils_1.randomIntFromInterval)(1, 3);
            const clothItem = server.generateItem(enums_2.Items.CLOTH, clothCount);
            if (clothItem) {
                lootItems.push(clothItem);
            }
        }
        // Only spawn lootbag if there is loot
        if (lootItems.length === 0)
            return;
        const characterId = (0, utils_1.generateRandomGuid)();
        const lootbag = new lootbag_1.Lootbag(characterId, server.getTransientId(characterId), enums_2.ModelIds.LOOT_BAG_CLEAN, new Float32Array([
            this.state.position[0] + 0.7,
            this.state.position[1],
            this.state.position[2] + 0.7
        ]), new Float32Array([0, 0, 0, 0]), server);
        const container = lootbag.getContainer();
        for (const item of lootItems) {
            server.addContainerItem(lootbag, item, container);
        }
        server._lootbags[characterId] = lootbag;
    }
    OnFullCharacterDataRequest(server, client) {
        server.sendData(client, "LightweightToFullNpc", this.pGetFull(server));
        if (this.onReadyCallback) {
            this.onReadyCallback(client);
            delete this.onReadyCallback;
        }
    }
    OnExplosiveHit(server, sourceEntity) {
        let damage = this.health + this.health / 2;
        const distance = (0, utils_1.getDistance)(sourceEntity.state.position, this.state.position);
        if (distance > 5)
            return;
        if (distance > 1)
            damage /= distance;
        this.damage(server, {
            entity: sourceEntity.characterId,
            damage: damage
        });
    }
    OnProjectileHit(server, damageInfo) {
        if (server.isHeadshotOnly &&
            damageInfo.hitReport?.hitLocation != "HEAD" &&
            this.isAlive)
            return;
        const client = server.getClientByCharId(damageInfo.entity);
        if (client && this.isAlive) {
            const hasHelmetBefore = this.hasHelmet(server);
            const hasArmorBefore = this.hasArmor(server);
            server.sendHitmarker(client, damageInfo.hitReport?.hitLocation, this.hasHelmet(server), this.hasArmor(server), hasHelmetBefore, hasArmorBefore);
        }
        switch (damageInfo.hitReport?.hitLocation) {
            case "HEAD":
            case "GLASSES":
            case "NECK":
                damageInfo.damage *= 4;
                break;
            default:
                break;
        }
        this.damage(server, damageInfo);
    }
    OnMeleeHit(server, damageInfo) {
        if (!this.isAlive)
            return; // prevent dead npc despawning from melee dmg
        damageInfo.damage = damageInfo.damage / 1.5;
        this.damage(server, damageInfo);
    }
    destroy(server) {
        return server.deleteEntity(this.characterId, server._npcs);
    }
    initNpcData() {
        switch (this.actorModelId) {
            case enums_2.ModelIds.ZOMBIE_SCREAMER:
                this.nameId = enums_2.StringIds.BANSHEE;
                this.npcId = enums_2.NpcIds.ZOMBIE;
                break;
            case enums_2.ModelIds.ZOMBIE_FEMALE_WALKER:
            case enums_2.ModelIds.ZOMBIE_MALE_WALKER:
                this.nameId = enums_2.StringIds.ZOMBIE_WALKER;
                this.rewardItems = [
                    {
                        itemDefId: enums_2.Items.BRAIN_INFECTED,
                        weight: 10
                    }
                ];
                this.npcId = enums_2.NpcIds.ZOMBIE;
                break;
            case enums_2.ModelIds.DEER_BUCK:
            case enums_2.ModelIds.DEER:
                this.nameId = enums_2.StringIds.DEER;
                this.rewardItems = [
                    {
                        itemDefId: enums_2.Items.MEAT_VENISON,
                        weight: 30
                    },
                    {
                        itemDefId: enums_2.Items.ANIMAL_FAT,
                        weight: 20
                    },
                    {
                        itemDefId: enums_2.Items.DEER_BLADDER,
                        weight: 10
                    }
                ];
                this.npcId = enums_2.NpcIds.DEER;
                break;
            case enums_2.ModelIds.WOLF:
                this.nameId = enums_2.StringIds.WOLF;
                this.rewardItems = [
                    {
                        itemDefId: enums_2.Items.MEAT_WOLF,
                        weight: 30
                    },
                    {
                        itemDefId: enums_2.Items.ANIMAL_FAT,
                        weight: 20
                    }
                ];
                this.npcId = enums_2.NpcIds.WOLF;
                break;
            case enums_2.ModelIds.BEAR:
                this.nameId = enums_2.StringIds.BEAR;
                this.rewardItems = [
                    {
                        itemDefId: enums_2.Items.MEAT_BEAR,
                        weight: 40
                    },
                    {
                        itemDefId: enums_2.Items.ANIMAL_FAT,
                        weight: 20
                    }
                ];
                this.npcId = enums_2.NpcIds.BEAR;
                break;
        }
    }
    OnPlayerSelect(server, client) {
        // Only one at a time
        if (this.isSelected) {
            return;
        }
        this.isSelected = true;
        // Unlock selection after 5sec
        // It's easier to do it that way that to make a whole sys with the utilizeHudTimer
        setTimeout(() => {
            this.isSelected = false;
        }, 5_100);
        const skinningKnife = client.character.getItemById(enums_2.Items.SKINNING_KNIFE);
        if (!this.isAlive && skinningKnife) {
            server.utilizeHudTimer(client, this.nameId, 5000, 0, () => {
                switch (this.actorModelId) {
                    case enums_2.ModelIds.ZOMBIE_FEMALE_WALKER:
                    case enums_2.ModelIds.ZOMBIE_MALE_WALKER:
                    case enums_2.ModelIds.ZOMBIE_SCREAMER:
                        const emptySyringe = client.character.getItemById(enums_2.Items.SYRINGE_EMPTY);
                        if (emptySyringe) {
                            client.character.lootContainerItem(server, server.generateItem(enums_2.Items.SYRINGE_INFECTED_BLOOD));
                            server.removeInventoryItem(client.character, emptySyringe);
                            return;
                        }
                        this.triggerAwards(server, client, this.rewardItems);
                        break;
                    case enums_2.ModelIds.DEER_BUCK:
                    case enums_2.ModelIds.DEER:
                        this.triggerAwards(server, client, this.rewardItems);
                        break;
                    case enums_2.ModelIds.BEAR:
                        this.triggerAwards(server, client, this.rewardItems);
                        break;
                    case enums_2.ModelIds.WOLF:
                        this.triggerAwards(server, client, this.rewardItems);
                        break;
                }
                server.damageItem(client.character, skinningKnife, 200);
                server.deleteEntity(this.characterId, server._npcs);
            });
        }
    }
    triggerAwards(server, client, rewardItems) {
        const ranges = [];
        const preRewardedItems = [];
        let cumulativeWeight = 0;
        for (const reward of rewardItems) {
            const range = {
                start: cumulativeWeight,
                end: cumulativeWeight + reward.weight,
                item: reward
            };
            ranges.push(range);
            cumulativeWeight = range.end;
        }
        const totalWeight = rewardItems.reduce((sum, item) => sum + item.weight, 0);
        let count = 1;
        let selectedRange = ranges[0];
        for (let i = 0; i < rewardItems.length; i++) {
            const randomValue = Math.random() * totalWeight;
            for (const range of ranges) {
                if (randomValue >= range.start && randomValue < range.end) {
                    selectedRange = range;
                    break; // Break out of the loop once a range is chosen
                }
            }
            if (!preRewardedItems.includes(selectedRange.item.itemDefId)) {
                preRewardedItems.push(selectedRange.item.itemDefId);
                if (Math.random() <= 0.4 &&
                    selectedRange.item.itemDefId != enums_2.Items.BRAIN_INFECTED) {
                    // 40% chance to spawn double rewards
                    count = 2;
                }
                const rewardItem = server.generateItem(selectedRange.item.itemDefId, count);
                if (rewardItem)
                    client.character.lootContainerItem(server, rewardItem);
            }
        }
    }
    OnInteractionString(server, client) {
        if (!this.isAlive && client.character.hasItem(enums_2.Items.SKINNING_KNIFE)) {
            switch (this.actorModelId) {
                case enums_2.ModelIds.ZOMBIE_FEMALE_WALKER:
                case enums_2.ModelIds.ZOMBIE_MALE_WALKER:
                case enums_2.ModelIds.ZOMBIE_SCREAMER:
                    if (client.character.hasItem(enums_2.Items.SYRINGE_EMPTY)) {
                        this.sendInteractionString(server, client, enums_2.StringIds.EXTRACT_BLOOD);
                        return;
                    }
                    this.sendInteractionString(server, client, enums_2.StringIds.HARVEST);
                    break;
                case enums_2.ModelIds.DEER_BUCK:
                case enums_2.ModelIds.DEER:
                case enums_2.ModelIds.WOLF:
                case enums_2.ModelIds.BEAR:
                    this.sendInteractionString(server, client, enums_2.StringIds.HARVEST);
                    break;
            }
        }
    }
    sendInteractionString(server, client, stringId) {
        server.sendData(client, "Command.InteractionString", {
            guid: this.characterId,
            stringId: stringId
        });
    }
    goTo(position) {
        const angleInRadians2 = Math.random() * (2 * Math.PI) - Math.PI;
        const angleInRadians = Math.atan2(position[1] - this.state.position[1], (0, utils_1.getDistance)(this.state.position, position));
        this.state.position = position;
        this.server.sendDataToAll("PlayerUpdatePosition", {
            transientId: this.transientId,
            positionUpdate: {
                sequenceTime: (0, utils_1.getCurrentServerTimeWrapper)().getTruncatedU32(),
                position: this.state.position,
                unknown3_int8: 0,
                stance: 66565,
                engineRPM: 0,
                orientation: angleInRadians2,
                frontTilt: 0,
                sideTilt: 0,
                angleChange: 0,
                verticalSpeed: angleInRadians,
                horizontalSpeed: 0.5
            }
        });
    }
}
exports.Npc = Npc;
