// h1z1-server/out/servers/ZoneServer2016/entities/plantingdiameter.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlantingDiameter = void 0;
const temporaryentity_1 = require("./temporaryentity");
const enums_1 = require("../models/enums");
class PlantingDiameter extends temporaryentity_1.TemporaryEntity {
    /** The time (milliseconds) at which the PlantingDiameter was placed */
    placementTime = Date.now();
    /** HashMap of the Plant occupying the seed slot */
    seedSlots = {};
    /** Time (milliseconds) at which the PlantingDiameter will disappear if unoccupied - Default: 1 day */
    disappearTimestamp = new Date().getTime() + 86400000;
    /** Returns true if the player uses a fertilizer nearby */
    isFertilized = false;
    /** Time (milliseconds) when a fertilizer was applied to a PlantingDiameter */
    fertilizedTimestamp = 0;
    /** CharacterId of the player who placed this object */
    ownerCharacterId;
    constructor(characterId, transientId, actorModelId, position, rotation, server, ownerCharacterId) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.npcRenderDistance = 30;
        this.ownerCharacterId = ownerCharacterId;
    }
    destroy(server) {
        for (const plant of Object.values(this.seedSlots)) {
            plant.destroy(server);
        }
        return server.deleteEntity(this.characterId, server._temporaryObjects);
    }
    OnInteractionString(server, client) {
        if (this.canUndoPlacement(server, client)) {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.UNDO_PLACEMENT
            });
        }
    }
    OnPlayerSelect(server, client) {
        if (this.canUndoPlacement(server, client)) {
            this.destroy(server);
            client.character.lootItem(server, server.generateItem(enums_1.Items.GROUND_TILLER));
            return;
        }
    }
    OnMeleeHit(server, damageInfo) {
        const client = server.getClientByCharId(damageInfo.entity), weapon = client?.character.getEquippedWeapon();
        if (!client || !weapon)
            return;
        if (weapon.itemDefinitionId != enums_1.Items.WEAPON_HAMMER_DEMOLITION)
            return;
        const parentFoundations = Object.values(server._constructionFoundations).filter((foundation) => foundation.isInside(this.state.position));
        for (const foundation of parentFoundations) {
            if (!foundation.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.DEMOLISH))
                return;
        }
        if (parentFoundations.length == 0 &&
            client.character.characterId != this.ownerCharacterId)
            return;
        for (const plant of Object.values(this.seedSlots)) {
            plant.destroy(server);
        }
        server.deleteEntity(this.characterId, server._temporaryObjects);
    }
    canUndoPlacement(server, client) {
        const weapon = client.character.getEquippedWeapon();
        if (!weapon)
            return false;
        if (weapon.itemDefinitionId != enums_1.Items.WEAPON_HAMMER_DEMOLITION) {
            return false;
        }
        if (Date.now() - this.placementTime > 120_000) {
            return false;
        }
        const parentFoundations = Object.values(server._constructionFoundations).filter((foundation) => foundation.isInside(this.state.position));
        const hasAllPermissions = parentFoundations.every((foundation) => foundation.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.DEMOLISH));
        if (hasAllPermissions && Object.keys(this.seedSlots).length == 0) {
            return true;
        }
        return false;
    }
}
exports.PlantingDiameter = PlantingDiameter;
