// h1z1-server/out/servers/ZoneServer2016/entities/plant.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.Plant = void 0;
const itemobject_1 = require("./itemobject");
const enums_1 = require("../models/enums");
class Plant extends itemobject_1.ItemObject {
    /** Current state the crop is in */
    growState = 0;
    /** Next time (milliseconds) that the crop will enter the next state */
    nextStateTime;
    /** Time (milliseconds) it takes for a crop to enter the next state - Default: 8hrs */
    growTime = 28800000;
    /** CharacterId of the PlantingDiamater the crop is occupying */
    parentObjectCharacterId;
    /** Current slot on a PlantingDiameter (4 slots in total) */
    slot;
    /** Returns true when a player uses fertilizer near the crop */
    isFertilized = false;
    isLightweight = false;
    /** Distance (H1Z1 meters) where the crop will render */
    npcRenderDistance = 30;
    constructor(characterId, transientId, actorModelId, position, rotation, server, spawnerId, item, parentObjectCharacterId, slot) {
        super(characterId, transientId, actorModelId, position, rotation, server, spawnerId, item);
        this.parentObjectCharacterId = parentObjectCharacterId;
        this.slot = slot;
        this.nextStateTime = new Date().getTime() + this.growTime;
        if (!server._temporaryObjects[parentObjectCharacterId])
            return;
        const parent = server._temporaryObjects[parentObjectCharacterId];
        if (parent.isFertilized)
            this.isFertilized = true;
        if (this.item.itemDefinitionId == enums_1.Items.SEED_CORN) {
            this.nameId = enums_1.StringIds.CORN;
        }
        else
            this.nameId = enums_1.StringIds.WHEAT;
    }
    grow(server) {
        if (this.growState == 3)
            return;
        this.growState++;
        switch (this.item.itemDefinitionId) {
            case enums_1.Items.SEED_CORN:
                switch (this.growState) {
                    case 1:
                        this.actorModelId = enums_1.ModelIds.CORN_CROPSTATE_1;
                        break;
                    case 2:
                        this.actorModelId = enums_1.ModelIds.CORN_CROPSTATE_2;
                        break;
                    case 3:
                        this.actorModelId = enums_1.ModelIds.CORN_CROPSTATE_3;
                        break;
                }
                break;
            case enums_1.Items.SEED_WHEAT:
                switch (this.growState) {
                    case 1:
                        this.actorModelId = enums_1.ModelIds.WHEAT_CROPSTATE_1;
                        break;
                    case 2:
                        this.actorModelId = enums_1.ModelIds.WEHAT_CROPSTATE_2;
                        break;
                    case 3:
                        this.actorModelId = enums_1.ModelIds.WHEAT_CROPSTATE_3;
                        break;
                }
        }
        server.sendDataToAllWithSpawnedEntity(server._plants, this.characterId, "Character.ReplaceBaseModel", {
            characterId: this.characterId,
            modelId: this.actorModelId
        });
        if (this.isFertilized) {
            const pos = this.state.position;
            server.sendDataToAllWithSpawnedEntity(
            // play burning effect & remove it after 15s
            server._plants, this.characterId, "Character.PlayWorldCompositeEffect", {
                characterId: this.characterId,
                effectId: enums_1.Effects.EFX_Crop_Fertilizer,
                position: new Float32Array([pos[0], pos[1], pos[2], 1]),
                effectTime: 180
            });
        }
        const timeToAdd = this.isFertilized ? this.growTime / 2 : this.growTime; // 4 or 8h based on fertilized or not
        this.nextStateTime = new Date().getTime() + timeToAdd;
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        if (this.growState != 3)
            return;
        for (const a in server._constructionFoundations) {
            const foundation = server._constructionFoundations[a];
            if (!foundation.isInside(this.state.position))
                continue;
            if (foundation.isSecured &&
                !foundation.getHasPermission(server, client.character.characterId, enums_1.ConstructionPermissionIds.CONTAINERS)) {
                server.sendChatText(client, "Construction: no permission");
                return;
            }
        }
        if (!server._temporaryObjects[this.parentObjectCharacterId]) {
            server.deleteEntity(this.characterId, server._plants);
            return;
        }
        const parent = server._temporaryObjects[this.parentObjectCharacterId];
        delete parent.seedSlots[this.slot];
        switch (this.item.itemDefinitionId) {
            case enums_1.Items.SEED_WHEAT:
                client.character.lootItem(server, server.generateItem(enums_1.Items.WHEAT));
                client.character.lootItem(server, server.generateItem(enums_1.Items.SEED_WHEAT, 2), 2);
                break;
            case enums_1.Items.SEED_CORN:
                client.character.lootItem(server, server.generateItem(enums_1.Items.CORN));
                client.character.lootItem(server, server.generateItem(enums_1.Items.SEED_CORN, 2), 2);
                break;
        }
        server.sendCompositeEffectToAllWithSpawnedEntity(server._plants, this, 5151);
        server.deleteEntity(this.characterId, server._plants);
    }
    OnInteractionString(server, client) {
        if (this.growState != 3)
            return;
        server.sendData(client, "Command.InteractionString", {
            guid: this.characterId,
            stringId: enums_1.StringIds.TAKE_ITEM
        });
    }
    OnFullCharacterDataRequest(server, client) {
        if (!this.isFertilized)
            return;
        server.sendData(client, "Command.PlayDialogEffect", {
            characterId: this.characterId,
            effectId: enums_1.Effects.EFX_Crop_Fertilizer
        });
    }
    destroy(server) {
        return server.deleteEntity(this.characterId, server._plants);
    }
}
exports.Plant = Plant;
