// h1z1-server/out/servers/ZoneServer2016/entities/taskprop.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskProp = void 0;
const enums_1 = require("../models/enums");
const baselightweightcharacter_1 = require("./baselightweightcharacter");
const utils_1 = require("../../../utils/utils");
class TaskProp extends baselightweightcharacter_1.BaseLightweightCharacter {
    /** ModelId of the TaskProp */
    actorModel;
    spawnerId;
    /** Id of the required item for the corresponding task */
    requiredItemId = 0;
    /** Array of items to be rewarded upon  */
    rewardItems = [];
    constructor(characterId, transientId, actorModelId, position, rotation, server, scale, zoneId, renderDistance, actorModel) {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.actorModel = actorModel;
        this.spawnerId = zoneId;
        this.scale = scale;
        this.npcRenderDistance = renderDistance;
        this.getTaskPropData();
        this.useSimpleStruct = true;
    }
    getRewardItemCount(itemId) {
        switch (itemId) {
            case enums_1.Items.AMMO_45:
            case enums_1.Items.AMMO_9MM:
            case enums_1.Items.AMMO_12GA:
            case enums_1.Items.AMMO_223:
                return (0, utils_1.randomIntFromInterval)(6, 8);
            case enums_1.Items.FIRST_AID:
            case enums_1.Items.ANTIBIOTICS:
                return 3;
        }
        return 1;
    }
    getRequiredItemCount(itemId) {
        switch (itemId) {
            case enums_1.Items.BRAIN_TREATED:
                return 10;
        }
        return 1;
    }
    getTaskPropData() {
        switch (this.actorModel) {
            case "Task_Patient_Safe_FileCabinet01.adr":
                this.nameId = enums_1.StringIds.LOCKED_CABINET;
                this.requiredItemId = enums_1.Items.WEICHS_WALLET;
                this.rewardItems = [enums_1.Items.AMMO_9MM, enums_1.Items.WEAPON_M9];
                break;
            case "Task_Patient_Hospital_Props_Desk01.adr":
                this.nameId = enums_1.StringIds.MORGUE_DESK;
                this.requiredItemId = enums_1.Items.WEICHS_REPORT;
                this.rewardItems = [enums_1.Items.WEICHS_WALLET];
                break;
            case "Task_Patient_Quarantine_FileCabinet01.adr":
                this.nameId = enums_1.StringIds.EXAMINATION_CABINET;
                this.requiredItemId = enums_1.Items.KLAVISK_NOTE;
                this.rewardItems = [enums_1.Items.WEICHS_REPORT];
                break;
            case "Task_Patient_Records_FileCabinet01.adr":
                this.nameId = enums_1.StringIds.RECORDS_CABINET;
                this.requiredItemId = enums_1.Items.DOCTORS_FILE;
                this.rewardItems = [enums_1.Items.KLAVISK_NOTE];
                break;
            case "Task_Nurse_Hospital_Props_DrugCabinet.adr":
                this.nameId = enums_1.StringIds.SMALL_PHARMACY_CABINET;
                this.requiredItemId = enums_1.Items.HANDWRITTEN_NOTE_CAROLINE;
                this.rewardItems = [
                    enums_1.Items.FIRST_AID,
                    enums_1.Items.ANTIBIOTICS,
                    enums_1.Items.VIAL_H1Z1_REDUCER,
                    enums_1.Items.EMPTY_SPECIMEN_BAG,
                    enums_1.Items.CAP_SCRUBS_GRAY,
                    enums_1.Items.PANTS_SCRUBS_GRAY,
                    enums_1.Items.SHIRT_SCRUBS_GRAY
                ];
                break;
            case "Task_Nurse_Hospital_Props_Desk01.adr":
                this.nameId = enums_1.StringIds.ICU_DESK;
                this.requiredItemId = enums_1.Items.CRACKED_CLIPBOARD;
                this.rewardItems = [enums_1.Items.HANDWRITTEN_NOTE_CAROLINE];
                break;
            case "Task_Hospital_Orderly_ToolCabinet01.adr":
                this.nameId = enums_1.StringIds.MAINTENANCE_TOOLBOX;
                this.requiredItemId = enums_1.Items.SMALL_KEY;
                this.rewardItems = [enums_1.Items.FIRST_AID, enums_1.Items.WEAPON_MACHETE01];
                break;
            case "Task_Orderly_Hospital_PaperDebris.adr":
                this.nameId = enums_1.StringIds.BATTERED_TRASHCAN;
                this.requiredItemId = enums_1.Items.TORN_LETTERHEAD;
                this.rewardItems = [
                    //Items.CRUMPLED_NOTE // TODO: This should spawn the nurse Zombie to retreive the small key
                    enums_1.Items.SMALL_KEY
                ];
                break;
            case "Task_Orderly_Hospital_Props_Desk01.adr":
                this.nameId = enums_1.StringIds.RADIOLOGY_DESK;
                this.requiredItemId = enums_1.Items.PHONE_CHARGED;
                this.rewardItems = [enums_1.Items.TORN_LETTERHEAD];
                break;
            case "Task_Hospital_Researcher_Radio.adr":
                this.nameId = enums_1.StringIds.LONG_RANGE_RADIO;
                break;
            case "Task_Hospital_Researcher_TreasureChest.adr":
                this.nameId = 12771;
                this.requiredItemId = enums_1.Items.BRAIN_TREATED;
                this.rewardItems = [enums_1.Items.CODED_MESSAGE];
                break;
            case "Task_Hospital_Researcher_Locker.adr":
                this.nameId = 12781;
                this.requiredItemId = enums_1.Items.LOCKER_KEY_F1;
                this.rewardItems = [
                    enums_1.Items.AMMO_45,
                    enums_1.Items.AMMO_9MM,
                    enums_1.Items.AMMO_12GA,
                    enums_1.Items.ANTIBIOTICS,
                    enums_1.Items.CAP_SCRUBS_BLUE,
                    enums_1.Items.PANTS_SCRUBS_BLUE,
                    enums_1.Items.SHIRT_SCRUBS_BLUE,
                    enums_1.Items.SURGEON_MASK_AQUA,
                    enums_1.Items.FIRST_AID
                ];
                break;
            case "Task_Hospital_Researcher_Locker02.adr":
                this.nameId = 12785;
                this.requiredItemId = enums_1.Items.LOCKER_KEY_F2;
                this.rewardItems = [
                    enums_1.Items.AMMO_45,
                    enums_1.Items.AMMO_9MM,
                    enums_1.Items.AMMO_12GA,
                    enums_1.Items.ANTIBIOTICS,
                    enums_1.Items.CAP_SCRUBS_BLUE,
                    enums_1.Items.PANTS_SCRUBS_BLUE,
                    enums_1.Items.SHIRT_SCRUBS_BLUE,
                    enums_1.Items.SURGEON_MASK_AQUA,
                    enums_1.Items.FIRST_AID
                ];
                break;
            case "Task_Hospital_Researcher_Locker03.adr":
                this.nameId = 12787;
                this.requiredItemId = enums_1.Items.LOCKER_KEY_F3;
                this.rewardItems = [
                    enums_1.Items.AMMO_45,
                    enums_1.Items.AMMO_9MM,
                    enums_1.Items.AMMO_12GA,
                    enums_1.Items.ANTIBIOTICS,
                    enums_1.Items.CAP_SCRUBS_BLUE,
                    enums_1.Items.PANTS_SCRUBS_BLUE,
                    enums_1.Items.SHIRT_SCRUBS_BLUE,
                    enums_1.Items.SURGEON_MASK_AQUA,
                    enums_1.Items.FIRST_AID
                ];
                break;
            case "Task_Hospital_Researcher_Locker04.adr":
                this.nameId = 12790;
                this.requiredItemId = enums_1.Items.LOCKER_KEY_F4;
                this.rewardItems = [
                    enums_1.Items.AMMO_45,
                    enums_1.Items.AMMO_9MM,
                    enums_1.Items.AMMO_12GA,
                    enums_1.Items.ANTIBIOTICS,
                    enums_1.Items.CAP_SCRUBS_BLUE,
                    enums_1.Items.PANTS_SCRUBS_BLUE,
                    enums_1.Items.SHIRT_SCRUBS_BLUE,
                    enums_1.Items.SURGEON_MASK_AQUA,
                    enums_1.Items.FIRST_AID
                ];
                break;
        }
    }
    destroy(server) {
        return server.deleteEntity(this.characterId, server._taskProps);
    }
    OnInteractionString(server, client) {
        switch (this.actorModel) {
            case "Task_Patient_Safe_FileCabinet01.adr":
                server.sendData(client, "Command.InteractionString", {
                    guid: this.characterId,
                    stringId: enums_1.StringIds.OPEN_LOCKED_CABINET
                });
                break;
            case "Task_Patient_Records_FileCabinet01.adr":
                server.sendData(client, "Command.InteractionString", {
                    guid: this.characterId,
                    stringId: enums_1.StringIds.SEARCH_RECORDS
                });
                break;
            case "Task_Nurse_Hospital_Props_DrugCabinet.adr":
                server.sendData(client, "Command.InteractionString", {
                    guid: this.characterId,
                    stringId: enums_1.StringIds.OPEN_CABINET
                });
                break;
            case "Task_Hospital_Orderly_ToolCabinet01.adr":
                server.sendData(client, "Command.InteractionString", {
                    guid: this.characterId,
                    stringId: enums_1.StringIds.OPEN_TOOLBOX
                });
                break;
            case "Task_Orderly_Hospital_PaperDebris.adr":
                server.sendData(client, "Command.InteractionString", {
                    guid: this.characterId,
                    stringId: enums_1.StringIds.BATTERED_TRASHCAN
                });
                break;
            case "Task_Patient_Hospital_Props_Desk01.adr":
            case "Task_Patient_Quarantine_FileCabinet01.adr":
            case "Task_Nurse_Hospital_Props_Desk01.adr":
            case "Task_Orderly_Hospital_Props_Desk01.adr":
                server.sendData(client, "Command.InteractionString", {
                    guid: this.characterId,
                    stringId: enums_1.StringIds.SEARCH
                });
                break;
            case "Task_Hospital_Researcher_Radio.adr":
                server.sendData(client, "Command.InteractionString", {
                    guid: this.characterId,
                    stringId: enums_1.StringIds.USE_TARGET
                });
                break;
            case "Common_Props_Bedroom_Mattress01.adr":
            case "Common_Props_MilitaryBase_BunkBed.adr":
            case "Common_Props_Bedroom_BedCombined01.adr":
            case "Common_Props_Bedroom_BedCombined02.adr":
            case "Common_Props_Bedroom_BedCombined03.adr":
            case "Common_Props_Bedroom_BedFrame01.adr":
                if (client.character._resources[enums_1.ResourceIds.ENDURANCE] <= 3501) {
                    server.sendData(client, "Command.InteractionString", {
                        guid: this.characterId,
                        stringId: enums_1.StringIds.REST
                    });
                }
                else {
                    server.sendData(client, "Command.InteractionString", {
                        guid: this.characterId,
                        stringId: enums_1.StringIds.NOT_TIRED
                    });
                }
                break;
            default:
                server.sendData(client, "Command.InteractionString", {
                    guid: this.characterId,
                    stringId: enums_1.StringIds.OPEN
                });
                break;
        }
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        switch (this.actorModel) {
            case "Task_Hospital_Researcher_Radio.adr":
                if (!client.character.hasItem(enums_1.Items.AIRDROP_CODE))
                    return;
                const item = client.character.getItemById(enums_1.Items.AIRDROP_CODE);
                if (!item || client.character.hasAirdropClearance)
                    return;
                server.utilizeHudTimer(client, enums_1.StringIds.LONG_RANGE_RADIO, 1000, 0, () => {
                    client.character.hasAirdropClearance = true;
                    server.sendAlert(client, "You hear an automated message requesting a code to call in an airdrop.");
                });
                break;
            case "Common_Props_Bedroom_Mattress01.adr":
            case "Common_Props_MilitaryBase_BunkBed.adr":
            case "Common_Props_Bedroom_BedCombined01.adr":
            case "Common_Props_Bedroom_BedCombined02.adr":
            case "Common_Props_Bedroom_BedCombined03.adr":
                if (client.character._resources[enums_1.ResourceIds.ENDURANCE] <= 3501) {
                    server.utilizeHudTimer(client, enums_1.StringIds.RESTING, 20000, 0, () => {
                        server.sleep(client);
                    });
                }
                break;
            default:
                const requiredItemCount = this.getRequiredItemCount(this.requiredItemId);
                const inventoryItemCount = client.character.getInventoryItemAmount(this.requiredItemId);
                if (requiredItemCount > inventoryItemCount) {
                    switch (this.requiredItemId) {
                        case enums_1.Items.LOCKER_KEY_F1:
                            server.sendAlert(client, "This locker requires key 207."); //String ID #12782
                            break;
                        case enums_1.Items.LOCKER_KEY_F2:
                            server.sendAlert(client, "This locker requires key 122."); //String ID #12786
                            break;
                        case enums_1.Items.LOCKER_KEY_F3:
                            server.sendAlert(client, "This locker requires key 591."); //String ID #12788
                            break;
                        case enums_1.Items.LOCKER_KEY_F4:
                            server.sendAlert(client, "This locker requires key 301."); //String ID #12791
                            break;
                    }
                    return;
                }
                const itemsPassed = [];
                const itemCount = this.rewardItems.length === 1 ? 1 : (0, utils_1.randomIntFromInterval)(2, 4);
                for (let x = 0; x < itemCount; x++) {
                    const randomIndex = (0, utils_1.randomIntFromInterval)(0, this.rewardItems.length - 1);
                    const item = this.rewardItems[randomIndex];
                    if (itemsPassed.map((sItem) => sItem.itemDefinitionId).includes(item))
                        continue; //TODO: Check if there's already a scrub received
                    itemsPassed.push({
                        itemDefinitionId: item,
                        count: this.getRewardItemCount(item)
                    });
                }
                const requiredItemInfo = {
                    itemDefinitionId: this.requiredItemId,
                    count: requiredItemCount
                };
                server.taskOption(client, 1000, this.nameId, requiredItemInfo, itemsPassed);
                break;
        }
    }
}
exports.TaskProp = TaskProp;
