// h1z1-server/out/servers/ZoneServer2016/entities/trapentity.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrapEntity = void 0;
const utils_1 = require("../../../utils/utils");
const enums_1 = require("../models/enums");
const basesimplenpc_1 = require("./basesimplenpc");
class TrapEntity extends basesimplenpc_1.BaseSimpleNpc {
    /** Returns true if a snare has been stepped on */
    isTriggered = false;
    cooldown = 1000;
    lastTrigger = Date.now();
    triggerRadiusX = 1;
    triggerRadiusY = 1;
    /** Distance (H1Z1 meters) where the TrapEntity will render */
    npcRenderDistance = 75;
    /** Id of the TrapEntity - See ServerItemDefinitions.json for more information */
    itemDefinitionId;
    worldOwned = false;
    ownerCharacterId;
    cubebounds;
    server;
    degradationTime = 3600_000 * 24;
    constructor(characterId, transientId, actorModelId, position, rotation, server, itemDefinitionId, worldOwned = false, ownerCharacterId = "") {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.itemDefinitionId = itemDefinitionId;
        this.worldOwned = worldOwned;
        this.ownerCharacterId = ownerCharacterId;
        this.server = server;
        const angle = -this.state.rotation[1];
        switch (itemDefinitionId) {
            case enums_1.Items.BARBED_WIRE:
                this.cubebounds = (0, utils_1.getCubeBounds)(position, 7.05, 2.15, angle, position[1] - 0.9, position[1] + 1.8);
        }
    }
    async arm() {
        switch (this.itemDefinitionId) {
            case enums_1.Items.PUNJI_STICKS:
            case enums_1.Items.PUNJI_STICK_ROW:
                this.server.aiManager.addEntity(this);
                this.cooldown = 500;
                this.triggerRadiusX = 0.8;
                this.triggerRadiusY = 0.5;
                break;
            case enums_1.Items.SNARE:
                this.server.aiManager.addEntity(this);
                this.cooldown = 200;
                this.triggerRadiusX = 1;
                this.triggerRadiusY = 0.5;
                break;
            case enums_1.Items.BARBED_WIRE:
                this.server.aiManager.addEntity(this);
                this.cooldown = 200;
                this.triggerRadiusX = 5;
                break;
            case enums_1.Items.TRAP_FIRE:
            case enums_1.Items.TRAP_FLASH:
                // Wait 10 seconds before activating the trap
                await new Promise((resolve) => setTimeout(resolve, 10000));
                this.server.aiManager.addEntity(this);
                this.cooldown = 90;
                this.triggerRadiusX = 1;
                this.triggerRadiusY = 0.5;
                break;
        }
    }
    async detonate(characterId) {
        const client = this.server.getClientByCharId(characterId);
        const server = this.server;
        if (!client) {
            return;
        }
        this.lastTrigger = Date.now();
        switch (this.itemDefinitionId) {
            case enums_1.Items.PUNJI_STICKS:
            case enums_1.Items.PUNJI_STICK_ROW:
                if (!server._traps[this.characterId]) {
                    return;
                }
                if (client.character.isAlive &&
                    !client.vehicle.mountedVehicle &&
                    !client.character.isSpectator) {
                    client.character.damage(server, {
                        entity: this.characterId,
                        causeBleed: true,
                        damage: 501
                    });
                    server.sendDataToAllWithSpawnedEntity(server._traps, this.characterId, "Character.PlayWorldCompositeEffect", {
                        characterId: "0x0",
                        effectId: enums_1.Effects.PFX_Impact_PunjiSticks_Blood,
                        position: client.character.state.position
                    });
                    server.sendDataToAllWithSpawnedEntity(server._traps, this.characterId, "Character.UpdateSimpleProxyHealth", this.pGetSimpleProxyHealth());
                    if (!this.worldOwned)
                        this.health -= 1000;
                }
                if (this.health <= 0) {
                    server.sendDataToAllWithSpawnedEntity(server._traps, this.characterId, "Character.PlayWorldCompositeEffect", {
                        characterId: "0x0",
                        effectId: enums_1.Effects.PFX_Damage_Crate_01m,
                        position: this.state.position
                    });
                    this.destroy();
                    return;
                }
                break;
            case enums_1.Items.SNARE:
                if (!server._traps[this.characterId]) {
                    return;
                }
                for (const a in server._clients) {
                    const client = server._clients[a];
                    if ((0, utils_1.getDistance)(client.character.state.position, this.state.position) <
                        1) {
                        client.character.damage(server, {
                            entity: this.characterId,
                            damage: 2000
                        });
                        client.character._resources[enums_1.ResourceIds.BLEEDING] += 41;
                        server.updateResourceToAllWithSpawnedEntity(client.character.characterId, client.character._resources[enums_1.ResourceIds.BLEEDING] > 0
                            ? client.character._resources[enums_1.ResourceIds.BLEEDING]
                            : 0, enums_1.ResourceIds.BLEEDING, enums_1.ResourceTypes.BLEEDING, server._characters);
                        server.sendDataToAllWithSpawnedEntity(server._traps, this.characterId, "Character.PlayWorldCompositeEffect", {
                            characterId: this.characterId,
                            effectId: enums_1.Effects.PFX_Impact_Knife_Metal_Vehicle,
                            position: server._traps[this.characterId].state.position
                        });
                        this.isTriggered = true;
                        server.applyMovementModifier(client, enums_1.MovementModifiers.SNARED);
                    }
                }
                if (this.isTriggered) {
                    this.destroy();
                    this.actorModelId = enums_1.ModelIds.SNARE;
                    server.worldObjectManager.createLootEntity(server, server.generateItem(enums_1.Items.SNARE), this.state.position, this.state.rotation, 15);
                }
                break;
            case enums_1.Items.BARBED_WIRE:
                if (!server._traps[this.characterId]) {
                    return;
                }
                for (const a in server._clients) {
                    const client = server._clients[a];
                    if (this.isInside(client.character.state.position) &&
                        client.character.isAlive &&
                        !client.character.isSpectator) {
                        client.character.damage(server, {
                            entity: this.characterId,
                            causeBleed: true,
                            damage: 501
                        });
                        server.sendDataToAllWithSpawnedEntity(server._traps, this.characterId, "Character.PlayWorldCompositeEffect", {
                            characterId: "0x0",
                            effectId: enums_1.Effects.PFX_Impact_PunjiSticks_Blood,
                            position: server._clients[a].character.state.position
                        });
                        server.sendDataToAllWithSpawnedEntity(server._traps, this.characterId, "Character.UpdateSimpleProxyHealth", this.pGetSimpleProxyHealth());
                        if (!this.worldOwned)
                            this.health -= 1000;
                    }
                }
                if (this.health <= 0) {
                    server.sendDataToAllWithSpawnedEntity(server._traps, this.characterId, "Character.PlayWorldCompositeEffect", {
                        characterId: "0x0",
                        effectId: enums_1.Effects.PFX_Damage_Crate_01m,
                        position: this.state.position
                    });
                    this.destroy();
                    return;
                }
                break;
            case enums_1.Items.TRAP_FIRE:
            case enums_1.Items.TRAP_FLASH:
                // Wait 10 seconds before activating the trap
                await new Promise((resolve) => setTimeout(resolve, 10000));
                if (!server._traps[this.characterId]) {
                    return;
                }
                for (const a in server._clients) {
                    const client = server._clients[a];
                    if ((0, utils_1.getDistance)(client.character.state.position, this.state.position) <
                        1) {
                        this.detonateTrap(server, {
                            entity: client.character.characterId,
                            damage: 0
                        });
                        this.isTriggered = true;
                    }
                }
                for (const a in server._vehicles) {
                    if ((0, utils_1.getDistance)(server._vehicles[a].state.position, this.state.position) < 2) {
                        server._vehicles[a].getPassengerList().map((passenger) => {
                            this.detonateTrap(server, { entity: passenger, damage: 0 });
                            this.isTriggered = true;
                        });
                    }
                }
                if (this.isTriggered) {
                    this.destroy();
                }
                break;
        }
    }
    destroy() {
        return this.server.deleteEntity(this.characterId, this.server._traps);
    }
    isInside(position) {
        switch (this.itemDefinitionId) {
            case enums_1.Items.BARBED_WIRE:
                return (0, utils_1.isInsideCube)(Array.from(position), this.cubebounds);
        }
    }
    OnProjectileHit(server, damageInfo) {
        const damage = damageInfo.damage * 6; // bullets do more to damage traps
        this.damage(server, { ...damageInfo, damage });
    }
    OnMeleeHit(server, damageInfo) {
        const client = server.getClientByCharId(damageInfo.entity), weapon = client?.character.getEquippedWeapon();
        if (!client || !weapon)
            return;
        const damage = damageInfo.damage * 3;
        this.damage(server, { ...damageInfo, damage });
        server.damageItem(client.character, weapon, 50);
    }
    damage(server, damageInfo) {
        if (this.worldOwned)
            return;
        this.health -= damageInfo.damage;
        server.sendDataToAllWithSpawnedEntity(server._traps, this.characterId, "Character.UpdateSimpleProxyHealth", this.pGetSimpleProxyHealth());
        if (this.health > 0)
            return;
        if (this.destroy())
            this.detonateTrap(server, damageInfo);
    }
    detonateTrap(server, damageInfo) {
        const client = server.getClientByCharId(damageInfo.entity);
        if (!client ||
            ![enums_1.Items.TRAP_FLASH, enums_1.Items.TRAP_FIRE].includes(this.itemDefinitionId))
            return;
        //TODO: Damage vehicles while driving over fire traps
        switch (this.itemDefinitionId) {
            case enums_1.Items.TRAP_FLASH:
                if ((0, utils_1.getDistance)(client.character.state.position, this.state.position) <= 5) {
                    server.addScreenEffect(client, server._screenEffects["FLASH"]);
                    server.sendDataToAllWithSpawnedEntity(server._characters, client.character.characterId, "Character.PlayAnimation", {
                        characterId: client.character.characterId,
                        animationName: "Action",
                        animationType: "ActionType",
                        unm4: 0,
                        unknownDword1: 0,
                        unknownByte1: 0,
                        unknownDword2: 0,
                        unknownByte1xda: 0,
                        unknownDword3: 9
                    });
                }
                break;
            case enums_1.Items.TRAP_FIRE:
                if ((0, utils_1.getDistance)(client.character.state.position, this.state.position) <= 5) {
                    server.sendDataToAllWithSpawnedEntity(server._characters, client.character.characterId, "Character.PlayAnimation", {
                        characterId: client.character.characterId,
                        animationName: "Action",
                        animationType: "ActionType",
                        unm4: 0,
                        unknownDword1: 0,
                        unknownByte1: 0,
                        unknownDword2: 0,
                        unknownByte1xda: 0,
                        unknownDword3: 11
                    });
                    server.sendDataToAllWithSpawnedEntity(server._traps, this.characterId, "Character.AddEffectTagCompositeEffect", {
                        characterId: client.character.characterId,
                        unknownDword1: enums_1.Effects.PFX_Fire_Person_loop,
                        effectId: enums_1.Effects.PFX_Fire_Person_loop,
                        unknownGuid: client.character.characterId,
                        unknownDword2: 3
                    });
                    const interval = 1000, duration = 10000;
                    let elapsedTime = 0;
                    const timerId = setInterval(() => {
                        elapsedTime += interval;
                        client.character.damage(server, {
                            entity: this.characterId,
                            causeBleed: false,
                            damage: 500
                        });
                        if (elapsedTime >= duration) {
                            clearInterval(timerId);
                            server.sendDataToAllWithSpawnedEntity(server._characters, client.character.characterId, "Character.RemoveEffectTagCompositeEffect", {
                                characterId: client.character.characterId,
                                effectId: enums_1.Effects.PFX_Fire_Person_loop,
                                newEffectId: 0
                            });
                        }
                    }, interval);
                }
                break;
        }
        if (!this.isTriggered) {
            server.sendCompositeEffectToAllInRange(600, "", this.state.position, enums_1.Effects.PFX_Impact_Explosion_FlashGrenade_Default);
        }
    }
    OnExplosiveHit(server, sourceEntity) {
        if (!(0, utils_1.isPosInRadius)(5, this.state.position, sourceEntity.state.position))
            return;
        this.destroy();
    }
}
exports.TrapEntity = TrapEntity;
