// h1z1-server/out/servers/ZoneServer2016/entities/treasurechest.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.TreasureChest = void 0;
const enums_1 = require("../models/enums");
const lootableprop_1 = require("./lootableprop");
const utils_1 = require("../../../utils/utils");
class TreasureChest extends lootableprop_1.LootableProp {
    /** Id of the worn letter that corresponds to the correct TreasureChest */
    requiredItemId = 0;
    /** Determines what items will spawn in the TreasureChest */
    rewardItems = [];
    /** Id of the chest container to spawn the items at */
    spawnerId;
    /** Time alloted before the Treasure Chest clears its contents */
    clearChestTimer;
    constructor(characterId, transientId, actorModelId, position, rotation, server, scale, spawnerId, renderDistance) {
        super(characterId, transientId, actorModelId, position, rotation, server, new Float32Array([1, 1, 1, 1]), 0, renderDistance);
        this.spawnerId = spawnerId;
        this.getTaskPropData();
    }
    getRewardItemCount(itemId) {
        switch (itemId) {
            case enums_1.Items.AMMO_223:
                return 24;
            case enums_1.Items.AMMO_12GA:
                return 12;
        }
        return 1;
    }
    getTaskPropData() {
        switch (this.spawnerId) {
            case enums_1.Items.CHEST_CHURCH_PV:
                this.requiredItemId = enums_1.Items.WORN_LETTER_CHURCH_PV;
                break;
            case enums_1.Items.CHEST_MISTY_DAM:
                this.requiredItemId = enums_1.Items.WORN_LETTER_MISTY_DAM;
                break;
            case enums_1.Items.CHEST_RUBY_LAKE:
                this.requiredItemId = enums_1.Items.WORN_LETTER_RUBY_LAKE;
                break;
            case enums_1.Items.CHEST_WATER_TOWER:
                this.requiredItemId = enums_1.Items.WORN_LETTER_WATER_TOWER;
                break;
            case enums_1.Items.CHEST_TOXIC_LAKE:
                this.requiredItemId = enums_1.Items.WORN_LETTER_TOXIC_LAKE;
                break;
            case enums_1.Items.CHEST_RADIO:
                this.requiredItemId = enums_1.Items.WORN_LETTER_RADIO;
                break;
            case enums_1.Items.CHEST_LJ_PV:
                this.requiredItemId = enums_1.Items.WORN_LETTER_LJ_PV;
                break;
            case enums_1.Items.CHEST_VILLAS:
                this.requiredItemId = enums_1.Items.WORN_LETTER_VILLAS;
                break;
        }
        this.rewardItems = [
            enums_1.Items.BACKPACK_MILITARY_GREEN_CAMO,
            enums_1.Items.COMPASS,
            enums_1.Items.ALL_PURPOSE_GOGGLES,
            enums_1.Items.AMMO_223,
            enums_1.Items.MRE01,
            enums_1.Items.AMMO_12GA
        ];
    }
    triggerRewards(server) {
        if (!this.requiredItemId)
            return;
        const removedItem = this.getItemById(this.requiredItemId);
        if (!removedItem)
            return;
        const rewardItems = this.rewardItems;
        if (Math.random() < 0.5) {
            rewardItems.push(enums_1.Items.WEAPON_SHOTGUN);
        }
        else {
            rewardItems.push(enums_1.Items.WEAPON_AR15);
        }
        const itemsPassed = rewardItems.map((item) => ({
            itemDefinitionId: item,
            count: this.getRewardItemCount(item)
        }));
        if (!server.removeInventoryItem(this, removedItem, 1))
            return;
        itemsPassed.forEach((itemInstance) => {
            const item = server.generateItem(itemInstance.itemDefinitionId, itemInstance.count, true);
            this.lootContainerItem(server, item, itemInstance.count, false);
        });
        const rewards = server.rewardManager.rewards;
        for (const reward of rewards) {
            if ((0, utils_1.chance)(5) && reward.dropChances > 0) {
                const item = server.generateItem(reward.itemId, 1, true);
                if (!item)
                    return;
                this.lootContainerItem(server, item, item?.stackCount, false);
            }
        }
    }
}
exports.TreasureChest = TreasureChest;
