// h1z1-server/out/servers/ZoneServer2016/entities/vehicle.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.Vehicle2016 = void 0;
const utils_1 = require("../../../utils/utils");
const enums_1 = require("../models/enums");
const baselootableentity_1 = require("./baselootableentity");
const loadouts_1 = require("../data/loadouts");
const constants_1 = require("../../../utils/constants");
const explosiveentity_1 = require("./explosiveentity");
const projectileentity_1 = require("./projectileentity");
const enums_2 = require("../../../utils/enums");
function getActorModelId(vehicleId) {
    switch (vehicleId) {
        case enums_1.VehicleIds.OFFROADER:
            return enums_1.ModelIds.OFFROADER;
        case enums_1.VehicleIds.PICKUP:
            return enums_1.ModelIds.PICKUP_TRUCK;
        case enums_1.VehicleIds.POLICECAR:
            return enums_1.ModelIds.POLICE_CAR;
        case enums_1.VehicleIds.ATV:
            return enums_1.ModelIds.ATV;
        case enums_1.VehicleIds.PARACHUTE:
            return enums_1.ModelIds.PARACHUTE;
        case enums_1.VehicleIds.SPECTATE:
            return enums_1.ModelIds.SPECTATE;
        default:
            return enums_1.ModelIds.OFFROADER;
    }
}
function getVehicleName(ModelId) {
    switch (ModelId) {
        case enums_1.ModelIds.OFFROADER:
            return enums_1.StringIds.OFFROADER;
        case enums_1.ModelIds.PICKUP_TRUCK: // pickup
            return enums_1.StringIds.PICKUP_TRUCK;
        case enums_1.ModelIds.POLICE_CAR: // policecar
            return enums_1.StringIds.POLICE_CAR;
        case enums_1.ModelIds.ATV: // atv
            return enums_1.StringIds.ATV;
        default:
            return enums_1.StringIds.OFFROADER;
    }
}
function getVehicleLoadoutId(vehicleId) {
    switch (vehicleId) {
        case enums_1.VehicleIds.PARACHUTE:
        case enums_1.VehicleIds.OFFROADER:
            return enums_1.LoadoutIds.VEHICLE_OFFROADER;
        case enums_1.VehicleIds.PICKUP:
            return enums_1.LoadoutIds.VEHICLE_PICKUP;
        case enums_1.VehicleIds.POLICECAR:
            return enums_1.LoadoutIds.VEHICLE_POLICECAR;
        case enums_1.VehicleIds.ATV:
            return enums_1.LoadoutIds.VEHICLE_ATV;
        case enums_1.VehicleIds.SPECTATE:
        default:
            return 5; // idk if this is right but these vehicles dont have a loadout
    }
}
function getDefaultLoadout(loadoutId) {
    switch (loadoutId) {
        case enums_1.LoadoutIds.VEHICLE_OFFROADER:
            return loadouts_1.vehicleDefaultLoadouts.offroader;
        case enums_1.LoadoutIds.VEHICLE_PICKUP:
            return loadouts_1.vehicleDefaultLoadouts.pickup;
        case enums_1.LoadoutIds.VEHICLE_POLICECAR:
            return loadouts_1.vehicleDefaultLoadouts.policecar;
        case enums_1.LoadoutIds.VEHICLE_ATV:
            return loadouts_1.vehicleDefaultLoadouts.atv;
        default:
            return [];
    }
}
function getHeadlightEffect(vehicleId) {
    switch (vehicleId) {
        case enums_1.VehicleIds.OFFROADER:
            return enums_1.Effects.VEH_Headlight_OffRoader_wShadows;
        case enums_1.VehicleIds.PICKUP:
            return enums_1.Effects.VEH_Headlight_PickupTruck_wShadows;
        case enums_1.VehicleIds.POLICECAR:
            return enums_1.Effects.VEH_Headlight_PoliceCar_wShadows;
        case enums_1.VehicleIds.ATV:
            return enums_1.Effects.VEH_Headlight_ATV_wShadows;
        default:
            return enums_1.Effects.VEH_Headlight_OffRoader_wShadows;
    }
}
function getHornEffect(vehicleId) {
    switch (vehicleId) {
        case enums_1.VehicleIds.OFFROADER:
            return enums_1.Effects.SFX_VEH_Offroader_Horn;
        case enums_1.VehicleIds.PICKUP:
            return enums_1.Effects.SFX_VEH_Pickup_Truck_Horn;
        case enums_1.VehicleIds.POLICECAR:
            return enums_1.Effects.SFX_VEH_Police_Car_Horn;
        case enums_1.VehicleIds.ATV:
            return enums_1.Effects.SFX_VEH_ATV_Horn;
        default:
            return enums_1.Effects.SFX_VEH_Offroader_Horn;
    }
}
function getTurboEffect(vehicleId) {
    switch (vehicleId) {
        case enums_1.VehicleIds.OFFROADER:
            return enums_1.Effects.VEH_Engine_Boost_OffRoader;
        case enums_1.VehicleIds.PICKUP:
            return enums_1.Effects.VEH_Engine_Boost_PickupTruck;
        case enums_1.VehicleIds.POLICECAR:
            return enums_1.Effects.VEH_Engine_Boost_PoliceCar;
        case enums_1.VehicleIds.ATV:
            return enums_1.Effects.VEH_Engine_Boost_ATV;
        default:
            return enums_1.Effects.VEH_Engine_Boost_OffRoader;
    }
}
function getHotwireEffect(vehicleId) {
    switch (vehicleId) {
        case enums_1.VehicleIds.OFFROADER:
            return enums_1.Effects.SFX_VEH_OffRoader_Hotwire;
        case enums_1.VehicleIds.PICKUP:
            return enums_1.Effects.SFX_VEH_Pickup_Truck_Hotwire;
        case enums_1.VehicleIds.POLICECAR:
            return enums_1.Effects.SFX_VEH_Police_Car_Hotwire;
        case enums_1.VehicleIds.ATV:
            return enums_1.Effects.SFX_VEH_ATV_Hotwire;
        default:
            return enums_1.Effects.SFX_VEH_OffRoader_Hotwire;
    }
}
class Vehicle2016 extends baselootableentity_1.BaseLootableEntity {
    /** Used to determine if the vehicle physics are currently managed by a client */
    isManaged = false;
    manager;
    /** Effect Id upon a car explosion */
    destroyedEffect = 0;
    /** Model Id to spawn after a car explosion */
    destroyedModel = 0;
    /** Effect Id of the vehicle when it collides with the world */
    minorDamageEffect = 0;
    majorDamageEffect = 0;
    criticalDamageEffect = 0;
    supercriticalDamageEffect = 0;
    /** Returns true when the engine is turned on */
    engineOn = false;
    /** Returns true when the player locks the vehicle */
    isLocked = false;
    /** Used to determine the position of the vehicle to transmit to the server */
    positionUpdate;
    /** Speed (H1Z1 RPMs) of the vehicle */
    engineRPM = 0;
    /** Used by resources to determine the fuel level */
    fuelUpdater;
    /** Returns true if the player is spectating or is parachuting -
     * spectating is treated as being inside a vehicle - a "hacky aircraft" if you will */
    isInvulnerable = false;
    onDismount;
    /** Used to update the health of the vehicle */
    resourcesUpdater;
    damageTimeout;
    vehicleManager;
    /** HashMap of seats - uses seatId (string) for indexing */
    seats = {};
    /** Id of a vehicle - See VehicleIds in enums.ts for more information */
    vehicleId;
    destroyedState = 0;
    currentDamageEffect = 0;
    /** The previous position of the vehicle that was last transmitted to the server,
     * used to determine FairPlay. */
    oldPos = {
        position: new Float32Array(),
        time: 0
    };
    shaderGroupId = 0;
    ownerCharacterId = "";
    droppedManagedClient; // for temporary fix
    isMountable = true;
    isDestroyed = false;
    constructor(characterId, transientId, actorModelId, position, rotation, server, gameTime, vehicleId, shaderGroupId = 0, ownerCharacterId = "") {
        super(characterId, transientId, actorModelId, position, rotation, server);
        this.positionUpdateType = enums_1.PositionUpdateType.MOVABLE;
        this._resources = {
            [enums_1.ResourceIds.CONDITION]: 100000,
            [enums_1.ResourceIds.FUEL]: 7500
        };
        this.state = {
            position: position,
            rotation: rotation,
            lookAt: new Float32Array([0, 0, 0, 1]),
            yaw: 0
        };
        this.vehicleId = vehicleId;
        if (!this.actorModelId)
            this.actorModelId = getActorModelId(this.vehicleId);
        this.getMaterialType(server, this.actorModelId);
        this.loadoutId = getVehicleLoadoutId(this.vehicleId);
        this.defaultLoadout = getDefaultLoadout(this.loadoutId);
        this.npcRenderDistance = 310;
        this.isInvulnerable =
            this.vehicleId == enums_1.VehicleIds.SPECTATE ||
                this.vehicleId == enums_1.VehicleIds.PARACHUTE;
        this.shaderGroupId = shaderGroupId;
        this.ownerCharacterId = ownerCharacterId;
        switch (this.vehicleId) {
            case enums_1.VehicleIds.PICKUP:
                this.seats = {
                    0: "",
                    1: "",
                    2: "",
                    3: "",
                    4: "",
                    5: "",
                    6: ""
                };
                break;
            case enums_1.VehicleIds.OFFROADER:
            case enums_1.VehicleIds.POLICECAR:
                this.seats = {
                    0: "",
                    1: "",
                    2: "",
                    3: "",
                    4: ""
                };
                break;
            case enums_1.VehicleIds.ATV:
                this.seats = {
                    0: "",
                    1: ""
                };
                break;
            default:
                this.seats = {
                    0: ""
                };
                break;
        }
        Object.seal(this.seats); // object can't be edited, but properties can
        this.positionUpdate = {
            ...(0, utils_1.createPositionUpdate)(this.state.position, this.state.rotation, gameTime)
        };
        this.nameId = getVehicleName(this.actorModelId);
        switch (this.vehicleId) {
            case enums_1.VehicleIds.PICKUP: {
                this.destroyedEffect = enums_1.Effects.VEH_Death_PickupTruck;
                this.destroyedModel = enums_1.ModelIds.PICKUP_DESTROYED;
                this.minorDamageEffect = enums_1.Effects.VEH_Damage_PickupTruck_Stage01;
                this.majorDamageEffect = enums_1.Effects.VEH_Damage_PickupTruck_Stage02;
                this.criticalDamageEffect = enums_1.Effects.VEH_Damage_PickupTruck_Stage03;
                this.supercriticalDamageEffect = enums_1.Effects.VEH_Damage_PickupTruck_Stage04;
                break;
            }
            case enums_1.VehicleIds.POLICECAR:
                this.destroyedEffect = enums_1.Effects.VEH_Death_PoliceCar;
                this.destroyedModel = enums_1.ModelIds.POLICE_CAR_DESTROYED;
                this.minorDamageEffect = enums_1.Effects.VEH_Damage_PoliceCar_Stage01;
                this.majorDamageEffect = enums_1.Effects.VEH_Damage_PoliceCar_Stage02;
                this.criticalDamageEffect = enums_1.Effects.VEH_Damage_PoliceCar_Stage03;
                this.supercriticalDamageEffect = enums_1.Effects.VEH_Damage_PoliceCar_Stage04;
                break;
            case enums_1.VehicleIds.ATV: {
                const allowedShaders = [838, 1143, 837, 1003, 1148];
                const randomShader = allowedShaders[Math.floor(Math.random() * allowedShaders.length)];
                if (this.shaderGroupId == 0) {
                    this.shaderGroupId = randomShader;
                }
                this.destroyedEffect = enums_1.Effects.VEH_Death_ATV;
                this.destroyedModel = enums_1.ModelIds.ATV_DESTROYED;
                this.minorDamageEffect = enums_1.Effects.VEH_Damage_ATV_Stage01;
                this.majorDamageEffect = enums_1.Effects.VEH_Damage_ATV_Stage02;
                this.criticalDamageEffect = enums_1.Effects.VEH_Damage_ATV_Stage03;
                this.supercriticalDamageEffect = enums_1.Effects.VEH_Damage_ATV_Stage04;
                break;
            }
            case enums_1.VehicleIds.PARACHUTE:
                this.destroyedEffect = 0;
                this.destroyedModel = 0;
                this.minorDamageEffect = 0;
                this.majorDamageEffect = 0;
                this.criticalDamageEffect = 0;
                this.supercriticalDamageEffect = 0;
                break;
            case enums_1.VehicleIds.OFFROADER:
            default: {
                const allowedShaders = [838, 1143, 837, 1003, 1148];
                const randomShader = allowedShaders[Math.floor(Math.random() * allowedShaders.length)];
                if (this.shaderGroupId == 0) {
                    this.shaderGroupId = randomShader;
                }
                this.destroyedEffect = enums_1.Effects.VEH_Death_OffRoader;
                this.destroyedModel = enums_1.ModelIds.OFFROADER_DESTROYED;
                this.minorDamageEffect = enums_1.Effects.VEH_Damage_OffRoader_Stage01;
                this.majorDamageEffect = enums_1.Effects.VEH_Damage_OffRoader_Stage02;
                this.criticalDamageEffect = enums_1.Effects.VEH_Damage_OffRoader_Stage03;
                this.supercriticalDamageEffect = enums_1.Effects.VEH_Damage_OffRoader_Stage04;
                break;
            }
        }
        server.aiManager.addEntity(this);
    }
    getSeatCount() {
        return Object.keys(this.seats).length;
    }
    getNextSeatId(server) {
        for (const seatId in this.seats) {
            const seat = this.seats[seatId], passenger = seat ? server._characters[seat] : undefined;
            if (!this.seats[seatId] || !passenger?.isAlive) {
                if (Number(seatId) > 4)
                    return -1;
                if (seatId === "2") {
                    if (!!this.seats[5])
                        continue;
                }
                if (seatId === "3") {
                    if (!!this.seats[6])
                        continue;
                }
                return Number(seatId);
            }
        }
        return -1;
    }
    getCharacterSeat(characterId) {
        for (const seatId in this.seats) {
            if (this.seats[seatId] === characterId) {
                return Number(seatId);
            }
        }
        return -1;
    }
    getPassengerList() {
        const passengers = [];
        for (const seatId in this.seats) {
            if (this.seats[seatId]) {
                passengers.push(this.seats[seatId]);
            }
        }
        return passengers;
    }
    removePassenger(characterId) {
        for (const seatId in this.seats) {
            if (this.seats[seatId] === characterId) {
                this.seats[seatId] = "";
                break;
            }
        }
    }
    pGetLightweightVehicle() {
        return {
            npcData: {
                ...this.pGetLightweight(),
                position: this.state.position,
                vehicleId: this.vehicleId,
                shaderGroupId: this.shaderGroupId
            },
            positionUpdate: this.positionUpdate
        };
    }
    pGetFull(server) {
        return {
            transientId: this.transientId,
            attachmentData: this.pGetAttachmentSlots(),
            characterId: this.characterId,
            resources: {
                data: this.pGetResources()
            },
            effectTags: [],
            unknownData1: {},
            targetData: {},
            unknownArray1: [],
            unknownArray2: [],
            unknownArray3: { data: [] },
            unknownArray4: {},
            unknownArray5: { data: [] },
            materialType: this.materialType,
            remoteWeapons: {
                isVehicle: true,
                data: {}
            },
            itemsData: {
                items: this.pGetInventoryItems(server),
                unknownDword1: 0
            }
        };
    }
    pGetFullVehicle(server) {
        return {
            npcData: {
                ...this.pGetFull(server)
            },
            engineState: this.engineOn ? 6 : 0,
            unknownArray1: [],
            unknownArray2: [],
            passengers: this.pGetPassengers(server),
            unknownArray3: [],
            unknownArray4: []
        };
    }
    pGetPassengers(server) {
        return this.getPassengerList().map((passenger) => {
            return {
                characterId: passenger,
                identity: {
                    characterName: server._characters[passenger]?.name ?? ""
                },
                unknownString1: server._characters[passenger]?.name ?? "",
                unknownByte1: 1
            };
        });
    }
    getInventoryItemId() {
        switch (this.loadoutId) {
            case enums_1.LoadoutIds.VEHICLE_OFFROADER:
                return enums_1.Items.CONTAINER_VEHICLE_OFFROADER;
            case enums_1.LoadoutIds.VEHICLE_PICKUP:
                return enums_1.Items.CONTAINER_VEHICLE_PICKUP;
            case enums_1.LoadoutIds.VEHICLE_POLICECAR:
                return enums_1.Items.CONTAINER_VEHICLE_POLICECAR;
            case enums_1.LoadoutIds.VEHICLE_ATV:
                return enums_1.Items.CONTAINER_VEHICLE_ATV;
            default:
                return 0;
        }
    }
    getTurboItemId() {
        switch (this.loadoutId) {
            case enums_1.LoadoutIds.VEHICLE_OFFROADER:
                return enums_1.Items.TURBO_OFFROADER;
            case enums_1.LoadoutIds.VEHICLE_PICKUP:
                return enums_1.Items.TURBO_PICKUP;
            case enums_1.LoadoutIds.VEHICLE_POLICECAR:
                return enums_1.Items.TURBO_POLICE;
            case enums_1.LoadoutIds.VEHICLE_ATV:
                return enums_1.Items.TURBO_ATV;
            default:
                return 0;
        }
    }
    getHeadlightsItemId() {
        switch (this.loadoutId) {
            case enums_1.LoadoutIds.VEHICLE_OFFROADER:
                return enums_1.Items.HEADLIGHTS_OFFROADER;
            case enums_1.LoadoutIds.VEHICLE_PICKUP:
                return enums_1.Items.HEADLIGHTS_PICKUP;
            case enums_1.LoadoutIds.VEHICLE_POLICECAR:
                return enums_1.Items.HEADLIGHTS_POLICE;
            case enums_1.LoadoutIds.VEHICLE_ATV:
                return enums_1.Items.HEADLIGHTS_ATV;
            default:
                return 0;
        }
    }
    getMotorItemId() {
        switch (this.loadoutId) {
            case enums_1.LoadoutIds.VEHICLE_OFFROADER:
                return enums_1.Items.VEHICLE_MOTOR_OFFROADER;
            case enums_1.LoadoutIds.VEHICLE_PICKUP:
                return enums_1.Items.VEHICLE_MOTOR_PICKUP;
            case enums_1.LoadoutIds.VEHICLE_POLICECAR:
                return enums_1.Items.VEHICLE_MOTOR_POLICECAR;
            case enums_1.LoadoutIds.VEHICLE_ATV:
                return enums_1.Items.VEHICLE_MOTOR_ATV;
            default:
                return 0;
        }
    }
    startDamageDelay(server) {
        this.damageTimeout = setTimeout(() => {
            this.damage(server, { entity: "", damage: 1000 });
            if (this._resources[enums_1.ResourceIds.CONDITION] < 20000 &&
                this._resources[enums_1.ResourceIds.CONDITION] > 0) {
                this.damageTimeout.refresh();
            }
            else {
                delete this.damageTimeout;
            }
        }, 1000);
    }
    getHealth() {
        return this._resources[enums_1.ResourceIds.CONDITION];
    }
    hasVehicleDamageBoost(weapon) {
        switch (weapon) {
            case enums_1.Items.WEAPON_308:
            case enums_1.Items.WEAPON_MAGNUM:
                return true;
            default:
                return false;
        }
    }
    damage(server, damageInfo) {
        if (this.isInvulnerable || this.isDestroyed)
            return;
        const oldHealth = this._resources[enums_1.ResourceIds.CONDITION];
        this._resources[enums_1.ResourceIds.CONDITION] -= this.hasVehicleDamageBoost(damageInfo.weapon)
            ? 8000
            : damageInfo.damage;
        const client = server.getClientByCharId(damageInfo.entity);
        if (client) {
            queueMicrotask(async () => {
                client.character.addCombatlogEntry(await server.generateDamageRecord(this.characterId, damageInfo, oldHealth));
            });
        }
        if (this._resources[enums_1.ResourceIds.CONDITION] <= 0) {
            if (client && !server._soloMode) {
                (0, utils_1.logClientActionToMongo)(server._db.collection(enums_2.DB_COLLECTIONS.KILLS), client, server._worldId, { type: enums_2.KILL_TYPE.VEHICLE });
            }
            this.destroy(server);
            if (client) {
                client.character.metrics.vehiclesDestroyed++;
            }
        }
        else {
            let damageeffect = 0;
            let allowSend = false;
            let startDamageTimeout = false;
            if (this._resources[enums_1.ResourceIds.CONDITION] <= 50000 &&
                this._resources[enums_1.ResourceIds.CONDITION] > 35000) {
                if (this.destroyedState != 1) {
                    damageeffect = this.minorDamageEffect;
                    allowSend = true;
                    this.destroyedState = 1;
                }
            }
            else if (this._resources[enums_1.ResourceIds.CONDITION] <= 35000 &&
                this._resources[enums_1.ResourceIds.CONDITION] > 20000) {
                if (this.destroyedState != 2) {
                    damageeffect = this.majorDamageEffect;
                    allowSend = true;
                    this.destroyedState = 2;
                }
            }
            else if (this._resources[enums_1.ResourceIds.CONDITION] <= 20000 &&
                this._resources[enums_1.ResourceIds.CONDITION] > 10000) {
                if (this.destroyedState != 3) {
                    damageeffect = this.criticalDamageEffect;
                    allowSend = true;
                    startDamageTimeout = true;
                    this.destroyedState = 3;
                }
            }
            else if (this._resources[enums_1.ResourceIds.CONDITION] <= 10000) {
                if (this.destroyedState != 4) {
                    damageeffect = this.supercriticalDamageEffect;
                    allowSend = true;
                    startDamageTimeout = true;
                    this.destroyedState = 4;
                }
            }
            else if (this._resources[enums_1.ResourceIds.CONDITION] > 50000 &&
                this.destroyedState != 0) {
                this.destroyedState = 0;
                damageeffect = 0;
                this.currentDamageEffect = 0;
                allowSend = true;
            }
            if (allowSend) {
                this.currentDamageEffect = damageeffect;
                server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Command.PlayDialogEffect", {
                    characterId: this.characterId,
                    effectId: damageeffect
                });
                if (!this.damageTimeout && startDamageTimeout) {
                    this.startDamageDelay(server);
                }
            }
            server.updateResourceToAllWithSpawnedEntity(this.characterId, this._resources[enums_1.ResourceIds.CONDITION], enums_1.ResourceIds.CONDITION, enums_1.ResourceTypes.CONDITION, server._vehicles);
        }
    }
    updateLoadout(server) {
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Loadout.SetLoadoutSlots", this.pGetLoadoutSlots());
    }
    getDriver(server) {
        const seat = this.seats[0];
        if (seat)
            return server._characters[seat];
    }
    doesPassengersHaveKey(server) {
        for (const seatId in this.seats) {
            const seat = this.seats[seatId], passenger = seat ? server._characters[seat] : undefined;
            if (passenger?.getItemById(enums_1.Items.VEHICLE_KEY)) {
                return true;
            }
        }
        return false;
    }
    startEngine(server) {
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Vehicle.Engine", {
            vehicleCharacterId: this.characterId,
            engineOn: true
        });
        this.engineOn = true;
        this.startResourceUpdater(server);
    }
    stopEngine(server) {
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Vehicle.Engine", {
            vehicleCharacterId: this.characterId,
            engineOn: false
        });
        this.engineOn = false;
        const driver = this.getDriver(server);
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Effect.RemoveEffect", {
            abilityEffectData: {
                unknownDword1: 4,
                abilityEffectId1: enums_1.VehicleEffects.MOTOR_RUN_OFFROADER_1,
                abilityEffectId2: enums_1.VehicleEffects.MOTOR_RUN_OFFROADER_2
            },
            targetCharacterData: {
                characterId: this.characterId
            },
            guid2: "0x0",
            targetCharacterId: driver?.characterId ?? ""
        });
    }
    getHeadlightState() {
        const headlightEffect = getHeadlightEffect(this.vehicleId), index = this.effectTags.indexOf(headlightEffect);
        return !(index <= -1);
    }
    getSirenState() {
        const sirenEffect = enums_1.Effects.VEH_SirenLight_PoliceCar, index = this.effectTags.indexOf(sirenEffect);
        return !(index <= -1);
    }
    getHornState() {
        const hornEffect = getHornEffect(this.vehicleId), index = this.effectTags.indexOf(hornEffect);
        return !(index <= -1);
    }
    getTurboState() {
        const turboEffect = getTurboEffect(this.vehicleId), index = this.effectTags.indexOf(turboEffect);
        return !(index <= -1);
    }
    sendNoPartsAlert(server, client) {
        server.sendAlert(client, "Parts may be required. Open vehicle loadout.");
    }
    sendNoFuelAlert(server, client) {
        server.sendAlert(client, "This vehicle will not run without fuel.  It can be created from animal fat or from corn based ethanol.");
    }
    setHeadlightState(server, state, client) {
        const headlightEffect = getHeadlightEffect(this.vehicleId), index = this.effectTags.indexOf(headlightEffect);
        if (state && index <= -1) {
            if (!this.hasBattery() || !this.hasHeadlights()) {
                if (client)
                    this.sendNoPartsAlert(server, client);
                return;
            }
            server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Character.AddEffectTagCompositeEffect", {
                characterId: this.characterId,
                effectId: headlightEffect,
                unknownDword1: headlightEffect,
                unknownDword2: headlightEffect
            });
            this.effectTags.push(headlightEffect);
            return;
        }
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Character.RemoveEffectTagCompositeEffect", {
            characterId: this.characterId,
            effectId: headlightEffect,
            newEffectId: 0
        });
        this.effectTags.splice(index, 1);
    }
    setSirenState(server, state, client) {
        if (this.vehicleId != enums_1.VehicleIds.POLICECAR)
            return;
        const sirenEffect = enums_1.Effects.VEH_SirenLight_PoliceCar, index = this.effectTags.indexOf(sirenEffect);
        if (state && index <= -1) {
            if (!this.hasBattery()) {
                if (client)
                    this.sendNoPartsAlert(server, client);
                return;
            }
            server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Character.AddEffectTagCompositeEffect", {
                characterId: this.characterId,
                effectId: sirenEffect,
                unknownDword1: sirenEffect,
                unknownDword2: sirenEffect
            });
            this.effectTags.push(sirenEffect);
            return;
        }
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Character.RemoveEffectTagCompositeEffect", {
            characterId: this.characterId,
            effectId: sirenEffect,
            newEffectId: 0
        });
        this.effectTags.splice(index, 1);
    }
    setHornState(server, state, client) {
        const hornEffect = getHornEffect(this.vehicleId), index = this.effectTags.indexOf(hornEffect);
        if (state && index <= -1) {
            if (!this.hasBattery()) {
                if (client)
                    this.sendNoPartsAlert(server, client);
                return;
            }
            server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Character.AddEffectTagCompositeEffect", {
                characterId: this.characterId,
                effectId: hornEffect,
                unknownDword1: hornEffect,
                unknownDword2: hornEffect
            });
            this.effectTags.push(hornEffect);
            return;
        }
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Character.RemoveEffectTagCompositeEffect", {
            characterId: this.characterId,
            effectId: hornEffect,
            newEffectId: 0
        });
        this.effectTags.splice(index, 1);
    }
    setTurboState(server, client, state) {
        const turboEffect = getTurboEffect(this.vehicleId), index = this.effectTags.indexOf(turboEffect);
        if (state && index <= -1) {
            if (!this.hasBattery()) {
                if (client)
                    this.sendNoPartsAlert(server, client);
                return;
            }
            server.abilitiesManager.addEffectTag(server, client, this, turboEffect, server._vehicles);
            return;
        }
        server.abilitiesManager.removeEffectTag(server, client, this, turboEffect, server._vehicles);
    }
    hasTurbo() {
        return (!!this.getLoadoutItemById(enums_1.Items.TURBO_OFFROADER) ||
            !!this.getLoadoutItemById(enums_1.Items.TURBO_PICKUP) ||
            !!this.getLoadoutItemById(enums_1.Items.TURBO_POLICE) ||
            !!this.getLoadoutItemById(enums_1.Items.TURBO_ATV));
    }
    hasHeadlights() {
        return (!!this.getLoadoutItemById(enums_1.Items.HEADLIGHTS_OFFROADER) ||
            !!this.getLoadoutItemById(enums_1.Items.HEADLIGHTS_PICKUP) ||
            !!this.getLoadoutItemById(enums_1.Items.HEADLIGHTS_POLICE) ||
            !!this.getLoadoutItemById(enums_1.Items.HEADLIGHTS_ATV));
    }
    hasBattery() {
        return !!this.getLoadoutItemById(enums_1.Items.BATTERY);
    }
    hasRequiredEngineParts() {
        return !!this.hasBattery() && !!this.getLoadoutItemById(enums_1.Items.SPARKPLUGS);
    }
    hasVehicleKey(server) {
        return (this.vehicleId == enums_1.VehicleIds.PARACHUTE ||
            !!this.getItemById(enums_1.Items.VEHICLE_KEY) ||
            this.doesPassengersHaveKey(server));
    }
    hasFuel() {
        return this._resources[enums_1.ResourceIds.FUEL] > 0;
    }
    hasRequiredComponents(server) {
        return (this.hasRequiredEngineParts() &&
            this.hasVehicleKey(server) &&
            this.hasFuel());
    }
    checkEngineRequirements(server, runEngine = true) {
        if (this.hasRequiredComponents(server) && !this.engineOn && runEngine) {
            this.startEngine(server);
            return;
        }
        const driver = this.getDriver(server), client = server.getClientByCharId(driver?.characterId || "");
        if (this.getHeadlightState() &&
            (!this.hasHeadlights() || !this.hasBattery())) {
            this.setHeadlightState(server, false);
        }
        if (this.getSirenState() && !this.hasBattery()) {
            this.setSirenState(server, false);
        }
        if (this.getHornState() && !this.hasBattery()) {
            this.setHornState(server, false);
        }
        if (!this.hasRequiredEngineParts()) {
            if (this.engineOn)
                this.stopEngine(server);
            if (client)
                this.sendNoPartsAlert(server, client);
            return;
        }
        if (!this.hasVehicleKey(server)) {
            if (this.engineOn)
                this.stopEngine(server);
            if (client)
                this.hotwire(server);
            return;
        }
        if (!this.hasFuel()) {
            if (this.engineOn)
                this.stopEngine(server);
            if (client)
                this.sendNoFuelAlert(server, client);
            return;
        }
    }
    removeHotwireEffect(server) {
        const hotwireEffect = getHotwireEffect(this.vehicleId), index = this.effectTags.indexOf(hotwireEffect);
        if (index <= -1)
            return;
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Character.RemoveEffectTagCompositeEffect", {
            characterId: this.characterId,
            effectId: hotwireEffect,
            newEffectId: 0
        });
        this.effectTags.splice(index, 1);
    }
    hotwire(server) {
        const driver = this.getDriver(server), client = server.getClientByCharId(driver?.characterId || "");
        if (!client)
            return;
        const hotwireEffect = getHotwireEffect(this.vehicleId), index = this.effectTags.indexOf(hotwireEffect);
        if (!hotwireEffect || index >= 0)
            return;
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Character.AddEffectTagCompositeEffect", {
            characterId: this.characterId,
            effectId: hotwireEffect,
            unknownDword1: hotwireEffect,
            unknownDword2: hotwireEffect
        });
        this.effectTags.push(hotwireEffect);
        server.utilizeHudTimer(client, enums_1.StringIds.HOTWIRE, 7000, 0, () => {
            this.removeHotwireEffect(server);
            this.startEngine(server);
        });
    }
    startResourceUpdater(server) {
        if (this.resourcesUpdater)
            return;
        this.resourcesUpdater = setTimeout(() => {
            if (!server._vehicles[this.characterId])
                return;
            if (!this.engineOn) {
                delete this.resourcesUpdater;
                return;
            }
            if (this.engineRPM) {
                const fuelLoss = this.engineRPM * 0.003 * (this.getTurboState() ? 4 : 1);
                this._resources[enums_1.ResourceIds.FUEL] -= fuelLoss;
            }
            if (this._resources[enums_1.ResourceIds.FUEL] < 0) {
                this._resources[enums_1.ResourceIds.FUEL] = 0;
            }
            if (this.engineOn && this._resources[enums_1.ResourceIds.FUEL] <= 0) {
                this.stopEngine(server);
                const driver = this.getDriver(server), client = server.getClientByCharId(driver?.characterId || "");
                if (client)
                    this.sendNoFuelAlert(server, client);
            }
            server.updateResourceToAllWithSpawnedEntity(this.characterId, this._resources[enums_1.ResourceIds.FUEL], enums_1.ResourceIds.FUEL, enums_1.ResourceTypes.FUEL, server._vehicles);
            this.resourcesUpdater.refresh();
        }, 3000);
    }
    setLockState(server, client, state) {
        server.sendData(client, "Vehicle.AccessType", {
            vehicleGuid: this.characterId,
            accessType: state ? 2 : 0
        });
        this.isLocked = state;
    }
    pGetLoadoutSlots() {
        return {
            characterId: this.characterId,
            loadoutId: this.loadoutId,
            loadoutData: {
                loadoutSlots: Object.values(this.getLoadoutSlots()).map((slotId) => {
                    return this.pGetLoadoutSlot(slotId);
                })
            },
            currentSlotId: this.currentLoadoutSlot
        };
    }
    isFlipped() {
        return Math.abs(this.positionUpdate.sideTilt) > 1.75;
    }
    flipVehicle(server) {
        let c;
        for (const a in server._clients) {
            if (server._clients[a].managedObjects.includes(this.characterId)) {
                c = server._clients[a];
            }
        }
        if (c) {
            this.positionUpdate.sideTilt = 0;
            server.sendData(c, "ClientUpdate.UpdateManagedLocation", {
                characterId: this.characterId,
                position: this.state.position,
                rotation: (0, utils_1.eul2quat)(new Float32Array([
                    this.positionUpdate.orientation,
                    this.positionUpdate.sideTilt,
                    this.positionUpdate.frontTilt
                ]))
            });
        }
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client, isInstant
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        if (client.vehicle.mountedVehicle) {
            server.dismountVehicle(client);
            return;
        }
        if (this.isFlipped()) {
            this.flipVehicle(server);
            return;
        }
        server.mountVehicle(client, this.characterId);
    }
    OnInteractionString(server, client) {
        if (client.vehicle.mountedVehicle)
            return;
        if (this.isFlipped()) {
            server.sendData(client, "Command.InteractionString", {
                guid: this.characterId,
                stringId: enums_1.StringIds.USE_TARGET
            });
            return;
        }
        server.sendData(client, "Command.InteractionString", {
            guid: this.characterId,
            stringId: enums_1.StringIds.ENTER_VEHICLE
        });
    }
    pGetItemData(server, item, containerDefId) {
        return {
            itemDefinitionId: item.itemDefinitionId,
            tintId: 0,
            guid: item.itemGuid,
            count: item.stackCount,
            itemSubData: {
                hasSubData: false
            },
            containerGuid: item.containerGuid,
            containerDefinitionId: containerDefId,
            containerSlotId: item.slotId,
            baseDurability: server.getItemBaseDurability(item.itemDefinitionId),
            currentDurability: item.currentDurability,
            maxDurabilityFromDefinition: server.getItemBaseDurability(item.itemDefinitionId),
            unknownBoolean1: true,
            ownerCharacterId: this.characterId,
            unknownDword9: 1,
            weaponData: this.pGetItemWeaponData(server, item)
        };
    }
    OnFullCharacterDataRequest(server, client) {
        if (this.vehicleId == enums_1.VehicleIds.SPECTATE) {
            return;
        }
        server.sendData(client, "LightweightToFullVehicle", this.pGetFullVehicle(server));
        Object.values(this._loadout).forEach((item) => {
            server.sendData(client, "ClientUpdate.ItemAdd", {
                characterId: this.characterId,
                data: this.pGetItemData(server, item, constants_1.LOADOUT_CONTAINER_ID)
            });
        });
        this.updateLoadout(server);
        // fix seat change crash related to our managed object workaround
        if (this.droppedManagedClient == client) {
            const seatId = this.getCharacterSeat(client.character.characterId);
            server.sendData(client, "Mount.MountResponse", {
                characterId: client.character.characterId,
                vehicleGuid: this.characterId, // vehicle guid
                seatId: seatId,
                isDriver: seatId == 0 ? 1 : 0, //isDriver
                identity: {}
            });
            delete this.droppedManagedClient;
        }
        server.sendData(client, "ResourceEvent", {
            eventData: {
                type: 1,
                value: {
                    characterId: this.characterId,
                    characterResources: this.pGetResources()
                }
            }
        });
        // disable this workaround for now
        /*for (const a in this.seats) {
          const seatId = this.getCharacterSeat(this.seats[a]);
          if (!this.seats[a]) continue;
          server.sendData(client, "Mount.MountResponse", {
            // mounts character
            characterId: this.seats[a],
            vehicleGuid: this.characterId, // vehicle guid
            seatId: seatId,
            unknownDword3: seatId === "0" ? 1 : 0, //isDriver
            identity: {},
          });
        }*/
        if (this.currentDamageEffect != 0) {
            server.sendData(client, "Command.PlayDialogEffect", {
                characterId: this.characterId,
                effectId: this.currentDamageEffect
            });
        }
        if (this.onReadyCallback &&
            (this.ownerCharacterId == "" ||
                this.ownerCharacterId == client.character.characterId)) {
            this.onReadyCallback(client);
            delete this.onReadyCallback;
        }
    }
    OnMeleeHit(server, damageInfo) {
        const client = server.getClientByCharId(damageInfo.entity), weapon = client?.character.getEquippedWeapon();
        damageInfo.damage = damageInfo.damage * 2;
        if (!client || weapon?.itemDefinitionId != enums_1.Items.WEAPON_WRENCH) {
            if (!server.isPvE) {
                this.damage(server, damageInfo);
            }
            return;
        }
        if (this._resources[enums_1.ResourceIds.CONDITION] < 100000) {
            this.damage(server, { ...damageInfo, damage: -2000 });
            server.damageItem(client.character, weapon, 80);
        }
    }
    OnProjectileHit(server, damageInfo) {
        if (!server.isPvE) {
            this.damage(server, damageInfo);
        }
    }
    destroy(server, disableExplosion = false) {
        this.isDestroyed = true;
        if (!server._vehicles[this.characterId])
            return false;
        this._resources[enums_1.ResourceIds.CONDITION] = 0;
        for (const c in server._clients) {
            if (this.characterId === server._clients[c].vehicle.mountedVehicle) {
                server.dismountVehicle(server._clients[c]);
            }
        }
        server.sendDataToAllWithSpawnedEntity(server._vehicles, this.characterId, "Character.Destroyed", {
            characterId: this.characterId,
            destroyedEffect: this.destroyedEffect,
            destroyedModel: this.destroyedModel,
            unknown3: 0,
            disableWeirdPhysics: false
        });
        const deleted = server.deleteEntity(this.characterId, server._vehicles);
        if (!disableExplosion) {
            server.explosionDamage(this);
        }
        //this.state.position[1] -= 0.4; // makes bags spawn under the map sometimes.
        // TODO: Have to revisit when the heightmap is implemented server side.
        // fix floating vehicle lootbags
        server.worldObjectManager.createLootbag(server, this);
        server.aiManager.removeEntity(this);
        return deleted;
    }
    async OnExplosiveHit(server, sourceEntity) {
        if (this.characterId == sourceEntity.characterId)
            return;
        let damage = 100000;
        switch (true) {
            case sourceEntity instanceof explosiveentity_1.ExplosiveEntity:
                damage = 100000;
                break;
            case sourceEntity instanceof projectileentity_1.ProjectileEntity:
                damage = sourceEntity.actorModelId == 0 ? 50000 : 100000;
                break;
            default:
                damage = 100000;
                break;
        }
        const distance = (0, utils_1.getDistance)(sourceEntity.state.position, this.state.position);
        if (distance > 1)
            damage /= distance;
        this.damage(server, {
            entity: sourceEntity.characterId,
            damage: damage
        });
    }
}
exports.Vehicle2016 = Vehicle2016;
