// h1z1-server/out/servers/ZoneServer2016/entities/watersource.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.WaterSource = void 0;
const enums_1 = require("../models/enums");
const taskprop_1 = require("./taskprop");
class WaterSource extends taskprop_1.TaskProp {
    /** The amount of interactions left on the WaterSource */
    usesLeft;
    /** Amount to be refilled in a water bottle */
    refillAmount;
    /** Returns true if the fire hydrant is currently releasing water */
    isHydrantGushing = false;
    /** Returns true if the hydrant is currently not able to be melee'd for water */
    isHydrantOnCooldown = false;
    constructor(characterId, transientId, actorModelId, position, rotation, server, scale, zoneId, renderDistance, actorModel, refillAmount) {
        super(characterId, transientId, actorModelId, position, rotation, server, scale, zoneId, renderDistance, actorModel);
        this.refillAmount = refillAmount;
    }
    replenish() {
        switch (this.actorModel) {
            case "Common_Props_Cabinets_BathroomSink.adr":
            case "Common_Props_Bathroom_Toilet01.adr":
                this.usesLeft = this.refillAmount;
                break;
            case "Common_Props_Dam_WaterValve01.adr":
            case "Common_Props_Well.adr":
            case "Common_Props_FireHydrant.adr":
                this.usesLeft = Number.MAX_SAFE_INTEGER;
                break;
        }
    }
    OnInteractionString(server, client) {
        switch (this.actorModel) {
            case "Common_Props_Dam_WaterValve01.adr":
            case "Common_Props_Cabinets_BathroomSink.adr":
            case "Common_Props_Bathroom_Toilet01.adr":
                if (this.usesLeft && this.usesLeft > 0) {
                    server.sendData(client, "Command.InteractionString", {
                        guid: this.characterId,
                        stringId: client.character.hasItem(enums_1.Items.WATER_EMPTY)
                            ? enums_1.StringIds.COLLECT_WATER
                            : enums_1.StringIds.DRINK_DIRTY_WATER
                    });
                }
                break;
            case "Common_Props_FireHydrant.adr":
                if (!this.isHydrantGushing) {
                    server.sendData(client, "Command.InteractionString", {
                        guid: this.characterId,
                        stringId: enums_1.StringIds.CHECK_WATER
                    });
                }
                else {
                    server.sendData(client, "Command.InteractionString", {
                        guid: this.characterId,
                        stringId: client.character.hasItem(enums_1.Items.WATER_EMPTY)
                            ? enums_1.StringIds.COLLECT_WATER
                            : enums_1.StringIds.DRINK_DIRTY_WATER
                    });
                }
                break;
            case "Common_Props_Well.adr":
                server.sendData(client, "Command.InteractionString", {
                    guid: this.characterId,
                    stringId: client.character.hasItem(enums_1.Items.WATER_EMPTY)
                        ? enums_1.StringIds.COLLECT_WATER
                        : enums_1.StringIds.DRINK_STAGNANT_WATER
                });
                break;
        }
    }
    /* eslint-disable @typescript-eslint/no-unused-vars */
    OnPlayerSelect(server, client
    /* eslint-enable @typescript-eslint/no-unused-vars */
    ) {
        if (this.actorModelId == enums_1.ModelIds.FIRE_HYDRANT && !this.isHydrantGushing) {
            server.utilizeHudTimer(client, enums_1.StringIds.FIRE_HYDRANT, 3000, 0, () => {
                if (this.isHydrantOnCooldown) {
                    server.sendChatText(client, "This fire hydrant has been drained recently. It will replenish soon.");
                }
                else {
                    server.sendChatText(client, "This fire hydrant is full of water!");
                }
            });
            return;
        }
        const bottle = client.character.getItemById(enums_1.Items.WATER_EMPTY), infiniteSources = [
            "Common_Props_Well.adr",
            "Common_Props_Dam_WaterValve01.adr",
            "Common_Props_FireHydrant.adr"
        ], hasUses = infiniteSources.includes(this.actorModel) ||
            (this.usesLeft && this.usesLeft > 0);
        if (!hasUses) {
            server.utilizeHudTimer(client, enums_1.StringIds.DIRTY_WATER, 250, 0, () => {
                server.sendAlert(client, "This water source is dry.");
            });
            return;
        }
        if (!bottle) {
            server.utilizeHudTimer(client, infiniteSources.includes(this.actorModel)
                ? enums_1.StringIds.WATER_WELL
                : enums_1.StringIds.DIRTY_WATER, 1000, 0, () => {
                client.character._resources[enums_1.ResourceIds.HYDRATION] += 1000;
                client.character.damage(server, { entity: "", damage: 1000 });
                server.updateResource(client, client.character.characterId, client.character._resources[enums_1.ResourceIds.HYDRATION], enums_1.ResourceIds.HYDRATION);
                if (this.usesLeft)
                    this.usesLeft--;
            });
            return;
        }
        server.utilizeHudTimer(client, infiniteSources.includes(this.actorModel)
            ? enums_1.StringIds.WATER_WELL
            : enums_1.StringIds.DIRTY_WATER, 1000, 0, () => {
            if (!server.removeInventoryItem(client.character, bottle))
                return;
            client.character.lootContainerItem(server, server.generateItem(infiniteSources.includes(this.actorModel)
                ? enums_1.Items.WATER_STAGNANT
                : enums_1.Items.WATER_DIRTY));
            if (this.usesLeft)
                this.usesLeft--;
        });
    }
    OnMeleeHit(server, damageInfo) {
        const client = server.getClientByCharId(damageInfo.entity), weapon = client?.character.getEquippedWeapon(), weaponId = weapon?.itemDefinitionId, activatableItems = [
            enums_1.Items.WEAPON_WRENCH,
            enums_1.Items.WEAPON_BRANCH,
            enums_1.Items.WEAPON_HAMMER,
            enums_1.Items.WEAPON_HAMMER_DEMOLITION
        ];
        if (this.actorModelId != enums_1.ModelIds.FIRE_HYDRANT)
            return;
        if (!client || !weaponId || !activatableItems.includes(weaponId))
            return;
        if (this.isHydrantOnCooldown) {
            if (!this.isHydrantGushing) {
                server.sendChatText(client, "This fire hydrant has been drained recently. It will replenish soon.");
            }
            return;
        }
        this.isHydrantGushing = true;
        this.isHydrantOnCooldown = true;
        const pos = this.state.position;
        server.sendDataToAllWithSpawnedEntity(server._taskProps, this.characterId, "Character.PlayWorldCompositeEffect", {
            characterId: this.characterId,
            effectId: enums_1.Effects.EFX_FireHydrant_Gushing,
            position: new Float32Array([pos[0], pos[1], pos[2], 1]),
            effectTime: 30
        });
        setTimeout(() => {
            this.isHydrantGushing = false;
        }, 30000);
        setTimeout(() => {
            this.isHydrantOnCooldown = false;
        }, 300000);
    }
}
exports.WaterSource = WaterSource;
