// h1z1-server/out/servers/SoeServer/soeclient.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils/utils");
const soeinputstream_1 = require("./soeinputstream");
const soeoutputstream_1 = require("./soeoutputstream");
const PacketsQueue_1 = require("./PacketsQueue");
const node_timers_1 = require("node:timers");
class SOEClient {
    sessionId = 0;
    address;
    port;
    family;
    crcSeed;
    crcLength = 2;
    clientUdpLength = 512;
    serverUdpLength = 512;
    packetsSentThisSec = 0;
    useEncryption = true;
    waitingQueue = new PacketsQueue_1.PacketsQueue();
    protocolName = "unset";
    unAckData = new Map();
    lastAckSend = new utils_1.wrappedUint16(-1);
    inputStream;
    outputStream;
    soeClientId;
    lastKeepAliveTimer = null;
    isDeleted = false;
    stats = {
        totalPhysicalPacketSent: 0,
        totalLogicalPacketSent: 0,
        packetsOutOfOrder: 0,
        packetResend: 0
    };
    avgPing = 0;
    pings = [];
    avgPingLen = 6;
    _statsResetTimer;
    delayedLogicalPackets = [];
    constructor(remote, crcSeed, cryptoKey) {
        this.soeClientId = SOEClient.getClientId(remote);
        this.address = remote.address;
        this.port = remote.port;
        this.family = remote.family;
        this.crcSeed = crcSeed;
        this.inputStream = new soeinputstream_1.SOEInputStream(cryptoKey);
        this.outputStream = new soeoutputstream_1.SOEOutputStream(cryptoKey);
        if (remote.address === "127.0.0.1") {
            this.outputStream = new soeoutputstream_1.SOEOutputStream(cryptoKey, 100);
        }
        else {
            this.outputStream = new soeoutputstream_1.SOEOutputStream(cryptoKey);
        }
        this._statsResetTimer = setInterval(() => this._resetStats(), 60000);
    }
    static getClientId(remote) {
        return remote.address + ":" + remote.port;
    }
    closeTimers() {
        (0, node_timers_1.clearInterval)(this._statsResetTimer);
    }
    _resetStats() {
        this.stats.totalPhysicalPacketSent = 0;
        this.stats.packetsOutOfOrder = 0;
        this.stats.packetResend = 0;
        this.stats.totalLogicalPacketSent = 0;
    }
    getNetworkStats() {
        const { totalPhysicalPacketSent: totalPacketSent, packetResend, packetsOutOfOrder } = this.stats;
        const packetLossRate = Number((packetResend / totalPacketSent).toFixed(3)) * 100;
        const packetOutOfOrderRate = Number((packetsOutOfOrder / totalPacketSent).toFixed(3)) * 100;
        return [
            `Packet loss rate ${packetLossRate}%`,
            `Packet outOfOrder rate ${packetOutOfOrderRate}%`,
            `Avg ping ${this.avgPing}ms`
        ];
    }
    addPing(ping) {
        if (ping > 0) {
            this.pings.push(ping);
        }
        if (this.pings.length > this.avgPingLen) {
            this.pings.shift();
        }
        this._updateAvgPing();
    }
    _updateAvgPing() {
        this.avgPing = (0, utils_1.toInt)(utils_1._.sum(this.pings) / this.pings.length);
    }
}
exports.default = SOEClient;
