// h1z1-server/out/packets/ClientProtocol/ClientProtocol_1080/items.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.itemsPackets = void 0;
exports.parseItemRequestSubData = parseItemRequestSubData;
exports.parseAccountItemRequestSubData = parseAccountItemRequestSubData;
const shared_1 = require("./shared");
function parseItemRequestSubData(data, offset) {
    const obj = {}, startOffset = offset;
    obj["unknownBoolean1"] = data.readUInt8(offset);
    offset += 1;
    if (!obj["unknownBoolean1"]) {
        obj["unknownDword1"] = data.readUInt32LE(offset);
        offset += 4;
        obj["unknownDword2"] = data.readUInt32LE(offset);
        offset += 4;
        obj["subType"] = data.readUInt32LE(offset);
        offset += 4;
        if (obj["subType"] == 1) {
            obj["count"] = data.readUInt32LE(offset);
            offset += 4;
            obj["unknownQword1"] = data.readUInt64String(offset);
            offset += 8;
            obj["unknownByte1"] = data.readUInt8(offset);
            offset += 1;
            return {
                value: obj,
                length: offset - startOffset
            };
        }
        obj["unknownDword3"] = data.readUInt32LE(offset);
        offset += 4;
        obj["unknownDword4"] = data.readUInt32LE(offset);
        offset += 4;
        obj["targetItemGuid"] = data.readUInt64String(offset);
        offset += 8;
        obj["unknownDword5"] = data.readUInt32LE(offset);
        offset += 4;
        obj["unknownDword6"] = data.readUInt32LE(offset);
        offset += 4;
        obj["characterId"] = data.readUInt64String(offset);
        offset += 8;
        obj["unknownQword1"] = data.readUInt64String(offset);
        offset += 8;
    }
    return {
        value: obj,
        length: offset - startOffset
    };
}
// This structure is not 100% correct, I just guessed at this point
function parseAccountItemRequestSubData(data, offset) {
    const obj = {}, startOffset = offset;
    obj["unknownBoolean1"] = data.readUInt8(offset);
    offset += 1;
    if (!obj["unknownBoolean1"]) {
        obj["unknownDword1"] = data.readUInt32LE(offset);
        offset += 4;
        obj["unknownDword2"] = data.readUInt32LE(offset);
        offset += 4;
        obj["unknownDword3"] = data.readUInt32LE(offset);
        offset += 4;
        obj["unknownDword4"] = data.readUInt32LE(offset);
        offset += 4;
        obj["unknownDword5"] = data.readUInt32LE(offset);
        offset += 4;
        obj["targetItemGuid"] = data.readUInt64String(offset);
        offset += 8;
        obj["unknownQword1"] = data.readUInt64String(offset);
        offset += 8;
    }
    return {
        value: obj,
        length: offset - startOffset
    };
}
exports.itemsPackets = [
    ["Items.LoadItemRentalDefinitionManager", 0xad01, {}],
    [
        "Items.SetItemTimerManager",
        0xad02,
        {
            fields: [
                // sub_140373A20
                { name: "unknownDword1", type: "uint32", defaultValue: 0 },
                { name: "unknownDword2", type: "uint32", defaultValue: 0 },
                { name: "unknownQword1", type: "uint64string", defaultValue: "0" },
                { name: "unknownQword2", type: "uint64string", defaultValue: "0" },
                // end sub_140373A20
                {
                    // sub_14037CF80
                    name: "unknownArray1",
                    type: "array",
                    defaultValue: [],
                    fields: [
                        { name: "unknownDword1_", type: "uint32", defaultValue: "0" },
                        // sub_140373A20
                        { name: "unknownDword1", type: "uint32", defaultValue: 0 },
                        { name: "unknownDword2", type: "uint32", defaultValue: 0 },
                        { name: "unknownQword1", type: "uint64string", defaultValue: "0" },
                        { name: "unknownQword2", type: "uint64string", defaultValue: "0" },
                        // end sub_140373A20
                        { name: "unknownDword2_", type: "uint32", defaultValue: "0" }
                    ]
                },
                { name: "unknownDword1_", type: "uint32", defaultValue: 0 },
                { name: "unknownDword2_", type: "uint32", defaultValue: 0 },
                { name: "unknownQword1_", type: "uint64string", defaultValue: "0" },
                {
                    // sub_14037D0F0
                    name: "unknownArray2",
                    type: "array",
                    defaultValue: [],
                    fields: [
                        { name: "unknownDword1_", type: "uint32", defaultValue: "0" },
                        {
                            // sub_1403702F0
                            name: "unknownData1",
                            type: "schema",
                            defaultValue: {},
                            fields: [
                                {
                                    // sub_140373AD0
                                    name: "unknownData1",
                                    type: "schema",
                                    defaultValue: {},
                                    fields: [
                                        {
                                            name: "unknownDword1",
                                            type: "uint32",
                                            defaultValue: 0
                                        },
                                        {
                                            name: "unknownDword2",
                                            type: "uint32",
                                            defaultValue: 0
                                        },
                                        {
                                            name: "unknownQword1",
                                            type: "uint64string",
                                            defaultValue: ""
                                        }
                                    ]
                                },
                                {
                                    name: "unknownDword1",
                                    type: "uint32",
                                    defaultValue: 0
                                },
                                {
                                    name: "unknownDword2",
                                    type: "uint32",
                                    defaultValue: 0
                                },
                                {
                                    name: "unknownByte1",
                                    type: "uint8",
                                    defaultValue: 0
                                }
                            ]
                        }
                    ]
                },
                {
                    // sub_14037CE40
                    name: "unknownArray3",
                    type: "array",
                    defaultValue: [],
                    fields: [
                        { name: "unknownDword1", type: "uint32", defaultValue: 0 },
                        {
                            // sub_1403701A0
                            name: "unknownData1",
                            type: "schema",
                            defaultValue: {},
                            fields: [
                                {
                                    name: "unknownData1",
                                    type: "schema",
                                    defaultValue: {},
                                    fields: [
                                        {
                                            name: "unknownDword1",
                                            type: "uint32",
                                            defaultValue: 0
                                        },
                                        {
                                            name: "unknownDword2",
                                            type: "uint32",
                                            defaultValue: 0
                                        },
                                        {
                                            name: "unknownQword1",
                                            type: "uint64string",
                                            defaultValue: ""
                                        }
                                    ]
                                },
                                {
                                    name: "unknownDword1",
                                    type: "uint32",
                                    defaultValue: 0
                                },
                                {
                                    name: "unknownDword2",
                                    type: "uint32",
                                    defaultValue: 0
                                },
                                {
                                    name: "unknownByte1",
                                    type: "uint8",
                                    defaultValue: 0
                                }
                            ]
                        }
                    ]
                },
                {
                    name: "unknownByte1_",
                    type: "uint8",
                    defaultValue: 0
                }
            ]
        }
    ],
    ["Items.SetItemLockTimer", 0xad03, {}],
    ["Items.SetItemTimers", 0xad04, {}],
    [
        "Items.SetItemTrialLockTimer",
        0xad05,
        {
            fields: [
                { name: "unknownDword1", type: "uint32", defaultValue: 0 },
                { name: "unknownDword2", type: "uint32", defaultValue: 0 },
                { name: "unknownQword1", type: "uint64string", defaultValue: "0" }
            ]
        }
    ],
    ["Items.SetItemTrialTimers", 0xad06, {}],
    ["Items.AddItemTrialTimer", 0xad07, {}],
    ["Items.RemoveItemTrialTimer", 0xad08, {}],
    ["Items.ExpireItemTrialTimer", 0xad09, {}],
    ["Items.UpdateItemTrialTimer", 0xad0a, {}],
    ["Items.SetItemRentalTimers", 0xad0b, {}], // sub_14037CE40 somewhere
    ["Items.AddItemRentalTimer", 0xad0c, {}],
    ["Items.RemoveItemRentalTimer", 0xad0d, {}],
    ["Items.ExpireItemRentalTimer", 0xad0e, {}],
    ["Items.UseClientItem", 0xad0f, {}],
    [
        "Items.SetAccountItemManager",
        0xad10,
        {
            fields: shared_1.accountItemsSchema
        }
    ],
    [
        "Items.AddAccountItem",
        0xad11,
        {
            fields: [
                ...shared_1.accountItemSchema,
                {
                    name: "unknownByte1",
                    type: "uint8",
                    defaultValue: 192
                }
            ]
        }
    ],
    [
        "Items.RemoveAccountItem",
        0xad12,
        {
            fields: [
                {
                    name: "itemId",
                    type: "uint64string",
                    defaultValue: ""
                },
                {
                    name: "itemDefinitionId",
                    type: "uint32",
                    defaultValue: 0
                }
            ]
        }
    ],
    [
        "Items.UpdateAccountItem",
        0xad13,
        {
            fields: [
                ...shared_1.accountItemSchema,
                {
                    name: "unknownByte1",
                    type: "uint8",
                    defaultValue: 192
                }
            ]
        }
    ],
    [
        "Items.SetEscrowAccountItemManager",
        0xad14,
        {
            fields: shared_1.escrowAccountItemsSchema
        }
    ],
    [
        "Items.AddEscrowAccountItem",
        0xad15,
        {
            fields: shared_1.escrowAccountItemSchema
        }
    ],
    [
        "Items.RemoveEscrowAccountItem",
        0xad16,
        {
            fields: [
                { name: "itemId", type: "uint64string", defaultValue: "0" },
                { name: "itemDefinitionId", type: "uint32", defaultValue: 0 }
            ]
        }
    ],
    [
        "Items.UpdateEscrowAccountItem",
        0xad17,
        {
            fields: shared_1.escrowAccountItemSchema
        }
    ],
    [
        "Items.AccountItemManagerStateChanged",
        0xad18,
        {
            fields: [
                { name: "escrowEnabled", type: "uint32", defaultValue: 0 },
                { name: "escrowServerConnected", type: "uint32", defaultValue: 0 },
                { name: "escrowAccountLoadSucceeded", type: "uint32", defaultValue: 0 },
                { name: "escrowAccountAllowTrading", type: "uint32", defaultValue: 0 }
            ]
        }
    ],
    ["Items.AddNewAccountItemRec", 0xad19, {}],
    ["Items.RemoveNewAccountItemRec", 0xad1a, {}],
    ["Items.RemoveNewAccountItemRecByItemId", 0xad1b, {}],
    ["Items.RemoveAllNewAccountItemRecs", 0xad1c, {}],
    [
        "Items.ReportNewRewardCrateAdded",
        0xad1d,
        {
            fields: shared_1.accountItemSchema
        }
    ],
    [
        "Items.ReportRewardCrateContents",
        0xad1e,
        {
            fields: [
                {
                    name: "winningRewards",
                    type: "array",
                    defaultValue: [],
                    fields: [
                        { name: "itemDefinitionId", type: "uint32", defaultValue: 0 },
                        { name: "itemCount", type: "uint32", defaultValue: 1 }
                    ]
                },
                {
                    name: "possibleRewards",
                    type: "array",
                    defaultValue: [],
                    fields: [
                        { name: "itemDefinitionId", type: "uint32", defaultValue: 0 },
                        { name: "itemCount", type: "uint32", defaultValue: 1 }
                    ]
                }
            ]
        }
    ],
    [
        "Items.SetEmoteItem",
        0xad1f,
        {
            fields: [
                { name: "unknownDword1", type: "uint32", defaultValue: 1 },
                { name: "unknownDword2", type: "uint32", defaultValue: 3154 }
            ]
        }
    ],
    [
        "Items.RemoveEmoteItem",
        0xad20,
        {
            fields: [{ name: "itemDefinitionId", type: "uint32", defaultValue: 0 }]
        }
    ],
    [
        "Items.SetSkinItemManager",
        0xad21,
        {
            fields: [
                { name: "unknownDword1", type: "uint32", defaultValue: 0 },
                { name: "unknownDword2", type: "uint32", defaultValue: 0 },
                { name: "unknownString1", type: "string", defaultValue: "" },
                ...shared_1.skinItemsSchema,
                ...shared_1.emoteItemsSchema,
                ...shared_1.itemCollectionSchema
            ]
        }
    ], // Related to unknownData8 in sendself
    [
        "Items.SetSkinItem",
        0xad22,
        {
            fields: [
                { name: "unknownDword1", type: "uint32", defaultValue: 1 },
                ...shared_1.skinItemSchema
            ]
        }
    ],
    ["Items.RemoveSkinItem", 0xad23, {}],
    ["Items.SetSkinItemCollectionCustomName", 0xad24, {}],
    ["Items.SelectSkinItemCollectionId", 0xad25, {}],
    [
        "Items.SetCurrentSkinItemCollection",
        0xad26,
        {
            fields: [
                { name: "unknownDword1", type: "uint32", defaultValue: 0 },
                { name: "unknownString1", type: "string", defaultValue: "" },
                ...shared_1.skinItemsSchema,
                ...shared_1.emoteItemsSchema
            ]
        }
    ], // Related to unknownData8 in sendself
    ["Items.RequestAddItemTimer", 0xad27, {}],
    ["Items.RequestTrialItem", 0xad28, {}],
    ["Items.RequestRentalItem", 0xad29, {}],
    [
        "Items.RequestUseItem",
        0xad2a,
        {
            fields: [
                { name: "itemCount", type: "uint32", defaultValue: 0 },
                { name: "unknownDword1", type: "uint32", defaultValue: 0 },
                { name: "itemUseOption", type: "uint32", defaultValue: 0 },
                { name: "characterId", type: "uint64string", defaultValue: "" },
                { name: "targetCharacterId", type: "uint64string", defaultValue: "" },
                { name: "sourceCharacterId", type: "uint64string", defaultValue: "" },
                { name: "itemGuid", type: "uint64string", defaultValue: "" },
                {
                    name: "itemSubData",
                    type: "custom",
                    defaultValue: {},
                    parser: parseItemRequestSubData
                }
            ]
        }
    ],
    [
        "Items.RequestUseAccountItem",
        0xad2b,
        {
            fields: [
                { name: "itemCount", type: "uint32", defaultValue: 0 },
                { name: "unknownDword2", type: "uint32", defaultValue: 0 },
                { name: "unknownDword3", type: "uint32", defaultValue: 0 },
                { name: "itemDefinitionId", type: "uint32", defaultValue: 0 },
                {
                    name: "itemSubData",
                    type: "custom",
                    defaultValue: {},
                    parser: parseAccountItemRequestSubData
                }
            ]
        }
    ],
    ["Items.RequestRemoveNewAccountItemRec", 0xad2c, {}],
    ["Items.RemoveNewAccountItemRecByItemId", 0xad2d, {}],
    ["Items.RequestRemoveAllNewAccountItemRecs", 0xad2e, {}],
    ["Items.RequestSetEmoteItem", 0xad2f, {}],
    ["Items.RequestSetSkinItem", 0xad30, {}],
    ["Items.SetSkinItemByItemId", 0xad31, {}],
    ["Items.RequestSetSkinItemFlags", 0xad32, {}],
    ["Items.RequestUnsetSkinItem", 0xad33, {}],
    ["Items.RequestRemoveSkinItem", 0xad34, {}],
    ["Items.SetSkinItemCollectionCustomName", 0xad35, {}],
    ["Items.RequestSelectSkinItemCollection", 0xad36, {}],
    ["Items.RequestOpenAccountCrate", 0xad37, {}],
    ["Items.RequestPreviewAccountCrateRewards", 0xad38, {}]
];
