// h1z1-server/out/servers/ZoneServer2016/models/enums.js
"use strict";
/* eslint-disable @typescript-eslint/no-duplicate-enum-values */
// ======================================================================
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReplicationPropertyHash = exports.UIElements = exports.GameModes = exports.DefaultSkinsBoots = exports.DefaultSkinsGators = exports.DefaultSkinsZeds = exports.DefaultSkinsMotorHelmet = exports.FramedBackpack = exports.Backpack1000 = exports.MilitaryTan = exports.DefaultSkinsBackpack = exports.DefaultSkinsConveys = exports.MeleeTypes = exports.MaterialTypes = exports.VehicleEffects = exports.AbilityIds = exports.ItemTypes = exports.HealTypes = exports.Effects = exports.WeaponDefinitionIds = exports.Items = exports.AccountItems = exports.Recipes = exports.TreeIds = exports.Abilities = exports.ResourceIndicators = exports.ResourceTypes = exports.ResourceIds = exports.EquipSlots = exports.LoadoutSlots = exports.LoadoutIds = exports.Characters = exports.NpcIds = exports.VehicleIds = exports.ItemClasses = exports.FilterIds = exports.ContainerErrors = exports.MovementModifiers = exports.StringIds = exports.ModelIds = exports.ConstructionPermissionIds = exports.ConstructionErrors = exports.ItemUseOptions = exports.PositionUpdateType = void 0;
var PositionUpdateType;
(function (PositionUpdateType) {
    PositionUpdateType[PositionUpdateType["STATIC"] = 0] = "STATIC";
    PositionUpdateType[PositionUpdateType["MOVABLE"] = 1] = "MOVABLE";
    // there are others we can get by looking at the vehicle packets
})(PositionUpdateType || (exports.PositionUpdateType = PositionUpdateType = {}));
var ItemUseOptions;
(function (ItemUseOptions) {
    ItemUseOptions[ItemUseOptions["EAT"] = 1] = "EAT";
    ItemUseOptions[ItemUseOptions["DRINK"] = 2] = "DRINK";
    ItemUseOptions[ItemUseOptions["USE"] = 3] = "USE";
    ItemUseOptions[ItemUseOptions["DROP"] = 4] = "DROP";
    ItemUseOptions[ItemUseOptions["SHRED"] = 6] = "SHRED";
    ItemUseOptions[ItemUseOptions["UNLOAD"] = 7] = "UNLOAD";
    ItemUseOptions[ItemUseOptions["UNPACK"] = 10] = "UNPACK";
    ItemUseOptions[ItemUseOptions["IGNITE"] = 11] = "IGNITE";
    ItemUseOptions[ItemUseOptions["REPAIR"] = 16] = "REPAIR";
    ItemUseOptions[ItemUseOptions["REFUEL"] = 17] = "REFUEL";
    ItemUseOptions[ItemUseOptions["SLICE"] = 18] = "SLICE";
    ItemUseOptions[ItemUseOptions["UNPACK_BUNDLE"] = 20] = "UNPACK_BUNDLE";
    ItemUseOptions[ItemUseOptions["APPLY_SKIN"] = 24] = "APPLY_SKIN";
    ItemUseOptions[ItemUseOptions["CALL_AIRDROP"] = 25] = "CALL_AIRDROP";
    ItemUseOptions[ItemUseOptions["OPEN_BAG"] = 26] = "OPEN_BAG";
    ItemUseOptions[ItemUseOptions["OPEN_PACKAGE"] = 46] = "OPEN_PACKAGE";
    ItemUseOptions[ItemUseOptions["USE_MEDICAL"] = 52] = "USE_MEDICAL";
    ItemUseOptions[ItemUseOptions["USE_AIRDROP"] = 58] = "USE_AIRDROP";
    ItemUseOptions[ItemUseOptions["LOOT"] = 59] = "LOOT";
    ItemUseOptions[ItemUseOptions["EQUIP"] = 60] = "EQUIP";
    ItemUseOptions[ItemUseOptions["MOVE"] = 61] = "MOVE";
    ItemUseOptions[ItemUseOptions["HOTWIRE_ATV"] = 62] = "HOTWIRE_ATV";
    ItemUseOptions[ItemUseOptions["LOOT_VEHICLE_LOADOUT"] = 66] = "LOOT_VEHICLE_LOADOUT";
    ItemUseOptions[ItemUseOptions["MOVE_VEHICLE_PARTS"] = 67] = "MOVE_VEHICLE_PARTS";
    ItemUseOptions[ItemUseOptions["HOTWIRE_OFFROADER"] = 70] = "HOTWIRE_OFFROADER";
    ItemUseOptions[ItemUseOptions["HOTWIRE_PICKUP"] = 71] = "HOTWIRE_PICKUP";
    ItemUseOptions[ItemUseOptions["HOTWIRE_POLICE"] = 72] = "HOTWIRE_POLICE";
    ItemUseOptions[ItemUseOptions["DROP_BATTERY"] = 73] = "DROP_BATTERY";
    ItemUseOptions[ItemUseOptions["LOOT_BATTERY"] = 75] = "LOOT_BATTERY";
    ItemUseOptions[ItemUseOptions["MOVE_BATTERY"] = 76] = "MOVE_BATTERY";
    ItemUseOptions[ItemUseOptions["DROP_SPARKS"] = 79] = "DROP_SPARKS";
    ItemUseOptions[ItemUseOptions["LOOT_SPARKS"] = 81] = "LOOT_SPARKS";
    ItemUseOptions[ItemUseOptions["MOVE_SPARKS"] = 82] = "MOVE_SPARKS";
    ItemUseOptions[ItemUseOptions["OPEN_CRATE"] = 85] = "OPEN_CRATE";
    ItemUseOptions[ItemUseOptions["SALVAGE"] = 87] = "SALVAGE";
    ItemUseOptions[ItemUseOptions["HOTWIRE_ATV_NO_PARTS"] = 90] = "HOTWIRE_ATV_NO_PARTS";
    ItemUseOptions[ItemUseOptions["HOTWIRE_OFFROADER_NO_PARTS"] = 91] = "HOTWIRE_OFFROADER_NO_PARTS";
    ItemUseOptions[ItemUseOptions["HOTWIRE_PICKUP_NO_PARTS"] = 92] = "HOTWIRE_PICKUP_NO_PARTS";
    ItemUseOptions[ItemUseOptions["HOTWIRE_POLICE_NO_PARTS"] = 93] = "HOTWIRE_POLICE_NO_PARTS";
})(ItemUseOptions || (exports.ItemUseOptions = ItemUseOptions = {}));
var ConstructionErrors;
(function (ConstructionErrors) {
    ConstructionErrors["UNKNOWN"] = "Unknown";
    ConstructionErrors["OVERLAP"] = "Construction overlap";
    ConstructionErrors["BUILD_PERMISSION"] = "No build permission";
    ConstructionErrors["DEMOLISH_PERMISSION"] = "No demolish permission";
    ConstructionErrors["UNKNOWN_PARENT"] = "Unknown parent";
    ConstructionErrors["UNKNOWN_SLOT"] = "Unknown slot";
    ConstructionErrors["UNKNOWN_CONSTRUCTION"] = "Unknown construction item";
    ConstructionErrors["OUT_OF_RANGE"] = "Out of range";
    ConstructionErrors["STACKED"] = "Construction stacked";
    ConstructionErrors["WALL_SLOT_FAILED"] = "Wall slot failed plz report this!";
    ConstructionErrors["TREESTACKED"] = "Construction tree stacked";
})(ConstructionErrors || (exports.ConstructionErrors = ConstructionErrors = {}));
var ConstructionPermissionIds;
(function (ConstructionPermissionIds) {
    ConstructionPermissionIds[ConstructionPermissionIds["BUILD"] = 1] = "BUILD";
    ConstructionPermissionIds[ConstructionPermissionIds["DEMOLISH"] = 2] = "DEMOLISH";
    ConstructionPermissionIds[ConstructionPermissionIds["CONTAINERS"] = 3] = "CONTAINERS";
    ConstructionPermissionIds[ConstructionPermissionIds["VISIT"] = 4] = "VISIT";
})(ConstructionPermissionIds || (exports.ConstructionPermissionIds = ConstructionPermissionIds = {}));
var ModelIds;
(function (ModelIds) {
    ModelIds[ModelIds["WEAPON_EMPTY"] = 9084] = "WEAPON_EMPTY";
    ModelIds[ModelIds["LOOT_BAG_BLOODY"] = 9581] = "LOOT_BAG_BLOODY";
    ModelIds[ModelIds["LOOT_BAG_CLEAN"] = 9391] = "LOOT_BAG_CLEAN";
    ModelIds[ModelIds["MILITARY_CRATE"] = 9218] = "MILITARY_CRATE";
    ModelIds[ModelIds["MILITARY_CRATE_PARACHUTE"] = 9219] = "MILITARY_CRATE_PARACHUTE";
    ModelIds[ModelIds["HOSPITAL_LAB_WORKBENCH"] = 9563] = "HOSPITAL_LAB_WORKBENCH";
    ModelIds[ModelIds["HOSPITAL_DESK"] = 9551] = "HOSPITAL_DESK";
    ModelIds[ModelIds["HOSPITAL_LAB_COUNTERTOP"] = 9553] = "HOSPITAL_LAB_COUNTERTOP";
    ModelIds[ModelIds["HOSPITAL_GROSSING_STATION"] = 9556] = "HOSPITAL_GROSSING_STATION";
    ModelIds[ModelIds["HOSPITAL_REFRIGERATOR"] = 9555] = "HOSPITAL_REFRIGERATOR";
    ModelIds[ModelIds["HOSPITAL_SINK_COUNTERTOP"] = 9554] = "HOSPITAL_SINK_COUNTERTOP";
    ModelIds[ModelIds["HOSPITAL_DRUG_CABNINET"] = 9552] = "HOSPITAL_DRUG_CABNINET";
    ModelIds[ModelIds["TREASURE_CHEST"] = 9347] = "TREASURE_CHEST";
    ModelIds[ModelIds["METAL_STORAGE_CHEST"] = 57] = "METAL_STORAGE_CHEST";
    ModelIds[ModelIds["CRATE_LONG"] = 8014] = "CRATE_LONG";
    ModelIds[ModelIds["CRATE_BOX_1"] = 8013] = "CRATE_BOX_1";
    ModelIds[ModelIds["CRATE_BOX_2"] = 9088] = "CRATE_BOX_2";
    ModelIds[ModelIds["BARELL_1"] = 9387] = "BARELL_1";
    ModelIds[ModelIds["GLASS_WINDOW_01"] = 8026] = "GLASS_WINDOW_01";
    ModelIds[ModelIds["TINTED_WINDOW_01"] = 9432] = "TINTED_WINDOW_01";
    ModelIds[ModelIds["GLASS_WINDOW_DESTROYED_1"] = 8027] = "GLASS_WINDOW_DESTROYED_1";
    ModelIds[ModelIds["GLASS_WINDOW_DESTROYED_2"] = 8028] = "GLASS_WINDOW_DESTROYED_2";
    ModelIds[ModelIds["GLASS_WINDOW_DESTROYED_3"] = 8029] = "GLASS_WINDOW_DESTROYED_3";
    ModelIds[ModelIds["TINTED_WINDOW_DESTROYED_1"] = 9433] = "TINTED_WINDOW_DESTROYED_1";
    ModelIds[ModelIds["TINTED_WINDOW_DESTROYED_2"] = 9434] = "TINTED_WINDOW_DESTROYED_2";
    ModelIds[ModelIds["TINTED_WINDOW_DESTROYED_3"] = 9435] = "TINTED_WINDOW_DESTROYED_3";
    ModelIds[ModelIds["GARBAGE_CAN_01"] = 9007] = "GARBAGE_CAN_01";
    ModelIds[ModelIds["ARMOIRE_01"] = 9006] = "ARMOIRE_01";
    ModelIds[ModelIds["ARMOIRE_02"] = 9124] = "ARMOIRE_02";
    ModelIds[ModelIds["DRESSER_01"] = 9005] = "DRESSER_01";
    ModelIds[ModelIds["DRESSER_02"] = 9125] = "DRESSER_02";
    ModelIds[ModelIds["OTTOMAN_01"] = 9149] = "OTTOMAN_01";
    ModelIds[ModelIds["OTTOMAN_02"] = 9150] = "OTTOMAN_02";
    ModelIds[ModelIds["TOOL_CABINET_01"] = 9074] = "TOOL_CABINET_01";
    ModelIds[ModelIds["TOOL_CABINET_02"] = 9035] = "TOOL_CABINET_02";
    ModelIds[ModelIds["BLUE_TOOL_CABINET"] = 9500] = "BLUE_TOOL_CABINET";
    ModelIds[ModelIds["RED_SILVER_TOOL_CABINET"] = 9501] = "RED_SILVER_TOOL_CABINET";
    ModelIds[ModelIds["CABINETS_BATHROOM_01"] = 9031] = "CABINETS_BATHROOM_01";
    ModelIds[ModelIds["CABINETS_BATHROOM_02"] = 9237] = "CABINETS_BATHROOM_02";
    ModelIds[ModelIds["CABINETS_BATHROOM_SINK"] = 9032] = "CABINETS_BATHROOM_SINK";
    ModelIds[ModelIds["CABINET_SET_01"] = 9352] = "CABINET_SET_01";
    ModelIds[ModelIds["CABINET_SET_02"] = 9353] = "CABINET_SET_02";
    ModelIds[ModelIds["CABINET_SET_03"] = 9362] = "CABINET_SET_03";
    ModelIds[ModelIds["CABINET_SET_04"] = 9363] = "CABINET_SET_04";
    ModelIds[ModelIds["CABINET_SET_05"] = 9364] = "CABINET_SET_05";
    ModelIds[ModelIds["CABINET_SET_06"] = 9365] = "CABINET_SET_06";
    ModelIds[ModelIds["CABINET_SET_07"] = 9366] = "CABINET_SET_07";
    ModelIds[ModelIds["CABINET_SET_08"] = 9367] = "CABINET_SET_08";
    ModelIds[ModelIds["CABINET_SET_09"] = 9368] = "CABINET_SET_09";
    ModelIds[ModelIds["CABINET_SET_10"] = 9369] = "CABINET_SET_10";
    ModelIds[ModelIds["CABINET_SET_11"] = 9370] = "CABINET_SET_11";
    ModelIds[ModelIds["CABINETS_KITCHEN_01"] = 9011] = "CABINETS_KITCHEN_01";
    ModelIds[ModelIds["CABINETS_KITCHEN_02"] = 9012] = "CABINETS_KITCHEN_02";
    ModelIds[ModelIds["CABINETS_KITCHEN_03"] = 9026] = "CABINETS_KITCHEN_03";
    ModelIds[ModelIds["CABINETS_KITCHEN_04"] = 9027] = "CABINETS_KITCHEN_04";
    ModelIds[ModelIds["CABINETS_KITCHEN_05"] = 9028] = "CABINETS_KITCHEN_05";
    ModelIds[ModelIds["CABINETS_KITCHEN_06"] = 9239] = "CABINETS_KITCHEN_06";
    ModelIds[ModelIds["OFFICE_CUBE_CABINET"] = 9229] = "OFFICE_CUBE_CABINET";
    ModelIds[ModelIds["OFFICE_CABINETS_01"] = 9226] = "OFFICE_CABINETS_01";
    ModelIds[ModelIds["OFFICE_CABINETS_02"] = 9227] = "OFFICE_CABINETS_02";
    ModelIds[ModelIds["OFFICE_CABINETS_03"] = 9228] = "OFFICE_CABINETS_03";
    ModelIds[ModelIds["PLYWOOD_METAL_GATE"] = 49] = "PLYWOOD_METAL_GATE";
    ModelIds[ModelIds["RESIDENTIAL_DOOR_1"] = 9009] = "RESIDENTIAL_DOOR_1";
    ModelIds[ModelIds["RESIDENTIAL_DOOR_2"] = 9165] = "RESIDENTIAL_DOOR_2";
    ModelIds[ModelIds["RESIDENTIAL_DOOR_3"] = 9167] = "RESIDENTIAL_DOOR_3";
    ModelIds[ModelIds["RESIDENTIAL_DOOR_4"] = 9169] = "RESIDENTIAL_DOOR_4";
    ModelIds[ModelIds["RESIDENTIAL_DOOR_5"] = 9171] = "RESIDENTIAL_DOOR_5";
    ModelIds[ModelIds["RESIDENTIAL_FRONT_DOOR_1"] = 9497] = "RESIDENTIAL_FRONT_DOOR_1";
    ModelIds[ModelIds["RESIDENTIAL_FRONT_DOOR_2"] = 9904] = "RESIDENTIAL_FRONT_DOOR_2";
    ModelIds[ModelIds["RESIDENTIAL_INTERIOR_DOOR"] = 9905] = "RESIDENTIAL_INTERIOR_DOOR";
    ModelIds[ModelIds["CAMPER_DOOR_1"] = 9333] = "CAMPER_DOOR_1";
    ModelIds[ModelIds["CAMPER_DOOR_GLASS_BROKEN_1"] = 9334] = "CAMPER_DOOR_GLASS_BROKEN_1";
    ModelIds[ModelIds["CAMPER_DOOR_GLASS_BROKEN_2"] = 9335] = "CAMPER_DOOR_GLASS_BROKEN_2";
    ModelIds[ModelIds["CHURCH_DOOR"] = 9136] = "CHURCH_DOOR";
    ModelIds[ModelIds["OFFICE_DOOR"] = 9224] = "OFFICE_DOOR";
    ModelIds[ModelIds["MENS_LEFT_DOOR"] = 9232] = "MENS_LEFT_DOOR";
    ModelIds[ModelIds["WOMENS_LEFT_DOOR"] = 9233] = "WOMENS_LEFT_DOOR";
    ModelIds[ModelIds["FREEZER_DOOR"] = 9243] = "FREEZER_DOOR";
    ModelIds[ModelIds["INDUSTRIAL_DOOR_1"] = 9903] = "INDUSTRIAL_DOOR_1";
    ModelIds[ModelIds["INDUSTRIAL_DOOR_2"] = 9246] = "INDUSTRIAL_DOOR_2";
    ModelIds[ModelIds["PUSH_DOOR"] = 9498] = "PUSH_DOOR";
    ModelIds[ModelIds["RESTAURANT_DOOR_GLASS_BROKEN_1"] = 9452] = "RESTAURANT_DOOR_GLASS_BROKEN_1";
    ModelIds[ModelIds["RESTAURANT_DOOR_GLASS_BROKEN_2"] = 9453] = "RESTAURANT_DOOR_GLASS_BROKEN_2";
    ModelIds[ModelIds["RESTAURANT_DOOR_GLASS_BROKEN_3"] = 9454] = "RESTAURANT_DOOR_GLASS_BROKEN_3";
    ModelIds[ModelIds["RESTAURANT_DOOR_1"] = 9455] = "RESTAURANT_DOOR_1";
    ModelIds[ModelIds["BUISNESS_DOOR_GLASS"] = 9183] = "BUISNESS_DOOR_GLASS";
    ModelIds[ModelIds["BUISNESS_DOOR_GLASS_BROKEN_1"] = 9184] = "BUISNESS_DOOR_GLASS_BROKEN_1";
    ModelIds[ModelIds["BUISNESS_DOOR_GLASS_BROKEN_2"] = 9185] = "BUISNESS_DOOR_GLASS_BROKEN_2";
    ModelIds[ModelIds["BUISNESS_DOOR_GLASS_BROKEN_3"] = 9186] = "BUISNESS_DOOR_GLASS_BROKEN_3";
    ModelIds[ModelIds["CABIN_DOOR"] = 9901] = "CABIN_DOOR";
    ModelIds[ModelIds["CABIN_DOOR_BROKEN"] = 9902] = "CABIN_DOOR_BROKEN";
    ModelIds[ModelIds["BUISNESS_DOORS_GLASS"] = 9897] = "BUISNESS_DOORS_GLASS";
    ModelIds[ModelIds["BUISNESS_DOORS_GLASS_BROKEN_1"] = 9898] = "BUISNESS_DOORS_GLASS_BROKEN_1";
    ModelIds[ModelIds["BUISNESS_DOORS_GLASS_BROKEN_2"] = 9899] = "BUISNESS_DOORS_GLASS_BROKEN_2";
    ModelIds[ModelIds["BUISNESS_DOORS_GLASS_BROKEN_3"] = 9900] = "BUISNESS_DOORS_GLASS_BROKEN_3";
    ModelIds[ModelIds["DESK"] = 9038] = "DESK";
    ModelIds[ModelIds["LOCKERS_LOCKER_POLICE"] = 9254] = "LOCKERS_LOCKER_POLICE";
    ModelIds[ModelIds["LOCKERS_LOCKER_WEAPONS"] = 9255] = "LOCKERS_LOCKER_WEAPONS";
    ModelIds[ModelIds["OFFROADER"] = 10060] = "OFFROADER";
    ModelIds[ModelIds["PICKUP_TRUCK"] = 9258] = "PICKUP_TRUCK";
    ModelIds[ModelIds["POLICE_CAR"] = 9301] = "POLICE_CAR";
    ModelIds[ModelIds["ATV"] = 9588] = "ATV";
    ModelIds[ModelIds["PARACHUTE"] = 9374] = "PARACHUTE";
    ModelIds[ModelIds["SPECTATE"] = 9371] = "SPECTATE";
    ModelIds[ModelIds["PICKUP_DESTROYED"] = 9315] = "PICKUP_DESTROYED";
    ModelIds[ModelIds["POLICE_CAR_DESTROYED"] = 9316] = "POLICE_CAR_DESTROYED";
    ModelIds[ModelIds["ATV_DESTROYED"] = 9593] = "ATV_DESTROYED";
    ModelIds[ModelIds["OFFROADER_DESTROYED"] = 7226] = "OFFROADER_DESTROYED";
    ModelIds[ModelIds["WRECKED_TRUCK_01"] = 9108] = "WRECKED_TRUCK_01";
    ModelIds[ModelIds["WRECKED_TRUCK_OD_01"] = 9178] = "WRECKED_TRUCK_OD_01";
    ModelIds[ModelIds["WRECKED_CAR_OD_01"] = 9179] = "WRECKED_CAR_OD_01";
    ModelIds[ModelIds["WRECKED_CAR"] = 9105] = "WRECKED_CAR";
    ModelIds[ModelIds["WRECKED_VAN"] = 9110] = "WRECKED_VAN";
    ModelIds[ModelIds["AIRDROP_CARGO_CONTAINER"] = 9384] = "AIRDROP_CARGO_CONTAINER";
    ModelIds[ModelIds["AIRDROP_PLANE"] = 9215] = "AIRDROP_PLANE";
    ModelIds[ModelIds["OFFICE_FILING_CABINET"] = 9225] = "OFFICE_FILING_CABINET";
    ModelIds[ModelIds["LAB_GLASS_CABINET"] = 9256] = "LAB_GLASS_CABINET";
    ModelIds[ModelIds["KITCHEN_FRIDGE"] = 9030] = "KITCHEN_FRIDGE";
    ModelIds[ModelIds["FILE_CABINETS"] = 9034] = "FILE_CABINETS";
    ModelIds[ModelIds["DUMPSTER"] = 9037] = "DUMPSTER";
    ModelIds[ModelIds["WASHER"] = 9396] = "WASHER";
    ModelIds[ModelIds["DRYER"] = 9395] = "DRYER";
    ModelIds[ModelIds["CAMPFIRE"] = 9041] = "CAMPFIRE";
    ModelIds[ModelIds["BARBEQUE"] = 9205] = "BARBEQUE";
    ModelIds[ModelIds["FURNACE"] = 36] = "FURNACE";
    ModelIds[ModelIds["HAND_SHOVEL"] = 9248] = "HAND_SHOVEL";
    ModelIds[ModelIds["ZOMBIE_SCREAMER"] = 9667] = "ZOMBIE_SCREAMER";
    ModelIds[ModelIds["ZOMBIE_FEMALE_WALKER"] = 9510] = "ZOMBIE_FEMALE_WALKER";
    ModelIds[ModelIds["ZOMBIE_MALE_WALKER"] = 9634] = "ZOMBIE_MALE_WALKER";
    ModelIds[ModelIds["DEER_BUCK"] = 9253] = "DEER_BUCK";
    ModelIds[ModelIds["DEER"] = 9002] = "DEER";
    ModelIds[ModelIds["WOLF"] = 9003] = "WOLF";
    ModelIds[ModelIds["BEAR"] = 9187] = "BEAR";
    ModelIds[ModelIds["RAVEN"] = 9230] = "RAVEN";
    ModelIds[ModelIds["CORN_CROPSTATE_1"] = 59] = "CORN_CROPSTATE_1";
    ModelIds[ModelIds["CORN_CROPSTATE_2"] = 60] = "CORN_CROPSTATE_2";
    ModelIds[ModelIds["CORN_CROPSTATE_3"] = 61] = "CORN_CROPSTATE_3";
    ModelIds[ModelIds["WHEAT_CROPSTATE_1"] = 9191] = "WHEAT_CROPSTATE_1";
    ModelIds[ModelIds["WEHAT_CROPSTATE_2"] = 9190] = "WEHAT_CROPSTATE_2";
    ModelIds[ModelIds["WHEAT_CROPSTATE_3"] = 9189] = "WHEAT_CROPSTATE_3";
    ModelIds[ModelIds["BARBED_WIRE"] = 28] = "BARBED_WIRE";
    ModelIds[ModelIds["BARBED_WIRE_POSTS"] = 43] = "BARBED_WIRE_POSTS";
    ModelIds[ModelIds["SNARE"] = 1974] = "SNARE";
    ModelIds[ModelIds["BATTERY"] = 70] = "BATTERY";
    ModelIds[ModelIds["FIRE_HYDRANT"] = 9337] = "FIRE_HYDRANT";
    ModelIds[ModelIds["SURVIVOR_MALE_HEAD_01"] = 9469] = "SURVIVOR_MALE_HEAD_01";
    ModelIds[ModelIds["SURVIVAL_FEMALE_HEAD_01"] = 9474] = "SURVIVAL_FEMALE_HEAD_01";
    ModelIds[ModelIds["ZOMBIE_FEMALE_HEAD"] = 9510] = "ZOMBIE_FEMALE_HEAD";
    ModelIds[ModelIds["ZOMBIE_MALE_HEAD"] = 9634] = "ZOMBIE_MALE_HEAD";
    ModelIds[ModelIds["GROUND_TAMPER"] = 9115] = "GROUND_TAMPER";
    ModelIds[ModelIds["DECK_EXPANSION"] = 9492] = "DECK_EXPANSION";
    ModelIds[ModelIds["METAL_SHACK_DOOR"] = 9181] = "METAL_SHACK_DOOR";
    ModelIds[ModelIds["METAL_SHACK"] = 9180] = "METAL_SHACK";
    ModelIds[ModelIds["SMALL_SHACK"] = 9192] = "SMALL_SHACK";
    ModelIds[ModelIds["DEW_COLLECTOR"] = 55] = "DEW_COLLECTOR";
    ModelIds[ModelIds["WOOD_SHACK"] = 9223] = "WOOD_SHACK";
    ModelIds[ModelIds["WOOD_SHACK_DOOR"] = 63] = "WOOD_SHACK_DOOR";
    ModelIds[ModelIds["RAMP"] = 9487] = "RAMP";
    ModelIds[ModelIds["FOUNDATION_STAIRS"] = 9488] = "FOUNDATION_STAIRS";
    ModelIds[ModelIds["METAL_GATE"] = 49] = "METAL_GATE";
    ModelIds[ModelIds["METAL_WALL"] = 50] = "METAL_WALL";
    ModelIds[ModelIds["UPPER_METAL_WALL"] = 9407] = "UPPER_METAL_WALL";
    ModelIds[ModelIds["SHELTER"] = 51] = "SHELTER";
    ModelIds[ModelIds["LARGE_SHELTER"] = 52] = "LARGE_SHELTER";
    ModelIds[ModelIds["UPPER_LEVEL_SHELTER"] = 9408] = "UPPER_LEVEL_SHELTER";
    ModelIds[ModelIds["UPPER_LEVEL_LARGE_SHELTER"] = 9411] = "UPPER_LEVEL_LARGE_SHELTER";
    ModelIds[ModelIds["STRUCTURE_STAIRS"] = 53] = "STRUCTURE_STAIRS";
    ModelIds[ModelIds["TOWER"] = 9493] = "TOWER";
    ModelIds[ModelIds["DECK_FOUNDATION"] = 9130] = "DECK_FOUNDATION";
    ModelIds[ModelIds["FENCES_WOOD_PLANKS_GREY_POSTS_1X2"] = 8037] = "FENCES_WOOD_PLANKS_GREY_POSTS_1X2";
    ModelIds[ModelIds["FENCES_WOOD_PLANKS_GREY_POSTS_1X1"] = 8036] = "FENCES_WOOD_PLANKS_GREY_POSTS_1X1";
    ModelIds[ModelIds["FENCES_WOOD_PLANKS_GREY_1X1"] = 8033] = "FENCES_WOOD_PLANKS_GREY_1X1";
    ModelIds[ModelIds["FENCES_WOOD_PLANKS_GREY_PLANK"] = 8035] = "FENCES_WOOD_PLANKS_GREY_PLANK";
    ModelIds[ModelIds["FENCES_WOOD_PLANKS_GREY_GAP_1X1"] = 8034] = "FENCES_WOOD_PLANKS_GREY_GAP_1X1";
    ModelIds[ModelIds["BARBED_WIRE_FENCE_1X2"] = 9114] = "BARBED_WIRE_FENCE_1X2";
    ModelIds[ModelIds["BARBED_WIRE_FENCE_1X1"] = 9113] = "BARBED_WIRE_FENCE_1X1";
    ModelIds[ModelIds["FENCE_RANCH_1X2"] = 9403] = "FENCE_RANCH_1X2";
    ModelIds[ModelIds["FENCE_RANCH_1X1"] = 9402] = "FENCE_RANCH_1X1";
    ModelIds[ModelIds["CHAIN_LINK_FENCE_1X1"] = 8024] = "CHAIN_LINK_FENCE_1X1";
    ModelIds[ModelIds["CHAIN_LINK_FENCE_1X2"] = 8025] = "CHAIN_LINK_FENCE_1X2";
    ModelIds[ModelIds["SIGNS_STOP_SIGN_01"] = 8021] = "SIGNS_STOP_SIGN_01";
    ModelIds[ModelIds["SIGNS_STOP_SIGN_01_DESTROYED"] = 8022] = "SIGNS_STOP_SIGN_01_DESTROYED";
    ModelIds[ModelIds["TRAFFIC_SIGNS_STOP_SIGN"] = 9927] = "TRAFFIC_SIGNS_STOP_SIGN";
    ModelIds[ModelIds["TRAFFIC_SIGN_POLE"] = 9942] = "TRAFFIC_SIGN_POLE";
    ModelIds[ModelIds["TRAFFIC_SIGN_POLE_DOUBLE"] = 9947] = "TRAFFIC_SIGN_POLE_DOUBLE";
    ModelIds[ModelIds["TRAFFIC_SIGNS_HIGHWAY_LARGE_SIGNS_ROAD_TWO_POST_01"] = 9948] = "TRAFFIC_SIGNS_HIGHWAY_LARGE_SIGNS_ROAD_TWO_POST_01";
    ModelIds[ModelIds["TRAFFIC_SIGN_HIGHWAY_WOOD_BASE"] = 9981] = "TRAFFIC_SIGN_HIGHWAY_WOOD_BASE";
    ModelIds[ModelIds["TRAFFIC_SIGNS_HIGHWAY_LARGE_SIGNS_WOOD_TWO_POST_01"] = 9982] = "TRAFFIC_SIGNS_HIGHWAY_LARGE_SIGNS_WOOD_TWO_POST_01";
    ModelIds[ModelIds["TRAFFIC_SIGNS_HIGHWAY"] = 9992] = "TRAFFIC_SIGNS_HIGHWAY";
    ModelIds[ModelIds["TRAFFIC_SIGNS_TOWN"] = 9918] = "TRAFFIC_SIGNS_TOWN";
    ModelIds[ModelIds["STREET_SIGNS_STREET_NAMES_SIGN_POST"] = 9950] = "STREET_SIGNS_STREET_NAMES_SIGN_POST";
    ModelIds[ModelIds["STREET_SIGNS_STREET_NAMES_SIGN"] = 9943] = "STREET_SIGNS_STREET_NAMES_SIGN";
    ModelIds[ModelIds["STREET_SIGN_POLE"] = 9944] = "STREET_SIGN_POLE";
})(ModelIds || (exports.ModelIds = ModelIds = {}));
var StringIds;
(function (StringIds) {
    StringIds[StringIds["ENTER_VEHICLE"] = 15] = "ENTER_VEHICLE";
    StringIds[StringIds["OFFROADER"] = 16] = "OFFROADER";
    StringIds[StringIds["WATER_WELL"] = 1006] = "WATER_WELL";
    StringIds[StringIds["COLLECT_WATER"] = 1008] = "COLLECT_WATER";
    StringIds[StringIds["DRINK_DIRTY_WATER"] = 1331] = "DRINK_DIRTY_WATER";
    StringIds[StringIds["DRINK_STAGNANT_WATER"] = 1461] = "DRINK_STAGNANT_WATER";
    StringIds[StringIds["DIRTY_WATER"] = 1099] = "DIRTY_WATER";
    StringIds[StringIds["TAKE_ITEM"] = 29] = "TAKE_ITEM";
    StringIds[StringIds["OPEN"] = 31] = "OPEN";
    StringIds[StringIds["CORN"] = 628] = "CORN";
    StringIds[StringIds["WHEAT"] = 1184] = "WHEAT";
    StringIds[StringIds["USE_TARGET"] = 1186] = "USE_TARGET";
    StringIds[StringIds["SEARCH"] = 1191] = "SEARCH";
    StringIds[StringIds["SEARCH_ALL_CABINETS"] = 9242] = "SEARCH_ALL_CABINETS";
    StringIds[StringIds["OPEN_AND_LOCK"] = 8944] = "OPEN_AND_LOCK";
    StringIds[StringIds["ENTER_ACCESS_CODE"] = 8943] = "ENTER_ACCESS_CODE";
    StringIds[StringIds["REST"] = 9041] = "REST";
    StringIds[StringIds["RESTING"] = 9051] = "RESTING";
    StringIds[StringIds["NOT_TIRED"] = 12159] = "NOT_TIRED";
    StringIds[StringIds["USE_IGNITABLE"] = 9224] = "USE_IGNITABLE";
    StringIds[StringIds["USE_DOOR"] = 78] = "USE_DOOR";
    StringIds[StringIds["UNDO_PLACEMENT"] = 12001] = "UNDO_PLACEMENT";
    StringIds[StringIds["OPEN_TARGET"] = 12156] = "OPEN_TARGET";
    StringIds[StringIds["ATV"] = 12552] = "ATV";
    StringIds[StringIds["PICKUP_TRUCK"] = 12537] = "PICKUP_TRUCK";
    StringIds[StringIds["POLICE_CAR"] = 12538] = "POLICE_CAR";
    StringIds[StringIds["PERMISSIONS_TARGET"] = 12982] = "PERMISSIONS_TARGET";
    StringIds[StringIds["CLAIM_TARGET"] = 11940] = "CLAIM_TARGET";
    StringIds[StringIds["EXTRACT_BLOOD"] = 1453] = "EXTRACT_BLOOD";
    StringIds[StringIds["HARVEST"] = 1009] = "HARVEST";
    StringIds[StringIds["LONG_RANGE_RADIO"] = 12809] = "LONG_RANGE_RADIO";
    StringIds[StringIds["RADIOLOGY_DESK"] = 12752] = "RADIOLOGY_DESK";
    StringIds[StringIds["BATTERED_TRASHCAN"] = 12835] = "BATTERED_TRASHCAN";
    StringIds[StringIds["MAINTENANCE_TOOLBOX"] = 12736] = "MAINTENANCE_TOOLBOX";
    StringIds[StringIds["ICU_DESK"] = 12750] = "ICU_DESK";
    StringIds[StringIds["OPEN_TOOLBOX"] = 12739] = "OPEN_TOOLBOX";
    StringIds[StringIds["OPEN_CABINET"] = 12732] = "OPEN_CABINET";
    StringIds[StringIds["SMALL_PHARMACY_CABINET"] = 12740] = "SMALL_PHARMACY_CABINET";
    StringIds[StringIds["SEARCH_RECORDS"] = 12735] = "SEARCH_RECORDS";
    StringIds[StringIds["RECORDS_CABINET"] = 12734] = "RECORDS_CABINET";
    StringIds[StringIds["EXAMINATION_CABINET"] = 12719] = "EXAMINATION_CABINET";
    StringIds[StringIds["MORGUE_DESK"] = 12696] = "MORGUE_DESK";
    StringIds[StringIds["LOCKED_CABINET"] = 12744] = "LOCKED_CABINET";
    StringIds[StringIds["OPEN_LOCKED_CABINET"] = 12747] = "OPEN_LOCKED_CABINET";
    StringIds[StringIds["AIRDROP_CODE"] = 12795] = "AIRDROP_CODE";
    StringIds[StringIds["ZOMBIE_WALKER"] = 60] = "ZOMBIE_WALKER";
    StringIds[StringIds["DEER"] = 1245] = "DEER";
    StringIds[StringIds["BEAR"] = 1246] = "BEAR";
    StringIds[StringIds["WOLF"] = 1213] = "WOLF";
    StringIds[StringIds["CHECK_WATER"] = 9833] = "CHECK_WATER";
    StringIds[StringIds["FIRE_HYDRANT"] = 14383] = "FIRE_HYDRANT";
    StringIds[StringIds["TACTICAL_ARMOR"] = 12073] = "TACTICAL_ARMOR";
    StringIds[StringIds["BADASS_ARMOR"] = 13065] = "BADASS_ARMOR";
    StringIds[StringIds["WOODEN_ARMOR"] = 11151] = "WOODEN_ARMOR";
    StringIds[StringIds["PLATED_ARMOR"] = 11153] = "PLATED_ARMOR";
    StringIds[StringIds["TACTICAL_HELMET"] = 9945] = "TACTICAL_HELMET";
    StringIds[StringIds["SLEIGH_HELMET"] = 12994] = "SLEIGH_HELMET";
    StringIds[StringIds["GENERAL_HELMET_1"] = 9114] = "GENERAL_HELMET_1";
    StringIds[StringIds["CREEPY_MASK"] = 12898] = "CREEPY_MASK";
    StringIds[StringIds["SCRAY_HALLOWEEN_MASK"] = 12858] = "SCRAY_HALLOWEEN_MASK";
    StringIds[StringIds["UNCLE_SAM_MASK"] = 14171] = "UNCLE_SAM_MASK";
    StringIds[StringIds["CONVEYS"] = 11895] = "CONVEYS";
    StringIds[StringIds["BANSHEE"] = 12310] = "BANSHEE";
    StringIds[StringIds["HOTWIRE"] = 14166] = "HOTWIRE";
})(StringIds || (exports.StringIds = StringIds = {}));
var MovementModifiers;
(function (MovementModifiers) {
    MovementModifiers[MovementModifiers["RESTED"] = 1.11] = "RESTED";
    MovementModifiers[MovementModifiers["SWIZZLE"] = 1.1] = "SWIZZLE";
    MovementModifiers[MovementModifiers["SNARED"] = 0.5] = "SNARED";
    MovementModifiers[MovementModifiers["CONVEYS"] = 1.15] = "CONVEYS";
    MovementModifiers[MovementModifiers["BOOTS"] = 1.07] = "BOOTS";
    MovementModifiers[MovementModifiers["ADRENALINE"] = 1.12] = "ADRENALINE";
})(MovementModifiers || (exports.MovementModifiers = MovementModifiers = {}));
var ContainerErrors;
(function (ContainerErrors) {
    ContainerErrors[ContainerErrors["NONE"] = 0] = "NONE";
    ContainerErrors[ContainerErrors["IN_USE"] = 1] = "IN_USE";
    ContainerErrors[ContainerErrors["WRONG_ITEM_TYPE"] = 2] = "WRONG_ITEM_TYPE";
    ContainerErrors[ContainerErrors["UNKNOWN_CONTAINER"] = 3] = "UNKNOWN_CONTAINER";
    ContainerErrors[ContainerErrors["UNKNOWN_SLOT"] = 4] = "UNKNOWN_SLOT";
    ContainerErrors[ContainerErrors["NO_ITEM_IN_SLOT"] = 5] = "NO_ITEM_IN_SLOT";
    ContainerErrors[ContainerErrors["INTERACTION_VALIDATION"] = 6] = "INTERACTION_VALIDATION";
    ContainerErrors[ContainerErrors["UNKNOWN"] = 99] = "UNKNOWN";
    // custom errors
    ContainerErrors[ContainerErrors["DOES_NOT_ACCEPT_ITEMS"] = 7] = "DOES_NOT_ACCEPT_ITEMS";
    ContainerErrors[ContainerErrors["NOT_MUTABLE"] = 8] = "NOT_MUTABLE";
    ContainerErrors[ContainerErrors["NOT_CONSTRUCTED"] = 9] = "NOT_CONSTRUCTED";
    ContainerErrors[ContainerErrors["NO_SPACE"] = 10] = "NO_SPACE";
    ContainerErrors[ContainerErrors["INVALID_LOADOUT_SLOT"] = 11] = "INVALID_LOADOUT_SLOT";
    ContainerErrors[ContainerErrors["NO_PERMISSION"] = 12] = "NO_PERMISSION";
    ContainerErrors[ContainerErrors["UNACCEPTED_ITEM"] = 13] = "UNACCEPTED_ITEM";
})(ContainerErrors || (exports.ContainerErrors = ContainerErrors = {}));
var FilterIds;
(function (FilterIds) {
    FilterIds[FilterIds["COOKING"] = 2] = "COOKING";
    FilterIds[FilterIds["FURNACE"] = 4] = "FURNACE";
    FilterIds[FilterIds["WEAPONS"] = 5] = "WEAPONS";
    FilterIds[FilterIds["HOUSING"] = 6] = "HOUSING";
    FilterIds[FilterIds["SURVIVAL"] = 7] = "SURVIVAL";
    FilterIds[FilterIds["COMPONENT"] = 8] = "COMPONENT";
    FilterIds[FilterIds["DEW_COLLECTOR"] = 9] = "DEW_COLLECTOR";
    FilterIds[FilterIds["ANIMAL_TRAP"] = 10] = "ANIMAL_TRAP";
})(FilterIds || (exports.FilterIds = FilterIds = {}));
var ItemClasses;
(function (ItemClasses) {
    ItemClasses[ItemClasses["WEAPONS_LONG"] = 25036] = "WEAPONS_LONG";
    ItemClasses[ItemClasses["WEAPONS_PISTOL"] = 4096] = "WEAPONS_PISTOL";
    ItemClasses[ItemClasses["WEAPONS_MELEES"] = 4098] = "WEAPONS_MELEES";
    ItemClasses[ItemClasses["WEAPONS_MELEES0"] = 25037] = "WEAPONS_MELEES0";
    ItemClasses[ItemClasses["WEAPONS_GENERIC"] = 25054] = "WEAPONS_GENERIC";
    ItemClasses[ItemClasses["WEAPONS_CROSSBOW"] = 25047] = "WEAPONS_CROSSBOW";
    ItemClasses[ItemClasses["WEAPON_THROWABLES"] = 25078] = "WEAPON_THROWABLES";
    ItemClasses[ItemClasses["WEAPONS_BOW"] = 25038] = "WEAPONS_BOW";
    ItemClasses[ItemClasses["POUCHES"] = 25013] = "POUCHES";
    ItemClasses[ItemClasses["BACKPACKS"] = 25004] = "BACKPACKS";
    ItemClasses[ItemClasses["CONTAINER_CHEST"] = 25002] = "CONTAINER_CHEST";
    ItemClasses[ItemClasses["CONTAINER_LEGS"] = 25003] = "CONTAINER_LEGS";
    ItemClasses[ItemClasses["THROWABLES"] = 25078] = "THROWABLES";
    ItemClasses[ItemClasses["BODY_ARMOR"] = 25041] = "BODY_ARMOR";
    ItemClasses[ItemClasses["FOOTWEAR"] = 25005] = "FOOTWEAR";
})(ItemClasses || (exports.ItemClasses = ItemClasses = {}));
var VehicleIds;
(function (VehicleIds) {
    VehicleIds[VehicleIds["OFFROADER"] = 1] = "OFFROADER";
    VehicleIds[VehicleIds["PICKUP"] = 2] = "PICKUP";
    VehicleIds[VehicleIds["POLICECAR"] = 3] = "POLICECAR";
    VehicleIds[VehicleIds["ATV"] = 5] = "ATV";
    VehicleIds[VehicleIds["PARACHUTE"] = 13] = "PARACHUTE";
    VehicleIds[VehicleIds["SPECTATE"] = 1337] = "SPECTATE";
})(VehicleIds || (exports.VehicleIds = VehicleIds = {}));
var NpcIds;
(function (NpcIds) {
    NpcIds[NpcIds["ZOMBIE"] = 3] = "ZOMBIE";
    NpcIds[NpcIds["DEER"] = 13] = "DEER";
    NpcIds[NpcIds["WOLF"] = 14] = "WOLF";
    NpcIds[NpcIds["BEAR"] = 825] = "BEAR";
    NpcIds[NpcIds["RABBIT"] = 864] = "RABBIT";
    NpcIds[NpcIds["RAVEN"] = 893] = "RAVEN";
})(NpcIds || (exports.NpcIds = NpcIds = {}));
var Characters;
(function (Characters) {
    Characters[Characters["MALE_WHITE"] = 1] = "MALE_WHITE";
    Characters[Characters["MALE_WHITE_BALD"] = 2] = "MALE_WHITE_BALD";
    Characters[Characters["FEMALE_WHITE_YOUNG"] = 3] = "FEMALE_WHITE_YOUNG";
    Characters[Characters["FEMALE_WHITE"] = 4] = "FEMALE_WHITE";
    Characters[Characters["MALE_BLACK"] = 5] = "MALE_BLACK";
    Characters[Characters["FEMALE_BLACK"] = 6] = "FEMALE_BLACK";
})(Characters || (exports.Characters = Characters = {}));
var LoadoutIds;
(function (LoadoutIds) {
    LoadoutIds[LoadoutIds["CHARACTER"] = 3] = "CHARACTER";
    LoadoutIds[LoadoutIds["VEHICLE_OFFROADER"] = 4] = "VEHICLE_OFFROADER";
    LoadoutIds[LoadoutIds["VEHICLE_PICKUP"] = 6] = "VEHICLE_PICKUP";
    LoadoutIds[LoadoutIds["VEHICLE_POLICECAR"] = 8] = "VEHICLE_POLICECAR";
    LoadoutIds[LoadoutIds["VEHICLE_ATV"] = 14] = "VEHICLE_ATV";
})(LoadoutIds || (exports.LoadoutIds = LoadoutIds = {}));
var LoadoutSlots;
(function (LoadoutSlots) {
    LoadoutSlots[LoadoutSlots["PRIMARY"] = 1] = "PRIMARY";
    LoadoutSlots[LoadoutSlots["SECONDARY"] = 3] = "SECONDARY";
    LoadoutSlots[LoadoutSlots["TERTIARY"] = 4] = "TERTIARY";
    LoadoutSlots[LoadoutSlots["FISTS"] = 7] = "FISTS";
    LoadoutSlots[LoadoutSlots["CHEST"] = 10] = "CHEST";
    LoadoutSlots[LoadoutSlots["HEAD"] = 11] = "HEAD";
    LoadoutSlots[LoadoutSlots["BACK"] = 12] = "BACK";
    LoadoutSlots[LoadoutSlots["FEET"] = 13] = "FEET";
    LoadoutSlots[LoadoutSlots["LEGS"] = 14] = "LEGS";
    LoadoutSlots[LoadoutSlots["BELT"] = 25] = "BELT";
    LoadoutSlots[LoadoutSlots["EYES"] = 29] = "EYES";
    LoadoutSlots[LoadoutSlots["EXTERNAL_CONTAINER"] = 31] = "EXTERNAL_CONTAINER";
    LoadoutSlots[LoadoutSlots["VEHICLE_BATTERY"] = 33] = "VEHICLE_BATTERY";
    LoadoutSlots[LoadoutSlots["ARMOR"] = 38] = "ARMOR";
    LoadoutSlots[LoadoutSlots["RADIO"] = 39] = "RADIO";
    LoadoutSlots[LoadoutSlots["ITEM1"] = 40] = "ITEM1";
    LoadoutSlots[LoadoutSlots["ITEM2"] = 41] = "ITEM2";
})(LoadoutSlots || (exports.LoadoutSlots = LoadoutSlots = {}));
var EquipSlots;
(function (EquipSlots) {
    EquipSlots[EquipSlots["HEAD"] = 1] = "HEAD";
    EquipSlots[EquipSlots["HANDS"] = 2] = "HANDS";
    EquipSlots[EquipSlots["CHEST"] = 3] = "CHEST";
    EquipSlots[EquipSlots["LEGS"] = 4] = "LEGS";
    EquipSlots[EquipSlots["FEET"] = 5] = "FEET";
    EquipSlots[EquipSlots["RHAND"] = 7] = "RHAND";
    EquipSlots[EquipSlots["BACKPACK"] = 10] = "BACKPACK";
    EquipSlots[EquipSlots["HAIR"] = 27] = "HAIR";
    EquipSlots[EquipSlots["FACE"] = 28] = "FACE";
    EquipSlots[EquipSlots["EYES"] = 29] = "EYES";
    EquipSlots[EquipSlots["ARMOR"] = 100] = "ARMOR";
})(EquipSlots || (exports.EquipSlots = EquipSlots = {}));
var ResourceIds;
(function (ResourceIds) {
    ResourceIds[ResourceIds["HEALTH"] = 1] = "HEALTH";
    ResourceIds[ResourceIds["HUNGER"] = 4] = "HUNGER";
    ResourceIds[ResourceIds["HYDRATION"] = 5] = "HYDRATION";
    ResourceIds[ResourceIds["STAMINA"] = 6] = "STAMINA";
    ResourceIds[ResourceIds["VIRUS"] = 12] = "VIRUS";
    ResourceIds[ResourceIds["BLEEDING"] = 21] = "BLEEDING";
    ResourceIds[ResourceIds["ENDURANCE"] = 22] = "ENDURANCE";
    ResourceIds[ResourceIds["COMFORT"] = 68] = "COMFORT";
    ResourceIds[ResourceIds["FUEL"] = 396] = "FUEL";
    ResourceIds[ResourceIds["CONDITION"] = 561] = "CONDITION";
    ResourceIds[ResourceIds["CONSTRUCTION_CONDITION"] = 567] = "CONSTRUCTION_CONDITION";
    ResourceIds[ResourceIds["TOXICITY"] = 611] = "TOXICITY";
})(ResourceIds || (exports.ResourceIds = ResourceIds = {}));
var ResourceTypes;
(function (ResourceTypes) {
    ResourceTypes[ResourceTypes["HEALTH"] = 1] = "HEALTH";
    ResourceTypes[ResourceTypes["HUNGER"] = 4] = "HUNGER";
    ResourceTypes[ResourceTypes["HYDRATION"] = 5] = "HYDRATION";
    ResourceTypes[ResourceTypes["STAMINA"] = 6] = "STAMINA";
    ResourceTypes[ResourceTypes["VIRUS"] = 12] = "VIRUS";
    ResourceTypes[ResourceTypes["BLEEDING"] = 21] = "BLEEDING";
    ResourceTypes[ResourceTypes["ENDURANCE"] = 22] = "ENDURANCE";
    ResourceTypes[ResourceTypes["COMFORT"] = 68] = "COMFORT";
    ResourceTypes[ResourceTypes["FUEL"] = 50] = "FUEL";
    ResourceTypes[ResourceTypes["CONDITION"] = 1] = "CONDITION";
    ResourceTypes[ResourceTypes["TOXICITY"] = 75] = "TOXICITY";
})(ResourceTypes || (exports.ResourceTypes = ResourceTypes = {}));
var ResourceIndicators;
(function (ResourceIndicators) {
    ResourceIndicators["BLEEDING_LIGHT"] = "BLEEDING_LIGHT";
    ResourceIndicators["BLEEDING_MODERATE"] = "BLEEDING_MODERATE";
    ResourceIndicators["BLEEDING_SEVERE"] = "BLEEDING_SEVERE";
    ResourceIndicators["TIRED"] = "TIRED";
    ResourceIndicators["VERY_TIRED"] = "VERY_TIRED";
    ResourceIndicators["EXHAUSTED"] = "EXHAUSTED";
    ResourceIndicators["WELL_RESTED"] = "WELL_RESTED";
    ResourceIndicators["HEALING_BANDAGE"] = "HEALING_BANDAGE";
    ResourceIndicators["HEALING_BANDAGE_DRESSED"] = "HEALING_BANDAGE_DRESSED";
    ResourceIndicators["HEALING_FIRST_AID"] = "HEALING_FIRST_AID";
    ResourceIndicators["IMMUNITY_BOOST_LOW"] = "IMMUNITY_BOOST_LOW";
    ResourceIndicators["IMMUNITY_BOOST_MEDIUM"] = "IMMUNITY_BOOST_MEDIUM";
    ResourceIndicators["COLD_MEDICINE"] = "COLD_MEDICINE";
    ResourceIndicators["SWIZZLE"] = "SWIZZLE";
    ResourceIndicators["DEHYDRATED"] = "DEHYDRATED";
    ResourceIndicators["STARVING"] = "STARVING";
    ResourceIndicators["COMFORT_PLUS"] = "COMFORT +";
    ResourceIndicators["COMFORT_PLUSPLUS"] = "COMFORT ++";
    ResourceIndicators["FOOD_POISONING"] = "FOOD POISONING";
    ResourceIndicators["BEES"] = "BEES!";
    ResourceIndicators["COFFEE_SUGAR"] = "COFFEE_SUGAR";
    ResourceIndicators["ADRENALINE"] = "ADRENALINE";
    ResourceIndicators["ADRENALINE_AFTER_EFFECTS"] = "ADRENALINE AFTER EFFECTS";
})(ResourceIndicators || (exports.ResourceIndicators = ResourceIndicators = {}));
var Abilities;
(function (Abilities) {
    Abilities[Abilities["VEHICLE_HEADLIGHTS"] = 99998] = "VEHICLE_HEADLIGHTS";
    Abilities[Abilities["VEHICLE_TURBO"] = 1111141] = "VEHICLE_TURBO";
    Abilities[Abilities["VEHICLE_SIREN"] = 1111295] = "VEHICLE_SIREN";
    Abilities[Abilities["VEHICLE_HORN"] = 1111301] = "VEHICLE_HORN";
})(Abilities || (exports.Abilities = Abilities = {}));
var TreeIds;
(function (TreeIds) {
    TreeIds[TreeIds["COMMONHAZEL"] = 0] = "COMMONHAZEL";
    TreeIds[TreeIds["WESTERNCEDARSAPLING"] = 1] = "WESTERNCEDARSAPLING";
    TreeIds[TreeIds["WESTERNCEDAR"] = 2] = "WESTERNCEDAR";
    TreeIds[TreeIds["GREENMAPLEDEAD"] = 3] = "GREENMAPLEDEAD";
    TreeIds[TreeIds["SAPLINGMAPLE"] = 4] = "SAPLINGMAPLE";
    TreeIds[TreeIds["REDMAPLE"] = 5] = "REDMAPLE";
    TreeIds[TreeIds["GREENMAPLE"] = 6] = "GREENMAPLE";
    TreeIds[TreeIds["BOULDERS"] = 7] = "BOULDERS";
    TreeIds[TreeIds["WHITEBIRCH"] = 8] = "WHITEBIRCH";
    TreeIds[TreeIds["DEVILCLUB"] = 9] = "DEVILCLUB";
    TreeIds[TreeIds["REDCEDAR"] = 10] = "REDCEDAR";
    TreeIds[TreeIds["VINEMAPLE"] = 11] = "VINEMAPLE";
    TreeIds[TreeIds["OREGONOAK"] = 12] = "OREGONOAK";
    TreeIds[TreeIds["BLACKBERRY"] = 13] = "BLACKBERRY";
    TreeIds[TreeIds["PAPERBIRCH"] = 14] = "PAPERBIRCH";
    TreeIds[TreeIds["BIGLEAFMAPLE"] = 15] = "BIGLEAFMAPLE";
    TreeIds[TreeIds["BIGLEAFMAPLE_YOUNG01"] = 16] = "BIGLEAFMAPLE_YOUNG01";
    TreeIds[TreeIds["BIGLEAFMAPLE_YOUNG02"] = 17] = "BIGLEAFMAPLE_YOUNG02";
    TreeIds[TreeIds["DECIDUOUSSHRUB01"] = 18] = "DECIDUOUSSHRUB01";
    TreeIds[TreeIds["BOULDER"] = 19] = "BOULDER";
    TreeIds[TreeIds["APPLETREE01"] = 20] = "APPLETREE01";
    TreeIds[TreeIds["BUSHES_LARGE01"] = 21] = "BUSHES_LARGE01";
})(TreeIds || (exports.TreeIds = TreeIds = {}));
/* I got these values from Z1BR - Jason */
var Recipes;
(function (Recipes) {
    Recipes[Recipes["GUNPOWDER"] = 2] = "GUNPOWDER";
    Recipes[Recipes["METAL_SHEET"] = 6] = "METAL_SHEET";
    Recipes[Recipes["PIPE"] = 7] = "PIPE";
    Recipes[Recipes["COMBAT_KNIFE"] = 8] = "COMBAT_KNIFE";
    Recipes[Recipes["SLEEPING_MAT"] = 18] = "SLEEPING_MAT";
    Recipes[Recipes["COFFEE"] = 22] = "COFFEE";
    Recipes[Recipes["TORCH"] = 23] = "TORCH";
    Recipes[Recipes["MAKESHIFT_HATCHET"] = 24] = "MAKESHIFT_HATCHET";
    Recipes[Recipes["FLARE"] = 25] = "FLARE";
    Recipes[Recipes["DEER_STEAK"] = 28] = "DEER_STEAK";
    Recipes[Recipes["FURNACE"] = 31] = "FURNACE";
    Recipes[Recipes["COFFEE_SUGAR"] = 39] = "COFFEE_SUGAR";
    Recipes[Recipes["FUEL"] = 40] = "FUEL";
    Recipes[Recipes["LANDMINE"] = 41] = "LANDMINE";
    Recipes[Recipes["SALINE"] = 42] = "SALINE";
    Recipes[Recipes["FIRST_AID"] = 43] = "FIRST_AID";
    Recipes[Recipes["MACHETTE"] = 46] = "MACHETTE";
    Recipes[Recipes["DEER_JERKY"] = 47] = "DEER_JERKY";
    Recipes[Recipes["STORAGE"] = 48] = "STORAGE";
    Recipes[Recipes["ANIMAL_TRAP"] = 51] = "ANIMAL_TRAP";
    Recipes[Recipes["DEW_COLLECTOR"] = 54] = "DEW_COLLECTOR";
    Recipes[Recipes["PUNJI_STICKS"] = 55] = "PUNJI_STICKS";
    Recipes[Recipes["METAL_SHARD"] = 61] = "METAL_SHARD";
    Recipes[Recipes["BARBED_WIRE"] = 65] = "BARBED_WIRE";
    Recipes[Recipes["COOKED_RABBIT"] = 66] = "COOKED_RABBIT";
    Recipes[Recipes["RABBIT_STEW"] = 67] = "RABBIT_STEW";
    Recipes[Recipes["WOLF_STEAK"] = 68] = "WOLF_STEAK";
    Recipes[Recipes["WOODEN_BARRICADE"] = 69] = "WOODEN_BARRICADE";
    Recipes[Recipes["NAIL"] = 71] = "NAIL";
    Recipes[Recipes["METAL_BRACKET"] = 72] = "METAL_BRACKET";
    Recipes[Recipes["FRAMED_BACKPACK"] = 76] = "FRAMED_BACKPACK";
    Recipes[Recipes["METAL_GATE"] = 78] = "METAL_GATE";
    Recipes[Recipes["METAL_WALL"] = 79] = "METAL_WALL";
    Recipes[Recipes["SHELTER"] = 80] = "SHELTER";
    Recipes[Recipes["LARGE_SHELTER"] = 81] = "LARGE_SHELTER";
    Recipes[Recipes["STRUCTURE_STAIRS"] = 82] = "STRUCTURE_STAIRS";
    Recipes[Recipes["BLACKBERRY_JUICE"] = 103] = "BLACKBERRY_JUICE";
    Recipes[Recipes["GRILLED_TROUT"] = 104] = "GRILLED_TROUT";
    Recipes[Recipes["PURIFIED_WATER_1"] = 107] = "PURIFIED_WATER_1";
    Recipes[Recipes["DECK_FOUNDATION"] = 109] = "DECK_FOUNDATION";
    Recipes[Recipes["SURVIVAL_BORSCHT"] = 110] = "SURVIVAL_BORSCHT";
    Recipes[Recipes["GROUND_TILLER"] = 112] = "GROUND_TILLER";
    Recipes[Recipes["ETHANOL"] = 113] = "ETHANOL";
    Recipes[Recipes["CORN_MASH"] = 114] = "CORN_MASH";
    Recipes[Recipes["ROASTED_CORN"] = 115] = "ROASTED_CORN";
    Recipes[Recipes["MOLOTOV_COCKTAIL"] = 116] = "MOLOTOV_COCKTAIL";
    Recipes[Recipes["ETHANOL_TORCH"] = 117] = "ETHANOL_TORCH";
    Recipes[Recipes["SATCHEL"] = 118] = "SATCHEL";
    Recipes[Recipes["SNARE"] = 119] = "SNARE";
    Recipes[Recipes["SHACK"] = 120] = "SHACK";
    Recipes[Recipes["WOODEN_DOOR"] = 121] = "WOODEN_DOOR";
    Recipes[Recipes["SMALL_SHACK"] = 122] = "SMALL_SHACK";
    Recipes[Recipes["IMPROVISED_COMPASS"] = 124] = "IMPROVISED_COMPASS";
    Recipes[Recipes["YEAST"] = 125] = "YEAST";
    Recipes[Recipes["BARBEQUE"] = 126] = "BARBEQUE";
    Recipes[Recipes["BEAR_STEAK"] = 128] = "BEAR_STEAK";
    Recipes[Recipes["BOW_DRILL"] = 129] = "BOW_DRILL";
    Recipes[Recipes["FLOUR"] = 131] = "FLOUR";
    Recipes[Recipes["SURVIVAL_BREAD"] = 132] = "SURVIVAL_BREAD";
    Recipes[Recipes["DEER_SANDWICH"] = 133] = "DEER_SANDWICH";
    Recipes[Recipes["BEAR_SANDWICH"] = 134] = "BEAR_SANDWICH";
    Recipes[Recipes["WOLF_SANDWICH"] = 135] = "WOLF_SANDWICH";
    Recipes[Recipes["RABBIT_SANDWICH"] = 136] = "RABBIT_SANDWICH";
    Recipes[Recipes["DEER_SCENT"] = 137] = "DEER_SCENT";
    Recipes[Recipes["BACKPACK_FRAME"] = 138] = "BACKPACK_FRAME";
    Recipes[Recipes["BASIC_SHACK"] = 139] = "BASIC_SHACK";
    Recipes[Recipes["BASIC_SHACK_DOOR"] = 140] = "BASIC_SHACK_DOOR";
    Recipes[Recipes["PURIFIED_WATER_2"] = 141] = "PURIFIED_WATER_2";
    Recipes[Recipes["HAMMER"] = 142] = "HAMMER";
    Recipes[Recipes["SMOKE_FLARE"] = 144] = "SMOKE_FLARE";
    Recipes[Recipes["HAND_SHOVEL"] = 145] = "HAND_SHOVEL";
    Recipes[Recipes["IED"] = 146] = "IED";
    Recipes[Recipes["GROUND_TAMPER"] = 147] = "GROUND_TAMPER";
    Recipes[Recipes["BLACKBERRY_PIE"] = 148] = "BLACKBERRY_PIE";
    Recipes[Recipes["SWIZZLE"] = 149] = "SWIZZLE";
    Recipes[Recipes["RIGGED_LIGHT"] = 154] = "RIGGED_LIGHT";
    Recipes[Recipes["METAL_DOOR"] = 166] = "METAL_DOOR";
    Recipes[Recipes["GUN_REPAIR_KIT"] = 171] = "GUN_REPAIR_KIT";
    Recipes[Recipes["WORKBENCH"] = 172] = "WORKBENCH";
    Recipes[Recipes["UPPER_METAL_WALL"] = 173] = "UPPER_METAL_WALL";
    Recipes[Recipes["UPPER_LEVEL_LARGE_SHELTER"] = 174] = "UPPER_LEVEL_LARGE_SHELTER";
    Recipes[Recipes["UPPER_LEVEL_SHELTER"] = 175] = "UPPER_LEVEL_SHELTER";
    Recipes[Recipes["UPPER_LEVEL_STAIRS"] = 176] = "UPPER_LEVEL_STAIRS";
    Recipes[Recipes["DEMOLITION_HAMMER"] = 177] = "DEMOLITION_HAMMER";
    Recipes[Recipes["GUN_PART_1"] = 193] = "GUN_PART_1";
    Recipes[Recipes["GUN_PART_2"] = 194] = "GUN_PART_2";
    Recipes[Recipes["GUN_PART_3"] = 195] = "GUN_PART_3";
    Recipes[Recipes["GUN_PART_4"] = 195] = "GUN_PART_4";
    Recipes[Recipes["GUN_PART_5"] = 197] = "GUN_PART_5";
    Recipes[Recipes["GUN_PART_6"] = 198] = "GUN_PART_6";
    Recipes[Recipes["GUN_PART_7"] = 199] = "GUN_PART_7";
    Recipes[Recipes["GUN_PART_8"] = 200] = "GUN_PART_8";
    Recipes[Recipes["CANDLE"] = 202] = "CANDLE";
    Recipes[Recipes["BEE_BOX"] = 203] = "BEE_BOX";
    Recipes[Recipes["WOODEN_BODY_ARMOR"] = 204] = "WOODEN_BODY_ARMOR";
    Recipes[Recipes["PLATED_BODY_ARMOR"] = 205] = "PLATED_BODY_ARMOR";
    Recipes[Recipes["DRESSED_BANDAGE"] = 206] = "DRESSED_BANDAGE";
    Recipes[Recipes["METAL_DOORWAY"] = 207] = "METAL_DOORWAY";
    Recipes[Recipes["SANDBAG_WALL"] = 208] = "SANDBAG_WALL";
    Recipes[Recipes["FOUNDATION_RAMP"] = 214] = "FOUNDATION_RAMP";
    Recipes[Recipes["FOUNDATION_STAIRS"] = 215] = "FOUNDATION_STAIRS";
    Recipes[Recipes["GUN_PART_9"] = 216] = "GUN_PART_9";
    Recipes[Recipes["LOOKOUT_TOWER"] = 217] = "LOOKOUT_TOWER";
    Recipes[Recipes["FOUNDATION_EXPANSION"] = 218] = "FOUNDATION_EXPANSION";
    Recipes[Recipes["FLAMING_ARROW_1"] = 220] = "FLAMING_ARROW_1";
    Recipes[Recipes["PARACHUTE_FLARE"] = 221] = "PARACHUTE_FLARE";
    Recipes[Recipes["BASIC_BANDANA"] = 222] = "BASIC_BANDANA";
    Recipes[Recipes["FINGERLESS_GLOVES"] = 223] = "FINGERLESS_GLOVES";
    Recipes[Recipes["MODIFIED_AK47"] = 224] = "MODIFIED_AK47";
    Recipes[Recipes["GUN_UPGRADE_KIT"] = 230] = "GUN_UPGRADE_KIT";
    Recipes[Recipes["MODIFIED_TOXIC_AK47"] = 231] = "MODIFIED_TOXIC_AK47";
    Recipes[Recipes["MEDICATED_BANDAGE"] = 232] = "MEDICATED_BANDAGE";
    Recipes[Recipes["SYRINGE_H1Z1_REDUCER"] = 233] = "SYRINGE_H1Z1_REDUCER";
    Recipes[Recipes["SYRINGE_H1Z1_SUPPRESSANT"] = 234] = "SYRINGE_H1Z1_SUPPRESSANT";
    Recipes[Recipes["SYRINGE_H1Z1_INFECTED_BLOOD"] = 235] = "SYRINGE_H1Z1_INFECTED_BLOOD";
    Recipes[Recipes["SUPPRESSANT_MIXTURE_K49"] = 236] = "SUPPRESSANT_MIXTURE_K49";
    Recipes[Recipes["VIAL_H1Z1_SUPPRESSANT"] = 237] = "VIAL_H1Z1_SUPPRESSANT";
    Recipes[Recipes["ANTIBIOTIC_CREAM"] = 239] = "ANTIBIOTIC_CREAM";
    Recipes[Recipes["IDENTIFIED_ANTI_VIRAL_BOTTLE"] = 240] = "IDENTIFIED_ANTI_VIRAL_BOTTLE";
    Recipes[Recipes["CHARGED_PHONE"] = 241] = "CHARGED_PHONE";
    Recipes[Recipes["TREATED_BRAIN"] = 242] = "TREATED_BRAIN";
    Recipes[Recipes["AIRDROP_CODE"] = 248] = "AIRDROP_CODE";
    Recipes[Recipes["TAN_GHILLIE_SUIT"] = 249] = "TAN_GHILLIE_SUIT";
    Recipes[Recipes["HUGZ_NEEDED_SCRUBS_SHIRT"] = 250] = "HUGZ_NEEDED_SCRUBS_SHIRT";
    Recipes[Recipes["KURUMA_MEDICAL_SCRUBS_SHIRT"] = 251] = "KURUMA_MEDICAL_SCRUBS_SHIRT";
    Recipes[Recipes["MILITARY_SCRUBS_SHIRT"] = 252] = "MILITARY_SCRUBS_SHIRT";
    Recipes[Recipes["HAPPY_SKULL_SCRUBS_SHIRT"] = 253] = "HAPPY_SKULL_SCRUBS_SHIRT";
    Recipes[Recipes["FIRE_TRAP"] = 258] = "FIRE_TRAP";
    Recipes[Recipes["TRAP_IGNITION_KIT"] = 259] = "TRAP_IGNITION_KIT";
    Recipes[Recipes["FLASH_TRAP"] = 260] = "FLASH_TRAP";
    Recipes[Recipes["GAS_TRAP"] = 261] = "GAS_TRAP";
    Recipes[Recipes["SHOCK_TRAP"] = 262] = "SHOCK_TRAP";
    Recipes[Recipes["REPAIR_BOX"] = 288] = "REPAIR_BOX";
    Recipes[Recipes["SCREAM_GRENADE"] = 289] = "SCREAM_GRENADE";
    Recipes[Recipes["HANDFUL_BLACKBERRIES"] = 290] = "HANDFUL_BLACKBERRIES";
    Recipes[Recipes["FLAMING_ARROW_2"] = 291] = "FLAMING_ARROW_2";
    Recipes[Recipes["MAKESHIFT_ARMOR"] = 293] = "MAKESHIFT_ARMOR";
    Recipes[Recipes["EXPLOSIVE_ARROW"] = 294] = "EXPLOSIVE_ARROW";
    Recipes[Recipes["PROCOAGULANT"] = 297] = "PROCOAGULANT";
    Recipes[Recipes["PURGE"] = 304] = "PURGE";
    Recipes[Recipes["FROSTBITE"] = 305] = "FROSTBITE";
    Recipes[Recipes["REAPER"] = 306] = "REAPER";
    Recipes[Recipes["BLAZE"] = 307] = "BLAZE";
    Recipes[Recipes["NAGAFENS_RAGE"] = 308] = "NAGAFENS_RAGE";
    Recipes[Recipes["BLUE_CRESCENDO"] = 309] = "BLUE_CRESCENDO";
    Recipes[Recipes["JAGGED_SKINNING_KNIFE"] = 310] = "JAGGED_SKINNING_KNIFE";
    Recipes[Recipes["PUNJI_STICK_ROW"] = 312] = "PUNJI_STICK_ROW";
    Recipes[Recipes["FIELD_BANDAGE"] = 325] = "FIELD_BANDAGE";
})(Recipes || (exports.Recipes = Recipes = {}));
var AccountItems;
(function (AccountItems) {
    AccountItems[AccountItems["FULL_METAL_PUMPKIN_MASK"] = 3771] = "FULL_METAL_PUMPKIN_MASK";
    AccountItems[AccountItems["HAUNTED_HOODIE"] = 3772] = "HAUNTED_HOODIE";
    AccountItems[AccountItems["DAY_OF_THE_DEAD_HOCKEY_MASK"] = 3773] = "DAY_OF_THE_DEAD_HOCKEY_MASK";
    AccountItems[AccountItems["FRANKENSWINE_BACKPACK"] = 3774] = "FRANKENSWINE_BACKPACK";
    AccountItems[AccountItems["KRINGLE_HOLIDAY_HAT"] = 2814] = "KRINGLE_HOLIDAY_HAT";
    AccountItems[AccountItems["REWARD_CRATE_SHOWDOWN"] = 3630] = "REWARD_CRATE_SHOWDOWN";
    AccountItems[AccountItems["REWARD_CRATE_ALPHA_LAUNCH"] = 3623] = "REWARD_CRATE_ALPHA_LAUNCH";
    AccountItems[AccountItems["REWARD_CRATE_PREDATOR_GOLD"] = 3211] = "REWARD_CRATE_PREDATOR_GOLD";
    AccountItems[AccountItems["REWARD_CRATE_PREDATOR_SILVER"] = 3210] = "REWARD_CRATE_PREDATOR_SILVER";
    AccountItems[AccountItems["REWARD_CRATE_PREDATOR_BRONZE"] = 3209] = "REWARD_CRATE_PREDATOR_BRONZE";
    AccountItems[AccountItems["REWARD_CRATE_PREDATOR"] = 3208] = "REWARD_CRATE_PREDATOR";
    AccountItems[AccountItems["REWARD_CRATE_EZW_GOLD"] = 3122] = "REWARD_CRATE_EZW_GOLD";
    AccountItems[AccountItems["REWARD_CRATE_EZW_SILVER"] = 3121] = "REWARD_CRATE_EZW_SILVER";
    AccountItems[AccountItems["REWARD_CRATE_EZW_BRONZE"] = 3120] = "REWARD_CRATE_EZW_BRONZE";
    AccountItems[AccountItems["REWARD_CRATE_EZW"] = 3119] = "REWARD_CRATE_EZW";
    AccountItems[AccountItems["REWARD_CRATE_RENEGADE_GOLD"] = 2893] = "REWARD_CRATE_RENEGADE_GOLD";
    AccountItems[AccountItems["REWARD_CRATE_RENEGADE_SILVER"] = 2892] = "REWARD_CRATE_RENEGADE_SILVER";
    AccountItems[AccountItems["REWARD_CRATE_RENEGADE_BRONZE"] = 2891] = "REWARD_CRATE_RENEGADE_BRONZE";
    AccountItems[AccountItems["REWARD_CRATE_RENEGADE"] = 2890] = "REWARD_CRATE_RENEGADE";
    AccountItems[AccountItems["REWARD_CRATE_WASTELAND_GOLD"] = 3028] = "REWARD_CRATE_WASTELAND_GOLD";
    AccountItems[AccountItems["REWARD_CRATE_WASTELAND_SILVER"] = 3027] = "REWARD_CRATE_WASTELAND_SILVER";
    AccountItems[AccountItems["REWARD_CRATE_WASTELAND_BRONZE"] = 3026] = "REWARD_CRATE_WASTELAND_BRONZE";
    AccountItems[AccountItems["REWARD_CRATE_WASTELAND"] = 3025] = "REWARD_CRATE_WASTELAND";
    AccountItems[AccountItems["REWARD_CRATE_BR_1"] = 3625] = "REWARD_CRATE_BR_1";
    AccountItems[AccountItems["REWARD_CRATE_INVITATIONAL"] = 3629] = "REWARD_CRATE_INVITATIONAL";
    AccountItems[AccountItems["REWARD_CRATE_WEARABLES"] = 3624] = "REWARD_CRATE_WEARABLES";
    AccountItems[AccountItems["REWARD_CRATE_MARAUDER"] = 3626] = "REWARD_CRATE_MARAUDER";
    AccountItems[AccountItems["REWARD_CRATE_MERCENARY"] = 3628] = "REWARD_CRATE_MERCENARY";
    AccountItems[AccountItems["REWARD_CRATE_RONIN"] = 3502] = "REWARD_CRATE_RONIN";
    AccountItems[AccountItems["REWARD_CRATE_INVITATIONAL_2016_GOLD"] = 3573] = "REWARD_CRATE_INVITATIONAL_2016_GOLD";
    AccountItems[AccountItems["REWARD_CRATE_INVITATIONAL_2016_SILVER"] = 3572] = "REWARD_CRATE_INVITATIONAL_2016_SILVER";
    AccountItems[AccountItems["REWARD_CRATE_INVITATIONAL_2016_BRONZE"] = 3571] = "REWARD_CRATE_INVITATIONAL_2016_BRONZE";
    AccountItems[AccountItems["REWARD_CRATE_INVITATIONAL_2016"] = 3570] = "REWARD_CRATE_INVITATIONAL_2016";
    AccountItems[AccountItems["REWARD_CRATE_INFERNAL_GOLD"] = 3824] = "REWARD_CRATE_INFERNAL_GOLD";
    AccountItems[AccountItems["REWARD_CRATE_INFERNAL_SILVER"] = 3823] = "REWARD_CRATE_INFERNAL_SILVER";
    AccountItems[AccountItems["REWARD_CRATE_INFERNAL_BRONZE"] = 3822] = "REWARD_CRATE_INFERNAL_BRONZE";
    AccountItems[AccountItems["REWARD_CRATE_INFERNAL"] = 3821] = "REWARD_CRATE_INFERNAL";
    AccountItems[AccountItems["REWARD_CRATE_VIGILANTE"] = 3627] = "REWARD_CRATE_VIGILANTE";
    AccountItems[AccountItems["REWARD_CRATE_H1EMU"] = 5063] = "REWARD_CRATE_H1EMU";
    AccountItems[AccountItems["REWARD_CRATE_H1EMUEXCLUSIVE"] = 5026] = "REWARD_CRATE_H1EMUEXCLUSIVE";
    AccountItems[AccountItems["REWARD_CRATE_VICTORY"] = 5064] = "REWARD_CRATE_VICTORY";
    AccountItems[AccountItems["REWARD_CRATE_BREAKOUT"] = 5255] = "REWARD_CRATE_BREAKOUT";
    AccountItems[AccountItems["REWARD_CRATE_SWIRL"] = 5256] = "REWARD_CRATE_SWIRL";
    AccountItems[AccountItems["REWARD_CRATE_INFERNAL_2025"] = 5333] = "REWARD_CRATE_INFERNAL_2025";
    AccountItems[AccountItems["REWARD_CRATE_FROST"] = 5375] = "REWARD_CRATE_FROST";
    AccountItems[AccountItems["MYSTERY_BAG_KOTK"] = 3239] = "MYSTERY_BAG_KOTK";
    AccountItems[AccountItems["KOTK_ELITE_BAG"] = 3240] = "KOTK_ELITE_BAG";
    AccountItems[AccountItems["SKIRMISH_SMALL_FIREARMS"] = 3274] = "SKIRMISH_SMALL_FIREARMS";
    AccountItems[AccountItems["SKIRMISH_SHOTTIES_SNIPERS"] = 3365] = "SKIRMISH_SHOTTIES_SNIPERS";
    AccountItems[AccountItems["SKIRMISH_IRONMAN"] = 3447] = "SKIRMISH_IRONMAN";
    AccountItems[AccountItems["SKIRMISH_REANIMATED"] = 3523] = "SKIRMISH_REANIMATED";
    AccountItems[AccountItems["SKIRMISH_SHOTTIES_SNIPERS_2"] = 3633] = "SKIRMISH_SHOTTIES_SNIPERS_2";
    AccountItems[AccountItems["MYSTERY_BAG_1"] = 1840] = "MYSTERY_BAG_1";
    AccountItems[AccountItems["ELITE_BAG"] = 1841] = "ELITE_BAG";
    AccountItems[AccountItems["MYSTERY_BAG_2"] = 1879] = "MYSTERY_BAG_2";
    AccountItems[AccountItems["ELITE_BAG_HARDCORE"] = 1880] = "ELITE_BAG_HARDCORE";
    AccountItems[AccountItems["MYSTERY_BAG_V2"] = 2033] = "MYSTERY_BAG_V2";
})(AccountItems || (exports.AccountItems = AccountItems = {}));
var Items;
(function (Items) {
    //#region WEAPONS
    Items[Items["WEAPON_AR15"] = 10] = "WEAPON_AR15";
    Items[Items["WEAPON_AK47"] = 2229] = "WEAPON_AK47";
    Items[Items["WEAPON_SHOTGUN"] = 1374] = "WEAPON_SHOTGUN";
    Items[Items["WEAPON_CROWBAR"] = 82] = "WEAPON_CROWBAR";
    Items[Items["WEAPON_COMBATKNIFE"] = 84] = "WEAPON_COMBATKNIFE";
    Items[Items["WEAPON_MACHETE01"] = 83] = "WEAPON_MACHETE01";
    Items[Items["WEAPON_HISREGARD_MACHETE"] = 2228] = "WEAPON_HISREGARD_MACHETE";
    Items[Items["WEAPON_TOXIC_COMBATKNIFE"] = 2494] = "WEAPON_TOXIC_COMBATKNIFE";
    Items[Items["WEAPON_KATANA"] = 2961] = "WEAPON_KATANA";
    Items[Items["WEAPON_BAT_WOOD"] = 1724] = "WEAPON_BAT_WOOD";
    Items[Items["WEAPON_MOURNING_WOOD"] = 1442] = "WEAPON_MOURNING_WOOD";
    Items[Items["WEAPON_GUITAR"] = 1733] = "WEAPON_GUITAR";
    Items[Items["WEAPON_AXE_WOOD"] = 58] = "WEAPON_AXE_WOOD";
    Items[Items["WEAPON_AXE_FIRE"] = 1745] = "WEAPON_AXE_FIRE";
    Items[Items["WEAPON_HAMMER"] = 1536] = "WEAPON_HAMMER";
    Items[Items["WEAPON_HATCHET"] = 3] = "WEAPON_HATCHET";
    Items[Items["WEAPON_PIPE"] = 1448] = "WEAPON_PIPE";
    Items[Items["WEAPON_BAT_ALUM"] = 1721] = "WEAPON_BAT_ALUM";
    Items[Items["WEAPON_BOW_MAKESHIFT"] = 113] = "WEAPON_BOW_MAKESHIFT";
    Items[Items["WEAPON_BOW_WOOD"] = 1720] = "WEAPON_BOW_WOOD";
    Items[Items["WEAPON_BOW_RECURVE"] = 1986] = "WEAPON_BOW_RECURVE";
    Items[Items["WEAPON_1911"] = 2] = "WEAPON_1911";
    Items[Items["WEAPON_M9"] = 1997] = "WEAPON_M9";
    Items[Items["WEAPON_308"] = 1373] = "WEAPON_308";
    Items[Items["WEAPON_BINOCULARS"] = 1542] = "WEAPON_BINOCULARS";
    Items[Items["WEAPON_CROSSBOW"] = 2246] = "WEAPON_CROSSBOW";
    Items[Items["WEAPON_R380"] = 1991] = "WEAPON_R380";
    Items[Items["WEAPON_MOLOTOV"] = 14] = "WEAPON_MOLOTOV";
    Items[Items["WEAPON_MAGNUM"] = 1718] = "WEAPON_MAGNUM";
    Items[Items["WEAPON_FLASHLIGHT"] = 1380] = "WEAPON_FLASHLIGHT";
    Items[Items["WEAPON_WRENCH"] = 1538] = "WEAPON_WRENCH";
    Items[Items["WEAPON_BRANCH"] = 1725] = "WEAPON_BRANCH";
    Items[Items["WEAPON_FISTS"] = 85] = "WEAPON_FISTS";
    Items[Items["WEAPON_FROSTBITE"] = 3445] = "WEAPON_FROSTBITE";
    Items[Items["WEAPON_BLAZE"] = 3446] = "WEAPON_BLAZE";
    Items[Items["WEAPON_NAGAFENS_RAGE"] = 3448] = "WEAPON_NAGAFENS_RAGE";
    Items[Items["WEAPON_PURGE"] = 3449] = "WEAPON_PURGE";
    Items[Items["WEAPON_REAPER"] = 3450] = "WEAPON_REAPER";
    Items[Items["WEAPON_HAMMER_DEMOLITION"] = 1903] = "WEAPON_HAMMER_DEMOLITION";
    Items[Items["WEAPON_TORCH"] = 5] = "WEAPON_TORCH";
    Items[Items["WEAPON_TORCH_ETHANOL"] = 1389] = "WEAPON_TORCH_ETHANOL";
    Items[Items["WEAPON_HATCHET_MAKESHIFT"] = 1708] = "WEAPON_HATCHET_MAKESHIFT";
    Items[Items["WEAPON_AK47_MODIFIED"] = 2399] = "WEAPON_AK47_MODIFIED";
    Items[Items["WEAPON_SPEAR"] = 1382] = "WEAPON_SPEAR";
    Items[Items["WEAPON_REMOVER"] = 1776] = "WEAPON_REMOVER";
    Items[Items["WEAPON_GREENWOOD_SNIPER"] = 5014] = "WEAPON_GREENWOOD_SNIPER";
    Items[Items["WEAPON_PATRIOTIC_SNIPER"] = 2908] = "WEAPON_PATRIOTIC_SNIPER";
    Items[Items["WEAPON_H1EMU_SNIPER"] = 5008] = "WEAPON_H1EMU_SNIPER";
    Items[Items["WEAPON_H1EMU_AR"] = 5024] = "WEAPON_H1EMU_AR";
    Items[Items["WEAPON_ROSE_GARDEN_SHOTGUN"] = 5036] = "WEAPON_ROSE_GARDEN_SHOTGUN";
    Items[Items["WEAPON_PATRIOTIC_AR"] = 2593] = "WEAPON_PATRIOTIC_AR";
    Items[Items["WEAPON_NINJA_AR"] = 3586] = "WEAPON_NINJA_AR";
    Items[Items["WEAPON_MERCENARY_AR"] = 2592] = "WEAPON_MERCENARY_AR";
    Items[Items["WEAPON_COMBAT_SHOTGUN"] = 2586] = "WEAPON_COMBAT_SHOTGUN";
    Items[Items["WEAPON_INFERNAL_SHOTGUN"] = 3720] = "WEAPON_INFERNAL_SHOTGUN";
    Items[Items["WEAPON_SHALLOW_M1911A1"] = 3550] = "WEAPON_SHALLOW_M1911A1";
    Items[Items["WEAPON_SHALLOW_SNIPER"] = 3584] = "WEAPON_SHALLOW_SNIPER";
    Items[Items["WEAPON_PREDATOR_SNIPER"] = 3172] = "WEAPON_PREDATOR_SNIPER";
    Items[Items["WEAPON_PREDATOR_SHOTGUN"] = 3165] = "WEAPON_PREDATOR_SHOTGUN";
    Items[Items["WEAPON_PREDATOR_AR"] = 3179] = "WEAPON_PREDATOR_AR";
    Items[Items["WEAPON_ANODIZED_AR"] = 2789] = "WEAPON_ANODIZED_AR";
    Items[Items["WEAPON_LIGHTNING_AR"] = 2834] = "WEAPON_LIGHTNING_AR";
    Items[Items["WEAPON_IREZUMI_AR"] = 3475] = "WEAPON_IREZUMI_AR";
    Items[Items["WEAPON_NEKO_CHAN_SHOTGUN"] = 3477] = "WEAPON_NEKO_CHAN_SHOTGUN";
    Items[Items["WEAPON_KUROMU_M9"] = 3381] = "WEAPON_KUROMU_M9";
    Items[Items["WEAPON_HEAVY_ASSAULT_MAGNUM"] = 3495] = "WEAPON_HEAVY_ASSAULT_MAGNUM";
    Items[Items["WEAPON_NEKO_CHAN_R380"] = 3489] = "WEAPON_NEKO_CHAN_R380";
    Items[Items["WEAPON_KUROMU_MAGNUM"] = 3481] = "WEAPON_KUROMU_MAGNUM";
    Items[Items["WEAPON_RAINBOW_SWIRL_SNIPER"] = 2909] = "WEAPON_RAINBOW_SWIRL_SNIPER";
    Items[Items["WEAPON_RAINBOW_SWIRL_AK47"] = 3055] = "WEAPON_RAINBOW_SWIRL_AK47";
    Items[Items["WEAPON_RAINBOW_SWIRL_AR"] = 3056] = "WEAPON_RAINBOW_SWIRL_AR";
    Items[Items["WEAPON_RAINBOW_SWIRL_SHOTGUN"] = 3057] = "WEAPON_RAINBOW_SWIRL_SHOTGUN";
    Items[Items["WEAPON_GRAFFITI_SNIPER"] = 2906] = "WEAPON_GRAFFITI_SNIPER";
    Items[Items["WEAPON_TIGER_BLOOD_SHOTGUN"] = 3123] = "WEAPON_TIGER_BLOOD_SHOTGUN";
    Items[Items["WEAPON_DESERT_WARFARE_AK"] = 3401] = "WEAPON_DESERT_WARFARE_AK";
    Items[Items["WEAPON_GRIMMMZ_AR"] = 3585] = "WEAPON_GRIMMMZ_AR";
    Items[Items["WEAPON_TOXIC_AK47"] = 2480] = "WEAPON_TOXIC_AK47";
    Items[Items["WEAPON_TOXIC_SHOTGUN"] = 2446] = "WEAPON_TOXIC_SHOTGUN";
    Items[Items["WEAPON_SHOWDOWN_AR"] = 2781] = "WEAPON_SHOWDOWN_AR";
    Items[Items["WEAPON_WILDSTYLE_AR"] = 2601] = "WEAPON_WILDSTYLE_AR";
    Items[Items["WEAPON_WILDSTYLE_AK47"] = 2600] = "WEAPON_WILDSTYLE_AK47";
    Items[Items["WEAPON_WILDSTYLE_SHOTGUN"] = 2597] = "WEAPON_WILDSTYLE_SHOTGUN";
    //#endregion
    //#region AMMO
    Items[Items["AMMO_223"] = 1429] = "AMMO_223";
    Items[Items["AMMO_12GA"] = 1511] = "AMMO_12GA";
    Items[Items["AMMO_45"] = 1428] = "AMMO_45";
    Items[Items["AMMO_9MM"] = 1998] = "AMMO_9MM";
    Items[Items["AMMO_308"] = 1469] = "AMMO_308";
    Items[Items["AMMO_380"] = 1992] = "AMMO_380";
    Items[Items["AMMO_762"] = 2325] = "AMMO_762";
    Items[Items["AMMO_44"] = 1719] = "AMMO_44";
    Items[Items["AMMO_ARROW"] = 112] = "AMMO_ARROW";
    Items[Items["AMMO_ARROW_EXPLOSIVE"] = 138] = "AMMO_ARROW_EXPLOSIVE";
    Items[Items["AMMO_ARROW_FLAMING"] = 1434] = "AMMO_ARROW_FLAMING";
    //#endregion
    //#region AMMO_BOX
    Items[Items["AMMO_BOX_223"] = 1430] = "AMMO_BOX_223";
    Items[Items["AMMO_BOX_45"] = 8] = "AMMO_BOX_45";
    //#endregion
    //#region PERISHABLE
    Items[Items["AIO_COLD_MEDICINE"] = 1742] = "AIO_COLD_MEDICINE";
    Items[Items["ANTIBIOTICS"] = 1388] = "ANTIBIOTICS";
    Items[Items["PAINKILLERS"] = 2509] = "PAINKILLERS";
    Items[Items["ADRENALINE_SHOT"] = 2512] = "ADRENALINE_SHOT";
    Items[Items["VITAMINS"] = 1512] = "VITAMINS";
    Items[Items["IMMUNITY_BOOSTERS"] = 1471] = "IMMUNITY_BOOSTERS";
    Items[Items["FIRST_AID"] = 78] = "FIRST_AID";
    Items[Items["BANDAGE"] = 24] = "BANDAGE";
    Items[Items["BANDAGE_DRESSED"] = 2214] = "BANDAGE_DRESSED";
    Items[Items["BUNDLE_GAUZE"] = 1750] = "BUNDLE_GAUZE";
    Items[Items["BUNDLE_WOODEN_ARROWS"] = 1782] = "BUNDLE_WOODEN_ARROWS";
    Items[Items["BUNDLE_WOODEN_ARROWS_1"] = 1892] = "BUNDLE_WOODEN_ARROWS_1";
    Items[Items["BUNDLE_EXPLOSIVE_ARROWS"] = 2400] = "BUNDLE_EXPLOSIVE_ARROWS";
    Items[Items["BUNDLE_FLAMING_ARROWS"] = 2401] = "BUNDLE_FLAMING_ARROWS";
    Items[Items["BUNDLE_WOODEN_ARROWS_2"] = 3379] = "BUNDLE_WOODEN_ARROWS_2";
    Items[Items["GAUZE"] = 1751] = "GAUZE";
    Items[Items["SWIZZLE"] = 1709] = "SWIZZLE";
    Items[Items["GROUND_COFFEE"] = 56] = "GROUND_COFFEE";
    Items[Items["CANNED_FOOD01"] = 7] = "CANNED_FOOD01";
    Items[Items["CANNED_FOOD02"] = 1393] = "CANNED_FOOD02";
    Items[Items["CANNED_FOOD03"] = 1394] = "CANNED_FOOD03";
    Items[Items["CANNED_FOOD04"] = 1391] = "CANNED_FOOD04";
    Items[Items["CANNED_FOOD05"] = 1392] = "CANNED_FOOD05";
    Items[Items["CANNED_FOOD06"] = 1396] = "CANNED_FOOD06";
    Items[Items["CANNED_FOOD07"] = 1397] = "CANNED_FOOD07";
    Items[Items["CANNED_FOOD08"] = 1398] = "CANNED_FOOD08";
    Items[Items["CANNED_FOOD09"] = 1399] = "CANNED_FOOD09";
    Items[Items["CANNED_FOOD10"] = 1400] = "CANNED_FOOD10";
    Items[Items["CANNED_FOOD11"] = 1401] = "CANNED_FOOD11";
    Items[Items["CANNED_FOOD25"] = 1431] = "CANNED_FOOD25";
    Items[Items["CANNED_FOOD21"] = 1395] = "CANNED_FOOD21";
    Items[Items["CANNED_FOOD26"] = 2536] = "CANNED_FOOD26";
    Items[Items["BLACKBERRY"] = 105] = "BLACKBERRY";
    Items[Items["BLACKBERRY_JUICE"] = 1361] = "BLACKBERRY_JUICE";
    Items[Items["BLACKBERRY_PIE"] = 1706] = "BLACKBERRY_PIE";
    Items[Items["BLACKBERRY_PIE_SLICE"] = 1726] = "BLACKBERRY_PIE_SLICE";
    Items[Items["BLACKBERRY_HANDFUL"] = 3214] = "BLACKBERRY_HANDFUL";
    Items[Items["WATER_PURE"] = 1371] = "WATER_PURE";
    Items[Items["WATER_STAGNANT"] = 1535] = "WATER_STAGNANT";
    Items[Items["WATER_DIRTY"] = 1368] = "WATER_DIRTY";
    Items[Items["MRE_APPLE"] = 1402] = "MRE_APPLE";
    Items[Items["MRE01"] = 1403] = "MRE01";
    Items[Items["MRE02"] = 1404] = "MRE02";
    Items[Items["MRE03"] = 1405] = "MRE03";
    Items[Items["MRE04"] = 1406] = "MRE04";
    Items[Items["MRE05"] = 1407] = "MRE05";
    Items[Items["MRE06"] = 1409] = "MRE06";
    Items[Items["MRE07"] = 1408] = "MRE07";
    Items[Items["MRE08"] = 1410] = "MRE08";
    Items[Items["MRE09"] = 1411] = "MRE09";
    Items[Items["MRE10"] = 1412] = "MRE10";
    Items[Items["MRE11"] = 1413] = "MRE11";
    Items[Items["MRE12"] = 1414] = "MRE12";
    Items[Items["SANDWICH_BEAR"] = 1459] = "SANDWICH_BEAR";
    Items[Items["STEAK_BEAR"] = 1451] = "STEAK_BEAR";
    Items[Items["SURVIVAL_BREAD"] = 1456] = "SURVIVAL_BREAD";
    Items[Items["MEAT_BEAR"] = 1450] = "MEAT_BEAR";
    Items[Items["COFFEE"] = 55] = "COFFEE";
    Items[Items["COFFEE_SUGAR"] = 71] = "COFFEE_SUGAR";
    Items[Items["STEAK_RABBIT"] = 117] = "STEAK_RABBIT";
    Items[Items["MEAT_RABBIT"] = 116] = "MEAT_RABBIT";
    Items[Items["SANDWICH_RABBIT"] = 1457] = "SANDWICH_RABBIT";
    Items[Items["STEW_RABBIT"] = 118] = "STEW_RABBIT";
    Items[Items["CORN"] = 107] = "CORN";
    Items[Items["CORN_ROASTED"] = 1387] = "CORN_ROASTED";
    Items[Items["JERKY_DEER"] = 21] = "JERKY_DEER";
    Items[Items["MEAT_VENISON"] = 20] = "MEAT_VENISON";
    Items[Items["SANDWICH_DEER"] = 1460] = "SANDWICH_DEER";
    Items[Items["STEAK_DEER"] = 61] = "STEAK_DEER";
    Items[Items["HONEY"] = 2192] = "HONEY";
    Items[Items["MOONSHINE"] = 1386] = "MOONSHINE";
    Items[Items["COLD_MEDICINE"] = 1742] = "COLD_MEDICINE";
    Items[Items["SURVIVAL_BORSCHT"] = 1379] = "SURVIVAL_BORSCHT";
    Items[Items["SANDWICH_WOLF"] = 1458] = "SANDWICH_WOLF";
    Items[Items["STEAK_WOLF"] = 1343] = "STEAK_WOLF";
    Items[Items["MEAT_WOLF"] = 1342] = "MEAT_WOLF";
    Items[Items["MEAT_ROTTEN"] = 1381] = "MEAT_ROTTEN";
    //#endregion
    //#region CONSTRUCTION
    Items[Items["GROUND_TAMPER"] = 124] = "GROUND_TAMPER";
    Items[Items["WOODEN_BARRICADE"] = 122] = "WOODEN_BARRICADE";
    Items[Items["SHACK"] = 1433] = "SHACK";
    Items[Items["SHACK_SMALL"] = 1440] = "SHACK_SMALL";
    Items[Items["SHACK_BASIC"] = 1468] = "SHACK_BASIC";
    Items[Items["SHELTER"] = 150] = "SHELTER";
    Items[Items["SHELTER_LARGE"] = 153] = "SHELTER_LARGE";
    Items[Items["SHELTER_UPPER_LARGE"] = 1897] = "SHELTER_UPPER_LARGE";
    Items[Items["SHELTER_UPPER"] = 1898] = "SHELTER_UPPER";
    Items[Items["FOUNDATION"] = 1378] = "FOUNDATION";
    Items[Items["FOUNDATION_EXPANSION"] = 2336] = "FOUNDATION_EXPANSION";
    Items[Items["FOUNDATION_RAMP"] = 2269] = "FOUNDATION_RAMP";
    Items[Items["FOUNDATION_STAIRS"] = 2270] = "FOUNDATION_STAIRS";
    Items[Items["METAL_GATE"] = 148] = "METAL_GATE";
    Items[Items["DOOR_METAL"] = 1881] = "DOOR_METAL";
    Items[Items["DOOR_WOOD"] = 1435] = "DOOR_WOOD";
    Items[Items["METAL_WALL"] = 149] = "METAL_WALL";
    Items[Items["METAL_WALL_UPPER"] = 1896] = "METAL_WALL_UPPER";
    Items[Items["DOOR_BASIC"] = 1470] = "DOOR_BASIC";
    Items[Items["LANDMINE"] = 74] = "LANDMINE";
    Items[Items["IED"] = 1699] = "IED";
    Items[Items["PUNJI_STICKS"] = 98] = "PUNJI_STICKS";
    Items[Items["PUNJI_STICK_ROW"] = 3580] = "PUNJI_STICK_ROW";
    Items[Items["SNARE"] = 1415] = "SNARE";
    Items[Items["ANIMAL_TRAP"] = 91] = "ANIMAL_TRAP";
    Items[Items["BARBED_WIRE"] = 108] = "BARBED_WIRE";
    Items[Items["BARBEQUE"] = 1447] = "BARBEQUE";
    Items[Items["BEE_BOX"] = 2034] = "BEE_BOX";
    Items[Items["CAMPFIRE"] = 15] = "CAMPFIRE";
    Items[Items["CANDLE"] = 1904] = "CANDLE";
    Items[Items["DEW_COLLECTOR"] = 97] = "DEW_COLLECTOR";
    Items[Items["FURNACE"] = 64] = "FURNACE";
    Items[Items["LOOKOUT_TOWER"] = 2272] = "LOOKOUT_TOWER";
    Items[Items["METAL_DOORWAY"] = 1969] = "METAL_DOORWAY";
    Items[Items["REPAIR_BOX"] = 2792] = "REPAIR_BOX";
    Items[Items["SLEEPING_MAT"] = 51] = "SLEEPING_MAT";
    Items[Items["STORAGE_BOX"] = 1982] = "STORAGE_BOX";
    Items[Items["STRUCTURE_STAIRS"] = 154] = "STRUCTURE_STAIRS";
    Items[Items["STRUCTURE_STAIRS_UPPER"] = 1900] = "STRUCTURE_STAIRS_UPPER";
    Items[Items["WORKBENCH"] = 1891] = "WORKBENCH";
    Items[Items["WORKBENCH_WEAPON"] = 3778] = "WORKBENCH_WEAPON";
    Items[Items["BARRICADE"] = 122] = "BARRICADE";
    //#endregion
    //#region WORN LETTERS
    Items[Items["WORN_LETTER_CHURCH_PV"] = 1752] = "WORN_LETTER_CHURCH_PV";
    Items[Items["WORN_LETTER_MISTY_DAM"] = 1753] = "WORN_LETTER_MISTY_DAM";
    Items[Items["WORN_LETTER_RUBY_LAKE"] = 1754] = "WORN_LETTER_RUBY_LAKE";
    Items[Items["WORN_LETTER_WATER_TOWER"] = 1755] = "WORN_LETTER_WATER_TOWER";
    Items[Items["WORN_LETTER_TOXIC_LAKE"] = 1756] = "WORN_LETTER_TOXIC_LAKE";
    Items[Items["WORN_LETTER_RADIO"] = 1757] = "WORN_LETTER_RADIO";
    Items[Items["WORN_LETTER_LJ_PV"] = 1758] = "WORN_LETTER_LJ_PV";
    Items[Items["WORN_LETTER_VILLAS"] = 1759] = "WORN_LETTER_VILLAS";
    Items[Items["CHEST_CHURCH_PV"] = 1760] = "CHEST_CHURCH_PV";
    Items[Items["CHEST_MISTY_DAM"] = 1761] = "CHEST_MISTY_DAM";
    Items[Items["CHEST_RUBY_LAKE"] = 1762] = "CHEST_RUBY_LAKE";
    Items[Items["CHEST_WATER_TOWER"] = 1763] = "CHEST_WATER_TOWER";
    Items[Items["CHEST_TOXIC_LAKE"] = 1764] = "CHEST_TOXIC_LAKE";
    Items[Items["CHEST_RADIO"] = 1765] = "CHEST_RADIO";
    Items[Items["CHEST_LJ_PV"] = 1766] = "CHEST_LJ_PV";
    Items[Items["CHEST_VILLAS"] = 1767] = "CHEST_VILLAS";
    //#endregion
    //#region COMPONENT
    Items[Items["SHARD_METAL"] = 114] = "SHARD_METAL";
    Items[Items["SHARD_BRASS"] = 3780] = "SHARD_BRASS";
    Items[Items["SHARD_PLASTIC"] = 3775] = "SHARD_PLASTIC";
    Items[Items["GUNPOWDER_REFINED"] = 3805] = "GUNPOWDER_REFINED";
    Items[Items["ALLOY_LEAD"] = 3779] = "ALLOY_LEAD";
    Items[Items["PROTOTYPE_MECHANISM"] = 3455] = "PROTOTYPE_MECHANISM";
    Items[Items["PROTOTYPE_TRIGGER_ASSEMBLY"] = 3456] = "PROTOTYPE_TRIGGER_ASSEMBLY";
    Items[Items["PROTOTYPE_RECEIVER"] = 3457] = "PROTOTYPE_RECEIVER";
    Items[Items["NAIL"] = 135] = "NAIL";
    Items[Items["BACKPACK_FRAME"] = 1466] = "BACKPACK_FRAME";
    Items[Items["METAL_BRACKET"] = 141] = "METAL_BRACKET";
    Items[Items["ANIMAL_FAT"] = 72] = "ANIMAL_FAT";
    Items[Items["FLOUR"] = 1455] = "FLOUR";
    Items[Items["WAX"] = 2193] = "WAX";
    Items[Items["HONEYCOMB"] = 2203] = "HONEYCOMB";
    Items[Items["PHONE_DEAD"] = 2635] = "PHONE_DEAD";
    Items[Items["PHONE_BATTERY"] = 2637] = "PHONE_BATTERY";
    Items[Items["PHONE_CHARGED"] = 2636] = "PHONE_CHARGED";
    Items[Items["CORN_MASH"] = 1385] = "CORN_MASH";
    Items[Items["YEAST"] = 1445] = "YEAST";
    Items[Items["DEER_SCENT"] = 1462] = "DEER_SCENT";
    Items[Items["DEER_BLADDER"] = 1463] = "DEER_BLADDER";
    Items[Items["DUCT_TAPE"] = 134] = "DUCT_TAPE";
    Items[Items["TRAP_IGNITION_KIT"] = 2831] = "TRAP_IGNITION_KIT";
    Items[Items["SALINE"] = 77] = "SALINE";
    Items[Items["WHEAT"] = 1438] = "WHEAT";
    Items[Items["TWINE"] = 142] = "TWINE";
    Items[Items["GUN_PART"] = 1890] = "GUN_PART";
    Items[Items["REPAIR_KIT_GUN"] = 1895] = "REPAIR_KIT_GUN";
    Items[Items["UPGRADE_KIT_GUN"] = 2419] = "UPGRADE_KIT_GUN";
    Items[Items["CHARCOAL"] = 26] = "CHARCOAL";
    Items[Items["METAL_BAR"] = 39] = "METAL_BAR";
    //#endregion
    //#region Quest Items
    Items[Items["ANTI_VIRAL_BOTTLE"] = 2671] = "ANTI_VIRAL_BOTTLE";
    Items[Items["HANDWRITTEN_NOTE_CAROLINE"] = 2611] = "HANDWRITTEN_NOTE_CAROLINE";
    Items[Items["ANTI_VIRAL_BOTTLE_EMPTY"] = 2612] = "ANTI_VIRAL_BOTTLE_EMPTY";
    Items[Items["GRENADE_SONIC_BROKEN"] = 3040] = "GRENADE_SONIC_BROKEN";
    Items[Items["VIAL_H1Z1_B_PLASMA"] = 3041] = "VIAL_H1Z1_B_PLASMA";
    Items[Items["VIAL_H1Z1_REDUCER"] = 2498] = "VIAL_H1Z1_REDUCER";
    Items[Items["BATTERIES_AA"] = 2833] = "BATTERIES_AA";
    Items[Items["BRAIN_TREATED"] = 2643] = "BRAIN_TREATED";
    Items[Items["BRAIN_INFECTED"] = 2642] = "BRAIN_INFECTED";
    Items[Items["SYRINGE_INFECTED_BLOOD"] = 1510] = "SYRINGE_INFECTED_BLOOD";
    Items[Items["EMPTY_SPECIMEN_BAG"] = 2641] = "EMPTY_SPECIMEN_BAG";
    Items[Items["DEAD_CELL_PHONE"] = 2635] = "DEAD_CELL_PHONE";
    Items[Items["WEAK_CELL_PHONE_BATTERY"] = 2637] = "WEAK_CELL_PHONE_BATTERY";
    Items[Items["CRACKED_CLIPBOARD"] = 2610] = "CRACKED_CLIPBOARD";
    Items[Items["DOCTORS_FILE"] = 2631] = "DOCTORS_FILE";
    Items[Items["WEICHS_WALLET"] = 2703] = "WEICHS_WALLET";
    Items[Items["WEICHS_REPORT"] = 2633] = "WEICHS_REPORT";
    Items[Items["KLAVISK_NOTE"] = 2632] = "KLAVISK_NOTE";
    Items[Items["REFRIGERATOR_NOTE"] = 3075] = "REFRIGERATOR_NOTE";
    Items[Items["DOCTORS_MEMO"] = 3076] = "DOCTORS_MEMO";
    Items[Items["TORN_LETTERHEAD"] = 2638] = "TORN_LETTERHEAD";
    Items[Items["SMALL_KEY"] = 2640] = "SMALL_KEY";
    Items[Items["CRUMPLED_NOTE"] = 2639] = "CRUMPLED_NOTE";
    //#endregion
    //#region Traps
    Items[Items["TRAP_FIRE"] = 2812] = "TRAP_FIRE";
    Items[Items["TRAP_FLASH"] = 2810] = "TRAP_FLASH";
    Items[Items["TRAP_GAS"] = 2811] = "TRAP_GAS";
    Items[Items["TRAP_SHOCK"] = 2832] = "TRAP_SHOCK";
    //#endregion
    //#region
    Items[Items["BACKPACK_FRAMED"] = 2111] = "BACKPACK_FRAMED";
    Items[Items["BACKPACK_SATCHEL"] = 1432] = "BACKPACK_SATCHEL";
    Items[Items["BACKPACK_MILITARY_TAN"] = 2124] = "BACKPACK_MILITARY_TAN";
    Items[Items["BACKPACK_MILITARY_GREEN_CAMO"] = 2121] = "BACKPACK_MILITARY_GREEN_CAMO";
    Items[Items["BACKPACK_BLUE_ORANGE"] = 2038] = "BACKPACK_BLUE_ORANGE";
    Items[Items["HELMET_MOTORCYCLE"] = 2170] = "HELMET_MOTORCYCLE";
    Items[Items["DEFAULT_TACTICAL_HELMET"] = 2172] = "DEFAULT_TACTICAL_HELMET";
    Items[Items["HAT_CAP"] = 2102] = "HAT_CAP";
    Items[Items["SHIRT_DEFAULT"] = 2088] = "SHIRT_DEFAULT";
    Items[Items["PANTS_DEFAULT"] = 2177] = "PANTS_DEFAULT";
    Items[Items["SHIRT_SCRUBS_BLUE"] = 2553] = "SHIRT_SCRUBS_BLUE";
    Items[Items["PANTS_SCRUBS_BLUE"] = 2557] = "PANTS_SCRUBS_BLUE";
    Items[Items["SURGEON_MASK_AQUA"] = 2569] = "SURGEON_MASK_AQUA";
    Items[Items["CAP_SCRUBS_BLUE"] = 2560] = "CAP_SCRUBS_BLUE";
    Items[Items["SHIRT_SCRUBS_GRAY"] = 2555] = "SHIRT_SCRUBS_GRAY";
    Items[Items["PANTS_SCRUBS_GRAY"] = 2558] = "PANTS_SCRUBS_GRAY";
    Items[Items["CAP_SCRUBS_GRAY"] = 2561] = "CAP_SCRUBS_GRAY";
    Items[Items["CONVEYS_RED"] = 2216] = "CONVEYS_RED";
    Items[Items["CONVEYS_BLUE"] = 2217] = "CONVEYS_BLUE";
    Items[Items["CONVEYS_WHITE"] = 2218] = "CONVEYS_WHITE";
    Items[Items["CONVEYS_DRAGON"] = 2398] = "CONVEYS_DRAGON";
    Items[Items["CONVEYS_CDN"] = 2795] = "CONVEYS_CDN";
    Items[Items["CONVEYS_NEKOCHAN"] = 3404] = "CONVEYS_NEKOCHAN";
    Items[Items["CONVEYS_ELLOHIME"] = 3574] = "CONVEYS_ELLOHIME";
    Items[Items["ZEDS_WHITE"] = 3708] = "ZEDS_WHITE";
    Items[Items["GATORS_RED"] = 2550] = "GATORS_RED";
    Items[Items["HAT_BEANIE"] = 2162] = "HAT_BEANIE";
    Items[Items["BOOTS_TAN"] = 2207] = "BOOTS_TAN";
    Items[Items["BOOTS_GRAY_BLUE"] = 2209] = "BOOTS_GRAY_BLUE";
    Items[Items["SUGAR"] = 57] = "SUGAR";
    Items[Items["BATTERY"] = 1696] = "BATTERY";
    Items[Items["SPARKPLUGS"] = 1701] = "SPARKPLUGS";
    Items[Items["SALT"] = 22] = "SALT";
    Items[Items["LIGHTER"] = 1436] = "LIGHTER";
    Items[Items["BOW_DRILL"] = 1452] = "BOW_DRILL";
    Items[Items["WATER_EMPTY"] = 1353] = "WATER_EMPTY";
    Items[Items["FUEL_BIOFUEL"] = 73] = "FUEL_BIOFUEL";
    Items[Items["FUEL_ETHANOL"] = 1384] = "FUEL_ETHANOL";
    Items[Items["WOOD_PLANK"] = 109] = "WOOD_PLANK";
    Items[Items["METAL_SHEET"] = 46] = "METAL_SHEET";
    Items[Items["METAL_SCRAP"] = 48] = "METAL_SCRAP";
    Items[Items["BROKEN_METAL_ITEM"] = 1354] = "BROKEN_METAL_ITEM";
    Items[Items["TARP"] = 155] = "TARP";
    Items[Items["WOOD_LOG"] = 16] = "WOOD_LOG";
    Items[Items["WOOD_STICK"] = 111] = "WOOD_STICK";
    Items[Items["GROUND_TILLER"] = 1383] = "GROUND_TILLER";
    Items[Items["FERTILIZER"] = 25] = "FERTILIZER";
    Items[Items["SEED_CORN"] = 1987] = "SEED_CORN";
    Items[Items["SEED_WHEAT"] = 1988] = "SEED_WHEAT";
    Items[Items["VIAL_EMPTY"] = 2510] = "VIAL_EMPTY";
    Items[Items["SYRINGE_EMPTY"] = 1508] = "SYRINGE_EMPTY";
    Items[Items["GHILLIE_SUIT"] = 92] = "GHILLIE_SUIT";
    Items[Items["GHILLIE_SUIT_TAN"] = 2570] = "GHILLIE_SUIT_TAN";
    Items[Items["HELMET_TACTICAL"] = 2172] = "HELMET_TACTICAL";
    Items[Items["RESPIRATOR"] = 2148] = "RESPIRATOR";
    Items[Items["NV_GOGGLES"] = 1700] = "NV_GOGGLES";
    Items[Items["ALL_PURPOSE_GOGGLES"] = 1693] = "ALL_PURPOSE_GOGGLES";
    Items[Items["GUNPOWDER"] = 11] = "GUNPOWDER";
    Items[Items["KEVLAR_DEFAULT"] = 2271] = "KEVLAR_DEFAULT";
    Items[Items["ARMOR_PLATED"] = 2205] = "ARMOR_PLATED";
    Items[Items["ARMOR_WOODEN"] = 2204] = "ARMOR_WOODEN";
    Items[Items["CLOTH"] = 23] = "CLOTH";
    Items[Items["METAL_PIPE"] = 47] = "METAL_PIPE";
    Items[Items["GRENADE_SMOKE"] = 2236] = "GRENADE_SMOKE";
    Items[Items["GRENADE_FLASH"] = 2235] = "GRENADE_FLASH";
    Items[Items["GRENADE_GAS"] = 2237] = "GRENADE_GAS";
    Items[Items["GRENADE_HE"] = 65] = "GRENADE_HE";
    Items[Items["GRENADE_SCREAM"] = 3022] = "GRENADE_SCREAM";
    Items[Items["MAP"] = 1985] = "MAP";
    Items[Items["COMPASS"] = 1441] = "COMPASS";
    Items[Items["FLARE"] = 1804] = "FLARE";
    Items[Items["FLARE_PARACHUTE"] = 1906] = "FLARE_PARACHUTE";
    Items[Items["FLARE_SMOKE"] = 1672] = "FLARE_SMOKE";
    Items[Items["BACKPACK_RASTA"] = 2393] = "BACKPACK_RASTA";
    Items[Items["WAIST_PACK"] = 1803] = "WAIST_PACK";
    Items[Items["BELT_POUCH"] = 1529] = "BELT_POUCH";
    Items[Items["FANNY_PACK_DEV"] = 1] = "FANNY_PACK_DEV";
    Items[Items["VEHICLE_KEY"] = 3460] = "VEHICLE_KEY";
    Items[Items["CODED_MESSAGE"] = 2722] = "CODED_MESSAGE";
    Items[Items["AIRDROP_CODE"] = 2675] = "AIRDROP_CODE";
    Items[Items["AIRDROP_TICKET"] = 1800] = "AIRDROP_TICKET";
    Items[Items["BANDANA_BASIC"] = 2323] = "BANDANA_BASIC";
    Items[Items["GLOVES_FINGERLESS"] = 2324] = "GLOVES_FINGERLESS";
    Items[Items["HAND_SHOVEL"] = 1697] = "HAND_SHOVEL";
    Items[Items["COMPASS_IMPROVISED"] = 1444] = "COMPASS_IMPROVISED";
    Items[Items["SKINNING_KNIFE"] = 110] = "SKINNING_KNIFE";
    Items[Items["RIGGED_LIGHT"] = 1748] = "RIGGED_LIGHT";
    Items[Items["SYRINGE_H1Z1_REDUCER"] = 1464] = "SYRINGE_H1Z1_REDUCER";
    Items[Items["LOCKER_KEY_F1"] = 2645] = "LOCKER_KEY_F1";
    Items[Items["LOCKER_KEY_F2"] = 2646] = "LOCKER_KEY_F2";
    Items[Items["LOCKER_KEY_F3"] = 2647] = "LOCKER_KEY_F3";
    Items[Items["LOCKER_KEY_F4"] = 2648] = "LOCKER_KEY_F4";
    Items[Items["EMERGENCY_RADIO"] = 2273] = "EMERGENCY_RADIO";
    Items[Items["GUN_REPAIR_KIT"] = 1895] = "GUN_REPAIR_KIT";
    Items[Items["WEAPON_REPAIR_KIT"] = 1467] = "WEAPON_REPAIR_KIT";
    Items[Items["HEADLIGHTS_OFFROADER"] = 9] = "HEADLIGHTS_OFFROADER";
    Items[Items["HEADLIGHTS_PICKUP"] = 1728] = "HEADLIGHTS_PICKUP";
    Items[Items["HEADLIGHTS_POLICE"] = 1730] = "HEADLIGHTS_POLICE";
    Items[Items["HEADLIGHTS_ATV"] = 2595] = "HEADLIGHTS_ATV";
    Items[Items["TURBO_OFFROADER"] = 90] = "TURBO_OFFROADER";
    Items[Items["TURBO_PICKUP"] = 1729] = "TURBO_PICKUP";
    Items[Items["TURBO_POLICE"] = 1731] = "TURBO_POLICE";
    Items[Items["TURBO_ATV"] = 2727] = "TURBO_ATV";
    Items[Items["VEHICLE_HOTWIRE"] = 3458] = "VEHICLE_HOTWIRE";
    Items[Items["VEHICLE_MOTOR_OFFROADER"] = 1344] = "VEHICLE_MOTOR_OFFROADER";
    Items[Items["VEHICLE_MOTOR_PICKUP"] = 1712] = "VEHICLE_MOTOR_PICKUP";
    Items[Items["VEHICLE_MOTOR_POLICECAR"] = 1722] = "VEHICLE_MOTOR_POLICECAR";
    Items[Items["VEHICLE_MOTOR_ATV"] = 2594] = "VEHICLE_MOTOR_ATV";
    Items[Items["VEHICLE_HORN"] = 1858] = "VEHICLE_HORN";
    Items[Items["VEHICLE_HORN_POLICECAR"] = 1735] = "VEHICLE_HORN_POLICECAR";
    Items[Items["VEHICLE_SIREN_POLICECAR"] = 1732] = "VEHICLE_SIREN_POLICECAR";
    Items[Items["CONTAINER_VEHICLE_OFFROADER"] = 1541] = "CONTAINER_VEHICLE_OFFROADER";
    Items[Items["CONTAINER_VEHICLE_PICKUP"] = 1783] = "CONTAINER_VEHICLE_PICKUP";
    Items[Items["CONTAINER_VEHICLE_POLICECAR"] = 1723] = "CONTAINER_VEHICLE_POLICECAR";
    Items[Items["CONTAINER_VEHICLE_ATV"] = 2728] = "CONTAINER_VEHICLE_ATV";
    Items[Items["CONTAINER_DROPPED_ITEMS"] = 3089] = "CONTAINER_DROPPED_ITEMS";
    Items[Items["CONTAINER_STORAGE"] = 1506] = "CONTAINER_STORAGE";
    Items[Items["CONTAINER_REPAIR_BOX"] = 2338] = "CONTAINER_REPAIR_BOX";
    Items[Items["CONTAINER_WRECKED_VAN"] = 1476] = "CONTAINER_WRECKED_VAN";
    Items[Items["CONTAINER_WRECKED_CAR"] = 1474] = "CONTAINER_WRECKED_CAR";
    Items[Items["CONTAINER_WRECKED_TRUCK"] = 1475] = "CONTAINER_WRECKED_TRUCK";
    Items[Items["CONTAINER_WEAPONS_LOCKER"] = 1705] = "CONTAINER_WEAPONS_LOCKER";
    Items[Items["CONTAINER_DESK"] = 1503] = "CONTAINER_DESK";
    Items[Items["CONTAINER_STASH"] = 1698] = "CONTAINER_STASH";
    Items[Items["CONTAINER_TOOL_CABINETS"] = 1507] = "CONTAINER_TOOL_CABINETS";
    Items[Items["CONTAINER_DUMPSTER"] = 1490] = "CONTAINER_DUMPSTER";
    Items[Items["CONTAINER_FILE_CABINET"] = 1504] = "CONTAINER_FILE_CABINET";
    Items[Items["CONTAINER_LOCKER"] = 1704] = "CONTAINER_LOCKER";
    Items[Items["CONTAINER_FRIDGE"] = 1472] = "CONTAINER_FRIDGE";
    Items[Items["CONTAINER_OTTOMAN"] = 1497] = "CONTAINER_OTTOMAN";
    Items[Items["CONTAINER_DRESSER"] = 1493] = "CONTAINER_DRESSER";
    Items[Items["CONTAINER_ARMOIRE"] = 1491] = "CONTAINER_ARMOIRE";
    // COMBINE THESE
    Items[Items["CONTAINER_CABINETS"] = 1480] = "CONTAINER_CABINETS";
    Items[Items["CONTAINER_CABINETS_BATHROOM"] = 1480] = "CONTAINER_CABINETS_BATHROOM";
    Items[Items["CONTAINER_CABINETS_CUBE"] = 1480] = "CONTAINER_CABINETS_CUBE";
    Items[Items["CONTAINER_CABINETS_KITCHEN"] = 1480] = "CONTAINER_CABINETS_KITCHEN";
    Items[Items["CONTAINER_GARBAGE_CAN"] = 1500] = "CONTAINER_GARBAGE_CAN";
    Items[Items["CONTAINER_FURNACE"] = 1711] = "CONTAINER_FURNACE";
    Items[Items["CONTAINER_BARBEQUE"] = 1477] = "CONTAINER_BARBEQUE";
    Items[Items["CONTAINER_CAMPFIRE"] = 1473] = "CONTAINER_CAMPFIRE";
    Items[Items["CONTAINER_DEW_COLLECTOR"] = 1540] = "CONTAINER_DEW_COLLECTOR";
    Items[Items["CONTAINER_ANIMAL_TRAP"] = 1539] = "CONTAINER_ANIMAL_TRAP";
    Items[Items["CONTAINER_BEE_BOX"] = 2035] = "CONTAINER_BEE_BOX";
    Items[Items["CONTAINER_DRUG_CABINET"] = 2525] = "CONTAINER_DRUG_CABINET";
    Items[Items["CONTAINER_MEDICAL_STATION"] = 2524] = "CONTAINER_MEDICAL_STATION";
    Items[Items["CONTAINER_HOSPITAL_DESK"] = 2677] = "CONTAINER_HOSPITAL_DESK";
    Items[Items["CONTAINER_GROSSING_STATION"] = 2529] = "CONTAINER_GROSSING_STATION";
    Items[Items["CONTAINER_HOSPITAL_REFRIGERATOR"] = 1472] = "CONTAINER_HOSPITAL_REFRIGERATOR";
    Items[Items["CONTAINER_HOSPITAL_CABINET"] = 2676] = "CONTAINER_HOSPITAL_CABINET";
    Items[Items["CONTAINER_MILITARY_CRATE"] = 1501] = "CONTAINER_MILITARY_CRATE";
    Items[Items["CONTAINER_LOOT_CACHE"] = 9069] = "CONTAINER_LOOT_CACHE";
    Items[Items["CONTAINER_WASHING_MACHINE"] = 1887] = "CONTAINER_WASHING_MACHINE";
    Items[Items["CONTAINER_DRYER"] = 1888] = "CONTAINER_DRYER";
    // Skin ref
    Items[Items["CAMO_TAN_FLANNEL_SHIRT"] = 2030] = "CAMO_TAN_FLANNEL_SHIRT";
    Items[Items["BLUE_FLANNEL_SHIRT"] = 2054] = "BLUE_FLANNEL_SHIRT";
    Items[Items["BROWN_FLANNEL_SHIRT"] = 2055] = "BROWN_FLANNEL_SHIRT";
    Items[Items["GREEN_FLANNEL_SHIRT"] = 2056] = "GREEN_FLANNEL_SHIRT";
    Items[Items["RED_FLANNEL_SHIRT"] = 2057] = "RED_FLANNEL_SHIRT";
    Items[Items["PAISLEY_FLANNEL_SHIRT"] = 2184] = "PAISLEY_FLANNEL_SHIRT";
    Items[Items["CAMO_GREEN_FLANNEL_SHIRT"] = 2187] = "CAMO_GREEN_FLANNEL_SHIRT";
    Items[Items["POLICE_SHIRT"] = 2346] = "POLICE_SHIRT";
    Items[Items["BASIC_HOODIE"] = 2373] = "BASIC_HOODIE";
    Items[Items["BLACK_ORANGE_HOODIE"] = 2374] = "BLACK_ORANGE_HOODIE";
    Items[Items["TWITCH_HOODIE"] = 2375] = "TWITCH_HOODIE";
    Items[Items["BLUE_BLACK_HOODIE"] = 2376] = "BLUE_BLACK_HOODIE";
    Items[Items["DOA_HOODIE"] = 2377] = "DOA_HOODIE";
    Items[Items["PRO_GRAMER_HOODIE"] = 2378] = "PRO_GRAMER_HOODIE";
    Items[Items["PARAMEDIC_UNIFORM"] = 2535] = "PARAMEDIC_UNIFORM";
    Items[Items["DRAGON_LODGE_PARKA"] = 3033] = "DRAGON_LODGE_PARKA";
    Items[Items["FLOWER_PRINT_PANTS"] = 2280] = "FLOWER_PRINT_PANTS";
    Items[Items["POLICE_SLACKS"] = 2347] = "POLICE_SLACKS";
    Items[Items["LEGGINGS"] = 2364] = "LEGGINGS";
    Items[Items["BONE_LEGGINGS"] = 2365] = "BONE_LEGGINGS";
    Items[Items["MUSCLE_LEGGINGS"] = 2366] = "MUSCLE_LEGGINGS";
    Items[Items["PARAMEDIC_SLACKS"] = 2534] = "PARAMEDIC_SLACKS";
    Items[Items["TWIN_GALAXIES_PANTS"] = 3875] = "TWIN_GALAXIES_PANTS";
    Items[Items["STARS_STRIPES_SHORTS"] = 2404] = "STARS_STRIPES_SHORTS";
    Items[Items["AVIATOR_CAP"] = 2058] = "AVIATOR_CAP";
    Items[Items["BLACK_BEANIE"] = 2162] = "BLACK_BEANIE";
    Items[Items["POLICE_HAT"] = 2344] = "POLICE_HAT";
    Items[Items["TOXIC_BODY_ARMOR"] = 2477] = "TOXIC_BODY_ARMOR";
    Items[Items["DEAD_END_ARMOR"] = 2860] = "DEAD_END_ARMOR";
    Items[Items["EZW_ARMORED_CHAMPIONSHIP_BELT"] = 3125] = "EZW_ARMORED_CHAMPIONSHIP_BELT";
    Items[Items["EZW_ARMORED_SHOULDERPADS"] = 3126] = "EZW_ARMORED_SHOULDERPADS";
    Items[Items["POLICE_BODY_ARMOR"] = 2274] = "POLICE_BODY_ARMOR";
    Items[Items["GORILLA_WARFARE"] = 3631] = "GORILLA_WARFARE";
    Items[Items["ARACHNID"] = 3721] = "ARACHNID";
    Items[Items["DEFAULT"] = 2170] = "DEFAULT";
    Items[Items["INFERNAL_DEMON_MASK"] = 3862] = "INFERNAL_DEMON_MASK";
    Items[Items["HOLIDAY_HAT"] = 2887] = "HOLIDAY_HAT";
    Items[Items["HOLIDAY_HAT_2"] = 2815] = "HOLIDAY_HAT_2";
    Items[Items["KRINGLE_HOLIDAY_HAT"] = 2818] = "KRINGLE_HOLIDAY_HAT";
    Items[Items["MASK_OF_JESTER"] = 3366] = "MASK_OF_JESTER";
    Items[Items["PATCHWORK_RUDOLPH_HELMET"] = 2817] = "PATCHWORK_RUDOLPH_HELMET";
    Items[Items["WASTELAND_SKULL_HELMET"] = 3064] = "WASTELAND_SKULL_HELMET";
    Items[Items["EZW_SPIKED_HELMET"] = 3124] = "EZW_SPIKED_HELMET";
    Items[Items["DRAGON_MOTORCYCLE_HELMET"] = 2828] = "DRAGON_MOTORCYCLE_HELMET";
    Items[Items["CIGAR_HOG_MASK"] = 2396] = "CIGAR_HOG_MASK";
    Items[Items["WHITE_UNICORN_FULL_HELMET"] = 3439] = "WHITE_UNICORN_FULL_HELMET";
    Items[Items["STAR_SPANGLED_HAT"] = 3470] = "STAR_SPANGLED_HAT";
    Items[Items["PUMPKIN_MASK"] = 2745] = "PUMPKIN_MASK";
    Items[Items["SCARECROW_MASK"] = 2750] = "SCARECROW_MASK";
    Items[Items["MILITARY_DEFAULT"] = 2124] = "MILITARY_DEFAULT";
    Items[Items["RASTA_BACKPACK"] = 2393] = "RASTA_BACKPACK";
    Items[Items["PINK_SKULLS_SURVIVOR_BACKPACK"] = 3046] = "PINK_SKULLS_SURVIVOR_BACKPACK";
    Items[Items["SNIPER_MILITARY_BACKPACK"] = 3169] = "SNIPER_MILITARY_BACKPACK";
    Items[Items["FULLY_GEARED_EXPLORER_BACKPACK"] = 4009] = "FULLY_GEARED_EXPLORER_BACKPACK";
    Items[Items["EVIL_CLOWN_GLASSES"] = 2253] = "EVIL_CLOWN_GLASSES";
    Items[Items["DEFAULT_GOGGLES"] = 1693] = "DEFAULT_GOGGLES";
    Items[Items["WHITE_BIKER_GLASSES"] = 2252] = "WHITE_BIKER_GLASSES";
    Items[Items["STEAMPUNK_GOGGLES"] = 2954] = "STEAMPUNK_GOGGLES";
    Items[Items["HAPPY_SKULL_SCRUBS_SHIRT"] = 2802] = "HAPPY_SKULL_SCRUBS_SHIRT";
    Items[Items["HAPPY_SKULL_SCRUBS_CAP"] = 2805] = "HAPPY_SKULL_SCRUBS_CAP";
    Items[Items["HAPPY_SKULL_SCRUBS_PANTS"] = 2808] = "HAPPY_SKULL_SCRUBS_PANTS";
    Items[Items["MILITARY_SCRUBS_SHIRT"] = 2801] = "MILITARY_SCRUBS_SHIRT";
    Items[Items["MILITARY_SCRUBS_CAP"] = 2804] = "MILITARY_SCRUBS_CAP";
    Items[Items["MILITARY_SCRUBS_PANTS"] = 2807] = "MILITARY_SCRUBS_PANTS";
    Items[Items["HUGZ_NEEDED_SCRUBS_SHIRT"] = 2803] = "HUGZ_NEEDED_SCRUBS_SHIRT";
    Items[Items["HUGZ_NEEDED_SCRUBS_CAP"] = 2806] = "HUGZ_NEEDED_SCRUBS_CAP";
    Items[Items["HUGZ_NEEDED_SCRUBS_PANTS"] = 2809] = "HUGZ_NEEDED_SCRUBS_PANTS";
    Items[Items["KURAMA_MEDICAL_SCRUBS_SHIRT"] = 2798] = "KURAMA_MEDICAL_SCRUBS_SHIRT";
    Items[Items["KURAMA_MEDICAL_SCRUBS_CAP"] = 2799] = "KURAMA_MEDICAL_SCRUBS_CAP";
    Items[Items["KURAMA_MEDICAL_SCRUBS_PANTS"] = 2800] = "KURAMA_MEDICAL_SCRUBS_PANTS";
    Items[Items["PINK_GATORS"] = 2567] = "PINK_GATORS";
    Items[Items["DOOMED_PUFFY_JACKET"] = 3977] = "DOOMED_PUFFY_JACKET";
    Items[Items["SILVER_METAL_FRAME_SUNGLASSES"] = 2255] = "SILVER_METAL_FRAME_SUNGLASSES";
    Items[Items["QUEST_KURAMA_MEDICAL_SCRUBS_SHIRT"] = 2649] = "QUEST_KURAMA_MEDICAL_SCRUBS_SHIRT";
    Items[Items["QUEST_KURAMA_MEDICAL_SCRUBS_CAP"] = 2650] = "QUEST_KURAMA_MEDICAL_SCRUBS_CAP";
    Items[Items["QUEST_KURAMA_MEDICAL_SCRUBS_PANTS"] = 2657] = "QUEST_KURAMA_MEDICAL_SCRUBS_PANTS";
    Items[Items["QUEST_HUGZ_NEEDED_SCRUBS_SHIRT"] = 2653] = "QUEST_HUGZ_NEEDED_SCRUBS_SHIRT";
    Items[Items["QUEST_HUGZ_NEEDED_SCRUBS_CAP"] = 2656] = "QUEST_HUGZ_NEEDED_SCRUBS_CAP";
    Items[Items["QUEST_HUGZ_NEEDED_SCRUBS_PANTS"] = 2660] = "QUEST_HUGZ_NEEDED_SCRUBS_PANTS";
    Items[Items["QUEST_HAPPY_SKULL_SCRUBS_SHIRT"] = 2651] = "QUEST_HAPPY_SKULL_SCRUBS_SHIRT";
    Items[Items["QUEST_HAPPY_SKULL_SCRUBS_PANTS"] = 2654] = "QUEST_HAPPY_SKULL_SCRUBS_PANTS";
    Items[Items["QUEST_HAPPY_SKULL_SCRUBS_CAP"] = 2658] = "QUEST_HAPPY_SKULL_SCRUBS_CAP";
    Items[Items["QUEST_MILITARY_SCRUBS_SHIRT"] = 2652] = "QUEST_MILITARY_SCRUBS_SHIRT";
    Items[Items["QUEST_MILITARY_SCRUBS_CAP"] = 2655] = "QUEST_MILITARY_SCRUBS_CAP";
    Items[Items["QUEST_MILITARY_SCRUBS_PANTS"] = 2659] = "QUEST_MILITARY_SCRUBS_PANTS";
    Items[Items["REWARD_SET_WOODLAND_GHILLIE"] = 2725] = "REWARD_SET_WOODLAND_GHILLIE";
    Items[Items["REWARD_SET_GHILLIE"] = 2724] = "REWARD_SET_GHILLIE";
    Items[Items["REWARD_SET_RED_FACE_BANDANA"] = 2326] = "REWARD_SET_RED_FACE_BANDANA";
    Items[Items["REWARD_SET_EVIL_CLOWN_BANDANA"] = 2327] = "REWARD_SET_EVIL_CLOWN_BANDANA";
    Items[Items["REWARD_SET_BLUE_FACE_BANDANA"] = 2328] = "REWARD_SET_BLUE_FACE_BANDANA";
    Items[Items["REWARD_SET_CAMO_FACE_BANDANA"] = 2329] = "REWARD_SET_CAMO_FACE_BANDANA";
    Items[Items["REWARD_SET_AMERICAN_FACE_BANDANA"] = 2330] = "REWARD_SET_AMERICAN_FACE_BANDANA";
    Items[Items["REWARD_SET_PINK_FACE_BANDANA"] = 2331] = "REWARD_SET_PINK_FACE_BANDANA";
    Items[Items["REWARD_SET_SKULL_FACE_BANDANA"] = 2332] = "REWARD_SET_SKULL_FACE_BANDANA";
    Items[Items["SKIN_WOODLAND_GHILLIE_SUIT_BOOTS"] = 2679] = "SKIN_WOODLAND_GHILLIE_SUIT_BOOTS";
    Items[Items["SKIN_GHILLIE_SUIT_BOOTS"] = 2680] = "SKIN_GHILLIE_SUIT_BOOTS";
    Items[Items["SKIN_GHILLIE_SUIT_GLOVES"] = 2670] = "SKIN_GHILLIE_SUIT_GLOVES";
    Items[Items["SKIN_GREEN_GLOVES"] = 2669] = "SKIN_GREEN_GLOVES";
    Items[Items["SKIN_RED_FACE_BANDANA"] = 2305] = "SKIN_RED_FACE_BANDANA";
    Items[Items["SKIN_RED_BIKER_SHADES"] = 2296] = "SKIN_RED_BIKER_SHADES";
    Items[Items["SKIN_EVIL_CLOWN_FACE_BANDANA"] = 2303] = "SKIN_EVIL_CLOWN_FACE_BANDANA";
    Items[Items["SKIN_EVIL_CLOWN_GLASSES"] = 2298] = "SKIN_EVIL_CLOWN_GLASSES";
    Items[Items["SKIN_BLUE_FACE_BANDANA"] = 2301] = "SKIN_BLUE_FACE_BANDANA";
    Items[Items["SKIN_BLUE_BIKER_SHADES"] = 2293] = "SKIN_BLUE_BIKER_SHADES";
    Items[Items["SKIN_CAMO_FACE_BANDANA"] = 2302] = "SKIN_CAMO_FACE_BANDANA";
    Items[Items["SKIN_GREEN_BIKER_SHADES"] = 2294] = "SKIN_GREEN_BIKER_SHADES";
    Items[Items["SKIN_AMERICAN_FACE_BANDANA"] = 2307] = "SKIN_AMERICAN_FACE_BANDANA";
    Items[Items["SKIN_WHITE_BIKER_SHADES"] = 2297] = "SKIN_WHITE_BIKER_SHADES";
    Items[Items["SKIN_PINK_FACE_BANDANA"] = 2304] = "SKIN_PINK_FACE_BANDANA";
    Items[Items["SKIN_PINK_BIKER_SHADES"] = 2295] = "SKIN_PINK_BIKER_SHADES";
    Items[Items["SKIN_SKULL_FACE_BANDANA"] = 2306] = "SKIN_SKULL_FACE_BANDANA";
    Items[Items["SKIN_BLACK_BIKER_SHADES"] = 2292] = "SKIN_BLACK_BIKER_SHADES";
    Items[Items["REWARD_CRATE_SHOWDOWN"] = 2761] = "REWARD_CRATE_SHOWDOWN";
    Items[Items["REWARD_CRATE_H1EMU"] = 5063] = "REWARD_CRATE_H1EMU";
    Items[Items["REWARD_CRATE_H1EMUEXCLUSIVE"] = 5026] = "REWARD_CRATE_H1EMUEXCLUSIVE";
    Items[Items["REWARD_CRATE_VICTORY"] = 5064] = "REWARD_CRATE_VICTORY";
    Items[Items["REWARD_CRATE_BREAKOUT"] = 5255] = "REWARD_CRATE_BREAKOUT";
    Items[Items["REWARD_CRATE_SWIRL"] = 5256] = "REWARD_CRATE_SWIRL";
    Items[Items["REWARD_CRATE_ALPHA_LAUNCH"] = 1797] = "REWARD_CRATE_ALPHA_LAUNCH";
    Items[Items["REWARD_CRATE_PREDATOR_GOLD"] = 3211] = "REWARD_CRATE_PREDATOR_GOLD";
    Items[Items["REWARD_CRATE_PREDATOR_SILVER"] = 3210] = "REWARD_CRATE_PREDATOR_SILVER";
    Items[Items["REWARD_CRATE_PREDATOR_BRONZE"] = 3209] = "REWARD_CRATE_PREDATOR_BRONZE";
    Items[Items["REWARD_CRATE_PREDATOR"] = 3207] = "REWARD_CRATE_PREDATOR";
    Items[Items["REWARD_CRATE_EZW_GOLD"] = 3122] = "REWARD_CRATE_EZW_GOLD";
    Items[Items["REWARD_CRATE_EZW_SILVER"] = 3121] = "REWARD_CRATE_EZW_SILVER";
    Items[Items["REWARD_CRATE_EZW_BRONZE"] = 3120] = "REWARD_CRATE_EZW_BRONZE";
    Items[Items["REWARD_CRATE_EZW"] = 3118] = "REWARD_CRATE_EZW";
    Items[Items["REWARD_CRATE_RENEGADE_GOLD"] = 2893] = "REWARD_CRATE_RENEGADE_GOLD";
    Items[Items["REWARD_CRATE_RENEGADE_SILVER"] = 2892] = "REWARD_CRATE_RENEGADE_SILVER";
    Items[Items["REWARD_CRATE_RENEGADE_BRONZE"] = 2891] = "REWARD_CRATE_RENEGADE_BRONZE";
    Items[Items["REWARD_CRATE_RENEGADE"] = 2836] = "REWARD_CRATE_RENEGADE";
    Items[Items["REWARD_CRATE_WASTELAND_GOLD"] = 3028] = "REWARD_CRATE_WASTELAND_GOLD";
    Items[Items["REWARD_CRATE_WASTELAND_SILVER"] = 3027] = "REWARD_CRATE_WASTELAND_SILVER";
    Items[Items["REWARD_CRATE_WASTELAND_BRONZE"] = 3026] = "REWARD_CRATE_WASTELAND_BRONZE";
    Items[Items["REWARD_CRATE_WASTELAND"] = 2939] = "REWARD_CRATE_WASTELAND";
    Items[Items["REWARD_CRATE_BR_1"] = 2031] = "REWARD_CRATE_BR_1";
    Items[Items["REWARD_CRATE_BR_2"] = 2350] = "REWARD_CRATE_BR_2";
    Items[Items["REWARD_CRATE_INVITATIONAL"] = 2436] = "REWARD_CRATE_INVITATIONAL";
    Items[Items["REWARD_CRATE_WEARABLES"] = 2009] = "REWARD_CRATE_WEARABLES";
    Items[Items["REWARD_CRATE_MARAUDER"] = 2276] = "REWARD_CRATE_MARAUDER";
    Items[Items["REWARD_CRATE_MERCENARY"] = 2432] = "REWARD_CRATE_MERCENARY";
    Items[Items["REWARD_CRATE_RONIN"] = 3501] = "REWARD_CRATE_RONIN";
    Items[Items["REWARD_CRATE_INVITATIONAL_2016_GOLD"] = 3573] = "REWARD_CRATE_INVITATIONAL_2016_GOLD";
    Items[Items["REWARD_CRATE_INVITATIONAL_2016_SILVER"] = 3572] = "REWARD_CRATE_INVITATIONAL_2016_SILVER";
    Items[Items["REWARD_CRATE_INVITATIONAL_2016_BRONZE"] = 3571] = "REWARD_CRATE_INVITATIONAL_2016_BRONZE";
    Items[Items["REWARD_CRATE_INVITATIONAL_2016"] = 3569] = "REWARD_CRATE_INVITATIONAL_2016";
    Items[Items["REWARD_CRATE_INFERNAL_GOLD"] = 3824] = "REWARD_CRATE_INFERNAL_GOLD";
    Items[Items["REWARD_CRATE_INFERNAL_SILVER"] = 3823] = "REWARD_CRATE_INFERNAL_SILVER";
    Items[Items["REWARD_CRATE_INFERNAL_BRONZE"] = 3822] = "REWARD_CRATE_INFERNAL_BRONZE";
    Items[Items["REWARD_CRATE_INFERNAL"] = 3821] = "REWARD_CRATE_INFERNAL";
    Items[Items["REWARD_HAPPY_SKULL_SCRUBS_PANTS"] = 2740] = "REWARD_HAPPY_SKULL_SCRUBS_PANTS";
    Items[Items["REWARD_HAPPY_SKULL_SCRUBS_SHIRT"] = 2741] = "REWARD_HAPPY_SKULL_SCRUBS_SHIRT";
    Items[Items["REWARD_HAPPY_SKULL_SCRUBS_CAP"] = 2739] = "REWARD_HAPPY_SKULL_SCRUBS_CAP";
    Items[Items["REWARD_MILITARY_SCRUBS_SHIRT"] = 2742] = "REWARD_MILITARY_SCRUBS_SHIRT";
    Items[Items["REWARD_MILITARY_SCRUBS_CAP"] = 2743] = "REWARD_MILITARY_SCRUBS_CAP";
    Items[Items["REWARD_MILITARY_SCRUBS_PANTS"] = 2744] = "REWARD_MILITARY_SCRUBS_PANTS";
    Items[Items["REWARD_HUGZ_NEEDED_SCRUBS_SHIRT"] = 2733] = "REWARD_HUGZ_NEEDED_SCRUBS_SHIRT";
    Items[Items["REWARD_HUGZ_NEEDED_SCRUBS_CAP"] = 2735] = "REWARD_HUGZ_NEEDED_SCRUBS_CAP";
    Items[Items["REWARD_HUGZ_NEEDED_SCRUBS_PANTS"] = 2734] = "REWARD_HUGZ_NEEDED_SCRUBS_PANTS";
    Items[Items["REWARD_KURAMA_MEDICAL_SCRUBS_SHIRT"] = 2736] = "REWARD_KURAMA_MEDICAL_SCRUBS_SHIRT";
    Items[Items["REWARD_KURAMA_MEDICAL_SCRUBS_CAP"] = 2738] = "REWARD_KURAMA_MEDICAL_SCRUBS_CAP";
    Items[Items["REWARD_KURAMA_MEDICAL_SCRUBS_PANTS"] = 2737] = "REWARD_KURAMA_MEDICAL_SCRUBS_PANTS";
    Items[Items["MYSTERY_BAG_KOTK"] = 3239] = "MYSTERY_BAG_KOTK";
    Items[Items["KOTK_ELITE_BAG"] = 3240] = "KOTK_ELITE_BAG";
    Items[Items["SKIRMISH_SMALL_FIREARMS"] = 3274] = "SKIRMISH_SMALL_FIREARMS";
    Items[Items["SKIRMISH_SHOTTIES_SNIPERS"] = 3365] = "SKIRMISH_SHOTTIES_SNIPERS";
    Items[Items["SKIRMISH_IRONMAN"] = 3447] = "SKIRMISH_IRONMAN";
    Items[Items["SKIRMISH_REANIMATED"] = 3523] = "SKIRMISH_REANIMATED";
    Items[Items["SKIRMISH_SHOTTIES_SNIPERS_2"] = 3633] = "SKIRMISH_SHOTTIES_SNIPERS_2";
    Items[Items["MYSTERY_BAG_1"] = 1840] = "MYSTERY_BAG_1";
    Items[Items["ELITE_BAG"] = 1841] = "ELITE_BAG";
    Items[Items["MYSTERY_BAG_2"] = 1879] = "MYSTERY_BAG_2";
    Items[Items["ELITE_BAG_HARDCORE"] = 1880] = "ELITE_BAG_HARDCORE";
    Items[Items["MYSTERY_BAG_V2"] = 2033] = "MYSTERY_BAG_V2";
    Items[Items["ACCOUNT_ITEM_BR_CROWN"] = 3273] = "ACCOUNT_ITEM_BR_CROWN";
})(Items || (exports.Items = Items = {}));
var WeaponDefinitionIds;
(function (WeaponDefinitionIds) {
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_BOW_MAKESHIFT"] = 113] = "WEAPON_BOW_MAKESHIFT";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_BOW_RECURVE"] = 1399] = "WEAPON_BOW_RECURVE";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_CROSSBOW"] = 1414] = "WEAPON_CROSSBOW";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_BOW_WOOD"] = 1387] = "WEAPON_BOW_WOOD";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_SHOTGUN"] = 1374] = "WEAPON_SHOTGUN";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_308"] = 1373] = "WEAPON_308";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_NAGAFENS_RAGE"] = 1450] = "WEAPON_NAGAFENS_RAGE";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_REAPER"] = 1452] = "WEAPON_REAPER";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_FROSTBITE"] = 1448] = "WEAPON_FROSTBITE";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_BLAZE"] = 1449] = "WEAPON_BLAZE";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_PURGE"] = 1451] = "WEAPON_PURGE";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_AK47"] = 1405] = "WEAPON_AK47";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_AR15"] = 6] = "WEAPON_AR15";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_MAGNUM"] = 1388] = "WEAPON_MAGNUM";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_R380"] = 1400] = "WEAPON_R380";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_M9"] = 1401] = "WEAPON_M9";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_1911"] = 2] = "WEAPON_1911";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_CROWBAR"] = 9] = "WEAPON_CROWBAR";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_HAMMER"] = 1396] = "WEAPON_HAMMER";
    WeaponDefinitionIds[WeaponDefinitionIds["WEAPON_WRENCH"] = 1415] = "WEAPON_WRENCH";
})(WeaponDefinitionIds || (exports.WeaponDefinitionIds = WeaponDefinitionIds = {}));
var Effects;
(function (Effects) {
    Effects[Effects["Smoke_Orange"] = 3621] = "Smoke_Orange";
    Effects[Effects["Smoke_Green"] = 3622] = "Smoke_Green";
    Effects[Effects["Smoke_Purple"] = 3623] = "Smoke_Purple";
    Effects[Effects["Smoke_Yellow"] = 3624] = "Smoke_Yellow";
    Effects[Effects["Smoke_LimeGreen"] = 4537] = "Smoke_LimeGreen";
    Effects[Effects["Smoke_Red"] = 4538] = "Smoke_Red";
    Effects[Effects["Smoke_Blue"] = 4539] = "Smoke_Blue";
    Effects[Effects["Smoke_Teal"] = 4540] = "Smoke_Teal";
    Effects[Effects["Smoke_Magenta"] = 4541] = "Smoke_Magenta";
    Effects[Effects["TEST_XYZ_loop"] = 1] = "TEST_XYZ_loop";
    Effects[Effects["TEST_XYZ"] = 2] = "TEST_XYZ";
    Effects[Effects["TEST_XYZ_Mesh"] = 3] = "TEST_XYZ_Mesh";
    Effects[Effects["TEST_MipMaps"] = 4] = "TEST_MipMaps";
    Effects[Effects["TEST_Mesh_Sphere"] = 5] = "TEST_Mesh_Sphere";
    Effects[Effects["TEST_Mesh_Hemisphere"] = 6] = "TEST_Mesh_Hemisphere";
    Effects[Effects["TEST_NormalMap_Sphere"] = 7] = "TEST_NormalMap_Sphere";
    Effects[Effects["TEST_Null"] = 8] = "TEST_Null";
    Effects[Effects["TEST_Irradience_Spheres"] = 9] = "TEST_Irradience_Spheres";
    Effects[Effects["TEST_Number1"] = 21] = "TEST_Number1";
    Effects[Effects["TEST_Number2"] = 22] = "TEST_Number2";
    Effects[Effects["TEST_Number3"] = 23] = "TEST_Number3";
    Effects[Effects["TEST_Number4"] = 24] = "TEST_Number4";
    Effects[Effects["TEST_Number5"] = 25] = "TEST_Number5";
    Effects[Effects["TEST_Number6"] = 26] = "TEST_Number6";
    Effects[Effects["TEST_Number7"] = 27] = "TEST_Number7";
    Effects[Effects["TEST_Number8"] = 28] = "TEST_Number8";
    Effects[Effects["EFX_Blowing_SandLeaves_Tan_Flat"] = 96] = "EFX_Blowing_SandLeaves_Tan_Flat";
    Effects[Effects["EFX_Blowing_SandLeaves_Tan"] = 97] = "EFX_Blowing_SandLeaves_Tan";
    Effects[Effects["PFX_Eyes_Red"] = 98] = "PFX_Eyes_Red";
    Effects[Effects["PFX_Impact_BulletLight_Default"] = 99] = "PFX_Impact_BulletLight_Default";
    Effects[Effects["PFX_Mzl_PistolBase"] = 100] = "PFX_Mzl_PistolBase";
    Effects[Effects["PFX_Mzl_PistolBase_IronSight"] = 101] = "PFX_Mzl_PistolBase_IronSight";
    Effects[Effects["PFX_Mzl_Magnum"] = 102] = "PFX_Mzl_Magnum";
    Effects[Effects["PFX_Mzl_Magnum_IronSight"] = 103] = "PFX_Mzl_Magnum_IronSight";
    Effects[Effects["PFX_Mzl_Shotgun_PumpAction"] = 105] = "PFX_Mzl_Shotgun_PumpAction";
    Effects[Effects["PFX_Mzl_Shotgun_PumpAction_Ironsight"] = 106] = "PFX_Mzl_Shotgun_PumpAction_Ironsight";
    Effects[Effects["VEH_Death_OffRoader"] = 135] = "VEH_Death_OffRoader";
    Effects[Effects["EFX_SnowFalling_01"] = 150] = "EFX_SnowFalling_01";
    Effects[Effects["EFX_SmokePlume_01"] = 156] = "EFX_SmokePlume_01";
    Effects[Effects["EFX_Ash_Area"] = 160] = "EFX_Ash_Area";
    Effects[Effects["EFX_Dust_Floating_Area"] = 184] = "EFX_Dust_Floating_Area";
    Effects[Effects["EFX_Candle_Flame_01"] = 185] = "EFX_Candle_Flame_01";
    Effects[Effects["EFX_Candle_Flame_02"] = 186] = "EFX_Candle_Flame_02";
    Effects[Effects["EFX_Candle_Flame_03"] = 187] = "EFX_Candle_Flame_03";
    Effects[Effects["EFX_Bees_BeeHiveBox"] = 188] = "EFX_Bees_BeeHiveBox";
    Effects[Effects["PFX_Bee_Swarm_Attack"] = 189] = "PFX_Bee_Swarm_Attack";
    Effects[Effects["PFX_Fire_Molotov"] = 190] = "PFX_Fire_Molotov";
    Effects[Effects["MAT_Molotov"] = 191] = "MAT_Molotov";
    Effects[Effects["PFX_Fire_Molotov_Persist"] = 192] = "PFX_Fire_Molotov_Persist";
    Effects[Effects["PFX_Damage_Crate_01m"] = 163] = "PFX_Damage_Crate_01m";
    Effects[Effects["PFX_Damage_Sign_Stop"] = 164] = "PFX_Damage_Sign_Stop";
    Effects[Effects["PFX_Damage_GlassWindow_House"] = 165] = "PFX_Damage_GlassWindow_House";
    Effects[Effects["PFX_Impact_BulletLight_Asphalt"] = 168] = "PFX_Impact_BulletLight_Asphalt";
    Effects[Effects["PFX_Fire_Torch"] = 169] = "PFX_Fire_Torch";
    Effects[Effects["PFX_Fire_Torch_Blue"] = 183] = "PFX_Fire_Torch_Blue";
    Effects[Effects["PFX_Flare"] = 170] = "PFX_Flare";
    Effects[Effects["MAT_Engine_Idle_OffRoader"] = 171] = "MAT_Engine_Idle_OffRoader";
    Effects[Effects["VEH_Engine_Idle_OffRoader"] = 193] = "VEH_Engine_Idle_OffRoader";
    Effects[Effects["VEH_Engine_Thrust_OffRoader"] = 172] = "VEH_Engine_Thrust_OffRoader";
    Effects[Effects["VEH_Headlight_OffRoader"] = 173] = "VEH_Headlight_OffRoader";
    Effects[Effects["VEH_Lights_CargoContainer"] = 271] = "VEH_Lights_CargoContainer";
    Effects[Effects["VEH_Lights_Crate"] = 272] = "VEH_Lights_Crate";
    Effects[Effects["VEH_Headlight_OffRoader_wShadows"] = 273] = "VEH_Headlight_OffRoader_wShadows";
    Effects[Effects["VEH_Lights_C_130"] = 274] = "VEH_Lights_C_130";
    Effects[Effects["VEH_SirenLight_PoliceCar"] = 275] = "VEH_SirenLight_PoliceCar";
    Effects[Effects["VEH_Brakes_OffRoader"] = 174] = "VEH_Brakes_OffRoader";
    Effects[Effects["VEH_Dust_OffRoader_Generic"] = 175] = "VEH_Dust_OffRoader_Generic";
    Effects[Effects["VEH_Dust_OffRoader_Dirt"] = 176] = "VEH_Dust_OffRoader_Dirt";
    Effects[Effects["VEH_Dust_OffRoader_Grass"] = 177] = "VEH_Dust_OffRoader_Grass";
    Effects[Effects["PFX_Mzl_AssaultRifle"] = 178] = "PFX_Mzl_AssaultRifle";
    Effects[Effects["PFX_Mzl_AssaultRifle_Ironsight"] = 179] = "PFX_Mzl_AssaultRifle_Ironsight";
    Effects[Effects["VEH_Damage_OffRoader_Stage03"] = 180] = "VEH_Damage_OffRoader_Stage03";
    Effects[Effects["VEH_Damage_OffRoader_Stage02"] = 181] = "VEH_Damage_OffRoader_Stage02";
    Effects[Effects["VEH_Damage_OffRoader_Stage01"] = 182] = "VEH_Damage_OffRoader_Stage01";
    Effects[Effects["VEH_Engine_Idle_PoliceCar"] = 277] = "VEH_Engine_Idle_PoliceCar";
    Effects[Effects["VEH_Engine_Thrust_PoliceCar"] = 278] = "VEH_Engine_Thrust_PoliceCar";
    Effects[Effects["VEH_Engine_Boost_PoliceCar"] = 279] = "VEH_Engine_Boost_PoliceCar";
    Effects[Effects["VEH_Headlight_PoliceCar"] = 280] = "VEH_Headlight_PoliceCar";
    Effects[Effects["VEH_Headlight_PoliceCar_wShadows"] = 281] = "VEH_Headlight_PoliceCar_wShadows";
    Effects[Effects["VEH_Brakes_PoliceCar"] = 282] = "VEH_Brakes_PoliceCar";
    Effects[Effects["VEH_Damage_PoliceCar_Stage03"] = 283] = "VEH_Damage_PoliceCar_Stage03";
    Effects[Effects["VEH_Damage_PoliceCar_Stage02"] = 284] = "VEH_Damage_PoliceCar_Stage02";
    Effects[Effects["VEH_Damage_PoliceCar_Stage01"] = 285] = "VEH_Damage_PoliceCar_Stage01";
    Effects[Effects["VEH_Death_PoliceCar"] = 286] = "VEH_Death_PoliceCar";
    Effects[Effects["VEH_Dust_PoliceCar_Generic"] = 287] = "VEH_Dust_PoliceCar_Generic";
    Effects[Effects["VEH_Dust_PoliceCar_Dirt"] = 288] = "VEH_Dust_PoliceCar_Dirt";
    Effects[Effects["VEH_Dust_PoliceCar_Grass"] = 289] = "VEH_Dust_PoliceCar_Grass";
    Effects[Effects["VEH_Engine_Idle_PickupTruck"] = 317] = "VEH_Engine_Idle_PickupTruck";
    Effects[Effects["VEH_Engine_Thrust_PickupTruck"] = 318] = "VEH_Engine_Thrust_PickupTruck";
    Effects[Effects["VEH_Engine_Boost_PickupTruck"] = 319] = "VEH_Engine_Boost_PickupTruck";
    Effects[Effects["VEH_Headlight_PickupTruck"] = 320] = "VEH_Headlight_PickupTruck";
    Effects[Effects["VEH_Headlight_PickupTruck_wShadows"] = 321] = "VEH_Headlight_PickupTruck_wShadows";
    Effects[Effects["VEH_Brakes_PickupTruck"] = 322] = "VEH_Brakes_PickupTruck";
    Effects[Effects["VEH_Damage_PickupTruck_Stage03"] = 323] = "VEH_Damage_PickupTruck_Stage03";
    Effects[Effects["VEH_Damage_PickupTruck_Stage02"] = 324] = "VEH_Damage_PickupTruck_Stage02";
    Effects[Effects["VEH_Damage_PickupTruck_Stage01"] = 325] = "VEH_Damage_PickupTruck_Stage01";
    Effects[Effects["VEH_Death_PickupTruck"] = 326] = "VEH_Death_PickupTruck";
    Effects[Effects["VEH_Dust_PickupTruck_Generic"] = 327] = "VEH_Dust_PickupTruck_Generic";
    Effects[Effects["VEH_Dust_PickupTruck_Dirt"] = 328] = "VEH_Dust_PickupTruck_Dirt";
    Effects[Effects["VEH_Dust_PickupTruck_Grass"] = 329] = "VEH_Dust_PickupTruck_Grass";
    Effects[Effects["PFX_Damage_Barricade01"] = 241] = "PFX_Damage_Barricade01";
    Effects[Effects["PFX_Death_Barricade01"] = 242] = "PFX_Death_Barricade01";
    Effects[Effects["MAT_Impact_OffRoader"] = 300] = "MAT_Impact_OffRoader";
    Effects[Effects["PFX_Impact_OffRoader_Generic"] = 301] = "PFX_Impact_OffRoader_Generic";
    Effects[Effects["PFX_Impact_OffRoader_Flesh"] = 302] = "PFX_Impact_OffRoader_Flesh";
    Effects[Effects["PFX_Impact_OffRoader_Zombie"] = 303] = "PFX_Impact_OffRoader_Zombie";
    Effects[Effects["PFX_Trail_OffRoader_Zombie"] = 304] = "PFX_Trail_OffRoader_Zombie";
    Effects[Effects["MAT_Engine_Idle_ATV"] = 351] = "MAT_Engine_Idle_ATV";
    Effects[Effects["VEH_Engine_Idle_ATV"] = 352] = "VEH_Engine_Idle_ATV";
    Effects[Effects["VEH_Engine_Thrust_ATV"] = 353] = "VEH_Engine_Thrust_ATV";
    Effects[Effects["VEH_Engine_Boost_ATV"] = 354] = "VEH_Engine_Boost_ATV";
    Effects[Effects["VEH_Headlight_ATV_wShadows"] = 355] = "VEH_Headlight_ATV_wShadows";
    Effects[Effects["VEH_Brakes_ATV"] = 356] = "VEH_Brakes_ATV";
    Effects[Effects["VEH_Death_ATV"] = 357] = "VEH_Death_ATV";
    Effects[Effects["VEH_Damage_ATV_Stage03"] = 358] = "VEH_Damage_ATV_Stage03";
    Effects[Effects["VEH_Damage_ATV_Stage02"] = 359] = "VEH_Damage_ATV_Stage02";
    Effects[Effects["VEH_Damage_ATV_Stage01"] = 360] = "VEH_Damage_ATV_Stage01";
    Effects[Effects["MAT_WheelDust_ATV"] = 361] = "MAT_WheelDust_ATV";
    Effects[Effects["VEH_Dust_ATV_Generic"] = 362] = "VEH_Dust_ATV_Generic";
    Effects[Effects["VEH_Dust_ATV_Dirt"] = 363] = "VEH_Dust_ATV_Dirt";
    Effects[Effects["VEH_Dust_ATV_Grass"] = 364] = "VEH_Dust_ATV_Grass";
    Effects[Effects["VEH_Dust_ATV_Water"] = 365] = "VEH_Dust_ATV_Water";
    Effects[Effects["VEH_Dust_ATV_Water_Idle"] = 366] = "VEH_Dust_ATV_Water_Idle";
    Effects[Effects["MAT_Impact_ATV"] = 367] = "MAT_Impact_ATV";
    Effects[Effects["PFX_Trail_Projectile_Arrow_Regular"] = 500] = "PFX_Trail_Projectile_Arrow_Regular";
    Effects[Effects["PFX_Trail_Projectile_Bullet_Small_1P"] = 501] = "PFX_Trail_Projectile_Bullet_Small_1P";
    Effects[Effects["PFX_Trail_Projectile_Arrow_Flaming"] = 502] = "PFX_Trail_Projectile_Arrow_Flaming";
    Effects[Effects["PFX_Impact_TreeLeaves_Hazel"] = 701] = "PFX_Impact_TreeLeaves_Hazel";
    Effects[Effects["PFX_Impact_TreeLeaves_Hazel_small"] = 702] = "PFX_Impact_TreeLeaves_Hazel_small";
    Effects[Effects["PFX_Impact_TreeLeaves_Blackberry"] = 703] = "PFX_Impact_TreeLeaves_Blackberry";
    Effects[Effects["PFX_Impact_TreeLeaves_DevilClub"] = 704] = "PFX_Impact_TreeLeaves_DevilClub";
    Effects[Effects["PFX_Impact_TreeLeaves_VineMaple"] = 705] = "PFX_Impact_TreeLeaves_VineMaple";
    Effects[Effects["MAT_BulletLight"] = 1000] = "MAT_BulletLight";
    Effects[Effects["MAT_BulletHeavy"] = 1007] = "MAT_BulletHeavy";
    Effects[Effects["MAT_BulletShotgun"] = 1006] = "MAT_BulletShotgun";
    Effects[Effects["MAT_Blade"] = 1001] = "MAT_Blade";
    Effects[Effects["MAT_Blade_ForehandSlash"] = 1011] = "MAT_Blade_ForehandSlash";
    Effects[Effects["MAT_Blade_BackhandSlash"] = 1012] = "MAT_Blade_BackhandSlash";
    Effects[Effects["MAT_Blade_ForehandChop"] = 1013] = "MAT_Blade_ForehandChop";
    Effects[Effects["MAT_Blade_BackhandChop"] = 1014] = "MAT_Blade_BackhandChop";
    Effects[Effects["MAT_Blade_ForehandUppercut"] = 1015] = "MAT_Blade_ForehandUppercut";
    Effects[Effects["MAT_Blade_ForwardStab"] = 5115] = "MAT_Blade_ForwardStab";
    Effects[Effects["MAT_Guitar"] = 1016] = "MAT_Guitar";
    Effects[Effects["MAT_Guitar_ForehandSlash"] = 1021] = "MAT_Guitar_ForehandSlash";
    Effects[Effects["MAT_Guitar_BackhandSlash"] = 1022] = "MAT_Guitar_BackhandSlash";
    Effects[Effects["MAT_Guitar_ForehandChop"] = 1023] = "MAT_Guitar_ForehandChop";
    Effects[Effects["MAT_Guitar_BackhandChop"] = 1024] = "MAT_Guitar_BackhandChop";
    Effects[Effects["MAT_Guitar_ForehandUppercut"] = 1025] = "MAT_Guitar_ForehandUppercut";
    Effects[Effects["MAT_Guitar_ForwardStab"] = 1026] = "MAT_Guitar_ForwardStab";
    Effects[Effects["MAT_Torch"] = 1002] = "MAT_Torch";
    Effects[Effects["MAT_Knife"] = 1003] = "MAT_Knife";
    Effects[Effects["MAT_Fists"] = 1004] = "MAT_Fists";
    Effects[Effects["MAT_Fists_RightHook"] = 1041] = "MAT_Fists_RightHook";
    Effects[Effects["MAT_Fists_LeftHook"] = 1042] = "MAT_Fists_LeftHook";
    Effects[Effects["MAT_Fists_RightStraight"] = 1043] = "MAT_Fists_RightStraight";
    Effects[Effects["MAT_Fists_LeftStraight"] = 1044] = "MAT_Fists_LeftStraight";
    Effects[Effects["MAT_Fists_RightUppercut"] = 1045] = "MAT_Fists_RightUppercut";
    Effects[Effects["MAT_Fists_LeftUppercut"] = 1046] = "MAT_Fists_LeftUppercut";
    Effects[Effects["MAT_Arrow"] = 1005] = "MAT_Arrow";
    Effects[Effects["MAT_Arrow_Flaming"] = 1008] = "MAT_Arrow_Flaming";
    Effects[Effects["PFX_Impact_BulletLight_Metal"] = 1143] = "PFX_Impact_BulletLight_Metal";
    Effects[Effects["PFX_Impact_BulletLight_Metal_Vehicle"] = 1144] = "PFX_Impact_BulletLight_Metal_Vehicle";
    Effects[Effects["PFX_Impact_BulletLight_Dirt"] = 1148] = "PFX_Impact_BulletLight_Dirt";
    Effects[Effects["PFX_Impact_BulletLight_Rubber"] = 1150] = "PFX_Impact_BulletLight_Rubber";
    Effects[Effects["PFX_Impact_BulletLight_Flesh"] = 1165] = "PFX_Impact_BulletLight_Flesh";
    Effects[Effects["PFX_Impact_BulletLight_Zombie"] = 1166] = "PFX_Impact_BulletLight_Zombie";
    Effects[Effects["MAT_WheelDust_OffRoader"] = 1067] = "MAT_WheelDust_OffRoader";
    Effects[Effects["FX_Fire_Barrel"] = 1202] = "FX_Fire_Barrel";
    Effects[Effects["EFX_Fire_Barrel"] = 1203] = "EFX_Fire_Barrel";
    Effects[Effects["EFX_Fire_Dumpster"] = 1204] = "EFX_Fire_Dumpster";
    Effects[Effects["EFX_Fire_PlayerStructures"] = 1214] = "EFX_Fire_PlayerStructures";
    Effects[Effects["PFX_Fire_Person"] = 1205] = "PFX_Fire_Person";
    Effects[Effects["PFX_Fire_Zombie_loop"] = 1206] = "PFX_Fire_Zombie_loop";
    Effects[Effects["EFX_Fire_Campfire"] = 1207] = "EFX_Fire_Campfire";
    Effects[Effects["EFX_Fire_Campfire_Smoldering"] = 1208] = "EFX_Fire_Campfire_Smoldering";
    Effects[Effects["EFX_Fire_TrashCan_Small"] = 1209] = "EFX_Fire_TrashCan_Small";
    Effects[Effects["EFX_Light_StreetLight"] = 1210] = "EFX_Light_StreetLight";
    Effects[Effects["EFX_Light_StreetLight_wShadows"] = 1211] = "EFX_Light_StreetLight_wShadows";
    Effects[Effects["PFX_Fire_Person_loop"] = 1212] = "PFX_Fire_Person_loop";
    Effects[Effects["PFX_Fire_Zombie"] = 1213] = "PFX_Fire_Zombie";
    Effects[Effects["P2P_HealingBeam"] = 1262] = "P2P_HealingBeam";
    Effects[Effects["PFX_Impact_BulletLight_Concrete"] = 1290] = "PFX_Impact_BulletLight_Concrete";
    Effects[Effects["PFX_Impact_BulletLight_Wood"] = 1291] = "PFX_Impact_BulletLight_Wood";
    Effects[Effects["PFX_Impact_BulletLight_Glass"] = 1297] = "PFX_Impact_BulletLight_Glass";
    Effects[Effects["PFX_Impact_BulletLight_Grass"] = 1327] = "PFX_Impact_BulletLight_Grass";
    Effects[Effects["PFX_Impact_Blade_Flesh"] = 1600] = "PFX_Impact_Blade_Flesh";
    Effects[Effects["PFX_Impact_Blade_Concrete"] = 1601] = "PFX_Impact_Blade_Concrete";
    Effects[Effects["PFX_Impact_Blade_Dirt"] = 1602] = "PFX_Impact_Blade_Dirt";
    Effects[Effects["PFX_Impact_Blade_Glass"] = 1603] = "PFX_Impact_Blade_Glass";
    Effects[Effects["PFX_Impact_Blade_Grass"] = 1604] = "PFX_Impact_Blade_Grass";
    Effects[Effects["PFX_Impact_Blade_Metal"] = 1605] = "PFX_Impact_Blade_Metal";
    Effects[Effects["PFX_Impact_Blade_Rock"] = 1606] = "PFX_Impact_Blade_Rock";
    Effects[Effects["PFX_Impact_Blade_Wood"] = 1607] = "PFX_Impact_Blade_Wood";
    Effects[Effects["PFX_Impact_Blade_Zombie"] = 1608] = "PFX_Impact_Blade_Zombie";
    Effects[Effects["PFX_Impact_Blade_Water"] = 1609] = "PFX_Impact_Blade_Water";
    Effects[Effects["PFX_Impact_Blade_Metal_Vehicle"] = 1610] = "PFX_Impact_Blade_Metal_Vehicle";
    Effects[Effects["PFX_Impact_Knife_Flesh"] = 1620] = "PFX_Impact_Knife_Flesh";
    Effects[Effects["PFX_Impact_Knife_Concrete"] = 1621] = "PFX_Impact_Knife_Concrete";
    Effects[Effects["PFX_Impact_Knife_Dirt"] = 1622] = "PFX_Impact_Knife_Dirt";
    Effects[Effects["PFX_Impact_Knife_Glass"] = 1623] = "PFX_Impact_Knife_Glass";
    Effects[Effects["PFX_Impact_Knife_Grass"] = 1624] = "PFX_Impact_Knife_Grass";
    Effects[Effects["PFX_Impact_Knife_Metal"] = 1625] = "PFX_Impact_Knife_Metal";
    Effects[Effects["PFX_Impact_Knife_Rock"] = 1626] = "PFX_Impact_Knife_Rock";
    Effects[Effects["PFX_Impact_Knife_Wood"] = 1627] = "PFX_Impact_Knife_Wood";
    Effects[Effects["PFX_Impact_Knife_Zombie"] = 1628] = "PFX_Impact_Knife_Zombie";
    Effects[Effects["PFX_Impact_Knife_Water"] = 1629] = "PFX_Impact_Knife_Water";
    Effects[Effects["PFX_Impact_Knife_Metal_Vehicle"] = 1630] = "PFX_Impact_Knife_Metal_Vehicle";
    Effects[Effects["PFX_Impact_Torch_Flesh"] = 1640] = "PFX_Impact_Torch_Flesh";
    Effects[Effects["PFX_Impact_Torch_Concrete"] = 1641] = "PFX_Impact_Torch_Concrete";
    Effects[Effects["PFX_Impact_Torch_Dirt"] = 1642] = "PFX_Impact_Torch_Dirt";
    Effects[Effects["PFX_Impact_Torch_Glass"] = 1643] = "PFX_Impact_Torch_Glass";
    Effects[Effects["PFX_Impact_Torch_Grass"] = 1644] = "PFX_Impact_Torch_Grass";
    Effects[Effects["PFX_Impact_Torch_Metal"] = 1645] = "PFX_Impact_Torch_Metal";
    Effects[Effects["PFX_Impact_Torch_Rock"] = 1646] = "PFX_Impact_Torch_Rock";
    Effects[Effects["PFX_Impact_Torch_Wood"] = 1647] = "PFX_Impact_Torch_Wood";
    Effects[Effects["PFX_Impact_Torch_Zombie"] = 1648] = "PFX_Impact_Torch_Zombie";
    Effects[Effects["PFX_Impact_Torch_Water"] = 1649] = "PFX_Impact_Torch_Water";
    Effects[Effects["PFX_Impact_Torch_Metal_Vehicle"] = 1650] = "PFX_Impact_Torch_Metal_Vehicle";
    Effects[Effects["PFX_Impact_Fists_Flesh"] = 1660] = "PFX_Impact_Fists_Flesh";
    Effects[Effects["PFX_Impact_Fists_Concrete"] = 1661] = "PFX_Impact_Fists_Concrete";
    Effects[Effects["PFX_Impact_Fists_Dirt"] = 1662] = "PFX_Impact_Fists_Dirt";
    Effects[Effects["PFX_Impact_Fists_Glass"] = 1663] = "PFX_Impact_Fists_Glass";
    Effects[Effects["PFX_Impact_Fists_Grass"] = 1664] = "PFX_Impact_Fists_Grass";
    Effects[Effects["PFX_Impact_Fists_Metal"] = 1665] = "PFX_Impact_Fists_Metal";
    Effects[Effects["PFX_Impact_Fists_Rock"] = 1666] = "PFX_Impact_Fists_Rock";
    Effects[Effects["PFX_Impact_Fists_Wood"] = 1667] = "PFX_Impact_Fists_Wood";
    Effects[Effects["PFX_Impact_Fists_Zombie"] = 1668] = "PFX_Impact_Fists_Zombie";
    Effects[Effects["PFX_Impact_Fists_Metal_Vehicle"] = 1670] = "PFX_Impact_Fists_Metal_Vehicle";
    Effects[Effects["PFX_Impact_Arrow_Flesh"] = 1680] = "PFX_Impact_Arrow_Flesh";
    Effects[Effects["PFX_Impact_Arrow_Concrete"] = 1681] = "PFX_Impact_Arrow_Concrete";
    Effects[Effects["PFX_Impact_Arrow_Dirt"] = 1682] = "PFX_Impact_Arrow_Dirt";
    Effects[Effects["PFX_Impact_Arrow_Glass"] = 1683] = "PFX_Impact_Arrow_Glass";
    Effects[Effects["PFX_Impact_Arrow_Grass"] = 1684] = "PFX_Impact_Arrow_Grass";
    Effects[Effects["PFX_Impact_Arrow_Metal"] = 1685] = "PFX_Impact_Arrow_Metal";
    Effects[Effects["PFX_Impact_Arrow_Rock"] = 1686] = "PFX_Impact_Arrow_Rock";
    Effects[Effects["PFX_Impact_Arrow_Wood"] = 1687] = "PFX_Impact_Arrow_Wood";
    Effects[Effects["PFX_Impact_Arrow_Zombie"] = 1688] = "PFX_Impact_Arrow_Zombie";
    Effects[Effects["PFX_Impact_Arrow_Water"] = 1689] = "PFX_Impact_Arrow_Water";
    Effects[Effects["PFX_Impact_Arrow_Metal_Vehicle"] = 1690] = "PFX_Impact_Arrow_Metal_Vehicle";
    Effects[Effects["PFX_Impact_Arrow_Rubber"] = 1691] = "PFX_Impact_Arrow_Rubber";
    Effects[Effects["PFX_Flashlight_1P"] = 1705] = "PFX_Flashlight_1P";
    Effects[Effects["PFX_Flashlight_3P"] = 1706] = "PFX_Flashlight_3P";
    Effects[Effects["PFX_Impact_Guitar_Flesh"] = 1720] = "PFX_Impact_Guitar_Flesh";
    Effects[Effects["PFX_Impact_Guitar_Concrete"] = 1721] = "PFX_Impact_Guitar_Concrete";
    Effects[Effects["PFX_Impact_Guitar_Dirt"] = 1722] = "PFX_Impact_Guitar_Dirt";
    Effects[Effects["PFX_Impact_Guitar_Glass"] = 1723] = "PFX_Impact_Guitar_Glass";
    Effects[Effects["PFX_Impact_Guitar_Grass"] = 1724] = "PFX_Impact_Guitar_Grass";
    Effects[Effects["PFX_Impact_Guitar_Metal"] = 1725] = "PFX_Impact_Guitar_Metal";
    Effects[Effects["PFX_Impact_Guitar_Rock"] = 1726] = "PFX_Impact_Guitar_Rock";
    Effects[Effects["PFX_Impact_Guitar_Wood"] = 1727] = "PFX_Impact_Guitar_Wood";
    Effects[Effects["PFX_Impact_Guitar_Zombie"] = 1728] = "PFX_Impact_Guitar_Zombie";
    Effects[Effects["PFX_Impact_Guitar_Water"] = 1729] = "PFX_Impact_Guitar_Water";
    Effects[Effects["PFX_Impact_Guitar_Metal_Vehicle"] = 1730] = "PFX_Impact_Guitar_Metal_Vehicle";
    Effects[Effects["PFX_Impact_Blade_Zombie_ForehandSlash"] = 1811] = "PFX_Impact_Blade_Zombie_ForehandSlash";
    Effects[Effects["PFX_Impact_Blade_Zombie_BackhandSlash"] = 1812] = "PFX_Impact_Blade_Zombie_BackhandSlash";
    Effects[Effects["PFX_Impact_Blade_Zombie_ForehandChop"] = 1813] = "PFX_Impact_Blade_Zombie_ForehandChop";
    Effects[Effects["PFX_Impact_Blade_Zombie_BackhandChop"] = 1814] = "PFX_Impact_Blade_Zombie_BackhandChop";
    Effects[Effects["PFX_Impact_Blade_Zombie_ForehandUppercut"] = 1815] = "PFX_Impact_Blade_Zombie_ForehandUppercut";
    Effects[Effects["PFX_Impact_Blade_Zombie_ForwardStab"] = 5114] = "PFX_Impact_Blade_Zombie_ForwardStab";
    Effects[Effects["PFX_Impact_Blade_Human_ForehandSlash"] = 1816] = "PFX_Impact_Blade_Human_ForehandSlash";
    Effects[Effects["PFX_Impact_Blade_Human_BackhandSlash"] = 1817] = "PFX_Impact_Blade_Human_BackhandSlash";
    Effects[Effects["PFX_Impact_Blade_Human_ForehandChop"] = 1818] = "PFX_Impact_Blade_Human_ForehandChop";
    Effects[Effects["PFX_Impact_Blade_Human_BackhandChop"] = 1819] = "PFX_Impact_Blade_Human_BackhandChop";
    Effects[Effects["PFX_Impact_Blade_Human_ForehandUppercut"] = 1820] = "PFX_Impact_Blade_Human_ForehandUppercut";
    Effects[Effects["PFX_Impact_Blade_Human_ForwardStab"] = 5113] = "PFX_Impact_Blade_Human_ForwardStab";
    Effects[Effects["PFX_Impact_Guitar_Zombie_ForehandSlash"] = 1823] = "PFX_Impact_Guitar_Zombie_ForehandSlash";
    Effects[Effects["PFX_Impact_Guitar_Zombie_BackhandSlash"] = 1824] = "PFX_Impact_Guitar_Zombie_BackhandSlash";
    Effects[Effects["PFX_Impact_Guitar_Zombie_ForehandChop"] = 1825] = "PFX_Impact_Guitar_Zombie_ForehandChop";
    Effects[Effects["PFX_Impact_Guitar_Zombie_BackhandChop"] = 1826] = "PFX_Impact_Guitar_Zombie_BackhandChop";
    Effects[Effects["PFX_Impact_Guitar_Zombie_ForehandUppercut"] = 1827] = "PFX_Impact_Guitar_Zombie_ForehandUppercut";
    Effects[Effects["PFX_Impact_Guitar_Zombie_ForwardStab"] = 1828] = "PFX_Impact_Guitar_Zombie_ForwardStab";
    Effects[Effects["PFX_Impact_Guitar_Human_ForehandSlash"] = 1829] = "PFX_Impact_Guitar_Human_ForehandSlash";
    Effects[Effects["PFX_Impact_Guitar_Human_BackhandSlash"] = 1830] = "PFX_Impact_Guitar_Human_BackhandSlash";
    Effects[Effects["PFX_Impact_Guitar_Human_ForehandChop"] = 1831] = "PFX_Impact_Guitar_Human_ForehandChop";
    Effects[Effects["PFX_Impact_Guitar_Human_BackhandChop"] = 1832] = "PFX_Impact_Guitar_Human_BackhandChop";
    Effects[Effects["PFX_Impact_Guitar_Human_ForehandUppercut"] = 1833] = "PFX_Impact_Guitar_Human_ForehandUppercut";
    Effects[Effects["PFX_Impact_Guitar_Human_ForwardStab"] = 1834] = "PFX_Impact_Guitar_Human_ForwardStab";
    Effects[Effects["PFX_Impact_Fists_Zombie_RightHook"] = 1841] = "PFX_Impact_Fists_Zombie_RightHook";
    Effects[Effects["PFX_Impact_Fists_Zombie_LeftHook"] = 1842] = "PFX_Impact_Fists_Zombie_LeftHook";
    Effects[Effects["PFX_Impact_Fists_Zombie_RightStraight"] = 1843] = "PFX_Impact_Fists_Zombie_RightStraight";
    Effects[Effects["PFX_Impact_Fists_Zombie_LeftStraight"] = 1844] = "PFX_Impact_Fists_Zombie_LeftStraight";
    Effects[Effects["PFX_Impact_Fists_Zombie_RightUppercut"] = 1845] = "PFX_Impact_Fists_Zombie_RightUppercut";
    Effects[Effects["PFX_Impact_Fists_Zombie_LeftUppercut"] = 1846] = "PFX_Impact_Fists_Zombie_LeftUppercut";
    Effects[Effects["PFX_Impact_Fists_Flesh_RightHook"] = 1851] = "PFX_Impact_Fists_Flesh_RightHook";
    Effects[Effects["PFX_Impact_Fists_Flesh_LeftHook"] = 1852] = "PFX_Impact_Fists_Flesh_LeftHook";
    Effects[Effects["PFX_Impact_Fists_Flesh_RightStraight"] = 1853] = "PFX_Impact_Fists_Flesh_RightStraight";
    Effects[Effects["PFX_Impact_Fists_Flesh_LeftStraight"] = 1854] = "PFX_Impact_Fists_Flesh_LeftStraight";
    Effects[Effects["PFX_Impact_Fists_Flesh_RightUppercut"] = 1855] = "PFX_Impact_Fists_Flesh_RightUppercut";
    Effects[Effects["PFX_Impact_Fists_Flesh_LeftUppercut"] = 1856] = "PFX_Impact_Fists_Flesh_LeftUppercut";
    Effects[Effects["PFX_Impact_ExplosionGround_Dirt_01m_DEPRECATED"] = 1867] = "PFX_Impact_ExplosionGround_Dirt_01m_DEPRECATED";
    Effects[Effects["PFX_Impact_Explosion_FragGrenade_Dirt_08m"] = 1873] = "PFX_Impact_Explosion_FragGrenade_Dirt_08m";
    Effects[Effects["PFX_Impact_Explosion_ExplosiveArrow_Dirt_08m"] = 1874] = "PFX_Impact_Explosion_ExplosiveArrow_Dirt_08m";
    Effects[Effects["PFX_Impact_Explosion_Landmine_Dirt_10m"] = 1875] = "PFX_Impact_Explosion_Landmine_Dirt_10m";
    Effects[Effects["PFX_Impact_Explosion_AirdropBomb_Dirt_10m"] = 1876] = "PFX_Impact_Explosion_AirdropBomb_Dirt_10m";
    Effects[Effects["PFX_Impact_Explosion_FragGrenade_Metal_08m"] = 2082] = "PFX_Impact_Explosion_FragGrenade_Metal_08m";
    Effects[Effects["PFX_Impact_Explosion_ExplosiveArrow_Metal_08m"] = 2083] = "PFX_Impact_Explosion_ExplosiveArrow_Metal_08m";
    Effects[Effects["PFX_Impact_Explosion_Landmine_Metal_10m"] = 2084] = "PFX_Impact_Explosion_Landmine_Metal_10m";
    Effects[Effects["PFX_Impact_Explosion_AirdropBomb_Metal_10m"] = 2085] = "PFX_Impact_Explosion_AirdropBomb_Metal_10m";
    Effects[Effects["PFX_Impact_ExplosionGround_Concrete_02m_DEPRECATED"] = 2104] = "PFX_Impact_ExplosionGround_Concrete_02m_DEPRECATED";
    Effects[Effects["PFX_Impact_Explosion_FragGrenade_Concrete_08m"] = 2109] = "PFX_Impact_Explosion_FragGrenade_Concrete_08m";
    Effects[Effects["PFX_Impact_Explosion_ExplosiveArrow_Concrete_08m"] = 2110] = "PFX_Impact_Explosion_ExplosiveArrow_Concrete_08m";
    Effects[Effects["PFX_Impact_Explosion_Landmine_Concrete_10m"] = 2111] = "PFX_Impact_Explosion_Landmine_Concrete_10m";
    Effects[Effects["PFX_Impact_Explosion_AirdropBomb_Concrete_10m"] = 2112] = "PFX_Impact_Explosion_AirdropBomb_Concrete_10m";
    Effects[Effects["PFX_Impact_Explosion_FragGrenade_Wood_08m"] = 2118] = "PFX_Impact_Explosion_FragGrenade_Wood_08m";
    Effects[Effects["PFX_Impact_Explosion_ExplosiveArrow_Wood_08m"] = 2119] = "PFX_Impact_Explosion_ExplosiveArrow_Wood_08m";
    Effects[Effects["PFX_Impact_Explosion_Landmine_Wood_10m"] = 2120] = "PFX_Impact_Explosion_Landmine_Wood_10m";
    Effects[Effects["PFX_Impact_Explosion_AirdropBomb_Wood_10m"] = 2121] = "PFX_Impact_Explosion_AirdropBomb_Wood_10m";
    Effects[Effects["PFX_Impact_Explosion_FragGrenade_Grass_08m"] = 2136] = "PFX_Impact_Explosion_FragGrenade_Grass_08m";
    Effects[Effects["PFX_Impact_Explosion_ExplosiveArrow_Grass_08m"] = 2137] = "PFX_Impact_Explosion_ExplosiveArrow_Grass_08m";
    Effects[Effects["PFX_Impact_Explosion_Landmine_Grass_10m"] = 2138] = "PFX_Impact_Explosion_Landmine_Grass_10m";
    Effects[Effects["PFX_Impact_Explosion_AirdropBomb_Grass_10m"] = 2139] = "PFX_Impact_Explosion_AirdropBomb_Grass_10m";
    Effects[Effects["PFX_Impact_Explosion_FragGrenade_Rock_08m"] = 2145] = "PFX_Impact_Explosion_FragGrenade_Rock_08m";
    Effects[Effects["PFX_Impact_Explosion_ExplosiveArrow_Rock_08m"] = 2146] = "PFX_Impact_Explosion_ExplosiveArrow_Rock_08m";
    Effects[Effects["PFX_Impact_Explosion_Landmine_Rock_10m"] = 2147] = "PFX_Impact_Explosion_Landmine_Rock_10m";
    Effects[Effects["PFX_Impact_Explosion_AirdropBomb_Rock_10m"] = 2148] = "PFX_Impact_Explosion_AirdropBomb_Rock_10m";
    Effects[Effects["PFX_Impact_BulletLight_Rock"] = 3560] = "PFX_Impact_BulletLight_Rock";
    Effects[Effects["PFX_RallyPoint_Smoke_Orange"] = 3621] = "PFX_RallyPoint_Smoke_Orange";
    Effects[Effects["PFX_RallyPoint_Smoke_Green"] = 3622] = "PFX_RallyPoint_Smoke_Green";
    Effects[Effects["PFX_RallyPoint_Smoke_Purple"] = 3623] = "PFX_RallyPoint_Smoke_Purple";
    Effects[Effects["PFX_RallyPoint_Smoke_Yellow"] = 3624] = "PFX_RallyPoint_Smoke_Yellow";
    Effects[Effects["PFX_RallyPoint_Smoke_LimeGreen"] = 4537] = "PFX_RallyPoint_Smoke_LimeGreen";
    Effects[Effects["PFX_RallyPoint_Smoke_Red"] = 4538] = "PFX_RallyPoint_Smoke_Red";
    Effects[Effects["PFX_RallyPoint_Smoke_Blue"] = 4539] = "PFX_RallyPoint_Smoke_Blue";
    Effects[Effects["PFX_RallyPoint_Smoke_Teal"] = 4540] = "PFX_RallyPoint_Smoke_Teal";
    Effects[Effects["PFX_RallyPoint_Smoke_Magenta"] = 4541] = "PFX_RallyPoint_Smoke_Magenta";
    Effects[Effects["PFX_Impact_BulletLight_Water"] = 3773] = "PFX_Impact_BulletLight_Water";
    Effects[Effects["PFX_Impact_Explosion_FragGrenade_Water_08m"] = 4149] = "PFX_Impact_Explosion_FragGrenade_Water_08m";
    Effects[Effects["PFX_Impact_Explosion_ExplosiveArrow_Water_08m"] = 4150] = "PFX_Impact_Explosion_ExplosiveArrow_Water_08m";
    Effects[Effects["PFX_Impact_Explosion_Landmine_Water_10m"] = 4151] = "PFX_Impact_Explosion_Landmine_Water_10m";
    Effects[Effects["PFX_Impact_Explosion_AirdropBomb_Water_10m"] = 4152] = "PFX_Impact_Explosion_AirdropBomb_Water_10m";
    Effects[Effects["PFX_Impact_Mat_Metal_Vehicle_01m"] = 4707] = "PFX_Impact_Mat_Metal_Vehicle_01m";
    Effects[Effects["PFX_Impact_Mat_Metal_Vehicle_03m"] = 4708] = "PFX_Impact_Mat_Metal_Vehicle_03m";
    Effects[Effects["PFX_Impact_Mat_Metal_Vehicle_06m"] = 4709] = "PFX_Impact_Mat_Metal_Vehicle_06m";
    Effects[Effects["PFX_Impact_Mat_Metal_Vehicle_09m"] = 4710] = "PFX_Impact_Mat_Metal_Vehicle_09m";
    Effects[Effects["MAT_Explosion_FragGrenade_08m"] = 2301] = "MAT_Explosion_FragGrenade_08m";
    Effects[Effects["MAT_Explosion_ExplosiveArrow_08m"] = 2302] = "MAT_Explosion_ExplosiveArrow_08m";
    Effects[Effects["MAT_Explosion_Landmine_10m"] = 2303] = "MAT_Explosion_Landmine_10m";
    Effects[Effects["MAT_Explosion_AirdropBomb_10m"] = 2304] = "MAT_Explosion_AirdropBomb_10m";
    Effects[Effects["PFX_Impact_GrenadeSmoke_loop"] = 2333] = "PFX_Impact_GrenadeSmoke_loop";
    Effects[Effects["PFX_Impact_GrenadeSmoke_loop_noSound"] = 2334] = "PFX_Impact_GrenadeSmoke_loop_noSound";
    Effects[Effects["PFX_Impact_GrenadePoison_loop"] = 2335] = "PFX_Impact_GrenadePoison_loop";
    Effects[Effects["PFX_Impact_GrenadePoison_loop_noSound"] = 2336] = "PFX_Impact_GrenadePoison_loop_noSound";
    Effects[Effects["SFX_Run_Player"] = 2984] = "SFX_Run_Player";
    Effects[Effects["SFX_Run_Player_Dirt"] = 2980] = "SFX_Run_Player_Dirt";
    Effects[Effects["SFX_Run_Player_Grass"] = 2981] = "SFX_Run_Player_Grass";
    Effects[Effects["SFX_Run_Player_Concrete"] = 2982] = "SFX_Run_Player_Concrete";
    Effects[Effects["SFX_Run_Player_Metal"] = 2983] = "SFX_Run_Player_Metal";
    Effects[Effects["SFX_Run_Player_Water"] = 3633] = "SFX_Run_Player_Water";
    Effects[Effects["SFX_Run_Player_WaterShallow"] = 3634] = "SFX_Run_Player_WaterShallow";
    Effects[Effects["SFX_Run_Player_Ice"] = 3937] = "SFX_Run_Player_Ice";
    Effects[Effects["SFX_Run_Player_Metal_Grate"] = 4390] = "SFX_Run_Player_Metal_Grate";
    Effects[Effects["SFX_Walk_Player"] = 3625] = "SFX_Walk_Player";
    Effects[Effects["SFX_Walk_Player_Dirt"] = 3626] = "SFX_Walk_Player_Dirt";
    Effects[Effects["SFX_Walk_Player_Grass"] = 3627] = "SFX_Walk_Player_Grass";
    Effects[Effects["SFX_Walk_Player_Metal"] = 3628] = "SFX_Walk_Player_Metal";
    Effects[Effects["SFX_Walk_Player_Concrete"] = 3629] = "SFX_Walk_Player_Concrete";
    Effects[Effects["SFX_Walk_Player_Water"] = 3630] = "SFX_Walk_Player_Water";
    Effects[Effects["SFX_Walk_Player_Snow"] = 3631] = "SFX_Walk_Player_Snow";
    Effects[Effects["SFX_Walk_Player_WaterShallow"] = 3632] = "SFX_Walk_Player_WaterShallow";
    Effects[Effects["SFX_Walk_Player_Ice"] = 3938] = "SFX_Walk_Player_Ice";
    Effects[Effects["SFX_Walk_Player_Metal_Grate"] = 4393] = "SFX_Walk_Player_Metal_Grate";
    Effects[Effects["SFX_Jump_Player"] = 3635] = "SFX_Jump_Player";
    Effects[Effects["SFX_Jump_Player_Dirt"] = 3636] = "SFX_Jump_Player_Dirt";
    Effects[Effects["SFX_Jump_Player_Grass"] = 3637] = "SFX_Jump_Player_Grass";
    Effects[Effects["SFX_Jump_Player_Metal"] = 3638] = "SFX_Jump_Player_Metal";
    Effects[Effects["SFX_Jump_Player_Concrete"] = 3639] = "SFX_Jump_Player_Concrete";
    Effects[Effects["SFX_Jump_Player_Water"] = 3640] = "SFX_Jump_Player_Water";
    Effects[Effects["SFX_Jump_Player_WaterShallow"] = 3641] = "SFX_Jump_Player_WaterShallow";
    Effects[Effects["SFX_Jump_Player_Metal_Grate"] = 4392] = "SFX_Jump_Player_Metal_Grate";
    Effects[Effects["SFX_Land_Player"] = 3642] = "SFX_Land_Player";
    Effects[Effects["SFX_Land_Player_Dirt"] = 3643] = "SFX_Land_Player_Dirt";
    Effects[Effects["SFX_Land_Player_Grass"] = 3644] = "SFX_Land_Player_Grass";
    Effects[Effects["SFX_Land_Player_Metal"] = 3645] = "SFX_Land_Player_Metal";
    Effects[Effects["SFX_Land_Player_Concrete"] = 3646] = "SFX_Land_Player_Concrete";
    Effects[Effects["SFX_Land_Player_Water"] = 3647] = "SFX_Land_Player_Water";
    Effects[Effects["SFX_Land_Player_WaterShallow"] = 3648] = "SFX_Land_Player_WaterShallow";
    Effects[Effects["SFX_Land_Player_Metal_Grate"] = 4391] = "SFX_Land_Player_Metal_Grate";
    Effects[Effects["VEH_Engine_Boost_OffRoader"] = 5016] = "VEH_Engine_Boost_OffRoader";
    Effects[Effects["PFX_Compass_Needle"] = 5018] = "PFX_Compass_Needle";
    Effects[Effects["SFX_Walk_Player_Wood"] = 5020] = "SFX_Walk_Player_Wood";
    Effects[Effects["SFX_Run_Player_Wood"] = 5021] = "SFX_Run_Player_Wood";
    Effects[Effects["SFX_Jump_Player_Wood"] = 5022] = "SFX_Jump_Player_Wood";
    Effects[Effects["SFX_Land_Player_Wood"] = 5023] = "SFX_Land_Player_Wood";
    Effects[Effects["PFX_TreeFall_Start_Small"] = 5024] = "PFX_TreeFall_Start_Small";
    Effects[Effects["PFX_TreeFall_Land_CameraShake_Small"] = 5025] = "PFX_TreeFall_Land_CameraShake_Small";
    Effects[Effects["PFX_TreeFall_Start_Big"] = 5026] = "PFX_TreeFall_Start_Big";
    Effects[Effects["PFX_TreeFall_Land_CameraShake_Big"] = 5027] = "PFX_TreeFall_Land_CameraShake_Big";
    Effects[Effects["EFX_Fire_Furnace"] = 5028] = "EFX_Fire_Furnace";
    Effects[Effects["PFX_Fishing_Bobbing_Trout"] = 5029] = "PFX_Fishing_Bobbing_Trout";
    Effects[Effects["PFX_Fishing_Struggling_Trout"] = 5030] = "PFX_Fishing_Struggling_Trout";
    Effects[Effects["SFX_Idle_Player"] = 5032] = "SFX_Idle_Player";
    Effects[Effects["SFX_Idle_Player_Water"] = 5033] = "SFX_Idle_Player_Water";
    Effects[Effects["PFX_Fire_Lighter"] = 5034] = "PFX_Fire_Lighter";
    Effects[Effects["PFX_Scent_Human_Big"] = 5035] = "PFX_Scent_Human_Big";
    Effects[Effects["PFX_Scent_Human_Medium"] = 5036] = "PFX_Scent_Human_Medium";
    Effects[Effects["PFX_Scent_Human_Small"] = 5037] = "PFX_Scent_Human_Small";
    Effects[Effects["PFX_Impact_AirDrop_Large"] = 5038] = "PFX_Impact_AirDrop_Large";
    Effects[Effects["PFX_Impact_BulletShotgun_Default"] = 1300] = "PFX_Impact_BulletShotgun_Default";
    Effects[Effects["PFX_Impact_BulletShotgun_Asphalt"] = 1301] = "PFX_Impact_BulletShotgun_Asphalt";
    Effects[Effects["PFX_Impact_BulletShotgun_Flesh"] = 1302] = "PFX_Impact_BulletShotgun_Flesh";
    Effects[Effects["PFX_Impact_BulletShotgun_Concrete"] = 1304] = "PFX_Impact_BulletShotgun_Concrete";
    Effects[Effects["PFX_Impact_BulletShotgun_Dirt"] = 1306] = "PFX_Impact_BulletShotgun_Dirt";
    Effects[Effects["PFX_Impact_BulletShotgun_Glass"] = 1307] = "PFX_Impact_BulletShotgun_Glass";
    Effects[Effects["PFX_Impact_BulletShotgun_Grass"] = 1308] = "PFX_Impact_BulletShotgun_Grass";
    Effects[Effects["PFX_Impact_BulletShotgun_Metal"] = 1309] = "PFX_Impact_BulletShotgun_Metal";
    Effects[Effects["PFX_Impact_BulletShotgun_Rock"] = 1311] = "PFX_Impact_BulletShotgun_Rock";
    Effects[Effects["PFX_Impact_BulletShotgun_Rubber"] = 1313] = "PFX_Impact_BulletShotgun_Rubber";
    Effects[Effects["PFX_Impact_BulletShotgun_Metal_Vehicle"] = 1314] = "PFX_Impact_BulletShotgun_Metal_Vehicle";
    Effects[Effects["PFX_Impact_BulletShotgun_Water"] = 1315] = "PFX_Impact_BulletShotgun_Water";
    Effects[Effects["PFX_Impact_BulletShotgun_Wood"] = 1316] = "PFX_Impact_BulletShotgun_Wood";
    Effects[Effects["PFX_Impact_BulletShotgun_Zombie"] = 1317] = "PFX_Impact_BulletShotgun_Zombie";
    Effects[Effects["PFX_Bleeding_Human_medium"] = 5041] = "PFX_Bleeding_Human_medium";
    Effects[Effects["MAT_Explosion_FlashGrenade"] = 4657] = "MAT_Explosion_FlashGrenade";
    Effects[Effects["PFX_Impact_Explosion_FlashGrenade_Default"] = 4658] = "PFX_Impact_Explosion_FlashGrenade_Default";
    Effects[Effects["PFX_Impact_Explosion_FlashGrenade_Water"] = 4659] = "PFX_Impact_Explosion_FlashGrenade_Water";
    Effects[Effects["PFX_Bleeding_Human_moderate_loop"] = 5042] = "PFX_Bleeding_Human_moderate_loop";
    Effects[Effects["EFX_FireHydrant_Gushing"] = 5043] = "EFX_FireHydrant_Gushing";
    Effects[Effects["EFX_Fire_Barbecue"] = 5044] = "EFX_Fire_Barbecue";
    Effects[Effects["PFX_Light_Rigged_to_Battery"] = 5046] = "PFX_Light_Rigged_to_Battery";
    Effects[Effects["PFX_Circle_Bomb"] = 5047] = "PFX_Circle_Bomb";
    Effects[Effects["SFX_Door_Wood_Open"] = 5048] = "SFX_Door_Wood_Open";
    Effects[Effects["SFX_Door_Wood_Close"] = 5049] = "SFX_Door_Wood_Close";
    Effects[Effects["PFX_Impact_Zombie_Flesh"] = 5050] = "PFX_Impact_Zombie_Flesh";
    Effects[Effects["EFX_Flare_Parachute_Rocket_Red"] = 5051] = "EFX_Flare_Parachute_Rocket_Red";
    Effects[Effects["EFX_Water_Level_RainCatcher"] = 5052] = "EFX_Water_Level_RainCatcher";
    Effects[Effects["EFX_Flare_ParachuteRocket"] = 5053] = "EFX_Flare_ParachuteRocket";
    Effects[Effects["EFX_Light_SuperMarket_OutsideLight"] = 5054] = "EFX_Light_SuperMarket_OutsideLight";
    Effects[Effects["EFX_Crop_Fertilizer"] = 5056] = "EFX_Crop_Fertilizer";
    Effects[Effects["SFX_Player_Handcuff_Cuff"] = 5057] = "SFX_Player_Handcuff_Cuff";
    Effects[Effects["SFX_Player_Handcuff_Uncuff"] = 5058] = "SFX_Player_Handcuff_Uncuff";
    Effects[Effects["SFX_Land_Player_Metal_Hollow"] = 5059] = "SFX_Land_Player_Metal_Hollow";
    Effects[Effects["SFX_Land_Player_Barbed_Wire"] = 5060] = "SFX_Land_Player_Barbed_Wire";
    Effects[Effects["SFX_Land_Player_Chain_Link"] = 5061] = "SFX_Land_Player_Chain_Link";
    Effects[Effects["SFX_Jump_Player_Chain_Link"] = 5062] = "SFX_Jump_Player_Chain_Link";
    Effects[Effects["SFX_Run_Player_Chain_Link"] = 5063] = "SFX_Run_Player_Chain_Link";
    Effects[Effects["SFX_Walk_Player_Chain_Link"] = 5064] = "SFX_Walk_Player_Chain_Link";
    Effects[Effects["SFX_Walk_Player_Barbed_Wire"] = 5065] = "SFX_Walk_Player_Barbed_Wire";
    Effects[Effects["SFX_Run_Player_Barbed_Wire"] = 5066] = "SFX_Run_Player_Barbed_Wire";
    Effects[Effects["SFX_Jump_Player_Barbed_Wire"] = 5067] = "SFX_Jump_Player_Barbed_Wire";
    Effects[Effects["SFX_Jump_Player_Metal_Hollow"] = 5068] = "SFX_Jump_Player_Metal_Hollow";
    Effects[Effects["SFX_Run_Player_Metal_Hollow"] = 5069] = "SFX_Run_Player_Metal_Hollow";
    Effects[Effects["SFX_Walk_Player_Metal_Hollow"] = 5070] = "SFX_Walk_Player_Metal_Hollow";
    Effects[Effects["SFX_PLAYER_RX_Breathing_Slow"] = 5071] = "SFX_PLAYER_RX_Breathing_Slow";
    Effects[Effects["SFX_PLAYER_RX_Breathing_Med"] = 5072] = "SFX_PLAYER_RX_Breathing_Med";
    Effects[Effects["SFX_PLAYER_RX_Breathing_Fast"] = 5073] = "SFX_PLAYER_RX_Breathing_Fast";
    Effects[Effects["SFX_PLAYER_RX_Breathing_Exhausted"] = 5074] = "SFX_PLAYER_RX_Breathing_Exhausted";
    Effects[Effects["SFX_Door_Placeable_Metal_Open"] = 5075] = "SFX_Door_Placeable_Metal_Open";
    Effects[Effects["SFX_Door_Placeable_Metal_Close"] = 5076] = "SFX_Door_Placeable_Metal_Close";
    Effects[Effects["SFX_Door_Placeable_Wood_Shack_Open"] = 5077] = "SFX_Door_Placeable_Wood_Shack_Open";
    Effects[Effects["SFX_Door_Placeable_Wood_Shack_Close"] = 5078] = "SFX_Door_Placeable_Wood_Shack_Close";
    Effects[Effects["SFX_Door_Placeable_Wood_Shed_Open"] = 5079] = "SFX_Door_Placeable_Wood_Shed_Open";
    Effects[Effects["SFX_Door_Placeable_Wood_Shed_Close"] = 5080] = "SFX_Door_Placeable_Wood_Shed_Close";
    Effects[Effects["SFX_Gate_Placeable_Metal_Open"] = 5081] = "SFX_Gate_Placeable_Metal_Open";
    Effects[Effects["SFX_Gate_Placeable_Metal_Close"] = 5082] = "SFX_Gate_Placeable_Metal_Close";
    Effects[Effects["SFX_Door_Wood_Restaurant_Open"] = 5083] = "SFX_Door_Wood_Restaurant_Open";
    Effects[Effects["SFX_Door_Wood_Restaurant_Close"] = 5084] = "SFX_Door_Wood_Restaurant_Close";
    Effects[Effects["SFX_Door_Glass_Business_Open"] = 5085] = "SFX_Door_Glass_Business_Open";
    Effects[Effects["SFX_Door_Glass_Business_Close"] = 5086] = "SFX_Door_Glass_Business_Close";
    Effects[Effects["SFX_Door_Metal_Business_Open"] = 5087] = "SFX_Door_Metal_Business_Open";
    Effects[Effects["SFX_Door_Metal_Business_Close"] = 5088] = "SFX_Door_Metal_Business_Close";
    Effects[Effects["SFX_Door_Office_Open"] = 5089] = "SFX_Door_Office_Open";
    Effects[Effects["SFX_Door_Office_Close"] = 5090] = "SFX_Door_Office_Close";
    Effects[Effects["SFX_Door_Wood_Church_Open"] = 5091] = "SFX_Door_Wood_Church_Open";
    Effects[Effects["SFX_Door_Wood_Church_Close"] = 5092] = "SFX_Door_Wood_Church_Close";
    Effects[Effects["SFX_Door_Metal_Freezer_Open"] = 5093] = "SFX_Door_Metal_Freezer_Open";
    Effects[Effects["SFX_Door_Metal_Freezer_Close"] = 5094] = "SFX_Door_Metal_Freezer_Close";
    Effects[Effects["SFX_Door_Metal_Industrial_Open"] = 5095] = "SFX_Door_Metal_Industrial_Open";
    Effects[Effects["SFX_Door_Metal_Industrial_Close"] = 5096] = "SFX_Door_Metal_Industrial_Close";
    Effects[Effects["SFX_Leaderboard_Flip_Start"] = 5098] = "SFX_Leaderboard_Flip_Start";
    Effects[Effects["SFX_Leaderboard_Flip_Stop"] = 5099] = "SFX_Leaderboard_Flip_Stop";
    Effects[Effects["PFX_Impact_BulletLight_Headshot_Human"] = 5100] = "PFX_Impact_BulletLight_Headshot_Human";
    Effects[Effects["PFX_Impact_BulletLight_Headshot_Zombie"] = 5101] = "PFX_Impact_BulletLight_Headshot_Zombie";
    Effects[Effects["VEH_Dust_OffRoader_Water"] = 5103] = "VEH_Dust_OffRoader_Water";
    Effects[Effects["EFX_Waterfall_Dam"] = 5104] = "EFX_Waterfall_Dam";
    Effects[Effects["PFX_Bleeding_Human_severe_loop"] = 5105] = "PFX_Bleeding_Human_severe_loop";
    Effects[Effects["PFX_Bleeding_Human_minor_loop"] = 5106] = "PFX_Bleeding_Human_minor_loop";
    Effects[Effects["PFX_Bleeding_Zombie_minor_loop"] = 5108] = "PFX_Bleeding_Zombie_minor_loop";
    Effects[Effects["PFX_Bleeding_Zombie_moderate_loop"] = 5109] = "PFX_Bleeding_Zombie_moderate_loop";
    Effects[Effects["PFX_Bleeding_Zombie_severe_loop"] = 5110] = "PFX_Bleeding_Zombie_severe_loop";
    Effects[Effects["PFX_Impact_PunjiSticks_Blood"] = 5116] = "PFX_Impact_PunjiSticks_Blood";
    Effects[Effects["PFX_Impact_BulletHeavy_Water"] = 5118] = "PFX_Impact_BulletHeavy_Water";
    Effects[Effects["VEH_Dust_OffRoader_Water_Idle"] = 5119] = "VEH_Dust_OffRoader_Water_Idle";
    Effects[Effects["PFX_Impact_ArrowFlaming_Water"] = 5121] = "PFX_Impact_ArrowFlaming_Water";
    Effects[Effects["PFX_Impact_Fists_Default"] = 5122] = "PFX_Impact_Fists_Default";
    Effects[Effects["SFX_Item_PickUp_BackPack"] = 5139] = "SFX_Item_PickUp_BackPack";
    Effects[Effects["SFX_Item_PickUp_Ammo"] = 5140] = "SFX_Item_PickUp_Ammo";
    Effects[Effects["SFX_Item_PickUp_Arrow"] = 5141] = "SFX_Item_PickUp_Arrow";
    Effects[Effects["SFX_Item_PickUp_Arrows"] = 5142] = "SFX_Item_PickUp_Arrows";
    Effects[Effects["SFX_Item_PickUp_Bat"] = 5143] = "SFX_Item_PickUp_Bat";
    Effects[Effects["SFX_Item_PickUp_Blade"] = 5144] = "SFX_Item_PickUp_Blade";
    Effects[Effects["SFX_Item_PickUp_Blunt_Weapon"] = 5145] = "SFX_Item_PickUp_Blunt_Weapon";
    Effects[Effects["SFX_Item_PickUp_Body_Armor"] = 5146] = "SFX_Item_PickUp_Body_Armor";
    Effects[Effects["SFX_Item_PickUp_Can"] = 5147] = "SFX_Item_PickUp_Can";
    Effects[Effects["SFX_Item_PickUp_Clothing"] = 5148] = "SFX_Item_PickUp_Clothing";
    Effects[Effects["SFX_Item_PickUp_FirstAid"] = 5149] = "SFX_Item_PickUp_FirstAid";
    Effects[Effects["SFX_Item_PickUp_GasCan"] = 5150] = "SFX_Item_PickUp_GasCan";
    Effects[Effects["SFX_Item_PickUp_Generic"] = 5151] = "SFX_Item_PickUp_Generic";
    Effects[Effects["SFX_Item_PickUp_Grenade"] = 5152] = "SFX_Item_PickUp_Grenade";
    Effects[Effects["SFX_Item_PickUp_Helmet"] = 5153] = "SFX_Item_PickUp_Helmet";
    Effects[Effects["SFX_Item_PickUp_Molotov"] = 5154] = "SFX_Item_PickUp_Molotov";
    Effects[Effects["SFX_Item_PickUp_Package"] = 5155] = "SFX_Item_PickUp_Package";
    Effects[Effects["SFX_Item_PickUp_Pistol"] = 5156] = "SFX_Item_PickUp_Pistol";
    Effects[Effects["SFX_Item_PickUp_Rifle"] = 5157] = "SFX_Item_PickUp_Rifle";
    Effects[Effects["SFX_Item_PickUp_WaterBottle"] = 5158] = "SFX_Item_PickUp_WaterBottle";
    Effects[Effects["SFX_Item_PickUp_Milk_Jug"] = 5159] = "SFX_Item_PickUp_Milk_Jug";
    Effects[Effects["EFX_Insects_Cockroaches_01"] = 5126] = "EFX_Insects_Cockroaches_01";
    Effects[Effects["EFX_Insects_Maggots_01"] = 5127] = "EFX_Insects_Maggots_01";
    Effects[Effects["EFX_Insects_Termites_01"] = 5128] = "EFX_Insects_Termites_01";
    Effects[Effects["EFX_Trash_Blowing_01"] = 5129] = "EFX_Trash_Blowing_01";
    Effects[Effects["EFX_Birds_Flying_01"] = 5132] = "EFX_Birds_Flying_01";
    Effects[Effects["EFX_Birds_Flocking_01"] = 5133] = "EFX_Birds_Flocking_01";
    Effects[Effects["EFX_DandelionSeeds_01"] = 5135] = "EFX_DandelionSeeds_01";
    Effects[Effects["EFX_Insects_Butterflies_01"] = 5160] = "EFX_Insects_Butterflies_01";
    Effects[Effects["EFX_Insects_Fireflies_01"] = 5161] = "EFX_Insects_Fireflies_01";
    Effects[Effects["EFX_Insects_Flies_Swarm_01"] = 5162] = "EFX_Insects_Flies_Swarm_01";
    Effects[Effects["EFX_Insects_Gnats_01"] = 5163] = "EFX_Insects_Gnats_01";
    Effects[Effects["EFX_Insects_Mosquitos_01"] = 5164] = "EFX_Insects_Mosquitos_01";
    Effects[Effects["EFX_Insects_Moths_01"] = 5165] = "EFX_Insects_Moths_01";
    Effects[Effects["PFX_Impact_Blade_Metal_Chainlink"] = 5166] = "PFX_Impact_Blade_Metal_Chainlink";
    Effects[Effects["PFX_Impact_Fists_Metal_Chainlink"] = 5167] = "PFX_Impact_Fists_Metal_Chainlink";
    Effects[Effects["PFX_Impact_Knife_Metal_Chainlink"] = 5168] = "PFX_Impact_Knife_Metal_Chainlink";
    Effects[Effects["PFX_Impact_BulletLight_Metal_Chainlink"] = 5169] = "PFX_Impact_BulletLight_Metal_Chainlink";
    Effects[Effects["PFX_Impact_Arrow_Metal_Chainlink"] = 5170] = "PFX_Impact_Arrow_Metal_Chainlink";
    Effects[Effects["PFX_Seasonal_Halloween_Bats_ExplosionSkeleton"] = 5171] = "PFX_Seasonal_Halloween_Bats_ExplosionSkeleton";
    Effects[Effects["PFX_Seasonal_Halloween_Bats_SwarmSkeleton_loop"] = 5172] = "PFX_Seasonal_Halloween_Bats_SwarmSkeleton_loop";
    Effects[Effects["PFX_Damage_Fence_ChainLink_01"] = 5173] = "PFX_Damage_Fence_ChainLink_01";
    Effects[Effects["PFX_Damage_Fence_ChainLink_02"] = 5174] = "PFX_Damage_Fence_ChainLink_02";
    Effects[Effects["PFX_Damage_Fence_ChainLink_Industrial_01"] = 5175] = "PFX_Damage_Fence_ChainLink_Industrial_01";
    Effects[Effects["PFX_Damage_Fence_ChainLink_Industrial_Gate"] = 5176] = "PFX_Damage_Fence_ChainLink_Industrial_Gate";
    Effects[Effects["EFX_FireHydrant_Dripping"] = 5178] = "EFX_FireHydrant_Dripping";
    Effects[Effects["SFX_Bomb_Falling"] = 5179] = "SFX_Bomb_Falling";
    Effects[Effects["PFX_Impact_BulletHeavy_Dirt"] = 5180] = "PFX_Impact_BulletHeavy_Dirt";
    Effects[Effects["PFX_Impact_BulletHeavy_Grass"] = 5181] = "PFX_Impact_BulletHeavy_Grass";
    Effects[Effects["PFX_Impact_BulletHeavy_Wood"] = 5182] = "PFX_Impact_BulletHeavy_Wood";
    Effects[Effects["PFX_Impact_BulletHeavy_Rock"] = 5183] = "PFX_Impact_BulletHeavy_Rock";
    Effects[Effects["PFX_Impact_BulletHeavy_Glass"] = 5185] = "PFX_Impact_BulletHeavy_Glass";
    Effects[Effects["PFX_Impact_BulletHeavy_Metal"] = 5187] = "PFX_Impact_BulletHeavy_Metal";
    Effects[Effects["PFX_Impact_BulletHeavy_Metal_Vehicle"] = 5189] = "PFX_Impact_BulletHeavy_Metal_Vehicle";
    Effects[Effects["PFX_Impact_BulletHeavy_Rubber"] = 5190] = "PFX_Impact_BulletHeavy_Rubber";
    Effects[Effects["PFX_Poison_GasTrap_AOE_01"] = 5191] = "PFX_Poison_GasTrap_AOE_01";
    Effects[Effects["PFX_Trap_Shock_Trigger"] = 5192] = "PFX_Trap_Shock_Trigger";
    Effects[Effects["PFX_Impact_BulletHeavy_Asphalt"] = 5193] = "PFX_Impact_BulletHeavy_Asphalt";
    Effects[Effects["PFX_Impact_Blade_Asphalt"] = 5194] = "PFX_Impact_Blade_Asphalt";
    Effects[Effects["PFX_Char_Zombie_Screamer_Scream"] = 5203] = "PFX_Char_Zombie_Screamer_Scream";
    Effects[Effects["VEH_DeadLoop_ATV"] = 5206] = "VEH_DeadLoop_ATV";
    Effects[Effects["VEH_DeadLoop_OffRoader"] = 5207] = "VEH_DeadLoop_OffRoader";
    Effects[Effects["VEH_DeadLoop_PickupTruck"] = 5208] = "VEH_DeadLoop_PickupTruck";
    Effects[Effects["VEH_DeadLoop_PoliceCar"] = 5209] = "VEH_DeadLoop_PoliceCar";
    Effects[Effects["PFX_Seasonal_Holiday_Snow_skel"] = 5211] = "PFX_Seasonal_Holiday_Snow_skel";
    Effects[Effects["PFX_Seasonal_Holiday_Snow_Explosion"] = 5214] = "PFX_Seasonal_Holiday_Snow_Explosion";
    Effects[Effects["PFX_Trap_Blinded_Trigger"] = 5215] = "PFX_Trap_Blinded_Trigger";
    Effects[Effects["TEST_TimeDilation"] = 5223] = "TEST_TimeDilation";
    Effects[Effects["VEH_Damage_ATV_Stage04"] = 5226] = "VEH_Damage_ATV_Stage04";
    Effects[Effects["VEH_Damage_OffRoader_Stage04"] = 5227] = "VEH_Damage_OffRoader_Stage04";
    Effects[Effects["VEH_Damage_PickupTruck_Stage04"] = 5228] = "VEH_Damage_PickupTruck_Stage04";
    Effects[Effects["VEH_Damage_PoliceCar_Stage04"] = 5229] = "VEH_Damage_PoliceCar_Stage04";
    Effects[Effects["PFX_Impact_BulletLight_Textile"] = 5217] = "PFX_Impact_BulletLight_Textile";
    Effects[Effects["PFX_Impact_BulletHeavy_Textile"] = 5225] = "PFX_Impact_BulletHeavy_Textile";
    Effects[Effects["PFX_Impact_Blade_Textile"] = 5230] = "PFX_Impact_Blade_Textile";
    Effects[Effects["PFX_Impact_Fists_Textile"] = 5231] = "PFX_Impact_Fists_Textile";
    Effects[Effects["PFX_Impact_Fists_Asphalt"] = 5232] = "PFX_Impact_Fists_Asphalt";
    Effects[Effects["PFX_Impact_Arrow_Textile"] = 5236] = "PFX_Impact_Arrow_Textile";
    Effects[Effects["PFX_Impact_Arrow_Asphalt"] = 5237] = "PFX_Impact_Arrow_Asphalt";
    Effects[Effects["PFX_Char_Zombie_Screamer_Ambient"] = 5239] = "PFX_Char_Zombie_Screamer_Ambient";
    Effects[Effects["PFX_Char_Zombie_Exploder_Ambient"] = 5241] = "PFX_Char_Zombie_Exploder_Ambient";
    Effects[Effects["PFX_Char_Zombie_Exploder_Explosion"] = 5242] = "PFX_Char_Zombie_Exploder_Explosion";
    Effects[Effects["PFX_Char_Zombie_Gasser_ExplosionGasCloud"] = 5243] = "PFX_Char_Zombie_Gasser_ExplosionGasCloud";
    Effects[Effects["PFX_Char_Zombie_Gasser_Ambient"] = 5244] = "PFX_Char_Zombie_Gasser_Ambient";
    Effects[Effects["PFX_Char_Zombie_Screamer_Scream_Short"] = 5245] = "PFX_Char_Zombie_Screamer_Scream_Short";
    Effects[Effects["SFX_Player_Collar_Beep"] = 5247] = "SFX_Player_Collar_Beep";
    Effects[Effects["EFX_Light_BusStop"] = 5246] = "EFX_Light_BusStop";
    Effects[Effects["EFX_Light_GasLamp"] = 5251] = "EFX_Light_GasLamp";
    Effects[Effects["EFX_Light_Wall_Alley"] = 5252] = "EFX_Light_Wall_Alley";
    Effects[Effects["PFX_Explosion_OilBarrel_04m"] = 5250] = "PFX_Explosion_OilBarrel_04m";
    Effects[Effects["PFX_Meteor"] = 5253] = "PFX_Meteor";
    Effects[Effects["TEST_UnlitSprites_Fog_Issues"] = 5255] = "TEST_UnlitSprites_Fog_Issues";
    Effects[Effects["SFX_Horn_Truck_Oneshot"] = 5256] = "SFX_Horn_Truck_Oneshot";
    Effects[Effects["EFX_Light_Wall_Alley_Damaged01"] = 5257] = "EFX_Light_Wall_Alley_Damaged01";
    Effects[Effects["EFX_Light_GasLamp_Damaged01_Flickering_Light"] = 5258] = "EFX_Light_GasLamp_Damaged01_Flickering_Light";
    Effects[Effects["EFX_Light_GasLamp_Damaged02_Sporadic_Sparks"] = 5259] = "EFX_Light_GasLamp_Damaged02_Sporadic_Sparks";
    Effects[Effects["PFX_Weapon_Grenade_Screamer_Ambient"] = 5260] = "PFX_Weapon_Grenade_Screamer_Ambient";
    Effects[Effects["PFX_Damage_Light_GasLamp_Lit"] = 5261] = "PFX_Damage_Light_GasLamp_Lit";
    Effects[Effects["PFX_Char_Zombie_Gasser_Explosion"] = 5262] = "PFX_Char_Zombie_Gasser_Explosion";
    Effects[Effects["PFX_Char_Zombie_Gasser_GasCloud"] = 5263] = "PFX_Char_Zombie_Gasser_GasCloud";
    Effects[Effects["SFX_Player_Collar_Activate_Unsafe"] = 5264] = "SFX_Player_Collar_Activate_Unsafe";
    Effects[Effects["SFX_Player_Collar_Deactivate_Safe"] = 5265] = "SFX_Player_Collar_Deactivate_Safe";
    Effects[Effects["PFX_Char_Glasses_Goggles_Glow"] = 5266] = "PFX_Char_Glasses_Goggles_Glow";
    Effects[Effects["PFX_Char_Glasses_Goggles_Glow_Female"] = 5267] = "PFX_Char_Glasses_Goggles_Glow_Female";
    Effects[Effects["PFX_Waypoint_Green"] = 5268] = "PFX_Waypoint_Green";
    Effects[Effects["PFX_Events_GasCloud_WasteManagement_01"] = 5269] = "PFX_Events_GasCloud_WasteManagement_01";
    Effects[Effects["SFX_Grenade_Screamer_Loop"] = 5270] = "SFX_Grenade_Screamer_Loop";
    Effects[Effects["EFX_Fire_OilBarrel"] = 5271] = "EFX_Fire_OilBarrel";
    Effects[Effects["EFX_Fire_OilBarrel_Small"] = 5272] = "EFX_Fire_OilBarrel_Small";
    Effects[Effects["VEH_Helicopter_BladesLightsProjector"] = 5273] = "VEH_Helicopter_BladesLightsProjector";
    Effects[Effects["SFX_Proj_Bullet_By_Generic"] = 5275] = "SFX_Proj_Bullet_By_Generic";
    Effects[Effects["SFX_Proj_Bullet_By_Sniper"] = 5276] = "SFX_Proj_Bullet_By_Sniper";
    Effects[Effects["VEH_Helicopter_ProjectorLights_Green"] = 5278] = "VEH_Helicopter_ProjectorLights_Green";
    Effects[Effects["VEH_Helicopter_ProjectorLights_Red"] = 5279] = "VEH_Helicopter_ProjectorLights_Red";
    Effects[Effects["PFX_Char_Zombie_Screamer_Ambient_ShortBurst"] = 5281] = "PFX_Char_Zombie_Screamer_Ambient_ShortBurst";
    Effects[Effects["SFX_Proj_Arrow_By"] = 5282] = "SFX_Proj_Arrow_By";
    Effects[Effects["SFX_Proj_Fire_Arrow_By"] = 5283] = "SFX_Proj_Fire_Arrow_By";
    Effects[Effects["PFX_Damage_GraveStone_03"] = 5287] = "PFX_Damage_GraveStone_03";
    Effects[Effects["PFX_Damage_GraveStone_02"] = 5288] = "PFX_Damage_GraveStone_02";
    Effects[Effects["PFX_Damage_PatioChair_01"] = 5289] = "PFX_Damage_PatioChair_01";
    Effects[Effects["PFX_MedicalMister_SurroundingMist"] = 5290] = "PFX_MedicalMister_SurroundingMist";
    Effects[Effects["TEST_Bug_TimeScalar"] = 5291] = "TEST_Bug_TimeScalar";
    Effects[Effects["EFX_TV"] = 5292] = "EFX_TV";
    Effects[Effects["PFX_Damage_TV"] = 5293] = "PFX_Damage_TV";
    Effects[Effects["PFX_Damage_Fence_Farm01"] = 5294] = "PFX_Damage_Fence_Farm01";
    Effects[Effects["PFX_Damage_Fence_Farm02"] = 5295] = "PFX_Damage_Fence_Farm02";
    Effects[Effects["PFX_Damage_Fence_ResidTall01"] = 5296] = "PFX_Damage_Fence_ResidTall01";
    Effects[Effects["PFX_Damage_Fence_ResidTallGate"] = 5297] = "PFX_Damage_Fence_ResidTallGate";
    Effects[Effects["PFX_Damage_Building_Interior_WallHole01"] = 5298] = "PFX_Damage_Building_Interior_WallHole01";
    Effects[Effects["PFX_Char_Death_Explosion_Body"] = 5299] = "PFX_Char_Death_Explosion_Body";
    Effects[Effects["PFX_Impact_Explosion_FragGrenade_Asphalt_08m"] = 5300] = "PFX_Impact_Explosion_FragGrenade_Asphalt_08m";
    Effects[Effects["PFX_Impact_Explosion_FragGrenade_Default_08m"] = 5301] = "PFX_Impact_Explosion_FragGrenade_Default_08m";
    Effects[Effects["PFX_Impact_Explosion_Landmine_Asphalt_10m"] = 5303] = "PFX_Impact_Explosion_Landmine_Asphalt_10m";
    Effects[Effects["PFX_Impact_Explosion_Landmine_Default_10m"] = 5304] = "PFX_Impact_Explosion_Landmine_Default_10m";
    Effects[Effects["PFX_Fire_Molotov_Impact"] = 5308] = "PFX_Fire_Molotov_Impact";
    Effects[Effects["PFX_Fire_Molotov_Impact_Water"] = 5309] = "PFX_Fire_Molotov_Impact_Water";
    Effects[Effects["EFX_Light_Residential_Bedroom_LampLarge"] = 5310] = "EFX_Light_Residential_Bedroom_LampLarge";
    Effects[Effects["EFX_Light_Residential_Bedroom_LampSmall"] = 5311] = "EFX_Light_Residential_Bedroom_LampSmall";
    Effects[Effects["EFX_Stoplight_Flashing_All_wStreetlight"] = 5312] = "EFX_Stoplight_Flashing_All_wStreetlight";
    Effects[Effects["EFX_Stoplight_Flashing_All"] = 5313] = "EFX_Stoplight_Flashing_All";
    Effects[Effects["EFX_Stoplight_Flashing_3_w_Streetlight"] = 5314] = "EFX_Stoplight_Flashing_3_w_Streetlight";
    Effects[Effects["EFX_Stoplight_Flashing_2_w_Streetlight"] = 5315] = "EFX_Stoplight_Flashing_2_w_Streetlight";
    Effects[Effects["VEH_Helicopter_BladesLights"] = 5319] = "VEH_Helicopter_BladesLights";
    Effects[Effects["PFX_Impact_Explosion_ExplosiveArrow_Asphalt_08m"] = 5320] = "PFX_Impact_Explosion_ExplosiveArrow_Asphalt_08m";
    Effects[Effects["PFX_Impact_Explosion_ExplosiveArrow_Default_08m"] = 5321] = "PFX_Impact_Explosion_ExplosiveArrow_Default_08m";
    Effects[Effects["PFX_Trail_Projectile_Arrow_Explosive"] = 5322] = "PFX_Trail_Projectile_Arrow_Explosive";
    Effects[Effects["PFX_Damage_Light_StreetLight_Flickering"] = 5323] = "PFX_Damage_Light_StreetLight_Flickering";
    Effects[Effects["PFX_Damage_Light_StreetLight_Sparks"] = 5324] = "PFX_Damage_Light_StreetLight_Sparks";
    Effects[Effects["TEST_LitParticles"] = 5327] = "TEST_LitParticles";
    Effects[Effects["PFX_Impact_Explosion_AirdropBomb_Default_10m"] = 5328] = "PFX_Impact_Explosion_AirdropBomb_Default_10m";
    Effects[Effects["PFX_Impact_Explosion_AirdropBomb_Asphalt_10m"] = 5330] = "PFX_Impact_Explosion_AirdropBomb_Asphalt_10m";
    Effects[Effects["VEH_DeadLoop_ATV_Tire_LeftFront"] = 5334] = "VEH_DeadLoop_ATV_Tire_LeftFront";
    Effects[Effects["VEH_DeadLoop_ATV_Tire_RightFront"] = 5335] = "VEH_DeadLoop_ATV_Tire_RightFront";
    Effects[Effects["VEH_DeadLoop_OffRoader_Tire_LeftFront"] = 5336] = "VEH_DeadLoop_OffRoader_Tire_LeftFront";
    Effects[Effects["VEH_DeadLoop_OffRoader_Tire_Spare"] = 5337] = "VEH_DeadLoop_OffRoader_Tire_Spare";
    Effects[Effects["VEH_DeadLoop_PickupTruck_Tire_LeftFront"] = 5338] = "VEH_DeadLoop_PickupTruck_Tire_LeftFront";
    Effects[Effects["VEH_DeadLoop_PickupTruck_Tire_RightRear"] = 5339] = "VEH_DeadLoop_PickupTruck_Tire_RightRear";
    Effects[Effects["VEH_DeadLoop_PoliceCar_Tire_LeftFront"] = 5340] = "VEH_DeadLoop_PoliceCar_Tire_LeftFront";
    Effects[Effects["PFX_Impact_Arrow_Default"] = 5342] = "PFX_Impact_Arrow_Default";
    Effects[Effects["PFX_Impact_BulletHeavy_Default"] = 5343] = "PFX_Impact_BulletHeavy_Default";
    Effects[Effects["PFX_Damage_Tree_City"] = 5329] = "PFX_Damage_Tree_City";
    Effects[Effects["VEH_Dust_OffRoader_Cardboard"] = 5344] = "VEH_Dust_OffRoader_Cardboard";
    Effects[Effects["VEH_Dust_ATV_Cardboard"] = 5346] = "VEH_Dust_ATV_Cardboard";
    Effects[Effects["EFX_Ceiling_Fan_01"] = 5347] = "EFX_Ceiling_Fan_01";
    Effects[Effects["MAT_VehicleTreads_OffRoader"] = 5349] = "MAT_VehicleTreads_OffRoader";
    Effects[Effects["MAT_VehicleTreads_PickupTruck"] = 5350] = "MAT_VehicleTreads_PickupTruck";
    Effects[Effects["MAT_VehicleTreads_PoliceCar"] = 5351] = "MAT_VehicleTreads_PoliceCar";
    Effects[Effects["MAT_VehicleTreads_ATV"] = 5352] = "MAT_VehicleTreads_ATV";
    Effects[Effects["VEH_Treads_ATV_Dirt"] = 5353] = "VEH_Treads_ATV_Dirt";
    Effects[Effects["VEH_Treads_OffRoader_Dirt"] = 5354] = "VEH_Treads_OffRoader_Dirt";
    Effects[Effects["VEH_Treads_PickupTruck_Dirt"] = 5355] = "VEH_Treads_PickupTruck_Dirt";
    Effects[Effects["VEH_Treads_PoliceCar_Dirt"] = 5356] = "VEH_Treads_PoliceCar_Dirt";
    Effects[Effects["PFX_Trail_Projectile_Grenade"] = 5357] = "PFX_Trail_Projectile_Grenade";
    Effects[Effects["PFX_Trail_Projectile_Molotov"] = 5358] = "PFX_Trail_Projectile_Molotov";
    Effects[Effects["SFX_Veh_ATV_Cable_Release"] = 5359] = "SFX_Veh_ATV_Cable_Release";
    Effects[Effects["EFX_Vent_Fan_Spinning"] = 5361] = "EFX_Vent_Fan_Spinning";
    Effects[Effects["PFX_Impact_ArrowFlaming_Default"] = 5363] = "PFX_Impact_ArrowFlaming_Default";
    Effects[Effects["PFX_Char_Drink_WaterBottle_L"] = 5364] = "PFX_Char_Drink_WaterBottle_L";
    Effects[Effects["PFX_Char_Drink_WaterBottle_R"] = 5365] = "PFX_Char_Drink_WaterBottle_R";
    Effects[Effects["PFX_Impact_ArrowFlaming_Asphalt"] = 5367] = "PFX_Impact_ArrowFlaming_Asphalt";
    Effects[Effects["PFX_Impact_ArrowFlaming_Metal_Chainlink"] = 5368] = "PFX_Impact_ArrowFlaming_Metal_Chainlink";
    Effects[Effects["PFX_Impact_ArrowFlaming_Concrete"] = 5369] = "PFX_Impact_ArrowFlaming_Concrete";
    Effects[Effects["PFX_Impact_ArrowFlaming_Dirt"] = 5370] = "PFX_Impact_ArrowFlaming_Dirt";
    Effects[Effects["PFX_Impact_ArrowFlaming_Flesh"] = 5371] = "PFX_Impact_ArrowFlaming_Flesh";
    Effects[Effects["PFX_Impact_ArrowFlaming_Glass"] = 5372] = "PFX_Impact_ArrowFlaming_Glass";
    Effects[Effects["PFX_Impact_ArrowFlaming_Grass"] = 5373] = "PFX_Impact_ArrowFlaming_Grass";
    Effects[Effects["PFX_Impact_ArrowFlaming_Metal"] = 5374] = "PFX_Impact_ArrowFlaming_Metal";
    Effects[Effects["PFX_Impact_ArrowFlaming_Rock"] = 5376] = "PFX_Impact_ArrowFlaming_Rock";
    Effects[Effects["PFX_Impact_ArrowFlaming_Rubber"] = 5377] = "PFX_Impact_ArrowFlaming_Rubber";
    Effects[Effects["PFX_Impact_ArrowFlaming_Textile"] = 5378] = "PFX_Impact_ArrowFlaming_Textile";
    Effects[Effects["PFX_Impact_ArrowFlaming_Metal_Vehicle"] = 5379] = "PFX_Impact_ArrowFlaming_Metal_Vehicle";
    Effects[Effects["PFX_Impact_ArrowFlaming_Wood"] = 5380] = "PFX_Impact_ArrowFlaming_Wood";
    Effects[Effects["PFX_Impact_ArrowFlaming_Zombie"] = 5381] = "PFX_Impact_ArrowFlaming_Zombie";
    Effects[Effects["TEST_StressTest"] = 5366] = "TEST_StressTest";
    Effects[Effects["PFX_Char_Zombie_Spitter_AttackSpit_Projectile"] = 5382] = "PFX_Char_Zombie_Spitter_AttackSpit_Projectile";
    Effects[Effects["PFX_Char_Zombie_Spitter_AttackSpit_Impact"] = 5384] = "PFX_Char_Zombie_Spitter_AttackSpit_Impact";
    Effects[Effects["PFX_Impact_Knife_Human_ForehandSlash"] = 5387] = "PFX_Impact_Knife_Human_ForehandSlash";
    Effects[Effects["PFX_Impact_Knife_Human_ForwardStab"] = 5388] = "PFX_Impact_Knife_Human_ForwardStab";
    Effects[Effects["PFX_Impact_Knife_Zombie_ForwardStab"] = 5389] = "PFX_Impact_Knife_Zombie_ForwardStab";
    Effects[Effects["PFX_Impact_Knife_Zombie_ForehandSlash"] = 5390] = "PFX_Impact_Knife_Zombie_ForehandSlash";
    Effects[Effects["PFX_Impact_Blunt_Human_Forehand"] = 5391] = "PFX_Impact_Blunt_Human_Forehand";
    Effects[Effects["PFX_Impact_Blunt_Human_ForehandUppercut"] = 5392] = "PFX_Impact_Blunt_Human_ForehandUppercut";
    Effects[Effects["PFX_Impact_Blunt_Zombie_Forehand"] = 5393] = "PFX_Impact_Blunt_Zombie_Forehand";
    Effects[Effects["PFX_Impact_Blunt_Zombie_ForehandUppercut"] = 5394] = "PFX_Impact_Blunt_Zombie_ForehandUppercut";
    Effects[Effects["PFX_Char_Zombie_Feeding_Hands"] = 5461] = "PFX_Char_Zombie_Feeding_Hands";
    Effects[Effects["PFX_Char_Zombie_Feeding_HeadAndHands"] = 5465] = "PFX_Char_Zombie_Feeding_HeadAndHands";
    Effects[Effects["PFX_Char_Zombie_Feeding_Head"] = 5468] = "PFX_Char_Zombie_Feeding_Head";
    Effects[Effects["SFX_WPN_Grenade_Land"] = 5495] = "SFX_WPN_Grenade_Land";
    Effects[Effects["PFX_Char_Zombie_Flinch_Up"] = 5496] = "PFX_Char_Zombie_Flinch_Up";
    Effects[Effects["PFX_Char_Zombie_Flinch_Down"] = 5497] = "PFX_Char_Zombie_Flinch_Down";
    Effects[Effects["PFX_Char_Zombie_Flinch_Left"] = 5498] = "PFX_Char_Zombie_Flinch_Left";
    Effects[Effects["PFX_Char_Zombie_Flinch_Right"] = 5499] = "PFX_Char_Zombie_Flinch_Right";
    Effects[Effects["PFX_Impact_Arrow_Carpet"] = 5395] = "PFX_Impact_Arrow_Carpet";
    Effects[Effects["PFX_Impact_Arrow_Metal_Barbedwire"] = 5396] = "PFX_Impact_Arrow_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_Arrow_Metal_Grate"] = 5397] = "PFX_Impact_Arrow_Metal_Grate";
    Effects[Effects["PFX_Impact_Arrow_Metal_Hard"] = 5398] = "PFX_Impact_Arrow_Metal_Hard";
    Effects[Effects["PFX_Impact_Arrow_Metal_Hollow"] = 5399] = "PFX_Impact_Arrow_Metal_Hollow";
    Effects[Effects["PFX_Impact_ArrowFlaming_Carpet"] = 5400] = "PFX_Impact_ArrowFlaming_Carpet";
    Effects[Effects["PFX_Impact_ArrowFlaming_Metal_Barbedwire"] = 5401] = "PFX_Impact_ArrowFlaming_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_ArrowFlaming_Metal_Grate"] = 5402] = "PFX_Impact_ArrowFlaming_Metal_Grate";
    Effects[Effects["PFX_Impact_ArrowFlaming_Metal_Hard"] = 5403] = "PFX_Impact_ArrowFlaming_Metal_Hard";
    Effects[Effects["PFX_Impact_ArrowFlaming_Metal_Hollow"] = 5404] = "PFX_Impact_ArrowFlaming_Metal_Hollow";
    Effects[Effects["PFX_Impact_BulletHeavy_Carpet"] = 5405] = "PFX_Impact_BulletHeavy_Carpet";
    Effects[Effects["PFX_Impact_BulletHeavy_Concrete"] = 5406] = "PFX_Impact_BulletHeavy_Concrete";
    Effects[Effects["PFX_Impact_BulletHeavy_Metal_Barbedwire"] = 5408] = "PFX_Impact_BulletHeavy_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_BulletHeavy_Metal_Chainlink"] = 5409] = "PFX_Impact_BulletHeavy_Metal_Chainlink";
    Effects[Effects["PFX_Impact_BulletHeavy_Metal_Grate"] = 5410] = "PFX_Impact_BulletHeavy_Metal_Grate";
    Effects[Effects["PFX_Impact_BulletHeavy_Metal_Hard"] = 5411] = "PFX_Impact_BulletHeavy_Metal_Hard";
    Effects[Effects["PFX_Impact_BulletHeavy_Metal_Hollow"] = 5412] = "PFX_Impact_BulletHeavy_Metal_Hollow";
    Effects[Effects["PFX_Impact_BulletHeavy_Flesh"] = 5414] = "PFX_Impact_BulletHeavy_Flesh";
    Effects[Effects["PFX_Impact_BulletHeavy_Zombie"] = 5415] = "PFX_Impact_BulletHeavy_Zombie";
    Effects[Effects["PFX_Impact_BulletLight_Carpet"] = 5416] = "PFX_Impact_BulletLight_Carpet";
    Effects[Effects["PFX_Impact_BulletLight_Metal_Barbedwire"] = 5417] = "PFX_Impact_BulletLight_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_BulletLight_Metal_Grate"] = 5418] = "PFX_Impact_BulletLight_Metal_Grate";
    Effects[Effects["PFX_Impact_BulletLight_Metal_Hard"] = 5419] = "PFX_Impact_BulletLight_Metal_Hard";
    Effects[Effects["PFX_Impact_BulletLight_Metal_Hollow"] = 5420] = "PFX_Impact_BulletLight_Metal_Hollow";
    Effects[Effects["PFX_Impact_BulletShotgun_Carpet"] = 5421] = "PFX_Impact_BulletShotgun_Carpet";
    Effects[Effects["PFX_Impact_BulletShotgun_Metal_Barbedwire"] = 5422] = "PFX_Impact_BulletShotgun_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_BulletShotgun_Metal_Chainlink"] = 5423] = "PFX_Impact_BulletShotgun_Metal_Chainlink";
    Effects[Effects["PFX_Impact_BulletShotgun_Metal_Grate"] = 5424] = "PFX_Impact_BulletShotgun_Metal_Grate";
    Effects[Effects["PFX_Impact_BulletShotgun_Metal_Hard"] = 5425] = "PFX_Impact_BulletShotgun_Metal_Hard";
    Effects[Effects["PFX_Impact_BulletShotgun_Metal_Hollow"] = 5426] = "PFX_Impact_BulletShotgun_Metal_Hollow";
    Effects[Effects["PFX_Impact_Fists_Carpet"] = 5427] = "PFX_Impact_Fists_Carpet";
    Effects[Effects["PFX_Impact_Fists_Metal_Barbedwire"] = 5428] = "PFX_Impact_Fists_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_Fists_Metal_Hard"] = 5429] = "PFX_Impact_Fists_Metal_Hard";
    Effects[Effects["PFX_Impact_Fists_Metal_Hollow"] = 5430] = "PFX_Impact_Fists_Metal_Hollow";
    Effects[Effects["PFX_Impact_Guitar_Asphalt"] = 5431] = "PFX_Impact_Guitar_Asphalt";
    Effects[Effects["PFX_Impact_Guitar_Carpet"] = 5432] = "PFX_Impact_Guitar_Carpet";
    Effects[Effects["PFX_Impact_Guitar_Metal_Chainlink"] = 5433] = "PFX_Impact_Guitar_Metal_Chainlink";
    Effects[Effects["PFX_Impact_Guitar_Metal_Hard"] = 5434] = "PFX_Impact_Guitar_Metal_Hard";
    Effects[Effects["PFX_Impact_Guitar_Metal_Hollow"] = 5435] = "PFX_Impact_Guitar_Metal_Hollow";
    Effects[Effects["PFX_Impact_Guitar_Metal_Barbedwire"] = 5436] = "PFX_Impact_Guitar_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_Guitar_Textile"] = 5437] = "PFX_Impact_Guitar_Textile";
    Effects[Effects["PFX_Impact_Guitar_Default"] = 5442] = "PFX_Impact_Guitar_Default";
    Effects[Effects["PFX_Impact_Blade_Carpet"] = 5443] = "PFX_Impact_Blade_Carpet";
    Effects[Effects["PFX_Impact_Blade_Metal_Barbedwire"] = 5444] = "PFX_Impact_Blade_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_Blade_Metal_Hard"] = 5445] = "PFX_Impact_Blade_Metal_Hard";
    Effects[Effects["PFX_Impact_Blade_Metal_Hollow"] = 5446] = "PFX_Impact_Blade_Metal_Hollow";
    Effects[Effects["PFX_Impact_Blade_Metal_Grate"] = 5447] = "PFX_Impact_Blade_Metal_Grate";
    Effects[Effects["PFX_Impact_Knife_Carpet"] = 5448] = "PFX_Impact_Knife_Carpet";
    Effects[Effects["PFX_Impact_Knife_Metal_Barbedwire"] = 5449] = "PFX_Impact_Knife_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_Knife_Metal_Hard"] = 5450] = "PFX_Impact_Knife_Metal_Hard";
    Effects[Effects["PFX_Impact_Knife_Metal_Hollow"] = 5451] = "PFX_Impact_Knife_Metal_Hollow";
    Effects[Effects["PFX_Impact_Knife_Textile"] = 5452] = "PFX_Impact_Knife_Textile";
    Effects[Effects["PFX_Impact_Torch_Carpet"] = 5453] = "PFX_Impact_Torch_Carpet";
    Effects[Effects["PFX_Impact_Torch_Default"] = 5454] = "PFX_Impact_Torch_Default";
    Effects[Effects["PFX_Impact_Torch_Metal_Hard"] = 5455] = "PFX_Impact_Torch_Metal_Hard";
    Effects[Effects["PFX_Impact_Torch_Metal_Hollow"] = 5456] = "PFX_Impact_Torch_Metal_Hollow";
    Effects[Effects["PFX_Impact_Torch_Metal_Barbedwire"] = 5457] = "PFX_Impact_Torch_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_Torch_Metal_Grate"] = 5458] = "PFX_Impact_Torch_Metal_Grate";
    Effects[Effects["PFX_Impact_Torch_Metal_Chainlink"] = 5459] = "PFX_Impact_Torch_Metal_Chainlink";
    Effects[Effects["PFX_Impact_Blunt_Asphalt"] = 5460] = "PFX_Impact_Blunt_Asphalt";
    Effects[Effects["PFX_Impact_Blunt_Concrete"] = 5462] = "PFX_Impact_Blunt_Concrete";
    Effects[Effects["PFX_Impact_Blunt_Carpet"] = 5463] = "PFX_Impact_Blunt_Carpet";
    Effects[Effects["PFX_Impact_Blunt_Dirt"] = 5464] = "PFX_Impact_Blunt_Dirt";
    Effects[Effects["PFX_Impact_Blunt_Flesh"] = 5467] = "PFX_Impact_Blunt_Flesh";
    Effects[Effects["PFX_Impact_Blunt_Zombie"] = 5469] = "PFX_Impact_Blunt_Zombie";
    Effects[Effects["PFX_Impact_Blunt_Metal"] = 5470] = "PFX_Impact_Blunt_Metal";
    Effects[Effects["PFX_Impact_Blunt_Metal_Barbedwire"] = 5471] = "PFX_Impact_Blunt_Metal_Barbedwire";
    Effects[Effects["PFX_Impact_Blunt_Metal_Chainlink"] = 5472] = "PFX_Impact_Blunt_Metal_Chainlink";
    Effects[Effects["PFX_Impact_Blunt_Metal_Grate"] = 5473] = "PFX_Impact_Blunt_Metal_Grate";
    Effects[Effects["PFX_Impact_Blunt_Metal_Hard"] = 5474] = "PFX_Impact_Blunt_Metal_Hard";
    Effects[Effects["PFX_Impact_Blunt_Metal_Hollow"] = 5475] = "PFX_Impact_Blunt_Metal_Hollow";
    Effects[Effects["PFX_Impact_Blunt_Metal_Vehicle"] = 5476] = "PFX_Impact_Blunt_Metal_Vehicle";
    Effects[Effects["PFX_Impact_Blunt_Rock"] = 5477] = "PFX_Impact_Blunt_Rock";
    Effects[Effects["PFX_Impact_Blunt_Textile"] = 5478] = "PFX_Impact_Blunt_Textile";
    Effects[Effects["PFX_Impact_Blunt_Water"] = 5479] = "PFX_Impact_Blunt_Water";
    Effects[Effects["PFX_Impact_Blunt_Wood"] = 5480] = "PFX_Impact_Blunt_Wood";
    Effects[Effects["PFX_Impact_Blunt_Default"] = 5481] = "PFX_Impact_Blunt_Default";
    Effects[Effects["PFX_Impact_Blade_Default"] = 5482] = "PFX_Impact_Blade_Default";
    Effects[Effects["PFX_Impact_Knife_Metal_Grate"] = 5483] = "PFX_Impact_Knife_Metal_Grate";
    Effects[Effects["PFX_Impact_Fists_Metal_Grate"] = 5484] = "PFX_Impact_Fists_Metal_Grate";
    Effects[Effects["PFX_Impact_Guitar_Metal_Grate"] = 5485] = "PFX_Impact_Guitar_Metal_Grate";
    Effects[Effects["PFX_Impact_BulletShotgun_Textile"] = 5486] = "PFX_Impact_BulletShotgun_Textile";
    Effects[Effects["PFX_Impact_Knife_Default"] = 5487] = "PFX_Impact_Knife_Default";
    Effects[Effects["PFX_Impact_Knife_Asphalt"] = 5488] = "PFX_Impact_Knife_Asphalt";
    Effects[Effects["PFX_Impact_Blunt_Glass"] = 5489] = "PFX_Impact_Blunt_Glass";
    Effects[Effects["PFX_Impact_Blunt_Grass"] = 5490] = "PFX_Impact_Blunt_Grass";
    Effects[Effects["PFX_Impact_Torch_Asphalt"] = 5491] = "PFX_Impact_Torch_Asphalt";
    Effects[Effects["PFX_Impact_Torch_Textile"] = 5492] = "PFX_Impact_Torch_Textile";
    Effects[Effects["PFX_Impact_Fists_Water"] = 5493] = "PFX_Impact_Fists_Water";
    Effects[Effects["MAT_Blunt"] = 5494] = "MAT_Blunt";
    Effects[Effects["PFX_Damage_Light_StreetLightParking_Flickering"] = 5501] = "PFX_Damage_Light_StreetLightParking_Flickering";
    Effects[Effects["PFX_Damage_Light_StreetLightParking_Sparks"] = 5502] = "PFX_Damage_Light_StreetLightParking_Sparks";
    Effects[Effects["EFX_Light_StreetLightParking"] = 5504] = "EFX_Light_StreetLightParking";
    Effects[Effects["PFX_Impact_Blunt_Zombie_Backhand"] = 5505] = "PFX_Impact_Blunt_Zombie_Backhand";
    Effects[Effects["PFX_Impact_Blunt_Human_Backhand"] = 5506] = "PFX_Impact_Blunt_Human_Backhand";
    Effects[Effects["MAT_Blunt_ForehandChop"] = 5508] = "MAT_Blunt_ForehandChop";
    Effects[Effects["MAT_Blunt_BackhandChop"] = 5509] = "MAT_Blunt_BackhandChop";
    Effects[Effects["MAT_Blunt_ForehandUppercut"] = 5510] = "MAT_Blunt_ForehandUppercut";
    Effects[Effects["MAT_Blunt_BackhandSlash"] = 5512] = "MAT_Blunt_BackhandSlash";
    Effects[Effects["MAT_Blunt_ForehandSlash"] = 5513] = "MAT_Blunt_ForehandSlash";
    Effects[Effects["MAT_Knife_ForehandSlash"] = 5514] = "MAT_Knife_ForehandSlash";
    Effects[Effects["MAT_Knife_ForehandChop"] = 5515] = "MAT_Knife_ForehandChop";
    Effects[Effects["MAT_Knife_ForewardStab"] = 5516] = "MAT_Knife_ForewardStab";
    Effects[Effects["EFX_Insects_Flies_Large_01"] = 5517] = "EFX_Insects_Flies_Large_01";
    Effects[Effects["EFX_Insects_Flies_Small_01"] = 5518] = "EFX_Insects_Flies_Small_01";
    Effects[Effects["EFX_Insects_Butterflies_Yellow_Small_01"] = 5519] = "EFX_Insects_Butterflies_Yellow_Small_01";
    Effects[Effects["EFX_Insects_Flies_Small_02"] = 5520] = "EFX_Insects_Flies_Small_02";
    Effects[Effects["EFX_Props_Commercial_Props_FreshFoodWallCoolers"] = 5523] = "EFX_Props_Commercial_Props_FreshFoodWallCoolers";
    Effects[Effects["EFX_Props_Farm_Props_Crops_GreenPlantsRow_Row01"] = 5524] = "EFX_Props_Farm_Props_Crops_GreenPlantsRow_Row01";
    Effects[Effects["EFX_Props_Farm_Props_Remains_Livestock01"] = 5525] = "EFX_Props_Farm_Props_Remains_Livestock01";
    Effects[Effects["EFX_Props_Farm_Props_Remains_Livestock02"] = 5526] = "EFX_Props_Farm_Props_Remains_Livestock02";
    Effects[Effects["EFX_Light_Commercial_Props_PendantLighting"] = 5527] = "EFX_Light_Commercial_Props_PendantLighting";
    Effects[Effects["EFX_Props_Light_Dam_Props_LightFlood"] = 5528] = "EFX_Props_Light_Dam_Props_LightFlood";
    Effects[Effects["EFX_Props_Light_Dam_Props_LightLamp"] = 5529] = "EFX_Props_Light_Dam_Props_LightLamp";
    Effects[Effects["EFX_Props_Light_Dam_Props_LightWall"] = 5530] = "EFX_Props_Light_Dam_Props_LightWall";
    Effects[Effects["EFX_Props_Light_FireStation_Structures_Buildings_FireHouse01_AddOnGarage01_Ext"] = 5531] = "EFX_Props_Light_FireStation_Structures_Buildings_FireHouse01_AddOnGarage01_Ext";
    Effects[Effects["EFX_Props_Light_FireStation_Structures_Buildings_FireHouse01_Ext"] = 5532] = "EFX_Props_Light_FireStation_Structures_Buildings_FireHouse01_Ext";
    Effects[Effects["PFX_Trail_Projectile_DragonBreath"] = 5521] = "PFX_Trail_Projectile_DragonBreath";
    Effects[Effects["PFX_Mzl_DragonBreath"] = 5522] = "PFX_Mzl_DragonBreath";
    Effects[Effects["EFX_Props_Light_Dam_Props_LightEmergency"] = 5533] = "EFX_Props_Light_Dam_Props_LightEmergency";
    Effects[Effects["EFX_Insects_Flies_Tiny_01"] = 5535] = "EFX_Insects_Flies_Tiny_01";
    Effects[Effects["EFX_Waterfall_Dam_Z2"] = 5538] = "EFX_Waterfall_Dam_Z2";
    Effects[Effects["EFX_Insects_Flies_Box"] = 5537] = "EFX_Insects_Flies_Box";
    Effects[Effects["PFX_Char_Zombie_Spitter_Ambient"] = 5541] = "PFX_Char_Zombie_Spitter_Ambient";
    Effects[Effects["PFX_Char_Zombie_Spitter_Explosion"] = 5542] = "PFX_Char_Zombie_Spitter_Explosion";
    Effects[Effects["PFX_Char_Zombie_Spitter_ExplosionGasCloud"] = 5543] = "PFX_Char_Zombie_Spitter_ExplosionGasCloud";
    Effects[Effects["PFX_Char_Zombie_Spitter_GasCloud"] = 5544] = "PFX_Char_Zombie_Spitter_GasCloud";
    Effects[Effects["PFX_Char_Zombie_Spitter_PowerUp"] = 5549] = "PFX_Char_Zombie_Spitter_PowerUp";
    Effects[Effects["PFX_Seasonal_Fireworks_TripleBurst"] = 5551] = "PFX_Seasonal_Fireworks_TripleBurst";
    Effects[Effects["EFX_Props_Light_City_Structures_Blockade_Floodlight01"] = 5552] = "EFX_Props_Light_City_Structures_Blockade_Floodlight01";
    Effects[Effects["EFX_Props_Common_Props_RadioTower01"] = 5553] = "EFX_Props_Common_Props_RadioTower01";
    Effects[Effects["EFX_Props_City_Props_Tree01"] = 5554] = "EFX_Props_City_Props_Tree01";
    Effects[Effects["SFX_VEH_Offroader_Horn"] = 5556] = "SFX_VEH_Offroader_Horn";
    Effects[Effects["SFX_VEH_Pickup_Truck_Horn"] = 5557] = "SFX_VEH_Pickup_Truck_Horn";
    Effects[Effects["SFX_VEH_Police_Car_Horn"] = 5558] = "SFX_VEH_Police_Car_Horn";
    Effects[Effects["SFX_VEH_ATV_Horn"] = 5559] = "SFX_VEH_ATV_Horn";
    Effects[Effects["EFX_Trash_Blowing_01_loop"] = 5560] = "EFX_Trash_Blowing_01_loop";
    Effects[Effects["SFX_VEH_ATV_Hotwire"] = 5561] = "SFX_VEH_ATV_Hotwire";
    Effects[Effects["SFX_VEH_OffRoader_Hotwire"] = 5562] = "SFX_VEH_OffRoader_Hotwire";
    Effects[Effects["SFX_VEH_Pickup_Truck_Hotwire"] = 5563] = "SFX_VEH_Pickup_Truck_Hotwire";
    Effects[Effects["SFX_VEH_Police_Car_Hotwire"] = 5564] = "SFX_VEH_Police_Car_Hotwire";
    Effects[Effects["EFX_Props_Dam_WaterValve_Drip"] = 5555] = "EFX_Props_Dam_WaterValve_Drip";
    Effects[Effects["EFX_Props_Dam_ControlWall"] = 5565] = "EFX_Props_Dam_ControlWall";
    Effects[Effects["EFX_Props_Dam_DeskStation"] = 5566] = "EFX_Props_Dam_DeskStation";
    Effects[Effects["EFX_Props_Light_Campground_Props_Lantern01"] = 5568] = "EFX_Props_Light_Campground_Props_Lantern01";
    Effects[Effects["TEST_WeaponFX_Fire"] = 5571] = "TEST_WeaponFX_Fire";
    Effects[Effects["EFX_Props_Farm_Props_WeatherVane"] = 5572] = "EFX_Props_Farm_Props_WeatherVane";
    Effects[Effects["EFX_Leaves_Falling_01_10m_loop"] = 5573] = "EFX_Leaves_Falling_01_10m_loop";
    Effects[Effects["EFX_Leaves_Falling_01_30m_loop"] = 5574] = "EFX_Leaves_Falling_01_30m_loop";
    Effects[Effects["EFX_Leaves_Falling_02_10m_loop"] = 5575] = "EFX_Leaves_Falling_02_10m_loop";
    Effects[Effects["EFX_Leaves_Falling_02_30m_loop"] = 5576] = "EFX_Leaves_Falling_02_30m_loop";
    Effects[Effects["PFX_Damage_Paper_Mail_CommunalMailbox"] = 5578] = "PFX_Damage_Paper_Mail_CommunalMailbox";
    Effects[Effects["PFX_Damage_Paper_Mail_PostalBox"] = 5579] = "PFX_Damage_Paper_Mail_PostalBox";
    Effects[Effects["PFX_Damage_Paper_Newspapers_Vending"] = 5580] = "PFX_Damage_Paper_Newspapers_Vending";
    Effects[Effects["PFX_Damage_Paper_Trash_CityGarbageCan_01"] = 5581] = "PFX_Damage_Paper_Trash_CityGarbageCan_01";
    Effects[Effects["PFX_Damage_Paper_Trash_MunicipalGarbageCan"] = 5582] = "PFX_Damage_Paper_Trash_MunicipalGarbageCan";
    Effects[Effects["EFX_Props_Hospital_Fountain_Water"] = 5583] = "EFX_Props_Hospital_Fountain_Water";
    Effects[Effects["PFX_Damage_Paper_Trash_CityGarbageCan_02"] = 5584] = "PFX_Damage_Paper_Trash_CityGarbageCan_02";
    Effects[Effects["PFX_Damage_TrafficSigns_Highway"] = 5585] = "PFX_Damage_TrafficSigns_Highway";
    Effects[Effects["PFX_Damage_TrafficSigns_Highway_TwoPosts"] = 5586] = "PFX_Damage_TrafficSigns_Highway_TwoPosts";
    Effects[Effects["PFX_Damage_TrafficSigns_StopSign"] = 5587] = "PFX_Damage_TrafficSigns_StopSign";
    Effects[Effects["PFX_Damage_TrafficSigns_StreetSign"] = 5588] = "PFX_Damage_TrafficSigns_StreetSign";
    Effects[Effects["PFX_Damage_TrafficSigns_Town"] = 5589] = "PFX_Damage_TrafficSigns_Town";
    Effects[Effects["PFX_Damage_TrafficSigns_Highway_TwoPosts_Persist"] = 5590] = "PFX_Damage_TrafficSigns_Highway_TwoPosts_Persist";
    Effects[Effects["PFX_Damage_TrafficSigns_StopSign_Persist"] = 5591] = "PFX_Damage_TrafficSigns_StopSign_Persist";
    Effects[Effects["PFX_Damage_TrafficSigns_Highway_Persist"] = 5592] = "PFX_Damage_TrafficSigns_Highway_Persist";
    Effects[Effects["PFX_Damage_TrafficSigns_Town_Persist"] = 5593] = "PFX_Damage_TrafficSigns_Town_Persist";
    Effects[Effects["PFX_Damage_TrafficSigns_StreetSign_Persist"] = 5594] = "PFX_Damage_TrafficSigns_StreetSign_Persist";
    Effects[Effects["EFX_Birds_Swarming_01"] = 5595] = "EFX_Birds_Swarming_01";
    Effects[Effects["PFX_Damage_ParkingMeter"] = 5596] = "PFX_Damage_ParkingMeter";
    Effects[Effects["PFX_Damage_TrafficSigns_StreetSign_StreetName"] = 5597] = "PFX_Damage_TrafficSigns_StreetSign_StreetName";
    Effects[Effects["PFX_Damage_TrafficSigns_Highway_Wood_TwoPosts"] = 5598] = "PFX_Damage_TrafficSigns_Highway_Wood_TwoPosts";
    Effects[Effects["PFX_Damage_Tent_Campground"] = 5600] = "PFX_Damage_Tent_Campground";
    Effects[Effects["PFX_Damage_Light_Commercial_Props_PendantLighting"] = 5601] = "PFX_Damage_Light_Commercial_Props_PendantLighting";
    Effects[Effects["PFX_Damage_Fence_Farm01_Post"] = 5602] = "PFX_Damage_Fence_Farm01_Post";
    Effects[Effects["PFX_Damage_Cooler_Campground"] = 5603] = "PFX_Damage_Cooler_Campground";
    Effects[Effects["PFX_Damage_Barrels_CabinBasket01"] = 5604] = "PFX_Damage_Barrels_CabinBasket01";
    Effects[Effects["PFX_Damage_TableAndChairs_Campground_FoldingChair"] = 5605] = "PFX_Damage_TableAndChairs_Campground_FoldingChair";
    Effects[Effects["PFX_Damage_TableAndChairs_Church_FoldingChair"] = 5606] = "PFX_Damage_TableAndChairs_Church_FoldingChair";
    Effects[Effects["PFX_Damage_TableAndChairs_Residential_PatioChair"] = 5607] = "PFX_Damage_TableAndChairs_Residential_PatioChair";
    Effects[Effects["PFX_Damage_TableAndChairs_Residential_PatioTable"] = 5608] = "PFX_Damage_TableAndChairs_Residential_PatioTable";
    Effects[Effects["PFX_Damage_TableAndChairs_Restaurant_CafeChair"] = 5609] = "PFX_Damage_TableAndChairs_Restaurant_CafeChair";
    Effects[Effects["PFX_Damage_TableAndChairs_Restaurant_CafeTable"] = 5610] = "PFX_Damage_TableAndChairs_Restaurant_CafeTable";
    Effects[Effects["PFX_Damage_TableAndChairs_Restaurant_CafeUmbrella"] = 5611] = "PFX_Damage_TableAndChairs_Restaurant_CafeUmbrella";
    Effects[Effects["PFX_Damage_TableAndChairs_Restaurant_SquareChair"] = 5612] = "PFX_Damage_TableAndChairs_Restaurant_SquareChair";
    Effects[Effects["PFX_Impact_Katana_Poison"] = 5613] = "PFX_Impact_Katana_Poison";
    Effects[Effects["PFX_Damage_Z1_Mailbox"] = 5614] = "PFX_Damage_Z1_Mailbox";
    Effects[Effects["PFX_Damage_Z1_BarbedWireFence1x1"] = 5615] = "PFX_Damage_Z1_BarbedWireFence1x1";
    Effects[Effects["PFX_Damage_Z1_BarbedWireFence1x2"] = 5616] = "PFX_Damage_Z1_BarbedWireFence1x2";
    Effects[Effects["PFX_Damage_Barriers_IndustrialCrowdBarrier"] = 5617] = "PFX_Damage_Barriers_IndustrialCrowdBarrier";
    Effects[Effects["PFX_Damage_Z1_GarbageCan_02"] = 5618] = "PFX_Damage_Z1_GarbageCan_02";
    Effects[Effects["PFX_Damage_Crate_Industrial"] = 5619] = "PFX_Damage_Crate_Industrial";
    Effects[Effects["PFX_Damage_Light_Bedroom_LampSmall"] = 5620] = "PFX_Damage_Light_Bedroom_LampSmall";
    Effects[Effects["PFX_Damage_Light_Bedroom_LampTall"] = 5621] = "PFX_Damage_Light_Bedroom_LampTall";
    Effects[Effects["PFX_Damage_Light_City_BlockadeFloodlight"] = 5622] = "PFX_Damage_Light_City_BlockadeFloodlight";
    Effects[Effects["PFX_Damage_Light_Dam_EmergencyLight"] = 5623] = "PFX_Damage_Light_Dam_EmergencyLight";
    Effects[Effects["PFX_Damage_Light_Dam_FloodLight"] = 5624] = "PFX_Damage_Light_Dam_FloodLight";
    Effects[Effects["PFX_Damage_Light_Dam_LampLight"] = 5625] = "PFX_Damage_Light_Dam_LampLight";
    Effects[Effects["PFX_Damage_Light_Dam_WallLight"] = 5626] = "PFX_Damage_Light_Dam_WallLight";
    Effects[Effects["EFX_BoD_Helicopter_Blades"] = 5627] = "EFX_BoD_Helicopter_Blades";
    Effects[Effects["EFX_BoD_Helicopter_DustKickUp"] = 5628] = "EFX_BoD_Helicopter_DustKickUp";
    Effects[Effects["PFX_Damage_ShoppingCart"] = 5629] = "PFX_Damage_ShoppingCart";
    Effects[Effects["PFX_Damage_RoadBlock_Gate_Military"] = 5630] = "PFX_Damage_RoadBlock_Gate_Military";
    Effects[Effects["EFX_CrateOpening_Fanfare_Blue"] = 5631] = "EFX_CrateOpening_Fanfare_Blue";
    Effects[Effects["EFX_CrateOpening_Fanfare_Green"] = 5632] = "EFX_CrateOpening_Fanfare_Green";
    Effects[Effects["EFX_CrateOpening_Fanfare_Gold"] = 5633] = "EFX_CrateOpening_Fanfare_Gold";
    Effects[Effects["EFX_CrateOpening_Fanfare_Purple"] = 5634] = "EFX_CrateOpening_Fanfare_Purple";
    Effects[Effects["PFX_Damage_Fence_ChainLink_High1x1Corner"] = 5635] = "PFX_Damage_Fence_ChainLink_High1x1Corner";
    Effects[Effects["PFX_Damage_Fence_ChainLink_High1x1"] = 5636] = "PFX_Damage_Fence_ChainLink_High1x1";
    Effects[Effects["PFX_Damage_Fence_ChainLink_High1x2"] = 5637] = "PFX_Damage_Fence_ChainLink_High1x2";
    Effects[Effects["EFX_BoD_Fire_Plume_Huge_Line"] = 5638] = "EFX_BoD_Fire_Plume_Huge_Line";
    Effects[Effects["EFX_BoD_Fire_Plume_Large"] = 5639] = "EFX_BoD_Fire_Plume_Large";
    Effects[Effects["EFX_BoD_Fire_Plume_Medium"] = 5640] = "EFX_BoD_Fire_Plume_Medium";
    Effects[Effects["EFX_BoD_Steam_01"] = 5641] = "EFX_BoD_Steam_01";
    Effects[Effects["EFX_BoD_Steam_02"] = 5642] = "EFX_BoD_Steam_02";
    Effects[Effects["PFX_Damage_RoadBlock_WoodPost_Military"] = 5643] = "PFX_Damage_RoadBlock_WoodPost_Military";
    Effects[Effects["EFX_CrateOpening_CrateSpawn"] = 5644] = "EFX_CrateOpening_CrateSpawn";
    Effects[Effects["EFX_Light_Hospital_CeilingLight"] = 5645] = "EFX_Light_Hospital_CeilingLight";
    Effects[Effects["EFX_CrateOpening_Crate_Ambient"] = 5648] = "EFX_CrateOpening_Crate_Ambient";
    Effects[Effects["EFX_BoD_SnowGust_01"] = 5649] = "EFX_BoD_SnowGust_01";
    Effects[Effects["SFX_AMB_Emit_Vending_Machine"] = 5650] = "SFX_AMB_Emit_Vending_Machine";
    Effects[Effects["EFX_BoD_C130_DustKickUp"] = 5651] = "EFX_BoD_C130_DustKickUp";
    Effects[Effects["EFX_BoD_C130_Blades"] = 5652] = "EFX_BoD_C130_Blades";
    Effects[Effects["EFX_BoD_Snow_Blowing_Slow_01"] = 5653] = "EFX_BoD_Snow_Blowing_Slow_01";
    Effects[Effects["EFX_BoD_Grass_Blowing_Slow_01"] = 5655] = "EFX_BoD_Grass_Blowing_Slow_01";
    Effects[Effects["EFX_BoD_Blood_Splatter"] = 5656] = "EFX_BoD_Blood_Splatter";
    Effects[Effects["EFX_BoD_RunwayLights"] = 5657] = "EFX_BoD_RunwayLights";
    Effects[Effects["EFX_BoD_RunwayLights_Start"] = 5658] = "EFX_BoD_RunwayLights_Start";
    Effects[Effects["EFX_Windmiill_Spinning"] = 5659] = "EFX_Windmiill_Spinning";
    Effects[Effects["PFX_Damage_Light_Campground_Lantern01"] = 5660] = "PFX_Damage_Light_Campground_Lantern01";
    Effects[Effects["EFX_CrateOpening_CrateDespawn"] = 5661] = "EFX_CrateOpening_CrateDespawn";
    Effects[Effects["PFX_Damage_Fence_ChainLink_Industrial_Corner"] = 5662] = "PFX_Damage_Fence_ChainLink_Industrial_Corner";
    Effects[Effects["PFX_Damage_Fence_ChainLink_Industrial_Breach"] = 5663] = "PFX_Damage_Fence_ChainLink_Industrial_Breach";
    Effects[Effects["PFX_Damage_Fence_ChainLink_Industrial_Straight"] = 5664] = "PFX_Damage_Fence_ChainLink_Industrial_Straight";
    Effects[Effects["PFX_Damage_Fence_ChainLink_Industrial_Post"] = 5665] = "PFX_Damage_Fence_ChainLink_Industrial_Post";
    Effects[Effects["PFX_CrateItem_Helmet_DemonMask_Green"] = 5668] = "PFX_CrateItem_Helmet_DemonMask_Green";
    Effects[Effects["PFX_CrateItem_Helmet_DemonMask_Orange"] = 5669] = "PFX_CrateItem_Helmet_DemonMask_Orange";
})(Effects || (exports.Effects = Effects = {}));
var HealTypes;
(function (HealTypes) {
    HealTypes[HealTypes["OTHER"] = 1] = "OTHER";
    HealTypes[HealTypes["BANDAGE"] = 2] = "BANDAGE";
    HealTypes[HealTypes["MEDKIT"] = 3] = "MEDKIT";
})(HealTypes || (exports.HealTypes = HealTypes = {}));
var ItemTypes;
(function (ItemTypes) {
    ItemTypes[ItemTypes["WEAPON"] = 20] = "WEAPON";
    ItemTypes[ItemTypes["CONTAINER"] = 34] = "CONTAINER";
})(ItemTypes || (exports.ItemTypes = ItemTypes = {}));
var AbilityIds;
(function (AbilityIds) {
    AbilityIds[AbilityIds["HEADLIGHTS_OFFROADER"] = 99998] = "HEADLIGHTS_OFFROADER";
    AbilityIds[AbilityIds["MOTOR_OFFROADER"] = 1111153] = "MOTOR_OFFROADER";
    AbilityIds[AbilityIds["TURBO_OFFROADER"] = 1111141] = "TURBO_OFFROADER";
    AbilityIds[AbilityIds["TURBO_PICKP_TRUCK"] = 1111292] = "TURBO_PICKP_TRUCK";
    AbilityIds[AbilityIds["TURBO_POLICE_CAR"] = 1111294] = "TURBO_POLICE_CAR";
    AbilityIds[AbilityIds["TURBO_ATV"] = 1111611] = "TURBO_ATV";
    AbilityIds[AbilityIds["ABILITY_POLICE_CAR"] = 1111301] = "ABILITY_POLICE_CAR";
    AbilityIds[AbilityIds["NV_GOGGLES"] = 1111272] = "NV_GOGGLES";
})(AbilityIds || (exports.AbilityIds = AbilityIds = {}));
var VehicleEffects;
(function (VehicleEffects) {
    VehicleEffects[VehicleEffects["MOTOR_RUN_OFFROADER_1"] = 90001] = "MOTOR_RUN_OFFROADER_1";
    VehicleEffects[VehicleEffects["MOTOR_RUN_OFFROADER_2"] = 100042] = "MOTOR_RUN_OFFROADER_2";
    VehicleEffects[VehicleEffects["TURBO_PICKUP_TRUCK"] = 90068] = "TURBO_PICKUP_TRUCK";
    VehicleEffects[VehicleEffects["TURBO_OFFROADER"] = 90000] = "TURBO_OFFROADER";
    VehicleEffects[VehicleEffects["TURBO_POLICE_CAR"] = 90069] = "TURBO_POLICE_CAR";
    VehicleEffects[VehicleEffects["TURBO_ATV"] = 90193] = "TURBO_ATV";
})(VehicleEffects || (exports.VehicleEffects = VehicleEffects = {}));
var MaterialTypes;
(function (MaterialTypes) {
    MaterialTypes[MaterialTypes["DEFAULT"] = 0] = "DEFAULT";
    MaterialTypes[MaterialTypes["METAL"] = 2] = "METAL";
    MaterialTypes[MaterialTypes["FLESH"] = 3] = "FLESH";
    MaterialTypes[MaterialTypes["METAL_HARD"] = 4] = "METAL_HARD";
    MaterialTypes[MaterialTypes["WOOD"] = 5] = "WOOD";
    MaterialTypes[MaterialTypes["GLASS"] = 6] = "GLASS";
    MaterialTypes[MaterialTypes["GRASS"] = 7] = "GRASS";
    MaterialTypes[MaterialTypes["CONCRETE"] = 10] = "CONCRETE";
    MaterialTypes[MaterialTypes["ICE"] = 12] = "ICE";
    MaterialTypes[MaterialTypes["SNOW"] = 15] = "SNOW";
    MaterialTypes[MaterialTypes["WATER"] = 17] = "WATER";
    MaterialTypes[MaterialTypes["VEHICLE_GROUND"] = 25] = "VEHICLE_GROUND";
    MaterialTypes[MaterialTypes["VEHICLE_AIR"] = 26] = "VEHICLE_AIR";
    MaterialTypes[MaterialTypes["ASPHALT"] = 27] = "ASPHALT";
    MaterialTypes[MaterialTypes["CERAMIC"] = 44] = "CERAMIC";
    MaterialTypes[MaterialTypes["DIRTLIGHTBROWN"] = 45] = "DIRTLIGHTBROWN";
    MaterialTypes[MaterialTypes["DIRTDARKBROWN"] = 46] = "DIRTDARKBROWN";
    MaterialTypes[MaterialTypes["ROCKBROWN"] = 47] = "ROCKBROWN";
    MaterialTypes[MaterialTypes["ROCKGREY"] = 48] = "ROCKGREY";
    MaterialTypes[MaterialTypes["SANDBROWN"] = 49] = "SANDBROWN";
    MaterialTypes[MaterialTypes["SANDRED"] = 50] = "SANDRED";
    MaterialTypes[MaterialTypes["WATERSHALLOW"] = 51] = "WATERSHALLOW";
    MaterialTypes[MaterialTypes["CHAINLINK"] = 52] = "CHAINLINK";
    MaterialTypes[MaterialTypes["RUBBER"] = 53] = "RUBBER";
    MaterialTypes[MaterialTypes["METAL_GRATE"] = 54] = "METAL_GRATE";
    MaterialTypes[MaterialTypes["PLASTIC"] = 55] = "PLASTIC";
    MaterialTypes[MaterialTypes["PLASTIC_SOFT"] = 56] = "PLASTIC_SOFT";
    MaterialTypes[MaterialTypes["CARPET"] = 57] = "CARPET";
    MaterialTypes[MaterialTypes["ZOMBIE"] = 58] = "ZOMBIE";
    MaterialTypes[MaterialTypes["WOODBURNT"] = 59] = "WOODBURNT";
    MaterialTypes[MaterialTypes["CARDBOARD"] = 60] = "CARDBOARD";
    MaterialTypes[MaterialTypes["LEATHER"] = 61] = "LEATHER";
    MaterialTypes[MaterialTypes["KEVLAR"] = 62] = "KEVLAR";
    MaterialTypes[MaterialTypes["METAL_HOLLOW"] = 63] = "METAL_HOLLOW";
    MaterialTypes[MaterialTypes["CHAINLINK_FENCE"] = 64] = "CHAINLINK_FENCE";
    MaterialTypes[MaterialTypes["BARBED_WIRE"] = 65] = "BARBED_WIRE";
    MaterialTypes[MaterialTypes["BULLET_PENETRATION_NO_FX"] = 66] = "BULLET_PENETRATION_NO_FX";
    MaterialTypes[MaterialTypes["WATER_PUDDLE"] = 67] = "WATER_PUDDLE";
    MaterialTypes[MaterialTypes["DEFAULT_BULLET_PENETRATION"] = 69] = "DEFAULT_BULLET_PENETRATION";
})(MaterialTypes || (exports.MaterialTypes = MaterialTypes = {}));
var MeleeTypes;
(function (MeleeTypes) {
    MeleeTypes[MeleeTypes["FISTS"] = 1] = "FISTS";
    MeleeTypes[MeleeTypes["BLUNT"] = 2] = "BLUNT";
    MeleeTypes[MeleeTypes["BLADE"] = 3] = "BLADE";
    MeleeTypes[MeleeTypes["KNIFE"] = 4] = "KNIFE";
    MeleeTypes[MeleeTypes["GUITAR"] = 5] = "GUITAR";
})(MeleeTypes || (exports.MeleeTypes = MeleeTypes = {}));
var DefaultSkinsConveys;
(function (DefaultSkinsConveys) {
    DefaultSkinsConveys[DefaultSkinsConveys["RED_CONVEYS"] = 2216] = "RED_CONVEYS";
    DefaultSkinsConveys[DefaultSkinsConveys["WHITE_CONVEYS"] = 2218] = "WHITE_CONVEYS";
    DefaultSkinsConveys[DefaultSkinsConveys["DEFAULT_BLUE_CONVEYS"] = 2217] = "DEFAULT_BLUE_CONVEYS";
})(DefaultSkinsConveys || (exports.DefaultSkinsConveys = DefaultSkinsConveys = {}));
var DefaultSkinsBackpack;
(function (DefaultSkinsBackpack) {
    DefaultSkinsBackpack[DefaultSkinsBackpack["BLUE_BACKPACK"] = 2113] = "BLUE_BACKPACK";
    DefaultSkinsBackpack[DefaultSkinsBackpack["BLUE_AND_GRAY_PACKPACK"] = 2114] = "BLUE_AND_GRAY_PACKPACK";
    DefaultSkinsBackpack[DefaultSkinsBackpack["GREEN_BACKPACK"] = 2115] = "GREEN_BACKPACK";
    DefaultSkinsBackpack[DefaultSkinsBackpack["ORANGE_BACKPACK"] = 2116] = "ORANGE_BACKPACK";
    DefaultSkinsBackpack[DefaultSkinsBackpack["RED_BACKPACK"] = 2117] = "RED_BACKPACK";
    DefaultSkinsBackpack[DefaultSkinsBackpack["DEFAULT_BLUE_ORANGE"] = 2038] = "DEFAULT_BLUE_ORANGE";
})(DefaultSkinsBackpack || (exports.DefaultSkinsBackpack = DefaultSkinsBackpack = {}));
var MilitaryTan;
(function (MilitaryTan) {
    MilitaryTan[MilitaryTan["BLACK_MILITARY"] = 2118] = "BLACK_MILITARY";
    MilitaryTan[MilitaryTan["BLUE_MILITARY"] = 2119] = "BLUE_MILITARY";
    MilitaryTan[MilitaryTan["BLUE_CAMO_MILITARY"] = 2120] = "BLUE_CAMO_MILITARY";
    MilitaryTan[MilitaryTan["GREEN_CAMO_MILITARY"] = 2121] = "GREEN_CAMO_MILITARY";
    MilitaryTan[MilitaryTan["TAN_CAMO_MILITARY"] = 2122] = "TAN_CAMO_MILITARY";
    MilitaryTan[MilitaryTan["RED_BLACK_MILITARY"] = 2123] = "RED_BLACK_MILITARY";
    MilitaryTan[MilitaryTan["TAN_MILITARY"] = 2124] = "TAN_MILITARY";
    MilitaryTan[MilitaryTan["RASTA_MILITARY"] = 2393] = "RASTA_MILITARY";
    MilitaryTan[MilitaryTan["SHOWDOWN_MILITARY"] = 2777] = "SHOWDOWN_MILITARY";
    MilitaryTan[MilitaryTan["CROWN_MILITARY"] = 2778] = "CROWN_MILITARY";
    MilitaryTan[MilitaryTan["BLUE_SURVIVOR_BACKPACK"] = 3042] = "BLUE_SURVIVOR_BACKPACK";
    MilitaryTan[MilitaryTan["GREEN_SURVIVOR_BACKPACK"] = 3043] = "GREEN_SURVIVOR_BACKPACK";
    MilitaryTan[MilitaryTan["ORANGE_SURVIVOR_BACKPACK"] = 3044] = "ORANGE_SURVIVOR_BACKPACK";
    MilitaryTan[MilitaryTan["CAMO_SURVIVOR_BACKPACK"] = 3045] = "CAMO_SURVIVOR_BACKPACK";
    MilitaryTan[MilitaryTan["PINK_SKULLS_SURVIVOR_BACKPACK"] = 3046] = "PINK_SKULLS_SURVIVOR_BACKPACK";
    MilitaryTan[MilitaryTan["RED_GREEN_PURPLE_SURVIVOR_BACKPACK"] = 3047] = "RED_GREEN_PURPLE_SURVIVOR_BACKPACK";
    MilitaryTan[MilitaryTan["SNIPER_MILITARY"] = 3169] = "SNIPER_MILITARY";
    MilitaryTan[MilitaryTan["HEAVY_ASSAULT_MILITARY"] = 3403] = "HEAVY_ASSAULT_MILITARY";
    MilitaryTan[MilitaryTan["TWITCH_TACTICAL_MILITARY"] = 3583] = "TWITCH_TACTICAL_MILITARY";
    MilitaryTan[MilitaryTan["REINFORCED_TACTICAL_MILITARY"] = 4008] = "REINFORCED_TACTICAL_MILITARY";
    MilitaryTan[MilitaryTan["FULLY_GEARED_EXPLORER_BACKPACK"] = 4009] = "FULLY_GEARED_EXPLORER_BACKPACK";
    MilitaryTan[MilitaryTan["DIGITAL_CAMO_MILITARY"] = 4010] = "DIGITAL_CAMO_MILITARY";
    MilitaryTan[MilitaryTan["ALPACA_SURVIVOR_BACKPACK"] = 4011] = "ALPACA_SURVIVOR_BACKPACK";
    MilitaryTan[MilitaryTan["H1EMU_BLACK_RED_MILITARY"] = 5012] = "H1EMU_BLACK_RED_MILITARY";
    MilitaryTan[MilitaryTan["UNICORN_MILITARY"] = 5022] = "UNICORN_MILITARY";
    MilitaryTan[MilitaryTan["H1EMU_WHITE_RED_MILITARY"] = 5033] = "H1EMU_WHITE_RED_MILITARY";
    MilitaryTan[MilitaryTan["NEW_GREEN_CAMO_MILITARY"] = 5081] = "NEW_GREEN_CAMO_MILITARY";
    MilitaryTan[MilitaryTan["ASUS_ROG_MILITARY"] = 5083] = "ASUS_ROG_MILITARY";
    MilitaryTan[MilitaryTan["BLACK_MAGIC_MILITARY"] = 5086] = "BLACK_MAGIC_MILITARY";
    MilitaryTan[MilitaryTan["GREY_BROWN_LEATHER_WESTER_MILITARY"] = 5087] = "GREY_BROWN_LEATHER_WESTER_MILITARY";
    MilitaryTan[MilitaryTan["RAINBOW_CHECK_WESTERN_MILITARY"] = 5088] = "RAINBOW_CHECK_WESTERN_MILITARY";
})(MilitaryTan || (exports.MilitaryTan = MilitaryTan = {}));
var Backpack1000;
(function (Backpack1000) {
    Backpack1000[Backpack1000["CAMOUFLAGE_BACKPACK"] = 2051] = "CAMOUFLAGE_BACKPACK";
    Backpack1000[Backpack1000["GREY_YELLOW_BACKPACK"] = 2072] = "GREY_YELLOW_BACKPACK";
    Backpack1000[Backpack1000["BLACK_BACKPACK"] = 2112] = "BLACK_BACKPACK";
    Backpack1000[Backpack1000["BLUE_BACKPACK"] = 2113] = "BLUE_BACKPACK";
    Backpack1000[Backpack1000["BLUE_AND_GRAY_PACKPACK"] = 2114] = "BLUE_AND_GRAY_PACKPACK";
    Backpack1000[Backpack1000["GREEN_BACKPACK"] = 2115] = "GREEN_BACKPACK";
    Backpack1000[Backpack1000["ORANGE_BACKPACK"] = 2116] = "ORANGE_BACKPACK";
    Backpack1000[Backpack1000["RED_BACKPACK"] = 2117] = "RED_BACKPACK";
    Backpack1000[Backpack1000["DEFAULT_BLUE_ORANGE"] = 2038] = "DEFAULT_BLUE_ORANGE";
    Backpack1000[Backpack1000["BLACK_RED_STRIPED_BACKPACK"] = 3643] = "BLACK_RED_STRIPED_BACKPACK";
    Backpack1000[Backpack1000["TAN_CAMO_BACKPACK"] = 3644] = "TAN_CAMO_BACKPACK";
    Backpack1000[Backpack1000["FRANKENSWINE_BACKPACK"] = 3742] = "FRANKENSWINE_BACKPACK";
    Backpack1000[Backpack1000["FOUNDERS_BACKPACK"] = 3798] = "FOUNDERS_BACKPACK";
})(Backpack1000 || (exports.Backpack1000 = Backpack1000 = {}));
var FramedBackpack;
(function (FramedBackpack) {
    FramedBackpack[FramedBackpack["ANIMAL_HIDE_FRAMED_BACKPACK"] = 1995] = "ANIMAL_HIDE_FRAMED_BACKPACK";
    FramedBackpack[FramedBackpack["FRAMED_BACKPACK"] = 2111] = "FRAMED_BACKPACK";
})(FramedBackpack || (exports.FramedBackpack = FramedBackpack = {}));
var DefaultSkinsMotorHelmet;
(function (DefaultSkinsMotorHelmet) {
    // BLACK_MOTORCYCLE_HELMET = 2168 - skin is light blue
    DefaultSkinsMotorHelmet[DefaultSkinsMotorHelmet["GRAY_MOTOR_HELMET"] = 2169] = "GRAY_MOTOR_HELMET";
    // WHITE_MOTOR_HELMET = 2171, skin is light red
    DefaultSkinsMotorHelmet[DefaultSkinsMotorHelmet["DEFAULT_RED_MOTOR_HELMET"] = 2170] = "DEFAULT_RED_MOTOR_HELMET";
})(DefaultSkinsMotorHelmet || (exports.DefaultSkinsMotorHelmet = DefaultSkinsMotorHelmet = {}));
var DefaultSkinsZeds;
(function (DefaultSkinsZeds) {
    DefaultSkinsZeds[DefaultSkinsZeds["NAVY_ZEDS"] = 3709] = "NAVY_ZEDS";
    DefaultSkinsZeds[DefaultSkinsZeds["BROWN_ZEDS"] = 3710] = "BROWN_ZEDS";
    DefaultSkinsZeds[DefaultSkinsZeds["GREEN_ZEDS"] = 3711] = "GREEN_ZEDS";
    DefaultSkinsZeds[DefaultSkinsZeds["DEFAULT_WHITE_ZEDS"] = 3708] = "DEFAULT_WHITE_ZEDS";
})(DefaultSkinsZeds || (exports.DefaultSkinsZeds = DefaultSkinsZeds = {}));
var DefaultSkinsGators;
(function (DefaultSkinsGators) {
    DefaultSkinsGators[DefaultSkinsGators["GREEN_GATORS"] = 2563] = "GREEN_GATORS";
    DefaultSkinsGators[DefaultSkinsGators["AQUA_GATORS"] = 2564] = "AQUA_GATORS";
    DefaultSkinsGators[DefaultSkinsGators["PURPLE_GATORS"] = 2565] = "PURPLE_GATORS";
    DefaultSkinsGators[DefaultSkinsGators["WHITE_GATORS"] = 2566] = "WHITE_GATORS";
    DefaultSkinsGators[DefaultSkinsGators["PINK_GATORS"] = 2567] = "PINK_GATORS";
    DefaultSkinsGators[DefaultSkinsGators["BROWN_GATORS"] = 3212] = "BROWN_GATORS";
    DefaultSkinsGators[DefaultSkinsGators["DARK_GREY_GATORS"] = 3485] = "DARK_GREY_GATORS";
    DefaultSkinsGators[DefaultSkinsGators["GRAY_BLUE_GATORS"] = 3486] = "GRAY_BLUE_GATORS";
    DefaultSkinsGators[DefaultSkinsGators["DEFAULT_RED_GATORS"] = 2550] = "DEFAULT_RED_GATORS";
})(DefaultSkinsGators || (exports.DefaultSkinsGators = DefaultSkinsGators = {}));
var DefaultSkinsBoots;
(function (DefaultSkinsBoots) {
    DefaultSkinsBoots[DefaultSkinsBoots["BROWN_BOOTS"] = 2206] = "BROWN_BOOTS";
    DefaultSkinsBoots[DefaultSkinsBoots["TAN_BOOTS"] = 2207] = "TAN_BOOTS";
    DefaultSkinsBoots[DefaultSkinsBoots["GRAY_BOOTS"] = 2208] = "GRAY_BOOTS";
    DefaultSkinsBoots[DefaultSkinsBoots["DEFAULT_GRAY_BLUE_BOOTS"] = 2209] = "DEFAULT_GRAY_BLUE_BOOTS";
})(DefaultSkinsBoots || (exports.DefaultSkinsBoots = DefaultSkinsBoots = {}));
var GameModes;
(function (GameModes) {
    GameModes[GameModes["SURVIVAL"] = 0] = "SURVIVAL";
    GameModes[GameModes["BATTLE_ROYALE"] = 1] = "BATTLE_ROYALE";
    GameModes[GameModes["BATTLE_ROYALE_ZOMBIES"] = 3] = "BATTLE_ROYALE_ZOMBIES";
    GameModes[GameModes["BATTLE_ROYALE_2"] = 5] = "BATTLE_ROYALE_2";
    GameModes[GameModes["BATTLE_ROYALE_5"] = 7] = "BATTLE_ROYALE_5";
    GameModes[GameModes["BATTLE_ROYALE_GREEN_DAWN"] = 10] = "BATTLE_ROYALE_GREEN_DAWN";
    GameModes[GameModes["BATTLE_ROYALE_REANIMATOR"] = 14] = "BATTLE_ROYALE_REANIMATOR";
})(GameModes || (exports.GameModes = GameModes = {}));
var UIElements;
(function (UIElements) {
    UIElements[UIElements["MAP"] = 12] = "MAP";
    UIElements[UIElements["INVENTORY"] = 15] = "INVENTORY";
})(UIElements || (exports.UIElements = UIElements = {}));
var ReplicationPropertyHash;
(function (ReplicationPropertyHash) {
    ReplicationPropertyHash[ReplicationPropertyHash["UNKNOWN1"] = 1356143773] = "UNKNOWN1";
    ReplicationPropertyHash[ReplicationPropertyHash["ISWORLDITEM"] = 4136351497] = "ISWORLDITEM";
    ReplicationPropertyHash[ReplicationPropertyHash["UNKNOWN2"] = 3106403311] = "UNKNOWN2";
})(ReplicationPropertyHash || (exports.ReplicationPropertyHash = ReplicationPropertyHash = {}));
