// h1emu-core/h1emu_core.js

let imports = {};
imports['__wbindgen_placeholder__'] = module.exports;
let wasm;
const { TextDecoder, TextEncoder } = require(`util`);

function addToExternrefTable0(obj) {
    const idx = wasm.__externref_table_alloc();
    wasm.__wbindgen_export_2.set(idx, obj);
    return idx;
}

function handleError(f, args) {
    try {
        return f.apply(this, args);
    } catch (e) {
        const idx = addToExternrefTable0(e);
        wasm.__wbindgen_exn_store(idx);
    }
}

let cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });

cachedTextDecoder.decode();

let cachedUint8ArrayMemory0 = null;

function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}

function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}

function isLikeNone(x) {
    return x === undefined || x === null;
}

let WASM_VECTOR_LEN = 0;

let cachedTextEncoder = new TextEncoder('utf-8');

const encodeString = (typeof cachedTextEncoder.encodeInto === 'function'
    ? function (arg, view) {
    return cachedTextEncoder.encodeInto(arg, view);
}
    : function (arg, view) {
    const buf = cachedTextEncoder.encode(arg);
    view.set(buf);
    return {
        read: arg.length,
        written: buf.length
    };
});

function passStringToWasm0(arg, malloc, realloc) {

    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }

    let len = arg.length;
    let ptr = malloc(len, 1) >>> 0;

    const mem = getUint8ArrayMemory0();

    let offset = 0;

    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr + offset] = code;
    }

    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
        const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
        const ret = encodeString(arg, view);

        offset += ret.written;
        ptr = realloc(ptr, len, offset, 1) >>> 0;
    }

    WASM_VECTOR_LEN = offset;
    return ptr;
}

let cachedDataViewMemory0 = null;

function getDataViewMemory0() {
    if (cachedDataViewMemory0 === null || cachedDataViewMemory0.buffer.detached === true || (cachedDataViewMemory0.buffer.detached === undefined && cachedDataViewMemory0.buffer !== wasm.memory.buffer)) {
        cachedDataViewMemory0 = new DataView(wasm.memory.buffer);
    }
    return cachedDataViewMemory0;
}

function getArrayU8FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getUint8ArrayMemory0().subarray(ptr / 1, ptr / 1 + len);
}

function passArray8ToWasm0(arg, malloc) {
    const ptr = malloc(arg.length * 1, 1) >>> 0;
    getUint8ArrayMemory0().set(arg, ptr / 1);
    WASM_VECTOR_LEN = arg.length;
    return ptr;
}
/**
 * @param {Uint8Array} data
 * @param {number} crc_seed
 * @returns {Uint8Array}
 */
module.exports.append_crc_legacy = function(data, crc_seed) {
    const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.append_crc_legacy(ptr0, len0, crc_seed);
    var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
    wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
    return v2;
};

/**
 * @param {Uint8Array} data
 * @param {number} crc_seed
 * @returns {number}
 */
module.exports.crc32_legacy = function(data, crc_seed) {
    const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.crc32_legacy(ptr0, len0, crc_seed);
    return ret >>> 0;
};

/**
 * @param {string} string
 * @returns {number}
 */
module.exports.joaat = function(string) {
    const ptr0 = passStringToWasm0(string, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.joaat(ptr0, len0);
    return ret >>> 0;
};

/**
 * @returns {string}
 */
module.exports.generate_random_guid = function() {
    let deferred1_0;
    let deferred1_1;
    try {
        const ret = wasm.generate_random_guid();
        deferred1_0 = ret[0];
        deferred1_1 = ret[1];
        return getStringFromWasm0(ret[0], ret[1]);
    } finally {
        wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
    }
};

/**
 * @enum {0 | 1 | 2 | 3 | 4}
 */
module.exports.EncryptMethod = Object.freeze({
    EncryptMethodNone: 0, "0": "EncryptMethodNone",
    EncryptMethodUserSupplied: 1, "1": "EncryptMethodUserSupplied",
    EncryptMethodUserSupplied2: 2, "2": "EncryptMethodUserSupplied2",
    EncryptMethodXorBuffer: 3, "3": "EncryptMethodXorBuffer",
    EncryptMethodXor: 4, "4": "EncryptMethodXor",
});
/**
 * @enum {153}
 */
module.exports.ErrorType = Object.freeze({
    Deserializing: 153, "153": "Deserializing",
});
/**
 * @enum {0 | 1 | 2 | 3 | 4}
 */
module.exports.GatewayChannels = Object.freeze({
    Zone: 0, "0": "Zone",
    World: 1, "1": "World",
    UpdatePosition: 2, "2": "UpdatePosition",
    ShortCircuitZone: 3, "3": "ShortCircuitZone",
    Gateway: 4, "4": "Gateway",
});
/**
 * @enum {1 | 2 | 3 | 5 | 6 | 7 | 8 | 9 | 13 | 17 | 21 | 25 | 27 | 29 | 0}
 */
module.exports.SoeOpcode = Object.freeze({
    SessionRequest: 1, "1": "SessionRequest",
    SessionReply: 2, "2": "SessionReply",
    MultiPacket: 3, "3": "MultiPacket",
    Disconnect: 5, "5": "Disconnect",
    Ping: 6, "6": "Ping",
    NetStatusRequest: 7, "7": "NetStatusRequest",
    NetStatusReply: 8, "8": "NetStatusReply",
    Data: 9, "9": "Data",
    DataFragment: 13, "13": "DataFragment",
    OutOfOrder: 17, "17": "OutOfOrder",
    Ack: 21, "21": "Ack",
    Group: 25, "25": "Group",
    Ordered: 27, "27": "Ordered",
    FatalError: 29, "29": "FatalError",
    Unknown: 0, "0": "Unknown",
});

const AckPacketFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_ackpacket_free(ptr >>> 0, 1));

class AckPacket {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        AckPacketFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_ackpacket_free(ptr, 0);
    }
    /**
     * @returns {number}
     */
    get sequence() {
        const ret = wasm.__wbg_get_ackpacket_sequence(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {number} arg0
     */
    set sequence(arg0) {
        wasm.__wbg_set_ackpacket_sequence(this.__wbg_ptr, arg0);
    }
    /**
     * @param {number} sequence
     */
    constructor(sequence) {
        const ret = wasm.ackpacket_new(sequence);
        this.__wbg_ptr = ret >>> 0;
        AckPacketFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @returns {number}
     */
    get_sequence() {
        const ret = wasm.ackpacket_get_sequence(this.__wbg_ptr);
        return ret;
    }
}
module.exports.AckPacket = AckPacket;

const DataPacketFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_datapacket_free(ptr >>> 0, 1));

class DataPacket {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        DataPacketFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_datapacket_free(ptr, 0);
    }
    /**
     * @returns {number}
     */
    get sequence() {
        const ret = wasm.__wbg_get_datapacket_sequence(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {number} arg0
     */
    set sequence(arg0) {
        wasm.__wbg_set_datapacket_sequence(this.__wbg_ptr, arg0);
    }
    /**
     * @param {Uint8Array} data
     * @param {number} sequence
     */
    constructor(data, sequence) {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.datapacket_new(ptr0, len0, sequence);
        this.__wbg_ptr = ret >>> 0;
        DataPacketFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @returns {number}
     */
    get_sequence() {
        const ret = wasm.datapacket_get_sequence(this.__wbg_ptr);
        return ret;
    }
}
module.exports.DataPacket = DataPacket;

const GatewayProtocolFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_gatewayprotocol_free(ptr >>> 0, 1));

class GatewayProtocol {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        GatewayProtocolFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_gatewayprotocol_free(ptr, 0);
    }
    constructor() {
        const ret = wasm.gatewayprotocol_initialize();
        this.__wbg_ptr = ret >>> 0;
        GatewayProtocolFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @param {Uint8Array} data
     * @returns {string}
     */
    parse(data) {
        let deferred2_0;
        let deferred2_1;
        try {
            const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
            const len0 = WASM_VECTOR_LEN;
            const ret = wasm.gatewayprotocol_parse(this.__wbg_ptr, ptr0, len0);
            deferred2_0 = ret[0];
            deferred2_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
        }
    }
    /**
     * @param {bigint} character_id
     * @param {string} ticket
     * @param {string} client_protocol
     * @param {string} client_build
     * @returns {Uint8Array}
     */
    pack_login_request_packet(character_id, ticket, client_protocol, client_build) {
        const ptr0 = passStringToWasm0(ticket, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(client_protocol, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ptr2 = passStringToWasm0(client_build, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len2 = WASM_VECTOR_LEN;
        const ret = wasm.gatewayprotocol_pack_login_request_packet(this.__wbg_ptr, character_id, ptr0, len0, ptr1, len1, ptr2, len2);
        var v4 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v4;
    }
    /**
     * @param {boolean} logged_in
     * @returns {Uint8Array}
     */
    pack_login_reply_packet(logged_in) {
        const ret = wasm.gatewayprotocol_pack_login_reply_packet(this.__wbg_ptr, logged_in);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {Uint8Array} data
     * @param {number} channel
     * @returns {Uint8Array}
     */
    pack_tunnel_data_packet_for_client(data, channel) {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.gatewayprotocol_pack_tunnel_data_packet_for_client(this.__wbg_ptr, ptr0, len0, channel);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {Uint8Array} data
     * @param {number} channel
     * @returns {Uint8Array}
     */
    pack_tunnel_data_packet_for_server(data, channel) {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.gatewayprotocol_pack_tunnel_data_packet_for_server(this.__wbg_ptr, ptr0, len0, channel);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @returns {Uint8Array}
     */
    pack_channel_is_routable_packet() {
        const ret = wasm.gatewayprotocol_pack_channel_is_routable_packet(this.__wbg_ptr);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @returns {Uint8Array}
     */
    pack_channel_is_not_routable_packet() {
        const ret = wasm.gatewayprotocol_pack_channel_is_not_routable_packet(this.__wbg_ptr);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
}
module.exports.GatewayProtocol = GatewayProtocol;

const NetStatusReplyPacketFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_netstatusreplypacket_free(ptr >>> 0, 1));

class NetStatusReplyPacket {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NetStatusReplyPacketFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_netstatusreplypacket_free(ptr, 0);
    }
    /**
     * @returns {number}
     */
    get client_tick_count() {
        const ret = wasm.__wbg_get_netstatusreplypacket_client_tick_count(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {number} arg0
     */
    set client_tick_count(arg0) {
        wasm.__wbg_set_netstatusreplypacket_client_tick_count(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get server_tick_count() {
        const ret = wasm.__wbg_get_netstatusreplypacket_server_tick_count(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set server_tick_count(arg0) {
        wasm.__wbg_set_netstatusreplypacket_server_tick_count(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {bigint}
     */
    get client_packet_sent() {
        const ret = wasm.__wbg_get_netstatusreplypacket_client_packet_sent(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @param {bigint} arg0
     */
    set client_packet_sent(arg0) {
        wasm.__wbg_set_netstatusreplypacket_client_packet_sent(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {bigint}
     */
    get client_packet_received() {
        const ret = wasm.__wbg_get_netstatusreplypacket_client_packet_received(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @param {bigint} arg0
     */
    set client_packet_received(arg0) {
        wasm.__wbg_set_netstatusreplypacket_client_packet_received(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {bigint}
     */
    get server_packet_sent() {
        const ret = wasm.__wbg_get_netstatusreplypacket_server_packet_sent(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @param {bigint} arg0
     */
    set server_packet_sent(arg0) {
        wasm.__wbg_set_netstatusreplypacket_server_packet_sent(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {bigint}
     */
    get server_packet_received() {
        const ret = wasm.__wbg_get_netstatusreplypacket_server_packet_received(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @param {bigint} arg0
     */
    set server_packet_received(arg0) {
        wasm.__wbg_set_netstatusreplypacket_server_packet_received(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get unknown_field() {
        const ret = wasm.__wbg_get_netstatusreplypacket_unknown_field(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {number} arg0
     */
    set unknown_field(arg0) {
        wasm.__wbg_set_netstatusreplypacket_unknown_field(this.__wbg_ptr, arg0);
    }
    /**
     * @param {number} client_tick_count
     * @param {number} server_tick_count
     * @param {bigint} client_packet_sent
     * @param {bigint} client_packet_received
     * @param {bigint} server_packet_sent
     * @param {bigint} server_packet_received
     * @param {number} unknown_field
     */
    constructor(client_tick_count, server_tick_count, client_packet_sent, client_packet_received, server_packet_sent, server_packet_received, unknown_field) {
        const ret = wasm.netstatusreplypacket_new(client_tick_count, server_tick_count, client_packet_sent, client_packet_received, server_packet_sent, server_packet_received, unknown_field);
        this.__wbg_ptr = ret >>> 0;
        NetStatusReplyPacketFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @returns {number}
     */
    get_client_tick_count() {
        const ret = wasm.netstatusreplypacket_get_client_tick_count(this.__wbg_ptr);
        return ret;
    }
    /**
     * @returns {number}
     */
    get_server_tick_count() {
        const ret = wasm.netstatusreplypacket_get_server_tick_count(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {bigint}
     */
    get_client_packet_sent() {
        const ret = wasm.netstatusreplypacket_get_client_packet_sent(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @returns {bigint}
     */
    get_client_packet_received() {
        const ret = wasm.netstatusreplypacket_get_client_packet_received(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @returns {bigint}
     */
    get_server_packet_sent() {
        const ret = wasm.netstatusreplypacket_get_server_packet_sent(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @returns {bigint}
     */
    get_server_packet_received() {
        const ret = wasm.netstatusreplypacket_get_server_packet_received(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @returns {number}
     */
    get_unknown_field() {
        const ret = wasm.netstatusreplypacket_get_unknown_field(this.__wbg_ptr);
        return ret;
    }
}
module.exports.NetStatusReplyPacket = NetStatusReplyPacket;

const NetStatusRequestPacketFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_netstatusrequestpacket_free(ptr >>> 0, 1));

class NetStatusRequestPacket {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NetStatusRequestPacketFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_netstatusrequestpacket_free(ptr, 0);
    }
    /**
     * @returns {number}
     */
    get client_tick_count() {
        const ret = wasm.__wbg_get_netstatusreplypacket_client_tick_count(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {number} arg0
     */
    set client_tick_count(arg0) {
        wasm.__wbg_set_netstatusreplypacket_client_tick_count(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get last_client_update() {
        const ret = wasm.__wbg_get_netstatusrequestpacket_last_client_update(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set last_client_update(arg0) {
        wasm.__wbg_set_netstatusrequestpacket_last_client_update(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get average_update() {
        const ret = wasm.__wbg_get_netstatusrequestpacket_average_update(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set average_update(arg0) {
        wasm.__wbg_set_netstatusrequestpacket_average_update(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get shortest_update() {
        const ret = wasm.__wbg_get_netstatusrequestpacket_shortest_update(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set shortest_update(arg0) {
        wasm.__wbg_set_netstatusrequestpacket_shortest_update(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get longest_update() {
        const ret = wasm.__wbg_get_netstatusrequestpacket_longest_update(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set longest_update(arg0) {
        wasm.__wbg_set_netstatusrequestpacket_longest_update(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get last_server_update() {
        const ret = wasm.__wbg_get_netstatusreplypacket_server_tick_count(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set last_server_update(arg0) {
        wasm.__wbg_set_netstatusreplypacket_server_tick_count(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {bigint}
     */
    get packets_sent() {
        const ret = wasm.__wbg_get_netstatusreplypacket_client_packet_sent(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @param {bigint} arg0
     */
    set packets_sent(arg0) {
        wasm.__wbg_set_netstatusreplypacket_client_packet_sent(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {bigint}
     */
    get packets_received() {
        const ret = wasm.__wbg_get_netstatusreplypacket_client_packet_received(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @param {bigint} arg0
     */
    set packets_received(arg0) {
        wasm.__wbg_set_netstatusreplypacket_client_packet_received(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get unknown_field() {
        const ret = wasm.__wbg_get_netstatusreplypacket_unknown_field(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {number} arg0
     */
    set unknown_field(arg0) {
        wasm.__wbg_set_netstatusreplypacket_unknown_field(this.__wbg_ptr, arg0);
    }
    /**
     * @param {number} client_tick_count
     * @param {number} last_client_update
     * @param {number} average_update
     * @param {number} shortest_update
     * @param {number} longest_update
     * @param {number} last_server_update
     * @param {bigint} packets_sent
     * @param {bigint} packets_received
     * @param {number} unknown_field
     */
    constructor(client_tick_count, last_client_update, average_update, shortest_update, longest_update, last_server_update, packets_sent, packets_received, unknown_field) {
        const ret = wasm.netstatusrequestpacket_new(client_tick_count, last_client_update, average_update, shortest_update, longest_update, last_server_update, packets_sent, packets_received, unknown_field);
        this.__wbg_ptr = ret >>> 0;
        NetStatusRequestPacketFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @returns {number}
     */
    get_client_tick_count() {
        const ret = wasm.netstatusreplypacket_get_client_tick_count(this.__wbg_ptr);
        return ret;
    }
    /**
     * @returns {number}
     */
    get_last_client_update() {
        const ret = wasm.netstatusrequestpacket_get_last_client_update(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get_average_update() {
        const ret = wasm.netstatusrequestpacket_get_average_update(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get_shortest_update() {
        const ret = wasm.netstatusrequestpacket_get_shortest_update(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get_longest_update() {
        const ret = wasm.netstatusrequestpacket_get_longest_update(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get_last_server_update() {
        const ret = wasm.netstatusreplypacket_get_server_tick_count(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {bigint}
     */
    get_packets_sent() {
        const ret = wasm.netstatusreplypacket_get_client_packet_sent(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @returns {bigint}
     */
    get_packets_received() {
        const ret = wasm.netstatusreplypacket_get_client_packet_received(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @returns {number}
     */
    get_unknown_field() {
        const ret = wasm.netstatusreplypacket_get_unknown_field(this.__wbg_ptr);
        return ret;
    }
}
module.exports.NetStatusRequestPacket = NetStatusRequestPacket;

const RC4Finalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_rc4_free(ptr >>> 0, 1));

class RC4 {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        RC4Finalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_rc4_free(ptr, 0);
    }
    /**
     * @param {Uint8Array} given_key
     */
    constructor(given_key) {
        const ptr0 = passArray8ToWasm0(given_key, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.rc4_initialize(ptr0, len0);
        this.__wbg_ptr = ret >>> 0;
        RC4Finalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @param {Uint8Array} data
     * @returns {Uint8Array}
     */
    encrypt(data) {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.rc4_encrypt(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {Uint8Array} data
     * @returns {Uint8Array}
     */
    decrypt(data) {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.rc4_decrypt(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
}
module.exports.RC4 = RC4;

const SessionReplyPacketFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_sessionreplypacket_free(ptr >>> 0, 1));

class SessionReplyPacket {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        SessionReplyPacketFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_sessionreplypacket_free(ptr, 0);
    }
    /**
     * @returns {number}
     */
    get session_id() {
        const ret = wasm.__wbg_get_sessionreplypacket_session_id(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set session_id(arg0) {
        wasm.__wbg_set_sessionreplypacket_session_id(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get crc_seed() {
        const ret = wasm.__wbg_get_sessionreplypacket_crc_seed(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set crc_seed(arg0) {
        wasm.__wbg_set_sessionreplypacket_crc_seed(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get crc_length() {
        const ret = wasm.__wbg_get_sessionreplypacket_crc_length(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {number} arg0
     */
    set crc_length(arg0) {
        wasm.__wbg_set_sessionreplypacket_crc_length(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get encrypt_method() {
        const ret = wasm.__wbg_get_datapacket_sequence(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {number} arg0
     */
    set encrypt_method(arg0) {
        wasm.__wbg_set_datapacket_sequence(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get udp_length() {
        const ret = wasm.__wbg_get_sessionreplypacket_udp_length(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set udp_length(arg0) {
        wasm.__wbg_set_sessionreplypacket_udp_length(this.__wbg_ptr, arg0);
    }
    /**
     * @param {number} session_id
     * @param {number} crc_seed
     * @param {number} crc_length
     * @param {number} encrypt_method
     * @param {number} udp_length
     */
    constructor(session_id, crc_seed, crc_length, encrypt_method, udp_length) {
        const ret = wasm.sessionreplypacket_new(session_id, crc_seed, crc_length, encrypt_method, udp_length);
        this.__wbg_ptr = ret >>> 0;
        SessionReplyPacketFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @returns {number}
     */
    get_session_id() {
        const ret = wasm.sessionreplypacket_get_session_id(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get_crc_seed() {
        const ret = wasm.sessionreplypacket_get_crc_seed(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get_crc_length() {
        const ret = wasm.sessionreplypacket_get_crc_length(this.__wbg_ptr);
        return ret;
    }
    /**
     * @returns {number}
     */
    get_encrypt_method() {
        const ret = wasm.datapacket_get_sequence(this.__wbg_ptr);
        return ret;
    }
    /**
     * @returns {number}
     */
    get_udp_length() {
        const ret = wasm.sessionreplypacket_get_udp_length(this.__wbg_ptr);
        return ret >>> 0;
    }
}
module.exports.SessionReplyPacket = SessionReplyPacket;

const SessionRequestPacketFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_sessionrequestpacket_free(ptr >>> 0, 1));

class SessionRequestPacket {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        SessionRequestPacketFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_sessionrequestpacket_free(ptr, 0);
    }
    /**
     * @returns {number}
     */
    get session_id() {
        const ret = wasm.__wbg_get_sessionrequestpacket_session_id(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set session_id(arg0) {
        wasm.__wbg_set_sessionrequestpacket_session_id(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get crc_length() {
        const ret = wasm.__wbg_get_sessionrequestpacket_crc_length(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set crc_length(arg0) {
        wasm.__wbg_set_sessionrequestpacket_crc_length(this.__wbg_ptr, arg0);
    }
    /**
     * @returns {number}
     */
    get udp_length() {
        const ret = wasm.__wbg_get_sessionrequestpacket_udp_length(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} arg0
     */
    set udp_length(arg0) {
        wasm.__wbg_set_sessionrequestpacket_udp_length(this.__wbg_ptr, arg0);
    }
    /**
     * @param {number} session_id
     * @param {number} crc_length
     * @param {number} udp_length
     * @param {string} protocol
     */
    constructor(session_id, crc_length, udp_length, protocol) {
        const ptr0 = passStringToWasm0(protocol, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.sessionrequestpacket_new(session_id, crc_length, udp_length, ptr0, len0);
        this.__wbg_ptr = ret >>> 0;
        SessionRequestPacketFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @returns {number}
     */
    get_session_id() {
        const ret = wasm.sessionrequestpacket_get_session_id(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get_crc_length() {
        const ret = wasm.sessionrequestpacket_get_crc_length(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {number}
     */
    get_udp_length() {
        const ret = wasm.sessionrequestpacket_get_udp_length(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @returns {string}
     */
    get_protocol() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.sessionrequestpacket_get_protocol(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
}
module.exports.SessionRequestPacket = SessionRequestPacket;

const SoeprotocolFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_soeprotocol_free(ptr >>> 0, 1));

class Soeprotocol {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        SoeprotocolFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_soeprotocol_free(ptr, 0);
    }
    /**
     * @param {boolean} use_crc
     * @param {number} crc_seed
     */
    constructor(use_crc, crc_seed) {
        const ret = wasm.soeprotocol_initialize(use_crc, crc_seed);
        this.__wbg_ptr = ret >>> 0;
        SoeprotocolFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @param {SoeOpcode} packet_opcode
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack(packet_opcode, packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack(this.__wbg_ptr, packet_opcode, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_ordered(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_ordered(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {Uint8Array} data
     * @param {number} sequence
     * @returns {Uint8Array}
     */
    pack_ordered_packet(data, sequence) {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_ordered_packet(this.__wbg_ptr, ptr0, len0, sequence);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_session_request(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_session_request(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_session_request_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_session_request_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {number} session_id
     * @param {number} crc_length
     * @param {number} udp_length
     * @param {string} protocol
     * @returns {Uint8Array}
     */
    pack_session_request_packet(session_id, crc_length, udp_length, protocol) {
        const ptr0 = passStringToWasm0(protocol, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_session_request_packet(this.__wbg_ptr, session_id, crc_length, udp_length, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_session_reply(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_session_reply(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_session_reply_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_session_reply_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {number} session_id
     * @param {number} crc_seed
     * @param {number} crc_length
     * @param {number} encrypt_method
     * @param {number} udp_length
     * @returns {Uint8Array}
     */
    pack_session_reply_packet(session_id, crc_seed, crc_length, encrypt_method, udp_length) {
        const ret = wasm.soeprotocol_pack_session_reply_packet(this.__wbg_ptr, session_id, crc_seed, crc_length, encrypt_method, udp_length);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_net_status_request(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_net_status_request(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_net_status_request_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_net_status_request_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_net_status_reply(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_net_status_reply(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_net_status_reply_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_net_status_reply_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_multi(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_multi(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_multi_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_multi_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_group(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_group(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_group_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_group_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_data(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_data(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_data_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_data_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {Uint8Array} data
     * @param {number} sequence
     * @returns {Uint8Array}
     */
    pack_data_packet(data, sequence) {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_data_packet(this.__wbg_ptr, ptr0, len0, sequence);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_fragment_data(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_fragment_data(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_fragment_data_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_fragment_data_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {Uint8Array} data
     * @param {number} sequence
     * @returns {Uint8Array}
     */
    pack_fragment_data_packet(data, sequence) {
        const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_fragment_data_packet(this.__wbg_ptr, ptr0, len0, sequence);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_out_of_order(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_out_of_order(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_out_of_order_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_out_of_order_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {number} sequence
     * @returns {Uint8Array}
     */
    pack_out_of_order_packet(sequence) {
        const ret = wasm.soeprotocol_pack_out_of_order_packet(this.__wbg_ptr, sequence);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {string} packet
     * @returns {Uint8Array}
     */
    pack_ack(packet) {
        const ptr0 = passStringToWasm0(packet, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.soeprotocol_pack_ack(this.__wbg_ptr, ptr0, len0);
        var v2 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v2;
    }
    /**
     * @param {any} js_object
     * @returns {Uint8Array}
     */
    pack_ack_fromjs(js_object) {
        const ret = wasm.soeprotocol_pack_ack_fromjs(this.__wbg_ptr, js_object);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {number} sequence
     * @returns {Uint8Array}
     */
    pack_ack_packet(sequence) {
        const ret = wasm.soeprotocol_pack_ack_packet(this.__wbg_ptr, sequence);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * @param {Uint8Array} data
     * @returns {string}
     */
    parse(data) {
        let deferred2_0;
        let deferred2_1;
        try {
            const ptr0 = passArray8ToWasm0(data, wasm.__wbindgen_malloc);
            const len0 = WASM_VECTOR_LEN;
            const ret = wasm.soeprotocol_parse(this.__wbg_ptr, ptr0, len0);
            deferred2_0 = ret[0];
            deferred2_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
        }
    }
    /**
     * @returns {number}
     */
    get_crc_seed() {
        const ret = wasm.soeprotocol_get_crc_seed(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * @param {number} crc_seed
     */
    set_crc_seed(crc_seed) {
        wasm.soeprotocol_set_crc_seed(this.__wbg_ptr, crc_seed);
    }
    /**
     * @returns {boolean}
     */
    is_using_crc() {
        const ret = wasm.soeprotocol_is_using_crc(this.__wbg_ptr);
        return ret !== 0;
    }
    disable_crc() {
        wasm.soeprotocol_disable_crc(this.__wbg_ptr);
    }
    enable_crc() {
        wasm.soeprotocol_enable_crc(this.__wbg_ptr);
    }
}
module.exports.Soeprotocol = Soeprotocol;

const SubBasePacketFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_subbasepacket_free(ptr >>> 0, 1));

class SubBasePacket {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        SubBasePacketFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_subbasepacket_free(ptr, 0);
    }
    /**
     * @returns {number | undefined}
     */
    get sequence() {
        const ret = wasm.__wbg_get_subbasepacket_sequence(this.__wbg_ptr);
        return ret === 0xFFFFFF ? undefined : ret;
    }
    /**
     * @param {number | null} [arg0]
     */
    set sequence(arg0) {
        wasm.__wbg_set_subbasepacket_sequence(this.__wbg_ptr, isLikeNone(arg0) ? 0xFFFFFF : arg0);
    }
    /**
     * @param {string} name
     * @param {number | null} [sequence]
     * @param {Uint8Array | null} [data]
     */
    constructor(name, sequence, data) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        var ptr1 = isLikeNone(data) ? 0 : passArray8ToWasm0(data, wasm.__wbindgen_malloc);
        var len1 = WASM_VECTOR_LEN;
        const ret = wasm.subbasepacket_new(ptr0, len0, isLikeNone(sequence) ? 0xFFFFFF : sequence, ptr1, len1);
        this.__wbg_ptr = ret >>> 0;
        SubBasePacketFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @returns {string}
     */
    get_name() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.subbasepacket_get_name(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * @returns {number | undefined}
     */
    get_sequence() {
        const ret = wasm.subbasepacket_get_sequence(this.__wbg_ptr);
        return ret === 0xFFFFFF ? undefined : ret;
    }
}
module.exports.SubBasePacket = SubBasePacket;

const SubBasePacketsFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_subbasepackets_free(ptr >>> 0, 1));

class SubBasePackets {

    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        SubBasePacketsFinalization.unregister(this);
        return ptr;
    }

    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_subbasepackets_free(ptr, 0);
    }
    constructor() {
        const ret = wasm.subbasepackets_new();
        this.__wbg_ptr = ret >>> 0;
        SubBasePacketsFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
}
module.exports.SubBasePackets = SubBasePackets;

module.exports.__wbg_buffer_609cc3eee51ed158 = function(arg0) {
    const ret = arg0.buffer;
    return ret;
};

module.exports.__wbg_call_672a4d21634d4a24 = function() { return handleError(function (arg0, arg1) {
    const ret = arg0.call(arg1);
    return ret;
}, arguments) };

module.exports.__wbg_call_7cccdd69e0791ae2 = function() { return handleError(function (arg0, arg1, arg2) {
    const ret = arg0.call(arg1, arg2);
    return ret;
}, arguments) };

module.exports.__wbg_crypto_ed58b8e10a292839 = function(arg0) {
    const ret = arg0.crypto;
    return ret;
};

module.exports.__wbg_getRandomValues_bcb4912f16000dc4 = function() { return handleError(function (arg0, arg1) {
    arg0.getRandomValues(arg1);
}, arguments) };

module.exports.__wbg_msCrypto_0a36e2ec3a343d26 = function(arg0) {
    const ret = arg0.msCrypto;
    return ret;
};

module.exports.__wbg_new_a12002a7f91c75be = function(arg0) {
    const ret = new Uint8Array(arg0);
    return ret;
};

module.exports.__wbg_newnoargs_105ed471475aaf50 = function(arg0, arg1) {
    const ret = new Function(getStringFromWasm0(arg0, arg1));
    return ret;
};

module.exports.__wbg_newwithbyteoffsetandlength_d97e637ebe145a9a = function(arg0, arg1, arg2) {
    const ret = new Uint8Array(arg0, arg1 >>> 0, arg2 >>> 0);
    return ret;
};

module.exports.__wbg_newwithlength_a381634e90c276d4 = function(arg0) {
    const ret = new Uint8Array(arg0 >>> 0);
    return ret;
};

module.exports.__wbg_node_02999533c4ea02e3 = function(arg0) {
    const ret = arg0.node;
    return ret;
};

module.exports.__wbg_process_5c1d670bc53614b8 = function(arg0) {
    const ret = arg0.process;
    return ret;
};

module.exports.__wbg_randomFillSync_ab2cfe79ebbf2740 = function() { return handleError(function (arg0, arg1) {
    arg0.randomFillSync(arg1);
}, arguments) };

module.exports.__wbg_require_79b1e9274cde3c87 = function() { return handleError(function () {
    const ret = module.require;
    return ret;
}, arguments) };

module.exports.__wbg_set_65595bdd868b3009 = function(arg0, arg1, arg2) {
    arg0.set(arg1, arg2 >>> 0);
};

module.exports.__wbg_static_accessor_GLOBAL_88a902d13a557d07 = function() {
    const ret = typeof global === 'undefined' ? null : global;
    return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
};

module.exports.__wbg_static_accessor_GLOBAL_THIS_56578be7e9f832b0 = function() {
    const ret = typeof globalThis === 'undefined' ? null : globalThis;
    return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
};

module.exports.__wbg_static_accessor_SELF_37c5d418e4bf5819 = function() {
    const ret = typeof self === 'undefined' ? null : self;
    return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
};

module.exports.__wbg_static_accessor_WINDOW_5de37043a91a9c40 = function() {
    const ret = typeof window === 'undefined' ? null : window;
    return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
};

module.exports.__wbg_stringify_f7ed6987935b4a24 = function() { return handleError(function (arg0) {
    const ret = JSON.stringify(arg0);
    return ret;
}, arguments) };

module.exports.__wbg_subarray_aa9065fa9dc5df96 = function(arg0, arg1, arg2) {
    const ret = arg0.subarray(arg1 >>> 0, arg2 >>> 0);
    return ret;
};

module.exports.__wbg_versions_c71aa1626a93e0a1 = function(arg0) {
    const ret = arg0.versions;
    return ret;
};

module.exports.__wbindgen_init_externref_table = function() {
    const table = wasm.__wbindgen_export_2;
    const offset = table.grow(4);
    table.set(0, undefined);
    table.set(offset + 0, undefined);
    table.set(offset + 1, null);
    table.set(offset + 2, true);
    table.set(offset + 3, false);
    ;
};

module.exports.__wbindgen_is_function = function(arg0) {
    const ret = typeof(arg0) === 'function';
    return ret;
};

module.exports.__wbindgen_is_object = function(arg0) {
    const val = arg0;
    const ret = typeof(val) === 'object' && val !== null;
    return ret;
};

module.exports.__wbindgen_is_string = function(arg0) {
    const ret = typeof(arg0) === 'string';
    return ret;
};

module.exports.__wbindgen_is_undefined = function(arg0) {
    const ret = arg0 === undefined;
    return ret;
};

module.exports.__wbindgen_memory = function() {
    const ret = wasm.memory;
    return ret;
};

module.exports.__wbindgen_string_get = function(arg0, arg1) {
    const obj = arg1;
    const ret = typeof(obj) === 'string' ? obj : undefined;
    var ptr1 = isLikeNone(ret) ? 0 : passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    var len1 = WASM_VECTOR_LEN;
    getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
};

module.exports.__wbindgen_string_new = function(arg0, arg1) {
    const ret = getStringFromWasm0(arg0, arg1);
    return ret;
};

module.exports.__wbindgen_throw = function(arg0, arg1) {
    throw new Error(getStringFromWasm0(arg0, arg1));
};

const path = require('path').join(__dirname, 'h1emu_core_bg.wasm');
const bytes = require('fs').readFileSync(path);

const wasmModule = new WebAssembly.Module(bytes);
const wasmInstance = new WebAssembly.Instance(wasmModule, imports);
wasm = wasmInstance.exports;
module.exports.__wasm = wasm;

wasm.__wbindgen_start();
