// h1z1-server/out/servers/ZoneServer2015/classes/zoneclient.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2023 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoneClient = void 0;
const character_1 = require("./character");
class ZoneClient {
    currentPOI;
    firstLoading = false;
    isLoading = true;
    isInteracting = false;
    isAdmin = false;
    posAtLastRoutine = new Float32Array();
    posAtLogoutStart = new Float32Array();
    hudTimer;
    spawnedDTOs = [];
    spawnedEntities = [];
    managedObjects = [];
    vehicle;
    npcsToSpawn = [];
    npcsToSpawnTimer;
    character;
    loginSessionId;
    pingTimer;
    savePositionTimer;
    clearHudTimer;
    clearTimers;
    sessionId;
    soeClientId;
    constructor(sessionId, soeClientId, loginSessionId, characterId, generatedTransient) {
        this.sessionId = sessionId;
        this.soeClientId = soeClientId;
        this.isLoading = true;
        this.firstLoading = true;
        this.loginSessionId = loginSessionId;
        this.vehicle = {
            vehicleState: 0,
            falling: -1,
            vehicleSeat: 0
        };
        this.character = new character_1.Character(characterId, generatedTransient);
        this.spawnedEntities = [];
        this.managedObjects = [];
        this.clearTimers = () => {
            clearTimeout(this.npcsToSpawnTimer);
        };
        this.clearHudTimer = () => {
            clearTimeout(this.hudTimer);
            this.hudTimer = null;
            this.isInteracting = false;
        };
    }
}
exports.ZoneClient = ZoneClient;
