// h1z1-server/out/utils/constants.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOADOUT_CONTAINER_ID = exports.EXTERNAL_CONTAINER_GUID = exports.LOADOUT_CONTAINER_GUID = exports.OBSERVER_GUID = exports.DB_NAME = exports.DEFAULT_CRYPTO_KEY = exports.MAX_SEQUENCE = exports.DATA_HEADER_SIZE = exports.MAX_UINT32 = exports.MAX_UINT16 = exports.MAX_UINT8 = exports.MAX_TRANSIENT_ID = void 0;
exports.MAX_TRANSIENT_ID = 16777215;
exports.MAX_UINT8 = 0xff;
exports.MAX_UINT16 = 0xffff;
exports.MAX_UINT32 = 0xffffffff;
exports.DATA_HEADER_SIZE = 4;
exports.MAX_SEQUENCE = exports.MAX_UINT16;
exports.DEFAULT_CRYPTO_KEY = "F70IaxuU8C/w7FPXY1ibXw==";
exports.DB_NAME = "h1server";
exports.OBSERVER_GUID = "0xFAFAFAFAFAFAFAFA";
exports.LOADOUT_CONTAINER_GUID = "0xffffffffffffffff";
exports.EXTERNAL_CONTAINER_GUID = "0x0000000000000001";
exports.LOADOUT_CONTAINER_ID = 101;
