// h1z1-server/out/utils/enums.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOGIN_KICK_REASON = exports.KILL_TYPE = exports.DB_COLLECTIONS = exports.CUSTOM_PROFILES_IDS = exports.NAME_VALIDATION_STATUS = exports.CONNECTION_REJECTION_FLAGS = exports.GAME_VERSIONS = void 0;
var GAME_VERSIONS;
(function (GAME_VERSIONS) {
    GAME_VERSIONS[GAME_VERSIONS["H1Z1_15janv_2015"] = 1] = "H1Z1_15janv_2015";
    GAME_VERSIONS[GAME_VERSIONS["H1Z1_6dec_2016"] = 2] = "H1Z1_6dec_2016";
    GAME_VERSIONS[GAME_VERSIONS["H1Z1_KOTK_PS3"] = 3] = "H1Z1_KOTK_PS3";
})(GAME_VERSIONS || (exports.GAME_VERSIONS = GAME_VERSIONS = {}));
var CONNECTION_REJECTION_FLAGS;
(function (CONNECTION_REJECTION_FLAGS) {
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["ERROR"] = 0] = "ERROR";
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["LOCAL_BAN"] = 1] = "LOCAL_BAN";
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["GLOBAL_BAN"] = 2] = "GLOBAL_BAN";
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["VPN"] = 3] = "VPN";
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["HWID"] = 4] = "HWID";
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["UNVERIFIED"] = 5] = "UNVERIFIED";
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["SERVER_LOCKED"] = 6] = "SERVER_LOCKED";
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["SERVER_REBOOT"] = 7] = "SERVER_REBOOT";
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["CHARACTER_NOT_FOUND"] = 8] = "CHARACTER_NOT_FOUND";
    CONNECTION_REJECTION_FLAGS[CONNECTION_REJECTION_FLAGS["OTHER"] = 9] = "OTHER";
})(CONNECTION_REJECTION_FLAGS || (exports.CONNECTION_REJECTION_FLAGS = CONNECTION_REJECTION_FLAGS = {}));
var NAME_VALIDATION_STATUS;
(function (NAME_VALIDATION_STATUS) {
    NAME_VALIDATION_STATUS[NAME_VALIDATION_STATUS["AVAILABLE"] = 1] = "AVAILABLE";
    NAME_VALIDATION_STATUS[NAME_VALIDATION_STATUS["TAKEN"] = 2] = "TAKEN";
    NAME_VALIDATION_STATUS[NAME_VALIDATION_STATUS["INVALID"] = 3] = "INVALID";
    NAME_VALIDATION_STATUS[NAME_VALIDATION_STATUS["PROFANE"] = 4] = "PROFANE";
    NAME_VALIDATION_STATUS[NAME_VALIDATION_STATUS["RESERVED"] = 5] = "RESERVED";
})(NAME_VALIDATION_STATUS || (exports.NAME_VALIDATION_STATUS = NAME_VALIDATION_STATUS = {}));
var CUSTOM_PROFILES_IDS;
(function (CUSTOM_PROFILES_IDS) {
    CUSTOM_PROFILES_IDS[CUSTOM_PROFILES_IDS["CONSTRUCTION"] = 999] = "CONSTRUCTION";
})(CUSTOM_PROFILES_IDS || (exports.CUSTOM_PROFILES_IDS = CUSTOM_PROFILES_IDS = {}));
var DB_COLLECTIONS;
(function (DB_COLLECTIONS) {
    DB_COLLECTIONS["CONSTRUCTION_LOGS"] = "construction-logs";
    DB_COLLECTIONS["ADMINS"] = "admins";
    DB_COLLECTIONS["BANNED"] = "banned";
    DB_COLLECTIONS["BLACK_LIST_ENTRIES"] = "blackListEntries";
    DB_COLLECTIONS["CHARACTERS"] = "characters";
    DB_COLLECTIONS["ACCOUNT_ITEMS"] = "account-items";
    DB_COLLECTIONS["CHARACTERS_LIGHT"] = "characters-light";
    DB_COLLECTIONS["CHAT"] = "chats";
    DB_COLLECTIONS["CONSTRUCTION"] = "construction";
    DB_COLLECTIONS["CROPS"] = "crops";
    DB_COLLECTIONS["TRAPS"] = "traps";
    DB_COLLECTIONS["FINGERPRINTS"] = "fingerprints";
    DB_COLLECTIONS["PROPS"] = "props";
    DB_COLLECTIONS["SERVERS"] = "servers";
    DB_COLLECTIONS["USERS_SESSIONS"] = "user-sessions";
    DB_COLLECTIONS["VEHICLES"] = "vehicles";
    DB_COLLECTIONS["WEATHERS"] = "weathers";
    DB_COLLECTIONS["WORLD_CONSTRUCTIONS"] = "worldconstruction";
    DB_COLLECTIONS["WORLDS"] = "worlds";
    DB_COLLECTIONS["COMMAND_USED"] = "commands-used";
    DB_COLLECTIONS["FAIRPLAY"] = "fairplay-logs";
    DB_COLLECTIONS["FAIRPLAY_TEMP"] = "fairplay-temp";
    DB_COLLECTIONS["KILLS"] = "kills";
    DB_COLLECTIONS["MUTED"] = "muted";
    DB_COLLECTIONS["GROUPS"] = "groups";
    DB_COLLECTIONS["VPN_WHITELIST"] = "vpn-whitelist";
    DB_COLLECTIONS["CHALLENGES"] = "challenges";
    DB_COLLECTIONS["AUTHKEYS"] = "verified-authkeys";
    DB_COLLECTIONS["LOGIN_ERRORS"] = "login-errors";
    DB_COLLECTIONS["NODEJS_REPORTS"] = "nodejs-reports";
})(DB_COLLECTIONS || (exports.DB_COLLECTIONS = DB_COLLECTIONS = {}));
var KILL_TYPE;
(function (KILL_TYPE) {
    KILL_TYPE["PLAYER"] = "player";
    KILL_TYPE["ZOMBIE"] = "zombie";
    KILL_TYPE["WILDLIFE"] = "wildlife";
    KILL_TYPE["VEHICLE"] = "vehicle";
})(KILL_TYPE || (exports.KILL_TYPE = KILL_TYPE = {}));
var LOGIN_KICK_REASON;
(function (LOGIN_KICK_REASON) {
    LOGIN_KICK_REASON[LOGIN_KICK_REASON["UNDEFINED"] = 0] = "UNDEFINED";
    LOGIN_KICK_REASON[LOGIN_KICK_REASON["GLOBAL_BAN"] = 1] = "GLOBAL_BAN";
})(LOGIN_KICK_REASON || (exports.LOGIN_KICK_REASON = LOGIN_KICK_REASON = {}));
