// h1z1-server/out/utils/recast.js
"use strict";
// ======================================================================
//
//   GNU GENERAL PUBLIC LICENSE
//   Version 3, 29 June 2007
//   copyright (C) 2020 - 2021 Quentin Gruber
//   copyright (C) 2021 - 2026 H1emu community
//
//   https://github.com/QuentinGruber/h1z1-server
//   https://www.npmjs.com/package/h1z1-server
//
//   Based on https://github.com/psemu/soe-network
// ======================================================================
Object.defineProperty(exports, "__esModule", { value: true });
exports.NavManager = void 0;
const node_fs_1 = require("node:fs");
const recast_navigation_1 = require("recast-navigation");
const recast_navigation_2 = require("recast-navigation");
const recast_navigation_3 = require("recast-navigation");
const recast_navigation_4 = require("recast-navigation");
class NavManager {
    navmesh;
    crowd;
    navMeshQuery;
    lastTimeCall = Date.now();
    constructor() { }
    async loadNav() {
        const navData = new Uint8Array((0, node_fs_1.readFileSync)(__dirname + "/../../data/2016/navData/z1.bin"));
        await (0, recast_navigation_1.init)();
        const { navMesh } = (0, recast_navigation_3.importNavMesh)(navData);
        this.navmesh = navMesh;
        const maxAgents = 100;
        const maxAgentRadius = 0.6;
        this.navMeshQuery = new recast_navigation_2.NavMeshQuery(this.navmesh);
        this.crowd = new recast_navigation_4.Crowd(navMesh, { maxAgents, maxAgentRadius });
    }
    static Float32ToVec3(f) {
        return { x: f[0], y: f[1], z: f[2] };
    }
    static Vec3ToFloat32(v) {
        return new Float32Array([v.x, v.y, v.z]);
    }
    updt() {
        this.crowd.update(1 / 60);
    }
    getClosestNavPoint(pos) {
        const n = this.navMeshQuery.findNearestPoly(NavManager.Float32ToVec3(pos));
        return n.nearestPoint;
    }
    createAgent(pos) {
        const position = this.getClosestNavPoint(pos);
        const radius = 0.5;
        const { randomPoint: initialAgentPosition, success, status } = this.navMeshQuery.findRandomPointAroundCircle(position, radius);
        if (!success) {
            console.log((0, recast_navigation_1.statusToReadableString)(status));
        }
        const agent = this.crowd.addAgent(initialAgentPosition, {
            radius: 0.5,
            height: 2,
            maxAcceleration: 4.0,
            maxSpeed: 1.0,
            collisionQueryRange: 0.1,
            pathOptimizationRange: 0.0,
            separationWeight: 0
        });
        return agent;
    }
    testNavMesh(a, b) {
        console.time("calculating path");
        const start = NavManager.Float32ToVec3(a);
        const end = NavManager.Float32ToVec3(b);
        const { success, error, path } = this.navMeshQuery.computePath(start, end);
        console.log(success);
        console.log(error);
        console.log(path);
        console.timeEnd("calculating path");
        const pathNodes = [];
        if (path) {
            path.forEach((v) => {
                pathNodes.push(new Float32Array([v.x, v.y, v.z]));
            });
        }
        return pathNodes;
    }
}
exports.NavManager = NavManager;
